#save.image(file="workspace.2019.08.13.RData")
#load(file="workspace.2019.08.13.RData")

### used from 27 October, 2021

source("C:\\Data/comparative.analysis.healthy.sweden/microbiome_atlas_comparative/check.mgs.functions.r")

if (F) {
  
  genusFile = "J://Deposit//Project//genus.table.txt"
  speciesFile = "J://Deposit//Project//genus.table.txt"
  
  speciesTab = read.delim(speciesFile, header=F, row.names = 1)
  genusTab = read.delim(genusFile, header=F, row.names = 1)
  genumsMat = as.matrix(genusTab[,-1])
  rownames(genumsMat) = rownames(genusTab)
  genumsMat = getRelAbd(genumsMat)
  heatmap.2(genumsMat, trace="none", margins = )
  
  
}


loadLibrary=T
if (loadLibrary) {
  
  require(dplyr)
  #require(h20)
  require(betapart)
  require(viridis)
  require(colourpicker)
  require(circlize)
  require(markovchain)
  require(diagram)
  require(ggtern)
  require(survminer)
  require(survival)
  require(qqman)  
  require(igraph)
  require(momr)
  require(monocle)
  require(ppcor)
  library(ggplot2)
  library(zoo)
  require(reshape2)
  require(data.table)
  require(matrixStats)
  require(gplots)
  require(lme4)
  require(ade4)
  require(pls)
  require(car)
  require(lmerTest)
  require(corrplot)
  require(vegan)
  require(ape)
  require(Rtsne)
  require(RColorBrewer)
  require(randomForest)
  require(ropls)
  require(BiocManager)
  require(plotrix)
  #require(pROC)
  require(ROCR)
  require(networkD3)
  require(tidyverse)
  require(RColorBrewer)
  require(ggrepel)
  require(prabclus)
  require(dbscan)  
  require(mclust)
  require(DirichletMultinomial)
  require(densityClust)
  
  #install("C:/Users/SunjaeLee/Downloads/rPython/")
  #install.packages("http://genome.crg.es/~didac/ggsunburst/ggsunburst_0.0.10.tar.gz", repos=NULL, type="source")
  
  setwd("C:\\Data/comparative.analysis.healthy.sweden/")
  
  gutMsp1992Data = "C:\\Data/comparative.analysis.healthy.sweden/hs_10.4_1992_MSP_freeze2_20180905.RData"
  mgsRData = "C:\\Data/comparative.analysis.healthy.sweden/mgs.all.10M.RData"
  mgsCleanRData = "C:\\Data/comparative.analysis.healthy.sweden/mgs.clean.10M.RData"
  mgsClean2RData = "C:\\Data/comparative.analysis.healthy.sweden/mgs.clean.madagascar.10M.RData"
  mgsCleanUpdatedRData = "C:\\Data/comparative.analysis.healthy.sweden/mgs.clean.updated.20190501.10M.RData"
  
  basicMetaCleanRData = "C:\\Data/comparative.analysis.healthy.sweden/all.basic.clean.metadata.20190109.RData"
  basicMetaCleanRData2 = "C:\\Data/comparative.analysis.healthy.sweden/all.basic.clean.metadata.20190114.RData"
  basicMetaCleanRData3 = "C:\\Data/comparative.analysis.healthy.sweden/all.basic.clean.metadata.updated.20190501.RData"
  #basicMetaCleanRDataUpdated = "C:\\Data/comparative.analysis.healthy.sweden/all.basic.clean.metadata.madagascar.20190125.RData"
  #commented at 2020-06-19
  basicMetaCleanRDataUpdated2 = "C:\\Data/comparative.analysis.healthy.sweden/all.basic.clean.metadata.behcet.20190805.RData"
  
  hqDataFile="C:/Data/comparative.analysis.healthy.sweden/high.qualtiy.metadata.tables.txt"
  hqDataRData="C:/Data/comparative.analysis.healthy.sweden/high.qualtiy.metadata.tables.RData"
  load(hqDataRData)
  #hqDataTab = read.table(hqDataFile, header=T, sep="\t", stringsAsFactors = F)
  datasetOverviewFile = "C:/Data/comparative.analysis.healthy.sweden/atlas.dataset.overview.txt"
  datasetOverviewTab = read.delim(datasetOverviewFile, header=T, sep="\t", stringsAsFactors = F)
  keggPathDescTabRData = "C:/Data/comparative.analysis.healthy.sweden/keggPathDescTab.RData"
  #moduleDescTabRData = "C:/Data/comparative.analysis.healthy.sweden/moduleDescTab.RData"
  moduleDescTabRData = "J://Deposit/Project/2018_microbiome_atlas/functional.annotation/KEGG module/moduleDescTab.RData" 
  
  phageomeData = "C://Data/comparative.analysis.healthy.sweden/mergePhageMatFinal.RData"
  
  load(hqDataRData)
  load(phageomeData)
  
  load(keggPathDescTabRData)
  load(moduleDescTabRData)
  moduleDescTab$FullName = paste(moduleDescTab$moduleId, moduleDescTab$moduleName, sep = ":")
  
  exModules = moduleDescTab$FullName[moduleDescTab$system =="Module set"]
  moduleDescTab = moduleDescTab[moduleDescTab$system !="Module set",]
  
  #keggPathDescTab
  #load("colon.mgs.med.vec.fecies.RData")
  
  # pp = getTaxaAvgMat(mgs_med_vec,taxo,"phylum")
  # pp2 = rowMedians(pp)
  # names(pp2)=rownames(pp)
}

loadFunc = T
if (loadFunc) {
  
  #### general functions ####
  getAbSubj <- function(str) {
    subj = strsplit(str, split = "_")[[1]][1]
    return(subj)
  }
  getAbSubjs <- function(strVec) {
    return(sapply(strVec, getAbSubj))
  }
  
  getVisitIds <- function(str) {
    visitId = strsplit(str, split = "_")[[1]][1]
    return(visitId)
  }
  
  getFields <- function(str) {
    str = gsub("visit1_","",str)
    str = gsub("visit2_","",str)
    str = gsub("visit3_","",str)
    str = gsub("visit4_","",str)
    return(str)
  }
  
  getRelAbd <- function(targetMat) {
    relAbds = sweep(targetMat, MARGIN=2, colSums(targetMat), FUN="/")
    return(relAbds)
  }
  jacc <- function(x,y) {
    prod = sum(x*y)
    sumx = sum(x)
    sumy = sum(y)
    uni  = sumx + sumy - prod
    return(prod/uni)
  }
  
  checkOverlap <- function(xList, yList) {
    xLen = length(xList)
    yLen = length(yList)
    xy = sum((xList %in% yList))
    print(c(xLen, yLen, xy, xy/(xLen+yLen-xy)))
  }
  # contourMat = pcoaMgsMat
  # target1 = newCaseByEnteroList$ETF
  # target2 = newContByEnteroList$ETF
  # 
  getAnosimBoth <- function(mgsMat, target1, target2) {
    targetAll = unique(c(target1, target2))
    vegOut=vegdist(t(mgsMat[,targetAll]),"bray")
    targetClasses = rep("class1",length(targetAll))
    targetClasses[targetAll %in% target2] = "class2"
    set.seed(1)
    anosimOut = anosim(vegOut, targetClasses)
    print(anosimOut)
    return(anosimOut)
  }
  getPermanovaBoth <- function(mgsMat, target1, target2) {
    targetAll = unique(c(target1, target2))
    vegOut=vegdist(t(mgsMat[,targetAll]),"bray")
    targetClasses = rep("class1",length(targetAll))
    targetClasses[targetAll %in% target2] = "class2"
    set.seed(1)
    permanovaOut = adonis2(vegOut ~ targetClasses, permutations = 999, method="bray")
    print(permanovaOut)
    return(permanovaOut)
  }
  
  
  plotContourBoth <- function(contourMat, target1, target2) {
    targetAll = unique(c(target1, target2))
    k <- 9
    my.cols.1 <- (brewer.pal(k, "PuRd"))
    my.cols.2 <- (brewer.pal(k, "Blues"))
    
    
    targetInd1 = rownames(contourMat)%in% target1
    targetInd2 = rownames(contourMat)%in% target2
    
    targetMat1 = contourMat[targetInd1,]
    targetMat2 = contourMat[targetInd2,]
    targetMatAll = rbind(targetMat1, targetMat2)
    
    print(summary(out))
    z.1 <- kde2d(targetMat1[,1], targetMat1[,2], n=50)
    z.2 <- kde2d(targetMat2[,1], targetMat2[,2], n=50)
    
    plot(contourMat[targetAll,], xlab="PCoA1", ylab="PCoA2", col="#80808033", pch=19, cex=.4, xlim=c(-1,1),ylim=c(-1,1))
    contour(z.1, drawlabels=FALSE, nlevels=k, col=my.cols.1, add=TRUE)
    contour(z.2, drawlabels=FALSE, nlevels=k, col=my.cols.2, add=TRUE)
    
    abline(h=0, v=0, col="black")
    return(list(z.1, z.2))
    
  }
  
  plotContour <- function(contourMat) {
    k <- 11
    my.cols <- rev(brewer.pal(k, "RdYlBu"))
    
    z <- kde2d(contourMat[,1], contourMat[,2], n=50)
    
    plot(contourMat, xlab="X label", ylab="Y label", col="#80808033", pch=19, cex=.4, xlim=c(-1,1),ylim=c(-1,1))
    contour(z, drawlabels=FALSE, nlevels=k, col=my.cols, add=TRUE)
    
    abline(h=0, v=0, col="black")
    return(z)
    
  }
  
  getNumbersFromWellnessVisits <- function(visits) {
    visits = gsub("v","",visits)
    visits = as.numeric(visits)
    return(visits)
  }
  
  
  getChiPval <- function(chisq, df) {
    return(1 - pchisq(chisq,df))
  }
  
  writeJaccNetToCytoscape <- function(samples, jaccMat, metaTab, nodeFile, edgeFile, jaccCut=0.5) {
    col_vector = unlist(mapply(brewer.pal, qual_col_pals$maxcolors, rownames(qual_col_pals)))
    col_vector = c(col_vector,col_vector)
    
    currJaccMat = jaccMat[samples, samples]
    currJaccTab = melt(currJaccMat, variable.factor=F)
    currJaccTab$Var1 = as.character(currJaccTab$Var1)
    currJaccTab$Var2 = as.character(currJaccTab$Var2)
    currJaccTab = currJaccTab[currJaccTab$value >= jaccCut,]
    currJaccTab = currJaccTab[currJaccTab$Var1 != currJaccTab$Var2,]
    
    nodes = unique(c(currJaccTab$Var1, currJaccTab$Var2))
    subjects= metaTab$metadata.ID[match(samples,metaTab$sample.ID)]
    subtypes= metaTab$subtype[match(samples,metaTab$sample.ID)]
    types= metaTab$type[match(samples,metaTab$sample.ID)]
    
    nodesNotShown = samples[!samples %in% nodes]
    nodeTable = data.frame(node=samples,
                           subject=subjects,
                           type = types,
                           subtype = subtypes)
    edgeTable = currJaccTab 
    
    outList = list(nodeTable = nodeTable, 
                   edgeTable = edgeTable, 
                   nodesNotShown = nodesNotShown )
    
    writeModuleNetworkAnnot(outList, nodeFile, edgeFile)
    return(outList)
  }
  
  copyLowerToUpper <- function(m) {
    m[lower.tri(m)] <- t(m)[lower.tri(m)]
    return(m)
  }
  getJaccMat <- function(mat) {
    
    jaccMat = matrix(0,dim(mat)[2],dim(mat)[2])
    rownames(jaccMat) = colnames(mat)
    colnames(jaccMat) = colnames(mat)
    
    combs = t(combn(ncol(mat),2))
    jaccs = apply(combs, 1, function(x) jacc(mat[,x[1]], mat[,x[2]]))
    jaccMat[combs]=jaccs
    diag(jaccMat)=1
    jaccMat = copyLowerToUpper(jaccMat)
    return(jaccMat)
  }
  
  getJaccFrom <- function(strs1, strs2) {
    len1 = length(strs1)
    len2 = length(strs2)
    if (len1 == 0 | len2 ==0 ) return(NA)
    lenO = sum(strs1 %in% strs2)
    lenU = len1 + len2 - lenO
    return(lenO/lenU)
  }
  
  #### disease network ####
  makeSpeciesDiseaseNet <- function(speciesTab) {
    diseaseNet = igraph::graph.data.frame(speciesTab[,1:2], directed=F)
    diseaseNet = set_edge_attr(diseaseNet, "weight", value= speciesTab[,3])
    diseaseNet = igraph::simplify(diseaseNet, remove.multiple=T, remove.loops=T)
    return(diseaseNet)
  }
  
  
  makeJaccNet <- function(jaccMat, cutJacc = 0.5) {
    jaccElementsAll = rownames(jaccMat)
    jaccTable = melt(jaccMat)
    jaccTable = jaccTable[jaccTable[,3] >= cutJacc & jaccTable[,1] != jaccTable[,2], ]
    jaccTable[,1]=as.character(jaccTable[,1])
    jaccTable[,2]=as.character(jaccTable[,2])
    
    jaccElementsLeft = unique(c(jaccTable[,1], jaccTable[,2]))
    jaccElementsExcluded = jaccElementsAll[!jaccElementsAll %in% jaccElementsLeft]
    
    jaccNet = igraph::graph.data.frame(jaccTable[,1:2], directed=F)
    jaccNet = igraph::simplify(jaccNet, remove.multiple=T, remove.loops=T)
    
    jaccNet = add_vertices(jaccNet, length(jaccElementsExcluded), attr=list(name=jaccElementsExcluded))
    
    return(jaccNet)
  }
  
  
  makeModuleList <- function(jaccNet, weightMode = F, debug=F) {
    fc = cluster_walktrap(jaccNet)
    if (weightMode) fc = cluster_walktrap(jaccNet, weights = E(jaccNet)$weight)
    
    #fc = cluster_edge_betweenness(jaccNet)
    #fc = cluster_louvain(jaccNet)
    
    cluster = fc$membership
    geneCluster = data.frame(gene=V(jaccNet)$name, 
                             cluster=cluster,
                             stringsAsFactors=F)
    geneClusterList = split.data.frame(geneCluster, geneCluster$cluster)
    geneClusterList = lapply(geneClusterList, "[[", "gene")
    geneClusterSizes = do.call(c, lapply(geneClusterList, length))
    
    if (debug) {
      print("... done")
      print(paste("... modularity:", as.character(modularity(fc))))
      print(paste("... no. clusters:", as.character(length(geneClusterList))))
      print(paste("... no. genes of max cluster:", as.character(sort(geneClusterSizes,T)[1])))
      
    }
    return(geneClusterList)
  }
  
  
  getClusterScore  <- function(jaccNet, currClusterGenes) {
    netGenes=V(jaccNet)$name
    currInd = match(currClusterGenes, netGenes)
    currGraph = subgraph(jaccNet, currInd)
    cc = transitivity(currGraph, "global")
    return(cc)
  }
  getClusterScores <- function(jaccNet, geneClusterList) {
    ccScores = lapply(geneClusterList, function(currClusterGenes){
      #currGraph = igraph::subgraph(corNet, currClusterGenes)
      currGraph = subgraph(jaccNet, currClusterGenes)
      cc = transitivity(currGraph, "global")
      if (is.nan(cc)) cc = 0
      return(cc)
    })
    ccScores = unlist(ccScores)
    names(ccScores) = names(geneClusterList)
    return(ccScores)
  }
  
  getModuleMatrix <- function(jaccNet, modulePruned, debug=F) {
    numEdges = ecount(jaccNet)
    adjMat = as_adj(jaccNet)
    degMat = getDegMat(jaccNet)
    
    numModules = length(modulePruned)
    if (debug) print(numModules)
    moduleMat = matrix(0, numModules, numModules)
    for (i in 1:(numModules-1)) {
      if (debug) print(i)
      for (j in (i+1):numModules) {
        moduleMat[i,j] = areModuleInteract(adjMat, degMat, numEdges, modulePruned[[i]], modulePruned[[j]])
      }
    }  
    moduleMat[lower.tri(moduleMat)] = t(moduleMat)[lower.tri(moduleMat)]
    rownames(moduleMat) = names(modulePruned)
    colnames(moduleMat) = names(modulePruned)
    return(moduleMat)
  }
  
  prunListByNetworkGenesAndCut <- function(modList, networkGenes, clusterCut) {
    outList = lapply(modList, function(modGenes){
      modGenes = modGenes[modGenes %in% networkGenes]
    })
    outListLens = lapply(outList, length)
    outList = outList[outListLens>=clusterCut]
    return(outList)
    
  }
  areModuleInteract <- function(adjMat, degMat, numEdges, genesA, genesB) {
    observedEdges = observedEdgesBtw(adjMat, genesA, genesB)
    expectedEdges = expectedEdgesBtw(degMat, numEdges, genesA, genesB)
    diff=(observedEdges-expectedEdges)/expectedEdges
    return(diff)
  }
  getDegMat <- function(igraphNet) {
    #we regard network as undirected
    #degs = degree(igraphNet, mode="all")
    degs = igraph::degree(igraphNet, mode="total")
    names(degs) = V(igraphNet)$name
    degMat = as.matrix(degs) %*% t(as.matrix(degs))
    return(degMat)
  }
  expectedEdgesBtw <- function(degMat, numEdges, genesA, genesB) {
    degMatSel = degMat[genesA, genesB]/(2*numEdges)
    count = sum(degMatSel)
    return(count)
  }
  observedEdgesBtw <- function(adjMat, genesA, genesB) {
    adjMatSel = adjMat[genesA, genesB]
    count = sum(adjMatSel)
    return(count)
  }
  getModulePruned <- function(geneClusterList, cutCluster, debug=F) {
    geneClusterSizes = unlist(lapply(geneClusterList, length))
    names(geneClusterSizes) = names(geneClusterSizes)
    if (debug) print(table(geneClusterSizes>=cutCluster))
    geneClusterCut = names(geneClusterSizes[geneClusterSizes>=cutCluster])
    geneClusterList = geneClusterList[geneClusterCut]
    return(geneClusterList) 
  }
  getInteractionsFromModules <- function( coexpTable, moduleList, targets, debug=T ) {
    moduleNames = names(moduleList)
    if (!all(targets %in% moduleNames)) {
      print("... some targets were not shown in modulePruned")
      print("... please check it again")
      return(NULL)
    }
    
    targetModules = moduleList[targets]
    targetModuleGenes = unique.default(do.call(c, targetModules))
    targetModuleMaps = rep(NA, length(targetModuleGenes))
    for (i in 1:length(targets)) {
      #currTarget = targets[1]
      currTarget = targets[i]
      currTargetGenes = unlist(moduleList[[currTarget]])
      targetModuleMaps[targetModuleGenes %in% currTargetGenes] = currTarget
    }
    
    targetInd = coexpTable[,1]%in%targetModuleGenes & coexpTable[,2]%in%targetModuleGenes
    targetEdgeTable = coexpTable[targetInd, 1:3]
    colnames(targetEdgeTable) = c("Source", "Target","Corr")
    
    targetNodeTable = data.frame(node=targetModuleGenes,
                                 symbol=getSymbols(targetModuleGenes),
                                 moduleType=targetModuleMaps,
                                 stringsAsFactors=F)
    
    
    return(list(nodeTable = targetNodeTable, edgeTable = targetEdgeTable))
  }
  
  annotateModulesByCC <-function(jaccNet, moduleList,  cutCluster=5, cutCC = 0.5, debug=F) {
    
    modulePrunedList = getModulePruned(moduleList, cutCluster)
    moduleNames = names(modulePrunedList)
    
    # module matrix
    moduleMat = getModuleMatrix(jaccNet, modulePrunedList)
    print(moduleMat)
    # first axis
    moduleCC = getClusterScores(jaccNet, modulePrunedList)
    highCCModuleNames = names(moduleCC[moduleCC >= cutCC])
    
    
    # summary of two axes
    summaryTypes = rep("None", length(modulePrunedList))
    names(summaryTypes) = moduleNames 
    summaryTypes[moduleNames %in% highCCModuleNames ] = "HighCC"
    summaryTypes[!moduleNames %in% highCCModuleNames ] = "LowCC"
    
    moduleCut=1
    moduleMelt = melt(moduleMat)
    moduleMelt = moduleMelt[moduleMelt[,3]>moduleCut,]
    moduleMelt[] = lapply(moduleMelt, as.character)
    moduleMelt = moduleMelt[!is.na(moduleMelt$value),]
    modulesShown = unique.default(c(moduleMelt[,1], moduleMelt[,2]))
    moduleSizes = unlist(lapply(modulePrunedList, length))
    moduleAttr = data.frame(node=moduleNames, size=moduleSizes, summary = summaryTypes, cc = moduleCC, stringsAsFactors = F)
    
    modulesNotShown = moduleNames[!moduleNames %in% modulesShown]
    return(list(nodeTable = moduleAttr, edgeTable=moduleMelt, nodesNotShown = modulesNotShown ))
  }
  
  getEnrichMentGsByList <- function(refGeneSet, targetGenes, debug=F) {
    allRefGenes = unique.default(do.call(c, refGeneSet))
    if (debug) print(table(targetGenes %in% allRefGenes))
    targetGenes = targetGenes[targetGenes %in% allRefGenes]
    
    hyperPs = lapply(refGeneSet, function(currRefGenes) {
      hyperP = getHyperP( targetGenes, currRefGenes, allRefGenes, F ) 
      return(hyperP)
    })
    hyperPs = unlist(hyperPs, use.names = F)
    names(hyperPs) = names(refGeneSet)
    return(hyperPs)
  }
  
  write.edge.cytoscape <- function(edgeTable, nodesNotShown, outFile) {
    lenTable=dim(edgeTable)[1]
    lenNodesNotShown = length
    print(head(edgeTable))
    sink(outFile, append=F)
    cat("source")
    cat("\t")
    cat("target")
    cat("\n")
    for (i in 1:lenTable) {
      # ints = unlist(edgeTable[i,])
      cat(edgeTable[i,1])
      cat("\t")
      cat(edgeTable[i,2])
      cat("\n")
    }
    for (node in nodesNotShown) {
      cat(node)
      cat("\n")
    }
    sink() 
  }
  write.node.cytoscape <- function(nodeTable, outFile) {
    write.table(nodeTable, outFile, row.names = F, quote = F, sep="\t")
  }
  writeModuleNetworkAnnot <- function(moduleOut, nodeFile, edgeFile) {
    nodeTable = moduleOut[["nodeTable"]]
    edgeTable = moduleOut[["edgeTable"]]
    nodesNotShown = moduleOut[["nodesNotShown"]]
    write.node.cytoscape(nodeTable, nodeFile)
    write.edge.cytoscape(edgeTable, nodesNotShown, edgeFile)
  }
  
  
  
  #### Enterotyping ####
  
  getSubectIds <- function(strs) {
    newStrs = gsub("_v[1-4]","",strs)
    return(newStrs)
  }
  getEntCount <- function(enterotypes) {
    etf <- sum(enterotypes == "ET-Firmicutes")
    etb <- sum(enterotypes == "ET-Bacteroides")
    etp <- sum(enterotypes == "ET-Prevotella")
    return(c(etf, etb, etp))
  }
  getEntCountFrom <- function(basicMetaMap, cut0, cut1, by="GeneRichness") {
    entCount = getEntCount(basicMetaMap$enteroType[basicMetaMap[,by] <= cut1 & basicMetaMap[,by] > cut0])
    return(entCount)
  }
  getSurvivalRatesFrom <- function(survivalStats, cut0, cut1, by="mgs.abd") {
    survRates = survivalStats$surv.rates[survivalStats[,by] <= cut1 & survivalStats[,by] > cut0]
    return(survRates)
  }
  
  getSelectedStatFrom <- function(dataTab, cut0, cut1, by, target) {
    selected = dataTab[dataTab[,by] <= cut1 & dataTab[,by] > cut0, target]
    return(selected)
  }
  
  
  getMarkovStatFrom <- function(markovStatTab, cut0, cut1, by="meanAbd", target="inflow") {
    targetRates = markovStatTab[markovStatTab[,by] <= cut1 & markovStatTab[,by] > cut0, target]
    return(targetRates)
  }
  getBinTab <- function(basicMetaMap, by="GeneRichness") {
    richnessMin = min(basicMetaMap[,by])
    richness01 = quantile(basicMetaMap[,by], 0.1)
    richness02 = quantile(basicMetaMap[,by], 0.2)
    richness03 = quantile(basicMetaMap[,by], 0.3)
    richness04 = quantile(basicMetaMap[,by], 0.4)
    richness05 = quantile(basicMetaMap[,by], 0.5)
    richness06 = quantile(basicMetaMap[,by], 0.6)
    richness07 = quantile(basicMetaMap[,by], 0.7)
    richness08 = quantile(basicMetaMap[,by], 0.8)
    richness09 = quantile(basicMetaMap[,by], 0.9)
    richnessMax = max(basicMetaMap[,by])
    
    entTab00 = getEntCountFrom(basicMetaMap, richnessMin-1, richness01, by)
    entTab01 = getEntCountFrom(basicMetaMap, richness01, richness02, by)
    entTab02 = getEntCountFrom(basicMetaMap, richness02, richness03, by)
    entTab03 = getEntCountFrom(basicMetaMap, richness03, richness04, by)
    entTab04 = getEntCountFrom(basicMetaMap, richness04, richness05, by)
    entTab05 = getEntCountFrom(basicMetaMap, richness05, richness06, by)
    entTab06 = getEntCountFrom(basicMetaMap, richness06, richness07, by)
    entTab07 = getEntCountFrom(basicMetaMap, richness07, richness08, by)
    entTab08 = getEntCountFrom(basicMetaMap, richness08, richness09, by)
    entTab09 = getEntCountFrom(basicMetaMap, richness09, richnessMax+1, by)
    
    entTabByRichnes = rbind(entTab00,
                            entTab01,
                            entTab02,
                            entTab03,
                            entTab04,
                            entTab05,
                            entTab06,
                            entTab07,
                            entTab08,
                            entTab09)
    entTabByRichnes = sweep(entTabByRichnes, MARGIN=1, rowSums(entTabByRichnes)/100, FUN="/")
    return(entTabByRichnes)
  }
  getBinTabForSurvival <- function(survivalStats, by="mgs.abd") {
    abdMin = min(survivalStats[,by])
    abd02 = quantile(survivalStats[,by], 0.2)
    abd04 = quantile(survivalStats[,by], 0.4)
    abd06 = quantile(survivalStats[,by], 0.6)
    abd08 = quantile(survivalStats[,by], 0.8)
    abdMax = max(survivalStats[,by])
    
    survTab00 = getSurvivalRatesFrom(survivalStats, abdMin-1, abd02, by)
    survTab02 = getSurvivalRatesFrom(survivalStats, abd02, abd04, by)
    survTab04 = getSurvivalRatesFrom(survivalStats, abd04, abd06, by)
    survTab06 = getSurvivalRatesFrom(survivalStats, abd06, abd08, by)
    survTab08 = getSurvivalRatesFrom(survivalStats, abd08, abdMax+1, by)
    
    survListByAbd = list(per20=survTab00,
                         per40=survTab02,
                         per60=survTab04,
                         per80=survTab06,
                         per100=survTab08)
    
    return(survListByAbd)
  }
  getBinTabForSurvival2 <- function(survivalStats, by="mgs.abd") {
    abdMin = min(survivalStats[,by])
    abd02 = quantile(survivalStats[,by], 0.25)
    abd04 = quantile(survivalStats[,by], 0.5)
    abd06 = quantile(survivalStats[,by], 0.75)
    abdMax = max(survivalStats[,by])
    
    print(abdMin)
    print(abd02)
    print(abd04)
    print(abd06)
    print(abdMax)
    
    survTab00 = getSurvivalRatesFrom(survivalStats, abdMin-1, abd02, by)
    survTab02 = getSurvivalRatesFrom(survivalStats, abd02, abd04, by)
    survTab04 = getSurvivalRatesFrom(survivalStats, abd04, abd06, by)
    survTab06 = getSurvivalRatesFrom(survivalStats, abd06, abdMax+1, by)
    
    survListByAbd = list(per25=survTab00,
                         per50=survTab02,
                         per75=survTab04,
                         per100=survTab06)
    
    return(survListByAbd)
  }
  getRichnessBinByOutflowMean <- function(dataTab, by="outflow", target="richness") {
    qMin = min(outflowTab[,by])
    q01 = quantile(outflowTab[,by], 0.1)
    q02 = quantile(outflowTab[,by], 0.2)
    q03 = quantile(outflowTab[,by], 0.3)
    q04 = quantile(outflowTab[,by], 0.4)
    q05 = quantile(outflowTab[,by], 0.5)
    q06 = quantile(outflowTab[,by], 0.6)
    q07 = quantile(outflowTab[,by], 0.7)
    q08 = quantile(outflowTab[,by], 0.8)
    q09 = quantile(outflowTab[,by], 0.9)
    qMax = max(outflowTab[,by])
    
    selected00 = getSelectedStatFrom(outflowTab, qMin-1, q01, by, target)
    selected01 = getSelectedStatFrom(outflowTab, q01, q02, by, target)
    selected02 = getSelectedStatFrom(outflowTab, q02, q03, by, target)
    selected03 = getSelectedStatFrom(outflowTab, q03, q04, by, target)
    selected04 = getSelectedStatFrom(outflowTab, q04, q05, by, target)
    selected05 = getSelectedStatFrom(outflowTab, q05, q06, by, target)
    selected06 = getSelectedStatFrom(outflowTab, q06, q07, by, target)
    selected07 = getSelectedStatFrom(outflowTab, q07, q08, by, target)
    selected08 = getSelectedStatFrom(outflowTab, q08, q09, by, target)
    selected09 = getSelectedStatFrom(outflowTab, q09, qMax+1, by, target)
    
    selctedBins = list(per10=selected00,
                       per20=selected01,
                       per30=selected02,
                       per40=selected03,
                       per50=selected04,
                       per60=selected05,
                       per70=selected06,
                       per80=selected07,
                       per90=selected08,
                       per100=selected09)
    
    return(selctedBins)
  }
  
  makeKeggQuery <- function(pathList, mspKeggList, pathId, mspId, viewMode = F) {
    koList = mspKeggList[[mspId]]
    prefix1 = "https://www.kegg.jp/kegg-bin/show_pathway?map="
    prefix2 = "&multi_query=7165+red,blue"
    suffices = unlist(sapply(koList, function(ko) {
      suffix1 = "%0d%0a"
      suffix2 = "+red"
      concats = paste(suffix1, ko, suffix2, sep="")
      return(concats)
    }))
    suffices = paste(suffices, collapse = "")
    finalUrl = paste(prefix1, pathId, prefix2, suffices, sep = "")
    if (viewMode) browseURL(finalUrl)
    return(finalUrl)  
  }
  
  makeCurrTab <- function(currRow) {
    currTab = data.frame(case=currRow[1:3], cont=currRow[4:6])
    rownames(currTab) = c("ET-Firmicutes", "ET-Bacteroides", "ET-Prevotella")
    return(currTab)
  }
  grepEnteroTypesFrom <- function(sampleList, basicMetaMap) {
    newEnteroList = lapply(sampleList, function(currSamples){
      enterotypes = basicMetaMap$enteroType[basicMetaMap$sample.ID %in% currSamples]
      return(enterotypes)
    })
    return(newEnteroList)
  }
  
  grepGeneRichnesssFrom <- function(sampleList, basicMetaMap) {
    newRichnessList = lapply(sampleList, function(currSamples){
      richness = basicMetaMap$GeneRichness[basicMetaMap$sample.ID %in% currSamples]
      return(richness)
    })
    return(newRichnessList)
  }
  
  grepGeneRichnesssFromNew <- function(geoList, basicMetaMap) {
    newRichnessList = lapply(geoList, function(cohortList){
      return(lapply(cohortList, function(currSamples)  return(basicMetaMap$GeneRichness[basicMetaMap$sample.ID %in% currSamples])))
    })
    return(newRichnessList)
  }
  
  getEntTab <- function(enteroList) {
    entGeoTab = do.call(rbind,lapply(enteroList,  function(enterotypes){
      entCounts = getEntCount(enterotypes)
      return(entCounts)
    }))
    colnames(entGeoTab) = c("ET-Firmicutes","ET-Bacteroides","ET-Prevotella")
    return(entGeoTab)
  }
  
  getEntGeoPlot <- function(enteroList, ordered=T, margins=c(8,4,4,1)) {
    entGeoTab = do.call(rbind,lapply(enteroList,  function(enterotypes){
      entCounts = getEntCount(enterotypes)
      return(entCounts)
    }))
    colnames(entGeoTab) = c("ET-Firmicutes","ET-Bacteroides","ET-Prevotella")
    
    entGeoTab = sweep(entGeoTab, MARGIN=1, rowSums(entGeoTab)/100, FUN="/")
    if (ordered) entGeoTab = entGeoTab[order(entGeoTab[,2]),]
    if (ordered) entGeoTab = entGeoTab[order(entGeoTab[,1]),]
    par(mar=margins)
    
    etfCol = "#77e2c799"
    etbCol = "#bee24599"
    etpCol = "#f35b38"
    
    tradVec = c(255,127,0,120)/255
    #tradVec = c(255,127,0,200)/255
    tradCol = rgb(tradVec[1],tradVec[2],tradVec[3],tradVec[4])
    
    cjpVec = c(211,211,211,120)/255
    #cjpVec = c(211,211,211,200)/255
    cjpCol = rgb(cjpVec[1],cjpVec[2],cjpVec[3],cjpVec[4])
    
    eurVec = c(117,112,179,120)/255
    #eurVec = c(117,112,179,200)/255
    eurCol = rgb(eurVec[1],eurVec[2],eurVec[3],eurVec[4])
    
    etpCol = tradCol
    etbCol = cjpCol
    etfCol = eurCol
    
    barplot(t(entGeoTab),#[c(2,4,3,5,6,7,8,9),]
            col=c(etfCol,etbCol,etpCol), las=2)
    return(entGeoTab)
  }
  
  #### meta-analysis ####
  checkMinEffectSizeOfFeatures <- function(datasetList, abdMat) {
    
    datasetSeqs = seq_along(datasetList)
    features = rownames(abdMat)
    
    minEffectSizeList = list()
    for (i in datasetSeqs) {
      samplesI = datasetList[[i]]
      minEffectSize = rep(10, length(features))
      names(minEffectSize) = features
      
      for (j in datasetSeqs) {
        samplesJ = datasetList[[j]]
        if (i!=j) {
          print(c(i,j))
          currCaseMat = abdMat[, colnames(abdMat) %in% samplesI]
          currContMat = abdMat[, colnames(abdMat) %in% samplesJ]
          
          currEsVec <- getEffectSizeMat(currCaseMat, currContMat)
          minEffectSize <- rowMins(cbind(minEffectSize, currEsVec))
          
        }
      }
      minEffectSizeList[[i]] = minEffectSize
      
    }
    minEffectSizeMat = do.call(cbind, minEffectSizeList)
    colnames(minEffectSizeMat) = names(datasetList)
    rownames(minEffectSizeMat) = rownames(abdMat)
    return(minEffectSizeMat)
    
  }
  checkMaxEffectSizeOfFeatures <- function(datasetList, abdMat) {
    
    datasetSeqs = seq_along(datasetList)
    features = rownames(abdMat)
    
    maxEffectSizeList = list()
    for (i in datasetSeqs) {
      samplesI = datasetList[[i]]
      maxEffectSize = rep(0, length(features))
      names(maxEffectSize) = features
      
      for (j in datasetSeqs) {
        samplesJ = datasetList[[j]]
        if (i!=j) {
          print(c(i,j))
          currCaseMat = abdMat[, colnames(abdMat) %in% samplesI]
          currContMat = abdMat[, colnames(abdMat) %in% samplesJ]
          
          currEsVec <- getEffectSizeMat(currCaseMat, currContMat)
          maxEffectSize <- rowMaxs(cbind(maxEffectSize, currEsVec))
          
        }
      }
      maxEffectSizeList[[i]] = maxEffectSize
      
    }
    maxEffectSizeMat = do.call(cbind, maxEffectSizeList)
    colnames(maxEffectSizeMat) = names(datasetList)
    rownames(maxEffectSizeMat) = rownames(abdMat)
    return(maxEffectSizeMat)
    
  }
  checkBestFeaturesInES <- function(datasetList, abdMat, esCut=0.8) {
    
    datasetSeqs = seq_along(datasetList)
    features = rownames(abdMat)
    
    featureList = list()
    for (i in datasetSeqs) {
      samplesI = datasetList[[i]]
      featureCounts = rep(0, length(features))
      names(featureCounts) = features
      
      for (j in datasetSeqs) {
        samplesJ = datasetList[[j]]
        if (i!=j) {
          print(c(i,j))
          currCaseMat = abdMat[, colnames(abdMat) %in% samplesI]
          currContMat = abdMat[, colnames(abdMat) %in% samplesJ]
          
          currEsVec <- getEffectSizeMat(currCaseMat, currContMat)
          currEsVec <- currEsVec >= esCut
          featureCounts <- featureCounts + currEsVec
          
        }
      }
      featureList[[i]] = featureCounts
      
      
    }
    featureMat = do.call(cbind, featureList)
    colnames(featureMat) = names(datasetList)
    return(featureMat)
    
  }
  getAvgAbdMatBy <- function(abdMat, datasetList) {
    
    avgMat = do.call(cbind, lapply(datasetList, function(dsSamples) {
      currMat = abdMat[, colnames(abdMat) %in% dsSamples]
      return(rowMeans(currMat))
    }))
    
    rownames(avgMat) = rownames(abdMat)
    colnames(avgMat) = names(datasetList)
    
    return(avgMat)
  }
  
  #### volcano plots ####
  getVolcanoStat <- function(targetMat, hgcSamples, lgcSamples, taxo, speciesMode=F) {
    targetMean = rowMeans(targetMat)
    targetPart  = targetMean/sum(targetMean)
    
    
    hgcMat = targetMat[,hgcSamples]
    lgcMat = targetMat[,lgcSamples]
    
    relAbdMat = getRelAbd(targetMat)
    relAbdHgcMat = relAbdMat[,hgcSamples]
    relAbdLgcMat = relAbdMat[,lgcSamples]
    
    allMean = rowMeans(relAbdMat)
    hgcMean = rowMeans(relAbdHgcMat)
    lgcMean = rowMeans(relAbdLgcMat)
    
    lenSpecies = dim(hgcMat)[1]
    pValues = sapply(1:lenSpecies, function(ind) {
      mergedVec = c(hgcMat[ind,],lgcMat[ind,])
      mergedFac = c(rep("HGC",length(hgcMat[ind,])),
                    rep("LGC",length(lgcMat[ind,])))
      
      wilcoxOut = wilcox.test(mergedVec ~ mergedFac,)
      return(wilcoxOut$p.value)
    })
    pValues[is.nan(pValues)]=1
    adjPs = p.adjust(pValues,method="BH")
    logAdjPs = -log10(adjPs)
    logAdjPs[logAdjPs>300] = 300
    
    lfcs = sapply(1:lenSpecies,function(ind){
      hmeans = mean(hgcMat[ind,])
      lmeans = mean(lgcMat[ind,]) 
      return(hmeans/lmeans)
    })
    names(lfcs)=rownames(hgcMat)
    lfcs[is.nan(lfcs)]=1
    lfcs[lfcs==0]=2^-15
    lfcs[is.infinite(lfcs)]=2^15
    lfcs = log2(lfcs)
    
    statsTab = data.frame(lfc=lfcs, 
                          pvalue=pValues,
                          qvalue=adjPs,
                          sig=logAdjPs, 
                          relAbd=allMean, 
                          relAbd_HGC = hgcMean,
                          relAbd_LGC = lgcMean,
                          label=names(lfcs), 
                          stringsAsFactors = F)
    
    if (speciesMode) {
      statsTab = data.frame(lfc=lfcs, 
                            pvalue=pValues,
                            qvalue=adjPs,
                            sig=logAdjPs,
                            relAbd=allMean,
                            relAbd_HGC = hgcMean,
                            relAbd_LGC = lgcMean,
                            label=getSpeciesName(names(lfcs),taxo), 
                            stringsAsFactors = F)
      
    }
    
    return(statsTab)
    
    
  }
  getVolcanoStatPaired <- function(targetMat, hgcSamples, lgcSamples, taxo, speciesMode=F) {
    targetMean = rowMeans(targetMat)
    targetPart  = targetMean/sum(targetMean)
    
    
    hgcMat = targetMat[,match(hgcSamples, colnames(targetMat))]
    lgcMat = targetMat[,match(lgcSamples, colnames(targetMat))]
    
    relAbdMat = getRelAbd(targetMat)
    relAbdHgcMat = relAbdMat[,match(hgcSamples, colnames(targetMat))]
    relAbdLgcMat = relAbdMat[,match(lgcSamples, colnames(targetMat))]
    
    allMean = rowMeans(relAbdMat)
    hgcMean = rowMeans(relAbdHgcMat)
    lgcMean = rowMeans(relAbdLgcMat)
    
    lenSpecies = dim(hgcMat)[1]
    pValues = sapply(1:lenSpecies, function(ind) {
      mergedVec = c(hgcMat[ind,],lgcMat[ind,])
      mergedFac = c(rep("HGC",length(hgcMat[ind,])),
                    rep("LGC",length(lgcMat[ind,])))
      
      wilcoxOut = wilcox.test(mergedVec ~ mergedFac, paird=T)
      return(wilcoxOut$p.value)
    })
    pValues[is.nan(pValues)]=1
    adjPs = p.adjust(pValues,method="BH")
    logAdjPs = -log10(adjPs)
    logAdjPs[logAdjPs>300] = 300
    
    lfcs = sapply(1:lenSpecies,function(ind){
      hmeans = mean(hgcMat[ind,])
      lmeans = mean(lgcMat[ind,]) 
      return(hmeans/lmeans)
    })
    names(lfcs)=rownames(hgcMat)
    lfcs[is.nan(lfcs)]=1
    lfcs[lfcs==0]=2^-15
    lfcs[is.infinite(lfcs)]=2^15
    lfcs = log2(lfcs)
    
    statsTab = data.frame(lfc=lfcs, 
                          pvalue=pValues,
                          qvalue=adjPs,
                          sig=logAdjPs, 
                          relAbd=allMean, 
                          relAbd_HGC = hgcMean,
                          relAbd_LGC = lgcMean,
                          label=names(lfcs), 
                          stringsAsFactors = F)
    
    if (speciesMode) {
      statsTab = data.frame(lfc=lfcs, 
                            pvalue=pValues,
                            qvalue=adjPs,
                            sig=logAdjPs,
                            relAbd=allMean,
                            relAbd_HGC = hgcMean,
                            relAbd_LGC = lgcMean,
                            label=getSpeciesName(names(lfcs),taxo), 
                            stringsAsFactors = F)
      
    }
    
    return(statsTab)
    
    
  }
  
  
  
  
  #### disease signatures ####
  
  plsrDiseaseSignature <- function(sampleTab, featureTab) {
    healthInd <- sampleTab$type == "health"
    diseaseInd <- sampleTab$type == "disease"
    
    currTab = sampleTab[,c("type","geography","sequencer")]
    
    lapply(as.list(colnames(featureTab)), function(currFeature) {
      dataTab = cbind(currTab, feature=featureTab[,currFeature])
      healthTab = dataTab[healthInd,]
      diseaseTab = dataTab[diseaseInd,]
      dataTab = rbind(healthTab, diseaseTab)
      
      out <- tryCatch({
        plsrOut = plsr(feature ~ type + geography, ncomp=15, data=dataTab, validation = "LOO")
        summary(plsrOut)
        plot(RMSEP(plsrOut), legendpos = "topright")
        plot(plsrOut, ncomp = 15, asp = 1, line = TRUE)
        plot(plsrOut, plottype = "scores", comps = 1:4)
        
        coef(plsrOut)
        
      }, error = function(cond){
        message(cond)
        print(currFeature)
        return(c(NA,NA))
      })
    })
    statOut = do.call(rbind,lapply(as.list(colnames(featureTab)), function(currFeature){
      
      dataTab = cbind(currTab, feature=featureTab[,currFeature])
      healthTab = dataTab[healthInd,]
      diseaseTab = dataTab[diseaseInd,]
      dataTab = rbind(healthTab, diseaseTab)
      
      out <-tryCatch({
        lmerOut = plsr(feature ~   type + (1|geography), data=dataTab, na.action = na.exclude)
        
        beta = summary(lmerOut)$coefficients["typehealth","Estimate"]
        betaP = summary(lmerOut)$coefficients["typehealth","Pr(>|t|)"]
        geoCoeff = coef(lmerOut)$geography[,1]
        geoCoeffNames = rownames(coef(lmerOut)$geography)
        names(geoCoeff) = geoCoeffNames
        
        return(c(beta,betaP))
      },
      error = function(cond) {
        message(cond)
        print(currFeature)
        return(c(NA,NA))
      })
    }))
    rownames(statOut) = colnames(featureTab)
    colnames(statOut) = c("estimate", "pvalue")
    
    
    
    
  }
  lmerDiseaseSignature <- function(sampleTab, featureTab) {
    healthInd <- sampleTab$type == "health"
    diseaseInd <- sampleTab$type == "disease"
    
    currTab = sampleTab[,c("type","geography","sequencer")]
    geoOut = lapply(as.list(colnames(featureTab)), function(currFeature){
      
      dataTab = cbind(currTab, feature=featureTab[,currFeature])
      healthTab = dataTab[healthInd,]
      diseaseTab = dataTab[diseaseInd,]
      dataTab = rbind(healthTab, diseaseTab)
      out <-tryCatch({
        lmerOut = lmer(feature ~   type + (1|geography), data=dataTab, na.action = na.exclude)
        
        beta = summary(lmerOut)$coefficients["typehealth","Estimate"]
        betaP = summary(lmerOut)$coefficients["typehealth","Pr(>|t|)"]
        geoCoeff = coef(lmerOut)$geography[,1]
        geoCoeffNames = rownames(coef(lmerOut)$geography)
        names(geoCoeff) = geoCoeffNames
        
        return(geoCoeff)
      },
      error = function(cond) {
        message(cond)
        print(currFeature)
        return(rep(NA,18))
      })
    })
    geoMat = do.call(cbind, geoOut)
    colnames(geoMat) = colnames(featureTab)
    geoMat[is.na(geoMat)] = 0
    
    statOut = do.call(rbind,lapply(as.list(colnames(featureTab)), function(currFeature){
      
      dataTab = cbind(currTab, feature=featureTab[,currFeature])
      healthTab = dataTab[healthInd,]
      diseaseTab = dataTab[diseaseInd,]
      dataTab = rbind(healthTab, diseaseTab)
      
      out <-tryCatch({
        lmerOut = lmer(feature ~   type + (1|geography), data=dataTab, na.action = na.exclude)
        
        beta = summary(lmerOut)$coefficients["typehealth","Estimate"]
        betaP = summary(lmerOut)$coefficients["typehealth","Pr(>|t|)"]
        geoCoeff = coef(lmerOut)$geography[,1]
        geoCoeffNames = rownames(coef(lmerOut)$geography)
        names(geoCoeff) = geoCoeffNames
        
        return(c(beta,betaP))
      },
      error = function(cond) {
        message(cond)
        print(currFeature)
        return(c(NA,NA))
      })
    }))
    rownames(statOut) = colnames(featureTab)
    colnames(statOut) = c("estimate", "pvalue")
    
    
    return(list(geography=geoMat, disease=statOut))
  }
  
  
  
  
  #### pie plots ####
  getRelAbdOfTaxa <- function(targetMat, targetTaxo, taxo) {
    targetTaxoMat = getTaxaSumMat(targetMat, taxo, targetTaxo,T)
    targetTaxoMean = rowMeans(targetTaxoMat)
    relAbds = targetTaxoMean/sum(targetTaxoMean)
    return(relAbds)
  }
  
  getRelAbdWithOthersImputed <- function(currRelAbd, topList){
    newRelAbd = currRelAbd[match(topList, names(currRelAbd))]
    relAbdOthers = sum(currRelAbd[!names(currRelAbd)%in%topList]) 
    newRelAbd = c(newRelAbd, relAbdOthers)
    names(newRelAbd) = c(topList, "others")
    return(newRelAbd)
  }
  
  #### geography signals ####
  showPerBarPlotByCountry <- function(samplesByCountry, target, mergeMat, cutOff=1e-7) {
    perList <- lapply(samplesByCountry, function(currSamples){
      currVec = mergeMat[target,currSamples]
      present = sum(currVec >= cutOff)/length(currVec)
      absent = sum(currVec < cutOff)/length(currVec)
      return(c(present, absent))
    })
    perMat <- do.call(cbind, perList)
    
    par(mar=c(5,10,4,2))
    barplot(perMat, horiz = T, las=1)
    par(mar=c(5.1,4.1,4.1,2.1))
    return(perMat)
  }
  showBoxPlotsByCountry <- function(samplesByCountry, target, mergeMat) {
    abdList <- lapply(samplesByCountry, function(currSamples){
      return(mergeMat[target,currSamples])
    })
    par(mar=c(5,10,4,2))
    boxplot(abdList,horizontal = T,las=1, ylim=c(0,1e-5))
    par(mar=c(5.1,4.1,4.1,2.1))
    return(abdList)
  }
  
  #### hidden cluster analysis ####
  getKeyClusterFromFirst <- function(targetMat, eps=0.3, pts=10) {
    set.seed(1)
    #cl = hdbscan(targetMat, minPts = pts)
    cl = dbscan(targetMat, eps=eps, minPts = pts)
    #plot(cl, show_flat = TRUE)
    #plot(cl)
    print(sort(table(cl$cluster)))
    if (all(cl$cluster=="0")) return(NULL)
    
    keyCluster=names(sort(table(cl$cluster),decreasing = T))
    keyCluster = keyCluster[keyCluster!="0"]
    keyCluster = keyCluster[1]
    print(keyCluster)
    keyClusterInd = which(cl$cluster==keyCluster)
    keyClusterSamples = rownames(targetMat)[keyClusterInd]
    return(keyClusterSamples)
  }
  getKeyClusterFromSecond <- function(targetMat, pts=10) {
    set.seed(1)
    cl = hdbscan(targetMat, minPts = pts)
    plot(cl, show_flat = TRUE)
    plot(cl)
    print(sort(table(cl$cluster)))
    
    keyCluster=names(sort(table(cl$cluster),decreasing = T))[2]
    print(keyCluster)
    keyClusterInd = which(cl$cluster==keyCluster)
    keyClusterSamples = rownames(targetMat)[keyClusterInd]
    return(keyClusterSamples)
  }
  
  
  
  #### functional clusters ####
  getGenesFromVfTerms <- function(vfTerms, patricVfDescTab) {
    genes = patricVfDescTab$Gene[match(vfTerms, patricVfDescTab$RefSeq.Locus.Tag)]
    return(genes)
  }
  getProductsFromVfTerms <- function(vfTerms, patricVfDescTab) {
    products = patricVfDescTab$Product[match(vfTerms, patricVfDescTab$RefSeq.Locus.Tag)]
    return(products)
  }
  getClassesFromVfTerms <- function(vfTerms, patricVfDescTab) {
    classes = patricVfDescTab$Classification[match(vfTerms, patricVfDescTab$RefSeq.Locus.Tag)]
    return(classes)
  }
  grepMspsByKOs <- function(targetKOs, gutKoBestMat, taxo, cutRatio=0.75, debug=F) {
    if (debug) print(length(targetKOs))
    if (length(targetKOs)==1) {
      currMspsByKos = (gutKoBestMat[,targetKOs])
    } else {
      currMspsByKos = rowSums(gutKoBestMat[,targetKOs])
    }
    currMspsByKos = sort(currMspsByKos, decreasing = T)
    currMspsByKosTab = data.frame(msp=names(currMspsByKos), 
                                  species=getSpeciesName(names(currMspsByKos), taxo), 
                                  count = currMspsByKos, stringsAsFactors = F)
    
    cutMspsByKosTab = currMspsByKosTab[currMspsByKosTab$count >= (length(targetKOs)*cutRatio),]
    return(list(msps=cutMspsByKosTab$msp, wholeData=currMspsByKosTab))
    
  }
  
  grepBestMspsAssociatedWithModules <- function(funcModules, funcMat,debug=F) {
    
    mspList = lapply(funcModules, function(currFuncs) {
      lenCurrFuncs = length(currFuncs)
      if (length(currFuncs)==1) {
        currMspsByFuncs = funcMat[,currFuncs]
      } else {
        currMspsByFuncs = rowSums(funcMat[,currFuncs])
      }
      maxRatio = max(currMspsByFuncs)
      currMspsByFuncs = currMspsByFuncs[currMspsByFuncs==maxRatio]
      return(names(currMspsByFuncs))
    })
    return(mspList)
  }
  
  grepBestScoresMspsAssociatedWithModules <- function(funcModules, funcMat,debug=F) {
    
    scoreList = lapply(funcModules, function(currFuncs) {
      lenCurrFuncs = length(currFuncs)
      if (length(currFuncs)==1) {
        currMspsByFuncs = funcMat[,currFuncs]
      } else {
        currMspsByFuncs = rowSums(funcMat[,currFuncs])
      }
      maxRatio = max(currMspsByFuncs)/lenCurrFuncs
      return(maxRatio)
    })
    return(scoreList)
  }
  grepMspsAssociatedWithModules <- function(funcModules, funcMat, cutRatio=0.75, debug=F) {
    
    mspList = lapply(funcModules, function(currFuncs) {
      lenCurrFuncs = length(currFuncs)
      if (length(currFuncs)==1) {
        currMspsByFuncs = funcMat[,currFuncs]
      } else {
        currMspsByFuncs = rowSums(funcMat[,currFuncs])
      }
      currMspsByFuncs = sort(currMspsByFuncs, decreasing = T)
      currMspsByFuncs = currMspsByFuncs[currMspsByFuncs>=(lenCurrFuncs*cutRatio)]
      return(names(currMspsByFuncs))
    })
    return(mspList)
  }
  
  
  
  grepMspsByFuncModules <- function(funcKoModules, gutKoBestMat, taxo, targetModule, cutRatio=0.75, debug=F) {
    targetKOs = funcKoModules[[targetModule]]
    if (debug) print(length(targetKOs))
    if (length(targetKOs)==1) {
      currMspsByKos = (gutKoBestMat[,targetKOs])
    } else {
      currMspsByKos = rowSums(gutKoBestMat[,targetKOs])
    }
    currMspsByKos = sort(currMspsByKos, decreasing = T)
    currMspsByKosTab = data.frame(msp=names(currMspsByKos), 
                                  species=getSpeciesName(names(currMspsByKos), taxo), 
                                  count = currMspsByKos, stringsAsFactors = F)
    
    cutMspsByKosTab = currMspsByKosTab[currMspsByKosTab$count >= (length(targetKOs)*cutRatio),]
    return(list(msps=cutMspsByKosTab$msp, wholeData=currMspsByKosTab))
    
  }
  checkFractionOfGenus <- function(gutTaxo, targetMsps, plotMode = F) {
    gutGenusCount = table(gutTaxo$genus)
    gutGenusCountVec = as.numeric(gutGenusCount)
    names(gutGenusCountVec) = names(gutGenusCount)
    
    targetGenera = gutTaxo$genus[gutTaxo$MSP %in% targetMsps]
    targetGenusCount = table(targetGenera)
    targetGenusCountVec = as.numeric(targetGenusCount)
    names(targetGenusCountVec) = names(targetGenusCount)
    
    matchedGutGenusCountVec = gutGenusCountVec[match(names(targetGenusCountVec),names(gutGenusCountVec))]
    outVec = targetGenusCountVec*100/matchedGutGenusCountVec
    if (plotMode) {
      par(mar=c(4,7,3,1))
      showVec = sort(outVec)
      showVec = showVec[!grepl("unclassified", names(showVec))]
      barplot(showVec, xlab="found species (%)",
              col=colorRampPalette(c("#0000FF", "white", "#808080"))(length(outVec)), 
              cex.axis = 0.6, cex.names = 0.6,
              horiz = T, las=1)
      par(mar=c(5,4,4,1))
    }
    return(outVec)
    
  }
  findModulesBy <- function(funcModules, targetTerms) {
    areContained = do.call(c, lapply(funcModules, function(x) any(x %in% targetTerms) ))
    return(names(funcModules)[areContained])
  }
  
  
  #### function comparisons ####
  getChisqStat <- function(refFuncMat, targetFuncMat, funcs) {
    
    pvalues = sapply(funcs, function(currFunc) {
      refCurrFunc = refFuncMat[currFunc,]
      targetCurrFunc = targetFuncMat[currFunc,]
      
      refPresent = sum(refCurrFunc==1)
      refAbsent = sum(refCurrFunc==0)
      
      targetPresent = sum(targetCurrFunc==1)
      targetAbsent = sum(targetCurrFunc==0)
      
      contigTab = cbind(refFlow=c(present=refPresent, absent=refAbsent),
                        targetFlow=c(present=targetPresent, absent=targetAbsent))
      pvalue=chisq.test(contigTab)$p.value
      return(pvalue)
    }, simplify = F)
    pvalues = unlist(pvalues)
    
    oddsRatios = sapply(funcs, function(currFunc) {
      refCurrFunc = refFuncMat[currFunc,]
      targetCurrFunc = targetFuncMat[currFunc,]
      
      refPresent = sum(refCurrFunc==1)
      refAbsent = sum(refCurrFunc==0)
      
      targetPresent = sum(targetCurrFunc==1)
      targetAbsent = sum(targetCurrFunc==0)
      
      
      oddsRatio = (targetPresent*refAbsent)/(targetAbsent*refPresent)
      if (refPresent*targetAbsent == 0) return(Inf) 
      return(oddsRatio)
    }, simplify = F)
    oddsRatios = unlist(oddsRatios)
    
    chisqStatTab = data.frame(term=names(pvalues),
                              pvalue=pvalues,
                              oddsRatio=oddsRatios,
                              logP = -log10(pvalues),
                              logOdds = log(oddsRatios))
    
    return(chisqStatTab)
  }
  getFractionMat <- function(fMat, bMat, pMat, funcs) {
    ratios = sapply(funcs, function(currFunc) {
      fFunc = fMat[currFunc,]
      bFunc = bMat[currFunc,]
      pFunc = pMat[currFunc,]
      
      fPresent = sum(fFunc==1)
      fAbsent = sum(fFunc==0)
      
      bPresent = sum(bFunc==1)
      bAbsent = sum(bFunc==0)
      
      pPresent = sum(pFunc==1)
      pAbsent = sum(pFunc==0)
      
      fRatio = fPresent/(fPresent + fAbsent)
      bRatio = bPresent/(bPresent + bAbsent)
      pRatio = pPresent/(pPresent + pAbsent)
      
      return(c(fRatio, bRatio, pRatio))
    }, simplify = F)
    ratioMat = do.call(rbind, ratios)
    rownames(ratioMat) = funcs
    colnames(ratioMat) = c("F","B","P")
    return(ratioMat)      
    
  }
  getVarScoreMat <- function(funcMat, funcs, etfSpecies, etbSpecies, etpSpecies) {
    
    selFuncMat = funcMat[,colnames(funcMat) %in% funcs]
    varMat = apply(selFuncMat, 2, function(currFunc) {
      fClass = rep("no",dim(selFuncMat)[1])
      bClass = rep("no",dim(selFuncMat)[1])
      pClass = rep("no",dim(selFuncMat)[1])
      fClass[rownames(selFuncMat)%in% etfSpecies]="yes"
      bClass[rownames(selFuncMat)%in% etbSpecies]="yes"
      pClass[rownames(selFuncMat)%in% etpSpecies]="yes"
      
      currData = data.frame(fClass= fClass, bClass=bClass, pClass=pClass, func=currFunc)
      coefs = summary(lm(func~fClass+bClass+pClass, data=currData))$coefficients[,1]
      fstats = summary(lm(func~fClass+bClass+pClass, data=currData))$fstatistic 
      pvalue = pf(fstats[1], fstats[2], fstats[3], lower.tail = F) 
      
      fRatio = mean(currFunc[fClass=="yes"]) / mean(currFunc[fClass=="no"])
      bRatio = mean(currFunc[bClass=="yes"]) / mean(currFunc[bClass=="no"])
      pRatio = mean(currFunc[pClass=="yes"]) / mean(currFunc[pClass=="no"])
      
      fBeta = coefs[2]
      bBeta = coefs[3]
      pBeta = coefs[4]
      
      fBeta = fBeta^2
      bBeta = bBeta^2
      pBeta = pBeta^2
      
      sumBeta = fBeta+bBeta+pBeta
      vars = c(fBeta, bBeta, pBeta)/sumBeta
      out = c(vars, fRatio, bRatio, pRatio, pvalue)
      names(out) = c("fBeta", "bBeta", "pBeta", "fRatio", "bRatio", "pRatio", "pvalue")
      return(out)
      
    })
    
    varMat = t(varMat)
    varMat = as.data.frame(varMat)
    varMat$term = rownames(varMat)
    return(varMat)
  }
  getVarScoreMatTwo <- function(funcMat, funcs, inSpecies, outSpecies) {
    
    selFuncMat = funcMat[,colnames(funcMat) %in% funcs]
    varMat = apply(selFuncMat, 2, function(currFunc) {
      fClass = rep("no",dim(selFuncMat)[1])
      bClass = rep("no",dim(selFuncMat)[1])
      fClass[rownames(selFuncMat)%in% inSpecies]="yes"
      bClass[rownames(selFuncMat)%in% outSpecies]="yes"
      
      currData = data.frame(fClass= fClass, bClass=bClass,  func=currFunc)
      coefs = summary(lm(func~fClass+bClass , data=currData))$coefficients[,1]
      fstats = summary(lm(func~fClass+bClass , data=currData))$fstatistic 
      pvalue = pf(fstats[1], fstats[2], fstats[3], lower.tail = F) 
      
      fRatio = mean(currFunc[fClass=="yes"]) / mean(currFunc[fClass=="no"])
      bRatio = mean(currFunc[bClass=="yes"]) / mean(currFunc[bClass=="no"])
      
      fBeta = coefs[2]
      bBeta = coefs[3]
      
      fBeta = fBeta^2
      bBeta = bBeta^2
      
      sumBeta = fBeta+bBeta 
      vars = c(fBeta, bBeta )/sumBeta
      out = c(vars, fRatio, bRatio,   pvalue)
      names(out) = c("inBeta", "outBeta",  "inRatio", "outRatio",  "pvalue")
      return(out)
      
    })
    
    varMat = t(varMat)
    varMat = as.data.frame(varMat)
    varMat$term = rownames(varMat)
    return(varMat)
  }
  
  selectVarScoreMat <- function(varMat, pCut=1e-2) {
    pvalueInd = varMat$pvalue < pCut
    
    betaMaxInds = apply(varMat, 1, function(currRow){
      return(which.max(currRow[1:3]))
    })
    
    ratioMaxInds = apply(varMat, 1, function(currRow){
      return(which.max(currRow[4:6]))
    })
    
    consistentInds <- betaMaxInds == ratioMaxInds
    
    
    sigMat = varMat[pvalueInd & consistentInds, ] 
    #sigMat = sigMat[fInd | bInd | pInd, ] 
    return(sigMat)
  }
  selectVarScoreMatTwo <- function(varMat, pCut=1e-2) {
    pvalueInd = varMat$pvalue < pCut
    
    betaMaxInds = apply(varMat, 1, function(currRow){
      return(which.max(currRow[1:2]))
    })
    
    ratioMaxInds = apply(varMat, 1, function(currRow){
      return(which.max(currRow[3:4]))
    })
    
    consistentInds <- betaMaxInds == ratioMaxInds
    
    
    sigMat = varMat[pvalueInd & consistentInds, ] 
    #sigMat = sigMat[fInd | bInd | pInd, ] 
    return(sigMat)
  }
  
  
  
  
}

loadCols = T
if (loadCols) {
  qual_col_pals = brewer.pal.info[brewer.pal.info$category == 'qual',]
  col_vector = unlist(mapply(brewer.pal, qual_col_pals$maxcolors, rownames(qual_col_pals)))
  
  tradVec = c(255,127,0,120)/255
  tradCol = rgb(tradVec[1],tradVec[2],tradVec[3],tradVec[4])
  
  cjpVec = c(211,211,211,120)/255
  cjpCol = rgb(cjpVec[1],cjpVec[2],cjpVec[3],cjpVec[4])
  
  eurVec = c(117,112,179,120)/255
  eurCol = rgb(eurVec[1],eurVec[2],eurVec[3],eurVec[4])
  
  blueCol = rgb(5/255,48/255,97/255,0.5)
  
}

loadModuleMode = T
if (loadModuleMode) {
  koModuleListRData = "J://Deposit/Project/2018_microbiome_atlas/functional.annotation/KEGG module/koModuleList.20190219.RData"
  load(koModuleListRData)
  koModuleList = koModuleList[!names(koModuleList) %in% exModules]
  
  microbeModuleTab = moduleDescTab[moduleDescTab$isBacteriaModule|moduleDescTab$isArchaeaModule,]
  nonMicrobeModuleTab = moduleDescTab[!(moduleDescTab$isBacteriaModule|moduleDescTab$isArchaeaModule),]
  nonMicrobeModuleTab2 = moduleDescTab[grepl("eukaryotes",moduleDescTab$moduleName),]
  
  microbeModules = paste(microbeModuleTab$moduleId, microbeModuleTab$moduleName, sep = ":")
  nonMicrobeModules = paste(nonMicrobeModuleTab$moduleId, nonMicrobeModuleTab$moduleName, sep = ":")
  nonMicrobeModules2 = paste(nonMicrobeModuleTab2$moduleId, nonMicrobeModuleTab2$moduleName, sep = ":")
  nonMicrobeModules = unique(c(nonMicrobeModules,nonMicrobeModules2))
  
  subsystemList = unique(microbeModuleTab$subsystem)
  systemList = unique(microbeModuleTab$system)
  
  koBySubsystemList = sapply(subsystemList, function(subsys) {
    currModules = microbeModuleTab$FullName[microbeModuleTab$subsystem==subsys]
    currKOs = unique(unlist(koModuleList[currModules]))
    return(currKOs)
  })
  
  secretionModules = microbeModuleTab$FullName[microbeModuleTab$subsystem=="Bacterial secretion system"]
  secretionKOs = unique(unlist(koModuleList[secretionModules]))
  
  drugEffluxModules = microbeModuleTab$FullName[microbeModuleTab$subsystem=="Drug efflux transporter/pump"]
  drugEffluxKOs = unique(unlist(koModuleList[drugEffluxModules]))
  
  vitaminModules = microbeModuleTab$FullName[microbeModuleTab$subsystem=="Cofactor and vitamin biosynthesis"]
  vitaminKOs = unique(unlist(koModuleList[vitaminModules]))
  
  
  mineralIonModules = microbeModuleTab$FullName[microbeModuleTab$subsystem=="Mineral and organic ion transport system"]
  mineralIonKOs = unique(unlist(koModuleList[mineralIonModules]))
  
  metalB12Modules = microbeModuleTab$FullName[microbeModuleTab$subsystem=="Metallic cation, iron-siderophore and vitamin B12 transport system"]
  metalB12KOs = unique(unlist(koModuleList[metalB12Modules]))
  
  mineralKOs = unique(c(mineralIonKOs,metalB12KOs))
  
  carbLipidTransportModules = microbeModuleTab$FullName[microbeModuleTab$subsystem=="Saccharide, polyol, and lipid transport system"]
  carbLipidTransportKOs = unique(unlist(koModuleList[carbLipidTransportModules]))
  
  atpModules = microbeModuleTab$FullName[microbeModuleTab$subsystem=="ATP synthesis"]
  atpKOs = unique(unlist(koModuleList[atpModules]))
  
  aaKeywords = c("Serine and threonine metabolism",
                 "Cysteine and methionine metabolism",
                 "Branched-chain amino acid metabolism",
                 "Lysine metabolism",
                 "Arginine and proline metabolism",
                 "Histidine metabolism",
                 "Aromatic amino acid metabolism",
                 "Other amino acid metabolism")
  aminoAcidModules = microbeModuleTab$FullName[microbeModuleTab$subsystem%in% aaKeywords]
  aminoAcidKOs = unique(unlist(koModuleList[aminoAcidModules]))
  
  secondaryModules = microbeModuleTab$FullName[microbeModuleTab$subsystem == "Biosynthesis of secondary metabolites"]
  secondaryKOs = unique(unlist(koModuleList[secondaryModules]))
  
  nSugarModules = microbeModuleTab$FullName[microbeModuleTab$subsystem == "Nucleotide sugar"]
  nSugarKOs = unique(unlist(koModuleList[nSugarModules]))
  
  
  
  
}

loadPathMode = T
if (loadPathMode) {
  
  koPathwayMapRData = "J:\\Deposit\\Project\\2018_oral_catalog_paper\\rackham.data/koElementList.20180609.RData"
  load(koPathwayMapRData)
  
  koElementLength = do.call(c,lapply(koElementList, length))
  
  bacteriaPathFile="J:\\Deposit\\Project\\2018_oral_catalog_paper\\rackham.data/bacteria.kegg.pathway.ids.txt"
  bacteriaPaths = read.table(bacteriaPathFile, stringsAsFactors = F,colClasses = c("character"), sep="\t")[,]
  bacteriaPaths = paste("ko",bacteriaPaths,sep="")
  
  koPaths = names(koElementList)
  extractKoIds = sapply(koPaths, function(path) {
    id=strsplit(path,split = ":")[[1]][1]
    return(id)
  })
  names(extractKoIds)=NULL
  
  exPaths = c("ko00363:Bisphenol A degradation",
              "ko00196:Photosynthesis - antenna proteins",
              "ko00195:Photosynthesis")
  
  bacteriaKoElementList = koElementList[extractKoIds %in% bacteriaPaths]
  bacteriaKoElementList = bacteriaKoElementList[!names(bacteriaKoElementList)%in%exPaths]
  
  #recently added (2019.07.16)-->
  bacteriaKoElementList = lapply(bacteriaKoElementList, unique)
  
  bacteriaPathFullNames = names(bacteriaKoElementList)
  bacteriaKoElementLength = do.call(c,lapply(bacteriaKoElementList, length))
  
}

loadMode = T
if (loadMode) {
  
  funcModulesRData = "J:\\Deposit\\Project\\2018_microbiome_atlas\\functional.annotation//funcModules.20191227.RData"
  load(funcModulesRData)
  
  koDescMapFile = "J:\\Deposit\\Project\\2018_microbiome_atlas\\functional.annotation\\ko_desc_mapping.20190603.RData"
  load(koDescMapFile)
  koDescMap = koDescMap[!duplicated(koDescMap),]
  save(koDescMap, file=koDescMapFile)
  
  pcoaAllUpdatedData = "C:/Data/comparative.analysis.healthy.sweden/pcoaAllData.updated.20190805.RData"
  load(pcoaAllUpdatedData)
  
  pcoaMat = pcoaOut$vectors 
  pcoaGenusMat = pcoaGenusOut$vectors
  pcoaFamilyMat = pcoaFamilyOut$vectors
  pcoaClassMat = pcoaClassOut$vectors
  pcoaOrderMat = pcoaOrderOut$vectors
  dim(pcoaOut$vectors)
  
  gutMsp1992Data = "C:\\Data/comparative.analysis.healthy.sweden/hs_10.4_1992_MSP_freeze2_20180905.RData"
  load(gutMsp1992Data)
  rm(MSP_gc)
  rm(MSP_id)
  gutMsps = rownames(taxo)
  rm(taxo)
  
  mgePfamTermFile = "J:\\Deposit\\Project\\2018_microbiome_atlas\\MGE.pfam.terms.txt"
  mgePfamTerms=read.table(mgePfamTermFile, sep="\t",header=F, stringsAsFactors = F)[,]
  
  loadEmmanuelleRecentTaxa = T
  if (loadEmmanuelleRecentTaxa) {
    gssOssTabNewFile = "J://Deposit//Project//2018_oral_catalog_paper//catalogs (gut, oral), 20190523//MSP_set_class_tropism_20190523_short.txt"
    gssOssTabNew = read.table(gssOssTabNewFile, header=T, sep="\t", stringsAsFactors = F)
    rownames(gssOssTabNew) = gssOssTabNew$MSP
    taxo = gssOssTabNew
    
  }
  
  loadRccNsclcMetadata = T
  if (loadRccNsclcMetadata) {
    rccNsclcMetaFile = "J:\\Deposit\\Project\\2018_microbiome_atlas\\atlas.metadata.collected\\RCC.NSCLC.immunotherapy.metadata.part.txt"
    rccNsclcMetaTab = read.table(rccNsclcMetaFile, sep="\t", stringsAsFactors = F, header=T)
    respondingSamples = rccNsclcMetaTab$sample.ID[rccNsclcMetaTab$subtype %in% c("Stable","Partial response")]
    nonRespondingSamples = rccNsclcMetaTab$sample.ID[rccNsclcMetaTab$subtype %in% c("Progression","Dead")]
  }
  
  suppressMsps = c("msp_1974","msp_2153")
  suppressMsps2 = c("msp_1974","msp_2153", "msp_0864")
  gutMsps = gutMsps[!gutMsps %in% suppressMsps2]
  
  taxo = taxo[!rownames(taxo) %in% suppressMsps,]
  gutTaxo = taxo[gutMsps,]
  
  mgsInfoFile="C:\\Data/comparative.analysis.healthy.sweden/mgs.files.info.txt"
  mgsInfo = read.table(mgsInfoFile, header = T, sep = "\t",stringsAsFactors = F)
  
  gssSpecies = gutTaxo$MSP[gutTaxo$MSP %in% taxo$MSP[taxo$type_sh=="gss"]]
  ossSpecies = gutTaxo$MSP[gutTaxo$MSP %in% taxo$MSP[taxo$type_sh=="oss"]]
  gutOnlySpecies = gutTaxo$MSP[gutTaxo$MSP %in% taxo$MSP[taxo$type_sh=="gut"]]
  oralOnlySpecies = gutTaxo$MSP[gutTaxo$MSP %in% taxo$MSP[taxo$type_sh=="oral"]]
  
  save(gutTaxo, file="J:\\Deposit\\Project\\2018_microbiome_atlas\\gutTaxo.RData")
  
  # load(mgsClean2RData)
  # load(basicMetaCleanRData2)
  
  mgsCleanUpdatedRData2 = "C:\\Data/comparative.analysis.healthy.sweden/mgs.clean.updated.20190726.10M.RData"
  load(mgsCleanUpdatedRData2)
  
  mergeMatUpdated = mergeMatUpdated[!rownames(mergeMatUpdated)%in%suppressMsps,]
  
  basicMetaCleanRDataUpdated2 = "C:\\Data/comparative.analysis.healthy.sweden/all.basic.clean.metadata.behcet.20190805.RData"
  load(basicMetaCleanRDataUpdated2)
  
  
  metadataForReza = F
  if (metadataForReza) {
    #write.table(basicMetaMapUpdated, "C://Data//comparative.analysis.healthy.sweden//basic.metadata.reza.2019.11.26.txt", sep="\t")
    write.table(basicMetaMapUpdated, "C://Data//comparative.analysis.healthy.sweden//basic.metadata.reza.2020.06.19.txt", sep="\t")
    
  }
  
  fixWrongCode = F
  if (fixWrongCode) {
    basicMetaMapUpdated$WhatImputed[basicMetaMapUpdated$WhatImputed=="None"]="XXX"
    basicMetaMapUpdated$Imputation[basicMetaMapUpdated$Imputation=="XXX"]="None"
    save(basicMetaMapUpdated, file= basicMetaCleanRDataUpdated)
  }
  
  hqDataRData="C:/Data/comparative.analysis.healthy.sweden/high.qualtiy.metadata.tables.RData"
  load(hqDataRData)
  
  taxoByPhyla = split(rownames(taxo),taxo$phylum)
  taxoByClass = split(rownames(taxo),taxo$class)
  taxoByFamily = split(rownames(taxo),taxo$family)
  taxoByGenus = split(rownames(taxo),taxo$genus)
  hqDataTab = hqDataTab
  
  taxaLevelMode = T
  if (taxaLevelMode) {
    # genusMat = getTaxaSumMat(mergeMat, taxo, "genus")
    # familyMat = getTaxaSumMat(mergeMat, taxo, "family")
    # classMat = getTaxaSumMat(mergeMat, taxo, "class")
    # orderMat = getTaxaSumMat(mergeMat, taxo, "order")
    # phylumMat = getTaxaSumMat(mergeMat, taxo, "phylum")
    
    genusMatUpdated = getTaxaSumMat(mergeMatUpdated, taxo, "genus")
    familyMatUpdated = getTaxaSumMat(mergeMatUpdated, taxo, "family")
    orderMatUpdated = getTaxaSumMat(mergeMatUpdated, taxo, "order")
    classMatUpdated = getTaxaSumMat(mergeMatUpdated, taxo, "class")
    phylumMatUpdated = getTaxaSumMat(mergeMatUpdated, taxo, "phylum")
  }
  
  if (F) {
    mgsRichness = colSums(mergeMat>0)
    basicMetaMap$MgsRichness = mgsRichness
    basicMetaMap$GeneRichness = mergeVec[match(basicMetaMap$sample.ID, names(mergeVec))]
    save(basicMetaMap, file=basicMetaCleanRData3)
    
    phageRichness = colSums(mergedPhageMatFinal>0)
    table(basicMetaMap$sample.ID == names(phageRichness))
    basicMetaMap$PhageRichness = phageRichness
    save(basicMetaMap, file=basicMetaCleanRData3)
  }
  
  mergeMatRelAbd = getRelAbd(mergeMatUpdated)
  mergeMatCumSum = apply(mergeMatRelAbd, 2, function(currCol){
    return(cumsum(sort(currCol, decreasing=T)))
  })
  
  checkMode = F
  if (checkMode) {
    cutValue = 0.9
    whereCuts90 = apply(mergeMatCumSum,2, function(currCol){
      out = min(which(currCol >= cutValue))
      return(out)
    })
    cutValue = 0.5
    whereCuts50 = apply(mergeMatCumSum,2, function(currCol){
      out = min(which(currCol >= cutValue))
      return(out)
    })
    
    boxplot(list(Q0.9=whereCuts90, 
                 Q0.5=whereCuts50))
  }
  
}

loadUniqSampleInfo = T
if (loadUniqSampleInfo) {
  
  cohortNameMode = T
  if (cohortNameMode) {
    cohortIdByDiseases = list(Ankylosing = "Ankylosing_CN_id9",
                              Behcet = "Behcet_CN_id52",
                              CVD = "CVD_CN_id10",
                              GDM = "GDM_CN_id12",
                              Cirrhosis = c("Cirrhosis_CN_id20", "Cirrhosis_UK_id27"),
                              CD = "CD_CN_id6",
                              RA = "RA_CN_id31",
                              T2D = c("T2D_CN_id2","T2D_SW_id14"),
                              Atherosclerosis = "Atherosclerosis_SW_id1",
                              CRC = c("CRC_JP_id45", "CRC_US_id11", "CRC_IT_id46","CRC_DE_id44"),
                              Obesity = c("Obesity_DK_id16", "Obesity_DK_id25"),
                              CFS = "CFS_US_id34",
                              Melanoma = "Melanoma_US_id3",
                              NAFLD = c("NAFLD_US_id7","NAFLD_IT_id40","NAFLD_ES_id40"),
                              T1D = c("T1D_FI_id21","T1D_LU_id35"),
                              PD="PD_DE_id8",
                              IBD="IBD_ES_id17", 
                              NSCLC="NSCLC_FR_id26",
                              RCC="RCC_FR_id41")
    
    cohortIdByGeos = list(China = c("Ankylosing_CN_id9",
                                    "Behcet_CN_id52",
                                    "Cirrhosis_CN_id20",
                                    "CVD_CN_id10",
                                    "GDM_CN_id12",
                                    "CD_CN_id6",
                                    "RA_CN_id31",
                                    "T2D_CN_id2",
                                    "China_id10",
                                    "China_id12",
                                    "China_id2",
                                    "China_id20",
                                    "China_id31",
                                    "China_id52",
                                    "China_id6",
                                    "China_id9"),
                          
                          Japan = c("CRC_JP_id45", 
                                    "Japan_id45"),
                          
                          UK = c("Cirrhosis_UK_id27",
                                 "UK_id39",
                                 "UK_id29"),
                          
                          US = c("CFS_US_id34",
                                 "CRC_US_id11",
                                 "Melanoma_NR_US_id3",
                                 "Melanoma_R_US_id3",
                                 "NAFLD_US_id7",
                                 "US_id11",
                                 "US_id34",
                                 "US_id36",
                                 "US_id43"),
                          
                          Sweden = c("Sweden_id1",
                                     "Sweden_id14",
                                     "Sweden_id28",
                                     "Atherosclerosis_SW_id1",
                                     "IGT_SW_id14",
                                     "T2D_SW_id14"),
                          
                          Denmark = c("Obesity_DK_id16",
                                      "Obesity_DK_id25",
                                      "Denmark_id16",
                                      "Denmark_id25"),
                          Italy = c("CRC_IT_id46",
                                    "NAFLD_IT_id40",
                                    "Italy_id37", 
                                    "Italy_id46"),
                          
                          Finland = c("Finland_id21",
                                      "T1D_FI_id21"),
                          Germany = c("CRC_DE_id44",
                                      "PD_DE_id8",
                                      "Germany_id44", 
                                      "Germany_id8"),
                          Spain = c("IBD_ES_id17", 
                                    "NAFLD_ES_id40",
                                    "T2D_ES_id53",
                                    "Spain_id17"), 
                          Luxembourg = c("Luxembourg_id35", 
                                         "T1D_LU_id35" ),
                          France = c("NSCLC_FR_id26",
                                     "RCC_FR_id41"), 
                          
                          Fiji = "Fiji_id49",
                          Madagascar = "Madagascar_id42",
                          Mongo = "Mongo_id33",
                          India = "India_id50",
                          Peru = "Peru_id32",
                          Tanzania = "Tanzania_id37",
                          Thailand = "Thailand_id43")
  }
  
  diseaseMode = T
  if (diseaseMode) {
    diseaseUniqSubjectList = list()
    diseaseDatasetsNoDup = c("Atherosclerosis_SW_id1",
                             "CVD_CN_id10",
                             "GDM_CN_id12",
                             "T2D_CN_id2",
                             "T1D_FI_id21",
                             "Cirrhosis_UK_id27",
                             "CFS_US_id34",
                             "NAFLD_IT_id40",
                             "NAFLD_ES_id40",
                             "CRC_JP_id45",
                             "CRC_IT_id46",
                             "Behcet_CN_id52",
                             "CD_CN_id6",
                             "Ankylosing_CN_id9")
    
    
    
    diseaseUniqSubjectList$Atherosclerosis_SW_id1 = basicMetaMapUpdated$sample.ID[basicMetaMapUpdated$type=="Case" 
                                                                                  & basicMetaMapUpdated$dataset.ID=="id1"]
    
    diseaseUniqSubjectList$CVD_CN_id10 = basicMetaMapUpdated$sample.ID[basicMetaMapUpdated$type=="Case" 
                                                                       & basicMetaMapUpdated$dataset.ID=="id10"]
    
    diseaseUniqSubjectList$GDM_CN_id12 = basicMetaMapUpdated$sample.ID[basicMetaMapUpdated$type=="Case" 
                                                                       & basicMetaMapUpdated$dataset.ID=="id12"]
    
    diseaseUniqSubjectList$T2D_CN_id2 = basicMetaMapUpdated$sample.ID[basicMetaMapUpdated$type=="Case" 
                                                                      & basicMetaMapUpdated$dataset.ID=="id2"]
    
    diseaseUniqSubjectList$T1D_FI_id21 = basicMetaMapUpdated$sample.ID[basicMetaMapUpdated$type=="Case" 
                                                                       & basicMetaMapUpdated$dataset.ID=="id21"]
    
    diseaseUniqSubjectList$Cirrhosis_UK_id27 = basicMetaMapUpdated$sample.ID[basicMetaMapUpdated$type=="Case" 
                                                                             & basicMetaMapUpdated$dataset.ID=="id27"]
    
    diseaseUniqSubjectList$CFS_US_id34 = basicMetaMapUpdated$sample.ID[basicMetaMapUpdated$type=="Case" 
                                                                       & basicMetaMapUpdated$dataset.ID=="id34"]
    
    diseaseUniqSubjectList$NAFLD_IT_id40 = basicMetaMapUpdated$sample.ID[basicMetaMapUpdated$type=="Case" 
                                                                         & basicMetaMapUpdated$dataset.ID=="id40"
                                                                         & basicMetaMapUpdated$Geography == "Italy"]
    
    diseaseUniqSubjectList$NAFLD_ES_id40 = basicMetaMapUpdated$sample.ID[basicMetaMapUpdated$type=="Case" 
                                                                         & basicMetaMapUpdated$dataset.ID=="id40"
                                                                         & basicMetaMapUpdated$Geography == "Spain"]
    
    diseaseUniqSubjectList$CRC_JP_id45 = basicMetaMapUpdated$sample.ID[basicMetaMapUpdated$type=="Case" 
                                                                       & basicMetaMapUpdated$dataset.ID=="id45"]
    
    diseaseUniqSubjectList$CRC_IT_id46 = basicMetaMapUpdated$sample.ID[basicMetaMapUpdated$type=="Case" 
                                                                       & basicMetaMapUpdated$dataset.ID=="id46"]
    
    diseaseUniqSubjectList$Behcet_CN_id52 = basicMetaMapUpdated$sample.ID[basicMetaMapUpdated$type=="Case" 
                                                                          & basicMetaMapUpdated$dataset.ID=="id52"]
    
    diseaseUniqSubjectList$CD_CN_id6 = basicMetaMapUpdated$sample.ID[basicMetaMapUpdated$type=="Case" 
                                                                     & basicMetaMapUpdated$dataset.ID=="id6"]
    
    diseaseUniqSubjectList$Ankylosing_CN_id9 = basicMetaMapUpdated$sample.ID[basicMetaMapUpdated$type=="Case" 
                                                                             & basicMetaMapUpdated$dataset.ID=="id9"]
    
    
    diseaseDatasetsToCheck = c("CRC_US_id11",
                               "IGT_SW_id14",
                               "T2D_SW_id14",
                               "Obesity_DK_id16",
                               "IBD_ES_id17",
                               "Cirrhosis_CN_id20",
                               "Obesity_DK_id25",
                               "NSCLC_FR_id26",
                               "NSCLC_R_FR_id26",
                               "NSCLC_NR_FR_id26",
                               "Melanoma_US_id3",
                               "Melanoma_R_US_id3",
                               "Melanoma_NR_US_id3",
                               "RA_CN_id31",
                               "T1D_LU_id35",
                               "RCC_FR_id41",
                               "RCC_R_FR_id41",
                               "RCC_NR_FR_id41",
                               "CRC_DE_id44",
                               "T2D_ES_id53",
                               "T2D_Metformin_ES_id53",
                               "NAFLD_US_id7",
                               "LiverFibrosis_US_id7",
                               "PD_DE_id8")
    
    ### id11-CRC ###
    id11DiseaseTab = basicMetaMapUpdated[basicMetaMapUpdated$dataset.ID=="id11" & 
                                           basicMetaMapUpdated$type=="Case",]
    id11DiseaseTab = id11DiseaseTab[order(id11DiseaseTab$sample.ID),]
    id11DiseaseUniqSamples = id11DiseaseTab$sample.ID[!duplicated(id11DiseaseTab$metadata.ID)]
    diseaseUniqSubjectList$CRC_US_id11 = id11DiseaseUniqSamples
    
    ### id14-T2D ###
    diseaseUniqSubjectList$T2D_SW_id14 = basicMetaMapUpdated$sample.ID[basicMetaMapUpdated$type=="Case" 
                                                                       & basicMetaMapUpdated$dataset.ID=="id14"
                                                                       & basicMetaMapUpdated$subtype=="T2D"]
    ### id14-IGT ###
    diseaseUniqSubjectList$IGT_SW_id14 = basicMetaMapUpdated$sample.ID[basicMetaMapUpdated$type=="Case" 
                                                                       & basicMetaMapUpdated$dataset.ID=="id14"
                                                                       & basicMetaMapUpdated$subtype=="IGT"]
    ### id16-Obesity ###
    id16DiseaseTab = basicMetaMapUpdated[basicMetaMapUpdated$dataset.ID=="id16" & 
                                           basicMetaMapUpdated$type=="Case",]
    id16DiseaseTab = id16DiseaseTab[order(id16DiseaseTab$sample.ID),]
    id16DiseaseUniqSamples = id16DiseaseTab$sample.ID[!duplicated(id16DiseaseTab$metadata.ID)]
    diseaseUniqSubjectList$Obesity_DK_id16 = id16DiseaseUniqSamples
    
    ### id17-IBD ###
    id17DiseaseTab = basicMetaMapUpdated[basicMetaMapUpdated$dataset.ID=="id17" & 
                                           basicMetaMapUpdated$type=="Case",]
    id17DiseaseTab = id17DiseaseTab[order(id17DiseaseTab$sample.ID),]
    id17DiseaseUniqSamples = id17DiseaseTab$sample.ID[!duplicated(id17DiseaseTab$metadata.ID)]
    diseaseUniqSubjectList$IBD_ES_id17 = id17DiseaseUniqSamples
    
    ### id20-cirrhosis ###
    id20DiseaseTab = basicMetaMapUpdated[basicMetaMapUpdated$dataset.ID=="id20" & 
                                           basicMetaMapUpdated$type=="Case",]
    id20DiseaseTab = id20DiseaseTab[order(id20DiseaseTab$sample.ID),]
    id20DiseaseUniqSamples = id20DiseaseTab$sample.ID[!duplicated(id20DiseaseTab$metadata.ID)]
    diseaseUniqSubjectList$Cirrhosis_CN_id20 = id20DiseaseUniqSamples
    
    ### id25-Obesity ###
    id25DiseaseTab = basicMetaMapUpdated[basicMetaMapUpdated$dataset.ID=="id25" & 
                                           basicMetaMapUpdated$type=="Case",]
    id25DiseaseTab = id25DiseaseTab[order(id25DiseaseTab$sample.ID),]
    id25DiseaseUniqSamples = id25DiseaseTab$sample.ID[!duplicated(id25DiseaseTab$metadata.ID)]
    diseaseUniqSubjectList$Obesity_DK_id25 = id25DiseaseUniqSamples
    
    ### id26-NSCLC ###
    id26DiseaseTab = basicMetaMapUpdated[basicMetaMapUpdated$dataset.ID=="id26" & 
                                           basicMetaMapUpdated$type=="Case" &
                                           basicMetaMapUpdated$subtype=="Before",]
    id26DiseaseTab = id26DiseaseTab[order(id26DiseaseTab$sample.ID),]
    id26DiseaseUniqSamples = id26DiseaseTab$sample.ID[!duplicated(id26DiseaseTab$metadata.ID)]
    diseaseUniqSubjectList$NSCLC_FR_id26 = id26DiseaseUniqSamples
    diseaseUniqSubjectList$NSCLC_R_FR_id26 = id26DiseaseUniqSamples[id26DiseaseUniqSamples %in% respondingSamples]
    diseaseUniqSubjectList$NSCLC_NR_FR_id26 = id26DiseaseUniqSamples[id26DiseaseUniqSamples %in% nonRespondingSamples]
    
    ### id3-melanoma ###
    diseaseUniqSubjectList$Melanoma_US_id3 = basicMetaMapUpdated$sample.ID[basicMetaMapUpdated$type=="Case" 
                                                                           & basicMetaMapUpdated$dataset.ID=="id3"]
    
    ### id3-reponse ###
    diseaseUniqSubjectList$Melanoma_R_US_id3 = basicMetaMapUpdated$sample.ID[basicMetaMapUpdated$type=="Case" 
                                                                             & basicMetaMapUpdated$dataset.ID=="id3"
                                                                             & basicMetaMapUpdated$subtype %in% c("Response","Stable")]
    ### id3-non-reponse ###
    diseaseUniqSubjectList$Melanoma_NR_US_id3 = basicMetaMapUpdated$sample.ID[basicMetaMapUpdated$type=="Case" 
                                                                              & basicMetaMapUpdated$dataset.ID=="id3"
                                                                              & basicMetaMapUpdated$subtype == "Progression"]
    
    ### id31-RA ###
    id31DiseaseTab = basicMetaMapUpdated[basicMetaMapUpdated$dataset.ID=="id31" & 
                                           basicMetaMapUpdated$type=="Case" ,]
    id31DiseaseTab = id31DiseaseTab[order(id31DiseaseTab$sample.ID),]
    id31DiseaseUniqSamples = id31DiseaseTab$sample.ID[!duplicated(id31DiseaseTab$metadata.ID)]
    diseaseUniqSubjectList$RA_CN_id31 = id31DiseaseUniqSamples
    
    ### id35-T1D ###
    id35DiseaseTab = basicMetaMapUpdated[basicMetaMapUpdated$dataset.ID=="id35" & 
                                           basicMetaMapUpdated$type=="Case" ,]
    id35DiseaseTab = id35DiseaseTab[order(id35DiseaseTab$subtype),]
    id35DiseaseUniqSamples = id35DiseaseTab$sample.ID[!duplicated(id35DiseaseTab$metadata.ID)]
    diseaseUniqSubjectList$T1D_LU_id35 = id35DiseaseUniqSamples
    
    ### id41-RCC ###
    id41DiseaseTab = basicMetaMapUpdated[basicMetaMapUpdated$dataset.ID=="id41" & 
                                           basicMetaMapUpdated$type=="Case" &
                                           basicMetaMapUpdated$subtype=="before",]
    id41DiseaseTab = id41DiseaseTab[order(id41DiseaseTab$sample.ID),]
    id41DiseaseUniqSamples = id41DiseaseTab$sample.ID[!duplicated(id41DiseaseTab$metadata.ID)]
    diseaseUniqSubjectList$RCC_FR_id41 = id41DiseaseUniqSamples
    diseaseUniqSubjectList$RCC_R_FR_id26 = id41DiseaseUniqSamples[id41DiseaseUniqSamples %in% respondingSamples]
    diseaseUniqSubjectList$RCC_NR_FR_id26 = id41DiseaseUniqSamples[id41DiseaseUniqSamples %in% nonRespondingSamples]
    
    ### id44-CRC ###
    id44DiseaseTab = basicMetaMapUpdated[basicMetaMapUpdated$dataset.ID=="id44" & 
                                           basicMetaMapUpdated$type=="Case",]
    id44DiseaseTab = id44DiseaseTab[order(id44DiseaseTab$sample.ID),]
    id44DiseaseUniqSamples = id44DiseaseTab$sample.ID[!duplicated(id44DiseaseTab$metadata.ID)]
    diseaseUniqSubjectList$CRC_DE_id44 = id44DiseaseUniqSamples
    
    ### id53-T2D ###
    id53DiseaseTab = basicMetaMapUpdated[basicMetaMapUpdated$dataset.ID=="id53" & 
                                           basicMetaMapUpdated$type=="Case" &
                                           basicMetaMapUpdated$subtype %in% c("P0","M0"),]
    id53DiseaseTab = id53DiseaseTab[order(id53DiseaseTab$sample.ID),]
    id53DiseaseUniqSamples = id53DiseaseTab$sample.ID[!duplicated(id53DiseaseTab$metadata.ID)]
    diseaseUniqSubjectList$T2D_ES_id53 = id53DiseaseUniqSamples
    
    ### id53-T2D-metformin ###
    if (F) {
      id53DiseaseTab = basicMetaMapUpdated[basicMetaMapUpdated$dataset.ID=="id53" & 
                                             basicMetaMapUpdated$type=="Case" &
                                             basicMetaMapUpdated$subtype %in% c("M2","M4","P/M6"),]
      id53DiseaseTab = id53DiseaseTab[order(id53DiseaseTab$sample.ID),]
      id53DiseaseUniqSamples = id53DiseaseTab$sample.ID[!duplicated(id53DiseaseTab$metadata.ID)]
      diseaseUniqSubjectList$T2D_Metofrmin_ES_id53 = id53DiseaseUniqSamples
      
    }
    
    ### id7-NAFLD ###
    id7DiseaseTab = basicMetaMapUpdated[basicMetaMapUpdated$dataset.ID=="id7" & 
                                          basicMetaMapUpdated$type=="Case",]
    id7DiseaseTab = id7DiseaseTab[order(id7DiseaseTab$subtype),]
    id7DiseaseUniqSamples = id7DiseaseTab$sample.ID[!duplicated(id7DiseaseTab$metadata.ID)]
    diseaseUniqSubjectList$NAFLD_US_id7 = id7DiseaseUniqSamples
    
    
    ### id8-PD ###
    id8DiseaseTab = basicMetaMapUpdated[basicMetaMapUpdated$dataset.ID=="id8" & 
                                          basicMetaMapUpdated$type=="Case" ,]
    id8DiseaseTab = id8DiseaseTab[order(id8DiseaseTab$subtype),]
    id8DiseaseUniqSamples = id8DiseaseTab$sample.ID[!duplicated(id8DiseaseTab$metadata.ID)]
    diseaseUniqSubjectList$PD_DE_id8 = id8DiseaseUniqSamples
    
    diseaseUniqAll = unique(do.call(c, diseaseUniqSubjectList))
    
    diseaseUniqList = lapply(cohortIdByDiseases, function(currCohorts){
      samples = unique(do.call(c, diseaseUniqSubjectList[currCohorts]))
      return(samples)
    })
    
  }
  
  normMode = T
  if (normMode) {
    
    normUniqSubjectList = list()
    
    normDatasetsNoDup = c("Sweden_id1",
                          "China_id10", 
                          "China_id12",
                          "Sweden_id14",
                          "China_id2",
                          "Finland_id21",
                          "UK_id29",
                          "Peru_id32",
                          "Mongo_id33",
                          "US_id34",
                          "Tanzania_id37",
                          "UK_id39",
                          "Madagascar_id42",
                          "Germany_id44",
                          "Japan_id45",
                          "Italy_id46",
                          "India_id50",
                          "China_id52",
                          "China_id6",
                          "China_id9",
                          "Italy_id37")
    
    normUniqSubjectList$Sweden_id1 = basicMetaMapUpdated$sample.ID[basicMetaMapUpdated$type=="Control" 
                                                                   & basicMetaMapUpdated$dataset.ID=="id1"]
    normUniqSubjectList$China_id10 = basicMetaMapUpdated$sample.ID[basicMetaMapUpdated$type=="Control" 
                                                                   & basicMetaMapUpdated$dataset.ID=="id10"]
    normUniqSubjectList$China_id12 = basicMetaMapUpdated$sample.ID[basicMetaMapUpdated$type=="Control" 
                                                                   & basicMetaMapUpdated$dataset.ID=="id12"]
    normUniqSubjectList$Sweden_id14 = basicMetaMapUpdated$sample.ID[basicMetaMapUpdated$type=="Control" 
                                                                    & basicMetaMapUpdated$dataset.ID=="id14"]
    normUniqSubjectList$China_id2 = basicMetaMapUpdated$sample.ID[basicMetaMapUpdated$type=="Control" 
                                                                  & basicMetaMapUpdated$dataset.ID=="id2"]
    normUniqSubjectList$Finland_id21 = basicMetaMapUpdated$sample.ID[basicMetaMapUpdated$type=="Control" 
                                                                     & basicMetaMapUpdated$dataset.ID=="id21"]
    normUniqSubjectList$UK_id29 = basicMetaMapUpdated$sample.ID[basicMetaMapUpdated$type=="Case" 
                                                                & basicMetaMapUpdated$dataset.ID=="id29"]
    normUniqSubjectList$Peru_id32 = basicMetaMapUpdated$sample.ID[basicMetaMapUpdated$type=="Case" 
                                                                  & basicMetaMapUpdated$dataset.ID=="id32"]
    normUniqSubjectList$Mongo_id33 = basicMetaMapUpdated$sample.ID[basicMetaMapUpdated$type=="Case" 
                                                                   & basicMetaMapUpdated$dataset.ID=="id33"]
    normUniqSubjectList$US_id34 = basicMetaMapUpdated$sample.ID[basicMetaMapUpdated$type=="Control" 
                                                                & basicMetaMapUpdated$dataset.ID=="id34"]
    normUniqSubjectList$Tanzania_id37 = basicMetaMapUpdated$sample.ID[basicMetaMapUpdated$type=="Case" 
                                                                      & basicMetaMapUpdated$dataset.ID=="id37"]
    normUniqSubjectList$UK_id39 = basicMetaMapUpdated$sample.ID[basicMetaMapUpdated$type=="Case" 
                                                                & basicMetaMapUpdated$dataset.ID=="id39"]
    normUniqSubjectList$Madagascar_id42 = basicMetaMapUpdated$sample.ID[basicMetaMapUpdated$type=="Case" 
                                                                        & basicMetaMapUpdated$dataset.ID=="id42"]
    normUniqSubjectList$Germany_id44 = basicMetaMapUpdated$sample.ID[basicMetaMapUpdated$type=="Control" 
                                                                     & basicMetaMapUpdated$dataset.ID=="id44"]
    normUniqSubjectList$Japan_id45 = basicMetaMapUpdated$sample.ID[basicMetaMapUpdated$type=="Control" 
                                                                   & basicMetaMapUpdated$dataset.ID=="id45"]
    normUniqSubjectList$Italy_id46 = basicMetaMapUpdated$sample.ID[basicMetaMapUpdated$type=="Control" 
                                                                   & basicMetaMapUpdated$dataset.ID=="id46"]
    normUniqSubjectList$India_id50 = basicMetaMapUpdated$sample.ID[basicMetaMapUpdated$type=="Case" 
                                                                   & basicMetaMapUpdated$dataset.ID=="id50"]
    normUniqSubjectList$China_id52 = basicMetaMapUpdated$sample.ID[basicMetaMapUpdated$type=="Control" 
                                                                   & basicMetaMapUpdated$dataset.ID=="id52"]
    normUniqSubjectList$China_id6 = basicMetaMapUpdated$sample.ID[basicMetaMapUpdated$type=="Control" 
                                                                  & basicMetaMapUpdated$dataset.ID=="id6"]
    normUniqSubjectList$China_id9 = basicMetaMapUpdated$sample.ID[basicMetaMapUpdated$type=="Control" 
                                                                  & basicMetaMapUpdated$dataset.ID=="id9"]
    
    normUniqSubjectList$Italy_id37 = basicMetaMapUpdated$sample.ID[basicMetaMapUpdated$type=="Control" 
                                                                   & basicMetaMapUpdated$dataset.ID=="id37"]
    
    normDatasetsToCheck = c("US_id11", #duplicated
                            "Denmark_id16", 
                            "Spain_id17",
                            "China_id20",
                            "Denmark_id25",
                            "Sweden_id28",
                            "China_id31",
                            "Luxembourg_id35",
                            "US_id36",
                            "Thailand_id43",
                            "US_id43",
                            "Fiji_id49",
                            "Germany_id8")
    
    
    ### id11 ###
    id11NormTab = basicMetaMapUpdated[basicMetaMapUpdated$dataset.ID=="id11" & 
                                        basicMetaMapUpdated$type=="Control",]
    id11NormTab = id11NormTab[order(id11NormTab$sample.ID),]
    id11NormUniqSamples = id11NormTab$sample.ID[!duplicated(id11NormTab$metadata.ID)]
    normUniqSubjectList$US_id11 = id11NormUniqSamples
    
    ### id16 ###
    id16NormTab = basicMetaMapUpdated[basicMetaMapUpdated$dataset.ID=="id16" & 
                                        basicMetaMapUpdated$type=="Control",]
    id16NormTab = id16NormTab[order(id16NormTab$sample.ID),]
    id16NormUniqSamples = id16NormTab$sample.ID[!duplicated(id16NormTab$metadata.ID)]
    normUniqSubjectList$Denmark_id16 = id16NormUniqSamples
    
    ### id17 ###
    id17NormTab = basicMetaMapUpdated[basicMetaMapUpdated$dataset.ID=="id17" & 
                                        basicMetaMapUpdated$type=="Control",]
    id17NormTab = id17NormTab[order(id17NormTab$sample.ID),]
    id17NormUniqSamples = id17NormTab$sample.ID[!duplicated(id17NormTab$metadata.ID)]
    normUniqSubjectList$Spain_id17 = id17NormUniqSamples
    
    ### id20 ###
    id20NormTab = basicMetaMapUpdated[basicMetaMapUpdated$dataset.ID=="id20" & 
                                        basicMetaMapUpdated$type=="Control",]
    id20NormTab = id20NormTab[order(id20NormTab$sample.ID),]
    id20NormUniqSamples = id20NormTab$sample.ID[!duplicated(id20NormTab$metadata.ID)]
    normUniqSubjectList$China_id20 = id20NormUniqSamples
    
    ### id25 ###
    id25NormTab = basicMetaMapUpdated[basicMetaMapUpdated$dataset.ID=="id25" & 
                                        basicMetaMapUpdated$type=="Control",]
    id25NormTab = id25NormTab[order(id25NormTab$sample.ID),]
    id25NormUniqSamples = id25NormTab$sample.ID[!duplicated(id25NormTab$metadata.ID)]
    normUniqSubjectList$Denmark_id25 = id25NormUniqSamples
    
    ### id28 ###
    id28NormTab = basicMetaMapUpdated[basicMetaMapUpdated$dataset.ID=="id28" & 
                                        basicMetaMapUpdated$type=="Case",]
    id28NormTab = id28NormTab[order(id28NormTab$subtype),]
    id28NormUniqSamples = id28NormTab$sample.ID[!duplicated(id28NormTab$metadata.ID)]
    normUniqSubjectList$Sweden_id28 = id28NormUniqSamples
    
    ### id31 ###
    id31NormTab = basicMetaMapUpdated[basicMetaMapUpdated$dataset.ID=="id31" & 
                                        basicMetaMapUpdated$type=="Control",]
    id31NormTab = id31NormTab[order(id31NormTab$sample.ID),]
    id31NormUniqSamples = id31NormTab$sample.ID[!duplicated(id31NormTab$metadata.ID)]
    normUniqSubjectList$China_id31 = id31NormUniqSamples
    
    ### id35 ###
    id35NormTab = basicMetaMapUpdated[basicMetaMapUpdated$dataset.ID=="id35" & 
                                        basicMetaMapUpdated$type=="Control",]
    id35NormTab = id35NormTab[order(id35NormTab$subtype),]
    id35NormUniqSamples = id35NormTab$sample.ID[!duplicated(id35NormTab$metadata.ID)]
    normUniqSubjectList$Luxembourg_id35 = id35NormUniqSamples
    
    ### id36 ###
    id36NormTab = basicMetaMapUpdated[basicMetaMapUpdated$dataset.ID=="id36" & 
                                        basicMetaMapUpdated$type=="Case",]
    id36NormTab = id36NormTab[order(id36NormTab$sample.ID),]
    id36NormUniqSamples = id36NormTab$sample.ID[!duplicated(id36NormTab$metadata.ID)]
    normUniqSubjectList$US_id36 = id36NormUniqSamples
    
    ### id43-Thailand ###
    id43NormThaiTab = basicMetaMapUpdated[basicMetaMapUpdated$dataset.ID=="id43" & 
                                            basicMetaMapUpdated$type=="Control",]
    id43NormThaiTab = id43NormThaiTab[order(id43NormThaiTab$sample.ID),]
    id43NormThaiUniqSamples = id43NormThaiTab$sample.ID[!duplicated(id43NormThaiTab$metadata.ID)]
    normUniqSubjectList$Thailand_id43 = id43NormThaiUniqSamples
    
    ### id43-US ###
    id43NormUsTab = basicMetaMapUpdated[basicMetaMapUpdated$dataset.ID=="id43" & 
                                          basicMetaMapUpdated$type=="Case" &
                                          basicMetaMapUpdated$subtype=="Control",]
    id43NormUsTab = id43NormUsTab[order(id43NormUsTab$sample.ID),]
    id43NormUsUniqSamples = id43NormUsTab$sample.ID[!duplicated(id43NormUsTab$metadata.ID)]
    normUniqSubjectList$US_id43 = id43NormUsUniqSamples
    
    ### id49 ###
    id49NormTab = basicMetaMapUpdated[basicMetaMapUpdated$dataset.ID=="id49" & 
                                        basicMetaMapUpdated$type=="Case",]
    id49NormTab = id49NormTab[order(id49NormTab$sample.ID),]
    id49NormUniqSamples = id49NormTab$sample.ID[!duplicated(id49NormTab$metadata.ID)]
    normUniqSubjectList$Fiji_id49 = id49NormUniqSamples
    
    
    ### id8 ###
    id8NormTab = basicMetaMapUpdated[basicMetaMapUpdated$dataset.ID=="id8" & 
                                       basicMetaMapUpdated$type=="Control",]
    id8NormTab = id8NormTab[order(id8NormTab$sample.ID),]
    id8NormUniqSamples = id8NormTab$sample.ID[!duplicated(id8NormTab$metadata.ID)]
    normUniqSubjectList$Germany_id8 = id8NormUniqSamples
    normUniqAll = unique(do.call(c, normUniqSubjectList))
    
    normUniqCountryList = list()
    normUniqCountryList$China = c(normUniqSubjectList$China_id10,
                                  normUniqSubjectList$China_id12,
                                  normUniqSubjectList$China_id2,
                                  normUniqSubjectList$China_id20,
                                  normUniqSubjectList$China_id31,
                                  normUniqSubjectList$China_id52,
                                  normUniqSubjectList$China_id6,
                                  normUniqSubjectList$China_id9)
    normUniqCountryList$Denmark = c(normUniqSubjectList$Denmark_id16,
                                    normUniqSubjectList$Denmark_id25)
    normUniqCountryList$Fiji = normUniqSubjectList$Fiji_id49
    normUniqCountryList$Finland = normUniqSubjectList$Finland_id21
    normUniqCountryList$Germany = c(normUniqSubjectList$Germany_id44,
                                    normUniqSubjectList$Germany_id8)
    normUniqCountryList$India = normUniqSubjectList$India_id50
    normUniqCountryList$Italy = normUniqSubjectList$Italy_id46
    normUniqCountryList$Japan = normUniqSubjectList$Japan_id45
    normUniqCountryList$Luxembourg = normUniqSubjectList$Luxembourg_id35
    normUniqCountryList$Madagascar = normUniqSubjectList$Madagascar_id42
    normUniqCountryList$Mongo = normUniqSubjectList$Mongo_id33
    normUniqCountryList$Peru = normUniqSubjectList$Peru_id32
    normUniqCountryList$Spain = normUniqSubjectList$Spain_id17
    normUniqCountryList$Sweden = c(normUniqSubjectList$Sweden_id1,
                                   normUniqSubjectList$Sweden_id14,
                                   normUniqSubjectList$Sweden_id28)
    normUniqCountryList$Tanzania = normUniqSubjectList$Tanzania_id37
    normUniqCountryList$Thailand = normUniqSubjectList$Thailand_id43
    normUniqCountryList$UK = c(normUniqSubjectList$UK_id29, 
                               normUniqSubjectList$UK_id39)
    normUniqCountryList$US = c(normUniqSubjectList$US_id11,
                               normUniqSubjectList$US_id34,
                               normUniqSubjectList$US_id36,
                               normUniqSubjectList$US_id43)
    
    europeanCountries = c("Germany","Luxembourg","Spain","Sweden","Denmark", "Italy","UK")
    industrializedCountries = c("China","Denmark","Finland","Germany",
                                "Japan","Luxembourg","Spain","Sweden",
                                "Italy", #updated at 2019-09-26
                                "UK","US")
    westernCountries = c("Denmark","Finland","Germany",
                         "Luxembourg","Spain","Sweden",
                         "Italy", "UK","US")
    traditionalCountries = c("Thailand","Tanzania","Peru",
                             "Mongo","Madagascar","India","Fiji")
    
    industrializedSamples = do.call(c, normUniqCountryList[industrializedCountries])
    traditionalSamples = do.call(c, normUniqCountryList[traditionalCountries])
    westernSamples = do.call(c, normUniqCountryList[westernCountries])
    europeanSamples = do.call(c, normUniqCountryList[europeanCountries])
    
    
    
    
  }
  
  mixedMode = T
  if (mixedMode) {
    mixedUniqSubjectList = c(normUniqSubjectList, 
                             diseaseUniqSubjectList)
    mixedUniqAll = unique(do.call(c,mixedUniqSubjectList))
    
    mixedUniqCountryList = c(normUniqCountryList, 
                             diseaseUniqSubjectList)
    
    
    newEtfCountriesSamples = c(normUniqCountryList$Denmark,
                               normUniqCountryList$Finland,
                               normUniqCountryList$Germany,
                               normUniqCountryList$Italy,
                               normUniqCountryList$Luxembourg,
                               normUniqCountryList$Spain, 
                               normUniqCountryList$Sweden,
                               normUniqCountryList$UK)
    newEtbCountriesSamples = c(normUniqCountryList$China,
                               normUniqCountryList$US,
                               normUniqCountryList$Japan)
    newEtpCountriesSamples = c(normUniqCountryList$Fiji,
                               normUniqCountryList$India,
                               normUniqCountryList$Madagascar,
                               normUniqCountryList$Mongo,
                               normUniqCountryList$Peru,
                               normUniqCountryList$Tanzania,
                               normUniqCountryList$Thailand)
    
    etCountryList = list(etFirmicutes = newEtfCountriesSamples,
                         etBacteroides = newEtbCountriesSamples,
                         etPrevotella = newEtpCountriesSamples)
    
    
    mixedUniqCountryForMetaList = c(mixedUniqCountryList, 
                                    etCountryList)
    
  }
  
  perturbMode = T
  if (perturbMode) {
    
  }
  
  if (F) {
    View(basicMetaMapUpdated) 
    cor.test(mergeMatUpdated["msp_0058",],basicMetaMapUpdated$BMI, use="pairwise.complete.obs", method="spearman")
    cor.test(mergeMatUpdated["msp_0213",],basicMetaMapUpdated$BMI, use="pairwise.complete.obs", method="spearman")
    cor.test(mergeMatUpdated["msp_0009",],basicMetaMapUpdated$BMI, use="pairwise.complete.obs", method="spearman")
    cor.test(mergeMatUpdated["msp_0206",],basicMetaMapUpdated$BMI, use="pairwise.complete.obs", method="spearman")
    cor.test(mergeMatUpdated["msp_0924",],basicMetaMapUpdated$BMI, use="pairwise.complete.obs", method="spearman")
  }
  
  
}

checkSamplesUsedForMGWAS = T
if (checkSamplesUsedForMGWAS) {
  exItalySamples = c("SRR1930253", "SRR1930255", "SRR1931170", "SRR1931173")
  mixedUniqAll2 = mixedUniqAll[!mixedUniqAll %in% exItalySamples]
  
  usedBasicMetaMap = basicMetaMapUpdated[basicMetaMapUpdated$sample.ID %in% mixedUniqAll2,]
}

checkFuncAnnotNum = F
if (checkFuncAnnotNum) {
  
  loadFuncMat = T
  if (loadFuncMat) {
    funcMatRData = "j://Deposit/Project/2018_microbiome_atlas/functional.annotation/funcMat.20190808.RData"
    load(funcMatRData)
    funcMat = funcMat[!rownames(funcMat) %in% suppressMsps2, ]
    
    funcCounts = colSums(funcMat)
    
    table(names(funcCounts) %in% koTerms)
    table(names(funcCounts) %in% pfamTerms)
    table(names(funcCounts) %in% antismashTerms)
    table(names(funcCounts) %in% arTerms)
    table(names(funcCounts) %in% cazyTerms)
    table(names(funcCounts) %in% vfTerms)
    table(names(funcCounts) %in% jgiTerms)
    
    
    
    
  }
  
  loadAnnots = T
  if (loadAnnots) {koTerms # seven types of functional/phenotypic annotations
    
    ### virulence factors ###
    vfMatRData = "j://Deposit/Project/2018_microbiome_atlas/functional.annotation/PATRIC/vfBestMat.RData"
    load(vfMatRData) #vfMatDigit
    vfTerms = colnames(vfMatDigit)
    patricVfDescTabFile = "J://Deposit/Project/2018_microbiome_atlas/functional.annotation/PATRIC/PATRIC_sp_gene.txt"
    patricVfDescTab = read.table(patricVfDescTabFile, header=T, sep="\t", stringsAsFactors = F)
    
    ### JGI phenotype ###
    jgiMatRData = "j://Deposit/Project/2018_microbiome_atlas/functional.annotation/jgiMat.RData"
    load(jgiMatRData)
    jgiTerms = colnames(jgiMat)
    
    ### reactions ###
    gemRData = "J://Deposit/Project/2018_microbiome_atlas//atlas.GEM.model.related/GEM.from.Reza.amazing/absentPresentInModels.RData"
    load(gemRData)
    reactionTerms = colnames(gemMat)
    
    ### antibiotic resistance ###
    mustardMatRData = "j://Deposit/Project/2018_microbiome_atlas/functional.annotation/mustard.mat.RData"
    load(mustardMatRData)
    arTerms = colnames(mustardMat)
    
    ### antismash - secondary metabolism ###
    antismashMatRData = "J://Deposit/Project/2018_microbiome_atlas/functional.annotation/msp1992.igc2.antismashMat.RData"
    load(antismashMatRData)
    dim(antismashMat)
    antismashMat[1:3,1:3]
    antismashMelt = melt(antismashMat)
    head(antismashMelt)
    antismashMelt = antismashMelt[antismashMelt$value!=0,]
    dim(antismashMelt)
    heatmap.2(antismashMat, trace="none")
    
    writeMode = F
    if (writeMode) write.table(antismashMelt, "antismash.list.txt", sep="\t")
    antismashTerms = colnames(antismashMat)
    
    
    
    
    
    ### Cazy  ### 
    newGutCazyMatRData = "J://Deposit/Project/2018_microbiome_atlas/functional.annotation/CAZY.INRA/igc2.new.cazy.mat.RData"
    load(newGutCazyMatRData)
    colnames(cazyMat) = paste("cazy", colnames(cazyMat), sep=".")
    cazyTerms = colnames(cazyMat)
    
    ### PFAM ###
    igc2PfamMatRData = "j://Deposit/Project/2018_microbiome_atlas/functional.annotation/PFAM/igc2PfamMat.RData"
    load(igc2PfamMatRData)
    pfamTerms = colnames(pfamMat)
    
    ### KEGG ###
    gutKoBestMatRData = "j://Deposit/Project/2018_microbiome_atlas/functional.annotation/gutKoBestMat.1990.RData"
    load(gutKoBestMatRData)
    koTerms = colnames(gutKoBestMat)
    
    ### KEGG pathway ###
    # already loaded from above scripts
    # data: bacteriaKoElementList
    
    
    
    
  }
  
}

loadOverallInfos = T
if (loadOverallInfos) {
  
  allSamplesByCountries = split(basicMetaMapUpdated$sample.ID,
                                basicMetaMapUpdated$Geography)
  
  caseTab = basicMetaMapUpdated[basicMetaMapUpdated$type=="Case",]
  contTab = basicMetaMapUpdated[basicMetaMapUpdated$type=="Control",]
  
  thaiLong = basicMetaMapUpdated$sample.ID[basicMetaMapUpdated$dataset.ID=="id43" & basicMetaMapUpdated$subtype=="Hmong1st"]
  thaiShort = basicMetaMapUpdated$sample.ID[basicMetaMapUpdated$dataset.ID=="id43" & basicMetaMapUpdated$subtype=="Karen1st"]
  usLocal = basicMetaMapUpdated$sample.ID[basicMetaMapUpdated$dataset.ID=="id43" & basicMetaMapUpdated$subtype=="Control"]
  
  caseList = split(caseTab$sample.ID, paste("Case",caseTab$dataset.ID, caseTab$Geography, sep="."))
  contList = split(contTab$sample.ID, paste("Cont", contTab$dataset.ID, contTab$Geography, sep="."))
  caseContList = c(caseList, contList)
  
  newDiseaseSets = c("Case.id1.Sweden","Case.id10.China",
                     "Case.id11.US","Case.id12.China",
                     "Case.id14.Sweden","Case.id16.Denmark",
                     "Case.id17.Spain","Case.id2.China",
                     "Case.id20.China","Case.id21.Finland",
                     "Case.id25.Denmark","Case.id26.France",
                     "Case.id27.UK","Case.id3.US",
                     "Case.id31.China","Case.id34.US","Case.id35.Luxembourg",
                     "Case.id40.Italy","Case.id40.Spain",
                     "Case.id41.France","Case.id44.Germany",
                     "Case.id45.Japan", "Case.id46.Italy",
                     "Case.id6.China", "Case.id7.US",
                     "Case.id8.Germany", "Case.id9.China",
                     "Case.id52.China", "Case.id53.Spain")  
  
  names(newDiseaseSets) = c("Atherosclerosis_SE",
                            "CVD_CN","CRC_US",
                            "GDM_CN","T2D_SE",
                            "Obesity_DK_16","IBD_ES","T2D_CN",
                            "Cirrhosis_CN","T1D_FI",
                            "Obesity_DK_25","NSCLC_FR",
                            "Cirrhosis_GB","Melanoma_US",
                            "Rheumatoid_CN", "CFS_US","T1D_LU",
                            "NAFLD_IT","NAFLD_ES",
                            "RCC_FR","CRC_DE", "CRC_JP",
                            "CRC_IT", "Crohn_CN",
                            "NAFLD_US","Parkinson_DE",
                            "Ankylosing_CN",
                            "Behcet_CN","T2D_ES") 
  
  newDiseaseList = caseContList[match(newDiseaseSets,names(caseContList))]
  names(newDiseaseList) = names(newDiseaseSets)
  
  newDiseaseIds = unique(do.call(c, newDiseaseList))
  
  
  newNormalSets = c("Case.id28.Sweden", "Cont.id1.Sweden",
                    "Cont.id14.Sweden",
                    "Cont.id16.Denmark","Cont.id25.Denmark",
                    "Cont.id21.Finland",
                    "Cont.id35.Luxembourg",
                    "Case.id29.UK","Case.id39.UK", 
                    "Case.id36.US","Cont.id11.US","Cont.id34.US",
                    "Cont.id2.China","Cont.id6.China",
                    "Cont.id9.China","Cont.id10.China",
                    "Cont.id12.China","Cont.id20.China",
                    "Cont.id31.China",
                    "Cont.id45.Japan",
                    "Cont.id8.Germany","Cont.id44.Germany",
                    "Cont.id46.Italy","Cont.id37.Italy",
                    "Cont.id17.Spain", "Cont.id52.China", "Cont.id32.US", "Case.id43.US")
  
  names(newNormalSets)=c("Sweden_id28", "Sweden_id1",
                         "Sweden_id14",
                         "Denmark_id16","Denmark_id25",
                         "Finland_id21",
                         "Luxembourg_id35",
                         "UK_id29", "UK_id39", 
                         "US_id36", "US_id11","US_id34",
                         "China_id2","China_id6",
                         "China_id9","China_id10",
                         "China_id12","China_id20",
                         "China_id31",
                         "Japan_id45",
                         "Germany_id8","Germany_id44",
                         "Italy_id46","Italy_id37",
                         "Spain_id17", "China_id52", "US_id32", "US_id43")
  
  newNormalList = caseContList[match(newNormalSets,names(caseContList))]
  names(newNormalList) = names(newNormalSets)
  
  #newNormalList$US_id43 = usLocal
  
  newNormalIds = unique(do.call(c, newNormalList))
  
  ### traditional populations ###
  
  newNonWesternSets = c("Case.id32.Peru", "Case.id33.Mongo",
                        "Cont.id43.Thai","Case.id49.Fiji",
                        "Case.id50.India","Case.id37.Tanzania",
                        "Case.id42.Madagascar")
  
  names(newNonWesternSets) = c("Peru_id32", "Mongo_id33", 
                               "Thai_id43", "Fiji_id49", 
                               "India_id50","Tanzania_id37",
                               "Madagascar_id42")
  
  newNonWesternList = caseContList[match(newNonWesternSets,names(caseContList))]
  names(newNonWesternList) = names(newNonWesternSets)
  
  newNonWesternIds = unique(do.call(c, newNonWesternList))
  
  newPerturbCaseExistingSets = list()
  #metformin (metaHIT), immunotherapy (RCC, NSCLC, Melanoma), Longitudinal changes (wellness)
  
  newPerturbCaseSets = c("Case.id51.Denmark","Case.id48.US", "Case.id38.US")
  names(newPerturbCaseSets) = c("antibiotics_DK","foodborne_US", "Preterm_US")
  
  newPerturbCaseList = caseContList[match(newPerturbCaseSets, names(caseContList))]
  names(newPerturbCaseList) = names(newPerturbCaseSets)
  
  newPerturbCaseList$LongImmigrant = thaiLong
  newPerturbCaseList$ShortImmigrant = thaiShort
  
  newNormalAllList = c(newNormalList,newNonWesternList)
  newNormalAllIds = unique(do.call(c, newNormalAllList))
  newNormalAllIds_wo_FI = unique(do.call(c, newNormalAllList[names(newNormalAllList) != "Finland_id21"]))
  
  ###
  basicMetaMapNormal = basicMetaMapUpdated[basicMetaMapUpdated$sample.ID %in% newNormalAllIds,]
  
  normalSamplesByCountries = split(basicMetaMapNormal$sample.ID,
                                   basicMetaMapNormal$Geography)
  
  ###
  
  newWesternIds = newNormalAllIds[!newNormalAllIds %in% newNonWesternIds]
  
  
  
  dataListByGeo = list(China=c("China_id2", "China_id6", "China_id10", "China_id12", "China_id20", "China_id31", "China_id2", "China_id52" ), 
                       Japan="Japan_id45",
                       US=c("US_id43", "US_id36","US_id11","US_id34"),
                       UK=c("UK_id29","UK_id39"),
                       Germany=c("Germany_id8","Germany_id44"),
                       Italy=c("Italy_id46","Italy_id37"),
                       Spain="Spain_id17",
                       Sweden=c("Sweden_id28","Sweden_id1","Sweden_id14"),
                       Denmark=c("Denmark_id16","Denmark_id25"),
                       Luxembourg="Luxembourg_id35",
                       Peru="Peru_id32",
                       Mongo="Mongo_id33",
                       Thai="Thai_id43",
                       Fiji="Fiji_id49",
                       India="India_id50",
                       Tanzania="Tanzania_id37",
                       Madagascar="Madagascar_id42")
  
  
  newGeoList = lapply(dataListByGeo, 
                      function(currDatasets) {
                        currAllSamples = do.call(c,newNormalAllList[currDatasets])
                        names(currAllSamples) = NULL
                        return(currAllSamples)
                      })
  
  newContByEnteroList = list(ETB=c(newGeoList$China, 
                                   newGeoList$Japan,
                                   newGeoList$US),
                             ETF=c(newGeoList$UK,
                                   newGeoList$Germany,
                                   newGeoList$Italy,
                                   newGeoList$Spain,
                                   newGeoList$Sweden,
                                   newGeoList$Denmark,
                                   newGeoList$Luxembourg),
                             ETP=c(newGeoList$Madagascar,
                                   newGeoList$Tanzania,
                                   newGeoList$India,
                                   newGeoList$Fiji,
                                   newGeoList$Thai,
                                   newGeoList$Mongo,
                                   newGeoList$Peru))
  
  newCaseByEnteroList = list(ETB=c(newDiseaseList$CVD_CN,
                                   newDiseaseList$GDM_CN,
                                   newDiseaseList$T2D_CN,
                                   newDiseaseList$Cirrhosis_CN,
                                   newDiseaseList$Rheumatoid_CN,
                                   newDiseaseList$Crohn_CN,
                                   newDiseaseList$Ankylosing_CN,
                                   newDiseaseList$Preterm_US,
                                   newDiseaseList$CRC_US,
                                   newDiseaseList$Melanoma_US,
                                   newDiseaseList$CFS_US,
                                   newDiseaseList$NAFLD_US,
                                   newDiseaseList$CRC_JP, 
                                   newDiseaseList$Behcet_CN),
                             ETF=c(newDiseaseList$Atherosclerosis_SE,
                                   newDiseaseList$T2D_SE,
                                   newDiseaseList$Obesity_DK_16,
                                   newDiseaseList$Obesity_DK_25,
                                   newDiseaseList$IBD_ES,
                                   newDiseaseList$NAFLD_ES,
                                   newDiseaseList$T1D_LU,
                                   newDiseaseList$Cirrhosis_GB,
                                   newDiseaseList$NAFLD_IT,
                                   newDiseaseList$CRC_IT,
                                   newDiseaseList$CRC_DE,
                                   newDiseaseList$Parkinson_DE))
  
  # newDiseaseList$NSCLC_FR
  # newDiseaseList$RCC_FR
  # newDiseaseList$T1D_FI
  
  
  diseaseIdTab = data.frame(sample=newDiseaseIds, type="disease")
  normalIdTab = data.frame(sample=newNormalAllIds, type="normal")
  allIdTab = rbind(diseaseIdTab, 
                   normalIdTab)
  
  
  write.table(allIdTab, "C://Data//comparative.analysis.healthy.sweden//disease.normal.IDs.for.reza.2019.11.26.txt", sep="\t")
  
}

checkInflowOutflowEnrichment = F
if (checkInflowOutflowEnrichment) {
  markovStatTabFinalFile = "C://Data//comparative.analysis.healthy.sweden//markovStatTab.final.3.txt"
  markovStatTabFinal = read.table(markovStatTabFinalFile, sep="\t", stringsAsFactors = F)
  
  table(markovStatTabFinal$inflow>=0.333)
  table(markovStatTabFinal$outflow>=0.333)
  
  inflowSpecies = rownames(markovStatTabFinal[markovStatTabFinal$inflow>=0.5,])
  outflowSpecies = rownames(markovStatTabFinal[markovStatTabFinal$outflow>=0.5,])
  
  table(inflowSpecies %in% outflowSpecies)
  ###
  mergeScaleMat =  t(scale(t(mergeMatUpdated)))
  inflowSignatureSum = colSums(mergeScaleMat[inflowSpecies,])/sqrt(length(inflowSpecies))
  outflowSignatureSum = colSums(mergeScaleMat[outflowSpecies,])/sqrt(length(outflowSpecies))
  #totalSignatures = outflowSignatureSum - inflowSignatureSum
  totalSignatures = outflowSignatureSum
  
  basicMetaMapUpdated$inflowOutflow=totalSignatures
  
  lowCut=as.numeric(quantile(totalSignatures, 0.1))
  highCut=as.numeric(quantile(totalSignatures, 0.9))
  
  hqDataTab$inflowOutflow = basicMetaMapUpdated$inflowOutflow[match(hqDataTab$sample.ID, basicMetaMapUpdated$sample.ID)]
  hqDataTab$inflowType = "NA"
  hqDataTab$inflowType[hqDataTab$inflowOutflow >= highCut] = "moreOutflow"
  hqDataTab$inflowType[hqDataTab$inflowOutflow <= lowCut] = "moreInflow"
  
  hqContTab = hqDataTab[hqDataTab$type=="Control",]
  hqCaseTab = hqDataTab[hqDataTab$type=="Case",]
  
  # hqContTab$GeneRichness = basicMetaMapUpdated$GeneRichness[match(hqContTab$sample.ID, basicMetaMapUpdated$sample.ID)]
  # hqContTab$enteroType = basicMetaMapUpdated$enteroType[match(hqContTab$sample.ID, basicMetaMapUpdated$sample.ID)]
  hqContTab_SW = hqContTab[hqContTab$Geography=="Sweden",]
  hqContTab_CN = hqContTab[hqContTab$Geography=="China",]
  
  checkFeatures = c("FastingGlucose", "HDL", "LDL", "TG", "Cholesterol", "ALT", "CRP", "Creatinine")
  for(feature in checkFeatures) {
    print(feature)
    #feature = "HDL"
    
    currTab = hqDataTab[!is.na(hqDataTab[,feature]),]
    print(table(currTab$inflowType))
    
    hgFeatures = currTab[currTab$inflowType == "moreOutflow",feature]
    lgFeatures = currTab[currTab$inflowType == "moreInflow",feature] 
    
    hgGenders = currTab$Gender[currTab$inflowType == "moreOutflow"]
    lgGenders = currTab$Gender[currTab$inflowType == "moreInflow"] 
    table(hgGenders)
    table(lgGenders) 
    
    
    print(table(hgFeatures<1.03)/length(hgFeatures))
    print(table(lgFeatures<1.03)/length(lgFeatures))
    
    barplot(cbind(table(hgFeatures<1.03)/length(hgFeatures),
                  table(lgFeatures<1.03)/length(lgFeatures)),
            col=c(eurCol, cjpCol))
    
    featureList = list(outflow=hgFeatures,
                       inflow=lgFeatures)
    
    boxplot(featureList, las=2, main=feature)
    print(wilcox.test(hgFeatures, lgFeatures)) 
    
    print(mean(hgFeatures) / mean(lgFeatures)) 
    
    ###
    hgMaleCheck <- hgFeatures>110 & hgGenders == "Male"
    lgMaleCheck <- lgFeatures>110 & lgGenders == "Male"
    
    hgFemaleCheck <- hgFeatures>90 & hgGenders == "Female"
    lgFemaleCheck <- lgFeatures>90 & lgGenders == "Female"
    
    hgCheck = hgMaleCheck | hgFemaleCheck
    lgCheck = lgMaleCheck | lgFemaleCheck
    
    
  }
  
  
}

checkHqMetadata = T
if (checkHqMetadata) {
  
  hAllCutNumber = 727622.5
  lAllCutNumber = 441994.5
  
  hqDataTab$GeneRichness = basicMetaMapUpdated$GeneRichness[match(hqDataTab$sample.ID, basicMetaMapUpdated$sample.ID)]
  hqDataTab$enteroType = basicMetaMapUpdated$enteroType[match(hqDataTab$sample.ID, basicMetaMapUpdated$sample.ID)]
  hqDataTab$richnessType = "NA"
  hqDataTab$richnessType[hqDataTab$GeneRichness >= hAllCutNumber] = "HG"
  hqDataTab$richnessType[hqDataTab$GeneRichness <= lAllCutNumber] = "LG"
  # hqDataTab$richnessType[hqDataTab$GeneRichness >= hHalfCutNumber] = "HG"
  # hqDataTab$richnessType[hqDataTab$GeneRichness < hHalfCutNumber] = "LG"
  
  hqContTab = hqDataTab[hqDataTab$type=="Control",]
  hqCaseTab = hqDataTab[hqDataTab$type=="Case",]
  
  # hqContTab$GeneRichness = basicMetaMapUpdated$GeneRichness[match(hqContTab$sample.ID, basicMetaMapUpdated$sample.ID)]
  # hqContTab$enteroType = basicMetaMapUpdated$enteroType[match(hqContTab$sample.ID, basicMetaMapUpdated$sample.ID)]
  hqContTab_SW = hqContTab[hqContTab$Geography=="Sweden",]
  hqContTab_CN = hqContTab[hqContTab$Geography=="China",]
  
}

checkRichnessWithHqMetadata = T
if (checkRichnessWithHqMetadata) {
  checkFeatures = c("FastingGlucose", "HDL", "LDL", "TG", "Cholesterol", "ALT", "CRP", "Creatinine")
  for(feature in checkFeatures) {
    print(feature)
    #feature = "HDL"
    
    currTab = hqDataTab[!is.na(hqDataTab[,feature]),]
    print(table(currTab$richnessType))
    
    hgFeatures = currTab[currTab$richnessType == "HG",feature]
    lgFeatures = currTab[currTab$richnessType == "LG",feature] 
    
    hgGenders = currTab$Gender[currTab$richnessType == "HG"]
    lgGenders = currTab$Gender[currTab$richnessType == "LG"] 
    table(hgGenders)
    table(lgGenders) 
    
    
    print(table(hgFeatures<1.03)/length(hgFeatures))
    print(table(lgFeatures<1.03)/length(lgFeatures))
    
    barplot(cbind(table(hgFeatures<1.03)/length(hgFeatures),
                  table(lgFeatures<1.03)/length(lgFeatures)),
            col=c(eurCol, cjpCol))
    
    featureList = list(HG=hgFeatures,
                       LG=lgFeatures)
    
    boxplot(featureList, las=2, main=feature)
    print(wilcox.test(hgFeatures, lgFeatures)) 
    
    print(mean(hgFeatures) / mean(lgFeatures)) 
    
    ###
    hgMaleCheck <- hgFeatures>110 & hgGenders == "Male"
    lgMaleCheck <- lgFeatures>110 & lgGenders == "Male"
    
    hgFemaleCheck <- hgFeatures>90 & hgGenders == "Female"
    lgFemaleCheck <- lgFeatures>90 & lgGenders == "Female"
    
    hgCheck = hgMaleCheck | hgFemaleCheck
    lgCheck = lgMaleCheck | lgFemaleCheck
    
    
  }
  
}

checkEnteroWithHqMetada = T
if (checkEnteroWithHqMetada) {
  
  ##chinese control mode ##
  checkFeatures = c("FastingGlucose", "HDL", "LDL", "TG", "Cholesterol", "ALT", "Creatinine")
  for(feature in checkFeatures) {
    print(feature)
    feature = "TG"
    
    currTab = hqContTab_CN[!is.na(hqContTab_CN[,feature]),]
    print(table(currTab$enteroType))
    
    etbFeatures = currTab[currTab$enteroType == "ET-Bacteroides",feature]
    etfFeatures = currTab[currTab$enteroType == "ET-Firmicutes",feature]
    etpFeatures = currTab[currTab$enteroType == "ET-Prevotella",feature]
    
    etbGenders = currTab$Gender[currTab$enteroType == "ET-Bacteroides"]
    etfGenders = currTab$Gender[currTab$enteroType == "ET-Firmicutes"]
    etpGenders = currTab$Gender[currTab$enteroType == "ET-Prevotella"]
    
    table(etbGenders)
    table(etfGenders)
    table(etpGenders)
    
    featureList = list(ETB=etbFeatures,
                       ETF=etfFeatures,
                       ETP=etpFeatures)
    
    table(etbFeatures>1.7)/length(etbFeatures)
    table(etfFeatures>1.7)/length(etfFeatures)
    table(etpFeatures>1.7)/length(etpFeatures)
    
    boxplot(featureList, las=2)
    print(t.test(etbFeatures, etfFeatures))
    print(t.test(etbFeatures, etpFeatures))
    
    print(mean(etbFeatures) / mean(etfFeatures))
    print(mean(etbFeatures) / mean(etpFeatures)) 
    
    
    
  }
  ###Creatinine --> B >> P
  ###TG ---> B >> F
  
  ##swedish control mode##
  checkFeatures = c("FastingGlucose", "HDL", "LDL", "TG", "Cholesterol", "CRP", "Creatinine")
  for(feature in checkFeatures) {
    print(feature)
    #feature = "TG"
    
    currTab = hqContTab_SW[!is.na(hqContTab_SW[,feature]),]
    print(table(currTab$enteroType))
    
    etbFeatures = currTab[currTab$enteroType == "ET-Bacteroides",feature]
    etfFeatures = currTab[currTab$enteroType == "ET-Firmicutes",feature]
    etpFeatures = currTab[currTab$enteroType == "ET-Prevotella",feature]
    
    
    
    featureList = list(ETB=etbFeatures,
                       ETF=etfFeatures,
                       ETP=etpFeatures)
    
    table(etbFeatures>1.7)/length(etbFeatures)
    table(etfFeatures>1.7)/length(etfFeatures)
    table(etpFeatures>1.7)/length(etpFeatures)
    
    boxplot(featureList, las=2)
    print(t.test(etbFeatures, etfFeatures))
    #print(t.test(etbFeatures, etpFeatures))
    
    print(mean(etbFeatures) / mean(etfFeatures))
    print(mean(etbFeatures) / mean(etpFeatures)) 
    
    
    
  }
  
  ##all control mode##
  checkFeatures = c("FastingGlucose", "HDL", "LDL", "TG", "Cholesterol", "Creatinine")
  for(feature in checkFeatures) {
    print(feature)
    #feature = "TG"
    
    currTab = hqContTab[!is.na(hqContTab[,feature]),]
    print(table(currTab$enteroType))
    
    etbFeatures = currTab[currTab$enteroType == "ET-Bacteroides",feature]
    etfFeatures = currTab[currTab$enteroType == "ET-Firmicutes",feature]
    etpFeatures = currTab[currTab$enteroType == "ET-Prevotella",feature]
    
    
    
    featureList = list(ETB=etbFeatures,
                       ETF=etfFeatures,
                       ETP=etpFeatures)
    
    table(etbFeatures>1.7)/length(etbFeatures)
    table(etfFeatures>1.7)/length(etfFeatures)
    table(etpFeatures>1.7)/length(etpFeatures)
    
    boxplot(featureList, las=2)
    print(t.test(etbFeatures, etfFeatures))
    #print(t.test(etbFeatures, etpFeatures))
    
    print(mean(etbFeatures) / mean(etfFeatures))
    print(mean(etbFeatures) / mean(etpFeatures)) 
    
    
    
  }
  
  ##all mode##
  checkFeatures = c("FastingGlucose", "HDL", "LDL", "TG", "Cholesterol", "Creatinine")
  checkFeatures = c("HDL", "TG", "Creatinine")
  for(feature in checkFeatures) {
    print(feature)
    #feature = "TG"
    
    currTab = hqDataTab[!is.na(hqDataTab[,feature]),]
    print(table(currTab$enteroType))
    
    etbFeatures = currTab[currTab$enteroType == "ET-Bacteroides",feature]
    etfFeatures = currTab[currTab$enteroType == "ET-Firmicutes",feature]
    etpFeatures = currTab[currTab$enteroType == "ET-Prevotella",feature]
    
    etbGenders = currTab$Gender[currTab$enteroType == "ET-Bacteroides"]
    etfGenders = currTab$Gender[currTab$enteroType == "ET-Firmicutes"]
    etpGenders = currTab$Gender[currTab$enteroType == "ET-Prevotella"]
    
    print(table(etbGenders))
    print(table(etfGenders))
    print(table(etpGenders))
    
    featureList = list(ETB=etbFeatures,
                       ETF=etfFeatures,
                       ETP=etpFeatures)
    
    table(etbFeatures>1.7)/length(etbFeatures)
    table(etfFeatures>1.7)/length(etfFeatures)
    table(etpFeatures>1.7)/length(etpFeatures)
    
    boxplot(featureList, las=2, main=feature)
    print(t.test(etbFeatures, etfFeatures))
    #
    print(t.test(etbFeatures, etpFeatures))
    
    print(mean(etbFeatures) / mean(etfFeatures))
    print(mean(etbFeatures) / mean(etpFeatures)) 
    
    
    
  }
  
  ### HDL, TG, Creatinine 
}

checkDistributions = T
if (checkDistributions) {
  require(fitdistrplus)
  require(logspline)
  
  mergeMatScaled = mergeMatUpdated*10^9
  mergeMatScaledLog = log2(mergeMatScaled)
  mergeMatScaledLog[mergeMatScaledLog == -Inf] = 0
  
  mergeMatScaledNonZero = mergeMatScaled[mergeMatScaled!=0]
  mergeMatScaledNonZero = log2(mergeMatScaledNonZero)
  
  qCutOff =(2^(4.042488))/10^9
  min(mergeMatUpdated[mergeMatUpdated!=0])
  
  mergeMatUpdatedImputed = mergeMatUpdated
  mergeMatUpdatedImputed[mergeMatUpdatedImputed<qCutOff] <- 0
  
  
  if (F) {
    fit.lnorm = fitdist(mergeMatScaledNonZero, "lnorm")
    plot(fit.lnorm)
    fit.lnorm$estimate
    #meanlog: 2.1114151
    #sdlog: 0.2811768
    fit.gamma = fitdist(mergeMatScaledNonZero, "gamma")
    plot(fit.gamma)
    fit.gamma$estimate
    #save(fit.gamma, file="C://Data//comparative.analysis.healthy.sweden//fit.gamma.RData")
    #shape: 13.107296
    #rate: 1.526737 
    
    q01 = qgamma(0.01, fit.gamma$estimate["shape"], fit.gamma$estimate["rate"])
    #4.042488
    
    
    q99 = qgamma(0.99, fit.gamma$estimate["shape"], fit.gamma$estimate["rate"])
    #15.04048
    
    mergeMatScaledCheckOutlier <- (mergeMatScaledLog > q01)*1
    
    mergeMatScaled = mergeMatUpdated*10^9
    mergeMatScaledLog = log2(mergeMatScaled)
    mergeMatScaledCheckOutlier[mergeMatScaledLog == -Inf] <- NA
    
    apply(mergeMatScaledCheckOutlier, 1, function(x) {
      table(x[!is.na(x)])
    })
    
  }
  
  ### check individuals by lowly abundant species
  ###
  
  
}

generateNewCohortIds  = T
if (generateNewCohortIds) {
  
  
}

convertTaxaToRadialDendrogram = F
if (convertTaxaToRadialDendrogram) {
  
  #superkingdom
  #phylum
  #class
  #order
  #family
  #genus
  #species
  getOutputFrom <- function(inputTaxo, seqs) {
    inputTaxoTab = unique(inputTaxo[,seqs])
    output = apply(inputTaxoTab,1, function(x) paste(x,collapse="."))
    finalOutput = apply(data.frame(str=output, msp=""), 1, function(x) paste(x,collapse=","))
    names(finalOutput)=NULL
    return(finalOutput)
  }
  getSpeciesOutputFrom <- function(inputTaxo, seqs) {
    inputTaxoTab = unique(inputTaxo[,seqs])
    msps = rownames(inputTaxoTab)
    output = apply(inputTaxoTab,1, function(x) paste(x,collapse="."))
    finalOutput = apply(data.frame(str=output, msp=msps), 1, function(x) paste(x,collapse=","))
    names(finalOutput)=NULL
    return(finalOutput)
  }
  
  getRadialDendroInput <- function(gutTaxo) {
    inputs = gutTaxo
    indUnclassified = grepl("unclassified", inputs$species)
    inputs$species[indUnclassified]=inputs$MSP[indUnclassified]
    inputs$root = "Life"
    
    #superkingdom
    seqs = c("root","superkingdom")
    output1 = getOutputFrom(inputs, seqs)
    
    #phylum
    seqs = c("root","superkingdom","phylum")
    output2 = getOutputFrom(inputs, seqs)
    
    #class
    seqs = c("root","superkingdom","phylum","class")
    output3 = getOutputFrom(inputs, seqs)
    
    #order
    seqs = c("root","superkingdom","phylum","class","order")
    output4 = getOutputFrom(inputs, seqs)
    
    #family
    seqs = c("root","superkingdom","phylum","class","order","family")
    output5 = getOutputFrom(inputs, seqs)
    
    #genus
    seqs = c("root","superkingdom","phylum","class","order","family","genus")
    output6 = getOutputFrom(inputs, seqs)
    
    #species
    seqs = c("root","superkingdom","phylum","class","order","family","genus","species")
    output7 = getSpeciesOutputFrom(inputs, seqs)
    
    outputFinal = c(output1,
                    output2,
                    output3,
                    output4,
                    output5,
                    output6) #output7
    
    View(outputFinal)
    #write.table(outputFinal, "J:\\Deposit\\Project\\2018_microbiome_atlas\\tree.of.life.input.txt", quote = F, row.names = F)
    write.table(outputFinal, "J:\\Deposit\\Project\\2018_microbiome_atlas\\tree.of.life.input.genus.end.csv", quote = F, row.names = F)
    
  }
  
  checkConsistency <- function(inputTaxo) {
    
    inputTaxo$phylum[inputTaxo$MSP == "msp_1346c"] = "unclassified Bacteria"
    inputTaxo$class[inputTaxo$MSP == "msp_1346c"] = "unclassified Bacteria"
    inputTaxo$order[inputTaxo$MSP == "msp_1346c"] = "unclassified Bacteria"
    inputTaxo$family[inputTaxo$MSP == "msp_1346c"] = "unclassified Bacteria"
    inputTaxo$genus[inputTaxo$MSP == "msp_1346c"] = "unclassified Bacteria"
    inputTaxo$species[inputTaxo$MSP == "msp_1346c"] = "unclassified Bacteria"
    
    inputTaxo$class[inputTaxo$family == "Peptostreptococcaceae"] = "Clostridia"
    inputTaxo$order[inputTaxo$family == "Peptostreptococcaceae"] = "Clostridiales"
    
    inputTaxo$order[inputTaxo$MSP == "msp_0422"] = "Clostridiales"
    inputTaxo$class[inputTaxo$MSP == "msp_0422"] = "Clostridia"
    
    inputTaxo$family[inputTaxo$genus == "Flavonifractor"] = "unclassified Clostridiales 6"
    #inputTaxo$family[inputTaxo$MSP == "msp_0213"] = "Ruminococcaceae"
    
    inputTaxo$family[inputTaxo$MSP == "msp_0613"] = "Christensenellaceae-Catabacteriaceae"
    inputTaxo$family[inputTaxo$MSP == "msp_1652"] = "unclassified Clostridiales 6"
    
    inputTaxo$genus[inputTaxo$MSP == "msp_1152"] = "unclassified Sutterellaceae"
    
    inputTaxo$species[inputTaxo$MSP == "msp_1296c"] = "unclassified Lachnospiraceae"
    
    # check parents of phylum
    currTaxo = inputTaxo[,c("phylum","superkingdom")]
    currTaxo = unique(currTaxo)
    checkCorrespondence = split(currTaxo[,2], currTaxo[,1])
    checkNotSingleParent = unlist(lapply(checkCorrespondence, length))!=1
    found=names(checkCorrespondence)[checkNotSingleParent]
    print(currTaxo[currTaxo[,1]%in%found,])
    
    # check parents of class
    currTaxo = inputTaxo[,c("class","phylum")]
    currTaxo = unique(currTaxo)
    checkCorrespondence = split(currTaxo[,2], currTaxo[,1])
    checkNotSingleParent = unlist(lapply(checkCorrespondence, length))!=1
    found = names(checkCorrespondence)[checkNotSingleParent]
    print(currTaxo[currTaxo[,1]%in%found,])
    
    # check parents of family
    currTaxo = inputTaxo[,c("family","class")]
    currTaxo = unique(currTaxo)
    checkCorrespondence = split(currTaxo[,2], currTaxo[,1])
    checkNotSingleParent = unlist(lapply(checkCorrespondence, length))!=1
    found = names(checkCorrespondence)[checkNotSingleParent]
    print(currTaxo[currTaxo[,1]%in%found,])
    
    # check parents of genus
    currTaxo = inputTaxo[,c("genus","family")]
    currTaxo = unique(currTaxo)
    checkCorrespondence = split(currTaxo[,2], currTaxo[,1])
    checkNotSingleParent = unlist(lapply(checkCorrespondence, length))!=1
    found = names(checkCorrespondence)[checkNotSingleParent]
    print(currTaxo[currTaxo[,1]%in%found,])
    
    
    # check parents of species
    currTaxo = inputTaxo[,c("species","genus")]
    currTaxo = unique(currTaxo)
    checkCorrespondence = split(currTaxo[,2], currTaxo[,1])
    checkNotSingleParent = unlist(lapply(checkCorrespondence, length))!=1
    found = names(checkCorrespondence)[checkNotSingleParent]
    print(currTaxo[currTaxo[,1]%in%found,])
    
    
    
    
  }
  
  checkConsistency(gutTaxo)
  
  getRadialDendroInputForR <- function(gutTaxo) {
    inputs = gutTaxo
    indUnclassified = grepl("unclassified", inputs$species)
    inputs$species[indUnclassified]=inputs$MSP[indUnclassified]
    inputs$root = "Life"
    
    Root = list()
    Root$name = "Life"
    
    ### superkingdom
    inSeqs = unique(inputs[,"superkingdom"])
    Root$children <- unname(sapply(inSeqs, function(x) {
      names(x) = NULL
      newX = list()
      newX["name"] = x
      return(newX)
    }, simplify = F))
    
    #radialNetwork(List = Root, fontSize = 7, opacity = 0.9)
    
    
    ### phylum
    seqs = c("superkingdom","phylum")
    inSeqs = unique(inputs[,seqs])
    inSeqList = split(inSeqs[,2],inSeqs[,1])
    Attached = lapply(Root$children, function(x){ #Root$children = 
      print(x)
      
      currName = x$name
      matchedChildren = inSeqList[[currName]]
      print(matchedChildren)
      newChildren = sapply(matchedChildren, function(currChild){
        names(currChild)=NULL
        a <- list()
        a["name"] = currChild
        return(a)
      }, simplify = F)
      
      newX = list()
      newX["name"] = currName
      newX["children"] = newChildren
      
      return(newX)
      
    })
    
    radialNetwork(List = Root, fontSize = 7, opacity = 0.9)
    
    
    ######
    x=lapply(Root$children, function(children){ #Root$children = 
      
      newX = lapply(children, function(x) {
        
        print(x)
        
        currName = x$name
        
        matchedChildren = inSeqList[currName]
        
        
        newChildren = sapply(matchedChildren, function(currChild){
          a <- list()
          a$name = currChild
          return(a)
        })
        
        return(newChildren)
        
      })
      return(newX)
    })
    
    ######
    radialNetwork(List = Root$children[[2]], fontSize = 7, opacity = 0.9)
    
    
    
    
    #superkingdom
    seqs = c("root","superkingdom")
    output1 = getOutputFrom(inputs, seqs)
    
    #phylum
    seqs = c("root","superkingdom","phylum")
    output2 = getOutputFrom(inputs, seqs)
    
    #class
    seqs = c("root","superkingdom","phylum","class")
    output3 = getOutputFrom(inputs, seqs)
    
    #order
    seqs = c("root","superkingdom","phylum","class","order")
    output4 = getOutputFrom(inputs, seqs)
    
    #family
    seqs = c("root","superkingdom","phylum","class","order","family")
    output5 = getOutputFrom(inputs, seqs)
    
    #genus
    seqs = c("root","superkingdom","phylum","class","order","family","genus")
    output6 = getOutputFrom(inputs, seqs)
    
    #species
    seqs = c("root","superkingdom","phylum","class","order","family","genus","species")
    output7 = getOutputFrom(inputs, seqs)
    
    outputFinal = c(output1,
                    output2,
                    output3,
                    output4,
                    output5,
                    output6,
                    output7)
    
    
    
  }
  
  testCode = F
  if (testCode) {
    require(networkD3)
    URL <- paste0(
      "https://cdn.rawgit.com/christophergandrud/networkD3/",
      "master/JSONdata//flare.json")
    
    Flare <- jsonlite::fromJSON(URL, simplifyDataFrame = FALSE)
    Flare$children = Flare$children[1:3]
    
    radialNetwork(List = Flare, fontSize = 7, opacity = 0.9)
    
  }
  
  
}

saveWebAtlasRawFile = F
if (saveWebAtlasRawFile) {
  
  basicMetaMapUpdatedWeb = basicMetaMapUpdated
  mergeMatUpdatedWeb = mergeMatUpdated
  mergeMatUpdatedWeb = mergeMatUpdatedWeb[!rownames(mergeMatUpdatedWeb) %in% suppressMsps2,]
  
  
  ### id28 - wellness ###
  id28SampleIDs = basicMetaMapUpdatedWeb$sample.ID[basicMetaMapUpdatedWeb$dataset.ID=="id28"]
  
  id28MetaIDs = basicMetaMapUpdatedWeb$metadata.ID[basicMetaMapUpdatedWeb$dataset.ID=="id28"]
  id28MetaIDsUnique = unique(id28MetaIDs)
  
  table(duplicated(id28SampleIDs))
  table(duplicated(id28MetaIDs))
  
  ### id29 - kcl twin ###
  id29SampleIDs = basicMetaMapUpdatedWeb$sample.ID[basicMetaMapUpdatedWeb$dataset.ID=="id29"]
  id29MetaIDs = id29SampleIDs # meta.ID is same
  table(duplicated(id29SampleIDs))
  
  ### id27 - kcl cirrhosis ###
  id27SampleIDs = basicMetaMapUpdatedWeb$sample.ID[basicMetaMapUpdatedWeb$dataset.ID=="id27"]
  id27MetaIDs = id27SampleIDs # meta.ID is same
  table(duplicated(id27SampleIDs))
  
  
  mergedMetaIds = c(id28MetaIDsUnique, id29MetaIDs, id27MetaIDs)
  lenMergedMetaIds = length(mergedMetaIds)
  
  set.seed(2)
  newIds = sort(paste("NEW", sample(1:500, size=lenMergedMetaIds), sep = "."))
  names(newIds) = (mergedMetaIds)
  
  id29NewMetaIDs = newIds[match(id29MetaIDs, names(newIds))]
  id27NewMetaIDs = newIds[match(id27MetaIDs, names(newIds))]
  id28NewMetaIDs = newIds[match(id28MetaIDs, names(newIds))] 
  
  id29NewSampleIDs = id29NewMetaIDs
  id27NewSampleIDs = id27NewMetaIDs
  id28NewSampleIDs = basicMetaMapUpdatedWeb$sample.ID[basicMetaMapUpdatedWeb$dataset.ID == "id28"]
  names(id28NewSampleIDs) = basicMetaMapUpdatedWeb$sample.ID[basicMetaMapUpdatedWeb$dataset.ID == "id28"]
  
  
  basicMetaMapUpdatedWeb$metadata.ID[ basicMetaMapUpdatedWeb$metadata.ID %in% names(id28NewMetaIDs)] = id28NewMetaIDs[basicMetaMapUpdatedWeb$metadata.ID[ basicMetaMapUpdatedWeb$metadata.ID %in% names(id28NewMetaIDs)]]
  basicMetaMapUpdatedWeb$metadata.ID[match(names(id27NewMetaIDs), basicMetaMapUpdatedWeb$metadata.ID)] = id27NewMetaIDs
  basicMetaMapUpdatedWeb$metadata.ID[match(names(id29NewMetaIDs), basicMetaMapUpdatedWeb$metadata.ID)] = id29NewMetaIDs
  id28NewSampleIDs = paste(basicMetaMapUpdatedWeb$metadata.ID[basicMetaMapUpdatedWeb$dataset.ID == "id28"], basicMetaMapUpdatedWeb$subtype[basicMetaMapUpdatedWeb$dataset.ID == "id28"], sep="_")
  names(id28NewSampleIDs) = basicMetaMapUpdatedWeb$sample.ID[basicMetaMapUpdatedWeb$dataset.ID == "id28"]
  
  
  
  basicMetaMapUpdatedWeb$sample.ID[match(names(id27NewSampleIDs), basicMetaMapUpdatedWeb$sample.ID)] = id27NewSampleIDs
  basicMetaMapUpdatedWeb$sample.ID[match(names(id29NewSampleIDs), basicMetaMapUpdatedWeb$sample.ID)] = id29NewSampleIDs
  basicMetaMapUpdatedWeb$sample.ID[match(names(id28NewSampleIDs), basicMetaMapUpdatedWeb$sample.ID)] = id28NewSampleIDs
  
  #basicMetaMapUpdatedWeb$sample.ID[basicMetaMapUpdatedWeb$dataset.ID == "id28"] = paste(basicMetaMapUpdatedWeb$metadata.ID[basicMetaMapUpdatedWeb$dataset.ID == "id28"], basicMetaMapUpdatedWeb$subtype[basicMetaMapUpdatedWeb$dataset.ID == "id28"], sep="_")
  selColumns = c("dataset.ID", "sample.ID", "metadata.ID", "type",
                 "subtype", "Age", "Gender", "BMI", "Geography", "Sequencer", "GeneRichness", "enteroType")
  basicMetaMapUpdatedWeb = basicMetaMapUpdatedWeb[,selColumns]
  basicMetaMapUpdatedWeb$Age [basicMetaMapUpdatedWeb$dataset.ID == "id27"] = NA
  basicMetaMapUpdatedWeb$BMI [basicMetaMapUpdatedWeb$dataset.ID == "id27"] = NA
  basicMetaMapUpdatedWeb$Gender [basicMetaMapUpdatedWeb$dataset.ID == "id27"] = NA
  
  basicMetaMapUpdatedWeb$Age [basicMetaMapUpdatedWeb$dataset.ID == "id29"] = NA
  basicMetaMapUpdatedWeb$BMI [basicMetaMapUpdatedWeb$dataset.ID == "id29"] = NA
  basicMetaMapUpdatedWeb$Gender [basicMetaMapUpdatedWeb$dataset.ID == "id29"] = NA
  
  colnames(mergeMatUpdatedWeb)[match(names(id28NewSampleIDs), colnames(mergeMatUpdatedWeb))] = id28NewSampleIDs
  colnames(mergeMatUpdatedWeb)[match(names(id27NewSampleIDs), colnames(mergeMatUpdatedWeb))] = id27NewSampleIDs
  colnames(mergeMatUpdatedWeb)[match(names(id29NewSampleIDs), colnames(mergeMatUpdatedWeb))] = id29NewSampleIDs
  
  table(id27NewSampleIDs %in% colnames(mergeMatUpdatedWeb))
  table(id28NewSampleIDs %in% colnames(mergeMatUpdatedWeb))
  table(id29NewSampleIDs %in% colnames(mergeMatUpdatedWeb))
  
  basicMetaMapUpdatedWeb = basicMetaMapUpdatedWeb[!basicMetaMapUpdatedWeb$dataset.ID %in% c("id27","id29"),]
  
  write.csv(basicMetaMapUpdatedWeb, "J:\\Deposit\\Project\\2018_microbiome_atlas\\atlas.webpage.detail\\raw data\\HGMA.web.metadata.csv", row.names = F)
  write.csv(mergeMatUpdatedWeb, "J:\\Deposit\\Project\\2018_microbiome_atlas\\atlas.webpage.detail\\raw data\\HGMA.web.MSP.abundance.matrix.csv")
  
  
}

dorinesInflammagingData = F
if (dorinesInflammagingData) {
  twinSamples = basicMetaMapUpdated$sample.ID[basicMetaMapUpdated$dataset.ID=="id29" ]
  ukCirrhosisSamples = basicMetaMapUpdated$sample.ID[basicMetaMapUpdated$dataset.ID=="id27" ]
  exUkSamples = c(twinSamples, ukCirrhosisSamples)
  
  basicMetaMapUpdatedNormal = basicMetaMapUpdated[basicMetaMapUpdated$sample.ID %in% newNormalIds,]
  mergeMatUpdatedNormal = mergeMatUpdated[,colnames(mergeMatUpdated) %in% newNormalIds]
  
  basicMetaMapUpdatedNormal = basicMetaMapUpdatedNormal[!basicMetaMapUpdatedNormal$sample.ID %in% exUkSamples,]
  mergeMatUpdatedNormal = mergeMatUpdatedNormal[,!colnames(mergeMatUpdatedNormal) %in% exUkSamples]
  
  save(basicMetaMapUpdatedNormal, mergeMatUpdatedNormal, file="C://Data//comparative.analysis.healthy.sweden//Dorines.metadata.msp.matrix.2020.0530.RData")
}

if (F) {
  write.table(basicMetaMapUpdated[basicMetaMapUpdated$dataset.ID=="id29",], 
              "C://Data//comparative.analysis.healthy.sweden//twin.metadata.atlas.txt", sep="\t", quote = F)
  
}

checkCovidTwinData = F
if (checkCovidTwinData) {
  covidMgsRData = "J:\\Deposit\\Project\\2020_COVID_kings_twin_data\\covid.final.mgs.med.vec.10M.RData"
  covidMgsTableFile = "J:\\Deposit\\Project\\2020_COVID_kings_twin_data\\covid.final.mgs.med.vec.10M.txt"
  
  
  load(covidMgsRData) 
  covidMgsMat = mgs_med_vec_10m
  
  rowSpecies = getSpeciesName(rownames(covidMgsMat), taxo)
  unclassifiedInd = grepl("unclassified", rowSpecies)
  rownames(covidMgsMat) = rowSpecies
  covidMgsMat = covidMgsMat[!unclassifiedInd,]
  
  write.table(covidMgsMat, covidMgsTableFile, sep="\t")
}



### very important part ###
checkGridForWellness = T
if (checkGridForWellness) {
  gridDataDir="J://Deposit//Project//2018_microbiome_atlas//GRiD_for_wellness_dataset//"
  gridList = list()
  for (file in dir(gridDataDir)) {
    if (file !="gridOut.RData") {
      sample = gsub(".GRiD.txt","",file)
      
      fullPath=paste(gridDataDir, file, sep="")
      print(fullPath)
      
      sampleTab = read.delim(fullPath, sep="\t", header=T, row.names=1)
      #print(head(sampleTab))
      
      gridVec = sampleTab[,"GRiD"]
      heteroVec = sampleTab[,"Species_heterogeneity"]
      names(gridVec) = rownames(sampleTab)
      gridTab = data.frame(name=names(gridVec), grid=gridVec, heterogeneity = heteroVec, sample=sample)
      
      gridList[[sample]] = gridTab
      
    }
  }
  
  heteroCut=0.3
  allGridTab = do.call(rbind.data.frame, gridList)
  allGridTabSel = allGridTab[allGridTab$heterogeneity <= heteroCut, ]
  allGridMat = acast(data=allGridTabSel, name~sample, value.var="grid")
  rownames(allGridMat)=gsub("_GUT_GENOME[0-9]+","",rownames(allGridMat))
  
  
  print(dim(allGridMat)) 
  ##
  ##
  colnames(allGridMat) = paste("X", colnames(allGridMat), sep="")
  
  
  maxGridBySubjects = colMaxs(allGridMat, na.rm=T)
  names(maxGridBySubjects) = colnames(allGridMat)
  
  
  
  wellnessCheckMode = F ### PLEASE RUN INFLOW/OUTFLOW SCRIPT FIRST
  if (wellnessCheckMode) {
    wellnessMgsMat
    inflowSpecies
    outflowSpecies
    
    allGridMat = allGridMat[,match(colnames(wellnessMgsMat), colnames(allGridMat))]
    allGridMat2 = allGridMat
    
    
    dim(allGridMat)
    
    iocClass = rep("commensal", dim(allGridMat)[1])
    names(iocClass) = rownames(allGridMat)
    iocClass[inflowSpecies] = "inflow"
    iocClass[outflowSpecies] = "outflow"
    iocClass[sharedSpecies] = "shared"
    
    totalMeans = apply(allGridMat, 1, function(x) mean(x, na.rm=T))
    totalMaxs = apply(allGridMat, 1, function(x) max(x, na.rm=T))
    
    allGridVisits = gsub("X[0-9]+_","",colnames(allGridMat))
    
    allGridTab = data.frame(name = rownames(allGridMat), 
                            species = getSpeciesName(rownames(allGridMat), gssOssTabNew),
                            class = iocClass,
                            allGridMat)
    
    
    commensal = rownames(allGridMat)
    commensal = commensal[!commensal %in% c(inflowSpecies, outflowSpecies, sharedSpecies)]
    
    allGridMat3= allGridMat
    allGridMat3[is.na(allGridMat)]=1
    
    cutOff=2
    highSpecies = rownames(allGridMat3)[rowMaxs(allGridMat3,na.rm=T)>=cutOff]
    highSamples = colnames(allGridMat3)[colMaxs(allGridMat3,na.rm=T)>=cutOff]
    
    rownames(allGridMat3) = getSpeciesName(rownames(allGridMat3), taxo)
    
    heatmap.2(log10(allGridMat3[highSpecies, highSamples]), trace="none")
    
    gridOut.inflow.outflow = list(inflow = rowMeans(allGridMat2[rownames(allGridMat2) %in% inflowSpecies,],na.rm=T), 
                                  outflow=rowMeans(allGridMat2[rownames(allGridMat2) %in% outflowSpecies,],na.rm=T), 
                                  commensal=rowMeans(allGridMat2[rownames(allGridMat2) %in% commensal,], na.rm=T))
    
    boxplot(gridOut.inflow.outflow)###
    
    gridOut.inflow.outflow = list(inflow = rowMeans(allGridMat3[rownames(allGridMat3) %in% inflowSpecies,],na.rm=T), 
                                  outflow=rowMeans(allGridMat3[rownames(allGridMat3) %in% outflowSpecies,],na.rm=T), 
                                  commensal=rowMeans(allGridMat3[rownames(allGridMat3) %in% commensal,], na.rm=T))
    
    gridMaxOut.inflow.outflow = list(inflow = rowMaxs(allGridMat3[rownames(allGridMat3) %in% inflowSpecies,],na.rm=T), 
                                     outflow=rowMaxs(allGridMat3[rownames(allGridMat3) %in% outflowSpecies,],na.rm=T), 
                                     commensal=rowMaxs(allGridMat3[rownames(allGridMat3) %in% commensal,], na.rm=T))
    wilcox.test(gridMaxOut.inflow.outflow$inflow, gridMaxOut.inflow.outflow$outflow)
    wilcox.test(gridMaxOut.inflow.outflow$outflow, gridMaxOut.inflow.outflow$commensal)
    
    boxplot(gridMaxOut.inflow.outflow)
    table(inflowSpecies %in% rownames(allGridMat2))
    
    highGridSpecies = rownames(allGridMat2)[rowMaxs(allGridMat2,na.rm=T)>2]
    table(highGridSpecies %in% inflowSpecies)
    table(highGridSpecies %in% outflowSpecies)
    table(highGridSpecies %in% sharedSpecies)
    table(highGridSpecies %in% commensal)
    
    inflowMat = allGridMat[rownames(allGridMat) %in% inflowSpecies,]
    outflowMat = allGridMat[rownames(allGridMat) %in% inflowSpecies,]
    
    levelplot(inflowMat)
    
    
    wellnessMgsMatSel = wellnessMgsMat[match(rownames(allGridMat), rownames(wellnessMgsMat)),]
    
    corList=list()
    for (ind in 1:229) {
      cor = cor(wellnessMgsMatSel[ind,], allGridMat[ind,], use = "pairwise.complete.obs")
      corList[[ind]]=cor
    }
    names(corList) = rownames(wellnessMgsMatSel)
    
    corList = do.call(c,corList)
    
    
    getGridMat <- function(currMgs) {
      markovSeqOfMsp = (sapply(wellnessUniqSubjs, function(subj) {
        currSamples = wellnessMetadata$sample.ID[wellnessMetadata$metadata.ID==subj]
        currSamples = currSamples[order(currSamples)]
        currMgsMat = allGridMat[,currSamples]
        #currMgsSubjVec = (currMgsMat[currMgs,] >0)*1
        currMgsSubjVec = currMgsMat[currMgs,]
        
      }, simplify = F))
      mat = do.call(rbind, markovSeqOfMsp)
      #class(mat) = "numeric"
      colnames(mat)=c("v1","v2","v3","v4")
      return(mat)
    }
    
    target="msp_0769"
    target="msp_0468"
    target="msp_0457"
    target="msp_0026"
    plot(wellnessMgsMatSel[target,], allGridMat[target,])
    
    currGridMat = getGridMat(target)
    levelplot(t(currGridMat), col.regions=colorRampPalette(c("white","gray"))(256),
              aspect=3, xlab="",ylab="",colorkey=T,scales=list(draw=F))
    
    boxplot(gridOut.inflow.outflow)
    t.test(gridOut.inflow.outflow$inflow, gridOut.inflow.outflow$commensal)
    t.test(gridOut.inflow.outflow$inflow, gridOut.inflow.outflow$outflow)
    
    ks.test(gridOut.inflow.outflow$inflow, gridOut.inflow.outflow$commensal, alternative = "less")
    ks.test(gridOut.inflow.outflow$inflow, gridOut.inflow.outflow$outflow, alternative = "less")
    ks.test(1:10, 100:200,  alternative = "less")
    
    highGridSpecies = rownames(allGridMat2)[rowMaxs(allGridMat2,na.rm=T)>2]
    
    
    
    ###
  }
  
  individualAnalysis = T
  if (individualAnalysis) {
    
    ###
    numCut = 34
    
    highRichnessSamples = wellnessMetadata[order(wellnessMetadata$GeneRichness,decreasing = T),"sample.ID"][1:numCut]
    lowRichnessSamples = wellnessMetadata[order(wellnessMetadata$GeneRichness,decreasing = F),"sample.ID"][1:numCut]
    
    inflowGrid.highRichnessSamples = rowMedians(allGridMat[rownames(allGridMat) %in% inflowSpecies,highRichnessSamples], na.rm = T)
    inflowGrid.lowRichnessSamples = rowMedians(allGridMat[rownames(allGridMat) %in% inflowSpecies,lowRichnessSamples], na.rm = T)
    outflowGrid.highRichnessSamples = rowMedians(allGridMat[rownames(allGridMat) %in% outflowSpecies,highRichnessSamples], na.rm = T)
    outflowGrid.lowRichnessSamples = rowMedians(allGridMat[rownames(allGridMat) %in% outflowSpecies,lowRichnessSamples], na.rm = T)
    commensalGrid.highRichnessSamples = rowMedians(allGridMat[rownames(allGridMat) %in% commensalSpecies,highRichnessSamples], na.rm = T)
    commensalGrid.lowRichnessSamples = rowMedians(allGridMat[rownames(allGridMat) %in% commensalSpecies,lowRichnessSamples], na.rm = T)
    
    boxplot(list(inflowGrid.highRichnessSamples,
                 outflowGrid.highRichnessSamples), ylim=c(1,2))
    
    
    boxplot(list(inflowGrid.lowRichnessSamples,
                 outflowGrid.lowRichnessSamples), ylim=c(1,2))
    
    ###
    numCut = 34
    
    highWeightSamples = names(sort(wellnessClinicalMatrix2["anthropometry_weight",], decreasing=T))[1:numCut]
    lowWeightSamples = names(sort(wellnessClinicalMatrix2["anthropometry_weight",], decreasing=F))[1:numCut]
    
    inflowGrid.highWeightSamples = rowMedians(allGridMat[rownames(allGridMat) %in% inflowSpecies,highWeightSamples], na.rm = T)
    inflowGrid.lowWeightSamples = rowMedians(allGridMat[rownames(allGridMat) %in% inflowSpecies,lowWeightSamples], na.rm = T)
    outflowGrid.highWeightSamples = rowMedians(allGridMat[rownames(allGridMat) %in% outflowSpecies,highWeightSamples], na.rm = T)
    outflowGrid.lowWeightSamples = rowMedians(allGridMat[rownames(allGridMat) %in% outflowSpecies,lowWeightSamples], na.rm = T)
    commensalGrid.highWeightSamples = rowMedians(allGridMat[rownames(allGridMat) %in% commensalSpecies,highWeightSamples], na.rm = T)
    commensalGrid.lowWeightSamples = rowMedians(allGridMat[rownames(allGridMat) %in% commensalSpecies,lowWeightSamples], na.rm = T)
    
    boxplot(list(inflowGrid.highWeightSamples,
                 outflowGrid.highWeightSamples), ylim=c(1,2))
    
    
    boxplot(list(inflowGrid.lowWeightSamples,
                 outflowGrid.lowWeightSamples), ylim=c(1,2))
    
    ###
    numCut = 34
    
    highTGSamples = names(sort(wellnessClinicalMatrix2["S.TG.A1",], decreasing=T)[sort(wellnessClinicalMatrix2["S.TG.A1",], decreasing=T)>2.3])
    lowTGSamples = names(sort(wellnessClinicalMatrix2["S.TG.A1",], decreasing=T)[sort(wellnessClinicalMatrix2["S.TG.A1",], decreasing=T)<1.7]) 
    
    
    inflowGrid.highTGSamples = rowMedians(allGridMat[rownames(allGridMat) %in% inflowSpecies,highTGSamples], na.rm = T)
    inflowGrid.lowTGSamples = rowMedians(allGridMat[rownames(allGridMat) %in% inflowSpecies,lowWeightSamples], na.rm = T)
    outflowGrid.highTGSamples = rowMedians(allGridMat[rownames(allGridMat) %in% outflowSpecies,highTGSamples], na.rm = T)
    outflowGrid.lowTGSamples = rowMedians(allGridMat[rownames(allGridMat) %in% outflowSpecies,lowWeightSamples], na.rm = T)
    commensalGrid.highTGSamples = rowMedians(allGridMat[rownames(allGridMat) %in% commensalSpecies,highTGSamples], na.rm = T)
    commensalGrid.lowTGSamples = rowMedians(allGridMat[rownames(allGridMat) %in% commensalSpecies,lowWeightSamples], na.rm = T)
    
    boxplot(list(inflowGrid.highTGSamples,
                 outflowGrid.highTGSamples), ylim=c(1,2))
    
    
    boxplot(list(inflowGrid.lowTGSamples,
                 outflowGrid.lowTGSamples), ylim=c(1,2))
    
    ###
    numCut = 34
    
    highBmiSamples = names(sort(wellnessClinicalMatrix2["bmi",], decreasing=T)[sort(wellnessClinicalMatrix2["bmi",], decreasing=T)>30])
    lowBmiSamples = names(sort(wellnessClinicalMatrix2["bmi",], decreasing=T)[sort(wellnessClinicalMatrix2["bmi",], decreasing=T)<30 & sort(wellnessClinicalMatrix2["bmi",], decreasing=T)>18.5]) 
    
    
    inflowGrid.highBmiSamples = rowMedians(allGridMat[rownames(allGridMat) %in% inflowSpecies,highBmiSamples], na.rm = T)
    inflowGrid.lowBmiSamples = rowMedians(allGridMat[rownames(allGridMat) %in% inflowSpecies,lowWeightSamples], na.rm = T)
    outflowGrid.highBmiSamples = rowMedians(allGridMat[rownames(allGridMat) %in% outflowSpecies,highBmiSamples], na.rm = T)
    outflowGrid.lowBmiSamples = rowMedians(allGridMat[rownames(allGridMat) %in% outflowSpecies,lowWeightSamples], na.rm = T)
    commensalGrid.highBmiSamples = rowMedians(allGridMat[rownames(allGridMat) %in% commensalSpecies,highBmiSamples], na.rm = T)
    commensalGrid.lowBmiSamples = rowMedians(allGridMat[rownames(allGridMat) %in% commensalSpecies,lowWeightSamples], na.rm = T)
    
    boxplot(list(inflowGrid.highBmiSamples,
                 outflowGrid.highBmiSamples), ylim=c(1,2))
    
    
    boxplot(list(inflowGrid.lowBmiSamples,
                 outflowGrid.lowBmiSamples), ylim=c(1,2))
    
    ###
    
    highInflowSamples = wellnessMetadata[order(wellnessMetadata$scaledInflow,decreasing = T),"sample.ID"][1:numCut]
    lowInflowSamples = wellnessMetadata[order(wellnessMetadata$scaledInflow,decreasing = F),"sample.ID"][1:numCut]
    
    highLowWellnessClinMat = cbind(wellnessClinicalMatrix[,highInflowSamples], wellnessClinicalMatrix[,lowInflowSamples])
    
    #significant parameters among higher inflow samples
    apply(highLowWellnessClinMat, 1, function(x) {
      wilcox.out = wilcox.test(x[1:34], x[35:68])
      return(wilcox.out)
    })
    
    #lower parameters among higher inflow samples
    apply(highLowWellnessClinMat, 1, function(x) {
      wilcox.out = wilcox.test(x[1:34], x[35:68], alternative="less")
      return(wilcox.out)
    })
    
    #higher parameters among higher inflow samples
    apply(highLowWellnessClinMat, 1, function(x) {
      wilcox.out = wilcox.test(x[1:34], x[35:68], alternative="greater")
      return(wilcox.out)
    })
    
    length(lowInflowSamples)
    table(lowInflowSamples %in% highInflowSamples)
    
    inflowGrid.highInflowSamples = rowMeans(allGridMat[rownames(allGridMat) %in% inflowSpecies,highInflowSamples], na.rm = T)
    inflowGrid.lowInflowSamples = rowMeans(allGridMat[rownames(allGridMat) %in% inflowSpecies,lowInflowSamples], na.rm = T)
    outflowGrid.highInflowSamples = rowMeans(allGridMat[rownames(allGridMat) %in% outflowSpecies,highInflowSamples], na.rm = T)
    outflowGrid.lowInflowSamples = rowMeans(allGridMat[rownames(allGridMat) %in% outflowSpecies,lowInflowSamples], na.rm = T)
    commensalGrid.highInflowSamples = rowMeans(allGridMat[rownames(allGridMat) %in% commensalSpecies,highInflowSamples], na.rm = T)
    commensalGrid.lowInflowSamples = rowMeans(allGridMat[rownames(allGridMat) %in% commensalSpecies,lowInflowSamples], na.rm = T)
    
    inflowGrid.highOutflowSamples = rowMeans(allGridMat[rownames(allGridMat) %in% inflowSpecies,highOutflowSamples], na.rm = T)
    inflowGrid.lowOutflowSamples = rowMeans(allGridMat[rownames(allGridMat) %in% inflowSpecies,lowOutflowSamples], na.rm = T)
    outflowGrid.highOutflowSamples = rowMeans(allGridMat[rownames(allGridMat) %in% outflowSpecies,highOutflowSamples], na.rm = T)
    outflowGrid.lowOutflowSamples = rowMeans(allGridMat[rownames(allGridMat) %in% outflowSpecies,lowOutflowSamples], na.rm = T)
    
    individual.grid.out = list(inflow.HI = inflowGrid.highInflowSamples,
                               inflow.LI = inflowGrid.lowInflowSamples,
                               outflow.LI = outflowGrid.lowInflowSamples,
                               outflow.HI = outflowGrid.highInflowSamples)
    
    wilcox.test(individual.grid.out$inflow.HI, individual.grid.out$inflow.LI, alternative = "greater")
    wilcox.test(individual.grid.out$inflow.HI, individual.grid.out$outflow.LI, alternative = "greater")
    wilcox.test(individual.grid.out$inflow.HI, individual.grid.out$outflow.HI, alternative = "greater")
    
    wilcox.test(inflowGrid.highInflowSamples, inflowGrid.lowInflowSamples, alternative = "greater")
    wilcox.test(inflowGrid.highOutflowSamples, outflowGrid.highOutflowSamples, alternative = "greater")
    
    
    inflowGrid.allSamples = rowMeans(allGridMat[rownames(allGridMat) %in% inflowSpecies,], na.rm = T)
    outflowGrid.allSamples = rowMeans(allGridMat[rownames(allGridMat) %in% outflowSpecies,], na.rm = T)
    
    wilcox.test(inflowGrid.allSamples, outflowGrid.allSamples, alternative = "greater")
    
    #####
    
    boxplot(list(inflowGrid.highInflowSamples,
                 outflowGrid.highInflowSamples), ylim=c(1,2))
    
    boxplot(list(inflowGrid.allSamples,
                 outflowGrid.allSamples), ylim=c(1,2))
    
    boxplot(list(inflowGrid.highOutflowSamples,
                 outflowGrid.highOutflowSamples), ylim=c(1,2))
    
    
    
    
    #####
    
    boxplot(list(inflowGrid.highInflowSamples,
                 inflowGrid.lowInflowSamples), ylim=c(1,2))
    
    
    boxplot(list(outflowGrid.highInflowSamples,
                 outflowGrid.lowInflowSamples), ylim=c(1,2))
    
    
    boxplot(list(inflowGrid.highInflowSamples,
                 outflowGrid.highInflowSamples), ylim=c(1,2))
    
    
    boxplot(list(inflowGrid.lowInflowSamples,
                 outflowGrid.lowInflowSamples), ylim=c(1,2))
    
    
    
    
    
  }
  
  presentStats = t(sapply(rownames(allGridMat), function(currMsp) {
    
    appearing = 0
    disappearing = 0
    
    
    apply(wellnessSeqTab, 1, function(currRow) {
      before = currRow["before"]
      after = currRow["after"]
      currBefore = allGridMat[currMsp, before]
      currAfter = allGridMat[currMsp, after]
      
      if (!is.na(currBefore) ) {
        if (!is.na(currAfter)) {
          appearing <<- appearing + 1
        }
        else {
          disappearing <<- disappearing + 1
        }
      }
      
      
    })
    
    return(c(appearing, disappearing))
    
    
  }))
  colnames(presentStats) = c("appearing","disappearing")
  presentStats = as.data.frame(presentStats)
  presentStats$species = getSpeciesName(rownames(presentStats),taxo)
  
  
  increaseStats = sapply(rownames(allGridMat), function(currMsp) {
    
    ratios = apply(wellnessSeqTab, 1, function(currRow) {
      before = currRow["before"]
      after = currRow["after"]
      currBefore = allGridMat[currMsp, before]
      currAfter = allGridMat[currMsp, after]
      
      ratio = NA
      if (!is.na(currBefore) && !is.na(currAfter) ) {
        
        ratio = (currAfter - currBefore) / currBefore
      }
      return(ratio)
      
    })
    
    return(ratios)
    
    
  }, simplify = F)
  names(increaseStats) = rownames(allGridMat)
  
  increaseValues = unlist(na.omit(unlist(increaseStats)))
  
  
  
  increaseMedian = unlist(lapply(increaseStats, function(x) median(x,na.rm=T)))
  increaseMean = unlist(lapply(increaseStats, function(x) mean(x,na.rm=T)))
  increaseMax = unlist(lapply(increaseStats, function(x) max(x,na.rm=T)))
  increaseMin = unlist(lapply(increaseStats, function(x) min(x,na.rm=T)))
  
  q99 = as.numeric(quantile(increaseValues, 0.99))
  q95 = as.numeric(quantile(increaseValues, 0.95))
  q90 = as.numeric(quantile(increaseValues, 0.90))
  q10 = as.numeric(quantile(increaseValues, 0.10))
  q05 = as.numeric(quantile(increaseValues, 0.05))
  q01 = as.numeric(quantile(increaseValues, 0.01))
  
  
  overSpecies =  names(increaseMax[increaseMax >= q90])
  underSpecies = names(increaseMin[increaseMin <= q10])
  
  inflowGridSpeciesList = list(over = inflow[overSpecies],
                               under = inflow[underSpecies])
  
  outflowGridSpeciesList = list(over = outflow[overSpecies],                   
                                under = outflow[underSpecies])
  
  boxplot(inflowGridSpeciesList)
  boxplot(outflowGridSpeciesList)
  
  t.test(inflowGridSpeciesList$over, inflowGridSpeciesList$under )
  t.test(outflowGridSpeciesList$over, outflowGridSpeciesList$under )
  
  
  
  
}

funcCompareByRichnessSignature = F
if (funcCompareByRichnessSignature) {
  
  loadRichnessSignature = T
  if (loadRichnessSignature) {
    statsMgsForAllSamples = "J:\\Deposit\\Project\\2018_microbiome_atlas\\atlas.volcanot.out\\richness.volcano.stats.mgs.for.all.txt"
    statsMgsForNormSamples = "J:\\Deposit\\Project\\2018_microbiome_atlas\\atlas.volcanot.out\\richness.volcano.stats.mgs.for.normal.txt"
    statsMgsForDiseaseSamples = "J:\\Deposit\\Project\\2018_microbiome_atlas\\atlas.volcanot.out\\richness.volcano.stats.mgs.for.disease.txt"
    statsMgsForIndustrialSamples  = "J:\\Deposit\\Project\\2018_microbiome_atlas\\atlas.volcanot.out\\richness.volcano.stats.mgs.for.industrial.txt"
    statsMgsForTraditionalSamples  = "J:\\Deposit\\Project\\2018_microbiome_atlas\\atlas.volcanot.out\\richness.volcano.stats.mgs.for.traditional.txt"
    adjpCut=1e-3
    statsMgsForNorm = read.table(statsMgsForNormSamples, sep="\t", header=T, stringsAsFactors = F)
    statsMgsForNormSel = statsMgsForNorm[statsMgsForNorm$qvalue <= adjpCut,]
    statsMgsForNormSelHgc = statsMgsForNormSel[statsMgsForNormSel$lfc > 2,]
    statsMgsForNormSelLgc = statsMgsForNormSel[statsMgsForNormSel$lfc < -2,]
    
    hgcSpecies = rownames(statsMgsForNormSelHgc)
    lgcSpecies = rownames(statsMgsForNormSelLgc)
  }
  
  checkFuncContrast = T
  if (checkFuncContrast) {
    load(funcMatRData)
    
    funcMatTranspose = t(funcMat)
    # hgcFuncMat = funcMatTranspose[,colnames(funcMatTranspose) %in% hgcSpecies]
    # lgcFuncMat = funcMatTranspose[,colnames(funcMatTranspose) %in% lgcSpecies]
    hgcFuncMat = funcMatTranspose[,colnames(funcMatTranspose) %in% hgcSpecies]
    lgcFuncMat = funcMatTranspose[,colnames(funcMatTranspose) %in% lgcSpecies]
    
    funcs = rownames(funcMatTranspose)
    pvalues = sapply(funcs, function(currFunc) {
      hgcCurrFunc = hgcFuncMat[currFunc,]
      lgcCurrFunc = lgcFuncMat[currFunc,]
      
      hgcAbsent = sum(hgcCurrFunc==0)
      hgcPresent = sum(hgcCurrFunc==1)
      
      lgcAbsent = sum(lgcCurrFunc==0)
      lgcPresent = sum(lgcCurrFunc==1)
      
      contigTab = cbind(HGC=c(present=hgcPresent, absent=hgcAbsent),
                        LGC=c(present=lgcPresent, absent=lgcAbsent))
      pvalue=chisq.test(contigTab)$p.value
      return(pvalue)
    }, simplify = F)
    pvalues = unlist(pvalues)
    
    oddsRatios = sapply(funcs, function(currFunc) {
      hgcCurrFunc = hgcFuncMat[currFunc,]
      lgcCurrFunc = lgcFuncMat[currFunc,]
      
      hgcAbsent = sum(hgcCurrFunc==0)
      hgcPresent = sum(hgcCurrFunc==1)
      
      lgcAbsent = sum(lgcCurrFunc==0)
      lgcPresent = sum(lgcCurrFunc==1)
      
      oddsRatio = (hgcPresent*lgcAbsent)/(hgcAbsent*lgcPresent)
      if (lgcPresent*hgcAbsent == 0) return(Inf) 
      return(oddsRatio)
    }, simplify = F)
    oddsRatios = unlist(oddsRatios)
    
    chisqStatTab = data.frame(term=names(pvalues),
                              pvalue=pvalues,
                              oddsRatio=oddsRatios,
                              logP = -log10(pvalues),
                              logOdds = log(oddsRatios))
    
    # View(chisqStatTab[grepl("cazy",rownames(chisqStatTab)),])
    # View(chisqStatTab[chisqStatTab$pvalue<1e-5,])
    
    View(chisqStatTab[mgePfamTerms,])
    View(chisqStatTab[arTerms,])
    View(chisqStatTab[antismashTerms,])
    
    
  }
  
}

funcCompareByCountryCluster = T
if (funcCompareByCountryCluster) {
  
  loadCountryInfo = T
  if (loadCountryInfo) {
    etfCountriesUpdated = c("UK","Germany","Italy","Spain","Sweden","Denmark","Luxembourg") #, "France", "Finland"
    etpCountriesUpdated = c("Peru","Mongo","Thailand","Fiji","Madagascar","Tanzania","India") #,"Fiji","India"
    etbCountriesUpdated = c("China","Japan","US")
  }
  
  loadFuncAnnot = T
  if (loadFuncAnnot) {
    ### virulence factors ###
    vfMatRData = "j://Deposit/Project/2018_microbiome_atlas/functional.annotation/PATRIC/vfBestMat.RData"
    load(vfMatRData) #vfMatDigit
    vfTerms = colnames(vfMatDigit)
    patricVfDescTabFile = "J://Deposit/Project/2018_microbiome_atlas/functional.annotation/PATRIC/PATRIC_sp_gene.txt"
    patricVfDescTab = read.table(patricVfDescTabFile, header=T, sep="\t", stringsAsFactors = F)
    
    ### JGI phenotype ###
    jgiMatRData = "j://Deposit/Project/2018_microbiome_atlas/functional.annotation/jgiMat.RData"
    load(jgiMatRData)
    jgiTerms = colnames(jgiMat)
    
    ### antibiotic resistance ###
    mustardMatRData = "j://Deposit/Project/2018_microbiome_atlas/functional.annotation/mustard.mat.RData"
    load(mustardMatRData)
    arTerms = colnames(mustardMat)
    
    ### antismash - secondary metabolism ###
    antismashMatRData = "J://Deposit/Project/2018_microbiome_atlas/functional.annotation/msp1992.igc2.antismashMat.RData"
    load(antismashMatRData)
    antismashMelt = melt(antismashMat)
    antismashMelt = antismashMelt[antismashMelt$value!=0,]
    antismashTerms = colnames(antismashMat)
    
    ### Cazy  ### 
    newGutCazyMatRData = "J://Deposit/Project/2018_microbiome_atlas/functional.annotation/CAZY.INRA/igc2.new.cazy.mat.RData"
    load(newGutCazyMatRData)
    colnames(cazyMat) = paste("cazy", colnames(cazyMat), sep=".")
    cazyTerms = colnames(cazyMat)
    
  }
  
  loadGeoSignal = T
  if (loadGeoSignal) {
    specificGroupCountriesMeltRData = "J://Deposit//Project//2018_microbiome_atlas//atlas.geography//specificGroupCountriesMelt.RData"
    specificGroupMgsByCountriesRData = "J://Deposit//Project//2018_microbiome_atlas//atlas.geography//specificGroupMgsByCountries.RData"
    specificGroupMgsAllRData = "J://Deposit//Project//2018_microbiome_atlas//atlas.geography//specificGroupMgsAll.RData"
    specificGroupMgsCountsByCountriesRData = "J://Deposit//Project//2018_microbiome_atlas//atlas.geography//specificGroupMgsCountsByCountries.RData" 
    
    load(specificGroupCountriesMeltRData)
    load(specificGroupMgsByCountriesRData)
    load(specificGroupMgsAllRData)
    load(specificGroupMgsCountsByCountriesRData)
    
    etfSpecies = unique(unlist(specificGroupMgsByCountries[etfCountriesUpdated]))
    etbSpecies = unique(unlist(specificGroupMgsByCountries[etbCountriesUpdated]))
    etpSpecies = unique(unlist(specificGroupMgsByCountries[etpCountriesUpdated]))
    
    etfOnlySpecies = etfSpecies[!etfSpecies %in% etbSpecies]
    etfOnlySpecies = etfOnlySpecies[!etfOnlySpecies %in% etpSpecies]
    etbOnlySpecies = etbSpecies[!etbSpecies %in% etfSpecies]
    etbOnlySpecies = etbOnlySpecies[!etbOnlySpecies %in% etpSpecies]
    etpOnlySpecies = etpSpecies[!etpSpecies %in% etfSpecies]
    etpOnlySpecies = etpOnlySpecies[!etpOnlySpecies %in% etbSpecies]
    
    length(etfSpecies)
    length(etbSpecies)
    length(etpSpecies)
    
    length(etfOnlySpecies)
    length(etbOnlySpecies)
    length(etpOnlySpecies)
    
    table(etfSpecies %in% etbSpecies)
    table(etfSpecies %in% etbSpecies)
    table(etbSpecies %in% etpSpecies)
    fbOverlap = etfSpecies[etfSpecies %in% etbSpecies]
    table(fbOverlap %in% etpSpecies)
    
    
    
  }
  
  loadFuncMode = T
  if (loadFuncMode) {
    funcMatRData = "j://Deposit/Project/2018_microbiome_atlas/functional.annotation/funcMat.20190808.RData"
    load(funcMatRData)
    funcRichness = rowSums(funcMat)
    
    mucinCazymes = c("cazy.GT27", "cazy.GT14", "cazy.GT11", "cazy.GT10", "cazy.GH89", "cazy.GH84", "cazy.GH33", "cazy.GH20", "cazy.GH29", "cazy.GH18", "cazy.GH123", "cazy.GH109", "cazy.CBM51", "cazy.CBM50", "cazy.CBM32")
    storageCazymes = c("cazy.CBM34", "cazy.CBM48", "cazy.CBM20", "cazy.GH13", "cazy.GH133", "cazy.GH32", "cazy.GH36", "cazy.GH37", "cazy.GH77", "cazy.GH57", "cazy.GT101", "cazy.GT26", "cazy.GT3", "cazy.GT35", "cazy.GT5")
    pectinCazymes = c("cazy.CBM77","cazy.CE8","cazy.GH105","cazy.GH28", "cazy.PL1")
    
    
    cazyRichness = rowSums(funcMat[,cazyTerms])
    #cazyRichness = rowSums(funcMat[,cazyTerms[!cazyTerms %in% mucinCazymes]])
    storageCazyRichness = rowSums(funcMat[,storageCazymes])
    mucinCazyRichness = rowSums(funcMat[,mucinCazymes])
    pectinCazyRichness = rowSums(funcMat[,colnames(funcMat)%in%pectinCazymes])
    
    arRichness = rowSums(funcMat[,arTerms])
    vfRichness = rowSums(funcMat[,vfTerms])
    secRichness = rowSums(funcMat[,antismashTerms])
    mgeRichness = rowSums(funcMat[,mgePfamTerms])
    
    targetRichness = pectinCazyRichness
    
    tradSpecies = unique(unlist(specificGroupMgsByCountries[traditionalCountries]))
    indSpecies = unique(unlist(specificGroupMgsByCountries[industrializedCountries]))
    
    
    boxplot(list(trad=targetRichness[tradSpecies],
                 industrial=targetRichness[indSpecies]))
  }
  
  chisqTestMode = T
  if (chisqTestMode) {
    
    
    cazyMode = T
    if (cazyMode) {
      mucinCazymes = c("cazy.GT27", "cazy.GT14", "cazy.GT11", "cazy.GT10", "cazy.GH89", "cazy.GH84", "cazy.GH33", "cazy.GH20", "cazy.GH29", "cazy.GH18", "cazy.GH123", "cazy.GH109", "cazy.CBM51", "cazy.CBM50", "cazy.CBM32")
      storageCazymes = c("cazy.CBM34", "cazy.CBM48", "cazy.CBM20", "cazy.GH13", "cazy.GH133", "cazy.GH32", "cazy.GH36", "cazy.GH37", "cazy.GH77", "cazy.GH57", "cazy.GT101", "cazy.GT26", "cazy.GT3", "cazy.GT35", "cazy.GT5")
      pectinCazymes = c("cazy.CBM77","cazy.CE8","cazy.GH105","cazy.GH28", "cazy.PL1")
      
      mucinCazymes =  gsub("cazy.","",mucinCazymes)
      storageCazymes =  gsub("cazy.","",storageCazymes)
      pectinCazymes =  gsub("cazy.","",pectinCazymes)
      
      pCut = 1e-2
      varMat = getVarScoreMat(funcMat, cazyTerms, etfSpecies, etbSpecies, etpSpecies)
      sigMat = selectVarScoreMat(varMat, pCut)
      
      colnames(sigMat)[1:3] = c("European","Pacific","NonWesternized")
      sigMat$lab = rownames(sigMat)
      sigMat$lab = gsub("cazy.","",sigMat$lab)
      sigMat$class = "others"
      sigMat$class[sigMat$lab %in% mucinCazymes] = "mucin"
      sigMat$class[sigMat$lab %in% storageCazymes] = "storage"
      sigMat$class[sigMat$lab %in% pectinCazymes] = "storage"
      sigMat$class = factor(sigMat$class, levels=c("others","mucin","storage"))
      View(sigMat)
      
      
      #ggtern(data=sigCazyMat, aes(x=F,y=B, z=P, label=lab)) + 
      ggtern(data=sigMat, aes(x=European,y=Pacific, z=NonWesternized, label=lab)) + 
        geom_point(aes(fill=class),
                   size = 4, 
                   shape = 21, 
                   color = "#80808011") +
        scale_fill_manual(values=c("#80808033","#0000ff88","#ff000088","#d4af3788")) +
        theme_ggtern(12, "") %+replace% ggplot2::theme_void(12, "") %+replace% theme(tern.axis.text = element_blank(), tern.axis.title = element_blank(), legend.position = "none" ) + 
        labs( x       = "",
              xarrow  = "Westernized",
              y       = "",
              yarrow  = "European",
              z       = "",
              zarrow  = "Non-westernized") 
      
      
      if (F) {
        
        
        funcMatTranspose = t(funcMat)
        funcMatTranspose = funcMatTranspose[cazyTerms,]
        
        etfFuncMat = funcMatTranspose[,colnames(funcMatTranspose) %in% etfSpecies]
        etbFuncMat = funcMatTranspose[,colnames(funcMatTranspose) %in% etbSpecies]
        etpFuncMat = funcMatTranspose[,colnames(funcMatTranspose) %in% etpSpecies]
        
        # etfFuncMat = funcMatTranspose[,colnames(funcMatTranspose) %in% etfSpecies]
        # etbFuncMat = funcMatTranspose[,colnames(funcMatTranspose) %in% etbSpecies]
        # etpFuncMat = funcMatTranspose[,colnames(funcMatTranspose) %in% etpSpecies]
        
        
        chisq_FB_cazy = getChisqStat(etfFuncMat, etbFuncMat, cazyTerms)
        chisq_BP_cazy = getChisqStat(etbFuncMat, etpFuncMat, cazyTerms)
        chisq_FP_cazy = getChisqStat(etfFuncMat, etpFuncMat, cazyTerms)
        
        
        pCut=1e-2
        sigCazy_FB = rownames(chisq_FB_cazy[!is.nan(chisq_FB_cazy$pvalue) & chisq_FB_cazy$pvalue<pCut,])
        sigCazy_BP = rownames(chisq_BP_cazy[!is.nan(chisq_BP_cazy$pvalue) & chisq_BP_cazy$pvalue<pCut,])
        sigCazy_FP = rownames(chisq_FP_cazy[!is.nan(chisq_FP_cazy$pvalue) & chisq_FP_cazy$pvalue<pCut,])
        sigCazy_union = unique(c(sigCazy_FB, sigCazy_BP, sigCazy_FP))
        
        sigCazyMat = getFractionMat(etfFuncMat, etbFuncMat, etpFuncMat, cazyTerms)
        sigCazyMat = as.data.frame(sigCazyMat)
        colnames(sigCazyMat) = c("European","Pacific","NonWesternized")
        sigCazyMat$lab = rownames(sigCazyMat)
        sigCazyMat$lab = gsub("cazy.","",sigCazyMat$lab)
        sigCazyMat$class = "others"
        sigCazyMat$class[sigCazyMat$lab %in% mucinCazymes] = "mucin"
        sigCazyMat$class[sigCazyMat$lab %in% storageCazymes] = "storage"
        sigCazyMat$class[sigCazyMat$lab %in% pectinCazymes] = "storage"
        sigCazyMat$class = factor(sigCazyMat$class, levels=c("others","mucin","storage"))
        View(sigCazyMat)
        
        
        
        
      }
      
      
    }
    
    antismashMode = T
    if(antismashMode) {
      
      
      pCut = 0.05
      varMat = getVarScoreMat(funcMat, antismashTerms, etfSpecies, etbSpecies, etpSpecies)
      sigMat = selectVarScoreMat(varMat, pCut)
      
      colnames(sigMat)[1:3] = c("European","Pacific","NonWesternized")
      sigMat$lab = rownames(sigMat)
      
      # sigMat$class = "others"
      # sigMat$class[sigMat$lab %in% mucinCazymes] = "mucin"
      # sigMat$class[sigMat$lab %in% storageCazymes] = "storage"
      # sigMat$class[sigMat$lab %in% pectinCazymes] = "storage"
      # sigMat$class = factor(sigMat$class, levels=c("others","mucin","storage"))
      View(sigMat)
      
      
      
      
      #ggtern(data=sigCazyMat, aes(x=F,y=B, z=P, label=lab)) + 
      ggtern(data=sigMat, aes(x=European,y=Pacific, z=NonWesternized, label=lab)) + 
        geom_point(fill="#80808033",#aes(fill=class),
                   size = 4, 
                   shape = 21, 
                   color = "#808080") +
        #scale_fill_manual(values=c("#80808033","#0000ff88","#ff000088","#d4af3788")) +
        theme_ggtern(12, "") %+replace% ggplot2::theme_void(12, "") %+replace% theme(tern.axis.text = element_blank(), tern.axis.title = element_blank(), legend.position = "none" ) + 
        labs( x       = "",
              xarrow  = "European",
              y       = "",
              yarrow  = "Westernized",
              z       = "",
              zarrow  = "Non-westernized") 
      
      
      if (F) {
        funcMatTranspose = t(funcMat)
        funcMatTranspose = funcMatTranspose[antismashTerms,]
        etfFuncMat = funcMatTranspose[,colnames(funcMatTranspose) %in% etfOnlySpecies]
        etbFuncMat = funcMatTranspose[,colnames(funcMatTranspose) %in% etbOnlySpecies]
        etpFuncMat = funcMatTranspose[,colnames(funcMatTranspose) %in% etpOnlySpecies]
        
        chisq_FB = getChisqStat(etfFuncMat, etbFuncMat, antismashTerms)
        chisq_BP = getChisqStat(etbFuncMat, etpFuncMat, antismashTerms)
        chisq_FP = getChisqStat(etfFuncMat, etpFuncMat, antismashTerms)
        
        pCut=0.05
        sig_FB = rownames(chisq_FB[!is.nan(chisq_FB$pvalue) & chisq_FB$pvalue<pCut,])
        sig_BP = rownames(chisq_BP[!is.nan(chisq_BP$pvalue) & chisq_BP$pvalue<pCut,])
        sig_FP = rownames(chisq_FP[!is.nan(chisq_FP$pvalue) & chisq_FP$pvalue<pCut,])
        sig_union = unique(c(sig_FB, sig_BP, sig_FP))
        
        sigMat = getFractionMat(etfFuncMat, etbFuncMat, etpFuncMat, sig_union)
        sigMat = as.data.frame(sigMat)
        colnames(sigMat) = c("European","Pacific","NonWesternized")
        sigMat$lab = rownames(sigMat)
        
        require(ggtern)
        #ggtern(data=sigCazyMat, aes(x=F,y=B, z=P, label=lab)) + 
        ggtern(data=sigMat, aes(x=European,y=Pacific, z=NonWesternized, label=lab)) + 
          geom_point(aes(fill=class),
                     size = 2, 
                     shape = 21, 
                     color = "black") +
          scale_fill_manual(values=c("#80808055","#0000ff88","#ff000088","#d4af3788")) +
          theme_rgbw() 
      }
      
    }
    
    vfMode = T
    if(vfMode) {
      #invasionProteins = c("APECO1_2093","t3801", "t0781", "MMAR_2285", "lmo1745", "STM2511")
      invasionProteins = c("MMAR_2285", "APECO1_2093", "Btr_1145", "lmo1847", "t0781", "t3801", "Lmon1_020100004517", "lmo0691")
      intracellularProteins = c("Rv2987c", "lmo0055", "lmo0136", "lmo0137", "lmo0810", "lmo2825", "Lmon1_020100002767", "SL1344_1828", "MMAR_2285")
      
      pCut = 1e-2
      varMat = getVarScoreMat(funcMat, vfTerms, etfSpecies, etbSpecies, etpSpecies)
      sigMat = selectVarScoreMat(varMat, pCut)
      
      colnames(sigMat)[1:3] = c("European","Pacific","NonWesternized")
      sigMat$lab = rownames(sigMat)
      sigMat$Prod = patricVfDescTab$Product[match(sigMat$lab, patricVfDescTab$RefSeq.Locus.Tag)]
      sigMat$Category = patricVfDescTab$Classification[match(sigMat$lab, patricVfDescTab$RefSeq.Locus.Tag)]
      sigMat$class = "others"
      sigMat$class[sigMat$lab %in% invasionProteins] = "invasion"
      sigMat$class[sigMat$lab %in% intracellularProteins] = "intracellular"
      sigMat$class = factor(sigMat$class, levels=c("others","invasion", "intracellular"))
      
      
      # sigMat$class = "others"
      # sigMat$class[sigMat$lab %in% mucinCazymes] = "mucin"
      # sigMat$class[sigMat$lab %in% storageCazymes] = "storage"
      # sigMat$class[sigMat$lab %in% pectinCazymes] = "storage"
      # sigMat$class = factor(sigMat$class, levels=c("others","mucin","storage"))
      View(sigMat)
      
      
      ggtern(data=sigMat, aes(x=European,y=Pacific, z=NonWesternized, label=lab)) + 
        geom_point(aes(fill=class),
                   size = 4, 
                   shape = 21, 
                   color = "#80808011") +
        scale_fill_manual(values=c("#80808033","#0000ff88","#ff000088","#d4af3788")) +
        theme_ggtern(12, "") %+replace% ggplot2::theme_void(12, "") %+replace% theme(tern.axis.text = element_blank(), tern.axis.title = element_blank(), legend.position = "none" ) + 
        labs( x       = "",
              xarrow  = "European",
              y       = "",
              yarrow  = "Westernized",
              z       = "",
              zarrow  = "Non-westernized") 
      
      if (F) {
        funcMatTranspose = t(funcMat)
        funcMatTranspose = funcMatTranspose[vfTerms,]
        etfFuncMat = funcMatTranspose[,colnames(funcMatTranspose) %in% etfOnlySpecies]
        etbFuncMat = funcMatTranspose[,colnames(funcMatTranspose) %in% etbOnlySpecies]
        etpFuncMat = funcMatTranspose[,colnames(funcMatTranspose) %in% etpOnlySpecies]
        
        chisq_FB = getChisqStat(etfFuncMat, etbFuncMat, vfTerms)
        chisq_BP = getChisqStat(etbFuncMat, etpFuncMat, vfTerms)
        chisq_FP = getChisqStat(etfFuncMat, etpFuncMat, vfTerms)
        
        pCut=1e-2
        sig_FB = rownames(chisq_FB[!is.nan(chisq_FB$pvalue) & chisq_FB$pvalue<pCut,])
        sig_BP = rownames(chisq_BP[!is.nan(chisq_BP$pvalue) & chisq_BP$pvalue<pCut,])
        sig_FP = rownames(chisq_FP[!is.nan(chisq_FP$pvalue) & chisq_FP$pvalue<pCut,])
        sig_union = unique(c(sig_FB, sig_BP, sig_FP))
        
        sigMat = getFractionMat(etfFuncMat, etbFuncMat, etpFuncMat, sig_union)
        sigMat = as.data.frame(sigMat)
        colnames(sigMat) = c("European","Pacific","NonWesternized")
        sigMat$lab = rownames(sigMat)
        sigMat$Prod = patricVfDescTab$Product[match(sigMat$lab, patricVfDescTab$RefSeq.Locus.Tag)]
        sigMat$Category = patricVfDescTab$Classification[match(sigMat$lab, patricVfDescTab$RefSeq.Locus.Tag)]
        sigMat$class = "others"
        sigMat$class[sigMat$lab %in% invasionProteins] = "invasion"
        sigMat$class = factor(sigMat$class, levels=c("others","invasion"))
        
        require(ggtern)
        #ggtern(data=sigCazyMat, aes(x=F,y=B, z=P, label=lab)) + 
        ggtern(data=sigMat, aes(x=European,y=Pacific, z=NonWesternized, label=lab)) + 
          geom_point(aes(fill=class),
                     size = 2, 
                     shape = 21, 
                     color = "black") +
          scale_fill_manual(values=c("#80808055","#0000ff88","#ff000088","#d4af3788")) +
          theme_rgbw() 
        
        View(sigMat)
      }
      
    }
    
    jgiMode = F
    if(jgiMode) {
      
      funcMatTranspose = t(funcMat)
      funcMatTranspose = funcMatTranspose[jgiTerms,]
      etfFuncMat = funcMatTranspose[,colnames(funcMatTranspose) %in% etfOnlySpecies]
      etbFuncMat = funcMatTranspose[,colnames(funcMatTranspose) %in% etbOnlySpecies]
      etpFuncMat = funcMatTranspose[,colnames(funcMatTranspose) %in% etpOnlySpecies]
      
      chisq_FB = getChisqStat(etfFuncMat, etbFuncMat, jgiTerms)
      chisq_BP = getChisqStat(etbFuncMat, etpFuncMat, jgiTerms)
      chisq_FP = getChisqStat(etfFuncMat, etpFuncMat, jgiTerms)
      
      pCut=1e-5
      sig_FB = rownames(chisq_FB[!is.nan(chisq_FB$pvalue) & chisq_FB$pvalue<pCut,])
      sig_BP = rownames(chisq_BP[!is.nan(chisq_BP$pvalue) & chisq_BP$pvalue<pCut,])
      sig_FP = rownames(chisq_FP[!is.nan(chisq_FP$pvalue) & chisq_FP$pvalue<pCut,])
      sig_union = unique(c(sig_FB, sig_BP, sig_FP))
      
      sigMat = getFractionMat(etfFuncMat, etbFuncMat, etpFuncMat, sig_union)
      sigMat = as.data.frame(sigMat)
      colnames(sigMat) = c("European","Pacific","NonWesternized")
      sigMat$lab = rownames(sigMat)
      
      
      # sigMat$Prod = patricVfDescTab$Product[match(sigMat$lab, patricVfDescTab$RefSeq.Locus.Tag)]
      # sigMat$Category = patricVfDescTab$Classification[match(sigMat$lab, patricVfDescTab$RefSeq.Locus.Tag)]
      # sigMat$class = "others"
      # sigMat$class[sigMat$lab %in% invasionProteins] = "invasion"
      # sigMat$class = factor(sigMat$class, levels=c("others","invasion"))
      # 
      
      
      
      require(ggtern)
      #ggtern(data=sigCazyMat, aes(x=F,y=B, z=P, label=lab)) + 
      ggtern(data=sigMat, aes(x=European,y=Pacific, z=NonWesternized, label=lab)) + 
        geom_point(#aes(fill=class),
          size = 2, 
          shape = 21, 
          color = "black") +
        scale_fill_manual(values=c("#80808055","#0000ff88","#ff000088","#d4af3788")) +
        theme_void() 
      
      View(sigMat)
    }
    
    arMode = F
    if(arMode) {
      
      
      pCut = 1e-2
      varMat = getVarScoreMat(funcMat, arTerms, etfSpecies, etbSpecies, etpSpecies)
      sigMat = selectVarScoreMat(varMat, pCut)
      
      colnames(sigMat)[1:3] = c("European","Pacific","NonWesternized")
      sigMat$lab = rownames(sigMat)
      
      
      # sigMat$class = "others"
      # sigMat$class[sigMat$lab %in% mucinCazymes] = "mucin"
      # sigMat$class[sigMat$lab %in% storageCazymes] = "storage"
      # sigMat$class[sigMat$lab %in% pectinCazymes] = "storage"
      # sigMat$class = factor(sigMat$class, levels=c("others","mucin","storage"))
      View(sigMat)
      
      
      ggtern(data=sigMat, aes(x=European,y=Pacific, z=NonWesternized, label=lab)) + 
        geom_point(fill="#80808033",#aes(fill=class),
                   size = 4, 
                   shape = 21, 
                   color = "#808080") +
        #scale_fill_manual(values=c("#80808033","#0000ff88","#ff000088","#d4af3788")) +
        theme_void() +
        labs( x       = "",
              xarrow  = "European",
              y       = "",
              yarrow  = "Westernized",
              z       = "",
              zarrow  = "Non-westernized") 
      
      if (F) {
        
      }
      funcMatTranspose = t(funcMat)
      funcMatTranspose = funcMatTranspose[arTerms,]
      etfFuncMat = funcMatTranspose[,colnames(funcMatTranspose) %in% etfOnlySpecies]
      etbFuncMat = funcMatTranspose[,colnames(funcMatTranspose) %in% etbOnlySpecies]
      etpFuncMat = funcMatTranspose[,colnames(funcMatTranspose) %in% etpOnlySpecies]
      
      chisq_FB = getChisqStat(etfFuncMat, etbFuncMat, arTerms)
      chisq_BP = getChisqStat(etbFuncMat, etpFuncMat, arTerms)
      chisq_FP = getChisqStat(etfFuncMat, etpFuncMat, arTerms)
      
      pCut=1e-5
      sig_FB = rownames(chisq_FB[!is.nan(chisq_FB$pvalue) & chisq_FB$pvalue<pCut,])
      sig_BP = rownames(chisq_BP[!is.nan(chisq_BP$pvalue) & chisq_BP$pvalue<pCut,])
      sig_FP = rownames(chisq_FP[!is.nan(chisq_FP$pvalue) & chisq_FP$pvalue<pCut,])
      sig_union = unique(c(sig_FB, sig_BP, sig_FP))
      
      sigMat = getFractionMat(etfFuncMat, etbFuncMat, etpFuncMat, sig_union)
      sigMat = as.data.frame(sigMat)
      colnames(sigMat) = c("European","Pacific","NonWesternized")
      sigMat$lab = rownames(sigMat)
      
      
      # sigMat$Prod = patricVfDescTab$Product[match(sigMat$lab, patricVfDescTab$RefSeq.Locus.Tag)]
      # sigMat$Category = patricVfDescTab$Classification[match(sigMat$lab, patricVfDescTab$RefSeq.Locus.Tag)]
      # sigMat$class = "others"
      # sigMat$class[sigMat$lab %in% invasionProteins] = "invasion"
      # sigMat$class = factor(sigMat$class, levels=c("others","invasion"))
      # 
      
      
      
      require(ggtern)
      #ggtern(data=sigCazyMat, aes(x=F,y=B, z=P, label=lab)) + 
      ggtern(data=sigMat, aes(x=European,y=Pacific, z=NonWesternized, label=lab)) + 
        geom_point(#aes(fill=class),
          size = 2, 
          shape = 21, 
          color = "black") +
        scale_fill_manual(values=c("#80808055","#0000ff88","#ff000088","#d4af3788")) +
        theme_rgbw() 
      
      View(sigMat)
    }
    
    arMode = F
    if(arMode) {
      
      funcMatTranspose = t(funcMat)
      funcMatTranspose = funcMatTranspose[arTerms,]
      etfFuncMat = funcMatTranspose[,colnames(funcMatTranspose) %in% etfOnlySpecies]
      etbFuncMat = funcMatTranspose[,colnames(funcMatTranspose) %in% etbOnlySpecies]
      etpFuncMat = funcMatTranspose[,colnames(funcMatTranspose) %in% etpOnlySpecies]
      
      chisq_FB = getChisqStat(etfFuncMat, etbFuncMat, arTerms)
      chisq_BP = getChisqStat(etbFuncMat, etpFuncMat, arTerms)
      chisq_FP = getChisqStat(etfFuncMat, etpFuncMat, arTerms)
      
      pCut=1e-5
      sig_FB = rownames(chisq_FB[!is.nan(chisq_FB$pvalue) & chisq_FB$pvalue<pCut,])
      sig_BP = rownames(chisq_BP[!is.nan(chisq_BP$pvalue) & chisq_BP$pvalue<pCut,])
      sig_FP = rownames(chisq_FP[!is.nan(chisq_FP$pvalue) & chisq_FP$pvalue<pCut,])
      sig_union = unique(c(sig_FB, sig_BP, sig_FP))
      
      sigMat = getFractionMat(etfFuncMat, etbFuncMat, etpFuncMat, sig_union)
      sigMat = as.data.frame(sigMat)
      colnames(sigMat) = c("European","Pacific","NonWesternized")
      sigMat$lab = rownames(sigMat)
      
      
      # sigMat$Prod = patricVfDescTab$Product[match(sigMat$lab, patricVfDescTab$RefSeq.Locus.Tag)]
      # sigMat$Category = patricVfDescTab$Classification[match(sigMat$lab, patricVfDescTab$RefSeq.Locus.Tag)]
      # sigMat$class = "others"
      # sigMat$class[sigMat$lab %in% invasionProteins] = "invasion"
      # sigMat$class = factor(sigMat$class, levels=c("others","invasion"))
      # 
      
      
      
      require(ggtern)
      #ggtern(data=sigCazyMat, aes(x=F,y=B, z=P, label=lab)) + 
      ggtern(data=sigMat, aes(x=European,y=Pacific, z=NonWesternized, label=lab)) + 
        geom_point(#aes(fill=class),
          size = 2, 
          shape = 21, 
          color = "black") +
        scale_fill_manual(values=c("#80808055","#0000ff88","#ff000088","#d4af3788")) +
        theme_rgbw() 
      
      View(sigMat)
    }
    
  }
  
}

funcCompareByInflowOutflow = T
if (funcCompareByInflowOutflow) {
  
  loadFuncAnnot = T
  if (loadFuncAnnot) {
    
    ### PFAM ###
    igc2PfamMatRData = "j://Deposit/Project/2018_microbiome_atlas/functional.annotation/PFAM/igc2PfamMat.RData"
    load(igc2PfamMatRData)
    pfamTerms = colnames(pfamMat)
    
    ### KEGG ###
    gutKoBestMatRData = "j://Deposit/Project/2018_microbiome_atlas/functional.annotation/gutKoBestMat.1990.RData"
    load(gutKoBestMatRData)
    koTerms = colnames(gutKoBestMat)
    
    
    ### virulence factors ###
    vfMatRData = "j://Deposit/Project/2018_microbiome_atlas/functional.annotation/PATRIC/vfBestMat.RData"
    load(vfMatRData) #vfMatDigit
    vfTerms = colnames(vfMatDigit)
    patricVfDescTabFile = "J://Deposit/Project/2018_microbiome_atlas/functional.annotation/PATRIC/PATRIC_sp_gene.txt"
    patricVfDescTab = read.table(patricVfDescTabFile, header=T, sep="\t", stringsAsFactors = F)
    
    ### JGI phenotype ###
    jgiMatRData = "j://Deposit/Project/2018_microbiome_atlas/functional.annotation/jgiMat.RData"
    load(jgiMatRData)
    jgiTerms = colnames(jgiMat)
    
    ### antibiotic resistance ###
    mustardMatRData = "j://Deposit/Project/2018_microbiome_atlas/functional.annotation/mustard.mat.RData"
    load(mustardMatRData)
    arTerms = colnames(mustardMat)
    
    ### antismash - secondary metabolism ###
    antismashMatRData = "J://Deposit/Project/2018_microbiome_atlas/functional.annotation/msp1992.igc2.antismashMat.RData"
    load(antismashMatRData)
    antismashMelt = melt(antismashMat)
    antismashMelt = antismashMelt[antismashMelt$value!=0,]
    antismashTerms = colnames(antismashMat)
    
    ### Cazy  ### 
    newGutCazyMatRData = "J://Deposit/Project/2018_microbiome_atlas/functional.annotation/CAZY.INRA/igc2.new.cazy.mat.RData"
    load(newGutCazyMatRData)
    colnames(cazyMat) = paste("cazy", colnames(cazyMat), sep=".")
    cazyTerms = colnames(cazyMat)
    
  }
  
  loadFuncMode = T
  if (loadFuncMode) {
    funcMatRData = "j://Deposit/Project/2018_microbiome_atlas/functional.annotation/funcMat.20190808.RData"
    load(funcMatRData)
  }
  
  loadInOutFlowStat = T
  if (loadInOutFlowStat) {
    markovStatTabFinalFile = "C://Data//comparative.analysis.healthy.sweden//markovStatTab.final.3.txt"
    markovStatTabFinal = read.table(markovStatTabFinalFile, sep="\t", stringsAsFactors = F)
    markovStatTabFinal$GssOss = NA
    markovStatTabFinal$GssOss[rownames(markovStatTabFinal) %in% gssSpecies] = "GSS"
    markovStatTabFinal$GssOss[rownames(markovStatTabFinal) %in% ossSpecies] = "OSS"
    #write.table(markovStatTabFinal, markovStatTabFinalFile, sep="\t")
    
    gutTaxoSel = gutTaxo[rownames(markovStatTabFinal),]
    gutGenusMat = acast(data=gutTaxoSel, MSP ~ genus )
    gutFamilyMat = acast(data=gutTaxoSel, MSP ~ family )
    gutClassMat = acast(data=gutTaxoSel, MSP ~ class )
    gutPhylumMat = acast(data=gutTaxoSel, MSP ~ phylum )
    gutGenusMat[is.na(gutGenusMat)] = 0
    gutFamilyMat[is.na(gutFamilyMat)] = 0
    gutClassMat[is.na(gutClassMat)] = 0
    gutPhylumMat[is.na(gutPhylumMat)] = 0
    gutGenusMat = gutGenusMat[,!grepl("unclassified",colnames(gutGenusMat))]
    gutFamilyMat = gutFamilyMat[,!grepl("unclassified",colnames(gutFamilyMat))]
    gutClassMat = gutClassMat[,!grepl("unclassified",colnames(gutClassMat))]
    gutPhylumMat = gutPhylumMat[,!grepl("unclassified",colnames(gutPhylumMat))]
    
    getSlideSums <- function(input, wd=10) {
      output = rep(0, length(input))
      slideSums = rollapply(input, width=wd, by=1, FUN=mean, align="left", partial=T)
      output[seq(slideSums)] = slideSums
      return(output)
    }
    getSignatures <- function(targetFlows, funcMatNew, currFcs, wd=10) {
      fcTab = data.frame(targetflow=targetFlows,
                         funcMatNew[,currFcs])
      set.seed(1)
      fcTab = fcTab[sample(nrow(fcTab)),]
      fcTab = fcTab[order(fcTab[,"targetflow"], decreasing = T),]
      #fcTab$slideSum = getSlideSums(fcTab[,"func"], wd)
      fcTab[,-1] = apply(fcTab[,-1], 2, function(x) getSlideSums(x, wd))
      fcTab = fcTab[order(fcTab[,"targetflow"], decreasing = F),]
      #print(head(fcTab))
      fcLevelInput = t(as.matrix(fcTab))
      #rownames(fcLevelInput) = c("Flow","Function", "slideSum")
      colnames(fcLevelInput) = NULL
      #outMat = fcLevelInput[c("Flow","slideSum"),]
      return(fcLevelInput)
      
    }
    getSignature <- function(targetFlows, funcMatNew, currFc, wd=10) {
      fcTab = data.frame(targetflow=targetFlows,
                         func = funcMatNew[,currFc])
      set.seed(1)
      fcTab = fcTab[sample(nrow(fcTab)),]
      fcTab = fcTab[order(fcTab[,"targetflow"], decreasing = T),]
      fcTab$slideSum = getSlideSums(fcTab[,"func"], wd)
      fcTab = fcTab[order(fcTab[,"targetflow"], decreasing = F),]
      #print(head(fcTab))
      fcLevelInput = t(as.matrix(fcTab))
      rownames(fcLevelInput) = c("Flow","Function", "slideSum")
      colnames(fcLevelInput) = NULL
      outMat = fcLevelInput[c("Flow","slideSum"),]
      return(outMat)
      
    }
    plotSignature <- function(outMat, as = 5) {
      out = levelplot(outMat, aspect = as, colorkey=F,
                      xlab="",ylab="",axes=F,
                      scales = list(tck = c(0,0), x=list(rot=90), draw=F),
                      col.regions = colorRampPalette(c("white","black"))(512))
      return(out)
    }
    
    
    
    #funcMatNew = funcMat[match(rownames(markovStatTabFinal), rownames(funcMat)),]
    #table(rownames(markovStatTabFinal) %in% rownames(funcMat))
    #all 1411 TRUE
    
    mucinCazymes = c("cazy.GT27", "cazy.GT14", "cazy.GT11", "cazy.GT10", "cazy.GH89", "cazy.GH84", "cazy.GH33", "cazy.GH20", "cazy.GH29", "cazy.GH18", "cazy.GH123", "cazy.GH109", "cazy.CBM51", "cazy.CBM50", "cazy.CBM32")
    storageCazymes = c("cazy.CBM34", "cazy.CBM48", "cazy.CBM20", "cazy.GH13", "cazy.GH133", "cazy.GH32", "cazy.GH36", "cazy.GH37", "cazy.GH77", "cazy.GH57", "cazy.GT101", "cazy.GT26", "cazy.GT3", "cazy.GT35", "cazy.GT5")
    pectinCazymes = c("cazy.CBM77","cazy.CE8","cazy.GH105","cazy.GH28", "cazy.PL1")
    
    if (F) {
      targetTerms = vfTerms
      targetTerms = arTerms
      targetTerms = antismashTerms
      targetTerms = cazyTerms
      targetTerms = pectinCazymes
      targetTerms = jgiTerms
      targetTerms = mgePfamTerms
      targetTerms = pfamTerms
      targetTerms = koTerms
      
      table(lmStatInMat[,"AdjP"]< adjCut) # 2086 terms were significant
      table(lmStatOutMat[,"AdjP"]< adjCut) # 294 terms were significant
      
      adjCut=1e-5
      table(rownames(lmStatInMat[lmStatInMat[,"AdjP"]< adjCut,]) %in% targetTerms)
      table(rownames(lmStatOutMat[lmStatOutMat[,"AdjP"]< adjCut,]) %in% targetTerms)
      
      
    }
    
    inflowAssocFunctionMat <- function(markovStatTabFinal, funcMat, targetFuncs = NULL) {
      funcMatNew = funcMat[match(rownames(markovStatTabFinal), rownames(funcMat)),]
      if (is.null(targetFuncs)) funcs = colnames(funcMatNew) 
      if (!is.null(targetFuncs)) funcs = targetFuncs
      
      inflows = markovStatTabFinal$inflow
      outflows = markovStatTabFinal$outflow
      names(inflows) = rownames(markovStatTabFinal)
      names(outflows) = rownames(markovStatTabFinal)
      
      
      lmStatInMat = do.call(rbind, sapply(funcs, function(fc) {
        funcProfiles = funcMatNew[,fc]
        lmOut  = lm(funcProfiles~inflows)
        pvalue = summary(lmOut)$coefficient["inflows","Pr(>|t|)"]
        estimate = summary(lmOut)$coefficient["inflows","Estimate"]
        return(c(estimate, pvalue))
        
      }, simplify = F))
      lmStatInMat = cbind(lmStatInMat, p.adjust(lmStatInMat[,2], "BH"))
      colnames(lmStatInMat) = c("Estimate", "Pvalue", "AdjP")
      print(dim(lmStatInMat))
      lmStatInMat = lmStatInMat[lmStatInMat[,"Estimate"]>0,]
      
      print(dim(lmStatInMat))
      return(lmStatInMat)
      #head(lmStatInMat[order(lmStatInMat[,"Pvalue"]),],n=20)
      View(lmStatInMat[lmStatInMat[,"AdjP"]< 1e-10,])
      tt = rownames(lmStatInMat[lmStatInMat[,"AdjP"]< 1e-10,])
      table(tt %in% vfTerms)
      
      View(data.frame(tt,getDescFromKoTerms(tt, koDescMap)))
      
      if (F) {
        lmStatInMat2 = lmStatInMat[rownames(lmStatInMat) %in% arTerms,]
        lmStatInMat2 = lmStatInMat[rownames(lmStatInMat) %in% vfTerms,]
        lmStatInMat2 = lmStatInMat[rownames(lmStatInMat) %in% mgePfamTerms,]
        lmStatInMat2 = lmStatInMat[rownames(lmStatInMat) %in% cazyTerms,]
        lmStatInMat2 = lmStatInMat[rownames(lmStatInMat) %in% storageCazymes,]
        lmStatInMat2 = lmStatInMat[rownames(lmStatInMat) %in% mucinCazymes,]
        lmStatInMat2 = lmStatInMat[rownames(lmStatInMat) %in% jgiTerms,]
        lmStatInMat2 = lmStatInMat[rownames(lmStatInMat) %in% antismashTerms,]
        
        
        head(lmStatInMat2[order(lmStatInMat2[,"AdjP"]),],n=20)
      }
      if (F) {
        
        currFc = "K18220"
        currFc = "K09759"
        currFc = "K06404"
        
        currFc = "Integrase_DNA"
        currFc = "MobC"
        currFc = "cazy.GT32"
        #currFc = "tetM"
        currFc = "Nonsporulating"
        currFcs = c("Anaerobe","Integrase_DNA", "MobC", "bacteriocin", "van", "cazy.GT5")
        currFcs = c("Integrase_DNA","MobC", "cazy.GT5")
        currFcs = c("Integrase_DNA","MobC")
        
        plotSignature(getSignatures(inflows, funcMatNew, currFcs, 100), 2)#, 1.2
        
        levelplot(getSignatures(outflows, funcMatNew, currFcs, 100), aspect = 1.5, 
                  xlab="",ylab="",axes=F,
                  scales = list(tck = c(0,0), x=list(rot=90), draw=F),
                  col.regions = colorRampPalette(c("white","black"))(512))
        
        
      }
      if (F) {
        
        
        plotSignature(getSignature(inflows, funcMatNew, currFc, 100))
        
        checkBoxPlot = F
        if (checkBoxPlot) {
          integraseBySplit = split(inflows, funcMatNew[,"Integrase_DNA"])
          boxplot(integraseBySplit)
        }
        
      }
      
      return(lmStatInMat)
    }
    outflowAssocFunctionMat <- function(markovStatTabFinal, funcMat, targetFuncs = NULL) {
      funcMatNew = funcMat[match(rownames(markovStatTabFinal), rownames(funcMat)),]
      if (is.null(targetFuncs)) funcs = colnames(funcMatNew) 
      if (!is.null(targetFuncs)) funcs = targetFuncs
      
      outflows = markovStatTabFinal$outflow
      inflows = markovStatTabFinal$inflow
      
      lmStatOutMat = do.call(rbind, sapply(funcs, function(fc) {
        funcProfiles = funcMatNew[,fc]
        lmOut  = lm(funcProfiles~outflows)
        pvalue = summary(lmOut)$coefficient["outflows","Pr(>|t|)"]
        estimate = summary(lmOut)$coefficient["outflows","Estimate"]
        return(c(estimate, pvalue))
        
      }, simplify = F))
      lmStatOutMat = cbind(lmStatOutMat, p.adjust(lmStatOutMat[,2], "BH"))
      colnames(lmStatOutMat) = c("Estimate", "Pvalue", "AdjP")
      lmStatOutMat = lmStatOutMat[lmStatOutMat[,"Estimate"]>0,]
      #head(lmStatOutMat[order(lmStatOutMat[,"AdjP"]),],n=20)
      return(lmStatOutMat)
      
      if (F) {
        lmStatOutMat2 = lmStatOutMat[rownames(lmStatOutMat) %in% arTerms,]
        lmStatOutMat2 = lmStatOutMat[rownames(lmStatOutMat) %in% cazyTerms,]
        lmStatOutMat2 = lmStatOutMat[rownames(lmStatOutMat) %in% vfTerms,]
        lmStatOutMat2 = lmStatOutMat[rownames(lmStatOutMat) %in% jgiTerms,]
        lmStatOutMat2 = lmStatOutMat[rownames(lmStatOutMat) %in% mgePfamTerms,]
        lmStatOutMat2 = lmStatOutMat[rownames(lmStatOutMat) %in% antismashTerms,]
        lmStatOutMat2 = lmStatOutMat[rownames(lmStatOutMat) %in% koTerms,]
        
        head(lmStatOutMat2[order(lmStatOutMat2[,"AdjP"]),],n=20)
        
        
      }
      
      if (F) {
        
        
        
        currFc = "siderophore"
        currFc = "t2pks-ladderane"
        currFc = "K00867"
        currFc = "lmo1003"
        #lmo1267 --> cell division trigger factor
        currFcs = c("Facultative", "Microaerophilic", "lmo1267", "K02248")
        
        plotSignature(getSignatures(outflows, funcMatNew, currFcs, 100), 1.5)
        
      }
      
      checkBoxPlot = F
      if (checkBoxPlot) {
        integraseBySplit = split(inflows, funcMatNew[,"Integrase_DNA"])
        boxplot(integraseBySplit)
      }
      
      return(lmStatOutMat)
    } 
    
    taxoMat = gutGenusMat
    inflowAssocTaxoMat <- function(markovStatTabFinal, taxoMat, targetFuncs = NULL) {
      taxoMatNew = taxoMat[match(rownames(markovStatTabFinal), rownames(taxoMat)),]
      if (is.null(targetFuncs)) taxa = colnames(taxoMatNew) 
      if (!is.null(targetFuncs)) taxa = targetTaxa
      
      inflows = markovStatTabFinal$inflow
      outflows = markovStatTabFinal$outflow
      
      lmStatInTaxaMat = do.call(rbind, sapply(taxa, function(currTaxo) {
        taxaProfiles = taxoMatNew[,currTaxo]
        lmOut  = lm(taxaProfiles~inflows)
        pvalue = summary(lmOut)$coefficient["inflows","Pr(>|t|)"]
        estimate = summary(lmOut)$coefficient["inflows","Estimate"]
        return(c(estimate, pvalue))
        
      }, simplify = F))
      lmStatInTaxaMat = cbind(lmStatInTaxaMat, p.adjust(lmStatInTaxaMat[,2], "BH"))
      colnames(lmStatInTaxaMat) = c("Estimate", "Pvalue", "AdjP")
      lmStatInTaxaMat = lmStatInTaxaMat[lmStatInTaxaMat[,"Estimate"]>0,]
      #head(lmStatInTaxaMat[order(lmStatInTaxaMat[,"Pvalue"]),],n=20)
      return(lmStatInTaxaMat)
      if (F) {
        
        currFc = "Lachnospiraceae"
        
        fcTab = data.frame(inflow=inflows,
                           func = taxoMatNew[,currFc])
        fcTab = fcTab[order(fcTab[,"inflow"], decreasing = T),]
        fcTab$slideSum = getSlideSums(fcTab[,"func"], 50)
        #fcTab$slideSum = (fcTab[,"func"])
        fcTab = fcTab[order(fcTab[,"inflow"], decreasing = F),]
        fcLevelInput = t(as.matrix(fcTab))
        rownames(fcLevelInput) = c("Inflow","Function", "slideSum")
        colnames(fcLevelInput) = NULL
        levelplot(fcLevelInput[c("Inflow","slideSum"),], aspect = 5, colorkey=F,
                  xlab="",ylab="",axes=F,
                  scales = list(tck = c(0,0), x=list(rot=90), draw=F),
                  col.regions = colorRampPalette(c("white","black"))(512))
        
        checkBoxPlot = F
        if (checkBoxPlot) {
          integraseBySplit = split(inflows, funcMatNew[,"Integrase_DNA"])
          boxplot(integraseBySplit)
        }
        
      }
      
      # currTaxo = "Lachnospiraceae"
      # plotSignature(getSignature(inflows, taxoMatNew, currTaxo, 100))
      # 
      currTaxa = c("Dorea", "Faecalibacterium")
      #plotSignature(getSignature(inflows, taxoMatNew, currTaxo, 50))
      plotSignature(getSignatures(inflows, taxoMatNew, currTaxa, 50), 2)
      
      return(lmStatInTaxaMat)
    }
    outflowAssocTaxoMat <- function(markovStatTabFinal, taxoMat, targetFuncs = NULL) {
      taxoMatNew = taxoMat[match(rownames(markovStatTabFinal), rownames(taxoMat)),]
      if (is.null(targetFuncs)) taxa = colnames(taxoMatNew) 
      if (!is.null(targetFuncs)) taxa = targetTaxa
      
      inflows = markovStatTabFinal$inflow
      outflows = markovStatTabFinal$outflow
      names(inflows) = rownames(markovStatTabFinal)
      names(outflows) = rownames(markovStatTabFinal)
      
      lmStatOutTaxaMat = do.call(rbind, sapply(taxa, function(currTaxo) {
        taxaProfiles = taxoMatNew[,currTaxo]
        lmOut  = lm(taxaProfiles~outflows)
        pvalue = summary(lmOut)$coefficient["outflows","Pr(>|t|)"]
        estimate = summary(lmOut)$coefficient["outflows","Estimate"]
        return(c(estimate, pvalue))
        
      }, simplify = F))
      lmStatOutTaxaMat = cbind(lmStatOutTaxaMat, p.adjust(lmStatOutTaxaMat[,2], "BH"))
      colnames(lmStatOutTaxaMat) = c("Estimate", "Pvalue", "AdjP")
      lmStatOutTaxaMat = lmStatOutTaxaMat[lmStatOutTaxaMat[,"Estimate"]>0,]
      #head(lmStatOutTaxaMat[order(lmStatOutTaxaMat[,"Pvalue"]),],n=20)
      return(lmStatOutTaxaMat)
      
      currTaxo = "Clostridiaceae"
      currTaxo = "Faecalicatena"
      currTaxo = "Clostridium"
      
      currTaxa = c("Faecalicatena", "Clostridium")
      #plotSignature(getSignature(outflows, taxoMatNew, currTaxo, 50))
      plotSignature(getSignatures(outflows, taxoMatNew, currTaxa, 50), 2)
      
      checkBoxPlot = F
      if (checkBoxPlot) {
        integraseBySplit = split(inflows, funcMatNew[,"Integrase_DNA"])
        boxplot(integraseBySplit)
      }
      
      return(lmStatOutTaxaMat)
      
    }
    
    inflowFuncStats = inflowAssocFunctionMat(markovStatTabFinal, funcMat)
    outflowFuncStats = outflowAssocFunctionMat(markovStatTabFinal, funcMat)
    
    inflowTaxaStats = inflowAssocTaxoMat(markovStatTabFinal, taxoMat)
    outflowTaxaStats = outflowAssocTaxoMat(markovStatTabFinal, taxoMat)
    
    if (F) {
      write.table(inflowFuncStats, "inflowFuncStats.txt", sep="\t")
      write.table(outflowFuncStats, "outflowFuncStats.txt", sep="\t")
      
      write.table(inflowTaxaStats, "inflowTaxaStats.txt", sep="\t")
      write.table(outflowTaxaStats, "outflowTaxaStats.txt", sep="\t")
      
    }
    
  }
  
  dim(inflowFuncStats)
  dim(outflowFuncStats)
  
  checkEnrichedFunctions <- function(funcList, statTab, cutOff = 1e-5, check=F ) {
    funcNames = names(funcList)
    
    statSelTab = statTab[statTab[,"AdjP"] <= cutOff,]
    allTerms = rownames(statTab)
    sigTerms = rownames(statSelTab)
    
    if (check) funcList = lapply(funcList, function(x) x[x %in% allTerms])
    funcFractions = do.call(c, lapply(funcList, function(x) {
      meanValues = mean(x %in% sigTerms)*100
      return(meanValues)
    }))
    return(funcFractions)
  } 
  
  #Com_YlbF --> regulator
  #MecA --> negative regulator
  compDomains = c("CinA", "Competence", "CoiA", "Com_YlbF", "ComJ", "MecA", "ComFB", "ComGF" )
  
  funcListForInOutflow = list(Virulence =vfTerms, 
                              mucin = mucinCazymes,
                              #competence=compDomains,
                              MGE=mgePfamTerms,
                              Storage = storageCazymes,
                              carbLipidTrans=carbLipidTransportKOs,
                              Vitamin=vitaminKOs,
                              aminoAcid=aminoAcidKOs)
  
  PTS = koBySubsystemList[["Phosphotransferase system (PTS)"]]
  
  # efflux = koBySubsystemList[["Drug efflux transporter/pump"]],
  # PTS = koBySubsystemList[["Phosphotransferase system (PTS)"]],
  # 
  
  
  cutOff = 1e-3
  inflowEnriched = checkEnrichedFunctions(funcListForInOutflow, inflowFuncStats, cutOff)
  outflowEnriched = checkEnrichedFunctions(funcListForInOutflow, outflowFuncStats, cutOff)
  testTab = rbind(data.frame(value = inflowEnriched, flowType="inflow",type= names(funcListForInOutflow), stringsAsFactors = F),
                  data.frame(value = outflowEnriched, flowType="outflow",type= names(funcListForInOutflow), stringsAsFactors = F))
  
  testTab$type[testTab$type=="efflux"] = "Drug efflux"
  testTab$type[testTab$type=="carbLipidTrans"] = "Nutrient transport"
  testTab$type[testTab$type=="Storage"] = "Storage carb."
  testTab$type[testTab$type=="aminoAcid"] = "Amino acid"
  
  
  ggplot(testTab) +    
    geom_bar(aes(x=as.factor(type), y=value, fill=flowType), stat="identity", alpha=0.5) +
    scale_fill_viridis(discrete=TRUE)  + coord_polar() +
    xlab("") + ylab("") +
    theme(
      panel.background = element_rect(fill = "white", colour = "black"),
      panel.grid.major = element_line(size = 0.5, linetype = 'solid',colour = "gray"), 
      legend.position = "none",
      axis.text = element_blank(),
      axis.title = element_blank(),
      axis.ticks = element_blank(),
      panel.grid = element_blank(),
      plot.margin = unit(rep(1,4), "cm")
    ) 
  
  accesoryTab =  data.frame(stringsAsFactors=FALSE,
                            value = c(5.952381, 4.545455, 5, 5.555555556, 14.70588235, 25, 43.68932039, 
                                      10.714286, 20.454545, 35, 50, 38.23529412, 43, 52.81553398),
                            flowType = c("inflow", "inflow", "inflow", "inflow", "inflow", "inflow", "inflow", 
                                         "outflow", "outflow", "outflow", "outflow", "outflow", "outflow", "outflow"),
                            type = c("Efflux","PTS", "Toluene", "Fluorobenzoate", "Xylene", "Benzoate", "ABC transporter", 
                                     "Efflux","PTS",  "Toluene", "Fluorobenzoate", "Xylene", "Benzoate", "ABC transporter"))
  ggplot(accesoryTab) +    
    geom_bar(aes(x=as.factor(type), y=value, fill=flowType), stat="identity", alpha=0.5) +
    scale_fill_viridis(discrete=TRUE)  + coord_polar() +
    xlab("") + ylab("") +
    theme(
      panel.background = element_rect(fill = "white", colour = "black"),
      panel.grid.major = element_line(size = 0.5, linetype = 'solid',colour = "gray"), 
      legend.position = "none",
      axis.text = element_blank(),
      axis.title = element_blank(),
      axis.ticks = element_blank(),
      panel.grid = element_blank(),
      plot.margin = unit(rep(1,4), "cm")
    ) 
  
  ggplot(accesoryTab) +    
    geom_bar(aes(x=as.factor(type), y=value, fill=flowType), stat="identity", alpha=0.5) +
    scale_fill_viridis(discrete=TRUE)  + coord_polar() +
    xlab("") + ylab("")
  # theme(
  #   legend.position = "none",
  #   axis.text = element_blank(),
  #   axis.title = element_blank(),
  #   panel.grid = element_blank(),
  #   plot.margin = unit(rep(-1,4), "cm") 
  # ) 
  # 
}

checkSThermophilusAndSalivarius = F
if (checkSThermophilusAndSalivarius) {
  wellnessGctNormSThermophilusSalivariusRData = "J:\\Deposit\\Project\\2018_microbiome_atlas\\gctTables\\Save.S.thermophilus.salivarius.gct.norm.out.RData"
  load(wellnessGctNormSThermophilusSalivariusRData)
  plotBarcode(gctNormSThermophilusMsp0833)
  plotBarcode(gctNormSSalivariusMsp0380)
}

longitudinalMode = T
if (longitudinalMode) {
  #write.table(gutTaxo, "gut.igc2.1989.taxonomy.for.saeed.txt", sep="\t", quote = F)
  
  parseClinMetaData = T
  if (parseClinMetaData) {
    wellnessClinData = "J://Deposit//Project//2017.Wellness//clinical.meta.data//20170822//wellness.meta.data.extended.99.subjects.170626.txt"
    wellnessClinTab = read.delim(wellnessClinData, sep="\t", stringsAsFactors = F)
    colnames(wellnessClinTab)[1] = "subject_id"
    
    
    ###ongoing
    meltClinMatrix <- function(wellnessClinTab) {
      defaultFields = c("subject_id","scapis_id","SUBJID","birthdate","height","imt.cca","blood.type","atherosclerosis","age.visit1")
      exFields = c("type","fasting","datetime","directanalysis","biobanksamples","faeces","other_smoking","other_smoking_hours","other_antibiotics","other_antibiotic_details","other_health_change","other_health_change_details")
      defaultTab = wellnessClinTab[,defaultFields]
      #View(defaultTab[defaultTab$atherosclerosis==2,])
      
      clinVisitTab = wellnessClinTab[,!colnames(wellnessClinTab) %in% defaultFields]
      rownames(clinVisitTab) = paste("X", wellnessClinTab$subject_id, sep="")
      
      clinVisitExMatch = sapply(colnames(clinVisitTab), function(currColumn){
        matched = sapply(exFields, function(exField) {
          return(grepl(exField, currColumn))
        })
        matched = any(matched)
        return(matched)
      })
      
      clinVisitTab = clinVisitTab[,!clinVisitExMatch]
      
      clinVisitIds = sapply(colnames(clinVisitTab), getVisitIds)
      clinVisitIdsUnique = unique(clinVisitIds)
      clinVisitFields = sapply(colnames(clinVisitTab), getFields)
      clinVisitFieldsUnique = unique(clinVisitFields)
      
      
      #subject_visit by clinical parameters
      
      mergedClinVisitTab = do.call(rbind, sapply(clinVisitFieldsUnique, function(currField) {
        
        data = clinVisitTab[,clinVisitFields==currField]
        visitInfo = clinVisitIds[clinVisitFields==currField]
        visitInfo = gsub("isit","",visitInfo)
        colnames(data) = visitInfo
        data = as.matrix(data)
        melted = melt(data)
        melted$field = currField
        melted$Var3 = paste(melted$Var1, melted$Var2, sep = "_")
        return(melted)
      }, simplify = F))
      
      mergedClinVisitMat = acast(mergedClinVisitTab, Var3 ~ field, value.var = "value")
      dim(mergedClinVisitMat)
      
      return(mergedClinVisitMat)
    }
    
    clinicalMatrix = meltClinMatrix(wellnessClinTab)
    colnames(clinicalMatrix) = gsub("analysis_code_","",colnames(clinicalMatrix))
    colnames(clinicalMatrix) = gsub("_result","",colnames(clinicalMatrix))
    #cbind(colnames(clinicalMatrix))
    
    
    checkBMI = F
    if (checkBMI) {
      clinicalVisits = gsub("X[0-9]+_","",rownames(clinicalMatrix))
      bmi = clinicalMatrix[,"anthropometry_weight"] / ( clinicalMatrix[,"anthropometry_height"]/100)^2
      
      clinicalMatrixNew = cbind(clinicalMatrix,bmi)
      bmiSplit = split ( bmi, clinicalVisits)
      
      wellnessBmi = wellnessMetadata$BMI
      names(wellnessBmi) = wellnessMetadata$sample.ID
      
      wellnessAge = wellnessMetadata$Age
      names(wellnessAge) = wellnessMetadata$sample.ID
    }
    
    correctBMI = F # wellness BMI has been corrected --> 2020.06.19
    if (correctBMI) {
      
      wellnessSampleIds = basicMetaMapUpdated$sample.ID [basicMetaMapUpdated$dataset.ID=="id28"]
      wellnessBMIs = basicMetaMapUpdated$BMI [basicMetaMapUpdated$dataset.ID=="id28"]
      
      
      
      correctSampleIds = rownames(clinicalMatrix)
      correctBMIs = clinicalMatrix[,"anthropometry_weight"] / ( clinicalMatrix[,"anthropometry_height"]/100)^2
      
      #wellnessSampleIds[!wellnessSampleIds %in% correctSampleIds]
      #"X3016_v1" "X3016_v2" "X3913_v1"
      correct_bmi_X3016_v1 = 72.5 / (158/100)^2
      correct_bmi_X3016_v2 = 72.8 / (158/100)^2
      correct_bmi_X3913_v1 = 59.4 / (161/100)^2
      
      wellnessSampleIdsMatched = wellnessSampleIds[wellnessSampleIds %in% correctSampleIds]
      wellnessSampleIdsNotMatched = wellnessSampleIds[!wellnessSampleIds %in% correctSampleIds]
      
      correctSampleIdsMatched = correctSampleIds[correctSampleIds %in% wellnessSampleIdsMatched]
      correctBMIsMatched = correctBMIs[correctSampleIds %in% wellnessSampleIdsMatched]
      
      basicMetaMapUpdated$BMI[match(correctSampleIdsMatched, basicMetaMapUpdated$sample.ID)] = correctBMIsMatched
      basicMetaMapUpdated$BMI[basicMetaMapUpdated$sample.ID == "X3016_v1"] = correct_bmi_X3016_v1
      basicMetaMapUpdated$BMI[basicMetaMapUpdated$sample.ID == "X3016_v2"] = correct_bmi_X3016_v2
      basicMetaMapUpdated$BMI[basicMetaMapUpdated$sample.ID == "X3913_v1"] = correct_bmi_X3913_v1
      
      save(basicMetaMapUpdated, file = basicMetaCleanRDataUpdated2)
      
      
    }
    
    
    #save(clinicalMatrix, file="J://Deposit//Project//2018_microbiome_atlas//microbiome.atlas.Rproj//clinicalMatrix.RData")
    
    if (F) {
      clinicalMelt = melt(clinicalMatrix)
      colnames(clinicalMelt) = c("sample","clinical","value")
      clinicalMelt$subject = gsub("_v[1-4]","",clinicalMelt$sample)
      
      matchedMgsMat = mergeMatUpdated[,clinicalMelt$sample]
      #View(clinicalMatrix)
    }
    
    
    loadData = T
    if (loadData) {
      wellnessMetadata = basicMetaMapUpdated[basicMetaMapUpdated$dataset.ID=="id28",]
      
      mgsMat = mergeMatUpdated
      metaTab = basicMetaMapUpdated
      
      wellnessMetadata = metaTab[metaTab$dataset.ID=="id28",]
      wellnessSubjs = (wellnessMetadata$metadata.ID)
      missingSubjs1 = names(table(wellnessSubjs)[table(wellnessSubjs)==3])
      exceptionSubjs = c("X3452","X3834","X3918","X4215")
      missingSubjs2 = c("X3913","X3535","X3016")
      missingSubjs1 = missingSubjs1[!missingSubjs1 %in% exceptionSubjs]
      
      wellnessMetadata = wellnessMetadata[!wellnessMetadata$metadata.ID %in% missingSubjs1, ]
      wellnessMetadata = wellnessMetadata[!wellnessMetadata$metadata.ID %in% missingSubjs2, ]
      wellnessMetadata = wellnessMetadata[!wellnessMetadata$metadata.ID %in% exceptionSubjs, ]
      wellnessMgsMat = mergeMatUpdated[,match(wellnessMetadata$sample.ID, colnames(mergeMatUpdated))]
      wellnessMgsMatDigit = (wellnessMgsMat>0) * 1
      wellnessMgsMeans = rowMeans(wellnessMgsMat)
      wellnessSamples = wellnessMetadata$sample.ID
      wellnessSubjects = gsub("_v[1-4]","",wellnessSamples)
      wellnessVisits = gsub("X[0-9]+_","",wellnessSamples)
      wellnessSamplesBySubjects = split(wellnessSamples, wellnessSubjects)
      wellnessUniqSubjs = unique(wellnessMetadata$metadata.ID)
      wellnessClinicalMatrix = t(clinicalMatrix)[,match(wellnessSamples, rownames(clinicalMatrix))]
      
      checkSummaryClinical = T
      if (checkSummaryClinical) {
        
        wellnessClinicalMatrix
        
        
      }
      individualLevel = F
      if (individualLevel) {
        
        indMatList = sapply(wellnessUniqSubjs, function(uniqSubj) {
          currSamples = paste(uniqSubj, c("v1","v2","v3","v4"), sep="_")
          currMat = wellnessMgsMat[,currSamples]
          currMgsInds = apply(currMat, 1, function(x) any(x>0))
          return(names(currMgsInds[currMgsInds]))
          
        }, simplify = F)
        allSpeciesNumbersPerIndividual = lapply(indMatList, length)
        hist(unlist(allSpeciesNumbersPerIndividual), main="", las=1)
        subMatList = sapply(wellnessUniqSubjs, function(uniqSubj) {
          currSamples = paste(uniqSubj, c("v1","v2","v3","v4"), sep="_")
          currMat = wellnessMgsMat[,currSamples]
          currMatDigit = currMat > 0
          currMatJaccard = 1-jaccard(currMatDigit)
          
          
        }, simplify = F)
        
        subMatAvg = Reduce("+", subMatList)
        subMatAvg = subMatAvg / length(wellnessUniqSubjs)
        colnames(subMatAvg) = c("V1","V2","V3","V4")
        rownames(subMatAvg) = c("V1","V2","V3","V4")
        
        heatmap.2(subMatAvg, 
                  lwid = c(2,8),
                  lhei = c(2,8),
                  key = F, 
                  key.title = "", key.ylab = "", key.xlab = "",
                  key.xtickfun = function() {
                    breaks = pretty(parent.frame()$breaks)
                    breaks = breaks[c(1,length(breaks))]
                    list(at = parent.frame()$scale01(breaks),
                         labels = breaks)
                  },
                  key.ytickfun = function() {
                    breaks = pretty(parent.frame()$breaks)
                    breaks = breaks[c(1,length(breaks))]
                    list(at = parent.frame()$scale01(breaks),
                         labels = breaks)
                  },
                  cellnote=as.matrix(round(subMatAvg, digits=3)),
                  notecol="black",
                  key.par = list(cex=1.0),
                  cexRow = 1.7, cexCol = 1.7,
                  col=colorRampPalette(c("white","skyblue"))(10),
                  trace="none",
                  sepcolor = "black",
                  colsep=0:4,
                  rowsep=0:4,
                  sepwidth = c(0.01,0.01))
        if (F) {
          diag(subMatAvg) = NA
          col <- colorRampPalette(c("#BB4444", "#EE9988", "#FFFFFF", "#77AADD", "#4477AA"))
          col3 <- colorRampPalette(c("white", "blue")) 
          corrplot(subMatAvg, type = "upper", order = "hclust",
                   col = col3(2), cl.lim = c(0.73, 1),
                   cl.pos = "n", 
                   method="number",
                   diag=F,bg = "lightblue")
        }
        
      }
      
      unionLevel = F
      if (unionLevel) {
        v1Sums = rowSums(wellnessMgsMat[,wellnessVisits=="v1"] )
        v2Sums = rowSums(wellnessMgsMat[,wellnessVisits=="v2"] )
        v3Sums = rowSums(wellnessMgsMat[,wellnessVisits=="v3"] )
        v4Sums = rowSums(wellnessMgsMat[,wellnessVisits=="v4"] )
        vAllSums = rowSums(wellnessMgsMat)
        
        v1Digit = v1Sums > 0
        v2Digit = v2Sums > 0
        v3Digit = v3Sums > 0
        v4Digit = v4Sums > 0
        
        visitDigitMat = cbind(v1Digit, 
                              v2Digit, 
                              v3Digit, 
                              v4Digit)
        visitJaccard = 1-jaccard(visitDigitMat)
        colnames(visitJaccard) = c("V1","V2","V3","V4")
        rownames(visitJaccard) = c("V1","V2","V3","V4")
        
        heatmap.2(visitJaccard, 
                  key = T, key.title = "", key.ylab = "", key.xlab = "",
                  cexRow = 2, cexCol = 2,
                  col=colorRampPalette(c("white","skyblue"))(10),
                  trace="none",
                  sepcolor = "black",
                  colsep=0:4,
                  rowsep=0:4,
                  sepwidth = c(0.01,0.01))
        
        v1Mgs = names(v1Sums[v1Sums!=0])
        v2Mgs = names(v2Sums[v2Sums!=0])
        v3Mgs = names(v3Sums[v3Sums!=0])
        v4Mgs = names(v4Sums[v4Sums!=0])
        
        
        # library(VennDiagram)
        # myCol <- brewer.pal(4, "Pastel2")
        # venn.diagram(
        #   x = list(v1Mgs, v2Mgs, v3Mgs, v4Mgs), 
        #   category.names = c("V1","V2","V3","V4"),
        #   filename="venn.diagram.by.visits.png", 
        #   output=T, 
        #   fill = myCol
        #   )
      }
      
    }
    
    
    getPairTabMode = T
    if (getPairTabMode) {
      
      getSeqVisitPairTab <- function(wellnessMetadata, wellnessUniqSubjs) {
        wellnessSeqTab  = do.call(rbind,sapply(wellnessUniqSubjs, function(subj) {
          currMetadata = wellnessMetadata[wellnessMetadata$metadata.ID == subj,]
          currVisits = currMetadata$subtype
          currVisits = gsub("v","",currVisits)
          currVisits = as.numeric(currVisits)
          seqTab = rbind(c(1,2),
                         c(2,3),
                         c(3,4))
          seqNameTab = apply(seqTab, 1:2, function(x) paste(subj, "_v", x, sep=""))
          return(seqNameTab)
        }, simplify = F))
        colnames(wellnessSeqTab) = c("before", "after")
        wellnessSeqTab = data.frame(wellnessSeqTab, stringsAsFactors = F)
        wellnessSeqTab$subj = gsub("_v[1-4]","", wellnessSeqTab$before)
        wellnessSeqTab$GeneRichness_bf = wellnessMetadata$GeneRichness[match(wellnessSeqTab$before, wellnessMetadata$sample.ID)]
        wellnessSeqTab$GeneRichness_af = wellnessMetadata$GeneRichness[match(wellnessSeqTab$after, wellnessMetadata$sample.ID)]
        wellnessSeqTab$MgsRichness_bf = wellnessMetadata$MgsRichness[match(wellnessSeqTab$before, wellnessMetadata$sample.ID)]
        wellnessSeqTab$MgsRichness_af = wellnessMetadata$MgsRichness[match(wellnessSeqTab$after, wellnessMetadata$sample.ID)]
        
        return(wellnessSeqTab)
        
      } 
      wellnessSeqTab = getSeqVisitPairTab(wellnessMetadata, wellnessUniqSubjs)
      
    } 
    
    if (F) {
      save(wellnessMgsMat, 
           file="C://Data//comparative.analysis.healthy.sweden//wellnessMgsMat.RData")
      
      save(wellnessMetadata, 
           file="C://Data//comparative.analysis.healthy.sweden//wellnessMetadata.RData")
      
    }
    
    generateMergedClinMgsMat = F
    if (generateMergedClinMgsMat) {
      richness = wellnessMetadata$GeneRichness[match(colnames(wellnessMgsMat), wellnessMetadata$sample.ID)]
      wellnessMgsMat = wellnessMgsMat[rowSums(wellnessMgsMat)!=0,]
      wellnessMgsMatNew = rbind(richness,
                                scaledInflowMeans,
                                scaledOutflowMeans,
                                scaledWellnessMgsMat) #
      
      dim(wellnessClinicalMatrix)
      dim(wellnessMgsMatNew)
      clinFeatures = rownames(wellnessClinicalMatrix)
      mgsFeatures = rownames(scaledWellnessMgsMat)
      
      lenClinFeatures = dim(wellnessClinicalMatrix)[1]
      lenMgsFeatures = dim(scaledWellnessMgsMat)[1]
      
      wellnessClinMgsMat = rbind(wellnessClinicalMatrix, scaledWellnessMgsMat)
      wellnessClinMgsTab = as.data.frame(t(wellnessClinMgsMat))
      wellnessClinMgsTab$sample = rownames(wellnessClinMgsTab)
      wellnessClinMgsTab$subject = gsub("_v[1-4]", "", wellnessClinMgsTab$sample)
      wellnessClinMgsTab$subject = factor(wellnessClinMgsTab$subject)
      wellnessClinMgsTab$visit = gsub("^X[0-9]+_", "", wellnessClinMgsTab$sample)
      
      ###lmer test
      library(MuMIn)
      
      
      ###checking models ####
      
      mixedModelOnlyInOutTab = do.call(rbind.data.frame,sapply(clinFeatures, function(currClin) {
        
        currData = data.frame(clin = wellnessClinicalMatrix[currClin,],
                              inflow = wellnessMgsMatNew["scaledInflowMeans",],
                              outflow = wellnessMgsMatNew["scaledOutflowMeans",],
                              subj = wellnessClinMgsTab$subject,
                              stringsAsFactors = F)
        
        lmerClinMgs = lmer(clin ~ inflow + outflow + (1|subj), data = currData)
        exp.var = r.squaredGLMM(lmerClinMgs)[1]
        
        varNames = rownames(summary(lmerClinMgs)$coefficients)
        tInflow = ifelse( any(varNames == "inflow"), summary(lmerClinMgs)$coefficients["inflow","Estimate"], NA)
        tOutflow = ifelse( any(varNames == "outflow"), summary(lmerClinMgs)$coefficients["outflow","Estimate"], NA)
        
        pInflow = ifelse( any(varNames == "inflow"), summary(lmerClinMgs)$coefficients["inflow","Pr(>|t|)"], NA)
        pOutflow = ifelse( any(varNames == "outflow"), summary(lmerClinMgs)$coefficients["outflow","Pr(>|t|)"], NA)
        
        return(list(exp.var = exp.var, pInflow=pInflow,  pOutflow=pOutflow, tInflow = tInflow, tOutflow = tOutflow))
        # 
        # 
        # mixedModelOuts = do.call(rbind.data.frame, sapply(mgsFeatures, function(currMgs) {
        #   currData = data.frame(clin = wellnessClinicalMatrix[currClin,],
        #                         mgs = wellnessMgsMatNew[currMgs,],
        #                         #inflow = wellnessMgsMatNew["scaledInflowMeans",],
        #                         outflow = wellnessMgsMatNew["scaledOutflowMeans",],
        #                         subj = wellnessClinMgsTab$subject,
        #                         stringsAsFactors = F)
        #   lmerClinMgs = lmer(clin ~ mgs + outflow + (1|subj), data = currData)
        #   if (dim(summary(lmerClinMgs)$coefficients)[1]<3) {
        #     #result = list(clin=currClin, mgs = currMgs, exp.var = NA, tEst = NA, pvalue = NA)
        #     #print(paste(currClin, currMgs, result[1],result[2],result[3], sep = " "))
        #     return(list(pMgs = NA,  pOutflow = NA, tMgs = NA, tOutflow = NA))
        #     return(result)
        #   }
        #   #print(summary(lmerClinMgs))
        #   print(which(mgsFeatures == currMgs))
        #   varNames = rownames(summary(lmerClinMgs)$coefficients)
        #   
        #   tMgs = ifelse( any(varNames == "mgs"), summary(lmerClinMgs)$coefficients["mgs","Estimate"], NA)
        #   tOutflow = ifelse( any(varNames == "outflow"), summary(lmerClinMgs)$coefficients["outflow","Estimate"], NA)
        #   
        #   pMgs = ifelse( any(varNames == "mgs"), summary(lmerClinMgs)$coefficients["mgs","Pr(>|t|)"], NA)
        #   pOutflow = ifelse( any(varNames == "outflow"), summary(lmerClinMgs)$coefficients["outflow","Pr(>|t|)"], NA)
        #   
        #   return(list(pMgs=pMgs,  pOutflow=pOutflow, tMgs = tMgs, tOutflow = tOutflow))
        #   # pvalue_t = summary(lmerClinMgs)$coefficients["mgs","Pr(>|t|)"]
        #   # exp.var = r.squaredGLMM(lmerClinMgs)[1,"R2m"]
        #   # #anovaPvalue = Anova(lmerClinMgs)$`Pr(>Chisq)`
        #   # #result = c(exp.var, estimate, pvalue_t)
        #   # result = list(clin=currClin, mgs = currMgs, exp.var = exp.var, tEst = estimate, pvalue = pvalue_t)
        #   # #print(paste(currClin, currMgs, result[1],result[2],result[3], sep = " "))
        #   # return(result)
        #   
        # }, simplify=F))
        # mixedModelOuts$species = getSpeciesName(rownames(mixedModelOuts),taxo)
        # 
        # mixedModelIns = do.call(rbind.data.frame, sapply(mgsFeatures, function(currMgs) {
        #   currData = data.frame(clin = wellnessClinicalMatrix[currClin,],
        #                         mgs = wellnessMgsMatNew[currMgs,],
        #                         inflow = wellnessMgsMatNew["scaledInflowMeans",],
        #                         #outflow = wellnessMgsMatNew["scaledOutflowMeans",],
        #                         subj = wellnessClinMgsTab$subject,
        #                         stringsAsFactors = F)
        #   lmerClinMgs = lmer(clin ~ mgs + inflow + (1|subj), data = currData)
        #   if (dim(summary(lmerClinMgs)$coefficients)[1]<3) {
        #     #result = list(clin=currClin, mgs = currMgs, exp.var = NA, tEst = NA, pvalue = NA)
        #     #print(paste(currClin, currMgs, result[1],result[2],result[3], sep = " "))
        #     return(list(pMgs = NA,  pInflow = NA, tMgs = NA, tInflow = NA))
        #     return(result)
        #   }
        #   #print(summary(lmerClinMgs))
        #   print(which(mgsFeatures == currMgs))
        #   varNames = rownames(summary(lmerClinMgs)$coefficients)
        #   
        #   tMgs = ifelse( any(varNames == "mgs"), summary(lmerClinMgs)$coefficients["mgs","Estimate"], NA)
        #   tInflow = ifelse( any(varNames == "inflow"), summary(lmerClinMgs)$coefficients["inflow","Estimate"], NA)
        #   
        #   pMgs = ifelse( any(varNames == "mgs"), summary(lmerClinMgs)$coefficients["mgs","Pr(>|t|)"], NA)
        #   pInflow = ifelse( any(varNames == "inflow"), summary(lmerClinMgs)$coefficients["inflow","Pr(>|t|)"], NA)
        #   
        #   return(list(pMgs=pMgs,  pInflow=pInflow, tMgs = tMgs, tInflow = tInflow))
        #   # pvalue_t = summary(lmerClinMgs)$coefficients["mgs","Pr(>|t|)"]
        #   # exp.var = r.squaredGLMM(lmerClinMgs)[1,"R2m"]
        #   # #anovaPvalue = Anova(lmerClinMgs)$`Pr(>Chisq)`
        #   # #result = c(exp.var, estimate, pvalue_t)
        #   # result = list(clin=currClin, mgs = currMgs, exp.var = exp.var, tEst = estimate, pvalue = pvalue_t)
        #   # #print(paste(currClin, currMgs, result[1],result[2],result[3], sep = " "))
        #   # return(result)
        #   
        # }, simplify=F))
        # mixedModelIns$species = getSpeciesName(rownames(mixedModelIns),taxo)
        # 
        # mixedModelIns = do.call(rbind.data.frame, sapply(mgsFeatures, function(currMgs) {
        #   currData = data.frame(clin = wellnessClinicalMatrix[currClin,],
        #                         mgs = wellnessMgsMatNew[currMgs,],
        #                         inflow = wellnessMgsMatNew["scaledInflowMeans",],
        #                         #outflow = wellnessMgsMatNew["scaledOutflowMeans",],
        #                         subj = wellnessClinMgsTab$subject,
        #                         stringsAsFactors = F)
        #   lmerClinMgs = lmer(clin ~ mgs + inflow + (1|subj), data = currData)
        #   if (dim(summary(lmerClinMgs)$coefficients)[1]<3) {
        #     #result = list(clin=currClin, mgs = currMgs, exp.var = NA, tEst = NA, pvalue = NA)
        #     #print(paste(currClin, currMgs, result[1],result[2],result[3], sep = " "))
        #     return(list(mgs=NA, inflow=NA))
        #     return(result)
        #   }
        #   print(summary(lmerClinMgs))
        # 
        #   estMgs = summary(lmerClinMgs)$coefficients["mgs","Estimate"]
        #   estInflow = summary(lmerClinMgs)$coefficients["inflow","Estimate"]
        #   #estOutflow = summary(lmerClinMgs)$coefficients["outflow","Estimate"]
        # 
        #   pMgs = summary(lmerClinMgs)$coefficients["mgs","Pr(>|t|)"]
        #   pInflow = summary(lmerClinMgs)$coefficients["inflow","Pr(>|t|)"]
        #   #pOutflow = summary(lmerClinMgs)$coefficients["outflow","Pr(>|t|)"]
        # 
        #   return(list(mgs=pMgs,  inflow=pInflow))
        #   # pvalue_t = summary(lmerClinMgs)$coefficients["mgs","Pr(>|t|)"]
        #   # exp.var = r.squaredGLMM(lmerClinMgs)[1,"R2m"]
        #   # #anovaPvalue = Anova(lmerClinMgs)$`Pr(>Chisq)`
        #   # #result = c(exp.var, estimate, pvalue_t)
        #   # result = list(clin=currClin, mgs = currMgs, exp.var = exp.var, tEst = estimate, pvalue = pvalue_t)
        #   # #print(paste(currClin, currMgs, result[1],result[2],result[3], sep = " "))
        #   # return(result)
        # 
        # }, simplify=F))
        
        
        return(mixedModelOuts)
      }, simplify = F))
      #mixedModelOnlyInOutTabRData = "J://Deposit//Project//2018_microbiome_atlas//microbiome.atlas.Rproj//mixedModelOnlyInOutTab.RData"
      mixedModelOnlyInOutTabRData = "J://Deposit//Project//2018_microbiome_atlas//microbiome.atlas.Rproj//mixedModelOnlyInOutTab.updated.2020.0507.RData"
      #save(mixedModelOnlyInOutTab, file=mixedModelOnlyInOutTabRData)
      
      mixedModelOnlyInOutSelTab = mixedModelOnlyInOutTab[mixedModelOnlyInOutTab$pInflow<0.05 | mixedModelOnlyInOutTab$pOutflow < 0.05,]
      mixedModelOnlyInOutSelTab$logPInflow = -log10(mixedModelOnlyInOutSelTab$pInflow)
      mixedModelOnlyInOutSelTab$logPOutflow = -log10(mixedModelOnlyInOutSelTab$pOutflow)
      mixedModelOnlyInOutSelTab$clin = rownames(mixedModelOnlyInOutSelTab)
      mixedModelOnlyInOutSelMelt = melt(as.matrix(mixedModelOnlyInOutSelTab[,c("logPInflow","logPOutflow")]))
      mixedModelOnlyInOutSelMelt = mixedModelOnlyInOutSelMelt[mixedModelOnlyInOutSelMelt$Var1 !="P.GLUK.A1", ]
      
      ggplot(mixedModelOnlyInOutSelMelt, aes(x=Var2, y=Var1, size=value, colour=value)) +
        scale_size(range = c(1,15))+
        scale_color_gradient(low="white", high = "gray") +
        geom_point() + 
        xlab("") + ylab("")+
        theme(panel.grid.major = element_blank(), 
              panel.grid.minor = element_blank(),
              axis.text.x = element_blank(),#
              axis.text.y = element_blank(),#
              legend.position = "none",
              panel.background = element_rect(fill = "white", 
                                              colour = "black", 
                                              size = 0.5, 
                                              linetype = "solid"))
      
      
      
      
      mixedModelMgsOutTab = do.call(rbind,sapply(clinFeatures, function(currClin) {
        print(which(clinFeatures == currClin))
        mixedModelResult = do.call(rbind.data.frame, sapply(mgsFeatures, function(currMgs) {
          currData = data.frame(clin = wellnessClinicalMatrix[currClin,],
                                mgs = wellnessMgsMatNew[currMgs,],
                                subj = wellnessClinMgsTab$subject,
                                stringsAsFactors = F)
          lmerClinMgs = lmer(clin ~ mgs + (1|subj), data = currData)
          if (dim(summary(lmerClinMgs)$coefficients)[1]<2) {
            #result = list(clin=currClin, mgs = currMgs, exp.var = NA, tEst = NA, pvalue = NA)
            #print(paste(currClin, currMgs, result[1],result[2],result[3], sep = " "))
            return(list(pMgs=NA,  tMgs = NA, exp.var = exp.vNAar))
            return(result)
          }
          #print(summary(lmerClinMgs))
          #print(which(mgsFeatures == currMgs))
          varNames = rownames(summary(lmerClinMgs)$coefficients)
          exp.var = r.squaredGLMM(lmerClinMgs)[1]
          
          tMgs = ifelse( any(varNames == "mgs"), summary(lmerClinMgs)$coefficients["mgs","Estimate"], NA)
          pMgs = ifelse( any(varNames == "mgs"), summary(lmerClinMgs)$coefficients["mgs","Pr(>|t|)"], NA)
          
          return(list(pMgs=pMgs,   tMgs = tMgs, exp.var = exp.var))
          # pvalue_t = summary(lmerClinMgs)$coefficients["mgs","Pr(>|t|)"]
          # exp.var = r.squaredGLMM(lmerClinMgs)[1,"R2m"]
          # #anovaPvalue = Anova(lmerClinMgs)$`Pr(>Chisq)`
          # #result = c(exp.var, estimate, pvalue_t)
          # result = list(clin=currClin, mgs = currMgs, exp.var = exp.var, tEst = estimate, pvalue = pvalue_t)
          # #print(paste(currClin, currMgs, result[1],result[2],result[3], sep = " "))
          # return(result)
          
        }, simplify=F))
        return(mixedModelResult)
      }, simplify = F))
      
      mixedModelMgsOutTabRData = "J://Deposit//Project//2018_microbiome_atlas//microbiome.atlas.Rproj//mixedModelMgsOutTab.RData"
      mixedModelMgsOutTab$clin =  gsub(".msp_.*$", "", rownames(mixedModelMgsOutTab))
      mixedModelMgsOutTab$mgs =  gsub("^.*.msp", "msp", rownames(mixedModelMgsOutTab))
      mixedModelMgsOutTab$species = getSpeciesName(mixedModelMgsOutTab$mgs, taxo)
      #save(mixedModelMgsOutTab, file=mixedModelMgsOutTabRData)
      mixedModelMgsOutTabSel = mixedModelMgsOutTab[mixedModelMgsOutTab$exp.var >= 0.1, ]
      mixedModelOutTab = do.call(rbind.data.frame,sapply(clinFeatures, function(currClin) {
        
        mixedModelOuts = do.call(rbind.data.frame, sapply(mgsFeatures, function(currMgs) {
          currData = data.frame(clin = wellnessClinicalMatrix[currClin,],
                                mgs = wellnessMgsMatNew[currMgs,],
                                #inflow = wellnessMgsMatNew["scaledInflowMeans",],
                                outflow = wellnessMgsMatNew["scaledOutflowMeans",],
                                subj = wellnessClinMgsTab$subject,
                                stringsAsFactors = F)
          lmerClinMgs = lmer(clin ~ mgs + outflow + (1|subj), data = currData)
          if (dim(summary(lmerClinMgs)$coefficients)[1]<3) {
            #result = list(clin=currClin, mgs = currMgs, exp.var = NA, tEst = NA, pvalue = NA)
            #print(paste(currClin, currMgs, result[1],result[2],result[3], sep = " "))
            return(list(pMgs = NA,  pOutflow = NA, tMgs = NA, tOutflow = NA))
            return(result)
          }
          #print(summary(lmerClinMgs))
          print(which(mgsFeatures == currMgs))
          varNames = rownames(summary(lmerClinMgs)$coefficients)
          
          tMgs = ifelse( any(varNames == "mgs"), summary(lmerClinMgs)$coefficients["mgs","Estimate"], NA)
          tOutflow = ifelse( any(varNames == "outflow"), summary(lmerClinMgs)$coefficients["outflow","Estimate"], NA)
          
          pMgs = ifelse( any(varNames == "mgs"), summary(lmerClinMgs)$coefficients["mgs","Pr(>|t|)"], NA)
          pOutflow = ifelse( any(varNames == "outflow"), summary(lmerClinMgs)$coefficients["outflow","Pr(>|t|)"], NA)
          
          return(list(pMgs=pMgs,  pOutflow=pOutflow, tMgs = tMgs, tOutflow = tOutflow))
          # pvalue_t = summary(lmerClinMgs)$coefficients["mgs","Pr(>|t|)"]
          # exp.var = r.squaredGLMM(lmerClinMgs)[1,"R2m"]
          # #anovaPvalue = Anova(lmerClinMgs)$`Pr(>Chisq)`
          # #result = c(exp.var, estimate, pvalue_t)
          # result = list(clin=currClin, mgs = currMgs, exp.var = exp.var, tEst = estimate, pvalue = pvalue_t)
          # #print(paste(currClin, currMgs, result[1],result[2],result[3], sep = " "))
          # return(result)
          
        }, simplify=F))
        mixedModelOuts$species = getSpeciesName(rownames(mixedModelOuts),taxo)
        
        mixedModelIns = do.call(rbind.data.frame, sapply(mgsFeatures, function(currMgs) {
          currData = data.frame(clin = wellnessClinicalMatrix[currClin,],
                                mgs = wellnessMgsMatNew[currMgs,],
                                inflow = wellnessMgsMatNew["scaledInflowMeans",],
                                #outflow = wellnessMgsMatNew["scaledOutflowMeans",],
                                subj = wellnessClinMgsTab$subject,
                                stringsAsFactors = F)
          lmerClinMgs = lmer(clin ~ mgs + inflow + (1|subj), data = currData)
          if (dim(summary(lmerClinMgs)$coefficients)[1]<3) {
            #result = list(clin=currClin, mgs = currMgs, exp.var = NA, tEst = NA, pvalue = NA)
            #print(paste(currClin, currMgs, result[1],result[2],result[3], sep = " "))
            return(list(pMgs = NA,  pInflow = NA, tMgs = NA, tInflow = NA))
            return(result)
          }
          #print(summary(lmerClinMgs))
          print(which(mgsFeatures == currMgs))
          varNames = rownames(summary(lmerClinMgs)$coefficients)
          
          tMgs = ifelse( any(varNames == "mgs"), summary(lmerClinMgs)$coefficients["mgs","Estimate"], NA)
          tInflow = ifelse( any(varNames == "inflow"), summary(lmerClinMgs)$coefficients["inflow","Estimate"], NA)
          
          pMgs = ifelse( any(varNames == "mgs"), summary(lmerClinMgs)$coefficients["mgs","Pr(>|t|)"], NA)
          pInflow = ifelse( any(varNames == "inflow"), summary(lmerClinMgs)$coefficients["inflow","Pr(>|t|)"], NA)
          
          return(list(pMgs=pMgs,  pInflow=pInflow, tMgs = tMgs, tInflow = tInflow))
          # pvalue_t = summary(lmerClinMgs)$coefficients["mgs","Pr(>|t|)"]
          # exp.var = r.squaredGLMM(lmerClinMgs)[1,"R2m"]
          # #anovaPvalue = Anova(lmerClinMgs)$`Pr(>Chisq)`
          # #result = c(exp.var, estimate, pvalue_t)
          # result = list(clin=currClin, mgs = currMgs, exp.var = exp.var, tEst = estimate, pvalue = pvalue_t)
          # #print(paste(currClin, currMgs, result[1],result[2],result[3], sep = " "))
          # return(result)
          
        }, simplify=F))
        mixedModelIns$species = getSpeciesName(rownames(mixedModelIns),taxo)
        
        # mixedModelIns = do.call(rbind.data.frame, sapply(mgsFeatures, function(currMgs) {
        #   currData = data.frame(clin = wellnessClinicalMatrix[currClin,],
        #                         mgs = wellnessMgsMatNew[currMgs,],
        #                         inflow = wellnessMgsMatNew["scaledInflowMeans",],
        #                         #outflow = wellnessMgsMatNew["scaledOutflowMeans",],
        #                         subj = wellnessClinMgsTab$subject,
        #                         stringsAsFactors = F)
        #   lmerClinMgs = lmer(clin ~ mgs + inflow + (1|subj), data = currData)
        #   if (dim(summary(lmerClinMgs)$coefficients)[1]<3) {
        #     #result = list(clin=currClin, mgs = currMgs, exp.var = NA, tEst = NA, pvalue = NA)
        #     #print(paste(currClin, currMgs, result[1],result[2],result[3], sep = " "))
        #     return(list(mgs=NA, inflow=NA))
        #     return(result)
        #   }
        #   print(summary(lmerClinMgs))
        # 
        #   estMgs = summary(lmerClinMgs)$coefficients["mgs","Estimate"]
        #   estInflow = summary(lmerClinMgs)$coefficients["inflow","Estimate"]
        #   #estOutflow = summary(lmerClinMgs)$coefficients["outflow","Estimate"]
        # 
        #   pMgs = summary(lmerClinMgs)$coefficients["mgs","Pr(>|t|)"]
        #   pInflow = summary(lmerClinMgs)$coefficients["inflow","Pr(>|t|)"]
        #   #pOutflow = summary(lmerClinMgs)$coefficients["outflow","Pr(>|t|)"]
        # 
        #   return(list(mgs=pMgs,  inflow=pInflow))
        #   # pvalue_t = summary(lmerClinMgs)$coefficients["mgs","Pr(>|t|)"]
        #   # exp.var = r.squaredGLMM(lmerClinMgs)[1,"R2m"]
        #   # #anovaPvalue = Anova(lmerClinMgs)$`Pr(>Chisq)`
        #   # #result = c(exp.var, estimate, pvalue_t)
        #   # result = list(clin=currClin, mgs = currMgs, exp.var = exp.var, tEst = estimate, pvalue = pvalue_t)
        #   # #print(paste(currClin, currMgs, result[1],result[2],result[3], sep = " "))
        #   # return(result)
        # 
        # }, simplify=F))
        # 
        
        return(mixedModelOuts)
      }, simplify = F))
      
      mixedModelOutTab = do.call(rbind,sapply(clinFeatures[1], function(currClin) {
        mixedModelOuts = do.call(rbind.data.frame, sapply(mgsFeatures, function(currMgs) {
          currData = data.frame(clin = wellnessClinicalMatrix[currClin,],
                                mgs = wellnessMgsMatNew[currMgs,],
                                inflow = wellnessMgsMatNew["scaledInflowMeans",],
                                outflow = wellnessMgsMatNew["scaledOutflowMeans",],
                                subj = wellnessClinMgsTab$subject, 
                                stringsAsFactors = F)
          lmerClinMgs = lmer(clin ~ mgs + inflow + outflow + (1|subj), data = currData)
          if (dim(summary(lmerClinMgs)$coefficients)[1]<2) {
            result = list(clin=currClin, mgs = currMgs, exp.var = NA, tEst = NA, pvalue = NA)
            #print(paste(currClin, currMgs, result[1],result[2],result[3], sep = " "))
            return(result)
          }
          print(summary(lmerClinMgs))
          estimate = summary(lmerClinMgs)$coefficients["mgs","Estimate"]
          pvalue_t = summary(lmerClinMgs)$coefficients["mgs","Pr(>|t|)"]
          exp.var = r.squaredGLMM(lmerClinMgs)[1,"R2m"]
          #anovaPvalue = Anova(lmerClinMgs)$`Pr(>Chisq)`
          #result = c(exp.var, estimate, pvalue_t)
          result = list(clin=currClin, mgs = currMgs, exp.var = exp.var, tEst = estimate, pvalue = pvalue_t)
          #print(paste(currClin, currMgs, result[1],result[2],result[3], sep = " "))
          return(result)
          
        }, simplify=F))
        return(mixedModelOuts)
      }, simplify = F))
      
      pvalueList = sapply(clinFeatures, function(currClin) {
        pvalues = sapply(mgsFeatures, function(currMgs) {
          lmClinMgs = lm(wellnessClinMgsTab[,currClin] ~ wellnessClinMgsTab[,currMgs] + wellnessClinMgsTab$subject)
          anovaClinMgs = anova(lmClinMgs)
          #anovaClinMgs$`Sum Sq`[1]
          pvalue=anovaClinMgs$`Pr(>F)`[1]
          #est=summary(lmClinMgs)$coefficients[2,"Estimate"]
          #pvalue=summary(lmClinMgs)$coefficients[2,"Pr(>|t|)"]
          return(pvalue)
          
        })
      }, simplify = F)
      pvalueMat = do.call(rbind, pvalueList)
      pvalueMelt = melt(pvalueMat)
      pvalueMelt$species = getSpeciesName(pvalueMelt$Var2, taxo)
      pvalueMatRData = "wellness.mgs.clin.pvalueMat.RData"
      #save(pvalueMat, file=pvalueMatRData)
      
      estList = sapply(clinFeatures, function(currClin) {
        ests = sapply(mgsFeatures, function(currMgs) {
          lmClinMgs = lm(wellnessClinMgsTab[,currClin] ~ wellnessClinMgsTab[,currMgs] + wellnessClinMgsTab$subject)
          anovaClinMgs = anova(lmClinMgs)
          #est=summary(lmClinMgs)$coefficients[2,"Estimate"]
          #pvalue=summary(lmClinMgs)$coefficients[2,"Pr(>|t|)"]
          est=anovaClinMgs$`Sum Sq`[1]*100/sum(anovaClinMgs$`Sum Sq`)
          
          return(est)
          
        })
        return(ests)
      }, simplify = F)
      estMat = do.call(rbind, estList)
      estMelt = melt(estMat)
      estMelt$species = getSpeciesName(estMelt$Var2, taxo)
      estMatRData = "wellness.mgs.clin.estMat.RData"
      #save(estMat, file=estMatRData)
      
      
      tList = sapply(clinFeatures, function(currClin) {
        ts = sapply(mgsFeatures, function(currMgs) {
          lmClinMgs = lm(wellnessClinMgsTab[,currClin] ~ wellnessClinMgsTab[,currMgs] + wellnessClinMgsTab$subject)
          est=summary(lmClinMgs)$coefficients[2,"Estimate"]
          #pvalue=summary(lmClinMgs)$coefficients[2,"Pr(>|t|)"]
          #est=anovaClinMgs$`Sum Sq`[1]*100/sum(anovaClinMgs$`Sum Sq`)
          
          return(est)
          
        })
        return(ts)
      }, simplify = F)
      tMat = do.call(rbind, tList)
      tMelt = melt(tMat)
      tMelt$species = getSpeciesName(tMelt$Var2, taxo)
      tMatRData = "wellness.mgs.clin.tMat.RData"
      #save(tMat, file=tMatRData)
      
      table(estMelt$Var1 == pvalueMelt$Var1)
      table(estMelt$Var1 == tMelt$Var1)
      table(estMelt$Var2 == pvalueMelt$Var2)
      table(estMelt$Var2 == tMelt$Var2)
      
      allStatMelt = data.frame(clin=estMelt$Var1,
                               msp=estMelt$Var2,
                               var.exp=estMelt$value,
                               t=tMelt$value,
                               pvalue=pvalueMelt$value,
                               stringsAsFactors = F)
      
      allStatMelt = data.frame(clin=estMelt$Var1,
                               msp=estMelt$Var2,
                               var.exp=estMelt$value,
                               t=tMelt$value,
                               pvalue=pvalueMelt$value, 
                               species=estMelt$species)
      
      
      allStatMeltRData = "C:/Data/comparative.analysis.healthy.sweden/wellness.mgs.clin.all.StatMelt.RData"
      save(allStatMelt, file=allStatMeltRData)
      
      allStatMeltSel$clin
      mgsClins = unique(allStatMeltSel$clin)
      
      
      allStatMeltRData = "C:/Data/comparative.analysis.healthy.sweden/wellness.mgs.clin.all.StatMelt.with.in.out.RData"
      save(allStatMelt, file=allStatMeltRData)
      
      inClins = allStatMelt$clin[allStatMelt$msp=="scaledInflowMeans" & allStatMelt$pvalue < 1e-10]
      outClins = allStatMelt$clin[allStatMelt$msp=="scaledOutflowMeans" & allStatMelt$pvalue < 1e-10]
      richnessClins = allStatMelt$clin[allStatMelt$msp=="richness" & allStatMelt$pvalue < 1e-10]
      
      
      
      allStatSigPosMelt = allStatMelt[allStatMelt$var.exp>10 & allStatMelt$t>0 & allStatMelt$pvalue<1e-2,]
      allStatSigNegMelt = allStatMelt[allStatMelt$var.exp>10 & allStatMelt$t<0 & allStatMelt$pvalue<1e-2,]
      
      posMat = acast(allStatSigPosMelt, clin ~ species, value.var = "var.exp")
      negMat = acast(allStatSigNegMelt, clin ~ species, value.var = "var.exp")
      posMat = t(posMat)
      negMat = t(negMat)
      
      heatmap.2(posMat, trace="none",margins = c(5,10),
                key=F,lhei=c(1,20),
                cexCol = 0.6,cexRow = 0.6,
                sepcolor = "gray",
                colsep=seq(0, dim(posMat)[2]),
                rowsep=seq(0, dim(posMat)[1]),
                sepwidth = c(0.01,0.01),srtCol = 45,
                col = colorRampPalette(c("white","red"))(256))
      heatmap.2(negMat, trace="none",margins = c(5,10),
                key=F,lhei=c(1,20),
                cexCol = 0.6,cexRow = 0.6,
                sepcolor = "gray",
                colsep=seq(0, dim(posMat)[2]),
                rowsep=seq(0, dim(posMat)[1]),
                sepwidth = c(0.01,0.01),srtCol = 45,
                col = colorRampPalette(c("white","blue"))(256))
      
      
    }
    
    generatedClinAssocInOutflow = F
    if (generatedClinAssocInOutflow) {
      richness = wellnessMetadata$GeneRichness[match(colnames(wellnessMgsMat), wellnessMetadata$sample.ID)]
      inflowMat = rbind(richness,
                        scaledInflowMeans,
                        scaledOutflowMeans) #,wellnessMgsMat
      dim(wellnessClinicalMatrix)
      dim(richness)
      clinFeatures = rownames(wellnessClinicalMatrix)
      flowFeatures = rownames(inflowMat)
      
      lenClinFeatures = dim(wellnessClinicalMatrix)[1]
      lenFlowFeatures = dim(flowFeatures)[1]
      
      wellnessClinMgsMat = rbind(wellnessClinicalMatrix, richness)
      wellnessClinMgsTab = as.data.frame(t(wellnessClinMgsMat))
      wellnessClinMgsTab$sample = rownames(wellnessClinMgsTab)
      wellnessClinMgsTab$subject = gsub("_v[1-4]", "", wellnessClinMgsTab$sample)
      wellnessClinMgsTab$visit = gsub("^X[0-9]+_", "", wellnessClinMgsTab$sample)
      
      
      library(MuMIn)
      mixedModelOutTab2 = do.call(rbind,sapply(clinFeatures, function(currClin) {
        mixedModelOuts2 = do.call(rbind.data.frame, sapply(flowFeatures, function(currFlow) {
          currData = data.frame(clin = wellnessClinicalMatrix[currClin,],
                                mgs = inflowMat[currFlow,],
                                subj = wellnessClinMgsTab$subject, 
                                stringsAsFactors = F)
          lmerClinMgs = lmer(clin ~ mgs + (1|subj), data = currData)
          if (dim(summary(lmerClinMgs)$coefficients)[1]<2) {
            result = list(clin=currClin, mgs = currFlow, exp.var = NA, tEst = NA, pvalue = NA)
            #print(paste(currClin, currMgs, result[1],result[2],result[3], sep = " "))
            return(result)
          }
          estimate = summary(lmerClinMgs)$coefficients["mgs","Estimate"]
          pvalue_t = summary(lmerClinMgs)$coefficients["mgs","Pr(>|t|)"]
          exp.var = r.squaredGLMM(lmerClinMgs)[1,"R2m"]
          #anovaPvalue = Anova(lmerClinMgs)$`Pr(>Chisq)`
          #result = c(exp.var, estimate, pvalue_t)
          result = list(clin=currClin, mgs = currFlow, exp.var = exp.var, tEst = estimate, pvalue = pvalue_t)
          #print(paste(currClin, currMgs, result[1],result[2],result[3], sep = " "))
          return(result)
          
        }, simplify=F))
        return(mixedModelOuts2)
      }, simplify = F))
      
      
      pvalueList = sapply(clinFeatures, function(currClin) {
        pvalues = sapply(mgsFeatures, function(currMgs) {
          lmClinMgs = lm(wellnessClinMgsTab[,currClin] ~ wellnessClinMgsTab[,currMgs] + wellnessClinMgsTab$subject)
          anovaClinMgs = anova(lmClinMgs)
          #anovaClinMgs$`Sum Sq`[1]
          pvalue=anovaClinMgs$`Pr(>F)`[1]
          #est=summary(lmClinMgs)$coefficients[2,"Estimate"]
          #pvalue=summary(lmClinMgs)$coefficients[2,"Pr(>|t|)"]
          return(pvalue)
          
        })
      }, simplify = F)
      pvalueMat = do.call(rbind, pvalueList)
      pvalueMelt = melt(pvalueMat)
      
      estList = sapply(clinFeatures, function(currClin) {
        ests = sapply(mgsFeatures, function(currMgs) {
          lmClinMgs = lm(wellnessClinMgsTab[,currClin] ~ wellnessClinMgsTab[,currMgs] + wellnessClinMgsTab$subject)
          anovaClinMgs = anova(lmClinMgs)
          #est=summary(lmClinMgs)$coefficients[2,"Estimate"]
          #pvalue=summary(lmClinMgs)$coefficients[2,"Pr(>|t|)"]
          est=anovaClinMgs$`Sum Sq`[1]*100/sum(anovaClinMgs$`Sum Sq`)
          
          return(est)
          
        })
        return(ests)
      }, simplify = F)
      estMat = do.call(rbind, estList)
      estMelt = melt(estMat)
      
      tList = sapply(clinFeatures, function(currClin) {
        ts = sapply(mgsFeatures, function(currMgs) {
          lmClinMgs = lm(wellnessClinMgsTab[,currClin] ~ wellnessClinMgsTab[,currMgs] + wellnessClinMgsTab$subject)
          est=summary(lmClinMgs)$coefficients[2,"Estimate"]
          #pvalue=summary(lmClinMgs)$coefficients[2,"Pr(>|t|)"]
          #est=anovaClinMgs$`Sum Sq`[1]*100/sum(anovaClinMgs$`Sum Sq`)
          
          return(est)
          
        })
        return(ts)
      }, simplify = F)
      tMat = do.call(rbind, tList)
      tMelt = melt(tMat)
      
      table(estMelt$Var1 == pvalueMelt$Var1)
      table(estMelt$Var1 == tMelt$Var1)
      table(estMelt$Var2 == pvalueMelt$Var2)
      table(estMelt$Var2 == tMelt$Var2)
      
      allStatMeltNew = data.frame(clin=as.character(estMelt$Var1),
                                  msp=as.character(estMelt$Var2),
                                  var.exp=estMelt$value,
                                  t=tMelt$value,
                                  pvalue=pvalueMelt$value, 
                                  stringsAsFactors = F)
      allStatMeltNew = allStatMeltNew[allStatMeltNew$clin !="anthropometry_height",]
      
      
      allStatMeltNewRData = "C:/Data/comparative.analysis.healthy.sweden/wellness.InOutflow.clin.all.StatMelt.new.RData"
      save(allStatMeltNew, file=allStatMeltNewRData)
      
      allStatMeltNewSel = allStatMeltNew[allStatMeltNew$pvalue<1e-14,]
      length(unique(allStatMeltNewSel$clin))
      unique(allStatMeltNewSel$clin)
      
      allStatMeltNewSelFile = "C:/Data/comparative.analysis.healthy.sweden/wellness.InOutflow.clin.all.StatMelt.new.sel.pvalue.1e_14.txt"
      write.table(allStatMeltNewSel, allStatMeltNewSelFile, sep="\t", row.names = F, quote = F )
      
      allStatMeltNewSel$logP = log10(allStatMeltNewSel$pvalue)
      
      
    }
    
  }
  
  loadMixedEffectModelResult = T
  if (loadMixedEffectModelResult) {
    mixedModelOnlyInOutTabRData = "J://Deposit//Project//2018_microbiome_atlas//microbiome.atlas.Rproj//mixedModelOnlyInOutTab.RData"
    mixedModelMgsOutTabRData = "J://Deposit//Project//2018_microbiome_atlas//microbiome.atlas.Rproj//mixedModelMgsOutTab.RData"
    
    load(mixedModelOnlyInOutTabRData)
    load(mixedModelMgsOutTabRData)
  }
  
  loadMergedClinMgsMat = F
  if (loadMergedClinMgsMat) {
    allStatMeltRData = "C:/Data/comparative.analysis.healthy.sweden/wellness.mgs.clin.all.StatMelt.RData"
    load(allStatMeltRData) 
  }
  
  checkCorrelation = F
  if (checkCorrelation) {
    
    ### b. fragilis 2 ###
    fragilisVec = wellnessMgsMat["msp_0033",]
    #fragilisVec = wellnessMgsMat["msp_0010",]
    crpVec = clinicalMatrix[,"S.CRPUS.A1"]
    crpVec = crpVec[match(names(fragilisVec), names(crpVec))]
    
    troponinVec = clinicalMatrix[,"S.TNTHS.A1"]
    troponinVec = troponinVec[match(names(fragilisVec), names(crpVec))]    
    
    cor.test(crpVec, fragilisVec)
    cor.test(crpVec, fragilisVec, method="spearman")
    
    plot(crpVec, fragilisVec)
    abline(lm(fragilisVec~crpVec), col="red")
    plot(crpVec[fragilisVec!=0], fragilisVec[fragilisVec!=0])
    
    cor.test(troponinVec, fragilisVec)
    cor.test(troponinVec, fragilisVec, method="spearman")
    
    plot(troponinVec, fragilisVec)
    abline(lm(fragilisVec~troponinVec), col="red")
    
    
    ### clostridia bacterium UC5.1 ###
    
    cbUc51Vec = wellnessMgsMat["msp_0204",]
    crpVec = clinicalMatrix[,"S.CRPUS.A1"]
    crpVec = crpVec[match(names(cbUc51Vec), names(crpVec))]
    
    plot(crpVec, cbUc51Vec)
    plot(crpVec[cbUc51Vec!=0], cbUc51Vec[cbUc51Vec!=0])
    
    ### b. ovatus ###
    
    
    bOvatusVec = wellnessMgsMat["msp_0007",]
    crpVec = clinicalMatrix[,"S.CRPUS.A1"]
    crpVec = crpVec[match(names(bOvatusVec), names(crpVec))]
    
    plot(crpVec, bOvatusVec)
    abline(lm(bOvatusVec~crpVec), col="red")
    #plot(crpVec[bOvatusVec!=0], bOvatusVec[bOvatusVec!=0])
    
    ### testing ###
    testVec = wellnessMgsMat["msp_1436",]
    testVec[testVec!=0]
    
    
  }
  
  processMergedClinMgsNew = T
  if (processMergedClinMgsNew) {
    #mixedModelMgsOutTabSel = mixedModelMgsOutTab[mixedModelMgsOutTab$exp.var>=0.1,]
    mixedModelMgsOutTab$adjPMgs = p.adjust(mixedModelMgsOutTab$pMgs, method="BH")
    mixedModelMgsOutTabSel = mixedModelMgsOutTab[mixedModelMgsOutTab$exp.var>=0.1,]
    write.table(mixedModelMgsOutTabSel, file="C://Data//comparative.analysis.healthy.sweden//mixedModelMgsOutTabSel.txt", sep="\t")
    
    #mixedModelMgsOutTabSel2 = mixedModelMgsOutTab[mixedModelMgsOutTab$pMgs <=0.05,]
    
    sigMspByClinList = split(mixedModelMgsOutTabSel$mgs, mixedModelMgsOutTabSel$clin )
    outflowSigMspByClinList = lapply(sigMspByClinList, function(x) outflow[x])
    medOutflowSigMspByClinList = unlist(lapply(outflowSigMspByClinList, median))
    
    inflowSigMspByClinList = lapply(sigMspByClinList, function(x) inflow[x])
    medInflowSigMspByClinList = unlist(lapply(inflowSigMspByClinList, median))
    
    par(mar=c(5,10,1,1))
    boxplot(outflowSigMspByClinList[order(medOutflowSigMspByClinList)], horizontal=T, las=1)
    boxplot(inflowSigMspByClinList[order(medInflowSigMspByClinList)], horizontal=T, las=1)
    
    lapply(outflowSigMspByClinList[order(medOutflowSigMspByClinList)], function(currOutflow){
      
      print(wilcox.test(currOutflow, outflow, alternative = "greater"))
    }) 
    
    lapply(inflowSigMspByClinList[order(medInflowSigMspByClinList)], function(currOutflow){
      
      print(wilcox.test(currOutflow, inflow, alternative = "greater"))
    }) 
    
    
    head(mixedModelOnlyInOutTab)
    mixedModelOnlyInOutSelTab = mixedModelOnlyInOutTab[mixedModelOnlyInOutTab$pInflow<0.05 | mixedModelOnlyInOutTab$pOutflow < 0.05,]
    
    write.table(mixedModelOnlyInOutSelTab, file="C://Data//comparative.analysis.healthy.sweden//mixedModelOnlyInOutSelTab.txt", sep="\t")
    
    clinCounts = data.frame(name=factor(names(sort(table(mixedModelMgsOutTabSel$clin ))), levels = names(sort(table(mixedModelMgsOutTabSel$clin )))),
                            count=as.numeric(sort(table(mixedModelMgsOutTabSel$clin ))))
    ggplot(clinCounts, aes(x=rev(name), y=count, fill=count)) + 
      geom_bar(stat="identity", colour="black") +
      scale_fill_gradient( low = "white", high="gray")+
      #scale_fill_manual(values=c())+
      #coord_flip() +
      xlab("") + ylab("")+
      #scale_y_continuous(position = "right") +
      theme(panel.border = element_blank(),
            axis.line = element_line(color="black"),
            panel.grid.major = element_blank(), 
            panel.grid.minor = element_blank(),
            legend.position = "none",
            text = element_blank(),
            panel.background = element_blank()) 
    
    par(mar=c(5,10,1,1))
    
    barplot(sort(table(mixedModelMgsOutTabSel$clin )), 
            col="white",
            horiz=T, las=1)
    barplot(sort(table(mixedModelMgsOutTabSel$mgs )), 
            horiz=T, las=1)
    getSpeciesName(names(sort(table(mixedModelMgsOutTabSel2$mgs ),T)[1:10]),taxo)
    
    
    circosMode = T
    if (circosMode) {
      expVarCut=0.1# 10%
      
      generateCirCosPlot <- function(mixedModelMgsOutTab, expVarCut=10, taxo=taxo) {
        mixedModelMgsOutTabSel = mixedModelMgsOutTab[abs(mixedModelMgsOutTab$exp.var)>expVarCut,]
        
        length(unique(mixedModelMgsOutTabSel$clin)) # 10 clinical parameters
        length(unique(mixedModelMgsOutTabSel$mgs)) # 37 MSPs
        
        newSpeciesNames = paste(mixedModelMgsOutTabSel$species[grepl("unclassified", mixedModelMgsOutTabSel$species)], " (", mixedModelMgsOutTabSel$mgs[grepl("unclassified", mixedModelMgsOutTabSel$species)], ")", sep = "")
        mixedModelMgsOutTabSel$species[grepl("unclassified", mixedModelMgsOutTabSel$species)] <- newSpeciesNames
        
        if (F) {
          pdf("C:/Data/comparative.analysis.healthy.sweden/counts.mgs.assoc.clin.wellness.pdf", weight="", height="", family="Arial Narrow")
          mspByClinFeatures = split(allStatMeltSel$msp, allStatMeltSel$clin)
          #mspByClinFeatures = split(allStatMeltSel2$msp, allStatMeltSel2$clin)
          mspCountsByClin = unlist(lapply(mspByClinFeatures, length))
          par(mar=c(5,15,4,1))
          barplot(sort(mspCountsByClin),horiz = T, las=1, col=colorRampPalette(c("gold","white","purple"))(length(mspCountsByClin)))
          dev.off()
        }
        
        circosPreInput = data.frame(from=getPhylaName(mixedModelMgsOutTabSel$mgs, taxo),
                                    msp = mixedModelMgsOutTabSel$msp,
                                    species = getSpeciesName(mixedModelMgsOutTabSel$mgs, taxo),
                                    to=mixedModelMgsOutTabSel$clin,
                                    value=(abs(mixedModelMgsOutTabSel$exp.var)),
                                    stringsAsFactors = F)
        
        circosInput = data.frame(from=getGenusName(mixedModelMgsOutTabSel$mgs, taxo),
                                 #msp = allStatMeltSel$msp,
                                 #species = getSpeciesName(allStatMeltSel$msp, taxo),
                                 to=mixedModelMgsOutTabSel$clin,
                                 value=(abs(mixedModelMgsOutTabSel$exp.var)),
                                 stringsAsFactors = F)
        
        circosInput = data.frame(from=getPhylaName(mixedModelMgsOutTabSel$mgs, taxo),
                                 #msp = allStatMeltSel$msp,
                                 #species = getSpeciesName(allStatMeltSel$msp, taxo),
                                 to=mixedModelMgsOutTabSel$clin,
                                 value=(abs(mixedModelMgsOutTabSel$exp.var)),
                                 stringsAsFactors = F)
        
        #cytoscapleOut = "j://Deposit//Project//2018_microbiome_atlas//functional.annotation//CytoscapeOut.clinical.mgs.assoc.txt"
        #write.table(circosInput, cytoscapleOut, sep="\t", row.names = F, quote = F)
        
        #circosInput$from[circosInput$from == "Candidatus Melainabacteria"] = "C. Melainabacteria"
        circosInput = circosInput[circosInput$from != "unclassified",]
        circosInput = circosInput[!grepl("unclassified",circosInput$from),]
        
        elements = (unique(c(circosInput$from, circosInput$to)))
        lenElements = length(elements)
        
        circosGridCols = (col_vector)[1:length(elements)]
        names(circosGridCols) = elements
        
        #circos.par(track.margin=c(0,0)) 
        set.seed(1)
        circos.initialize(elements, xlim=c(0,1))
        chordDiagram(circosInput)
        chordDiagram(circosInput, annotationTrack = "grid", preAllocateTracks = 1, grid.col = circosGridCols)
        circos.trackPlotRegion(track.index = 1,  panel.fun = function(x, y) {
          xlim = get.cell.meta.data("xlim")
          ylim = get.cell.meta.data("ylim")
          sector.name = get.cell.meta.data("sector.index")
          circos.text(mean(xlim), ylim[1] + .1, cex = 0.5,  sector.name, facing = "clockwise", niceFacing = TRUE, adj = c(0, 0.5))
          circos.axis(h = "top", labels = F, major.tick = F, 
                      sector.index = sector.name, track.index = 2)
        }, bg.border = NA)
        circos.clear() 
        
        return(circosInput)
      }
      
    }
    
  }
  
  processMergedClinMgsMatCircos = F
  if (processMergedClinMgsMatCircos) {
    exFeatures = c("height")
    
    lipidFeatures = c("LDL", "HDL", "TG", "ApoB", "ApoA1", "ApoB_ApoA1")
    glucoseFeatures = c("Glucose", "HbA1c")
    bloodFeatures = c("CRP", "Hemoglobin", "EVF", "MCV", "MCH", "MCHC", "Erythrocyte particle", "Platelets", "WBC", "Neutrophil", "Lymphocyte", "Monocyte", "Eosinophil", "Basophil")
    liverFeatures = c("ALAT", "GT")
    kidneyFeatures = c("Creatine", "Cystatin")
    heartFeatures = c("BNP", "Troponin T")
    
    wellnessDescFile = "C:/Data/comparative.analysis.healthy.sweden/wellness.description.txt"
    wellnessDescTab = read.table(wellnessDescFile, sep="\t", header=T, stringsAsFactors = F)
    allStatMeltRData = "C:/Data/comparative.analysis.healthy.sweden/wellness.mgs.clin.all.StatMelt.RData"
    load(allStatMeltRData)
    
    allStatMelt$msp = as.character(allStatMelt$msp)
    allStatMelt$clin = as.character(allStatMelt$clin)
    allStatMelt$species[is.na(allStatMelt$species)] = ""
    allStatMelt$species = as.character(allStatMelt$species)
    
    allStatMelt$clin = wellnessDescTab$Description[match(allStatMelt$clin, wellnessDescTab$Name)]
    allStatMelt$adjp = p.adjust(allStatMelt$pvalue, "BH")
    allStatMelt$var.exp[allStatMelt$t <0 ] = -1 * allStatMelt$var.exp[allStatMelt$t <0 ]
    allStatMeltSave = allStatMelt
    allStatMelt = allStatMelt[!allStatMelt$clin %in% exFeatures, ]
    allStatMeltSave = allStatMeltSave[!allStatMeltSave$clin %in% exFeatures, ]
    
    #expVarCut = 15
    expVarCut = 10
    
    circosMode = F
    if (circosMode) {
      allStatMeltSel2 = allStatMelt[(allStatMelt$pvalue) < 1e-20,]
      
      generateCirCosPlot <- function(allStatMelt, expVarCut, taxo=taxo) {
        allStatMeltSel = allStatMelt[abs(allStatMelt$var.exp)>expVarCut,]
        
        checkInOutflowMode = F
        if (checkInOutflowMode) {
          
          
          sigClinMsps = unique(allStatMeltSel$msp)
          selMspByClinList = split(as.character(allStatMeltSel$msp), 
                                   as.character(allStatMeltSel$clin))
          
          outflowSelMspByClinList = lapply(selMspByClinList, function(currMsp) outflow[names(outflow)%in% currMsp])
          mOutflowSelMspByClinList = unlist(lapply(outflowSelMspByClinList, median))
          #mOutflowSelMspByClinList = append(outflowSelMspByClinList[names(sort(mOutflowSelMspByClinList))], list(all=outflow))
          outflowSelMspNamesByClinList = lapply(outflowSelMspByClinList, names)
          
          
          outflowEnrichment = getEnrichMentGsByList(outflowSelMspNamesByClinList, names(outflow[outflow>=0.3]))
          
          
          par(mar=c(5,10,4,1))
          boxplot(outflowSelMspByClinList[(names(sort(mOutflowSelMspByClinList)))], horizontal=T, las=1)
          abline(v=0.3)
          
          sapply(1:23, function(ind) {
            print(names(mOutflowSelMspByClinList)[ind])
            print(wilcox.test(mOutflowSelMspByClinList[[ind]], mOutflowSelMspByClinList[[24]], alternative = "greater")$p.value)
            print("")
          })
          
          inflowSelMspByClinList = lapply(selMspByClinList, function(currMsp) inflow[names(inflow)%in% currMsp])
          mInflowSelMspByClinList = unlist(lapply(inflowSelMspByClinList, median))
          #mInflowSelMspByClinList = append(inflowSelMspByClinList[names(sort(mInflowSelMspByClinList))], list(all=inflow))
          inflowSelMspNamesByClinList = lapply(inflowSelMspByClinList, names)
          
          inflowEnrichment = getEnrichMentGsByList(inflowSelMspNamesByClinList, names(inflow[inflow>=0.3]))
          
          par(mar=c(5,10,4,1))
          boxplot(inflowSelMspByClinList[(names(sort(mInflowSelMspByClinList)))], horizontal=T, las=1)
          abline(v=0.3)
          
          sapply(1:23, function(ind) {
            print(names(mInflowSelMspByClinList)[ind])
            print(wilcox.test(mInflowSelMspByClinList[[ind]], mInflowSelMspByClinList[[24]], alternative = "greater")$p.value)
            print("")
          })
        }
        
        length(unique(allStatMeltSel$clin)) # 22 clinical parameters
        length(unique(allStatMeltSel$msp)) # 98 MSPs
        
        newSpeciesNames = paste(allStatMeltSel$species[grepl("unclassified", allStatMeltSel$species)], " (", allStatMeltSel$msp[grepl("unclassified", allStatMeltSel$species)], ")", sep = "")
        allStatMeltSel$species[grepl("unclassified", allStatMeltSel$species)] <- newSpeciesNames
        
        if (F) {
          pdf("C:/Data/comparative.analysis.healthy.sweden/counts.mgs.assoc.clin.wellness.pdf", weight="", height="", family="Arial Narrow")
          mspByClinFeatures = split(allStatMeltSel$msp, allStatMeltSel$clin)
          #mspByClinFeatures = split(allStatMeltSel2$msp, allStatMeltSel2$clin)
          mspCountsByClin = unlist(lapply(mspByClinFeatures, length))
          par(mar=c(5,15,4,1))
          barplot(sort(mspCountsByClin),horiz = T, las=1, col=colorRampPalette(c("gold","white","purple"))(length(mspCountsByClin)))
          dev.off()
        }
        
        circosInput = data.frame(from=getPhylaName(allStatMeltSel$msp, taxo),
                                 msp = allStatMeltSel$msp,
                                 species = getSpeciesName(allStatMeltSel$msp, taxo),
                                 to=allStatMeltSel$clin,
                                 value=(abs(allStatMeltSel$var.exp)),
                                 stringsAsFactors = F)
        
        absValues = (abs(allStatMeltSel$var.exp))
        circosInput = data.frame(from=getGenusName(allStatMeltSel$msp, taxo),
                                 #msp = allStatMeltSel$msp,
                                 #species = getSpeciesName(allStatMeltSel$msp, taxo),
                                 to=allStatMeltSel$clin,
                                 value=absValues,
                                 stringsAsFactors = F)
        cytoscapleOut = "j://Deposit//Project//2018_microbiome_atlas//functional.annotation//CytoscapeOut.clinical.mgs.assoc.txt"
        write.table(circosInput, cytoscapleOut, sep="\t", row.names = F, quote = F)
        
        #circosInput$from[circosInput$from == "Candidatus Melainabacteria"] = "C. Melainabacteria"
        circosInput = circosInput[circosInput$from != "unclassified",]
        circosInput = circosInput[!grepl("unclassified",circosInput$from),]
        
        elements = (unique(c(circosInput$from, circosInput$to)))
        lenElements = length(elements)
        
        circosGridCols = (col_vector)[1:length(elements)]
        names(circosGridCols) = elements
        
        #circos.par(track.margin=c(0,0)) 
        set.seed(1)
        circos.initialize(elements, xlim=c(0,1))
        chordDiagram(circosInput)
        chordDiagram(circosInput, annotationTrack = "grid", preAllocateTracks = 1, grid.col = circosGridCols)
        circos.trackPlotRegion(track.index = 1,  panel.fun = function(x, y) {
          xlim = get.cell.meta.data("xlim")
          ylim = get.cell.meta.data("ylim")
          sector.name = get.cell.meta.data("sector.index")
          circos.text(mean(xlim), ylim[1] + .1, cex = 0.5,  sector.name, facing = "clockwise", niceFacing = TRUE, adj = c(0, 0.5))
          circos.axis(h = "top", labels = F, major.tick = F, 
                      sector.index = sector.name, track.index = 2)
        }, bg.border = NA)
        circos.clear() 
        
        return(circosInput)
      }
      
      circosInput$inflow = inflow[match(circosInput$msp, names(inflow))]
      circosInput$outflow = outflow[match(circosInput$msp, names(outflow))]
      circosInput$prevalence = freqs_wellness[match(circosInput$msp, names(freqs_wellness))]
      circosInput$type = "transient"
      circosInput$type[circosInput$prevalence>0.9] = "colonized"
      circosInput$type[circosInput$prevalence<0.1] = "spontaneous"
      #write.table(circosInput, "circosInput.for.fig1b.txt", sep="\t")
      #require(lattice)
      #levelplot(t(as.matrix(circosInput$inflow)), aspect = 10, col.regions=colorRampPalette(c("gray", col_vector[4]))(256))
      
      table(circosInput$prevalence>0.9)
      table(circosInput$prevalence<0.1)
      
      barplot(as.matrix(cbind(c(Colonized=0, Transient=25, Spontaneous=94))),
              las=1, col=rev(c(col_vector[4],"gray")) )
      
    }
    
    barplotMode = F
    if (barplotMode) {
      
      allStatMelt$clin = as.character(allStatMelt$clin)
      allStatMelt = allStatMelt[!allStatMelt$clin %in% exFeatures, ]
      allStatMelt$adjp = p.adjust(allStatMelt$pvalue, "BH")
      allStatMelt$var.exp[allStatMelt$t <0 ] = -1 * allStatMelt$var.exp[allStatMelt$t <0 ]
      
      #expVarCut=10
      #expVarCut=15
      
      allStatMeltSel = allStatMelt[abs(allStatMelt$var.exp)>expVarCut,]
      allStatMeltSel$clin = as.character(allStatMeltSel$clin)
      allStatMeltSel$msp = as.character(allStatMeltSel$msp)
      #allStatMeltSel$species = as.character(allStatMeltSel$speices)
      
      allStatMeltSel$inflow = inflow[match(allStatMeltSel$msp, names(inflow))]
      allStatMeltSel$outflow = outflow[match(allStatMeltSel$msp, names(outflow))]
      allStatMeltSel2 = allStatMeltSel[allStatMeltSel$clin %in% bloodFeatures,]
      #boxplot(list(selected = allStatMeltSel2$outflow, all=outflow))
      
      mspByClinFeatures = split(allStatMeltSel$msp, allStatMeltSel$clin)
      #mspByClinFeatures = split(allStatMeltSel2$msp, allStatMeltSel2$clin)
      mspCountsByClin = unlist(lapply(mspByClinFeatures, length))
      par(mar=c(5,10,4,1))
      barplot(sort(mspCountsByClin),horiz = T, las=1,cex.names = 0.9,
              col=colorRampPalette(c("gold","white","purple"))(length(mspCountsByClin)))
      
    }
    
    
    # circosResult = generateCirCosPlot(allStatMelt, 15)
    #write.table(allStatMeltSel, "allStatMeltSel.txt", sep="\t")
    allStatMeltSel = allStatMeltSave[abs(allStatMeltSave$var.exp)>expVarCut,]
    allStatMeltSel$clin = as.character(allStatMeltSel$clin)
    allStatMeltSel$msp = as.character(allStatMeltSel$msp)
    
    allStatMeltSel$inflow = inflow[match(allStatMeltSel$msp, names(inflow))]
    allStatMeltSel$outflow = outflow[match(allStatMeltSel$msp, names(outflow))]
    
    expVarMat = acast(allStatMeltSel, clin ~ msp, value.var = "var.exp")
    expVarMat[is.na(expVarMat)]=0
    expVarInflow = allStatMeltSel$inflow[match(colnames(expVarMat), allStatMeltSel$msp)]
    expVarOutflow = allStatMeltSel$outflow[match(colnames(expVarMat), allStatMeltSel$msp)]
    expVarMat = rbind(expVarMat, expVarInflow)
    expVarMat = rbind(expVarMat, expVarOutflow)
    
    expVarCounts = apply(expVarMat,1, function(x) sum(x>0))
    
    getFeatureMat <- function(currExprVarMat, targetFeatures, expVarCut=10) {
      
      targetFlowMat = currExprVarMat[c("expVarInflow","expVarOutflow"),]
      targetExpVarMat = currExprVarMat[rownames(currExprVarMat) %in% targetFeatures,]
      
      targetExpVarMat = targetExpVarMat[rowMaxs(abs(targetExpVarMat))>expVarCut,]
      targetFlowMat = targetFlowMat[,colMaxs(abs(targetExpVarMat))>expVarCut]
      targetExpVarMat = targetExpVarMat[,colMaxs(abs(targetExpVarMat))>expVarCut]
      colnames(targetExpVarMat) = getSpeciesName(colnames(targetExpVarMat),taxo)
      #targetExpVarMat = rbind(targetExpVarMat, inflow=flowList$inflow)
      #targetExpVarMat = rbind(targetExpVarMat, outflow=flowList$outflow)
      targetExpVarMat = rbind(targetExpVarMat, targetFlowMat)
      
      #rownames(targetExpVarMat)[rownames(targetExpVarMat)=="expVarInflow"]="Inflow"
      #rownames(targetExpVarMat)[rownames(targetExpVarMat)=="expVarOutflow"]="Outflow"
      return(targetExpVarMat)
    }
    
    lipidExpVarMat = getFeatureMat(expVarMat, lipidFeatures)
    glucExpVarMat = getFeatureMat(expVarMat, glucoseFeatures) 
    bloodExpVarMat = getFeatureMat(expVarMat, bloodFeatures) 
    liverExpVarMat = getFeatureMat(expVarMat, liverFeatures) 
    glucExpVarMat = getFeatureMat(expVarMat, glucoseFeatures) 
    
    bloodMode = F
    if (bloodMode) {
      targetMat = bloodExpVarMat
      #targetMat = lipidExpVarMat
      targetMat = t(targetMat)
      targetFlows = targetMat[,c("expVarInflow","expVarOutflow")]
      #targetFlows = targetFlows[!grepl("unclassified", rownames(targetMat)),]
      targetMat = targetMat[!grepl("unclassified", rownames(targetMat)),]
      targetMat = targetMat[,!colnames(targetMat) %in% c("expVarInflow","expVarOutflow")]
      
      purple = colorRamp(c("white","purple"))
      gold = colorRamp(c("white","gold"))
      
      inflowCols = rgb((purple(targetFlows[,1]))/255)
      outflowCols = rgb((gold(targetFlows[,2]))/255)
      
      targetFlowMat = cbind(inflow=inflowCols, 
                            outflow=outflowCols)
      rownames(targetFlows) = rownames(targetMat)
      
      hr = heatmap.2((targetMat), trace="none",margins = c(5,13),
                     key=F,dendrogram = "none",
                     lhei=c(1,10),lwid = c(1,10),
                     cexCol = 0.8,cexRow = 0.8,
                     sepcolor = "black",
                     colsep=c(0, dim(targetMat)[2]),
                     rowsep=c(0, dim(targetMat)[1]),
                     sepwidth = c(0.01,0.01),
                     srtCol = 45,
                     col = colorRampPalette(c("blue","white","red"))(256))
      
      heatmap.2((targetFlows[rownames(targetMat)[rev(hr$rowInd)],]), 
                trace="none",margins = c(5,13),
                key=F,dendrogram = "none",Rowv = F,Colv=F,
                lhei=c(1,10),lwid = c(1,10),
                cexCol = 0.8,cexRow = 0.8,
                sepcolor = "black",
                colsep=c(0, dim(targetFlows)[2]),
                rowsep=c(0, dim(targetFlows)[1]),
                sepwidth = c(0.01,0.01),
                srtCol = 45,
                col = colorRampPalette(c("white","black"))(256))
      
      boxplot(targetFlows, yaxt="n",  xaxt="n", col=c("#8F7998", "#FBF2A9"))
      # text(0:71, -1, rep(0:23, 3), cex = 0.5)
      axis(2, at=seq(0, 0.8, 0.2), NA, cex.axis=.7, font=1, tck=.02)
      
      
    }
    
    # heatmap.2(t(expVarMat[rownames(expVarMat) %in% lipidFeatures,]), trace="none",margins = c(5,10),
    #           key=T,
    #           #lhei=c(2,20),
    #           cexCol = 0.6,cexRow = 0.6,
    #           #sepcolor = "gray",
    #           #colsep=seq(0, dim(posMat)[2]),
    #           #rowsep=seq(0, dim(posMat)[1]),
    #           #sepwidth = c(0.01,0.01),
    #           srtCol = 45,
    #           col = colorRampPalette(c("blue","white","red"))(256))
    # 
    
  }
  
  checkFlowsByVisit = T
  if (checkFlowsByVisit) {
    
    inCountMgsList = sapply(wellnessUniqSubjs, function(uniqSubj) {
      currSamples = paste(uniqSubj, c("v1","v2","v3","v4"), sep="_")
      currMat = wellnessMgsMat[,currSamples]
      
      v1Digit = currMat[,1] >0 
      v2Digit = currMat[,2] >0 
      v3Digit = currMat[,3] >0 
      v4Digit = currMat[,4] >0 
      
      v1Mgs = names(v1Digit[v1Digit])
      v2Mgs = names(v2Digit[v2Digit])
      v3Mgs = names(v3Digit[v3Digit])
      v4Mgs = names(v4Digit[v4Digit])
      
      inV12 = v2Mgs[!v2Mgs %in% v1Mgs]
      inV23 = v3Mgs[!v3Mgs %in% v2Mgs]
      inV34 = v4Mgs[!v4Mgs %in% v3Mgs]
      
      inAvgs = c(length(inV12)/length(v2Mgs),
                 length(inV23)/length(v3Mgs),
                 length(inV34)/length(v4Mgs))
      return(mean(inAvgs))
      
    })
    write.table(data.frame(names(inCountMgsList),inCountMgsList), "in.stats.txt", sep="\t")
    
    outCountMgsList = sapply(wellnessUniqSubjs, function(uniqSubj) {
      currSamples = paste(uniqSubj, c("v1","v2","v3","v4"), sep="_")
      currMat = wellnessMgsMat[,currSamples]
      
      v1Digit = currMat[,1] >0 
      v2Digit = currMat[,2] >0 
      v3Digit = currMat[,3] >0 
      v4Digit = currMat[,4] >0 
      
      v1Mgs = names(v1Digit[v1Digit])
      v2Mgs = names(v2Digit[v2Digit])
      v3Mgs = names(v3Digit[v3Digit])
      v4Mgs = names(v4Digit[v4Digit])
      
      outV12 = v1Mgs[!v1Mgs %in% v2Mgs]
      outV23 = v2Mgs[!v2Mgs %in% v3Mgs]
      outV34 = v3Mgs[!v3Mgs %in% v4Mgs]
      
      outAvgs = c(length(outV12)/length(v1Mgs),
                  length(outV23)/length(v2Mgs),
                  length(outV34)/length(v3Mgs))
      return(mean(outAvgs))
      
    })
    write.table(data.frame(names(outCountMgsList),outCountMgsList), "out.stats.txt", sep="\t")
    
    mean(outCountMgsList)
    inOutMgsList = sapply(wellnessUniqSubjs, function(uniqSubj) {
      currSamples = paste(uniqSubj, c("v1","v2","v3","v4"), sep="_")
      currMat = wellnessMgsMat[,currSamples]
      
      v1Digit = currMat[,1] >0 
      v2Digit = currMat[,2] >0 
      v3Digit = currMat[,3] >0 
      v4Digit = currMat[,4] >0 
      
      v1Mgs = names(v1Digit[v1Digit])
      v2Mgs = names(v2Digit[v2Digit])
      v3Mgs = names(v3Digit[v3Digit])
      v4Mgs = names(v4Digit[v4Digit])
      
      inV12 = v2Mgs[!v2Mgs %in% v1Mgs]
      inV23 = v3Mgs[!v3Mgs %in% v2Mgs]
      inV34 = v4Mgs[!v4Mgs %in% v3Mgs]
      
      outV12 = v1Mgs[!v1Mgs %in% v2Mgs]
      outV23 = v2Mgs[!v2Mgs %in% v3Mgs]
      outV34 = v3Mgs[!v3Mgs %in% v4Mgs]
      
      return(list(inV12=inV12,
                  inV23=inV23,
                  inV34=inV34,
                  outV12=outV12,
                  outV23=outV23,
                  outV34=outV34))
      
      
    }, simplify = F)
    inOutMgsJaccList = sapply(wellnessUniqSubjs, function(uniqSubj) {
      currSamples = paste(uniqSubj, c("v1","v2","v3","v4"), sep="_")
      currMat = wellnessMgsMat[,currSamples]
      
      v1Digit = currMat[,1] >0 
      v2Digit = currMat[,2] >0 
      v3Digit = currMat[,3] >0 
      v4Digit = currMat[,4] >0 
      
      v1Mgs = names(v1Digit[v1Digit])
      v2Mgs = names(v2Digit[v2Digit])
      v3Mgs = names(v3Digit[v3Digit])
      v4Mgs = names(v4Digit[v4Digit])
      
      inV12 = v2Mgs[!v2Mgs %in% v1Mgs]
      inV23 = v3Mgs[!v3Mgs %in% v2Mgs]
      inV34 = v4Mgs[!v4Mgs %in% v3Mgs]
      
      outV12 = v1Mgs[!v1Mgs %in% v2Mgs]
      outV23 = v2Mgs[!v2Mgs %in% v3Mgs]
      outV34 = v3Mgs[!v3Mgs %in% v4Mgs]
      jaccMatrix=cbind(c(getJaccFrom(inV12,inV12),
                         getJaccFrom(inV23,inV12),
                         getJaccFrom(inV34,inV12),
                         getJaccFrom(outV12,inV12),
                         getJaccFrom(outV23,inV12),
                         getJaccFrom(outV34,inV12)),
                       c(getJaccFrom(inV12,inV23),
                         getJaccFrom(inV23,inV23),
                         getJaccFrom(inV34,inV23),
                         getJaccFrom(outV12,inV23),
                         getJaccFrom(outV23,inV23),
                         getJaccFrom(outV34,inV23)),
                       c(getJaccFrom(inV12,inV34),
                         getJaccFrom(inV23,inV34),
                         getJaccFrom(inV34,inV34),
                         getJaccFrom(outV12,inV34),
                         getJaccFrom(outV23,inV34),
                         getJaccFrom(outV34,inV34)),
                       c(getJaccFrom(inV12,outV12),
                         getJaccFrom(inV23,outV12),
                         getJaccFrom(inV34,outV12),
                         getJaccFrom(outV12,outV12),
                         getJaccFrom(outV23,outV12),
                         getJaccFrom(outV34,outV12)),
                       c(getJaccFrom(inV12,outV23),
                         getJaccFrom(inV23,outV23),
                         getJaccFrom(inV34,outV23),
                         getJaccFrom(outV12,outV23),
                         getJaccFrom(outV23,outV23),
                         getJaccFrom(outV34,outV23)),
                       c(getJaccFrom(inV12,outV34),
                         getJaccFrom(inV23,outV34),
                         getJaccFrom(inV34,outV34),
                         getJaccFrom(outV12,outV34),
                         getJaccFrom(outV23,outV34),
                         getJaccFrom(outV34,outV34)))
      colnames(jaccMatrix) = c("inV12","inV23","inV34","outV12","outV23","outV34")
      rownames(jaccMatrix) = c("inV12","inV23","inV34","outV12","outV23","outV34")
      
      return(jaccMatrix)
      
      
    }, simplify = F)
    inOutMgsJaccReduced = Reduce("+",inOutMgsJaccList)
    inOutMgsJaccReduced = inOutMgsJaccReduced / length(inOutMgsJaccList)
    
    plotTileHeatmap2 <- function(jaccMat) {
      
      hr =  heatmap.2(jaccMat)
      dev.off()
      hrRows= rownames(jaccMat)[hr$rowInd]
      hrCols= colnames(jaccMat)[hr$colInd]
      colNews <- c(colorRampPalette(c("#e7f0fa", "#c9e2f6", "#95cbee", "#0099dc", "#4ab04a", "#ffd73e"))(10),
                   colorRampPalette(c("#eec73a", "#e29421", "#e29421", "#f05336","#ce472e"), bias=2)(90))
      
      
      quantile_range <- quantile(jaccMat, probs = seq(0, 1, 0.2))
      color_palette <- colorRampPalette(c("#3794bf", "#FFFFFF", "#df8640"))(length(quantile_range) - 1)
      label_text <- rollapply(round(quantile_range, 2), width = 2, by = 1, FUN = function(i) paste(i, collapse = " : "))
      theme_change <- theme(plot.background = element_blank(),
                            panel.grid.minor = element_blank(),
                            panel.grid.major = element_blank(),
                            panel.background = element_blank(),
                            panel.border = element_blank(),
                            axis.line = element_blank(),
                            axis.ticks = element_blank(),
                            #legend.position = "none",
                            #axis.text.x = element_blank(),
                            #axis.text.y = element_blank(),
                            axis.title.x = element_blank(),
                            axis.title.y = element_blank()
      )
      mod_mat <- matrix(findInterval(jaccMat, quantile_range, all.inside = TRUE), nrow = nrow(jaccMat))
      rownames(mod_mat) = rownames(jaccMat)
      colnames(mod_mat) = colnames(jaccMat)
      
      mod_tab = melt(mod_mat)
      mod_tab$Var1 = factor(mod_tab$Var1, levels = hrRows)
      mod_tab$Var2 = factor(mod_tab$Var2, levels = hrCols)
      
      
      p <- ggplot(mod_tab, aes(x = Var2, y = Var1, fill = factor(value))) +
        geom_tile(color = "black") +
        scale_fill_manual(values = color_palette, name = "", labels = label_text) +
        theme_change
      plot(p)
      return(list(matrix=mod_mat, table=mod_tab))
      
    }
    plotTileHeatmap2(inOutMgsJaccReduced)
    heatmap.2(inOutMgsJaccReduced, 
              lwid = c(3,8),
              lhei = c(3,8),
              key = T, 
              key.title = "", key.ylab = "", key.xlab = "",
              key.xtickfun = function() {
                breaks = pretty(parent.frame()$breaks)
                breaks = breaks[c(1,length(breaks))]
                list(at = parent.frame()$scale01(breaks),
                     labels = breaks)
              },
              key.ytickfun = function() {
                breaks = pretty(parent.frame()$breaks)
                breaks = breaks[c(1,length(breaks))]
                list(at = parent.frame()$scale01(breaks),
                     labels = breaks)
              },
              key.par = list(cex=1.0),
              cexRow = 1, cexCol = 1,
              col=colorRampPalette(c("white","skyblue"))(10),
              trace="none",
              sepcolor = "black",
              colsep=0:6,
              rowsep=0:6,
              sepwidth = c(0.01,0.01))
    
    inOutSpeciesUniqList = list(inV12= unique(unlist(lapply(inOutMgsList, "[", 1))),
                                inV23= unique(unlist(lapply(inOutMgsList, "[", 2))),
                                inV34= unique(unlist(lapply(inOutMgsList, "[", 3))),
                                outV12= unique(unlist(lapply(inOutMgsList, "[", 4))),
                                outV23= unique(unlist(lapply(inOutMgsList, "[", 5))),
                                outV34= unique(unlist(lapply(inOutMgsList, "[", 6))))
    
    inOutSpeciesCountsList = list(inV12= getCountsThroughTable(unlist(lapply(inOutMgsList, "[", 1)), T),
                                  inV23= getCountsThroughTable(unlist(lapply(inOutMgsList, "[", 2)), T),
                                  inV34= getCountsThroughTable(unlist(lapply(inOutMgsList, "[", 3)), T),
                                  outV12= getCountsThroughTable(unlist(lapply(inOutMgsList, "[", 4)), T),
                                  outV23= getCountsThroughTable(unlist(lapply(inOutMgsList, "[", 5)), T),
                                  outV34= getCountsThroughTable(unlist(lapply(inOutMgsList, "[", 6))), T)
    
    
    
    lapply(inOutSpeciesUniqList, length)
    
    allSpecies = rownames(wellnessMgsMat)
    diffAllPairs = lapply(inOutSpeciesCountsList[1:3], function(speciesInCounts) {
      
      diffList = lapply(inOutSpeciesCountsList[4:6], function(speciesOutCounts) {
        
        #unionSpecies = unique(c(names(speciesInCounts), names(speciesOutCounts)))
        diffs = sapply(allSpecies, function(currSpecies){
          
          InCounts = speciesInCounts[currSpecies]
          OutCounts = speciesOutCounts[currSpecies]
          
          if(is.na(InCounts)) InCounts = 0
          if(is.na(OutCounts)) OutCounts = 0
          diff = InCounts-OutCounts
          #if (InCounts == 0 & OutCounts == 0) return(NA)
          return(diff)
          
        }, simplify = F)
        diffs = unlist(diffs)
        return(diffs)
      })
      diffMat = do.call(cbind, diffList)
      return(diffMat) 
    })
    diffAllPairMat = do.call(cbind, diffAllPairs)
    colnames(diffAllPairMat)  = c("12.12", "12.23","12.34",
                                  "23.12", "23.23","23.34",
                                  "34.12", "34.23","34.34")
    rownames(diffAllPairMat) = unlist(lapply(strsplit(rownames(diffAllPairMat), split="\\."), "[",1))
    View(diffAllPairMat)
    
    diffMatrixList = sapply(1:nrow(diffAllPairMat), function(ind){
      outMat = matrix(diffAllPairMat[ind,], nrow=3, ncol=3)
      colnames(outMat)=c("out12","out23","out34")
      rownames(outMat)=c("in12","in23","in34")
      return(outMat)
    }, simplify = F)
    names(diffMatrixList) = rownames(diffAllPairMat)
    avgDiffMatrix = Reduce("+", diffMatrixList)
    avgDiffMatrix = avgDiffMatrix / 1977
    
    hist(diffAllPairs[[1]][[1]])
    hist(diffAllPairs[[1]][[2]])
    hist(diffAllPairs[[1]][[3]])
    
    
    sapply(1:3, function(indI) {
      sapply(1:3, function(indJ) {
        qq = diffAllPairs[[indI]][[indJ]][diffAllPairs[[indI]][[indJ]]>5]
        qqq = unlist(lapply(strsplit(names(qq), split = "\\."), "[",1))
        speciesNames = getSpeciesName(qqq, taxo)
        print(speciesNames)
        print(c(indI, indJ))
      })
    })
    
    perIndividuals = lapply(lapply(inOutMgsList, function(x) unique(unlist(x))), length)
    hist(unlist(perIndividuals), main="", las=1)
    
    overall = length(unique(unlist(unlist(inOutMgsList))))
    
    
  }
  
  checkSimilarity= F
  if (checkSimilarity) {
    
    #wellnessMgsAllData = mergeMatUpdated[,wellnessSamples]
    #allCor = cor(mergeMatUpdated)
    mean(allCor)
    
    wellnessMgsCor = cor(wellnessMgsMat)
    subjCorList = lapply(wellnessSamplesBySubjects, function(x) { 
      currMat = wellnessMgsCor[x,x]
      return(currMat[lower.tri(currMat)])
    })
    medianSubjCorList = unlist(lapply(subjCorList, median))
    boxplot(subjCorList[order(medianSubjCorList)], 
            cex.axis=0.7,
            las=2)
    abline(h=median(allCor))
    
    maxCorSamples = unlist(sapply(wellnessSamples, function(currSample){
      currCors = wellnessMgsCor[currSample,]
      currCors = currCors[names(currCors)!=currSample]
      maxInd = which.max(currCors)#
      return(names(currCors)[maxInd])
    }, simplify = T))
    
    
    pairedCheckTab = data.frame(sample=wellnessSamples,
                                matched=maxCorSamples, 
                                subj = gsub("_v[1-4]","",wellnessSamples),
                                matchedSubj = gsub("_v[1-4]","",maxCorSamples),
                                stringsAsFactors = F)
    
    tt=table(pairedCheckTab$subj == pairedCheckTab$matchedSubj)
    barplot(tt)
    
    ttt = unique(pairedCheckTab[pairedCheckTab$subj != pairedCheckTab$matchedSubj,"subj"])
    View(pairedCheckTab[pairedCheckTab$subj != pairedCheckTab$matchedSubj, ])
    
  }
  
  checkDistMode = F
  if (checkDistMode) {
    wellnessMetadata = basicMetaMapUpdated[basicMetaMapUpdated$dataset.ID=="id28",]
    
    mgsMat = mergeMatUpdated
    metaTab = basicMetaMapUpdated
    
    wellnessSubjs = (wellnessMetadata$metadata.ID)
    missingSubjs1 = names(table(wellnessSubjs)[table(wellnessSubjs)==3])
    exceptionSubjs = c("X3452","X3834","X3918","X4215")
    missingSubjs2 = c("X3913","X3535","X3016")
    missingSubjs1 = missingSubjs1[!missingSubjs1 %in% exceptionSubjs]
    
    wellnessMetadata = wellnessMetadata[!wellnessMetadata$metadata.ID %in% missingSubjs1, ]
    wellnessMetadata = wellnessMetadata[!wellnessMetadata$metadata.ID %in% missingSubjs2, ]
    wellnessMgsMat = mergeMatUpdated[,wellnessMetadata$sample.ID]
    
    
  }
  
  scatterPlot = T
  if (scatterPlot) {
    
    targetMsp = "msp_0033"
    targetClinParams = c("S.CRPUS.A1",
                         "S.TNTHS.A1",
                         "B.MONO.A7")
    targetMspVec = wellnessMgsMat[targetMsp,]
    targetClinVec = wellnessClinicalMatrix[targetClinParams[1],]
    
    plot(targetMspVec, wellnessClinicalMatrix[targetClinParams[1],])
    plot(targetMspVec, wellnessClinicalMatrix[targetClinParams[2],])
    
    plot(targetMspVec, wellnessClinicalMatrix[targetClinParams[3],])
    
    
    bFragilisCrpTab = data.frame(mgs=targetMspVec, 
                                 clin=as.numeric(scale(wellnessClinicalMatrix[targetClinParams[1],])),
                                 subj=wellnessSubjects,
                                 type="CRP",
                                 stringsAsFactors = F)
    bFragilisTntTab = data.frame(mgs=targetMspVec, 
                                 clin=as.numeric(scale(wellnessClinicalMatrix[targetClinParams[2],])),
                                 subj=wellnessSubjects,
                                 type="Troponin T",
                                 stringsAsFactors = F)
    bFragilisMonoTab = data.frame(mgs=targetMspVec, 
                                  clin=as.numeric(scale(wellnessClinicalMatrix[targetClinParams[3],])),
                                  subj=wellnessSubjects,
                                  type="Monocyte",
                                  stringsAsFactors = F)
    
    bFragilisAllTab = rbind(bFragilisCrpTab,
                            bFragilisTntTab,
                            bFragilisMonoTab)
    
    bFragilisTab = data.frame(mgs=targetMspVec, 
                              crp=wellnessClinicalMatrix[targetClinParams[1],],
                              tnt=wellnessClinicalMatrix[targetClinParams[2],],
                              mono=wellnessClinicalMatrix[targetClinParams[3],],
                              subj=wellnessSubjects,
                              stringsAsFactors = F)
    
    
    bFragilisTabPresent = bFragilisTab[bFragilisTab$mgs!=0,]
    
    ggplot(bFragilisTab, aes(x=mgs, y=crp)) +
      geom_point(colour="#87ceeb99",size=3)+
      geom_smooth(method = "lm",  se = F, colour="#000000AA")+
      theme(panel.grid.major = element_blank(), 
            panel.grid.minor = element_blank(),
            #legend.position = "none",
            text = element_blank(),
            panel.background = element_rect(fill = "white", 
                                            colour = "black", 
                                            size = 0.5, 
                                            linetype = "solid"))
    ggplot(bFragilisTab, aes(x=mgs, y=tnt)) +
      geom_point(colour="#ff000099",size=3)+
      geom_smooth(method = "lm",  se = F, colour="#000000AA")+
      theme(panel.grid.major = element_blank(), 
            panel.grid.minor = element_blank(),
            #legend.position = "none",
            text = element_blank(),
            panel.background = element_rect(fill = "white", 
                                            colour = "black", 
                                            size = 0.5, 
                                            linetype = "solid"))
    
    ggplot(bFragilisTab, aes(x=mgs, y=mono)) +
      geom_point(colour="#0B662399", size=3)+
      geom_smooth(method = "lm",  se = F, colour="#000000AA")+
      theme(panel.grid.major = element_blank(), 
            panel.grid.minor = element_blank(),
            #legend.position = "none",
            text = element_blank(),
            panel.background = element_rect(fill = "white", 
                                            colour = "black", 
                                            size = 0.5, 
                                            linetype = "solid")) 
    
    
  }
  
  newMarkovForFMT = T
  if (newMarkovForFMT) {
    
    getInflow <- function(metaTab, mgsMat) {
      uniqSubj = unique(metaTab$metadata.ID)
      scoresByMsp = do.call(rbind, sapply(rownames(mgsMat), function(currMgs){
        
        markovSeqOfMsp = (sapply(uniqSubj, function(subj) {
          currSamples = metaTab$sample.ID[metaTab$metadata.ID==subj]
          currVisits = metaTab$subtype[match(currSamples, metaTab$sample.ID)]
          currSamples = currSamples[order(currVisits)]
          currMgsMat = mgsMat[,currSamples]
          currMgsSubjVec = (currMgsMat[currMgs,] >0)*1
          currMgsSubjVec = as.character(currMgsSubjVec)
          
        }, simplify = F))
        
        myMarkovFit<-markovchainFit(data=markovSeqOfMsp,confidencelevel = .9,method = "mle")
        myMarkovMat = myMarkovFit$estimate@transitionMatrix
        myMarkovUpper = myMarkovFit$upperEndpointMatrix
        myMarkovLower = myMarkovFit$lowerEndpointMatrix
        myMarkovSE = myMarkovFit$standardError
        
        if(dim(myMarkovMat)[1]==1) {
          rname = rownames(myMarkovMat)
          #if (rname == "0") return(0)
          #if (rname == "1") return(Inf)
          if (rname == "0") {
            colonized = 0
            colonizedUpper = NA
            colonizedLower = NA
            colonizedSE = NA
            result = c(colonized, colonizedUpper, colonizedLower, colonizedSE)
            return(result)
          }
          
          if (rname == "1") {
            colonized = 1
            colonizedUpper = NA
            colonizedLower = NA
            colonizedSE = NA
            result = c(colonized, colonizedUpper, colonizedLower, colonizedSE)
            
            return(result)
          }
        }
        else {
          colonized = myMarkovMat["0","1"] 
          colonizedUpper = myMarkovUpper["0","1"]
          colonizedLower = myMarkovLower["0","1"]
          colonizedSE = myMarkovSE["0","1"]
          result = c(colonized, colonizedUpper, colonizedLower, colonizedSE)
          
          #preserved = myMarkovMat["1","1"] 
          #relativeRisk = colonized/ preserved
          #return(relativeRisk)
          return(result)
        }
        
      }, simplify = F))
      
      rownames(scoresByMsp) = rownames(mgsMat)
      colnames(scoresByMsp) = c("inflow","upper","lower","SE")
      return(scoresByMsp)
      
    }
    getOutflow <- function(metaTab, mgsMat) {
      
      uniqSubj = unique(metaTab$metadata.ID)
      scoresByMsp = do.call(rbind, sapply(rownames(mgsMat), function(currMgs){
        
        markovSeqOfMsp = (sapply(uniqSubj, function(subj) {
          currSamples = metaTab$sample.ID[metaTab$metadata.ID==subj]
          currVisits = metaTab$subtype[match(currSamples, metaTab$sample.ID)]
          currSamples = currSamples[order(currVisits)]
          currMgsMat = mgsMat[,currSamples]
          currMgsSubjVec = (currMgsMat[currMgs,] >0)*1
          currMgsSubjVec = as.character(currMgsSubjVec)
          
        }, simplify = F))
        
        myMarkovFit<-markovchainFit(data=markovSeqOfMsp,confidencelevel = .9,method = "mle")
        myMarkovMat = myMarkovFit$estimate@transitionMatrix
        myMarkovUpper = myMarkovFit$upperEndpointMatrix
        myMarkovLower = myMarkovFit$lowerEndpointMatrix
        myMarkovSE = myMarkovFit$standardError
        
        if(dim(myMarkovMat)[1]==1) {
          rname = rownames(myMarkovMat)
          #if (rname == "0") return(Inf)
          #if (rname == "1") return(0)
          if (rname == "0") {
            resistant = 1
            resistantUpper = NA
            resistantLower = NA
            resistantSE = NA
            result = c(resistant, resistantUpper, resistantLower, resistantSE)
            return(result) 
          }
          if (rname == "1") {
            resistant = 0
            resistantUpper = NA
            resistantLower = NA
            resistantSE = NA
            result = c(resistant, resistantUpper, resistantLower, resistantSE)
            return(result)
          }
        }
        else {
          resistant = myMarkovMat["1","0"] 
          resistantUpper = myMarkovUpper["1","0"] 
          resistantLower = myMarkovLower["1","0"] 
          resistantSE = myMarkovSE["1","0"]
          result = c(resistant, resistantUpper, resistantLower, resistantSE)
          
          #preserved = myMarkovMat["0","0"] 
          #relativeRisk = resistant/ preserved
          #return(relativeRisk)
          return(result)
          
        }
        
      }, simplify = F))
      rownames(scoresByMsp) = rownames(mgsMat)
      colnames(scoresByMsp) = c("outflow","upper","lower","SE")
      return(scoresByMsp)
      
    }
    
    
    fmtMgsRData = "J:\\Deposit\\Project\\2018_microbiome_atlas\\gctTables\\FMT.final.mgs.med.vec.10M.RData"
    fmtMetadataFile = "J:\\Deposit\\Project\\2018_microbiome_atlas\\atlas.metadata.collected//atlas.metadata.all//FMT.sample.metadata.txt"
    
    
    
    load(fmtMgsRData)
    fmtMgsMat = mgs_med_vec_10m
    fmtMetaTab = read.table(fmtMetadataFile, sep="\t", header=T, stringsAsFactors = F)
    
    fat_don_8_samples = c("ERR1297678", "ERR1297685", "ERR1297727", "ERR1297846")
    fat_don_11_samples = c("ERR1297659", "ERR1297777", "ERR1297822", "ERR1297674", "ERR1297686", "ERR1297800", "ERR1297843", "ERR1297755", "ERR1297774")
    
    fmtMgsMat[, fat_don_8_samples]
    
    fmtMetaTab$newId = paste( fmtMetaTab$metadata.ID, fmtMetaTab$subtype, sep="_" )
    
    fmtReplicatesBySamples = split(fmtMetaTab$sample.ID, fmtMetaTab$newId)
    fmtMgsMatNew = do.call(cbind, lapply(fmtReplicatesBySamples, function(currSamples) { 
      return(rowMeans(fmtMgsMat[,currSamples]))
    }))
    
    
    
    
    qCutOff =(2^(4.042488))/10^9
    min(americanMgsMat[americanMgsMat!=0])
    
    fmtMgsMatNewImputed = fmtMgsMatNew
    fmtMgsMatNewImputed[fmtMgsMatNewImputed < qCutOff] <- 0
    
    
    fmtEbiSampleIdMap = fmtMetaTab[,c("sample.ID","newId")]
    
    fmtMetaTab2 = fmtMetaTab
    fmtMetaTab2$sample.ID = fmtMetaTab2$newId
    fmtMetaTab2 = unique(fmtMetaTab2)
    fmtMetaTab2 = fmtMetaTab2[fmtMetaTab2$type != "Control",]
    #fmtMetaTab2 = fmtMetaTab2[fmtMetaTab2$subtype %in% c("v1","v2"),]
    fmtMetaTab2 = fmtMetaTab2[fmtMetaTab2$subtype %in% c("v1","v2","v4"),]
    
    
    fmtMgsMatNewImputed2 = fmtMgsMatNewImputed[,colnames(fmtMgsMatNewImputed) %in% fmtMetaTab2$sample.ID ]
    recipient_donor_8_mat = fmtMgsMatNewImputed[,colnames(fmtMgsMatNewImputed) %in% fat_don_8_samples ]
    donor_11_mat = fmtMgsMatNewImputed[,colnames(fmtMgsMatNewImputed) %in% fat_don_11_samples]
    
    fat_012_sample_ids = c("FAT_012_v1", "FAT_012_v2", "FAT_012_v3", "FAT_012_v4", "FAT_012_v5")
    fat_015_sample_ids = c("FAT_015_v1","FAT_015_v2","FAT_015_v4","FAT_015_v5")
    
    
    
    #donor_8_vec = rowMeans(fmtMgsMatNewImputed[,fat_don_8_samples])
    donor_8_vec = rowMeans(fmtMgsMat[,fat_don_8_samples])
    recipient_012_vec = rowMeans(fmtMgsMatNewImputed[,fat_012_sample_ids])
    
    before_012 = fmtMgsMatNewImputed2[,"FAT_012_v1"] ## donor 8
    after_012 = fmtMgsMatNewImputed2[,"FAT_012_v2"]
    after_012_v4 = fmtMgsMatNewImputed2[,"FAT_012_v4"]
    
    
    test = donor_8_vec & !before_012  
    test = donor_8_vec & before_012 & after_012
    test = donor_8_vec & before_012 & after_012 & after_012_v4
    
    table(test[commensalSpecies])
    
    
    
    donor8_only <- donor_8_vec & !before_012 # 104 species
    recipient012_only <- !donor_8_vec & before_012 # 86 species
    both8_012 <- donor_8_vec & before_012 # 134 species
    
    donor8_after_only <- donor_8_vec & !before_012 & after_012 # 45 species
    donor8_wo_after <- donor_8_vec & !before_012 & !after_012 # 59 species
    
    boxplot(list(donor = inflow[names(donor8_only[donor8_only])],
                 recipient = inflow[names(recipient012_only[recipient012_only])],
                 both = inflow[names(both8_012[both8_012])]))
    
    boxplot(list(donor = outflow[names(donor8_only[donor8_only])],
                 recipient = outflow[names(recipient012_only[recipient012_only])],
                 both = outflow[names(both8_012[both8_012])]))
    
    
    boxplot(list(left = inflow[names(donor8_after_only[donor8_after_only])], 
                 gone = inflow[names(donor8_wo_after[donor8_wo_after])],
                 recipient = inflow[names(recipient012_only[recipient012_only])]))
    
    boxplot(list(left = outflow[names(donor8_after_only[donor8_after_only])], 
                 gone = outflow[names(donor8_wo_after[donor8_wo_after])],
                 recipient = outflow[names(recipient012_only[recipient012_only])]))
    
    
    t.test(inflow[names(donor8_after_only[donor8_after_only])], inflow[names(both8_012[both8_012])])
    t.test(inflow[names(donor8_after_only[donor8_after_only])], inflow[names(donor8_wo_after[donor8_wo_after])])
    t.test(inflow[names(donor8_after_only[donor8_after_only])], inflow[names(recipient012_only[recipient012_only])])
    t.test(inflow[names(donor8_wo_after[donor8_wo_after])], inflow[names(recipient012_only[recipient012_only])])
    
    t.test(outflow[names(donor8_after_only[donor8_after_only])], outflow[names(donor8_wo_after[donor8_wo_after])])
    t.test(outflow[names(donor8_wo_after[donor8_wo_after])], outflow[names(recipient012_only[recipient012_only])])
    
    
    #donor_11_vec = rowMeans(fmtMgsMatNewImputed[,fat_don_11_samples])
    donor_11_vec = rowMeans(fmtMgsMat[,fat_don_11_samples])
    recipient_015_vec = rowMeans(fmtMgsMatNewImputed[,fat_015_sample_ids])
    
    
    before_015 = fmtMgsMatNewImputed2[,"FAT_015_v1"] ## donor 11
    after_015 = fmtMgsMatNewImputed2[,"FAT_015_v2"]
    after_015_v4 = fmtMgsMatNewImputed2[,"FAT_015_v4"]
    
    donor11_only <- donor_11_vec & !before_015 # 79 species
    recipient015_only <- !donor_11_vec & before_015 # 92 species
    both11_015 <- donor_11_vec & before_015 # 158 species
    
    donor11_after_only <- donor_11_vec & !before_015 & after_015 # 19 species
    donor11_wo_after <- donor_11_vec & !before_015 & !after_015 # 60 species
    recipient015_after_only <- !donor_11_vec & before_015  & after_015 # 46 species
    recipient015_wo_after <- !donor_11_vec & before_015  & !after_015 # 46 species
    
    
    
    boxplot(list(donor = inflow[names(donor11_only[donor11_only])],
                 recipient = inflow[names(recipient015_only[recipient015_only])],
                 both = inflow[names(both11_015[both11_015])]))
    
    boxplot(list(donor = outflow[names(donor11_only[donor11_only])],
                 recipient = outflow[names(recipient015_only[recipient015_only])],
                 both = outflow[names(both11_015[both11_015])]))
    
    
    boxplot(list(left = inflow[names(donor11_after_only[donor11_after_only])], 
                 gone = inflow[names(donor8_wo_after[donor8_wo_after])],
                 recipient = inflow[names(recipient015_only[recipient015_only])]))
    
    boxplot(list(left = outflow[names(donor11_after_only[donor11_after_only])], 
                 gone = outflow[names(donor11_wo_after[donor11_wo_after])],
                 recipient = outflow[names(recipient015_only[recipient015_only])]))
    
    t.test(inflow[names(donor11_after_only[donor11_after_only])], inflow[names(both11_015[both11_015])])
    t.test(inflow[names(donor11_after_only[donor11_after_only])], inflow[names(recipient015_only[recipient015_only])])
    
    t.test(inflow[names(donor11_after_only[donor11_after_only])], inflow[names(donor11_wo_after[donor11_wo_after])])
    t.test(inflow[names(donor11_after_only[donor11_after_only])], inflow[names(recipient015_only[recipient015_only])])
    
    t.test(outflow[names(donor11_after_only[donor11_after_only])], outflow[names(donor11_wo_after[donor11_wo_after])])
    t.test(outflow[names(donor11_wo_after[donor11_wo_after])], outflow[names(recipient015_only[recipient015_only])])
    
    
    
    
    inflowFmtMat = getInflow(fmtMetaTab2, fmtMgsMatNewImputed)
    outflowFmtMat = getOutflow(fmtMetaTab2, fmtMgsMatNewImputed)
    
    
    overlapSpecies = rownames(inflowFmtMat)[rownames(inflowFmtMat) %in% rownames(inflowOut)]
    plot(inflowOut[overlapSpecies,1], inflowFmtMat[overlapSpecies,1])
    plot(outflowOut[overlapSpecies,1], outflowFmtMat[overlapSpecies,1])
    
    ### donor profiles
    
    
    ### recipent metadata
    
    
    ### pre-FMT vs 2 day
    ### 2-day, 42 day, 84 day
  }
  
  newMarkovForLU_T1D = F
  if (newMarkovForLU_T1D) {
    exSubj = c("M04.4","M04.3","M04.1","M03.5","M03.1","M02.5","M01.5")
    lu.t1d.metaTab = basicMetaMapUpdated[basicMetaMapUpdated$dataset.ID=="id35",]
    lu.t1d.metaTab = lu.t1d.metaTab[!lu.t1d.metaTab$metadata.ID %in% exSubj, ]
    lu.t1d.metaTab$subtype = tolower(lu.t1d.metaTab$subtype)
    lu.t1d.mgsMat = mergeMatUpdatedImputed[,match(lu.t1d.metaTab$sample.ID,colnames(mergeMatUpdatedImputed))]
    dim(lu.t1d.mgsMat)
    lu.t1d.mgsMat = lu.t1d.mgsMat[rowSums(lu.t1d.mgsMat)!=0,]
    lu.t1d.mgsMat = lu.t1d.mgsMat[!apply(lu.t1d.mgsMat,1,all),]
    dim(lu.t1d.mgsMat)
    
    getInflow <- function(metaTab, mgsMat) {
      uniqSubj = unique(metaTab$metadata.ID)
      scoresByMsp = do.call(rbind, sapply(rownames(mgsMat), function(currMgs){
        
        markovSeqOfMsp = (sapply(uniqSubj, function(subj) {
          currSamples = metaTab$sample.ID[metaTab$metadata.ID==subj]
          currVisits = metaTab$subtype[match(currSamples, metaTab$sample.ID)]
          currSamples = currSamples[order(currVisits)]
          currMgsMat = mgsMat[,currSamples]
          currMgsSubjVec = (currMgsMat[currMgs,] >0)*1
          currMgsSubjVec = as.character(currMgsSubjVec)
          
        }, simplify = F))
        
        myMarkovFit<-markovchainFit(data=markovSeqOfMsp,confidencelevel = .9,method = "mle")
        myMarkovMat = myMarkovFit$estimate@transitionMatrix
        myMarkovUpper = myMarkovFit$upperEndpointMatrix
        myMarkovLower = myMarkovFit$lowerEndpointMatrix
        myMarkovSE = myMarkovFit$standardError
        
        if(dim(myMarkovMat)[1]==1) {
          rname = rownames(myMarkovMat)
          #if (rname == "0") return(0)
          #if (rname == "1") return(Inf)
          if (rname == "0") {
            colonized = 0
            colonizedUpper = NA
            colonizedLower = NA
            colonizedSE = NA
            result = c(colonized, colonizedUpper, colonizedLower, colonizedSE)
            return(result)
          }
          
          if (rname == "1") {
            colonized = 1
            colonizedUpper = NA
            colonizedLower = NA
            colonizedSE = NA
            result = c(colonized, colonizedUpper, colonizedLower, colonizedSE)
            
            return(result)
          }
        }
        else {
          colonized = myMarkovMat["0","1"] 
          colonizedUpper = myMarkovUpper["0","1"]
          colonizedLower = myMarkovLower["0","1"]
          colonizedSE = myMarkovSE["0","1"]
          result = c(colonized, colonizedUpper, colonizedLower, colonizedSE)
          
          #preserved = myMarkovMat["1","1"] 
          #relativeRisk = colonized/ preserved
          #return(relativeRisk)
          return(result)
        }
        
      }, simplify = F))
      
      rownames(scoresByMsp) = rownames(mgsMat)
      colnames(scoresByMsp) = c("inflow","upper","lower","SE")
      return(scoresByMsp)
      
    }
    getOutflow <- function(metaTab, mgsMat) {
      
      uniqSubj = unique(metaTab$metadata.ID)
      scoresByMsp = do.call(rbind, sapply(rownames(mgsMat), function(currMgs){
        
        markovSeqOfMsp = (sapply(uniqSubj, function(subj) {
          currSamples = metaTab$sample.ID[metaTab$metadata.ID==subj]
          currVisits = metaTab$subtype[match(currSamples, metaTab$sample.ID)]
          currSamples = currSamples[order(currVisits)]
          currMgsMat = mgsMat[,currSamples]
          currMgsSubjVec = (currMgsMat[currMgs,] >0)*1
          currMgsSubjVec = as.character(currMgsSubjVec)
          
        }, simplify = F))
        
        myMarkovFit<-markovchainFit(data=markovSeqOfMsp,confidencelevel = .9,method = "mle")
        myMarkovMat = myMarkovFit$estimate@transitionMatrix
        myMarkovUpper = myMarkovFit$upperEndpointMatrix
        myMarkovLower = myMarkovFit$lowerEndpointMatrix
        myMarkovSE = myMarkovFit$standardError
        
        if(dim(myMarkovMat)[1]==1) {
          rname = rownames(myMarkovMat)
          #if (rname == "0") return(Inf)
          #if (rname == "1") return(0)
          if (rname == "0") {
            resistant = 1
            resistantUpper = NA
            resistantLower = NA
            resistantSE = NA
            result = c(resistant, resistantUpper, resistantLower, resistantSE)
            return(result) 
          }
          if (rname == "1") {
            resistant = 0
            resistantUpper = NA
            resistantLower = NA
            resistantSE = NA
            result = c(resistant, resistantUpper, resistantLower, resistantSE)
            return(result)
          }
        }
        else {
          resistant = myMarkovMat["1","0"] 
          resistantUpper = myMarkovUpper["1","0"] 
          resistantLower = myMarkovLower["1","0"] 
          resistantSE = myMarkovSE["1","0"]
          result = c(resistant, resistantUpper, resistantLower, resistantSE)
          
          #preserved = myMarkovMat["0","0"] 
          #relativeRisk = resistant/ preserved
          #return(relativeRisk)
          return(result)
          
        }
        
      }, simplify = F))
      rownames(scoresByMsp) = rownames(mgsMat)
      colnames(scoresByMsp) = c("outflow","upper","lower","SE")
      return(scoresByMsp)
      
    }
    
    inflowLuMat = getInflow(lu.t1d.metaTab,lu.t1d.mgsMat)
    outflowLuMat = getOutflow(lu.t1d.metaTab,lu.t1d.mgsMat)
    
    luInflow = inflowLuMat[,1]
    luOutflow = outflowLuMat[,1]
    
    
    overlapLuSpecies = names(luOutflow)[names(luOutflow) %in% names(outflow)]
    
    
    
    
    plot(luInflow[ (overlapLuSpecies)], inflow[ (overlapLuSpecies)])
    abline(a=0,b=1,col="red")
    
    plot(luOutflow[ (overlapLuSpecies)], outflow[ (overlapLuSpecies)])
    abline(a=0,b=1,col="red")
    
    cor.test(luInflow[ (overlapLuSpecies)], inflow[ (overlapLuSpecies)])
    cor.test(luOutflow[ (overlapLuSpecies)], outflow[ (overlapLuSpecies)])
    
    
    plot(luInflow[names(inflow)], inflow[names(inflow)])
    plot(luOutflow[names(outflow)], outflow[names(outflow)])
    
    abline(a=0,b=1,col="red")
    cor.test(luInflow[names(inflow)], inflow[names(inflow)])
    cor.test(luOutflow[names(outflow)], outflow[names(outflow)])
    
    table(names(luInflow[names(inflow)]) == names(inflow))
    table(names(luOutflow[names(outflow)]) == names(outflow))
    
    par(mar=c(8,4,4,1))
    inflowList = list(Sweden=inflow[names(inflow)], Luxembourg=luInflow[names(inflow)])
    outflowList = list(Sweden=outflow[names(outflow)], Luxembourg=luOutflow[names(outflow)])
    
    boxplot(inflowList,las=2)
    wilcox.test(inflowList$Sweden, inflowList$Luxembourg)
    
    boxplot(outflowList,las=2)
    wilcox.test(outflowList$Sweden, outflowList$Luxembourg)
    
    ### T1D case ###
    lu.t1d.metaTab.t1d = lu.t1d.metaTab[lu.t1d.metaTab$type=="Case",]
    lu.t1d.mgsMat.t1d = lu.t1d.mgsMat[,match(lu.t1d.metaTab.t1d$sample.ID, colnames(lu.t1d.mgsMat))] # 687 x 21
    lu.t1d.mgsMat.t1d = lu.t1d.mgsMat.t1d[rowSums(lu.t1d.mgsMat.t1d)!=0,]
    lu.t1d.mgsMat.t1d = lu.t1d.mgsMat.t1d[!apply(lu.t1d.mgsMat.t1d,1,all),]
    
    inflowLuMat.t1d = getInflow(lu.t1d.metaTab.t1d, lu.t1d.mgsMat.t1d)
    outflowLuMat.t1d = getOutflow(lu.t1d.metaTab.t1d, lu.t1d.mgsMat.t1d)
    
    luInflow.t1d = inflowLuMat.t1d[,1]
    luOutflow.t1d = outflowLuMat.t1d[,1]
    
    
    
    ### healthy ###
    lu.t1d.metaTab.con = lu.t1d.metaTab[lu.t1d.metaTab$type=="Control",]
    lu.t1d.mgsMat.con = lu.t1d.mgsMat[,match(lu.t1d.metaTab.con$sample.ID, colnames(lu.t1d.mgsMat))] # 687 x 18
    lu.t1d.mgsMat.con = lu.t1d.mgsMat.con[rowSums(lu.t1d.mgsMat.con)!=0,]
    lu.t1d.mgsMat.con = lu.t1d.mgsMat.con[!apply(lu.t1d.mgsMat.con, 1, all),]
    
    
    inflowLuMat.con = getInflow(lu.t1d.metaTab.con, lu.t1d.mgsMat.con)
    outflowLuMat.con = getOutflow(lu.t1d.metaTab.con, lu.t1d.mgsMat.con)
    
    luInflow.con = inflowLuMat.con[,1]
    luOutflow.con = outflowLuMat.con[,1]
    
    plot(luInflow.con[names(inflow)], inflow[names(inflow)])
    abline(a=0,b=1,col="red")
    
    plot(luOutflow.con[names(outflow)], outflow[names(outflow)])
    abline(a=0,b=1,col="red")
    
    cor.test(luInflow.con[names(inflow)], inflow[names(inflow)])
    cor.test(luOutflow.con[names(outflow)], outflow[names(outflow)])
    
    ### comparing both ###
    plot(luInflow.con[names(luInflow.t1d)], inflow[names(luInflow.t1d)])
    abline(a=0,b=1,col="red")
    
    plot(luOutflow.con[names(luOutflow.t1d)], outflow[names(luOutflow.t1d)])
    abline(a=0,b=1,col="red")
    
    cor.test(luInflow.con[names(luInflow.t1d)], luInflow.t1d[names(luInflow.t1d)])
    cor.test(luOutflow.con[names(luOutflow.t1d)], luOutflow.t1d[names(luOutflow.t1d)])
    
  }
  
  newMarkovForAmerican =  T
  if (newMarkovForAmerican) {
    
    pathogenListFromMathieu = "j://Deposit//Project//2018_microbiome_atlas//pathogen.list.Ecker.et.al.from.Mathieu.txt"
    pathogenList = read.table(pathogenListFromMathieu, header=F, stringsAsFactors = F)[,]
    
    checkGammaFitMode = F
    if (checkGammaFitMode) {
      loadMode = T
      if (loadMode) {
        americanMgsFile = "J:\\Deposit\\Project\\2018_microbiome_atlas\\in.outflow.dataset\\PRJNA354235_vs_hs_10_4_igc2_trim80_id95_rmHosts90_raw_smart_shared_reads_20200505.down10M.fpkm.mgs_med.tsv"
        americanMetaFile= "J:\\Deposit\\Project\\2018_microbiome_atlas\\in.outflow.dataset\\american.metadata.txt"
        americanMgsMat = read.delim(americanMgsFile, sep="\t", stringsAsFactors = F,row.names = 1)
        americanMgsMat = as.matrix(americanMgsMat)
        
        americanMgsMat = americanMgsMat[!rownames(americanMgsMat) %in% suppressMsps2,] # 1989 x 244
        
        americanMetaTab = read.delim(americanMetaFile, sep="\t", stringsAsFactors = F) # 246 x 6
        americanMetaTab = americanMetaTab[americanMetaTab$secondary_sample_accession %in% colnames(americanMgsMat),]
        americanMetaTab = americanMetaTab[,!colnames(americanMetaTab) %in% c("sample_alias","mapped_read_count","timepoint_type") ]
        
        check = T
        if (check) {
          table(table(americanMetaTab$ind_name)==2)
          #all trues
          table(apply(matrix(americanMetaTab$timepoint,nrow = 2 ), 2, function(x) paste(x, collapse=":")))
          #SF05:SF07 --> 46
          #SF05:SF08 --> 38
          #SF06:SF07 --> 33
          #SF06:SF08 --> 38
          
          
        }
        americanMetaTab$timepoint[americanMetaTab$timepoint == "SF05"] = "v1"
        americanMetaTab$timepoint[americanMetaTab$timepoint == "SF06"] = "v2"
        americanMetaTab$timepoint[americanMetaTab$timepoint == "SF07"] = "v2"
        americanMetaTab$timepoint[americanMetaTab$timepoint == "SF08"] = "v2"
        
        colnames(americanMetaTab) = c("sample.ID","metadata.ID","subtype")
        americanMetaTab$type = "Case"
        
        qCutOff =(2^(4.042488))/10^9
        min(americanMgsMat[americanMgsMat!=0])
        
      }
      
      
      
      americanMgsMatScaled = americanMgsMat*10^9
      americanMgsMatScaledLog = log2(americanMgsMatScaled)
      americanMgsMatScaledLog[americanMgsMatScaledLog == -Inf] = 0
      
      americanMgsMatScaledNonZero = americanMgsMatScaled[americanMgsMatScaled!=0]
      americanMgsMatScaledNonZero = log2(americanMgsMatScaledNonZero)
      
      american.fit.gamma = fitdist(americanMgsMatScaledNonZero, "gamma")
      plot(american.fit.gamma)
      american.fit.gamma$estimate
      qgamma(0.01, american.fit.gamma$estimate["shape"], american.fit.gamma$estimate["rate"])
      #4.63775
      
      if (F) {
        qCutOff =(2^(4.042488))/10^9
        min(mergeMatUpdated[mergeMatUpdated!=0])
        
        mergeMatUpdatedImputed = mergeMatUpdated
        mergeMatUpdatedImputed[mergeMatUpdatedImputed<qCutOff] <- 0
        
      }
      
    }
    
    loadMode = T
    if (loadMode) {
      americanMgsFile = "J:\\Deposit\\Project\\2018_microbiome_atlas\\in.outflow.dataset\\PRJNA354235_vs_hs_10_4_igc2_trim80_id95_rmHosts90_raw_smart_shared_reads_20200505.down10M.fpkm.mgs_med.tsv"
      americanMetaFile= "J:\\Deposit\\Project\\2018_microbiome_atlas\\in.outflow.dataset\\american.metadata.txt"
      americanMgsMat = read.delim(americanMgsFile, sep="\t", stringsAsFactors = F,row.names = 1)
      americanMgsMat = as.matrix(americanMgsMat)
      
      americanMgsMat = americanMgsMat[!rownames(americanMgsMat) %in% suppressMsps2,] # 1989 x 244
      
      americanMetaTab = read.delim(americanMetaFile, sep="\t", stringsAsFactors = F) # 246 x 6
      americanMetaTab = americanMetaTab[americanMetaTab$secondary_sample_accession %in% colnames(americanMgsMat),]
      americanMetaTab = americanMetaTab[,!colnames(americanMetaTab) %in% c("sample_alias","mapped_read_count","timepoint_type") ]
      
      check = F
      if (check) {
        table(table(americanMetaTab$ind_name)==2)
        #all trues
        table(apply(matrix(americanMetaTab$timepoint,nrow = 2 ), 2, function(x) paste(x, collapse=":")))
        #SF05:SF07 --> 46
        #SF05:SF08 --> 38
        #SF06:SF07 --> 33
        #SF06:SF08 --> 38
      }
      
      americanMetaTab$timepoint[americanMetaTab$timepoint == "SF05"] = "v1"
      americanMetaTab$timepoint[americanMetaTab$timepoint == "SF06"] = "v2"
      americanMetaTab$timepoint[americanMetaTab$timepoint == "SF07"] = "v2"
      americanMetaTab$timepoint[americanMetaTab$timepoint == "SF08"] = "v2"
      
      colnames(americanMetaTab) = c("sample.ID","metadata.ID","subtype")
      americanMetaTab$type = "Case"
      
      qCutOff =(2^(4.042488))/10^9
      min(americanMgsMat[americanMgsMat!=0])
      
      americanMgsMatImputed = americanMgsMat
      americanMgsMatImputed[americanMgsMatImputed < qCutOff] <- 0
      americanMgsMatImputed = americanMgsMatImputed[rowSums(americanMgsMatImputed)!=0,]
      americanMgsMatImputed = americanMgsMatImputed[!apply(americanMgsMatImputed,1,all),] # 1077 x 114 
      americanMgsMatImputed = americanMgsMatImputed[,match(americanMetaTab$sample.ID,colnames(americanMgsMatImputed))]
      
    }
    
    getSurvMode = T
    if (getSurvMode) {
      getPairSeqs <- function(lenSeqs) {
        lenInds = lenSeqs-1
        pairList = sapply(1:lenInds, function(x) c(x, x+1), simplify = F)
        pairTab = do.call(rbind, pairList)
        return(pairTab)
      }
      getSurvVecByMgs <- function(mgsVec, mgsName, sampleNames, interval=3) {
        
        lenAll = length(mgsVec)
        lenInd = length(mgsVec)-1
        seqs = 1:lenInd
        #interval=3
        survTime = 0
        survEvent = "None"
        
        #####
        #check fist or second visit whether mgs exists
        
        mgsCutDigit <- mgsVec > 0
        presentInds = which(mgsCutDigit)
        if (length(presentInds)==0) {
          survTime = NA
          survEvent = "None"
          return(list(time=survTime, event=survEvent, sample=sampleNames[1], abd = 0, mgs=mgsName))
        }
        startInd = min(presentInds)
        lastInd = startInd
        for (ind in seq_along(presentInds)) {
          if (ind == length(presentInds)) break
          currInd = presentInds[ind]
          nextInd = presentInds[ind+1]
          diff = nextInd - currInd
          if (diff == 1) lastInd = nextInd
        }
        
        if (length(presentInds)!= 0) {
          survTime= (lastInd+1 - startInd)*interval
          survEvent = "Absent"
          if (lastInd == lenAll) {
            survTime = interval*lenInd
            survEvent = "None"
          }
        }
        res=list(time=survTime, event=survEvent, sample=sampleNames[startInd], abd = mgsVec[startInd], mgs=mgsName)
        #print(res)
        return(res)
        
      }
      getSurvivalRates <- function(metaTab, mgsMat, interval=3, halfTime=6) {
        
        uniqSubjs = unique(metaTab$metadata.ID)
        mgsPrevalence = apply(mgsMat, 1, function(currRow){
          mean(currRow>0)
        })
        mgsMeans = rowMeans(mgsMat)
        
        survList = do.call(c,sapply(uniqSubjs, function(subj) {
          currSamples = metaTab$sample.ID[metaTab$metadata.ID==subj]
          currSamples = currSamples[order(currSamples)]
          currMgsMat = mgsMat[,currSamples]
          currMgsMat = currMgsMat[rowMeans(currMgsMat)!=0,]
          currSubSurvVecThroughs = (sapply(rownames(currMgsMat), function(currMgs){
            currSubSurvVecThrough = getSurvVecByMgs(currMgsMat[currMgs,], currMgs, currSamples, interval)
            return(currSubSurvVecThrough)
          }, simplify = F))
          
          return(currSubSurvVecThroughs) 
          
        }, simplify = F))
        
        survTab = data.frame(time=do.call(c,lapply(survList, function(x) x$time)),
                             status=do.call(c,lapply(survList, function(x) x$event)),
                             sample=do.call(c,lapply(survList, function(x) x$sample)),
                             abd=do.call(c,lapply(survList, function(x) x$abd)),
                             mgs=do.call(c,lapply(survList, function(x) x$mgs)),
                             stringsAsFactors = F)
        survTab$logAbd = log10(survTab$abd)
        survivalRates = unlist(sapply(unique(survTab$mgs), function(currMsp) {
          currMspTab = survTab[survTab$mgs == currMsp,]
          if (F) {
            library("survminer")
            currMsp = "msp_0069"
            currMsp = "msp_0041"
            currMsp = "msp_0005"
            currMsp = "msp_0148c"
            currMsp = "msp_0742"
            
            
            currMspTab = survTab[survTab$mgs == currMsp,]
            currMspTab$status <- currMspTab$status == "Absent"
            currSurvFit=survfit(survival::Surv(time, status) ~ 1, data=currMspTab)
            ggsurvplot(currSurvFit, break.time.by = 3, color = "#2E9FDF")
            
          }
          halfTimeSummary = summary(survfit(survival::Surv(currMspTab$time, currMspTab$status == "Absent") ~ 1), times=halfTime)
          halfTimeSurvival = halfTimeSummary$surv
          return(halfTimeSurvival)
        }))
        
        survivalStats = data.frame(surv.rates=survivalRates[match(rownames(mgsMat), names(survivalRates))],
                                   mgs.abd = mgsMeans[match(rownames(mgsMat), names(mgsMeans))],
                                   freq= mgsPrevalence,
                                   mgs=rownames(mgsMat),
                                   species=getSpeciesName(rownames(mgsMat),taxo))
        
        return(survivalStats)
      }
      
      #survivalStats = getSurvivalRates(metaTab, mgsMat, decreaseCut)
      survivalStats_american_6mon = getSurvivalRates(americanMetaTab, americanMgsMatImputed, interval=6, halfTime = 6)
      
    }
    
    getInflow <- function(metaTab, mgsMat) {
      uniqSubj = unique(metaTab$metadata.ID)
      scoresByMsp = do.call(rbind, sapply(rownames(mgsMat), function(currMgs){
        
        markovSeqOfMsp = (sapply(uniqSubj, function(subj) {
          currSamples = metaTab$sample.ID[metaTab$metadata.ID==subj]
          currVisits = metaTab$subtype[match(currSamples, metaTab$sample.ID)]
          currSamples = currSamples[order(currVisits)]
          currMgsMat = mgsMat[,currSamples]
          currMgsSubjVec = (currMgsMat[currMgs,] >0)*1
          currMgsSubjVec = as.character(currMgsSubjVec)
          
        }, simplify = F))
        
        myMarkovFit<-markovchainFit(data=markovSeqOfMsp,confidencelevel = .9,method = "mle")
        myMarkovMat = myMarkovFit$estimate@transitionMatrix
        myMarkovUpper = myMarkovFit$upperEndpointMatrix
        myMarkovLower = myMarkovFit$lowerEndpointMatrix
        myMarkovSE = myMarkovFit$standardError
        
        if(dim(myMarkovMat)[1]==1) {
          rname = rownames(myMarkovMat)
          #if (rname == "0") return(0)
          #if (rname == "1") return(Inf)
          if (rname == "0") {
            colonized = 0
            colonizedUpper = NA
            colonizedLower = NA
            colonizedSE = NA
            result = c(colonized, colonizedUpper, colonizedLower, colonizedSE)
            return(result)
          }
          
          if (rname == "1") {
            colonized = 1
            colonizedUpper = NA
            colonizedLower = NA
            colonizedSE = NA
            result = c(colonized, colonizedUpper, colonizedLower, colonizedSE)
            
            return(result)
          }
        }
        else {
          colonized = myMarkovMat["0","1"] 
          colonizedUpper = myMarkovUpper["0","1"]
          colonizedLower = myMarkovLower["0","1"]
          colonizedSE = myMarkovSE["0","1"]
          result = c(colonized, colonizedUpper, colonizedLower, colonizedSE)
          
          #preserved = myMarkovMat["1","1"] 
          #relativeRisk = colonized/ preserved
          #return(relativeRisk)
          return(result)
        }
        
      }, simplify = F))
      
      rownames(scoresByMsp) = rownames(mgsMat)
      colnames(scoresByMsp) = c("inflow","upper","lower","SE")
      return(scoresByMsp)
      
    }
    getOutflow <- function(metaTab, mgsMat) {
      
      uniqSubj = unique(metaTab$metadata.ID)
      scoresByMsp = do.call(rbind, sapply(rownames(mgsMat), function(currMgs){
        
        markovSeqOfMsp = (sapply(uniqSubj, function(subj) {
          currSamples = metaTab$sample.ID[metaTab$metadata.ID==subj]
          currVisits = metaTab$subtype[match(currSamples, metaTab$sample.ID)]
          currSamples = currSamples[order(currVisits)]
          currMgsMat = mgsMat[,currSamples]
          currMgsSubjVec = (currMgsMat[currMgs,] >0)*1
          currMgsSubjVec = as.character(currMgsSubjVec)
          
        }, simplify = F))
        
        myMarkovFit<-markovchainFit(data=markovSeqOfMsp,confidencelevel = .9,method = "mle")
        myMarkovMat = myMarkovFit$estimate@transitionMatrix
        myMarkovUpper = myMarkovFit$upperEndpointMatrix
        myMarkovLower = myMarkovFit$lowerEndpointMatrix
        myMarkovSE = myMarkovFit$standardError
        
        if(dim(myMarkovMat)[1]==1) {
          rname = rownames(myMarkovMat)
          #if (rname == "0") return(Inf)
          #if (rname == "1") return(0)
          if (rname == "0") {
            resistant = 1
            resistantUpper = NA
            resistantLower = NA
            resistantSE = NA
            result = c(resistant, resistantUpper, resistantLower, resistantSE)
            return(result) 
          }
          if (rname == "1") {
            resistant = 0
            resistantUpper = NA
            resistantLower = NA
            resistantSE = NA
            result = c(resistant, resistantUpper, resistantLower, resistantSE)
            return(result)
          }
        }
        else {
          resistant = myMarkovMat["1","0"] 
          resistantUpper = myMarkovUpper["1","0"] 
          resistantLower = myMarkovLower["1","0"] 
          resistantSE = myMarkovSE["1","0"]
          result = c(resistant, resistantUpper, resistantLower, resistantSE)
          
          #preserved = myMarkovMat["0","0"] 
          #relativeRisk = resistant/ preserved
          #return(relativeRisk)
          return(result)
          
        }
        
      }, simplify = F))
      rownames(scoresByMsp) = rownames(mgsMat)
      colnames(scoresByMsp) = c("outflow","upper","lower","SE")
      return(scoresByMsp)
      
    }
    
    inflowAmericanMat = getInflow(americanMetaTab, americanMgsMatImputed)
    outflowAmericanMat = getOutflow(americanMetaTab, americanMgsMatImputed)
    
    americanInflow = inflowAmericanMat[,1]
    americanOutflow = outflowAmericanMat[,1]
    americanPrevalence = apply(americanMgsMatImputed, 1, function(x) sum(x>0)/length(x))
    
    americanInOutFlowTab = data.frame(inflow=americanInflow,
                                      outflow=americanOutflow,
                                      prevalence=americanPrevalence,
                                      species = getSpeciesName(names(americanInflow),gssOssTabNew) ) 
    
    
    ggplot(americanInOutFlowTab, aes(x=outflow, y=inflow, fill=inflow)) + 
      scale_fill_gradient2(midpoint = 0.2, low="#FFD70044", mid = "white", high = "purple") +
      geom_point(size=3, shape=21, colour="#d3d3d366")+ xlab("") + ylab("")+
      geom_vline(xintercept = 0.3, colour="gray",linetype="dashed")+
      geom_hline(yintercept = 0.3, colour="gray",linetype="dashed")+
      #geom_abline(slope=1, intercept = 0)+
      theme(panel.grid.major = element_blank(), 
            panel.grid.minor = element_blank(),
            legend.position = "none",
            text = element_text(size = 15),
            panel.background = element_rect(fill = "white", 
                                            colour = "black", 
                                            size = 0.5, 
                                            linetype = "solid"))
    
    americanInflowSpecies = names(americanInflow[americanInflow>=0.3])
    americanOutflowSpecies = names(americanOutflow[americanOutflow>=0.3])
    
    americanSharedpecies = americanInflowSpecies[americanInflowSpecies  %in% americanOutflowSpecies]
    
    americanInflowSpecies = americanInflowSpecies[!americanInflowSpecies %in% americanSharedpecies]
    americanOutflowSpecies = americanOutflowSpecies[!americanOutflowSpecies %in% americanSharedpecies]
    
    table(inflowSpecies %in% americanInflowSpecies)
    table(inflowSpecies %in% americanOutflowSpecies)
    
    table(outflowSpecies %in% americanInflowSpecies)
    table(outflowSpecies %in% americanOutflowSpecies)
    
    sequentialMode = T
    if (sequentialMode) {
      getSuccessionDisappearingCounts <- function(metaTab, mgsMat){
        uniqSubj = unique(metaTab$metadata.ID)
        countsByMsp = do.call(rbind,sapply(rownames(mgsMat), function(currMgs){
          
          currCountMat = do.call(rbind,sapply(uniqSubj, function(subj) {
            currSamples = metaTab$sample.ID[metaTab$metadata.ID==subj]
            currVisits = metaTab$subtype[match(currSamples, metaTab$sample.ID)]
            currSamples = currSamples[order(currVisits)]
            currMgsMat = mgsMat[,currSamples]
            currMgsSubjVec = (currMgsMat[currMgs,] >0)*1
            #currMgsSubjVec = as.character(currMgsSubjVec)
            currEnd=length(currMgsSubjVec) - 1
            currSeq = 1:currEnd
            
            count_t = 0
            count_t_plus_1 = 0
            for(ind in currSeq) {
              if (currMgsSubjVec[ind] == 1) {
                count_t = count_t + currMgsSubjVec[ind]
                count_t_plus_1 = count_t_plus_1 + currMgsSubjVec[ind + 1]
              }
            }
            return(c(count_t,count_t_plus_1))
            
          }, simplify = F))
          currCountSum = colSums(currCountMat)
          
          
          return(currCountSum)
          
        }, simplify = F))
        return(countsByMsp)
        
      }
      getSuccessionAppearingCounts <- function(metaTab, mgsMat){
        uniqSubj = unique(metaTab$metadata.ID)
        countsByMsp = do.call(rbind,sapply(rownames(mgsMat), function(currMgs){
          
          currCountMat = do.call(rbind,sapply(uniqSubj, function(subj) {
            currSamples = metaTab$sample.ID[metaTab$metadata.ID==subj]
            currVisits = metaTab$subtype[match(currSamples, metaTab$sample.ID)]
            currSamples = currSamples[order(currVisits)]
            currMgsMat = mgsMat[,currSamples]
            currMgsSubjVec = (currMgsMat[currMgs,] >0)*1
            #currMgsSubjVec = as.character(currMgsSubjVec)
            currEnd=length(currMgsSubjVec) - 1
            currSeq = 1:currEnd
            
            count_t = 0
            count_t_plus_1 = 0
            for(ind in currSeq) {
              if (currMgsSubjVec[ind] == 0) {
                count_t = count_t + currMgsSubjVec[ind]
                count_t_plus_1 = count_t_plus_1 + currMgsSubjVec[ind + 1]
              }
            }
            return(c(count_t,count_t_plus_1))
            
          }, simplify = F))
          currCountSum = colSums(currCountMat)
          
          
          return(currCountSum)
          
        }, simplify = F))
        return(countsByMsp)
        
      }
      
      getSuccessionDisappearingAbundances <- function(metaTab, mgsMat){
        uniqSubj = unique(metaTab$metadata.ID)
        countsByMsp = do.call(rbind,sapply(rownames(mgsMat), function(currMgs){
          
          currCountMat = do.call(rbind,sapply(uniqSubj, function(subj) {
            currSamples = metaTab$sample.ID[metaTab$metadata.ID==subj]
            currVisits = metaTab$subtype[match(currSamples, metaTab$sample.ID)]
            currSamples = currSamples[order(currVisits)]
            currMgsMat = mgsMat[,currSamples]
            currMgsSubjVec = currMgsMat[currMgs,]
            #currMgsSubjVec = as.character(currMgsSubjVec)
            currEnd=length(currMgsSubjVec) - 1
            currSeq = 1:currEnd
            
            count_t = 0
            count_t_plus_1 = 0
            for(ind in currSeq) {
              if (currMgsSubjVec[ind] > 0) {
                count_t = count_t + currMgsSubjVec[ind]
                count_t_plus_1 = count_t_plus_1 + currMgsSubjVec[ind + 1]
              }
            }
            return(c(count_t,count_t_plus_1))
            
          }, simplify = F))
          currCountSum = colMeans(currCountMat)
          
          
          return(currCountSum)
          
        }, simplify = F))
        return(countsByMsp)
        
      }
      getSuccessionAppearingAbundances <- function(metaTab, mgsMat){
        uniqSubj = unique(metaTab$metadata.ID)
        countsByMsp = do.call(rbind,sapply(rownames(mgsMat), function(currMgs){
          
          currCountMat = do.call(rbind,sapply(uniqSubj, function(subj) {
            currSamples = metaTab$sample.ID[metaTab$metadata.ID==subj]
            currVisits = metaTab$subtype[match(currSamples, metaTab$sample.ID)]
            currSamples = currSamples[order(currVisits)]
            currMgsMat = mgsMat[,currSamples]
            currMgsSubjVec = currMgsMat[currMgs,]  
            #currMgsSubjVec = as.character(currMgsSubjVec)
            currEnd=length(currMgsSubjVec) - 1
            currSeq = 1:currEnd
            
            count_t = 0
            count_t_plus_1 = 0
            for(ind in currSeq) {
              if (currMgsSubjVec[ind] == 0) {
                count_t = count_t + currMgsSubjVec[ind]
                count_t_plus_1 = count_t_plus_1 + currMgsSubjVec[ind + 1]
              }
            }
            return(c(count_t,count_t_plus_1))
            
          }, simplify = F))
          currCountSum = colMeans(currCountMat)
          
          
          return(currCountSum)
          
        }, simplify = F))
        return(countsByMsp)
        
      }
      
      americanDisappearingCounts = getSuccessionDisappearingCounts(americanMetaTab, americanMgsMatImputed)
      americanAppearingCounts = getSuccessionAppearingCounts(americanMetaTab, americanMgsMatImputed)
      
      americanDisappearingAbundances = getSuccessionDisappearingAbundances(americanMetaTab, americanMgsMatImputed)
      americanAppearingAbundances = getSuccessionAppearingAbundances(americanMetaTab, americanMgsMatImputed)
      
      
      wilcox.test(americanDisappearingAbundances[inflowSpecies,1], americanDisappearingAbundances[inflowSpecies,2], paired = T, alternative = "greater")
      wilcox.test(americanDisappearingAbundances[rownames(americanDisappearingAbundances) %in% outflowSpecies,1], americanDisappearingAbundances[rownames(americanDisappearingAbundances) %in% outflowSpecies,2], paired = T, alternative = "greater")
      
      wilcox.test(americanAppearingAbundances[inflowSpecies,1], americanAppearingAbundances[inflowSpecies,2], paired = T, alternative = "less")
      wilcox.test(americanAppearingAbundances[rownames(americanAppearingAbundances) %in% outflowSpecies,1], americanAppearingAbundances[rownames(americanAppearingAbundances) %in% outflowSpecies,2], paired = T, alternative = "less")
      
      
      plot(americanDisappearingAbundances[inflowSpecies,1], americanDisappearingAbundances[inflowSpecies,2],pch=16,col="#80808044")
      par(new=T)
      plot(americanDisappearingAbundances[rownames(americanDisappearingAbundances) %in% outflowSpecies,1],
           americanDisappearingAbundances[rownames(americanDisappearingAbundances) %in% outflowSpecies,2],col="#FF000066",pch=16)
      
      plot(americanAppearingAbundances[inflowSpecies,1], americanAppearingAbundances[inflowSpecies,2],pch=16,col="#80808044")
      par(new=T)
      plot(americanAppearingAbundances[rownames(americanAppearingAbundances) %in% outflowSpecies,1],
           americanAppearingAbundances[rownames(americanAppearingAbundances) %in% outflowSpecies,2],col="#FF000066",pch=16)
      
      
      plot(americanDisappearingCounts[inflowSpecies,1], americanDisappearingCounts[inflowSpecies,2],pch=16,col="#80808044",xlim=c(0,154),ylim=c(0,154))
      abline(lm(americanDisappearingCounts[inflowSpecies,2] ~ americanDisappearingCounts[inflowSpecies,1]))
      par(new=T)
      plot(americanDisappearingCounts[rownames(americanDisappearingCounts) %in% outflowSpecies,1],
           americanDisappearingCounts[rownames(americanDisappearingCounts) %in% outflowSpecies,2],col="#FF000066",pch=16,xlim=c(0,154),ylim=c(0,154))
      abline(lm(americanDisappearingCounts[rownames(americanDisappearingCounts) %in% outflowSpecies,2] ~ americanDisappearingCounts[rownames(americanDisappearingCounts) %in% outflowSpecies,1]))
      
      plot(americanAppearingCounts[inflowSpecies,1], americanAppearingCounts[inflowSpecies,2],pch=16,col="#80808044")
      par(new=T)
      plot(americanAppearingCounts[rownames(americanAppearingCounts) %in% outflowSpecies,1],
           americanAppearingCounts[rownames(americanAppearingCounts) %in% outflowSpecies,2],col="#FF000066",pch=16)
      
      
      wilcox.test(americanDisappearingCounts[inflowSpecies,1], americanDisappearingCounts[inflowSpecies,2], paired = T)
      wilcox.test(americanDisappearingCounts[rownames(americanDisappearingCounts) %in% outflowSpecies,1], americanDisappearingCounts[rownames(americanDisappearingCounts) %in% outflowSpecies,2], paired = T)
      
      wilcox.test(americanAppearingCounts[inflowSpecies,1], americanAppearingCounts[inflowSpecies,2], paired = T)
      wilcox.test(americanAppearingCounts[rownames(americanAppearingCounts) %in% outflowSpecies,1], americanAppearingCounts[rownames(americanAppearingCounts) %in% outflowSpecies,2], paired = T)
      
    }
    
    checkWellnessCountMode = T
    if (checkWellnessCountMode) {
      
      newInflowSpecies = inflowSpecies[order(inflow[inflowSpecies],decreasing = T)]
      newOutflowSpecies = outflowSpecies[order(outflow[outflowSpecies],decreasing = T)]
      newOutflowSpecies = newOutflowSpecies[newOutflowSpecies %in% rownames(americanAppearingCounts)]
      
      
      abundanceMat = cbind((americanAppearingAbundances[inflowSpecies,]), (americanDisappearingAbundances[inflowSpecies,]))
      
      plotBarcode((americanAppearingAbundances[newInflowSpecies,]))
      plotBarcode((americanDisappearingAbundances[newInflowSpecies,]))
      
      plotBarcode((americanAppearingAbundances[ newOutflowSpecies,]))
      plotBarcode((americanDisappearingAbundances[newOutflowSpecies,]))
      
      levelplot(t(americanAppearingCounts[newInflowSpecies,]), 
                col.regions= plasma(500),
                aspect = 2,
                xlab="",ylab="",
                at=seq(min(americanAppearingCounts), max(americanAppearingCounts), length.out=500),
                scales = list(draw=F))
      
      levelplot(t(americanDisappearingCounts[newInflowSpecies,]), 
                col.regions= plasma(500),
                aspect = 2,
                xlab="",ylab="",
                at=seq(min(americanDisappearingCounts), max(americanDisappearingCounts), length.out=500),
                scales = list(draw=F))
      
      levelplot(t(americanAppearingCounts[ newOutflowSpecies,]), 
                col.regions= plasma(500),
                aspect = 2,
                xlab="",ylab="",
                at=seq(min(americanAppearingCounts), max(americanAppearingCounts), length.out=500),
                scales = list(draw=F))
      
      levelplot(t(americanDisappearingCounts[  newOutflowSpecies,]), 
                col.regions= plasma(500),
                aspect = 2,
                xlab="",ylab="",
                at=seq(min(americanDisappearingCounts), max(americanDisappearingCounts), length.out=500),
                scales = list(draw=F))
      
      
    }
    
    wellnessMode = T
    if (wellnessMode) {
      overlapAmericanSpecies = names(americanOutflow)[names(americanOutflow) %in% names(outflow)]
      
      
      plot(americanInflow[overlapAmericanSpecies], inflow[overlapAmericanSpecies])
      abline(a=0,b=1,col="red")
      
      plot(americanOutflow[overlapAmericanSpecies], outflow[overlapAmericanSpecies])
      abline(a=0,b=1,col="red")
      
      cor.test(americanInflow[overlapAmericanSpecies], inflow[overlapAmericanSpecies], method="spearman")
      cor.test(americanOutflow[overlapAmericanSpecies], outflow[overlapAmericanSpecies], method="spearman")
      
      
      wellnessAmericanTab = data.frame(inflow.american = americanInflow[overlapAmericanSpecies],
                                       inflow.wellness = inflow[overlapAmericanSpecies],
                                       outflow.american = americanOutflow[overlapAmericanSpecies],
                                       outflow.wellness = outflow[overlapAmericanSpecies],
                                       prevalence.american = americanPrevalence[overlapAmericanSpecies],
                                       prevalence.wellness = freqs_wellness[overlapAmericanSpecies],
                                       species = getSpeciesName(overlapAmericanSpecies, gssOssTabNew))
      wellnessAmericanTab = wellnessAmericanTab[wellnessAmericanTab$prevalence.american >= 0.1 & wellnessAmericanTab$prevalence.wellness >= 0.1, ]
      
      cor.test(wellnessAmericanTab$inflow.american, wellnessAmericanTab$inflow.wellness, method="spearman")
      cor.test(wellnessAmericanTab$outflow.american, wellnessAmericanTab$outflow.wellness, method="spearman")
      
      
      ggplot(wellnessAmericanTab, aes(x=inflow.american, y=inflow.wellness)) + 
        geom_point(size=3, shape=16, colour="#00000055") + xlab("") + ylab("")+ #size=3, 
        #scale_fill_gradient2(midpoint = 0.2, low="#FFD70044", mid = "white", high = "purple") +
        geom_smooth(method = "lm",  se = T, colour="red")+
        scale_x_continuous(breaks = scales::pretty_breaks(n = 2)) +
        scale_y_continuous(breaks = scales::pretty_breaks(n = 2)) +
        
        #geom_hline(yintercept=6)+
        #geom_abline(slope=1, intercept = 0)+
        theme(panel.grid.major = element_blank(), 
              panel.grid.minor = element_blank(),
              legend.position = "none",
              text = element_text(size = 15),
              panel.background = element_rect(fill = "white", 
                                              colour = "black", 
                                              size = 0.5, 
                                              linetype = "solid"))
      
      ggplot(wellnessAmericanTab, aes(x=outflow.american, y=outflow.wellness)) + 
        geom_point(size=3, shape=16, colour="#00000055") + xlab("") + ylab("")+ #size=3, 
        #scale_fill_gradient2(midpoint = 0.2, low="#FFD70044", mid = "white", high = "purple") +
        geom_smooth(method = "lm",  se = T, colour="red")+
        scale_x_continuous(breaks = scales::pretty_breaks(n = 2)) +
        scale_y_continuous(breaks = scales::pretty_breaks(n = 2)) +
        
        #geom_hline(yintercept=6)+
        #geom_abline(slope=1, intercept = 0)+
        theme(panel.grid.major = element_blank(), 
              panel.grid.minor = element_blank(),
              legend.position = "none",
              text = element_text(size = 15),
              panel.background = element_rect(fill = "white", 
                                              colour = "black", 
                                              size = 0.5, 
                                              linetype = "solid"))
      diffMode = T
      if (diffMode) {
        
        inflowDiffOuts = do.call(rbind,sapply(overlapAmericanSpecies, function(currMsp){
          medInflow = inflowAmericanMat[currMsp,1]
          medInflowSe = inflowAmericanMat[currMsp,4]
          medZscore = medInflow / medInflowSe
          
          
          conInflow = inflowOut[currMsp,1]
          conInflowSe = inflowOut[currMsp,4]
          conZscore = conInflow / conInflowSe
          
          diffZscore = medZscore - conZscore
          pvalue = 2*pnorm(abs(diffZscore), lower.tail = F)
          
          if (medInflow == 0) {
            diffZscore = - conZscore 
            pvalue = 2*pnorm(abs(diffZscore), lower.tail = F)
            
          }
          if (conInflow == 0) {
            diffZscore = medZscore 
            pvalue = 2*pnorm(abs(diffZscore), lower.tail = F)
          }
          
          return(c(diffZscore, medInflow, conInflow, pvalue))
        },simplify=F))
        colnames(inflowDiffOuts)  = c("diff.z","American", "Wellness", "pvalue")
        inflowDiffOuts = data.frame(inflowDiffOuts)
        inflowDiffOuts$species = getSpeciesName(rownames(inflowDiffOuts),gssOssTabNew)
        inflowDiffOuts = inflowDiffOuts[!is.nan(inflowDiffOuts$pvalue) & inflowDiffOuts$pvalue < 0.01, ]
        
        inflowUpSpecies = rownames(inflowDiffOuts)[inflowDiffOuts$diff.z>0]
        inflowDnSpecies = rownames(inflowDiffOuts)[inflowDiffOuts$diff.z<0]
        inflowBothSpecies = rownames(inflowDiffOuts)
        
        
        write.table(inflowDiffOuts, "inflowDiffOuts.american.wellness.txt", sep="\t")
        
        outflowDiffOuts = do.call(rbind,sapply(overlapAmericanSpecies, function(currMsp){
          medOutflow = outflowAmericanMat[currMsp,1]
          medOutflowSe = outflowAmericanMat[currMsp,4]
          medZscore = medOutflow / medOutflowSe
          
          
          conOutflow = outflowOut[currMsp,1]
          conOutflowSe = outflowOut[currMsp,4]
          conZscore = conOutflow / conOutflowSe
          
          diffZscore = medZscore - conZscore
          pvalue = 2*pnorm(abs(diffZscore), lower.tail = F)
          
          if (medOutflow == 0) {
            diffZscore = - conZscore 
            pvalue = 2*pnorm(abs(diffZscore), lower.tail = F)
            
          }
          if (conOutflow == 0) {
            diffZscore = medZscore 
            pvalue = 2*pnorm(abs(diffZscore), lower.tail = F)
            
          }
          
          return(c(diffZscore, medOutflow, conOutflow, pvalue))
        },simplify=F))
        colnames(outflowDiffOuts)  = c("diff.z","American", "Wellness", "pvalue")
        outflowDiffOuts = data.frame(outflowDiffOuts)
        outflowDiffOuts$species = getSpeciesName(rownames(outflowDiffOuts),gssOssTabNew)
        #View(outflowDiffOuts)
        outflowDiffOuts = outflowDiffOuts[!is.nan(outflowDiffOuts$pvalue) & outflowDiffOuts$pvalue < 0.01, ]
        outflowUpSpecies = rownames(outflowDiffOuts)[outflowDiffOuts$diff.z>0]
        outflowDnSpecies = rownames(outflowDiffOuts)[outflowDiffOuts$diff.z<0]
        outflowBothSpecies = rownames(outflowDiffOuts)
        
        
        
        write.table(outflowDiffOuts, "outflowDiffOuts.american.wellness.txt", sep="\t")
        
        
        
      }
    }
    
    
  }
  
  newMarkovForDINAMIC = T
  if (newMarkovForDINAMIC) {
    
    pathogenListFromMathieu = "j://Deposit//Project//2018_microbiome_atlas//pathogen.list.Ecker.et.al.from.Mathieu.txt"
    pathogenList = read.table(pathogenListFromMathieu, header=F, stringsAsFactors = F)[,]
    
    
    checkGammaFitMode = F
    if (checkGammaFitMode) {
      loadAvgMode = T
      if (loadAvgMode) {
        exSubj = c("P31","P19")
        
        dinamicMgsFile = "J:\\Deposit\\Project\\2018_microbiome_atlas\\in.outflow.dataset\\dinamic_vs_hs_10_4_igc2_smart_shared_reads_29042020.down.10M.fpkm.msp.txt"
        dinamicMetaFile = "J:\\Deposit\\Project\\2018_microbiome_atlas\\in.outflow.dataset\\dinamic.metadata.txt"
        dinamicMgsMat = read.delim(dinamicMgsFile, sep="\t", stringsAsFactors = F,row.names = 1)
        dinamicMgsMat = as.matrix(dinamicMgsMat) # 1990 x 244
        dinamicMgsMat = dinamicMgsMat[!rownames(dinamicMgsMat) %in% suppressMsps2,] # 1989 x 244
        
        dinamicMetaTab = read.delim(dinamicMetaFile, sep="\t", stringsAsFactors = F) # 246 x 6
        dinamicMetaTab = dinamicMetaTab[dinamicMetaTab$ID_MGP %in% colnames(dinamicMgsMat),]
        dinamicMetaTab$Time_point[dinamicMetaTab$Time_point == "baseline"] = "v1"
        dinamicMetaTab$Time_point[dinamicMetaTab$Time_point == "4 weeks"] = "v2"
        dinamicMetaTab$Time_point[dinamicMetaTab$Time_point == "8 weeks"] = "v3"
        colnames(dinamicMetaTab) = c("sample.ID","metadata.ID","type","subtype","Gender","Age")
        
        
        dinamicMgsMat = dinamicMgsMat[rowSums(dinamicMgsMat)!=0,]
        dinamicMgsMat = dinamicMgsMat[!apply(dinamicMgsMat,1,all),] # 1268 x 244 
        
        dinamicMetaTab = dinamicMetaTab[!dinamicMetaTab$metadata.ID %in% exSubj,]
        dinamicMgsMat = dinamicMgsMat[,match(dinamicMetaTab$sample.ID,colnames(dinamicMgsMat))]
        
      }
      
      loadMedianMode = T
      if (loadMedianMode) {
        exSubj = c("P31","P19")
        
        
        dinamicMgsMedFile = "J:\\Deposit\\Project\\2018_microbiome_atlas\\in.outflow.dataset\\dinamic_vs_hs_10_4_igc2_smart_shared_reads_29042020_bis.down.10000000.fpkm.mgs.profiles.100.tsv"
        dinamicMetaFile = "J:\\Deposit\\Project\\2018_microbiome_atlas\\in.outflow.dataset\\dinamic.metadata.txt"
        
        dinamicMgsMedMat = read.delim(dinamicMgsMedFile, sep="\t", stringsAsFactors = F,row.names = 1)
        dinamicMgsMedMat = as.matrix(dinamicMgsMedMat)
        
        extractMedian <- function(dinamicMgsMedMat) {
          
          rowNames = rownames(dinamicMgsMedMat)
          mspNames = unlist(sapply(rowNames, function(x) lapply(strsplit(x, split="\\."), "[", 1)))
          #geneNames = unlist(sapply(rowNames, function(x) lapply(strsplit(x, split="\\."), "[", 2)))
          
          uniqueMsps = unique(mspNames)
          
          dinamicMgsMatNew = do.call(rbind,  sapply(uniqueMsps, function(currMsp) {
            currInds = which(mspNames == currMsp)
            #print(currInds)
            return(colMedians(dinamicMgsMedMat[currInds,]))
          }, simplify = F))
          colnames(dinamicMgsMatNew) = colnames(dinamicMgsMedMat)
          return(dinamicMgsMatNew)
        }
        
        dinamicMgsMatNew = extractMedian(dinamicMgsMedMat) # 1990 x 244
        #dim(dinamicMgsMatNew)
        #dinamicMgsMatNew = as.matrix(dinamicMgsMatNew) 
        
        
        dinamicMgsMat = dinamicMgsMatNew[!rownames(dinamicMgsMatNew) %in% suppressMsps2,] # 1989 x 244
        
        dinamicMetaTab = read.delim(dinamicMetaFile, sep="\t", stringsAsFactors = F) # 246 x 6
        dinamicMetaTab = dinamicMetaTab[dinamicMetaTab$ID_MGP %in% colnames(dinamicMgsMat),]
        dinamicMetaTab$Time_point[dinamicMetaTab$Time_point == "baseline"] = "v1"
        dinamicMetaTab$Time_point[dinamicMetaTab$Time_point == "4 weeks"] = "v2"
        dinamicMetaTab$Time_point[dinamicMetaTab$Time_point == "8 weeks"] = "v3"
        colnames(dinamicMetaTab) = c("sample.ID","metadata.ID","type","subtype","Gender","Age")
        
      }
      
      dinamicMgsMatScaled = dinamicMgsMat*10^9
      dinamicMgsMatScaledLog = log2(dinamicMgsMatScaled)
      dinamicMgsMatScaledLog[dinamicMgsMatScaledLog == -Inf] = 0
      
      dinamicMgsMatScaledNonZero = dinamicMgsMatScaled[dinamicMgsMatScaled!=0]
      dinamicMgsMatScaledNonZero = log2(dinamicMgsMatScaledNonZero)
      
      dinamic.fit.gamma = fitdist(dinamicMgsMatScaledNonZero, "gamma")
      plot(dinamic.fit.gamma)
      dinamic.fit.gamma$estimate
      
      
      if (F) {
        qCutOff =(2^(4.042488))/10^9
        min(mergeMatUpdated[mergeMatUpdated!=0])
        
        mergeMatUpdatedImputed = mergeMatUpdated
        mergeMatUpdatedImputed[mergeMatUpdatedImputed<qCutOff] <- 0
        
      }
      
    }
    
    medianSignalMode = T
    if (medianSignalMode) {
      
      loadMode = T
      if (loadMode) {
        exSubj = c("P31","P19")
        
        dinamicMgsMedFile = "J:\\Deposit\\Project\\2018_microbiome_atlas\\in.outflow.dataset\\dinamic_vs_hs_10_4_igc2_smart_shared_reads_29042020_bis.down.10000000.fpkm.mgs.profiles.100.tsv"
        dinamicMetaFile = "J:\\Deposit\\Project\\2018_microbiome_atlas\\in.outflow.dataset\\dinamic.metadata.txt"
        
        dinamicMgsMedMat = read.delim(dinamicMgsMedFile, sep="\t", stringsAsFactors = F,row.names = 1)
        dinamicMgsMedMat = as.matrix(dinamicMgsMedMat)
        
        extractMedian <- function(dinamicMgsMedMat) {
          
          rowNames = rownames(dinamicMgsMedMat)
          mspNames = unlist(sapply(rowNames, function(x) lapply(strsplit(x, split="\\."), "[", 1)))
          #geneNames = unlist(sapply(rowNames, function(x) lapply(strsplit(x, split="\\."), "[", 2)))
          
          uniqueMsps = unique(mspNames)
          
          dinamicMgsMatNew = do.call(rbind,  sapply(uniqueMsps, function(currMsp) {
            currInds = which(mspNames == currMsp)
            #print(currInds)
            return(colMedians(dinamicMgsMedMat[currInds,]))
          }, simplify = F))
          colnames(dinamicMgsMatNew) = colnames(dinamicMgsMedMat)
          return(dinamicMgsMatNew)
        }
        
        dinamicMgsMatNew = extractMedian(dinamicMgsMedMat) # 1990 x 244
        #dim(dinamicMgsMatNew)
        #dinamicMgsMatNew = as.matrix(dinamicMgsMatNew) 
        
        
        dinamicMgsMat = dinamicMgsMatNew[!rownames(dinamicMgsMatNew) %in% suppressMsps2,] # 1989 x 244
        
        dinamicMetaTab = read.delim(dinamicMetaFile, sep="\t", stringsAsFactors = F) # 246 x 6
        dinamicMetaTab = dinamicMetaTab[dinamicMetaTab$ID_MGP %in% colnames(dinamicMgsMat),]
        dinamicMetaTab$Time_point[dinamicMetaTab$Time_point == "baseline"] = "v1"
        dinamicMetaTab$Time_point[dinamicMetaTab$Time_point == "4 weeks"] = "v2"
        dinamicMetaTab$Time_point[dinamicMetaTab$Time_point == "8 weeks"] = "v3"
        colnames(dinamicMetaTab) = c("sample.ID","metadata.ID","type","subtype","Gender","Age")
        
        qCutOff =(2^(4.042488))/10^9
        min(dinamicMgsMat[dinamicMgsMat!=0])
        
        dinamicMgsMatImputed = dinamicMgsMat
        dinamicMgsMatImputed[dinamicMgsMatImputed < qCutOff] <- 0
        dinamicMgsMatImputed = dinamicMgsMatImputed[rowSums(dinamicMgsMatImputed)!=0,]
        dinamicMgsMatImputed = dinamicMgsMatImputed[!apply(dinamicMgsMatImputed,1,all),] # 1077 x 114 
        
        dinamicMetaTab = dinamicMetaTab[!dinamicMetaTab$metadata.ID %in% exSubj,]
        dinamicMgsMatImputed = dinamicMgsMatImputed[,match(dinamicMetaTab$sample.ID,colnames(dinamicMgsMatImputed))]
      }
      
      getSurvMode = T
      if (getSurvMode) {
        getPairSeqs <- function(lenSeqs) {
          lenInds = lenSeqs-1
          pairList = sapply(1:lenInds, function(x) c(x, x+1), simplify = F)
          pairTab = do.call(rbind, pairList)
          return(pairTab)
        }
        getSurvVecByMgs <- function(mgsVec, mgsName, sampleNames, interval=3) {
          
          lenAll = length(mgsVec)
          lenInd = length(mgsVec)-1
          seqs = 1:lenInd
          #interval=3
          survTime = 0
          survEvent = "None"
          
          #####
          #check fist or second visit whether mgs exists
          
          mgsCutDigit <- mgsVec > 0
          presentInds = which(mgsCutDigit)
          if (length(presentInds)==0) {
            survTime = NA
            survEvent = "None"
            return(list(time=survTime, event=survEvent, sample=sampleNames[1], abd = 0, mgs=mgsName))
          }
          startInd = min(presentInds)
          lastInd = startInd
          for (ind in seq_along(presentInds)) {
            if (ind == length(presentInds)) break
            currInd = presentInds[ind]
            nextInd = presentInds[ind+1]
            diff = nextInd - currInd
            if (diff == 1) lastInd = nextInd
          }
          
          if (length(presentInds)!= 0) {
            survTime= (lastInd+1 - startInd)*interval
            survEvent = "Absent"
            if (lastInd == lenAll) {
              survTime = interval*lenInd
              survEvent = "None"
            }
          }
          res=list(time=survTime, event=survEvent, sample=sampleNames[startInd], abd = mgsVec[startInd], mgs=mgsName)
          #print(res)
          return(res)
          
        }
        getSurvivalRates <- function(metaTab, mgsMat, interval=3, halfTime=6) {
          
          uniqSubjs = unique(metaTab$metadata.ID)
          mgsPrevalence = apply(mgsMat, 1, function(currRow){
            mean(currRow>0)
          })
          mgsMeans = rowMeans(mgsMat)
          
          survList = do.call(c,sapply(uniqSubjs, function(subj) {
            currSamples = metaTab$sample.ID[metaTab$metadata.ID==subj]
            currSamples = currSamples[order(currSamples)]
            currMgsMat = mgsMat[,currSamples]
            currMgsMat = currMgsMat[rowMeans(currMgsMat)!=0,]
            currSubSurvVecThroughs = (sapply(rownames(currMgsMat), function(currMgs){
              currSubSurvVecThrough = getSurvVecByMgs(currMgsMat[currMgs,], currMgs, currSamples, interval)
              return(currSubSurvVecThrough)
            }, simplify = F))
            
            return(currSubSurvVecThroughs) 
            
          }, simplify = F))
          
          survTab = data.frame(time=do.call(c,lapply(survList, function(x) x$time)),
                               status=do.call(c,lapply(survList, function(x) x$event)),
                               sample=do.call(c,lapply(survList, function(x) x$sample)),
                               abd=do.call(c,lapply(survList, function(x) x$abd)),
                               mgs=do.call(c,lapply(survList, function(x) x$mgs)),
                               stringsAsFactors = F)
          survTab$logAbd = log10(survTab$abd)
          survivalRates = unlist(sapply(unique(survTab$mgs), function(currMsp) {
            currMspTab = survTab[survTab$mgs == currMsp,]
            if (F) {
              library("survminer")
              currMsp = "msp_0069"
              currMsp = "msp_0041"
              currMsp = "msp_0005"
              currMsp = "msp_0148c"
              currMsp = "msp_0742"
              
              
              currMspTab = survTab[survTab$mgs == currMsp,]
              currMspTab$status <- currMspTab$status == "Absent"
              currSurvFit=survfit(survival::Surv(time, status) ~ 1, data=currMspTab)
              ggsurvplot(currSurvFit, break.time.by = 3, color = "#2E9FDF")
              
            }
            halfTimeSummary = summary(survfit(survival::Surv(currMspTab$time, currMspTab$status == "Absent") ~ 1), times=halfTime)
            halfTimeSurvival = halfTimeSummary$surv
            return(halfTimeSurvival)
          }))
          
          survivalStats = data.frame(surv.rates=survivalRates[match(rownames(mgsMat), names(survivalRates))],
                                     mgs.abd = mgsMeans[match(rownames(mgsMat), names(mgsMeans))],
                                     freq= mgsPrevalence,
                                     mgs=rownames(mgsMat),
                                     species=getSpeciesName(rownames(mgsMat),taxo))
          
          return(survivalStats)
        }
        
        #survivalStats = getSurvivalRates(metaTab, mgsMat, decreaseCut)
        survivalStats_DINAMIC_1mon = getSurvivalRates(dinamicMetaTab, dinamicMgsMatImputed, interval=1, halfTime = 1)
        
      }
      
      
      getInflow <- function(metaTab, mgsMat) {
        uniqSubj = unique(metaTab$metadata.ID)
        scoresByMsp = do.call(rbind, sapply(rownames(mgsMat), function(currMgs){
          
          markovSeqOfMsp = (sapply(uniqSubj, function(subj) {
            currSamples = metaTab$sample.ID[metaTab$metadata.ID==subj]
            currVisits = metaTab$subtype[match(currSamples, metaTab$sample.ID)]
            currSamples = currSamples[order(currVisits)]
            currMgsMat = mgsMat[,currSamples]
            currMgsSubjVec = (currMgsMat[currMgs,] >0)*1
            currMgsSubjVec = as.character(currMgsSubjVec)
            
          }, simplify = F))
          
          myMarkovFit<-markovchainFit(data=markovSeqOfMsp,confidencelevel = .9,method = "mle")
          myMarkovMat = myMarkovFit$estimate@transitionMatrix
          myMarkovUpper = myMarkovFit$upperEndpointMatrix
          myMarkovLower = myMarkovFit$lowerEndpointMatrix
          myMarkovSE = myMarkovFit$standardError
          
          if(dim(myMarkovMat)[1]==1) {
            rname = rownames(myMarkovMat)
            #if (rname == "0") return(0)
            #if (rname == "1") return(Inf)
            if (rname == "0") {
              colonized = 0
              colonizedUpper = NA
              colonizedLower = NA
              colonizedSE = NA
              result = c(colonized, colonizedUpper, colonizedLower, colonizedSE)
              return(result)
            }
            
            if (rname == "1") {
              colonized = 1
              colonizedUpper = NA
              colonizedLower = NA
              colonizedSE = NA
              result = c(colonized, colonizedUpper, colonizedLower, colonizedSE)
              
              return(result)
            }
          }
          else {
            colonized = myMarkovMat["0","1"] 
            colonizedUpper = myMarkovUpper["0","1"]
            colonizedLower = myMarkovLower["0","1"]
            colonizedSE = myMarkovSE["0","1"]
            result = c(colonized, colonizedUpper, colonizedLower, colonizedSE)
            
            #preserved = myMarkovMat["1","1"] 
            #relativeRisk = colonized/ preserved
            #return(relativeRisk)
            return(result)
          }
          
        }, simplify = F))
        
        rownames(scoresByMsp) = rownames(mgsMat)
        colnames(scoresByMsp) = c("inflow","upper","lower","SE")
        return(scoresByMsp)
        
      }
      getOutflow <- function(metaTab, mgsMat) {
        
        uniqSubj = unique(metaTab$metadata.ID)
        scoresByMsp = do.call(rbind, sapply(rownames(mgsMat), function(currMgs){
          
          markovSeqOfMsp = (sapply(uniqSubj, function(subj) {
            currSamples = metaTab$sample.ID[metaTab$metadata.ID==subj]
            currVisits = metaTab$subtype[match(currSamples, metaTab$sample.ID)]
            currSamples = currSamples[order(currVisits)]
            currMgsMat = mgsMat[,currSamples]
            currMgsSubjVec = (currMgsMat[currMgs,] >0)*1
            currMgsSubjVec = as.character(currMgsSubjVec)
            
          }, simplify = F))
          
          myMarkovFit<-markovchainFit(data=markovSeqOfMsp,confidencelevel = .9,method = "mle")
          myMarkovMat = myMarkovFit$estimate@transitionMatrix
          myMarkovUpper = myMarkovFit$upperEndpointMatrix
          myMarkovLower = myMarkovFit$lowerEndpointMatrix
          myMarkovSE = myMarkovFit$standardError
          
          if(dim(myMarkovMat)[1]==1) {
            rname = rownames(myMarkovMat)
            #if (rname == "0") return(Inf)
            #if (rname == "1") return(0)
            if (rname == "0") {
              resistant = 1
              resistantUpper = NA
              resistantLower = NA
              resistantSE = NA
              result = c(resistant, resistantUpper, resistantLower, resistantSE)
              return(result) 
            }
            if (rname == "1") {
              resistant = 0
              resistantUpper = NA
              resistantLower = NA
              resistantSE = NA
              result = c(resistant, resistantUpper, resistantLower, resistantSE)
              return(result)
            }
          }
          else {
            resistant = myMarkovMat["1","0"] 
            resistantUpper = myMarkovUpper["1","0"] 
            resistantLower = myMarkovLower["1","0"] 
            resistantSE = myMarkovSE["1","0"]
            result = c(resistant, resistantUpper, resistantLower, resistantSE)
            
            #preserved = myMarkovMat["0","0"] 
            #relativeRisk = resistant/ preserved
            #return(relativeRisk)
            return(result)
            
          }
          
        }, simplify = F))
        rownames(scoresByMsp) = rownames(mgsMat)
        colnames(scoresByMsp) = c("outflow","upper","lower","SE")
        return(scoresByMsp)
        
      }
      
      sequentialMode = F
      if (sequentialMode) {
        getSuccessionDisappearingCounts <- function(metaTab, mgsMat){
          uniqSubj = unique(metaTab$metadata.ID)
          countsByMsp = do.call(rbind,sapply(rownames(mgsMat), function(currMgs){
            
            currCountMat = do.call(rbind,sapply(uniqSubj, function(subj) {
              currSamples = metaTab$sample.ID[metaTab$metadata.ID==subj]
              currVisits = metaTab$subtype[match(currSamples, metaTab$sample.ID)]
              currSamples = currSamples[order(currVisits)]
              currMgsMat = mgsMat[,currSamples]
              currMgsSubjVec = (currMgsMat[currMgs,] >0)*1
              #currMgsSubjVec = as.character(currMgsSubjVec)
              currEnd=length(currMgsSubjVec) - 1
              currSeq = 1:currEnd
              
              count_t = 0
              count_t_plus_1 = 0
              for(ind in currSeq) {
                if (currMgsSubjVec[ind] == 1) {
                  count_t = count_t + currMgsSubjVec[ind]
                  count_t_plus_1 = count_t_plus_1 + currMgsSubjVec[ind + 1]
                }
              }
              return(c(count_t,count_t_plus_1))
              
            }, simplify = F))
            currCountSum = colSums(currCountMat)
            
            
            return(currCountSum)
            
          }, simplify = F))
          return(countsByMsp)
          
        }
        getSuccessionAppearingCounts <- function(metaTab, mgsMat){
          uniqSubj = unique(metaTab$metadata.ID)
          countsByMsp = do.call(rbind,sapply(rownames(mgsMat), function(currMgs){
            
            currCountMat = do.call(rbind,sapply(uniqSubj, function(subj) {
              currSamples = metaTab$sample.ID[metaTab$metadata.ID==subj]
              currVisits = metaTab$subtype[match(currSamples, metaTab$sample.ID)]
              currSamples = currSamples[order(currVisits)]
              currMgsMat = mgsMat[,currSamples]
              currMgsSubjVec = (currMgsMat[currMgs,] >0)*1
              #currMgsSubjVec = as.character(currMgsSubjVec)
              currEnd=length(currMgsSubjVec) - 1
              currSeq = 1:currEnd
              
              count_t = 0
              count_t_plus_1 = 0
              for(ind in currSeq) {
                if (currMgsSubjVec[ind] == 0) {
                  count_t = count_t + currMgsSubjVec[ind]
                  count_t_plus_1 = count_t_plus_1 + currMgsSubjVec[ind + 1]
                }
              }
              return(c(count_t,count_t_plus_1))
              
            }, simplify = F))
            currCountSum = colSums(currCountMat)
            
            
            return(currCountSum)
            
          }, simplify = F))
          return(countsByMsp)
          
        }
        getSuccessionAllCounts <- function(metaTab, mgsMat){
          
          uniqSubj = unique(metaTab$metadata.ID)
          countsByMsp = do.call(rbind,sapply(rownames(mgsMat), function(currMgs){
            
            currCountMat = do.call(rbind,sapply(uniqSubj, function(subj) {
              currSamples = metaTab$sample.ID[metaTab$metadata.ID==subj]
              currVisits = metaTab$subtype[match(currSamples, metaTab$sample.ID)]
              currSamples = currSamples[order(currVisits)]
              currMgsMat = mgsMat[,currSamples]
              currMgsSubjVec = (currMgsMat[currMgs,] >0)*1
              #currMgsSubjVec = as.character(currMgsSubjVec)
              currEnd=length(currMgsSubjVec) - 1
              currSeq = 1:currEnd
              
              count_t = 0
              count_t_plus_1 = 0
              for(ind in currSeq) {
                count_t = count_t + currMgsSubjVec[ind]
                count_t_plus_1 = count_t_plus_1 + currMgsSubjVec[ind + 1] 
              }
              return(c(count_t,count_t_plus_1))
              
            }, simplify = F))
            currCountSum = colSums(currCountMat)
            
            
            return(currCountSum)
            
          }, simplify = F))
          return(countsByMsp)
          
        }
        
        getSuccessionDisappearingAbundances <- function(metaTab, mgsMat){
          uniqSubj = unique(metaTab$metadata.ID)
          countsByMsp = do.call(rbind,sapply(rownames(mgsMat), function(currMgs){
            
            currCountMat = do.call(rbind,sapply(uniqSubj, function(subj) {
              currSamples = metaTab$sample.ID[metaTab$metadata.ID==subj]
              currVisits = metaTab$subtype[match(currSamples, metaTab$sample.ID)]
              currSamples = currSamples[order(currVisits)]
              currMgsMat = mgsMat[,currSamples]
              currMgsSubjVec = currMgsMat[currMgs,]
              #currMgsSubjVec = as.character(currMgsSubjVec)
              currEnd=length(currMgsSubjVec) - 1
              currSeq = 1:currEnd
              
              count_t = 0
              count_t_plus_1 = 0
              for(ind in currSeq) {
                if (currMgsSubjVec[ind] > 0) {
                  count_t = count_t + currMgsSubjVec[ind]
                  count_t_plus_1 = count_t_plus_1 + currMgsSubjVec[ind + 1]
                }
              }
              return(c(count_t,count_t_plus_1))
              
            }, simplify = F))
            currCountSum = colMeans(currCountMat)
            
            
            return(currCountSum)
            
          }, simplify = F))
          return(countsByMsp)
          
        }
        getSuccessionAppearingAbundances <- function(metaTab, mgsMat){
          uniqSubj = unique(metaTab$metadata.ID)
          countsByMsp = do.call(rbind,sapply(rownames(mgsMat), function(currMgs){
            
            currCountMat = do.call(rbind,sapply(uniqSubj, function(subj) {
              currSamples = metaTab$sample.ID[metaTab$metadata.ID==subj]
              currVisits = metaTab$subtype[match(currSamples, metaTab$sample.ID)]
              currSamples = currSamples[order(currVisits)]
              currMgsMat = mgsMat[,currSamples]
              currMgsSubjVec = currMgsMat[currMgs,]  
              #currMgsSubjVec = as.character(currMgsSubjVec)
              currEnd=length(currMgsSubjVec) - 1
              currSeq = 1:currEnd
              
              count_t = 0
              count_t_plus_1 = 0
              for(ind in currSeq) {
                if (currMgsSubjVec[ind] == 0) {
                  count_t = count_t + currMgsSubjVec[ind]
                  count_t_plus_1 = count_t_plus_1 + currMgsSubjVec[ind + 1]
                }
              }
              return(c(count_t,count_t_plus_1))
              
            }, simplify = F))
            currCountSum = colMeans(currCountMat)
            
            
            return(currCountSum)
            
          }, simplify = F))
          return(countsByMsp)
          
        }
        getSuccessionAllAbundances <- function(metaTab, mgsMat){
          uniqSubj = unique(metaTab$metadata.ID)
          countsByMsp = do.call(rbind,sapply(rownames(mgsMat), function(currMgs){
            
            currCountMat = do.call(rbind,sapply(uniqSubj, function(subj) {
              currSamples = metaTab$sample.ID[metaTab$metadata.ID==subj]
              currVisits = metaTab$subtype[match(currSamples, metaTab$sample.ID)]
              currSamples = currSamples[order(currVisits)]
              currMgsMat = mgsMat[,currSamples]
              currMgsSubjVec = currMgsMat[currMgs,]  
              #currMgsSubjVec = as.character(currMgsSubjVec)
              currEnd=length(currMgsSubjVec) - 1
              currSeq = 1:currEnd
              
              count_t = 0
              count_t_plus_1 = 0
              for(ind in currSeq) {
                count_t = count_t + currMgsSubjVec[ind]
                count_t_plus_1 = count_t_plus_1 + currMgsSubjVec[ind + 1]
              }
              return(c(count_t,count_t_plus_1))
              
            }, simplify = F))
            currCountSum = colMeans(currCountMat)
            
            
            return(currCountSum)
            
          }, simplify = F))
          return(countsByMsp)
          
        }
        
        dinamicDisappearingCounts = getSuccessionDisappearingCounts(dinamicMetaTab, dinamicMgsMatImputed)
        dinamicAppearingCounts = getSuccessionAppearingCounts(dinamicMetaTab, dinamicMgsMatImputed)
        dinamicAllCounts = getSuccessionAllCounts(dinamicMetaTab, dinamicMgsMatImputed)
        
        dinamicDisappearingAbundances = getSuccessionDisappearingAbundances(dinamicMetaTab, dinamicMgsMatImputed)
        dinamicAppearingAbundances = getSuccessionAppearingAbundances(dinamicMetaTab, dinamicMgsMatImputed)
        dinamicAllAbundances = getSuccessionAllAbundances(dinamicMetaTab, dinamicMgsMatImputed)
        
        dinamicDisappearingCounts = disappearingCounts
        dinamicAppearingCounts = appearingCounts
        dinamicDisappearingAbundances = disappearingAbundances
        dinamicAppearingAbundances = appearingAbundances
        
        
        
        wilcox.test(dinamicAllAbundances[inflowSpecies,1], dinamicAllAbundances[inflowSpecies,2], paired = T, alternative = "less")
        wilcox.test(dinamicAllAbundances[rownames(dinamicAllAbundances) %in% outflowSpecies,1], dinamicAllAbundances[rownames(dinamicAllAbundances) %in% outflowSpecies,2], paired = T, , alternative = "greater")
        
        wilcox.test(dinamicAllAbundances[inflowSpecies,1], dinamicAllAbundances[inflowSpecies,2], paired = T, alternative = "greater")
        wilcox.test(dinamicAllAbundances[rownames(dinamicAllAbundances) %in% outflowSpecies,1], dinamicAllAbundances[rownames(dinamicAllAbundances) %in% outflowSpecies,2], paired = T, , alternative = "less")
        
        
        wilcox.test(dinamicAllCounts[inflowSpecies,1], dinamicAllCounts[inflowSpecies,2], paired = T, alternative = "less")
        wilcox.test(dinamicAllCounts[rownames(dinamicAllCounts) %in% outflowSpecies,1], dinamicAllCounts[rownames(dinamicAllCounts) %in% outflowSpecies,2], paired = T, alternative = "greater")
        
        
        
        wilcox.test(dinamicDisappearingAbundances[inflowSpecies,1], dinamicDisappearingAbundances[inflowSpecies,2], paired = T)
        wilcox.test(dinamicDisappearingAbundances[rownames(dinamicDisappearingAbundances) %in% outflowSpecies,1], dinamicDisappearingAbundances[rownames(dinamicDisappearingAbundances) %in% outflowSpecies,2], paired = T)
        
        wilcox.test(dinamicAppearingAbundances[inflowSpecies,1], dinamicAppearingAbundances[inflowSpecies,2], paired = T)
        wilcox.test(dinamicAppearingAbundances[rownames(dinamicAppearingAbundances) %in% outflowSpecies,1], dinamicAppearingAbundances[rownames(dinamicAppearingAbundances) %in% outflowSpecies,2], paired = T)
        
        
        
        wilcox.test(dinamicDisappearingCounts[inflowSpecies,1], dinamicDisappearingCounts[inflowSpecies,2], paired = T)
        wilcox.test(dinamicDisappearingCounts[rownames(dinamicDisappearingCounts) %in% outflowSpecies,1], dinamicDisappearingCounts[rownames(dinamicDisappearingCounts) %in% outflowSpecies,2], paired = T)
        
        wilcox.test(dinamicAppearingCounts[inflowSpecies,1], dinamicAppearingCounts[inflowSpecies,2], paired = T)
        wilcox.test(dinamicAppearingCounts[rownames(dinamicAppearingCounts) %in% outflowSpecies,1], dinamicAppearingCounts[rownames(dinamicAppearingCounts) %in% outflowSpecies,2], paired = T)
        
      }
      
      
      checkWellnessCountMode = T
      if (checkWellnessCountMode) {
        
        newInflowSpecies = inflowSpecies[order(inflow[inflowSpecies],decreasing = T)]
        newOutflowSpecies = outflowSpecies[order(outflow[outflowSpecies],decreasing = T)]
        newOutflowSpecies = newOutflowSpecies[newOutflowSpecies %in% rownames(dinamicAllAbundances)]
        
        
        abundanceMat = cbind((appearingAbundances[inflowSpecies,]), (disappearingAbundances[inflowSpecies,]))
        
        
        plotBarcode((dinamicAllAbundances[newInflowSpecies,]))
        plotBarcode((dinamicAllAbundances[newOutflowSpecies,]))
        
        levelplot(t(dinamicAllCounts[newInflowSpecies,]), 
                  col.regions= plasma(500),
                  aspect = 2,
                  xlab="",ylab="",
                  at=seq(min(dinamicAllCounts), max(dinamicAllCounts), length.out=500),
                  scales = list(draw=F))
        
        levelplot(t(dinamicAllCounts[newOutflowSpecies,]), 
                  col.regions= plasma(500),
                  aspect = 2,
                  xlab="",ylab="",
                  at=seq(min(dinamicAllCounts), max(dinamicAllCounts), length.out=500),
                  scales = list(draw=F))
        
        
        plotBarcode((appearingAbundances[newInflowSpecies,]))
        plotBarcode((disappearingAbundances[newInflowSpecies,]))
        
        plotBarcode((appearingAbundances[ newOutflowSpecies,]))
        plotBarcode((disappearingAbundances[newOutflowSpecies,]))
        
        levelplot(t(appearingCounts[newInflowSpecies,]), 
                  col.regions= plasma(500),
                  aspect = 2,
                  xlab="",ylab="",
                  at=seq(min(appearingCounts), max(appearingCounts), length.out=500),
                  scales = list(draw=F))
        
        levelplot(t(disappearingCounts[newInflowSpecies,]), 
                  col.regions= plasma(500),
                  aspect = 2,
                  xlab="",ylab="",
                  at=seq(min(disappearingCounts), max(disappearingCounts), length.out=500),
                  scales = list(draw=F))
        
        levelplot(t(appearingCounts[ newOutflowSpecies,]), 
                  col.regions= plasma(500),
                  aspect = 2,
                  xlab="",ylab="",
                  at=seq(min(appearingCounts), max(appearingCounts), length.out=500),
                  scales = list(draw=F))
        
        levelplot(t(disappearingCounts[  newOutflowSpecies,]), 
                  col.regions= plasma(500),
                  aspect = 2,
                  xlab="",ylab="",
                  #at=seq(min(disappearingCounts), max(disappearingCounts), length.out=500),
                  scales = list(draw=F))
        
        
      }
      
      inflowDinamicMat = getInflow(dinamicMetaTab, dinamicMgsMatImputed)
      outflowDinamicMat = getOutflow(dinamicMetaTab, dinamicMgsMatImputed)
      
      dinamicInflow = inflowDinamicMat[,1]
      dinamicOutflow = outflowDinamicMat[,1]
      dinamicPrevalence = apply(dinamicMgsMatImputed, 1, function(x) sum(x>0)/length(x))
      
      dinamicInOutFlowTab = data.frame(inflow=dinamicInflow,
                                       outflow=dinamicOutflow,
                                       prevalence=dinamicPrevalence,
                                       species = getSpeciesName(names(dinamicInflow),gssOssTabNew) ) 
      
      
      ggplot(dinamicInOutFlowTab, aes(x=outflow, y=inflow, fill=inflow)) + 
        scale_fill_gradient2(midpoint = 0.2, low="#FFD70044", mid = "white", high = "purple") +
        geom_point(size=3, shape=21, colour="#d3d3d366")+ xlab("") + ylab("")+
        geom_vline(xintercept = 0.3, colour="gray",linetype="dashed")+
        geom_hline(yintercept = 0.3, colour="gray",linetype="dashed")+
        #geom_abline(slope=1, intercept = 0)+
        theme(panel.grid.major = element_blank(), 
              panel.grid.minor = element_blank(),
              legend.position = "none",
              text = element_text(size = 15),
              panel.background = element_rect(fill = "white", 
                                              colour = "black", 
                                              size = 0.5, 
                                              linetype = "solid"))
      
      dinamicInflowSpecies = names(dinamicInflow[dinamicInflow>=0.3])
      dinamicOutflowSpecies = names(dinamicOutflow[dinamicOutflow>=0.3])
      
      dinamicSharedpecies = dinamicInflowSpecies[dinamicInflowSpecies  %in% dinamicOutflowSpecies]
      
      dinamicInflowSpecies = dinamicInflowSpecies[!dinamicInflowSpecies %in% dinamicSharedpecies]
      dinamicOutflowSpecies = dinamicOutflowSpecies[!dinamicOutflowSpecies %in% dinamicSharedpecies]
      
      table(inflowSpecies %in% dinamicInflowSpecies)
      table(inflowSpecies %in% dinamicOutflowSpecies)
      
      table(outflowSpecies %in% dinamicInflowSpecies)
      table(outflowSpecies %in% dinamicOutflowSpecies)
      
      
      luT1DMode = F
      if (luT1DMode) {
        plot(dinamicInflow[names(luInflow)], inflow[names(luInflow)])
        abline(a=0,b=1,col="red")
        plot(dinamicOutflow[names(luOutflow)], outflow[names(luOutflow)])
        abline(a=0,b=1,col="red")
        
        cor.test(dinamicInflow[names(luInflow)], inflow[names(luInflow)])
        cor.test(dinamicOutflow[names(luOutflow)], outflow[names(luOutflow)])
        
      }
      
      wellnessMode = T
      if (wellnessMode) {
        
        overlapSpecies = names(dinamicOutflow)[names(dinamicOutflow) %in% names(outflow)]
        
        
        plot(dinamicInflow[overlapSpecies], inflow[overlapSpecies])
        abline(a=0,b=1,col="red")
        
        plot(dinamicOutflow[overlapSpecies], outflow[overlapSpecies])
        abline(a=0,b=1,col="red")
        
        cor.test(dinamicInflow[overlapSpecies], inflow[overlapSpecies], method="spearman")
        cor.test(dinamicOutflow[overlapSpecies], outflow[overlapSpecies], method="spearman")
        
        
        wellnessDinamicTab = data.frame(inflow.dinamic = dinamicInflow[overlapSpecies],
                                        inflow.wellness = inflow[overlapSpecies],
                                        outflow.dinamic = dinamicOutflow[overlapSpecies],
                                        outflow.wellness = outflow[overlapSpecies],
                                        prevalence.dinamic = dinamicPrevalence[overlapSpecies],
                                        prevalence.wellness = freqs_wellness[overlapSpecies],
                                        species = getSpeciesName(overlapSpecies, gssOssTabNew))
        #
        wellnessDinamicTab = wellnessDinamicTab[wellnessDinamicTab$prevalence.dinamic>=0.1 & wellnessDinamicTab$prevalence.wellness >= 0.1, ]
        
        cor.test(wellnessDinamicTab$inflow.dinamic, wellnessDinamicTab$inflow.wellness, method="spearman")
        cor.test(wellnessDinamicTab$outflow.dinamic, wellnessDinamicTab$outflow.wellness, method="spearman")
        
        
        ggplot(wellnessDinamicTab, aes(x=inflow.dinamic, y=inflow.wellness)) + 
          geom_point(size=3, shape=16, colour="#00000055") + xlab("") + ylab("")+ #size=3, 
          #scale_fill_gradient2(midpoint = 0.2, low="#FFD70044", mid = "white", high = "purple") +
          geom_smooth(method = "lm",  se = T, colour="red")+
          scale_x_continuous(breaks = scales::pretty_breaks(n = 2)) +
          scale_y_continuous(breaks = scales::pretty_breaks(n = 2)) +
          #geom_hline(yintercept=6)+
          #geom_abline(slope=1, intercept = 0)+
          theme(panel.grid.major = element_blank(), 
                panel.grid.minor = element_blank(),
                legend.position = "none",
                text = element_text(size = 15),
                panel.background = element_rect(fill = "white", 
                                                colour = "black", 
                                                size = 0.5, 
                                                linetype = "solid"))
        
        ggplot(wellnessDinamicTab, aes(x=outflow.dinamic, y=outflow.wellness)) + 
          geom_point(size=3, shape=16, colour="#00000055") + xlab("") + ylab("")+ #size=3, 
          #scale_fill_gradient2(midpoint = 0.2, low="#FFD70044", mid = "white", high = "purple") +
          geom_smooth(method = "lm",  se = T, colour="red")+
          scale_x_continuous(breaks = scales::pretty_breaks(n = 2)) +
          scale_y_continuous(breaks = scales::pretty_breaks(n = 2)) +
          #geom_hline(yintercept=6)+
          #geom_abline(slope=1, intercept = 0)+
          theme(panel.grid.major = element_blank(), 
                panel.grid.minor = element_blank(),
                legend.position = "none",
                text = element_text(size = 15),
                panel.background = element_rect(fill = "white", 
                                                colour = "black", 
                                                size = 0.5, 
                                                linetype = "solid"))
        
        
        ####
        par(mar=c(8,4,4,1))
        #inflowList = list(Sweden=inflow[names(inflow)], DINAMIC=dinamicInflow[names(inflow)])
        #outflowList = list(Sweden=outflow[names(outflow)], DINAMIC=dinamicOutflow[names(outflow)])
        inflowList = list(Sweden=inflow[(overlapSpecies)], DINAMIC=dinamicInflow[(overlapSpecies)])
        outflowList = list(Sweden=outflow[(overlapSpecies)], DINAMIC=dinamicOutflow[(overlapSpecies)])
        
        boxplot(inflowList,las=2)
        wilcox.test(inflowList$Sweden, inflowList$DINAMIC)
        
        boxplot(outflowList,las=2)
        wilcox.test(outflowList$Sweden, outflowList$DINAMIC)
        
        diffMode = T
        if (diffMode) {
          #overlapBothSpecies = names(dinamicMedOutflow)[names(dinamicMedOutflow) %in% names(dinamicContOutflow)]
          
          inflowDiffOuts = do.call(rbind,sapply(overlapSpecies, function(currMsp){
            medInflow = inflowDinamicMat[currMsp,1]
            medInflowSe = inflowDinamicMat[currMsp,4]
            medZscore = medInflow / medInflowSe
            
            
            conInflow = inflowOut[currMsp,1]
            conInflowSe = inflowOut[currMsp,4]
            conZscore = conInflow / conInflowSe
            
            diffZscore = medZscore - conZscore
            pvalue = 2*pnorm(abs(diffZscore), lower.tail = F)
            
            if (medInflow == 0) {
              diffZscore = - conZscore 
              pvalue = 2*pnorm(abs(diffZscore), lower.tail = F)
              
            }
            if (conInflow == 0) {
              diffZscore = medZscore 
              pvalue = 2*pnorm(abs(diffZscore), lower.tail = F)
            }
            
            return(c(diffZscore, medInflow, conInflow, pvalue))
          },simplify=F))
          colnames(inflowDiffOuts)  = c("diff.z","DINAMIC", "WELLNESS", "pvalue")
          inflowDiffOuts = data.frame(inflowDiffOuts)
          inflowDiffOuts$species = getSpeciesName(rownames(inflowDiffOuts),gssOssTabNew)
          inflowDiffOuts$prevalence.DINAMIC = dinamicPrevalence[overlapSpecies]
          inflowDiffOuts$prevalence.WELLNESS = freqs_wellness[overlapSpecies]
          inflowDiffOuts = inflowDiffOuts[inflowDiffOuts$prevalence.WELLNESS > 0.1 & inflowDiffOuts$prevalence.DINAMIC > 0.1, ]
          inflowDiffOuts = inflowDiffOuts[!is.nan(inflowDiffOuts$pvalue) & inflowDiffOuts$pvalue < 0.01, ]
          
          inflowUpSpecies = rownames(inflowDiffOuts)[inflowDiffOuts$diff.z>0]
          inflowDnSpecies = rownames(inflowDiffOuts)[inflowDiffOuts$diff.z<0]
          inflowBothSpecies = rownames(inflowDiffOuts)
          
          
          write.table(inflowDiffOuts, "inflowDiffOuts.wellness.dinamic.txt", sep="\t")
          
          ggplot(inflowDiffOuts, aes(x=DINAMIC, y=WELLNESS)) + 
            geom_point(size=3, shape=16, colour= ifelse( inflowDiffOuts$pvalue < 0.01, "#FF000077", "#00000055")) + xlab("") + ylab("")+ #size=3, 
            #scale_fill_gradient2(midpoint = 0.2, low="#FFD70044", mid = "white", high = "purple") +
            geom_smooth(method = "lm",  se = T, colour="red")+
            #geom_hline(yintercept=6)+
            #geom_abline(slope=1, intercept = 0)+
            theme(panel.grid.major = element_blank(), 
                  panel.grid.minor = element_blank(),
                  legend.position = "none",
                  text = element_text(size = 15),
                  panel.background = element_rect(fill = "white", 
                                                  colour = "black", 
                                                  size = 0.5, 
                                                  linetype = "solid"))
          
          outflowDiffOuts = do.call(rbind,sapply(overlapSpecies, function(currMsp){
            medOutflow = outflowDinamicMat[currMsp,1]
            medOutflowSe = outflowDinamicMat[currMsp,4]
            medZscore = medOutflow / medOutflowSe
            
            
            conOutflow = outflowOut[currMsp,1]
            conOutflowSe = outflowOut[currMsp,4]
            conZscore = conOutflow / conOutflowSe
            
            diffZscore = medZscore - conZscore
            pvalue = 2*pnorm(abs(diffZscore), lower.tail = F)
            
            if (medOutflow == 0) {
              diffZscore = - conZscore 
              pvalue = 2*pnorm(abs(diffZscore), lower.tail = F)
              
            }
            if (conOutflow == 0) {
              diffZscore = medZscore 
              pvalue = 2*pnorm(abs(diffZscore), lower.tail = F)
              
            }
            
            return(c(diffZscore, medOutflow, conOutflow, pvalue))
          },simplify=F))
          colnames(outflowDiffOuts)  = c("diff.z","DINAMIC", "WELLNESS", "pvalue")
          outflowDiffOuts = data.frame(outflowDiffOuts)
          outflowDiffOuts$species = getSpeciesName(rownames(outflowDiffOuts),gssOssTabNew)
          outflowDiffOuts$prevalence.DINAMIC = dinamicPrevalence[overlapSpecies]
          outflowDiffOuts$prevalence.WELLNESS = freqs_wellness[overlapSpecies]
          outflowDiffOuts = outflowDiffOuts[outflowDiffOuts$prevalence.WELLNESS > 0.1 & outflowDiffOuts$prevalence.DINAMIC > 0.1, ]
          
          #View(outflowDiffOuts)
          outflowDiffOuts = outflowDiffOuts[!is.nan(outflowDiffOuts$pvalue) & outflowDiffOuts$pvalue < 0.01, ]
          outflowUpSpecies = rownames(outflowDiffOuts)[outflowDiffOuts$diff.z>0]
          outflowDnSpecies = rownames(outflowDiffOuts)[outflowDiffOuts$diff.z<0]
          outflowBothSpecies = rownames(outflowDiffOuts)
          
          ggplot(outflowDiffOuts, aes(x=DINAMIC, y=WELLNESS)) + 
            geom_point(size=3, shape=16, colour= ifelse( outflowDiffOuts$pvalue < 0.01, "#FF000077", "#00000055")) + xlab("") + ylab("")+ #size=3, 
            #scale_fill_gradient2(midpoint = 0.2, low="#FFD70044", mid = "white", high = "purple") +
            geom_smooth(method = "lm",  se = T, colour="red")+
            #geom_hline(yintercept=6)+
            #geom_abline(slope=1, intercept = 0)+
            theme(panel.grid.major = element_blank(), 
                  panel.grid.minor = element_blank(),
                  legend.position = "none",
                  text = element_text(size = 15),
                  panel.background = element_rect(fill = "white", 
                                                  colour = "black", 
                                                  size = 0.5, 
                                                  linetype = "solid"))
          table(inflowUpSpecies)
          
          write.table(outflowDiffOuts, "outflowDiffOuts.wellness.dinamic.txt", sep="\t")
          
          
        }
      }
      
      contArmMode = T
      if (contArmMode) {
        ### control arm ###
        
        dinamicMetaTabCont = dinamicMetaTab[dinamicMetaTab$type=="CONTROL DIET",]
        dinamicMgsMatCont = dinamicMgsMatImputed[,match(dinamicMetaTabCont$sample.ID,colnames(dinamicMgsMatImputed))] # 1268 x 114
        dinamicMgsMatCont = dinamicMgsMatCont[rowSums(dinamicMgsMatCont)!=0,]
        dinamicMgsMatCont = dinamicMgsMatCont[!apply(dinamicMgsMatCont,1,all),] # 1077 x 114 
        
        dinamicPrevalenceCont = apply(dinamicMgsMatCont, 1, function(x) sum(x>0)/length(x))
        
        
        inflowDinamicContMat = getInflow(dinamicMetaTabCont, dinamicMgsMatCont)
        outflowDinamicContMat = getOutflow(dinamicMetaTabCont, dinamicMgsMatCont)
        
        dinamicContInflow = inflowDinamicContMat[,1]
        dinamicContOutflow = outflowDinamicContMat[,1]
        
        overlapContSpecies = names(dinamicContOutflow)[names(dinamicContOutflow) %in% names(outflow)]
        
        ################
        
        plot(dinamicContInflow[overlapContSpecies], inflow[overlapContSpecies])
        abline(a=0,b=1,col="red")
        
        plot(dinamicContOutflow[overlapContSpecies], outflow[overlapContSpecies])
        abline(a=0,b=1,col="red")
        
        cor.test(dinamicContInflow[overlapContSpecies], inflow[overlapContSpecies], method="spearman")
        cor.test(dinamicContOutflow[overlapContSpecies], outflow[overlapContSpecies], method="spearman")
        
        
        wellnessDinamicContTab = data.frame(inflow.dinamic.cont = dinamicContInflow[overlapContSpecies],
                                            inflow.wellness = inflow[overlapContSpecies],
                                            outflow.dinamic.cont = dinamicContOutflow[overlapContSpecies],
                                            outflow.wellness = outflow[overlapContSpecies],
                                            prevalence.dinamic = dinamicPrevalenceCont[overlapContSpecies],
                                            prevalence.wellness = freqs_wellness[overlapContSpecies],
                                            species = getSpeciesName(overlapContSpecies, gssOssTabNew))
        
        wellnessDinamicContTab = wellnessDinamicContTab[wellnessDinamicContTab$prevalence.dinamic>=0.1 & wellnessDinamicContTab$prevalence.wellness >= 0.1, ]
        
        cor.test(wellnessDinamicContTab$inflow.dinamic.cont, wellnessDinamicContTab$inflow.wellness , method="spearman")
        cor.test(wellnessDinamicContTab$outflow.dinamic.cont, wellnessDinamicContTab$outflow.wellness , method="spearman")
        
        ggplot(wellnessDinamicContTab, aes(x=inflow.dinamic.cont, y=inflow.wellness)) + 
          geom_point(size=3, shape=16, colour="#00000055") + xlab("") + ylab("")+ #size=3, 
          #scale_fill_gradient2(midpoint = 0.2, low="#FFD70044", mid = "white", high = "purple") +
          geom_smooth(method = "lm",  se = T, colour="red")+
          #geom_hline(yintercept=6)+
          #geom_abline(slope=1, intercept = 0)+
          theme(panel.grid.major = element_blank(), 
                panel.grid.minor = element_blank(),
                legend.position = "none",
                text = element_text(size = 15),
                panel.background = element_rect(fill = "white", 
                                                colour = "black", 
                                                size = 0.5, 
                                                linetype = "solid"))
        
        ggplot(wellnessDinamicContTab, aes(x=outflow.dinamic.cont, y=outflow.wellness)) + 
          geom_point(size=3, shape=16, colour="#00000055") + xlab("") + ylab("")+ #size=3, 
          #scale_fill_gradient2(midpoint = 0.2, low="#FFD70044", mid = "white", high = "purple") +
          geom_smooth(method = "lm",  se = T, colour="red")+
          #geom_hline(yintercept=6)+
          #geom_abline(slope=1, intercept = 0)+
          theme(panel.grid.major = element_blank(), 
                panel.grid.minor = element_blank(),
                legend.position = "none",
                text = element_text(size = 15),
                panel.background = element_rect(fill = "white", 
                                                colour = "black", 
                                                size = 0.5, 
                                                linetype = "solid"))
        
        
        ################
        # 
        # plot(dinamicContInflow[(overlapContSpecies)], inflow[(overlapContSpecies)])
        # abline(a=0,b=1,col="red")
        # 
        # plot(dinamicContOutflow[(overlapContSpecies)], outflow[(overlapContSpecies)])
        # abline(a=0,b=1,col="red")
        # 
        # cor.test(dinamicContInflow[(overlapContSpecies)], inflow[(overlapContSpecies)])
        # cor.test(dinamicContOutflow[(overlapContSpecies)], outflow[(overlapContSpecies)])
        # 
        # 
      }
      
      medArmMode = T
      if (medArmMode) {
        ### MD arm ###
        
        dinamicMetaTabMed = dinamicMetaTab[dinamicMetaTab$type=="MED DIET",]
        dinamicMgsMatMed = dinamicMgsMatImputed[,match(dinamicMetaTabMed$sample.ID,colnames(dinamicMgsMatImputed))]
        dinamicMgsMatMed = dinamicMgsMatMed[rowSums(dinamicMgsMatMed)!=0,]
        dinamicMgsMatMed = dinamicMgsMatMed[!apply(dinamicMgsMatMed,1,all),] # 1090 x 126 
        dinamicPrevalenceMed = apply(dinamicMgsMatMed, 1, function(x) sum(x>0)/length(x))
        
        
        inflowDinamicMedMat = getInflow(dinamicMetaTabMed, dinamicMgsMatMed)
        outflowDinamicMedMat = getOutflow(dinamicMetaTabMed, dinamicMgsMatMed)
        
        dinamicMedInflow = inflowDinamicMedMat[,1]
        dinamicMedOutflow = outflowDinamicMedMat[,1]
        
        overlapMedSpecies = names(dinamicMedOutflow)[names(dinamicMedOutflow) %in% names(outflow)]
        
        ################
        
        plot(dinamicMedInflow[overlapMedSpecies], inflow[overlapMedSpecies])
        abline(a=0,b=1,col="red")
        
        plot(dinamicMedOutflow[overlapMedSpecies], outflow[overlapMedSpecies])
        abline(a=0,b=1,col="red")
        
        cor.test(dinamicMedInflow[overlapMedSpecies], inflow[overlapMedSpecies], method="spearman")
        cor.test(dinamicMedOutflow[overlapMedSpecies], outflow[overlapMedSpecies], method="spearman")
        
        
        wellnessDinamicMedTab = data.frame(inflow.dinamic.med = dinamicMedInflow[overlapMedSpecies],
                                           inflow.wellness = inflow[overlapMedSpecies],
                                           outflow.dinamic.med = dinamicMedOutflow[overlapMedSpecies],
                                           outflow.wellness = outflow[overlapMedSpecies],
                                           prevalence.dinamic = dinamicPrevalenceMed[overlapMedSpecies],
                                           prevalence.wellness = freqs_wellness[overlapMedSpecies],
                                           species = getSpeciesName(overlapMedSpecies, gssOssTabNew))
        
        wellnessDinamicMedTab = wellnessDinamicMedTab[wellnessDinamicMedTab$prevalence.dinamic>=0.1 & wellnessDinamicMedTab$prevalence.wellness >= 0.1, ]
        
        cor.test(wellnessDinamicMedTab$inflow.dinamic.med, wellnessDinamicMedTab$inflow.wellness , method="spearman")
        cor.test(wellnessDinamicMedTab$outflow.dinamic.med, wellnessDinamicMedTab$outflow.wellness , method="spearman")
        
        ggplot(wellnessDinamicMedTab, aes(x=inflow.dinamic.med, y=inflow.wellness)) + 
          geom_point(size=3, shape=16, colour="#00000055") + xlab("") + ylab("")+ #size=3, 
          #scale_fill_gradient2(midpoint = 0.2, low="#FFD70044", mid = "white", high = "purple") +
          geom_smooth(method = "lm",  se = T, colour="red")+
          #geom_hline(yintercept=6)+
          #geom_abline(slope=1, intercept = 0)+
          theme(panel.grid.major = element_blank(), 
                panel.grid.minor = element_blank(),
                legend.position = "none",
                text = element_text(size = 15),
                panel.background = element_rect(fill = "white", 
                                                colour = "black", 
                                                size = 0.5, 
                                                linetype = "solid"))
        
        ggplot(wellnessDinamicMedTab, aes(x=outflow.dinamic.med, y=outflow.wellness)) + 
          geom_point(size=3, shape=16, colour="#00000055") + xlab("") + ylab("")+ #size=3, 
          #scale_fill_gradient2(midpoint = 0.2, low="#FFD70044", mid = "white", high = "purple") +
          geom_smooth(method = "lm",  se = T, colour="red")+
          #geom_hline(yintercept=6)+
          #geom_abline(slope=1, intercept = 0)+
          theme(panel.grid.major = element_blank(), 
                panel.grid.minor = element_blank(),
                legend.position = "none",
                text = element_text(size = 15),
                panel.background = element_rect(fill = "white", 
                                                colour = "black", 
                                                size = 0.5, 
                                                linetype = "solid"))
        
        #######
        
        plot(dinamicMedInflow[(overlapMedSpecies)], inflow[(overlapMedSpecies)])
        abline(a=0,b=1,col="red")
        
        plot(dinamicMedOutflow[(overlapMedSpecies)], outflow[(overlapMedSpecies)])
        abline(a=0,b=1,col="red")
        
        cor.test(dinamicMedInflow[(overlapMedSpecies)], inflow[(overlapMedSpecies)])
        cor.test(dinamicMedOutflow[(overlapMedSpecies)], outflow[(overlapMedSpecies)])
        
      }
      
      contVsMedMode = T
      if (contVsMedMode) {
        
        ### comparing both ###
        diffMode = T
        if (diffMode) {
          overlapBothSpecies = names(dinamicMedOutflow)[names(dinamicMedOutflow) %in% names(dinamicContOutflow)]
          
          inflowDiffOuts = do.call(rbind,sapply(overlapBothSpecies, function(currMsp){
            medInflow = inflowDinamicMedMat[currMsp,1]
            medInflowSe = inflowDinamicMedMat[currMsp,4]
            medZscore = medInflow / medInflowSe
            
            
            conInflow = inflowDinamicContMat[currMsp,1]
            conInflowSe = inflowDinamicContMat[currMsp,4]
            conZscore = conInflow / conInflowSe
            
            diffZscore = medZscore - conZscore
            pvalue = 2*pnorm(abs(diffZscore), lower.tail = F)
            
            if (medInflow == 0) {
              diffZscore = - conZscore 
              pvalue = 2*pnorm(abs(diffZscore), lower.tail = F)
              
            }
            if (conInflow == 0) {
              diffZscore = medZscore 
              pvalue = 2*pnorm(abs(diffZscore), lower.tail = F)
            }
            
            return(c(diffZscore, medInflow, conInflow, pvalue))
          },simplify=F))
          colnames(inflowDiffOuts)  = c("diff.z","MED", "CON", "pvalue")
          inflowDiffOuts = data.frame(inflowDiffOuts)
          inflowDiffOuts$species = getSpeciesName(rownames(inflowDiffOuts),gssOssTabNew)
          inflowDiffOuts = inflowDiffOuts[!is.nan(inflowDiffOuts$pvalue) & inflowDiffOuts$pvalue < 0.1, ]
          
          inflowUpSpecies = rownames(inflowDiffOuts)[inflowDiffOuts$diff.z>0]
          inflowDnSpecies = rownames(inflowDiffOuts)[inflowDiffOuts$diff.z<0]
          inflowBothSpecies = rownames(inflowDiffOuts)
          
          
          write.table(inflowDiffOuts, "inflowDiffOuts.txt", sep="\t")
          
          outflowDiffOuts = do.call(rbind,sapply(overlapBothSpecies, function(currMsp){
            medOutflow = outflowDinamicMedMat[currMsp,1]
            medOutflowSe = outflowDinamicMedMat[currMsp,4]
            medZscore = medOutflow / medOutflowSe
            
            
            conOutflow = outflowDinamicContMat[currMsp,1]
            conOutflowSe = outflowDinamicContMat[currMsp,4]
            conZscore = conOutflow / conOutflowSe
            
            diffZscore = medZscore - conZscore
            pvalue = 2*pnorm(abs(diffZscore), lower.tail = F)
            
            if (medOutflow == 0) {
              diffZscore = - conZscore 
              pvalue = 2*pnorm(abs(diffZscore), lower.tail = F)
              
            }
            if (conOutflow == 0) {
              diffZscore = medZscore 
              pvalue = 2*pnorm(abs(diffZscore), lower.tail = F)
              
            }
            
            return(c(diffZscore, medOutflow, conOutflow, pvalue))
          },simplify=F))
          colnames(outflowDiffOuts)  = c("diff.z","MED", "CON", "pvalue")
          outflowDiffOuts = data.frame(outflowDiffOuts)
          outflowDiffOuts$species = getSpeciesName(rownames(outflowDiffOuts),gssOssTabNew)
          #View(outflowDiffOuts)
          outflowDiffOuts = outflowDiffOuts[!is.nan(outflowDiffOuts$pvalue) & outflowDiffOuts$pvalue < 0.1, ]
          outflowUpSpecies = rownames(outflowDiffOuts)[outflowDiffOuts$diff.z>0]
          outflowDnSpecies = rownames(outflowDiffOuts)[outflowDiffOuts$diff.z<0]
          outflowBothSpecies = rownames(outflowDiffOuts)
          
          
          table(inflowUpSpecies)
          
          write.table(outflowDiffOuts, "outflowDiffOuts.txt", sep="\t")
          
        }
        
        corrMode = T
        if (corrMode) {
          overlapBothSpecies = names(dinamicMedOutflow)[names(dinamicMedOutflow) %in% names(dinamicContOutflow)]
          
          wellnessDinamicBothTab = data.frame(inflow.dinamic.med = dinamicMedInflow[overlapBothSpecies],
                                              inflow.dinamic.cont = dinamicContInflow[overlapBothSpecies],
                                              outflow.dinamic.med = dinamicMedOutflow[overlapBothSpecies],
                                              outflow.dinamic.cont = dinamicContOutflow[overlapBothSpecies],
                                              prevalence.dinamic.med = dinamicPrevalenceMed[overlapBothSpecies],
                                              prevalence.dinamic.cont = dinamicPrevalenceCont[overlapBothSpecies],
                                              msp = overlapBothSpecies,
                                              inflowChanged = overlapBothSpecies %in% inflowBothSpecies,
                                              outflowChanged =  overlapBothSpecies %in% outflowBothSpecies,
                                              species = getSpeciesName(overlapBothSpecies, gssOssTabNew),
                                              stringsAsFactors = F)
          
          wellnessDinamicBothTab = wellnessDinamicBothTab[wellnessDinamicBothTab$prevalence.dinamic.med>=0.1 & wellnessDinamicBothTab$prevalence.dinamic.cont >= 0.1, ]
          
          cor.test(wellnessDinamicBothTab$inflow.dinamic.med, wellnessDinamicBothTab$inflow.dinamic.cont , method="spearman")
          cor.test(wellnessDinamicBothTab$outflow.dinamic.med, wellnessDinamicBothTab$outflow.dinamic.cont , method="spearman")
          
          ggplot(wellnessDinamicBothTab, aes(x=inflow.dinamic.med, y=inflow.dinamic.cont)) + 
            #geom_point(size=3, shape=16, colour= ifelse( wellnessDinamicBothTab$inflowChanged, "#FF000077", "#00000055")) + xlab("") + ylab("")+ #size=3, 
            geom_point(size=3, shape=16, colour= "#00000055") + xlab("") + ylab("")+ #size=3, 
            #scale_fill_gradient2(midpoint = 0.2, low="#FFD70044", mid = "white", high = "purple") +
            geom_smooth(method = "lm",  se = T, colour="red")+
            #geom_hline(yintercept=6)+
            #geom_abline(slope=1, intercept = 0)+
            theme(panel.grid.major = element_blank(), 
                  panel.grid.minor = element_blank(),
                  legend.position = "none",
                  text = element_text(size = 15),
                  panel.background = element_rect(fill = "white", 
                                                  colour = "black", 
                                                  size = 0.5, 
                                                  linetype = "solid"))
          
          ggplot(wellnessDinamicBothTab, aes(x=outflow.dinamic.med, y=outflow.dinamic.cont)) + 
            #geom_point(size=3, shape=16, colour= ifelse( wellnessDinamicBothTab$outflowChanged, "#FF000077", "#00000055") ) + xlab("") + ylab("")+ #size=3, 
            geom_point(size=3, shape=16, colour= "#00000055" ) + xlab("") + ylab("")+ #size=3, 
            #scale_fill_gradient2(midpoint = 0.2, low="#FFD70044", mid = "white", high = "purple") +
            geom_smooth(method = "lm",  se = T, colour="red")+
            #geom_hline(yintercept=6)+
            #geom_abline(slope=1, intercept = 0)+
            theme(panel.grid.major = element_blank(), 
                  panel.grid.minor = element_blank(),
                  legend.position = "none",
                  text = element_text(size = 15),
                  panel.background = element_rect(fill = "white", 
                                                  colour = "black", 
                                                  size = 0.5, 
                                                  linetype = "solid"))
          
          
          #####
          plot(dinamicContInflow[(overlapBothSpecies)], dinamicMedInflow[(overlapBothSpecies)])
          abline(a=0,b=1,col="red")
          
          plot(dinamicContOutflow[(overlapBothSpecies)], dinamicMedOutflow[(overlapBothSpecies)])
          abline(a=0,b=1,col="red")
          
          cor.test(dinamicContInflow[(overlapBothSpecies)], dinamicMedInflow[(overlapBothSpecies)])
          cor.test(dinamicContOutflow[(overlapBothSpecies)], dinamicMedOutflow[(overlapBothSpecies)])
          
          allOverlap = names(dinamicMedInflow)[names(dinamicMedInflow) %in% names(dinamicContInflow)]
          allOverlap = allOverlap[allOverlap %in% names(dinamicInflow)]
          
          dinamicStatTab = data.frame(med.inflow=dinamicMedInflow[allOverlap],
                                      med.outflow=dinamicMedOutflow[allOverlap],
                                      con.inflow=dinamicContInflow[allOverlap],
                                      con.outflow=dinamicContOutflow[allOverlap],
                                      all.inflow=dinamicInflow[allOverlap],
                                      all.outflow=dinamicOutflow[allOverlap],
                                      sweden.inflow = inflow[allOverlap],
                                      sweden.outflow = outflow[allOverlap],
                                      pathogenic = ifelse(allOverlap %in% pathogenList, "Yes", ""),
                                      species = getSpeciesName(allOverlap,gssOssTabNew))
          
        }
        
      }
      
      wilcoxMode = T
      if (wilcoxMode) {
        
        
        dinamicMetaTab = dinamicMetaTab[order(dinamicMetaTab$subtype),]
        dinamicMetaTab = dinamicMetaTab[order(dinamicMetaTab$metadata.ID),]
        
        dinamicMetaTabCont = dinamicMetaTab[dinamicMetaTab$type=="CONTROL DIET",]
        dinamicMetaTabContV1 = dinamicMetaTab[dinamicMetaTab$type=="CONTROL DIET" & dinamicMetaTab$subtype=="v1",]
        dinamicMetaTabContV2 = dinamicMetaTab[dinamicMetaTab$type=="CONTROL DIET" & dinamicMetaTab$subtype=="v2",]
        dinamicMetaTabContV3 = dinamicMetaTab[dinamicMetaTab$type=="CONTROL DIET" & dinamicMetaTab$subtype=="v3",]
        
        
        dinamicMgsMatCont = dinamicMgsMatImputed[,match(dinamicMetaTabCont$sample.ID,colnames(dinamicMgsMatImputed))]# 1106 x 114
        dinamicMgsMatContV1 = dinamicMgsMatImputed[,match(dinamicMetaTabContV1$sample.ID,colnames(dinamicMgsMatImputed))]# 1106 x 38
        dinamicMgsMatContV2 = dinamicMgsMatImputed[,match(dinamicMetaTabContV2$sample.ID,colnames(dinamicMgsMatImputed))]# 1106 x 38
        dinamicMgsMatContV3 = dinamicMgsMatImputed[,match(dinamicMetaTabContV3$sample.ID,colnames(dinamicMgsMatImputed))]# 1106 x 38
        
        dinamicMgsMatContV1Samples = colnames(dinamicMgsMatContV1)
        dinamicMgsMatContV2Samples = colnames(dinamicMgsMatContV2)
        dinamicMgsMatContV3Samples = colnames(dinamicMgsMatContV3)
        
        dinamicVolcanoStatContV2 = getVolcanoStatPaired(dinamicMgsMatCont, dinamicMgsMatContV2Samples, dinamicMgsMatContV1Samples, gssOssTabNew, T)
        dinamicVolcanoStatContV3 = getVolcanoStatPaired(dinamicMgsMatCont, dinamicMgsMatContV3Samples, dinamicMgsMatContV1Samples, gssOssTabNew, T)
        
        
        dinamicMetaTabMed = dinamicMetaTab[dinamicMetaTab$type=="MED DIET",]
        dinamicMgsMatMed = dinamicMgsMatImputed[,match(dinamicMetaTabMed$sample.ID,colnames(dinamicMgsMatImputed))] # 1106 x 126
        dinamicMgsMatMedSamples = colnames(dinamicMgsMatMed)
        
        dinamicMgsMergeMat = cbind(dinamicMgsMatCont, dinamicMgsMatMed)
        
        dinamicVolcanoStat = getVolcanoStatPaired(dinamicMgsMergeMat, dinamicMgsMatMedSamples, dinamicMgsMatContSamples, gssOssTabNew, T)
        
        
      }
      
      
      
    }
    
    
    avgSignalMode = F
    if (avgSignalMode) {
      exSubj = c("P31","P19")
      
      dinamicMgsFile = "J:\\Deposit\\Project\\2018_microbiome_atlas\\in.outflow.dataset\\dinamic_vs_hs_10_4_igc2_smart_shared_reads_29042020.down.10M.fpkm.msp.txt"
      dinamicMetaFile = "J:\\Deposit\\Project\\2018_microbiome_atlas\\in.outflow.dataset\\dinamic.metadata.txt"
      dinamicMgsMat = read.delim(dinamicMgsFile, sep="\t", stringsAsFactors = F,row.names = 1)
      dinamicMgsMat = as.matrix(dinamicMgsMat) # 1990 x 244
      dinamicMgsMat = dinamicMgsMat[!rownames(dinamicMgsMat) %in% suppressMsps2,] # 1989 x 244
      
      dinamicMetaTab = read.delim(dinamicMetaFile, sep="\t", stringsAsFactors = F) # 246 x 6
      dinamicMetaTab = dinamicMetaTab[dinamicMetaTab$ID_MGP %in% colnames(dinamicMgsMat),]
      dinamicMetaTab$Time_point[dinamicMetaTab$Time_point == "baseline"] = "v1"
      dinamicMetaTab$Time_point[dinamicMetaTab$Time_point == "4 weeks"] = "v2"
      dinamicMetaTab$Time_point[dinamicMetaTab$Time_point == "8 weeks"] = "v3"
      colnames(dinamicMetaTab) = c("sample.ID","metadata.ID","type","subtype","Gender","Age")
      
      
      dinamicMgsMat = dinamicMgsMat[rowSums(dinamicMgsMat)!=0,]
      dinamicMgsMat = dinamicMgsMat[!apply(dinamicMgsMat,1,all),] # 1268 x 244 
      
      dinamicMetaTab = dinamicMetaTab[!dinamicMetaTab$metadata.ID %in% exSubj,]
      dinamicMgsMat = dinamicMgsMat[,match(dinamicMetaTab$sample.ID,colnames(dinamicMgsMat))]
      
      
      getInflow <- function(metaTab, mgsMat) {
        uniqSubj = unique(metaTab$metadata.ID)
        scoresByMsp = do.call(rbind, sapply(rownames(mgsMat), function(currMgs){
          
          markovSeqOfMsp = (sapply(uniqSubj, function(subj) {
            currSamples = metaTab$sample.ID[metaTab$metadata.ID==subj]
            currVisits = metaTab$subtype[match(currSamples, metaTab$sample.ID)]
            currSamples = currSamples[order(currVisits)]
            currMgsMat = mgsMat[,currSamples]
            currMgsSubjVec = (currMgsMat[currMgs,] >0)*1
            currMgsSubjVec = as.character(currMgsSubjVec)
            
          }, simplify = F))
          
          myMarkovFit<-markovchainFit(data=markovSeqOfMsp,confidencelevel = .9,method = "mle")
          myMarkovMat = myMarkovFit$estimate@transitionMatrix
          myMarkovUpper = myMarkovFit$upperEndpointMatrix
          myMarkovLower = myMarkovFit$lowerEndpointMatrix
          myMarkovSE = myMarkovFit$standardError
          
          if(dim(myMarkovMat)[1]==1) {
            rname = rownames(myMarkovMat)
            #if (rname == "0") return(0)
            #if (rname == "1") return(Inf)
            if (rname == "0") {
              colonized = 0
              colonizedUpper = NA
              colonizedLower = NA
              colonizedSE = NA
              result = c(colonized, colonizedUpper, colonizedLower, colonizedSE)
              return(result)
            }
            
            if (rname == "1") {
              colonized = 1
              colonizedUpper = NA
              colonizedLower = NA
              colonizedSE = NA
              result = c(colonized, colonizedUpper, colonizedLower, colonizedSE)
              
              return(result)
            }
          }
          else {
            colonized = myMarkovMat["0","1"] 
            colonizedUpper = myMarkovUpper["0","1"]
            colonizedLower = myMarkovLower["0","1"]
            colonizedSE = myMarkovSE["0","1"]
            result = c(colonized, colonizedUpper, colonizedLower, colonizedSE)
            
            #preserved = myMarkovMat["1","1"] 
            #relativeRisk = colonized/ preserved
            #return(relativeRisk)
            return(result)
          }
          
        }, simplify = F))
        
        rownames(scoresByMsp) = rownames(mgsMat)
        colnames(scoresByMsp) = c("inflow","upper","lower","SE")
        return(scoresByMsp)
        
      }
      getOutflow <- function(metaTab, mgsMat) {
        
        uniqSubj = unique(metaTab$metadata.ID)
        scoresByMsp = do.call(rbind, sapply(rownames(mgsMat), function(currMgs){
          
          markovSeqOfMsp = (sapply(uniqSubj, function(subj) {
            currSamples = metaTab$sample.ID[metaTab$metadata.ID==subj]
            currVisits = metaTab$subtype[match(currSamples, metaTab$sample.ID)]
            currSamples = currSamples[order(currVisits)]
            currMgsMat = mgsMat[,currSamples]
            currMgsSubjVec = (currMgsMat[currMgs,] >0)*1
            currMgsSubjVec = as.character(currMgsSubjVec)
            
          }, simplify = F))
          
          myMarkovFit<-markovchainFit(data=markovSeqOfMsp,confidencelevel = .9,method = "mle")
          myMarkovMat = myMarkovFit$estimate@transitionMatrix
          myMarkovUpper = myMarkovFit$upperEndpointMatrix
          myMarkovLower = myMarkovFit$lowerEndpointMatrix
          myMarkovSE = myMarkovFit$standardError
          
          if(dim(myMarkovMat)[1]==1) {
            rname = rownames(myMarkovMat)
            #if (rname == "0") return(Inf)
            #if (rname == "1") return(0)
            if (rname == "0") {
              resistant = 1
              resistantUpper = NA
              resistantLower = NA
              resistantSE = NA
              result = c(resistant, resistantUpper, resistantLower, resistantSE)
              return(result) 
            }
            if (rname == "1") {
              resistant = 0
              resistantUpper = NA
              resistantLower = NA
              resistantSE = NA
              result = c(resistant, resistantUpper, resistantLower, resistantSE)
              return(result)
            }
          }
          else {
            resistant = myMarkovMat["1","0"] 
            resistantUpper = myMarkovUpper["1","0"] 
            resistantLower = myMarkovLower["1","0"] 
            resistantSE = myMarkovSE["1","0"]
            result = c(resistant, resistantUpper, resistantLower, resistantSE)
            
            #preserved = myMarkovMat["0","0"] 
            #relativeRisk = resistant/ preserved
            #return(relativeRisk)
            return(result)
            
          }
          
        }, simplify = F))
        rownames(scoresByMsp) = rownames(mgsMat)
        colnames(scoresByMsp) = c("outflow","upper","lower","SE")
        return(scoresByMsp)
        
      }
      
      inflowDinamicMat = getInflow(dinamicMetaTab,dinamicMgsMat)
      outflowDinamicMat = getOutflow(dinamicMetaTab,dinamicMgsMat)
      
      dinamicInflow = inflowDinamicMat[,1]
      dinamicOutflow = outflowDinamicMat[,1]
      
      plot(dinamicInflow[names(luInflow)], inflow[names(luInflow)])
      abline(a=0,b=1,col="red")
      plot(dinamicOutflow[names(luOutflow)], outflow[names(luOutflow)])
      abline(a=0,b=1,col="red")
      
      cor.test(dinamicInflow[names(luInflow)], inflow[names(luInflow)])
      cor.test(dinamicOutflow[names(luOutflow)], outflow[names(luOutflow)])
      
      overlapSpecies = names(dinamicOutflow)[names(dinamicOutflow) %in% names(outflow)]
      
      
      plot(dinamicInflow[overlapSpecies], inflow[overlapSpecies])
      abline(a=0,b=1,col="red")
      
      plot(dinamicOutflow[overlapSpecies], outflow[overlapSpecies])
      abline(a=0,b=1,col="red")
      
      cor.test(dinamicInflow[overlapSpecies], inflow[overlapSpecies])
      cor.test(dinamicOutflow[overlapSpecies], outflow[overlapSpecies])
      
      
      
      ####
      par(mar=c(8,4,4,1))
      #inflowList = list(Sweden=inflow[names(inflow)], DINAMIC=dinamicInflow[names(inflow)])
      #outflowList = list(Sweden=outflow[names(outflow)], DINAMIC=dinamicOutflow[names(outflow)])
      inflowList = list(Sweden=inflow[(overlapSpecies)], DINAMIC=dinamicInflow[(overlapSpecies)])
      outflowList = list(Sweden=outflow[(overlapSpecies)], DINAMIC=dinamicOutflow[(overlapSpecies)])
      
      boxplot(inflowList,las=2)
      wilcox.test(inflowList$Sweden, inflowList$DINAMIC)
      
      boxplot(outflowList,las=2)
      wilcox.test(outflowList$Sweden, outflowList$DINAMIC)
      
      
      
      ### control arm ###
      
      dinamicMetaTabCont = dinamicMetaTab[dinamicMetaTab$type=="CONTROL DIET",]
      dinamicMgsMatCont = dinamicMgsMat[,match(dinamicMetaTabCont$sample.ID,colnames(dinamicMgsMat))] # 1268 x 114
      dinamicMgsMatCont = dinamicMgsMatCont[rowSums(dinamicMgsMatCont)!=0,]
      dinamicMgsMatCont = dinamicMgsMatCont[!apply(dinamicMgsMatCont,1,all),] # 1077 x 114 
      
      inflowDinamicContMat = getInflow(dinamicMetaTabCont, dinamicMgsMatCont)
      outflowDinamicContMat = getOutflow(dinamicMetaTabCont, dinamicMgsMatCont)
      
      dinamicContInflow = inflowDinamicContMat[,1]
      dinamicContOutflow = outflowDinamicContMat[,1]
      
      overlapContSpecies = names(dinamicContOutflow)[names(dinamicContOutflow) %in% names(outflow)]
      
      
      
      plot(dinamicContInflow[(overlapContSpecies)], inflow[(overlapContSpecies)])
      abline(a=0,b=1,col="red")
      
      plot(dinamicContOutflow[(overlapContSpecies)], outflow[(overlapContSpecies)])
      abline(a=0,b=1,col="red")
      
      cor.test(dinamicContInflow[(overlapContSpecies)], inflow[(overlapContSpecies)])
      cor.test(dinamicContOutflow[(overlapContSpecies)], outflow[(overlapContSpecies)])
      
      
      
      ### MD arm ###
      
      dinamicMetaTabMed = dinamicMetaTab[dinamicMetaTab$type=="MED DIET",]
      dinamicMgsMatMed = dinamicMgsMat[,match(dinamicMetaTabMed$sample.ID,colnames(dinamicMgsMat))]
      dinamicMgsMatMed = dinamicMgsMatMed[rowSums(dinamicMgsMatMed)!=0,]
      dinamicMgsMatMed = dinamicMgsMatMed[!apply(dinamicMgsMatMed,1,all),] # 1090 x 126 
      
      
      
      inflowDinamicMedMat = getInflow(dinamicMetaTabMed, dinamicMgsMatMed)
      outflowDinamicMedMat = getOutflow(dinamicMetaTabMed, dinamicMgsMatMed)
      
      dinamicMedInflow = inflowDinamicMedMat[,1]
      dinamicMedOutflow = outflowDinamicMedMat[,1]
      
      overlapMedSpecies = names(dinamicMedOutflow)[names(dinamicMedOutflow) %in% names(outflow)]
      
      
      plot(dinamicMedInflow[(overlapMedSpecies)], inflow[(overlapMedSpecies)])
      abline(a=0,b=1,col="red")
      
      plot(dinamicMedOutflow[(overlapMedSpecies)], outflow[(overlapMedSpecies)])
      abline(a=0,b=1,col="red")
      
      cor.test(dinamicMedInflow[(overlapMedSpecies)], inflow[(overlapMedSpecies)])
      cor.test(dinamicMedOutflow[(overlapMedSpecies)], outflow[(overlapMedSpecies)])
      
      
      ### comparing both ###
      
      overlapBothSpecies = names(dinamicMedOutflow)[names(dinamicMedOutflow) %in% names(dinamicContOutflow)]
      
      
      
      plot(dinamicContInflow[(overlapBothSpecies)], dinamicMedInflow[(overlapBothSpecies)])
      abline(a=0,b=1,col="red")
      
      plot(dinamicContOutflow[(overlapBothSpecies)], dinamicMedOutflow[(overlapBothSpecies)])
      abline(a=0,b=1,col="red")
      
      cor.test(dinamicContInflow[(overlapBothSpecies)], dinamicMedInflow[(overlapBothSpecies)])
      cor.test(dinamicContOutflow[(overlapBothSpecies)], dinamicMedOutflow[(overlapBothSpecies)])
      
      dinamicStatTab = data.frame(med.inflow=dinamicMedInflow,
                                  med.outflow=dinamicMedOutflow,
                                  con.inflow=dinamicContInflow,
                                  con.outflow=dinamicContOutflow,
                                  species = getSpeciesName(names(dinamicMedInflow),gssOssTabNew))
      
    }
    
    
    
  }
  
  newMarkovModeForWellnessV5andV6 = T
  if (newMarkovModeForWellnessV5andV6) {
    
    wellnessNewMetadataFile = "C:\\Data\\comparative.analysis.healthy.sweden\\wellnessMetadata.v5.v6.txt"
    wellnessNewMetadata = read.delim(wellnessNewMetadataFile, sep="\t", stringsAsFactors = F)
    wellnessNewMetadata$subtype = gsub("V","v",wellnessNewMetadata$subtype)
    
    wellnessNewMgsFile = "J:\\Deposit\\Project\\2018_microbiome_atlas\\atlas.visit5.6.metadata\\wellness.v5.v6.final.mgs.med.vec.10M.RData"
    load(wellnessNewMgsFile)
    
    wellnessNewMgsMat = mgs_med_vec_10m
    wellnessNewMgsMat = wellnessNewMgsMat[,colSums(wellnessNewMgsMat)!=0]
    
    if (F) {
      wellnessNewMgsMatScaled = wellnessNewMgsMat*10^9
      wellnessNewMgsMatScaledLog = log2(wellnessNewMgsMatScaled)
      wellnessNewMgsMatScaledLog[wellnessNewMgsMatScaledLog == -Inf] = 0
      
      wellnessNewMgsMatScaledNonZero = wellnessNewMgsMatScaled[wellnessNewMgsMatScaled!=0]
      wellnessNewMgsMatScaledNonZero = log2(wellnessNewMgsMatScaledNonZero)
      
      wellnessNew.fit.gamma = fitdist(wellnessNewMgsMatScaledNonZero, "gamma")
      plot(wellnessNew.fit.gamma)
      wellnessNew.fit.gamma$estimate
      qgamma(0.01, wellnessNew.fit.gamma$estimate["shape"], wellnessNew.fit.gamma$estimate["rate"])
      #4.63775
      
    }
    
    
    qCutOff =(2^(4.042488))/10^9
    #min(americanMgsMat[americanMgsMat!=0])
    
    wellnessNewMgsMatImputed = wellnessNewMgsMat
    wellnessNewMgsMatImputed[wellnessNewMgsMatImputed < qCutOff] <- 0
    wellnessNewMgsMatImputed = wellnessNewMgsMatImputed[rowSums(wellnessNewMgsMatImputed)!=0,]
    wellnessNewMgsMatImputed = wellnessNewMgsMatImputed[!apply(wellnessNewMgsMatImputed,1,all),]  
    wellnessNewMgsMatImputed = wellnessNewMgsMatImputed[,colnames(wellnessNewMgsMatImputed) %in% wellnessNewMetadata$sample.ID]
    wellnessNewMgsMatImputed = wellnessNewMgsMatImputed[,match(wellnessNewMetadata$sample.ID,colnames(wellnessNewMgsMatImputed))]
    
    wellnessNewMetadata = wellnessNewMetadata[,c("sample.ID","metadata.ID","type","subtype","Gender","Age")]
    
    v5v6Mode = F
    if (v5v6Mode) {
      
      getInflow <- function(metaTab, mgsMat) {
        mgsMat = mgsMat[,colnames(mgsMat) %in% metaTab$sample.ID]
        prevalence = apply(mgsMat, 1, function(currRow)mean(currRow>0))
        uniqSubj = unique(metaTab$metadata.ID)
        scoresByMsp = do.call(rbind, sapply(rownames(mgsMat), function(currMgs){
          
          markovSeqOfMsp = (sapply(uniqSubj, function(subj) {
            currSamples = metaTab$sample.ID[metaTab$metadata.ID==subj]
            currVisits = metaTab$subtype[match(currSamples, metaTab$sample.ID)]
            currSamples = currSamples[order(currVisits)]
            currMgsMat = mgsMat[,currSamples]
            currMgsSubjVec = (currMgsMat[currMgs,] >0)*1
            currMgsSubjVec = as.character(currMgsSubjVec)
            
          }, simplify = F))
          
          myMarkovFit<-markovchainFit(data=markovSeqOfMsp,confidencelevel = .9,method = "mle")
          myMarkovMat = myMarkovFit$estimate@transitionMatrix
          myMarkovUpper = myMarkovFit$upperEndpointMatrix
          myMarkovLower = myMarkovFit$lowerEndpointMatrix
          myMarkovSE = myMarkovFit$standardError
          
          if(dim(myMarkovMat)[1]==1) {
            rname = rownames(myMarkovMat)
            #if (rname == "0") return(0)
            #if (rname == "1") return(Inf)
            if (rname == "0") {
              colonized = 0
              colonizedUpper = NA
              colonizedLower = NA
              colonizedSE = NA
              result = c(colonized, colonizedUpper, colonizedLower, colonizedSE)
              return(result)
            }
            
            if (rname == "1") {
              colonized = 1
              colonizedUpper = NA
              colonizedLower = NA
              colonizedSE = NA
              result = c(colonized, colonizedUpper, colonizedLower, colonizedSE)
              
              return(result)
            }
          }
          else {
            colonized = myMarkovMat["0","1"] 
            colonizedUpper = myMarkovUpper["0","1"]
            colonizedLower = myMarkovLower["0","1"]
            colonizedSE = myMarkovSE["0","1"]
            result = c(colonized, colonizedUpper, colonizedLower, colonizedSE)
            
            #preserved = myMarkovMat["1","1"] 
            #relativeRisk = colonized/ preserved
            #return(relativeRisk)
            return(result)
          }
          
        }, simplify = F))
        
        rownames(scoresByMsp) = rownames(mgsMat)
        colnames(scoresByMsp) = c("inflow","upper","lower","SE")
        scoresByMsp = cbind(scoresByMsp, prevalence=prevalence)
        return(scoresByMsp)
        
      }
      getOutflow <- function(metaTab, mgsMat) {
        mgsMat = mgsMat[,colnames(mgsMat) %in% metaTab$sample.ID]
        prevalence = apply(mgsMat, 1, function(currRow)mean(currRow>0))
        uniqSubj = unique(metaTab$metadata.ID)
        scoresByMsp = do.call(rbind, sapply(rownames(mgsMat), function(currMgs){
          
          markovSeqOfMsp = (sapply(uniqSubj, function(subj) {
            currSamples = metaTab$sample.ID[metaTab$metadata.ID==subj]
            currVisits = metaTab$subtype[match(currSamples, metaTab$sample.ID)]
            currSamples = currSamples[order(currVisits)]
            currMgsMat = mgsMat[,currSamples]
            currMgsSubjVec = (currMgsMat[currMgs,] >0)*1
            currMgsSubjVec = as.character(currMgsSubjVec)
            
          }, simplify = F))
          
          myMarkovFit<-markovchainFit(data=markovSeqOfMsp,confidencelevel = .9,method = "mle")
          myMarkovMat = myMarkovFit$estimate@transitionMatrix
          myMarkovUpper = myMarkovFit$upperEndpointMatrix
          myMarkovLower = myMarkovFit$lowerEndpointMatrix
          myMarkovSE = myMarkovFit$standardError
          
          if(dim(myMarkovMat)[1]==1) {
            rname = rownames(myMarkovMat)
            #if (rname == "0") return(Inf)
            #if (rname == "1") return(0)
            if (rname == "0") {
              resistant = 1
              resistantUpper = NA
              resistantLower = NA
              resistantSE = NA
              result = c(resistant, resistantUpper, resistantLower, resistantSE)
              return(result) 
            }
            if (rname == "1") {
              resistant = 0
              resistantUpper = NA
              resistantLower = NA
              resistantSE = NA
              result = c(resistant, resistantUpper, resistantLower, resistantSE)
              return(result)
            }
          }
          else {
            resistant = myMarkovMat["1","0"] 
            resistantUpper = myMarkovUpper["1","0"] 
            resistantLower = myMarkovLower["1","0"] 
            resistantSE = myMarkovSE["1","0"]
            result = c(resistant, resistantUpper, resistantLower, resistantSE)
            
            #preserved = myMarkovMat["0","0"] 
            #relativeRisk = resistant/ preserved
            #return(relativeRisk)
            return(result)
            
          }
          
        }, simplify = F))
        rownames(scoresByMsp) = rownames(mgsMat)
        colnames(scoresByMsp) = c("outflow","upper","lower","SE")
        scoresByMsp = cbind(scoresByMsp, prevalence=prevalence)
        return(scoresByMsp)
        
      }
      
      inflowWellnessMat = getInflow(wellnessNewMetadata , wellnessNewMgsMatImputed)
      outflowWellnessMat = getOutflow(wellnessNewMetadata , wellnessNewMgsMatImputed)
      
      wellnessInflow = inflowWellnessMat[,1]
      wellnessOutflow = outflowWellnessMat[,1]
      wellnessPrevalence = inflowWellnessMat[,"prevalence"]
      
      wellnessInOutFlowTab = data.frame(inflow=wellnessInflow,
                                        outflow=wellnessOutflow,
                                        prevalence=wellnessPrevalence,
                                        species = getSpeciesName(names(wellnessInflow),gssOssTabNew) ) 
      
      wellnessInOutFlowTab$inflow.original = markovStatTabTemp[match(rownames(wellnessInOutFlowTab), rownames(markovStatTabTemp)),"inflow"]
      wellnessInOutFlowTab$outflow.original = markovStatTabTemp[match(rownames(wellnessInOutFlowTab), rownames(markovStatTabTemp)),"outflow"]
      wellnessInOutFlowTab$prevalence.original = markovStatTabTemp[match(rownames(wellnessInOutFlowTab), rownames(markovStatTabTemp)),"prevalence"]
      wellnessInOutFlowTab = wellnessInOutFlowTab[wellnessInOutFlowTab$prevalence >= 0.1 & wellnessInOutFlowTab$prevalence.original >= 0.1, ]
      
      cor.test(wellnessInOutFlowTab$inflow, wellnessInOutFlowTab$inflow.original, method="spearman" )
      cor.test(wellnessInOutFlowTab$outflow, wellnessInOutFlowTab$inflow.original, method="spearman" )
      
      
      
      ggplot(wellnessInOutFlowTab, aes(x=inflow, y=inflow.original)) + 
        geom_point(size=3, shape=16, colour="#00000055") + xlab("") + ylab("")+ #size=3, 
        #scale_fill_gradient2(midpoint = 0.2, low="#FFD70044", mid = "white", high = "purple") +
        geom_smooth(method = "lm",  se = T, colour="red")+
        scale_x_continuous(breaks = scales::pretty_breaks(n = 2)) +
        scale_y_continuous(breaks = scales::pretty_breaks(n = 2)) + 
        
        #geom_hline(yintercept=6)+
        #geom_abline(slope=1, intercept = 0)+
        theme(panel.grid.major = element_blank(), 
              panel.grid.minor = element_blank(),
              legend.position = "none",
              text = element_text(size = 15),
              panel.background = element_rect(fill = "white", 
                                              colour = "black", 
                                              size = 0.5, 
                                              linetype = "solid")) 
      ggplot(wellnessInOutFlowTab, aes(x=outflow, y=outflow.original)) + 
        geom_point(size=3, shape=16, colour="#00000055") + xlab("") + ylab("")+ #size=3, 
        #scale_fill_gradient2(midpoint = 0.2, low="#FFD70044", mid = "white", high = "purple") +
        geom_smooth(method = "lm",  se = T, colour="red")+
        scale_x_continuous(breaks = scales::pretty_breaks(n = 2)) +
        scale_y_continuous(breaks = scales::pretty_breaks(n = 2)) + 
        
        #geom_hline(yintercept=6)+
        #geom_abline(slope=1, intercept = 0)+
        theme(panel.grid.major = element_blank(), 
              panel.grid.minor = element_blank(),
              legend.position = "none",
              text = element_text(size = 15),
              panel.background = element_rect(fill = "white", 
                                              colour = "black", 
                                              size = 0.5, 
                                              linetype = "solid")) 
      
    }
    
    
    #wellnessMetadataV4 = wellnessMetadata[wellnessMetadata$metadata.ID %in% wellnessNewMetadata$metadata.ID &  wellnessMetadata$subtype %in% c("v1","v2","v3","v4") ,]
    wellnessMetadataV4 = wellnessMetadata[wellnessMetadata$metadata.ID %in% wellnessNewMetadata$metadata.ID &  wellnessMetadata$subtype %in% c( "v4") ,]
    wellnessMetadataV4 = wellnessMetadataV4[,c("sample.ID","metadata.ID","type","subtype","Gender","Age")]
    wellnessNewMetadata = wellnessNewMetadata[wellnessNewMetadata$metadata.ID %in% wellnessMetadataV4$metadata.ID,  ]
    wellnessNewMetadata = rbind(wellnessMetadataV4,wellnessNewMetadata)
    
    reloadMgsMatImputed = T
    if (reloadMgsMatImputed) {
      
      ######
      mgsMat = mergeMatUpdated
      metaTab = basicMetaMapUpdated
      
      wellnessSubjs = (wellnessMetadata$metadata.ID)
      missingSubjs1 = names(table(wellnessSubjs)[table(wellnessSubjs)==3])
      exceptionSubjs = c("X3452","X3834","X3918","X4215")
      missingSubjs2 = c("X3913","X3535","X3016")
      missingSubjs1 = missingSubjs1[!missingSubjs1 %in% exceptionSubjs]
      
      wellnessMetadata = wellnessMetadata[!wellnessMetadata$metadata.ID %in% missingSubjs1, ]
      wellnessMetadata = wellnessMetadata[!wellnessMetadata$metadata.ID %in% missingSubjs2, ]
      wellnessMetadata = wellnessMetadata[!wellnessMetadata$metadata.ID %in% exceptionSubjs, ]
      wellnessMgsMat = mergeMatUpdatedImputed[,wellnessMetadata$sample.ID]
      #wellnessMgsMat = mergeMatUpdated[,wellnessMetadata$sample.ID]
      
      wellnessSamples = wellnessMetadata$sample.ID
      wellnessUniqSubjs = unique(wellnessMetadata$metadata.ID)
      
      
      
    }
    
    wellnessMgsMat2 = wellnessMgsMat[,colnames(wellnessMgsMat) %in% wellnessNewMetadata$sample.ID ]
    wellnessSpeciesOverlap = rownames(wellnessMgsMat2)[rownames(wellnessMgsMat2) %in% rownames(wellnessNewMgsMatImputed)]
    wellnessMgsMat2 = wellnessMgsMat2[match(wellnessSpeciesOverlap, rownames(wellnessMgsMat2)),]
    wellnessNewMgsMatImputed = wellnessNewMgsMatImputed[match(wellnessSpeciesOverlap, rownames(wellnessNewMgsMatImputed)),]
    
    
    wellnessNewMgsMatImputed = cbind(wellnessMgsMat2,wellnessNewMgsMatImputed)
    wellnessNewMgsMatImputed = wellnessNewMgsMatImputed[,!is.na(colnames(wellnessNewMgsMatImputed))]
    
    wellnessNewMetadata = wellnessNewMetadata[wellnessNewMetadata$sample.ID %in% colnames(wellnessNewMgsMatImputed),]
    wellnessNewMgsMatImputed = wellnessNewMgsMatImputed[,colnames(wellnessNewMgsMatImputed) %in% wellnessNewMetadata$sample.ID]
    
    if (F) {
      qCutOff =(2^(4.042488))/10^9
      min(mergeMatUpdated[mergeMatUpdated!=0])
      
      mergeMatUpdatedImputed = mergeMatUpdated
      mergeMatUpdatedImputed[mergeMatUpdatedImputed<qCutOff] <- 0
      
    }
    
    getSurvMode = T
    if (getSurvMode) {
      getPairSeqs <- function(lenSeqs) {
        lenInds = lenSeqs-1
        pairList = sapply(1:lenInds, function(x) c(x, x+1), simplify = F)
        pairTab = do.call(rbind, pairList)
        return(pairTab)
      }
      getSurvVecByMgs <- function(mgsVec, mgsName, sampleNames, interval=3) {
        
        lenAll = length(mgsVec)
        lenInd = length(mgsVec)-1
        seqs = 1:lenInd
        #interval=3
        survTime = 0
        survEvent = "None"
        
        #####
        #check fist or second visit whether mgs exists
        
        mgsCutDigit <- mgsVec > 0
        presentInds = which(mgsCutDigit)
        if (length(presentInds)==0) {
          survTime = NA
          survEvent = "None"
          return(list(time=survTime, event=survEvent, sample=sampleNames[1], abd = 0, mgs=mgsName))
        }
        startInd = min(presentInds)
        lastInd = startInd
        for (ind in seq_along(presentInds)) {
          if (ind == length(presentInds)) break
          currInd = presentInds[ind]
          nextInd = presentInds[ind+1]
          diff = nextInd - currInd
          if (diff == 1) lastInd = nextInd
        }
        
        if (length(presentInds)!= 0) {
          survTime= (lastInd+1 - startInd)*interval
          survEvent = "Absent"
          if (lastInd == lenAll) {
            survTime = interval*lenInd
            survEvent = "None"
          }
        }
        res=list(time=survTime, event=survEvent, sample=sampleNames[startInd], abd = mgsVec[startInd], mgs=mgsName)
        #print(res)
        return(res)
        
      }
      getSurvivalRates <- function(metaTab, mgsMat, interval=3, halfTime=6) {
        
        uniqSubjs = unique(metaTab$metadata.ID)
        mgsPrevalence = apply(mgsMat, 1, function(currRow){
          mean(currRow>0)
        })
        mgsMeans = rowMeans(mgsMat)
        
        survList = do.call(c,sapply(uniqSubjs, function(subj) {
          currSamples = metaTab$sample.ID[metaTab$metadata.ID==subj]
          currSamples = currSamples[order(currSamples)]
          currMgsMat = mgsMat[,currSamples]
          currMgsMat = currMgsMat[rowMeans(currMgsMat)!=0,]
          currSubSurvVecThroughs = (sapply(rownames(currMgsMat), function(currMgs){
            currSubSurvVecThrough = getSurvVecByMgs(currMgsMat[currMgs,], currMgs, currSamples, interval)
            return(currSubSurvVecThrough)
          }, simplify = F))
          
          return(currSubSurvVecThroughs) 
          
        }, simplify = F))
        
        survTab = data.frame(time=do.call(c,lapply(survList, function(x) x$time)),
                             status=do.call(c,lapply(survList, function(x) x$event)),
                             sample=do.call(c,lapply(survList, function(x) x$sample)),
                             abd=do.call(c,lapply(survList, function(x) x$abd)),
                             mgs=do.call(c,lapply(survList, function(x) x$mgs)),
                             stringsAsFactors = F)
        survTab$logAbd = log10(survTab$abd)
        survivalRates = unlist(sapply(unique(survTab$mgs), function(currMsp) {
          currMspTab = survTab[survTab$mgs == currMsp,]
          if (F) {
            library("survminer")
            currMsp = "msp_0069"
            currMsp = "msp_0041"
            currMsp = "msp_0005"
            currMsp = "msp_0148c"
            currMsp = "msp_0742"
            
            
            currMspTab = survTab[survTab$mgs == currMsp,]
            currMspTab$status <- currMspTab$status == "Absent"
            currSurvFit=survfit(survival::Surv(time, status) ~ 1, data=currMspTab)
            ggsurvplot(currSurvFit, break.time.by = 3, color = "#2E9FDF")
            
          }
          halfTimeSummary = summary(survfit(survival::Surv(currMspTab$time, currMspTab$status == "Absent") ~ 1), times=halfTime)
          halfTimeSurvival = halfTimeSummary$surv
          return(halfTimeSurvival)
        }))
        
        survivalStats = data.frame(surv.rates=survivalRates[match(rownames(mgsMat), names(survivalRates))],
                                   mgs.abd = mgsMeans[match(rownames(mgsMat), names(mgsMeans))],
                                   freq= mgsPrevalence,
                                   mgs=rownames(mgsMat),
                                   species=getSpeciesName(rownames(mgsMat),taxo))
        
        return(survivalStats)
      }
      
      #survivalStats = getSurvivalRates(metaTab, mgsMat, decreaseCut)
      survivalStats_wellness_new = getSurvivalRates(wellnessNewMetadata, wellnessNewMgsMatImputed, interval=6, halfTime = 6)
    }
    
    getInflow <- function(metaTab, mgsMat) {
      mgsMat = mgsMat[,colnames(mgsMat) %in% metaTab$sample.ID]
      prevalence = apply(mgsMat, 1, function(currRow)mean(currRow>0))
      uniqSubj = unique(metaTab$metadata.ID)
      scoresByMsp = do.call(rbind, sapply(rownames(mgsMat), function(currMgs){
        
        markovSeqOfMsp = (sapply(uniqSubj, function(subj) {
          currSamples = metaTab$sample.ID[metaTab$metadata.ID==subj]
          currVisits = metaTab$subtype[match(currSamples, metaTab$sample.ID)]
          currSamples = currSamples[order(currVisits)]
          currMgsMat = mgsMat[,currSamples]
          currMgsSubjVec = (currMgsMat[currMgs,] >0)*1
          currMgsSubjVec = as.character(currMgsSubjVec)
          
        }, simplify = F))
        
        myMarkovFit<-markovchainFit(data=markovSeqOfMsp,confidencelevel = .9,method = "mle")
        myMarkovMat = myMarkovFit$estimate@transitionMatrix
        myMarkovUpper = myMarkovFit$upperEndpointMatrix
        myMarkovLower = myMarkovFit$lowerEndpointMatrix
        myMarkovSE = myMarkovFit$standardError
        
        if(dim(myMarkovMat)[1]==1) {
          rname = rownames(myMarkovMat)
          #if (rname == "0") return(0)
          #if (rname == "1") return(Inf)
          if (rname == "0") {
            colonized = 0
            colonizedUpper = NA
            colonizedLower = NA
            colonizedSE = NA
            result = c(colonized, colonizedUpper, colonizedLower, colonizedSE)
            return(result)
          }
          
          if (rname == "1") {
            colonized = 1
            colonizedUpper = NA
            colonizedLower = NA
            colonizedSE = NA
            result = c(colonized, colonizedUpper, colonizedLower, colonizedSE)
            
            return(result)
          }
        }
        else {
          colonized = myMarkovMat["0","1"] 
          colonizedUpper = myMarkovUpper["0","1"]
          colonizedLower = myMarkovLower["0","1"]
          colonizedSE = myMarkovSE["0","1"]
          result = c(colonized, colonizedUpper, colonizedLower, colonizedSE)
          
          #preserved = myMarkovMat["1","1"] 
          #relativeRisk = colonized/ preserved
          #return(relativeRisk)
          return(result)
        }
        
      }, simplify = F))
      
      rownames(scoresByMsp) = rownames(mgsMat)
      colnames(scoresByMsp) = c("inflow","upper","lower","SE")
      scoresByMsp = cbind(scoresByMsp, prevalence=prevalence)
      return(scoresByMsp)
      
    }
    getOutflow <- function(metaTab, mgsMat) {
      mgsMat = mgsMat[,colnames(mgsMat) %in% metaTab$sample.ID]
      prevalence = apply(mgsMat, 1, function(currRow)mean(currRow>0))
      uniqSubj = unique(metaTab$metadata.ID)
      scoresByMsp = do.call(rbind, sapply(rownames(mgsMat), function(currMgs){
        
        markovSeqOfMsp = (sapply(uniqSubj, function(subj) {
          currSamples = metaTab$sample.ID[metaTab$metadata.ID==subj]
          currVisits = metaTab$subtype[match(currSamples, metaTab$sample.ID)]
          currSamples = currSamples[order(currVisits)]
          currMgsMat = mgsMat[,currSamples]
          currMgsSubjVec = (currMgsMat[currMgs,] >0)*1
          currMgsSubjVec = as.character(currMgsSubjVec)
          
        }, simplify = F))
        
        myMarkovFit<-markovchainFit(data=markovSeqOfMsp,confidencelevel = .9,method = "mle")
        myMarkovMat = myMarkovFit$estimate@transitionMatrix
        myMarkovUpper = myMarkovFit$upperEndpointMatrix
        myMarkovLower = myMarkovFit$lowerEndpointMatrix
        myMarkovSE = myMarkovFit$standardError
        
        if(dim(myMarkovMat)[1]==1) {
          rname = rownames(myMarkovMat)
          #if (rname == "0") return(Inf)
          #if (rname == "1") return(0)
          if (rname == "0") {
            resistant = 1
            resistantUpper = NA
            resistantLower = NA
            resistantSE = NA
            result = c(resistant, resistantUpper, resistantLower, resistantSE)
            return(result) 
          }
          if (rname == "1") {
            resistant = 0
            resistantUpper = NA
            resistantLower = NA
            resistantSE = NA
            result = c(resistant, resistantUpper, resistantLower, resistantSE)
            return(result)
          }
        }
        else {
          resistant = myMarkovMat["1","0"] 
          resistantUpper = myMarkovUpper["1","0"] 
          resistantLower = myMarkovLower["1","0"] 
          resistantSE = myMarkovSE["1","0"]
          result = c(resistant, resistantUpper, resistantLower, resistantSE)
          
          #preserved = myMarkovMat["0","0"] 
          #relativeRisk = resistant/ preserved
          #return(relativeRisk)
          return(result)
          
        }
        
      }, simplify = F))
      rownames(scoresByMsp) = rownames(mgsMat)
      colnames(scoresByMsp) = c("outflow","upper","lower","SE")
      scoresByMsp = cbind(scoresByMsp, prevalence=prevalence)
      return(scoresByMsp)
      
    }
    
    inflowWellnessMat_original = getInflow(wellnessNewMetadata[wellnessNewMetadata$subtype %in% c("v1","v2","v3","v4"),], wellnessNewMgsMatImputed)
    outflowWellnessMat_original = getOutflow(wellnessNewMetadata[wellnessNewMetadata$subtype %in% c("v1","v2","v3","v4"),], wellnessNewMgsMatImputed)
    
    #inflowWellnessMat = getInflow(wellnessNewMetadata, wellnessNewMgsMatImputed)
    #outflowWellnessMat = getOutflow(wellnessNewMetadata, wellnessNewMgsMatImputed)
    
    inflowWellnessMat = getInflow(wellnessNewMetadata[wellnessNewMetadata$subtype %in% c( "v4","v5","v6"),], wellnessNewMgsMatImputed)
    outflowWellnessMat = getOutflow(wellnessNewMetadata[wellnessNewMetadata$subtype %in% c( "v4","v5","v6"),], wellnessNewMgsMatImputed)
    
    wellnessInflow = inflowWellnessMat[,1]
    wellnessOutflow = outflowWellnessMat[,1]
    wellnessPrevalence = inflowWellnessMat[,"prevalence"]
    
    wellnessInOutFlowTab = data.frame(inflow=wellnessInflow,
                                      outflow=wellnessOutflow,
                                      prevalence=wellnessPrevalence,
                                      species = getSpeciesName(names(wellnessInflow),gssOssTabNew) ) 
    
    v5v6PlusMode = F
    if (v5v6PlusMode) {
      
      
      wellnessInOutFlowTab$inflow.original = markovStatTabTemp[match(rownames(wellnessInOutFlowTab), rownames(markovStatTabTemp)),"inflow"]
      wellnessInOutFlowTab$outflow.original = markovStatTabTemp[match(rownames(wellnessInOutFlowTab), rownames(markovStatTabTemp)),"outflow"]
      wellnessInOutFlowTab$prevalence.original = markovStatTabTemp[match(rownames(wellnessInOutFlowTab), rownames(markovStatTabTemp)),"prevalence"]
      wellnessInOutFlowTab = wellnessInOutFlowTab[wellnessInOutFlowTab$prevalence >= 0.1 & wellnessInOutFlowTab$prevalence.original >= 0.1, ]
      
      cor.test(wellnessInOutFlowTab$inflow, wellnessInOutFlowTab$inflow.original, method="spearman" )
      cor.test(wellnessInOutFlowTab$outflow, wellnessInOutFlowTab$outflow.original, method="spearman" )
      
      save(wellnessInOutFlowTab, file="wellnessInOutFlowTab.RData")
      
      ggplot(wellnessInOutFlowTab, aes(x=inflow, y=inflow.original)) + 
        geom_point(size=3, shape=16, colour="#00000055") + xlab("") + ylab("")+ #size=3, 
        #scale_fill_gradient2(midpoint = 0.2, low="#FFD70044", mid = "white", high = "purple") +
        geom_smooth(method = "lm",  se = T, colour="red")+
        scale_x_continuous(breaks = scales::pretty_breaks(n = 2)) +
        scale_y_continuous(breaks = scales::pretty_breaks(n = 2)) + 
        
        #geom_hline(yintercept=6)+
        #geom_abline(slope=1, intercept = 0)+
        theme(panel.grid.major = element_blank(), 
              panel.grid.minor = element_blank(),
              legend.position = "none",
              text = element_text(size = 15),
              panel.background = element_rect(fill = "white", 
                                              colour = "black", 
                                              size = 0.5, 
                                              linetype = "solid")) 
      
      
      ggplot(wellnessInOutFlowTab, aes(x=outflow, y=outflow.original)) + 
        geom_point(size=3, shape=16, colour="#00000055") + xlab("") + ylab("")+ #size=3, 
        #scale_fill_gradient2(midpoint = 0.2, low="#FFD70044", mid = "white", high = "purple") +
        geom_smooth(method = "lm",  se = T, colour="red")+
        scale_x_continuous(breaks = scales::pretty_breaks(n = 2)) +
        scale_y_continuous(breaks = scales::pretty_breaks(n = 2)) + 
        
        #geom_hline(yintercept=6)+
        #geom_abline(slope=1, intercept = 0)+
        theme(panel.grid.major = element_blank(), 
              panel.grid.minor = element_blank(),
              legend.position = "none",
              text = element_text(size = 15),
              panel.background = element_rect(fill = "white", 
                                              colour = "black", 
                                              size = 0.5, 
                                              linetype = "solid")) 
      
    }
    
    
    ggplot(wellnessInOutFlowTab, aes(x=outflow, y=inflow, fill=inflow)) + 
      scale_fill_gradient2(midpoint = 0.2, low="#FFD70044", mid = "white", high = "purple") +
      geom_point(size=3, shape=21, colour="#d3d3d366")+ xlab("") + ylab("")+
      geom_vline(xintercept = 0.3, colour="gray",linetype="dashed")+
      geom_hline(yintercept = 0.3, colour="gray",linetype="dashed")+
      #geom_abline(slope=1, intercept = 0)+
      theme(panel.grid.major = element_blank(), 
            panel.grid.minor = element_blank(),
            legend.position = "none",
            text = element_text(size = 15),
            panel.background = element_rect(fill = "white", 
                                            colour = "black", 
                                            size = 0.5, 
                                            linetype = "solid"))
    
    wellnessInflowSpecies = names(wellnessInflow[wellnessInflow>=0.3])
    wellnessOutflowSpecies = names(wellnessOutflow[wellnessOutflow>=0.3])
    
    wellnessSharedpecies = wellnessInflowSpecies[wellnessInflowSpecies  %in% wellnessOutflowSpecies]
    
    wellnessInflowSpecies = wellnessInflowSpecies[!wellnessInflowSpecies %in% wellnessSharedpecies]
    wellnessOutflowSpecies = wellnessOutflowSpecies[!wellnessOutflowSpecies %in% wellnessSharedpecies]
    
    table(inflowSpecies %in% wellnessInflowSpecies)
    table(inflowSpecies %in% wellnessOutflowSpecies)
    
    table(outflowSpecies %in% wellnessInflowSpecies)
    table(outflowSpecies %in% wellnessOutflowSpecies)
    
    #################################
    #### comparing before/after #####
    #################################
    
    
    #overlapped = rownames(wellnessInOutFlowTab)[rownames(wellnessInOutFlowTab) %in% rownames(markovStatTabTemp)]
    
    #plot(markovStatTabTemp$inflow[match(overlapped, rownames(markovStatTabTemp))], wellnessInOutFlowTab$inflow[match(overlapped, rownames(wellnessInOutFlowTab))])
    
    
    wellnessNewTabMerged = data.frame(inflow.new.visits = inflowWellnessMat[,"inflow"],
                                      inflow.four.visits = inflowWellnessMat_original[,"inflow"],
                                      outflow.new.visits = outflowWellnessMat[,"outflow"],
                                      outflow.four.visits = outflowWellnessMat_original[,"outflow"],
                                      prevalence.new.visits = inflowWellnessMat[,"prevalence"],
                                      prevalence.four.visits = inflowWellnessMat[,"prevalence"],
                                      species = getSpeciesName(rownames(inflowWellnessMat), gssOssTabNew))
    
    
    wellnessNewTabMerged = wellnessNewTabMerged[wellnessNewTabMerged$prevalence.new.visits >= 0.1 & wellnessNewTabMerged$prevalence.four.visits >= 0.1, ]
    
    cor.test(wellnessNewTabMerged$inflow.new.visits, wellnessNewTabMerged$inflow.four.visits, method="spearman")
    cor.test(wellnessNewTabMerged$outflow.new.visits, wellnessNewTabMerged$outflow.four.visits, method="spearman")
    
    
    ggplot(wellnessNewTabMerged, aes(x=inflow.new.visits, y=inflow.four.visits)) + 
      geom_point(size=3, shape=16, colour="#00000055") + xlab("") + ylab("")+ #size=3, 
      #scale_fill_gradient2(midpoint = 0.2, low="#FFD70044", mid = "white", high = "purple") +
      geom_smooth(method = "lm",  se = T, colour="red")+
      scale_x_continuous(breaks = scales::pretty_breaks(n = 2)) +
      scale_y_continuous(breaks = scales::pretty_breaks(n = 2)) +
      
      #geom_hline(yintercept=6)+
      #geom_abline(slope=1, intercept = 0)+
      theme(panel.grid.major = element_blank(), 
            panel.grid.minor = element_blank(),
            legend.position = "none",
            text = element_text(size = 15),
            panel.background = element_rect(fill = "white", 
                                            colour = "black", 
                                            size = 0.5, 
                                            linetype = "solid"))
    
    ggplot(wellnessNewTabMerged, aes(x=outflow.new.visits, y=outflow.four.visits)) + 
      geom_point(size=3, shape=16, colour="#00000055") + xlab("") + ylab("")+ #size=3, 
      #scale_fill_gradient2(midpoint = 0.2, low="#FFD70044", mid = "white", high = "purple") +
      geom_smooth(method = "lm",  se = T, colour="red")+
      scale_x_continuous(breaks = scales::pretty_breaks(n = 2)) +
      scale_y_continuous(breaks = scales::pretty_breaks(n = 2)) +
      
      #geom_hline(yintercept=6)+
      #geom_abline(slope=1, intercept = 0)+
      theme(panel.grid.major = element_blank(), 
            panel.grid.minor = element_blank(),
            legend.position = "none",
            text = element_text(size = 15),
            panel.background = element_rect(fill = "white", 
                                            colour = "black", 
                                            size = 0.5, 
                                            linetype = "solid"))
    
    
  }
  
  newMarkovMode = T
  if (newMarkovMode) {
    getColonizationColonizingProbsNew <- function(metaTab, mgsMat) {
      if (F) {
        wellnessMetadata = metaTab[metaTab$dataset.ID=="id28",]
        wellnessSubjs = (wellnessMetadata$metadata.ID)
        missingSubjs1 = names(table(wellnessSubjs)[table(wellnessSubjs)==3])
        exceptionSubjs = c("X3452","X3834","X3918","X4215")
        missingSubjs2 = c("X3913","X3535","X3016")
        missingSubjs1 = missingSubjs1[!missingSubjs1 %in% exceptionSubjs]
        
        wellnessMetadata = wellnessMetadata[!wellnessMetadata$metadata.ID %in% missingSubjs1, ]
        wellnessMetadata = wellnessMetadata[!wellnessMetadata$metadata.ID %in% missingSubjs2, ]
        wellnessMgsMat = mergeMatUpdated[,wellnessMetadata$sample.ID]
        wellnessMgsMeans = rowMeans(wellnessMgsMat)
        wellnessSamples = wellnessMetadata$sample.ID
        wellnessUniqSubjs = unique(wellnessMetadata$metadata.ID)
        
      }
      if (T) {
        wellnessMetadata = basicMetaMapUpdated[basicMetaMapUpdated$dataset.ID=="id28",]
        
        #mgsMat = mergeMatUpdated
        #metaTab = basicMetaMapUpdated
        
        wellnessMetadata = metaTab[metaTab$dataset.ID=="id28",]
        wellnessSubjs = (wellnessMetadata$metadata.ID)
        missingSubjs1 = names(table(wellnessSubjs)[table(wellnessSubjs)==3])
        exceptionSubjs = c("X3452","X3834","X3918","X4215")
        missingSubjs2 = c("X3913","X3535","X3016")
        missingSubjs1 = missingSubjs1[!missingSubjs1 %in% exceptionSubjs]
        
        wellnessMetadata = wellnessMetadata[!wellnessMetadata$metadata.ID %in% missingSubjs1, ]
        wellnessMetadata = wellnessMetadata[!wellnessMetadata$metadata.ID %in% missingSubjs2, ]
        wellnessMetadata = wellnessMetadata[!wellnessMetadata$metadata.ID %in% exceptionSubjs, ]
        #wellnessMgsMat = mergeMatUpdated[,match(wellnessMetadata$sample.ID, colnames(mergeMatUpdated))]
        wellnessMgsMat = mgsMat
        wellnessSamples = wellnessMetadata$sample.ID
        wellnessSubjects = gsub("_v[1-4]","",wellnessSamples)
        wellnessVisits = gsub("X[0-9]+_","",wellnessSamples)
        wellnessUniqSubjs = unique(wellnessMetadata$metadata.ID)
        
      }
      
      
      scoresByMsp = do.call(rbind, sapply(rownames(mgsMat), function(currMgs){
        
        markovSeqOfMsp = (sapply(wellnessUniqSubjs, function(subj) {
          currSamples = wellnessMetadata$sample.ID[wellnessMetadata$metadata.ID==subj]
          currSamples = currSamples[order(currSamples)]
          currMgsMat = wellnessMgsMat[,currSamples]
          currMgsSubjVec = (currMgsMat[currMgs,] >0)*1
          currMgsSubjVec = as.character(currMgsSubjVec)
          
        }, simplify = F))
        
        myMarkovFit<-markovchainFit(data=markovSeqOfMsp,confidencelevel = .9,method = "mle")
        myMarkovMat = myMarkovFit$estimate@transitionMatrix
        myMarkovUpper = myMarkovFit$upperEndpointMatrix
        myMarkovLower = myMarkovFit$lowerEndpointMatrix
        myMarkovSE = myMarkovFit$standardError
        
        if(dim(myMarkovMat)[1]==1) {
          rname = rownames(myMarkovMat)
          #if (rname == "0") return(0)
          #if (rname == "1") return(Inf)
          if (rname == "0") {
            colonized = 0
            colonizedUpper = NA
            colonizedLower = NA
            colonizedSE = NA
            result = c(colonized, colonizedUpper, colonizedLower, colonizedSE)
            return(result)
          }
          
          if (rname == "1") {
            colonized = 1
            colonizedUpper = NA
            colonizedLower = NA
            colonizedSE = NA
            result = c(colonized, colonizedUpper, colonizedLower, colonizedSE)
            
            return(result)
          }
        }
        else {
          colonized = myMarkovMat["0","1"] 
          colonizedUpper = myMarkovUpper["0","1"]
          colonizedLower = myMarkovLower["0","1"]
          colonizedSE = myMarkovSE["0","1"]
          result = c(colonized, colonizedUpper, colonizedLower, colonizedSE)
          
          #preserved = myMarkovMat["1","1"] 
          #relativeRisk = colonized/ preserved
          #return(relativeRisk)
          return(result)
        }
        
      }, simplify = F))
      
      rownames(scoresByMsp) = rownames(mgsMat)
      colnames(scoresByMsp) = c("inflow","upper","lower","SE")
      return(scoresByMsp)
      
    }
    getColonizationResistantProbsNew <- function(metaTab, mgsMat) {
      if (F) {
        wellnessMetadata = metaTab[metaTab$dataset.ID=="id28",]
        wellnessSubjs = (wellnessMetadata$metadata.ID)
        missingSubjs1 = names(table(wellnessSubjs)[table(wellnessSubjs)==3])
        exceptionSubjs = c("X3452","X3834","X3918","X4215")
        missingSubjs2 = c("X3913","X3535","X3016")
        missingSubjs1 = missingSubjs1[!missingSubjs1 %in% exceptionSubjs]
        
        wellnessMetadata = wellnessMetadata[!wellnessMetadata$metadata.ID %in% missingSubjs1, ]
        wellnessMetadata = wellnessMetadata[!wellnessMetadata$metadata.ID %in% missingSubjs2, ]
        wellnessMgsMat = mergeMatUpdated[,wellnessMetadata$sample.ID]
        wellnessMgsMeans = rowMeans(wellnessMgsMat)
        wellnessSamples = wellnessMetadata$sample.ID
        wellnessUniqSubjs = unique(wellnessMetadata$metadata.ID)
        
      }
      if (T) {
        wellnessMetadata = basicMetaMapUpdated[basicMetaMapUpdated$dataset.ID=="id28",]
        
        #mgsMat = mergeMatUpdated
        #metaTab = basicMetaMapUpdated
        
        wellnessMetadata = metaTab[metaTab$dataset.ID=="id28",]
        wellnessSubjs = (wellnessMetadata$metadata.ID)
        missingSubjs1 = names(table(wellnessSubjs)[table(wellnessSubjs)==3])
        exceptionSubjs = c("X3452","X3834","X3918","X4215")
        missingSubjs2 = c("X3913","X3535","X3016")
        missingSubjs1 = missingSubjs1[!missingSubjs1 %in% exceptionSubjs]
        
        wellnessMetadata = wellnessMetadata[!wellnessMetadata$metadata.ID %in% missingSubjs1, ]
        wellnessMetadata = wellnessMetadata[!wellnessMetadata$metadata.ID %in% missingSubjs2, ]
        wellnessMetadata = wellnessMetadata[!wellnessMetadata$metadata.ID %in% exceptionSubjs, ]
        wellnessMgsMat = mergeMatUpdated[,match(wellnessMetadata$sample.ID, colnames(mergeMatUpdated))]
        wellnessMgsMat = mgsMat
        wellnessSamples = wellnessMetadata$sample.ID
        wellnessSubjects = gsub("_v[1-4]","",wellnessSamples)
        wellnessVisits = gsub("X[0-9]+_","",wellnessSamples)
        wellnessUniqSubjs = unique(wellnessMetadata$metadata.ID)
        
      }
      
      scoresByMsp = do.call(rbind, sapply(rownames(mgsMat), function(currMgs){
        
        markovSeqOfMsp = (sapply(wellnessUniqSubjs, function(subj) {
          currSamples = wellnessMetadata$sample.ID[wellnessMetadata$metadata.ID==subj]
          currSamples = currSamples[order(currSamples)]
          currMgsMat = wellnessMgsMat[,currSamples]
          currMgsSubjVec = (currMgsMat[currMgs,] >0)*1
          currMgsSubjVec = as.character(currMgsSubjVec)
          
        }, simplify = F))
        
        myMarkovFit<-markovchainFit(data=markovSeqOfMsp,confidencelevel = .9,method = "mle")
        myMarkovMat = myMarkovFit$estimate@transitionMatrix
        myMarkovUpper = myMarkovFit$upperEndpointMatrix
        myMarkovLower = myMarkovFit$lowerEndpointMatrix
        myMarkovSE = myMarkovFit$standardError
        
        if(dim(myMarkovMat)[1]==1) {
          rname = rownames(myMarkovMat)
          #if (rname == "0") return(Inf)
          #if (rname == "1") return(0)
          if (rname == "0") {
            resistant = 1
            resistantUpper = NA
            resistantLower = NA
            resistantSE = NA
            result = c(resistant, resistantUpper, resistantLower, resistantSE)
            return(result) 
          }
          if (rname == "1") {
            resistant = 0
            resistantUpper = NA
            resistantLower = NA
            resistantSE = NA
            result = c(resistant, resistantUpper, resistantLower, resistantSE)
            return(result)
          }
        }
        else {
          resistant = myMarkovMat["1","0"] 
          resistantUpper = myMarkovUpper["1","0"] 
          resistantLower = myMarkovLower["1","0"] 
          resistantSE = myMarkovSE["1","0"]
          result = c(resistant, resistantUpper, resistantLower, resistantSE)
          
          #preserved = myMarkovMat["0","0"] 
          #relativeRisk = resistant/ preserved
          #return(relativeRisk)
          return(result)
          
        }
        
      }, simplify = F))
      rownames(scoresByMsp) = rownames(mgsMat)
      colnames(scoresByMsp) = c("outflow","upper","lower","SE")
      return(scoresByMsp)
      
    }
    getColonizationSteadyStates <- function(metaTab, mgsMat) {
      if (F) {
        wellnessMetadata = metaTab[metaTab$dataset.ID=="id28",]
        wellnessSubjs = (wellnessMetadata$metadata.ID)
        missingSubjs1 = names(table(wellnessSubjs)[table(wellnessSubjs)==3])
        exceptionSubjs = c("X3452","X3834","X3918","X4215")
        missingSubjs2 = c("X3913","X3535","X3016")
        missingSubjs1 = missingSubjs1[!missingSubjs1 %in% exceptionSubjs]
        
        wellnessMetadata = wellnessMetadata[!wellnessMetadata$metadata.ID %in% missingSubjs1, ]
        wellnessMetadata = wellnessMetadata[!wellnessMetadata$metadata.ID %in% missingSubjs2, ]
        wellnessMgsMat = mergeMatUpdated[,wellnessMetadata$sample.ID]
        wellnessMgsMeans = rowMeans(wellnessMgsMat)
        wellnessSamples = wellnessMetadata$sample.ID
        wellnessUniqSubjs = unique(wellnessMetadata$metadata.ID)
        
      }
      if (T) {
        wellnessMetadata = basicMetaMapUpdated[basicMetaMapUpdated$dataset.ID=="id28",]
        
        #mgsMat = mergeMatUpdated
        #metaTab = basicMetaMapUpdated
        
        wellnessMetadata = metaTab[metaTab$dataset.ID=="id28",]
        wellnessSubjs = (wellnessMetadata$metadata.ID)
        missingSubjs1 = names(table(wellnessSubjs)[table(wellnessSubjs)==3])
        exceptionSubjs = c("X3452","X3834","X3918","X4215")
        missingSubjs2 = c("X3913","X3535","X3016")
        missingSubjs1 = missingSubjs1[!missingSubjs1 %in% exceptionSubjs]
        
        wellnessMetadata = wellnessMetadata[!wellnessMetadata$metadata.ID %in% missingSubjs1, ]
        wellnessMetadata = wellnessMetadata[!wellnessMetadata$metadata.ID %in% missingSubjs2, ]
        wellnessMetadata = wellnessMetadata[!wellnessMetadata$metadata.ID %in% exceptionSubjs, ]
        #wellnessMgsMat = mergeMatUpdated[,match(wellnessMetadata$sample.ID, colnames(mergeMatUpdated))]
        wellnessMgsMat = mgsMat
        wellnessSamples = wellnessMetadata$sample.ID
        wellnessSubjects = gsub("_v[1-4]","",wellnessSamples)
        wellnessVisits = gsub("X[0-9]+_","",wellnessSamples)
        wellnessUniqSubjs = unique(wellnessMetadata$metadata.ID)
        
      }
      
      stateMat = do.call(rbind, sapply(rownames(mgsMat), function(currMgs){
        
        markovSeqOfMsp = (sapply(wellnessUniqSubjs, function(subj) {
          currSamples = wellnessMetadata$sample.ID[wellnessMetadata$metadata.ID==subj]
          currSamples = currSamples[order(currSamples)]
          currMgsMat = wellnessMgsMat[,currSamples]
          currMgsSubjVec = (currMgsMat[currMgs,] >0)*1
          currMgsSubjVec = as.character(currMgsSubjVec)
          
        }, simplify = F))
        
        myMarkovFit<-markovchainFit(data=markovSeqOfMsp, confidencelevel = .9,method = "mle")
        tMatrix = myMarkovFit$estimate@transitionMatrix
        
        #print(currMgs)
        if (dim(tMatrix)[1]==1) {
          if(rownames(tMatrix)=="1")
            return(c("0"=0,"1"=tMatrix[1,1]))
          if(rownames(tMatrix)=="0")
            return(c("0"=tMatrix[1,1],"1"=0))
        }
        
        if (sum(tMatrix[2,])==0 ) {
          if (tMatrix[1,1]>tMatrix[1,2])
            return(c("0"=0,"1"=0))
          if (tMatrix[1,1]<tMatrix[1,2])
            return(c("0"=1,"1"=0))
          if (tMatrix[1,1]==tMatrix[1,2])
            return(c("0"=0.5,"1"=0))
        }
        mcData = new("markovchain", 
                     states = c("0","1"), 
                     transitionMatrix=tMatrix, 
                     name="test")
        ss = steadyStates(mcData)
        #print(ss)
        return(ss)
        
      }, simplify = F))
      rownames(stateMat) = rownames(mgsMat)
      return(stateMat)
      
    }
    getColonizationStationary <- function(metaTab, mgsMat) {
      if (F) {
        wellnessMetadata = metaTab[metaTab$dataset.ID=="id28",]
        wellnessSubjs = (wellnessMetadata$metadata.ID)
        missingSubjs1 = names(table(wellnessSubjs)[table(wellnessSubjs)==3])
        exceptionSubjs = c("X3452","X3834","X3918","X4215")
        missingSubjs2 = c("X3913","X3535","X3016")
        missingSubjs1 = missingSubjs1[!missingSubjs1 %in% exceptionSubjs]
        
        wellnessMetadata = wellnessMetadata[!wellnessMetadata$metadata.ID %in% missingSubjs1, ]
        wellnessMetadata = wellnessMetadata[!wellnessMetadata$metadata.ID %in% missingSubjs2, ]
        wellnessMgsMat = mergeMatUpdated[,wellnessMetadata$sample.ID]
        wellnessMgsMeans = rowMeans(wellnessMgsMat)
        wellnessSamples = wellnessMetadata$sample.ID
        wellnessUniqSubjs = unique(wellnessMetadata$metadata.ID)
        
      }
      if (T) {
        wellnessMetadata = basicMetaMapUpdated[basicMetaMapUpdated$dataset.ID=="id28",]
        
        #mgsMat = mergeMatUpdated
        #metaTab = basicMetaMapUpdated
        
        wellnessMetadata = metaTab[metaTab$dataset.ID=="id28",]
        wellnessSubjs = (wellnessMetadata$metadata.ID)
        missingSubjs1 = names(table(wellnessSubjs)[table(wellnessSubjs)==3])
        exceptionSubjs = c("X3452","X3834","X3918","X4215")
        missingSubjs2 = c("X3913","X3535","X3016")
        missingSubjs1 = missingSubjs1[!missingSubjs1 %in% exceptionSubjs]
        
        wellnessMetadata = wellnessMetadata[!wellnessMetadata$metadata.ID %in% missingSubjs1, ]
        wellnessMetadata = wellnessMetadata[!wellnessMetadata$metadata.ID %in% missingSubjs2, ]
        wellnessMetadata = wellnessMetadata[!wellnessMetadata$metadata.ID %in% exceptionSubjs, ]
        #wellnessMgsMat = mergeMatUpdated[,match(wellnessMetadata$sample.ID, colnames(mergeMatUpdated))]
        wellnessMgsMat = mgsMat
        wellnessSamples = wellnessMetadata$sample.ID
        wellnessSubjects = gsub("_v[1-4]","",wellnessSamples)
        wellnessVisits = gsub("X[0-9]+_","",wellnessSamples)
        wellnessUniqSubjs = unique(wellnessMetadata$metadata.ID)
        
      }
      
      stateVecs = do.call(c, sapply(rownames(wellnessMgsMat), function(currMgs){
        
        markovSeqOfMsp = (sapply(wellnessUniqSubjs, function(subj) {
          currSamples = wellnessMetadata$sample.ID[wellnessMetadata$metadata.ID==subj]
          currSamples = currSamples[order(currSamples)]
          currMgsMat = wellnessMgsMat[,currSamples]
          currMgsSubjVec = (currMgsMat[currMgs,] >0)*1
          currMgsSubjVec = as.character(currMgsSubjVec)
          
        }, simplify = F))
        
        stationarity = assessStationarity(markovSeqOfMsp, 10)
        return(stationarity$p.value)
        
        
      }, simplify = F))
      names(stateVecs) = rownames(mgsMat)
      return(stateVecs)
      
    }
    getFullVisitCounts <- function(metaTab, mgsMat) {
      
      if (T) {
        wellnessMetadata = basicMetaMapUpdated[basicMetaMapUpdated$dataset.ID=="id28",]
        
        #mgsMat = mergeMatUpdated
        #metaTab = basicMetaMapUpdated
        
        wellnessMetadata = metaTab[metaTab$dataset.ID=="id28",]
        wellnessSubjs = (wellnessMetadata$metadata.ID)
        missingSubjs1 = names(table(wellnessSubjs)[table(wellnessSubjs)==3])
        exceptionSubjs = c("X3452","X3834","X3918","X4215")
        missingSubjs2 = c("X3913","X3535","X3016")
        missingSubjs1 = missingSubjs1[!missingSubjs1 %in% exceptionSubjs]
        
        wellnessMetadata = wellnessMetadata[!wellnessMetadata$metadata.ID %in% missingSubjs1, ]
        wellnessMetadata = wellnessMetadata[!wellnessMetadata$metadata.ID %in% missingSubjs2, ]
        wellnessMetadata = wellnessMetadata[!wellnessMetadata$metadata.ID %in% exceptionSubjs, ]
        wellnessMgsMat = mergeMatUpdated[,match(wellnessMetadata$sample.ID, colnames(mergeMatUpdated))]
        #wellnessMgsMat = mgsMat
        wellnessSamples = wellnessMetadata$sample.ID
        wellnessSubjects = gsub("_v[1-4]","",wellnessSamples)
        wellnessVisits = gsub("X[0-9]+_","",wellnessSamples)
        wellnessUniqSubjs = unique(wellnessMetadata$metadata.ID)
        
      }
      
      visitCountsPerMspList = do.call(c, sapply(rownames(wellnessMgsMat), function(currMgs){
        
        individualCounts = unlist(sapply(wellnessUniqSubjs, function(subj) {
          currSamples = wellnessMetadata$sample.ID[wellnessMetadata$metadata.ID==subj]
          currSamples = currSamples[order(currSamples)]
          currMgsMat = wellnessMgsMat[,currSamples]
          currMgsAllVisits = all(currMgsMat[currMgs,] >0)
        }, simplify = F))
        return(sum(individualCounts))
      }, simplify = F))
      names(visitCountsPerMspList) = rownames(mgsMat)
      visitCountsPerMspList = visitCountsPerMspList*100/length(wellnessUniqSubjs)
      
      return(visitCountsPerMspList)
    }
    getSigMat <- function(currMgs) {
      markovSeqOfMsp = (sapply(wellnessUniqSubjs, function(subj) {
        currSamples = wellnessMetadata$sample.ID[wellnessMetadata$metadata.ID==subj]
        currSamples = currSamples[order(currSamples)]
        currMgsMat = wellnessMgsMat[,currSamples]
        currMgsSubjVec = (currMgsMat[currMgs,] >0)*1
        #currMgsSubjVec = currMgsMat[currMgs,]
        
      }, simplify = F))
      mat = do.call(rbind, markovSeqOfMsp)
      #class(mat) = "numeric"
      colnames(mat)=c("v1","v2","v3","v4")
      return(mat)
    }
    
    
    if (F) {
      
      mat = getSigMat("msp_1329") #test
      
      mat = getSigMat("msp_0172") #inflow
      mat = getSigMat("msp_0522") #outflow
      mat = getSigMat("msp_0313") #outflow
      
      
      mat = getSigMat("msp_0121") #commensal?
      mat = getSigMat("msp_0041") #commensal?
      mat = getSigMat("msp_0380") #trasient?
      
      mat = getSigMat("msp_1143") #inflow
      
      levelplot(t(mat), col.regions=c("white","gray"),aspect=3, xlab="",ylab="",colorkey=F,scales=list(draw=F))
      
      
      #plotBarcode(mat)  
      
      
      
      
    } 
    
    wellnessMetadata = basicMetaMapUpdated[basicMetaMapUpdated$dataset.ID=="id28",]
    
    briefCheckMetadata = F
    if (briefCheckMetadata) {
      maleRichness = wellnessMetadata$GeneRichness[wellnessMetadata$Gender=="Male"]
      femaleRichness = wellnessMetadata$GeneRichness[wellnessMetadata$Gender=="Female"]
      boxplot(list(m = maleRichness, f = maleRichness))  
    }
    
    reloadMgsMatImputed = T
    if (reloadMgsMatImputed) {
      
      ######
      mgsMat = mergeMatUpdated
      metaTab = basicMetaMapUpdated
      
      wellnessSubjs = (wellnessMetadata$metadata.ID)
      missingSubjs1 = names(table(wellnessSubjs)[table(wellnessSubjs)==3])
      exceptionSubjs = c("X3452","X3834","X3918","X4215")
      missingSubjs2 = c("X3913","X3535","X3016")
      missingSubjs1 = missingSubjs1[!missingSubjs1 %in% exceptionSubjs]
      
      wellnessMetadata = wellnessMetadata[!wellnessMetadata$metadata.ID %in% missingSubjs1, ]
      wellnessMetadata = wellnessMetadata[!wellnessMetadata$metadata.ID %in% missingSubjs2, ]
      wellnessMetadata = wellnessMetadata[!wellnessMetadata$metadata.ID %in% exceptionSubjs, ]
      wellnessMgsMat = mergeMatUpdatedImputed[,wellnessMetadata$sample.ID]
      #wellnessMgsMat = mergeMatUpdated[,wellnessMetadata$sample.ID]
      
      wellnessSamples = wellnessMetadata$sample.ID
      wellnessUniqSubjs = unique(wellnessMetadata$metadata.ID)
      
      
      
    }
    
    #save(wellnessMetadata, wellnessMgsMat, file="C://Data//comparative.analysis.healthy.sweden//wellness.output.RData" )
    reloadMgsMatNonImputd = F
    if (reloadMgsMatImputed) {
      
      ######
      mgsMat = mergeMatUpdated
      metaTab = basicMetaMapUpdated
      
      wellnessSubjs = (wellnessMetadata$metadata.ID)
      missingSubjs1 = names(table(wellnessSubjs)[table(wellnessSubjs)==3])
      exceptionSubjs = c("X3452","X3834","X3918","X4215")
      missingSubjs2 = c("X3913","X3535","X3016")
      missingSubjs1 = missingSubjs1[!missingSubjs1 %in% exceptionSubjs]
      
      wellnessMetadata = wellnessMetadata[!wellnessMetadata$metadata.ID %in% missingSubjs1, ]
      wellnessMetadata = wellnessMetadata[!wellnessMetadata$metadata.ID %in% missingSubjs2, ]
      wellnessMetadata = wellnessMetadata[!wellnessMetadata$metadata.ID %in% exceptionSubjs, ]
      wellnessMgsMat = mergeMatUpdated[,wellnessMetadata$sample.ID]
      #wellnessMgsMat = mergeMatUpdated[,wellnessMetadata$sample.ID]
      wellnessMgsMat = wellnessMgsMat[rowSums(wellnessMgsMat)!=0, ]
      wellnessSamples = wellnessMetadata$sample.ID
      wellnessUniqSubjs = unique(wellnessMetadata$metadata.ID)
      
      scaledWellnessMgsMat = t(scale(t(wellnessMgsMat)))
      
    }
    
    normalMgsMat = mergeMatUpdated[,newNormalAllIds]
    westernMgsMat = mergeMatUpdated[,newWesternIds]
    nonWesternMgsMat = mergeMatUpdated[,newNonWesternIds]
    diseaseMgsMat = mergeMatUpdated[,newDiseaseIds]
    
    
    freqs_normal = rowMeans(normalMgsMat>0)
    freqs_western = rowMeans(westernMgsMat>0)
    freqs_nonWestern = rowMeans(nonWesternMgsMat>0)
    freqs_disease = rowMeans(diseaseMgsMat>0)
    
    
    wellnessMgsMat = wellnessMgsMat[rowSums(wellnessMgsMat)!=0,]
    freqs_wellness = rowMeans(wellnessMgsMat>0)
    
    
    wellnessMgsMeans = rowMeans(wellnessMgsMat)
    wellnessMgsMedians = rowMedians(wellnessMgsMat)
    
    if (F) {
      par(mar=c(5,4,4,1))
      plot(ecdf(freqs_wellness))
      
      plot(freqs_wellness, wellnessMgsMeans, log="y")
      abline(v=c(0.1, 0.9))
    }
    
    newDefinition = F
    if (newDefinition) {
      mgsVisitCountFractions = getFullVisitCounts(basicMetaMapUpdated, mergeMatUpdated)
      
      passerbyCut=20
      colonizedCut=80
      passerby = names(mgsVisitCountFractions[mgsVisitCountFractions<=passerbyCut])
      colonized = names(mgsVisitCountFractions[mgsVisitCountFractions>=colonizedCut])
      transient = names(mgsVisitCountFractions[mgsVisitCountFractions<colonizedCut & mgsVisitCountFractions>passerbyCut])
    }
    
    oldDefinition = T
    if (oldDefinition) {
      passerbyCut=0.1
      colonizedCut=0.9
      passerby = names(freqs_wellness[freqs_wellness<=passerbyCut])
      colonized = names(freqs_wellness[freqs_wellness>=colonizedCut])
      transient = names(freqs_wellness[freqs_wellness<colonizedCut & freqs_wellness>passerbyCut])
      
      length(colonized)
      length(transient)
      length(passerby)
    }
    
    checkTaxonomy = F
    if (checkTaxonomy) {
      topCut=1:5
      par(mar=c(10,4,4,1))
      barplot(sort(getCountsThroughTable(taxo$phylum[taxo$MSP %in% colonized]),decreasing = T)[topCut], las=2, cex.names =  0.8)
      barplot(sort(getCountsThroughTable(taxo$phylum[taxo$MSP %in% passerby]),decreasing = T)[topCut], las=2, cex.names =  0.8)
      barplot(sort(getCountsThroughTable(taxo$phylum[taxo$MSP %in% transient]),decreasing = T)[topCut], las=2, cex.names =  0.8)
    }
    
    if (F) {
      # View(survivalStats[colonized,])
      # View(survivalStats[transient,])
      # View(survivalStats[passerby,])
      
      survEachList = list(colonized=survivalStats[colonized,"surv.rates"],
                          transient=survivalStats[transient,"surv.rates"],
                          passerby=survivalStats[passerby,"surv.rates"])
      boxplot(survEachList,las=2)
      
    }
    
    wellnessTransientMgsMat = wellnessMgsMat[transient,]
    wellnessPasserbyMgsMat = wellnessMgsMat[passerby,]
    wellnessColonizedMgsMat = wellnessMgsMat[colonized,]
    
    wellnessTransientMgsMatDigit <- wellnessTransientMgsMat > 0
    wellnessPasserbyMgsMatDigit <- wellnessPasserbyMgsMat > 0
    wellnessColonizedMgsMatDigit <- wellnessColonizedMgsMat > 0
    
    #length(colonized)
    #freqs_wellness["msp_0041"]
    
    #hist(freqs_wellness)
    # table(freqs_wellness < 0.2)
    # table(freqs_wellness > 0.8)
    # table(freqs_wellness < 0.1)
    # table(freqs_wellness > 0.9)
    # plot(density(freqs_wellness))
    # table(freqs_wellness==0)
    # freqs_all = rowMeans(mergeMatUpdatedImputed>0)
    #
    
    inflowOut = getColonizationColonizingProbsNew(wellnessMetadata, wellnessMgsMat)
    outflowOut = getColonizationResistantProbsNew(wellnessMetadata, wellnessMgsMat)
    
    if (F) { #lean vs obese
      leanMetadata = wellnessMetadata[wellnessMetadata$metadata.ID %in% leanOnly, ]
      obeseMetadata = wellnessMetadata[wellnessMetadata$metadata.ID %in% obese, ]
      leanMgsMat = wellnessMgsMat[,colnames(wellnessMgsMat) %in% leanMetadata$sample.ID  ]
      obeseMgsMat = wellnessMgsMat[,colnames(wellnessMgsMat) %in% obeseMetadata$sample.ID  ]
      
      leanPrevalence= rowMeans(leanMgsMat>0)
      leanLessPrevalnet= names(leanPrevalence[leanPrevalence<0.1])
      
      obesePrevalence= rowMeans(obeseMgsMat>0)
      obeseLessPrevalnet= names(obesePrevalence[obesePrevalence<0.1])
      
      exSpecies = unique(c(leanLessPrevalnet,obeseLessPrevalnet ))
      
      lean.inflowOut = getColonizationColonizingProbsNew(leanMetadata, leanMgsMat)
      lean.outflowOut = getColonizationResistantProbsNew(leanMetadata, leanMgsMat)
      
      lean.inflow = lean.inflowOut[,1]
      lean.outflow = lean.outflowOut[,1]
      
      obese.inflowOut = getColonizationColonizingProbsNew(obeseMetadata, obeseMgsMat)
      obese.outflowOut = getColonizationResistantProbsNew(obeseMetadata, obeseMgsMat)
      
      obese.inflow = obese.inflowOut[,1]
      obese.outflow = obese.outflowOut[,1]
      
      plot(lean.inflow[!names(lean.inflow) %in% exSpecies], obese.inflow[!names(lean.inflow) %in% exSpecies])
      plot(lean.outflow[!names(lean.outflow) %in% exSpecies], obese.outflow[!names(lean.outflow) %in% exSpecies])
      
      lean.obese.inflow = data.frame(lean.inflow[!names(lean.inflow) %in% exSpecies], obese.inflow[!names(lean.inflow) %in% exSpecies])
      lean.obese.inflow$species = getSpeciesName(rownames(lean.obese.inflow), gssOssTabNew)
      
      
      cor.test(lean.inflow[!names(lean.inflow) %in% exSpecies], obese.inflow[!names(lean.inflow) %in% exSpecies], method="spearman")
      cor.test(lean.outflow[!names(lean.outflow) %in% exSpecies], obese.outflow[!names(lean.outflow) %in% exSpecies], method="spearman")
      
    }
    
    #inflow = getColonizationColonizingProbs(wellnessMetadata, wellnessTransientMgsMatDigit)
    #outflow = getColonizationResistantProbs(wellnessMetadata, wellnessTransientMgsMatDigit)
    inflowOut = getColonizationColonizingProbsNew(wellnessMetadata, wellnessMgsMat)
    outflowOut = getColonizationResistantProbsNew(wellnessMetadata, wellnessMgsMat)
    
    inflow = inflowOut[,1]
    outflow = outflowOut[,1]
    names(inflow) = rownames(inflowOut)
    names(outflow) = rownames(outflowOut)
    
    inflowGenera = getGenusName(names(inflow), taxo)
    outflowGenera = getGenusName(names(outflow), taxo)
    
    wellnessInOutFlowTab = data.frame(inflow=inflow,
                                      outflow=outflow,
                                      prevalence=freqs_wellness,
                                      species = getSpeciesName(names(inflow),gssOssTabNew) ) 
    
    
    #newInflow = inflow
    #newOutflow = outflow
    
    #oldInflow = inflow
    #oldOutflow = outflow
    
    checkSequencingDepth = T
    if (checkSequencingDepth) {
      
      wellnessMgs5mRData = "J:\\Deposit\\Project\\2018_microbiome_atlas\\gctTables\\wellness.final.mgs.med.vec.5M.RData"
      wellnessMgs15mRData = "J:\\Deposit\\Project\\2018_microbiome_atlas\\gctTables\\wellness.final.mgs.med.vec.15M.RData"
      load(wellnessMgs5mRData) #mgs_med_vec_5m
      load(wellnessMgs15mRData) #mgs_med_vec_15m
      
      #wellnessMgsMat
      wellnessMgsMat_10m = wellnessMgsMat
      wellnessMgsMat_5m = mgs_med_vec_5m[match(rownames(wellnessMgsMat_10m), rownames(mgs_med_vec_5m)),]
      rownames(wellnessMgsMat_5m) = rownames(wellnessMgsMat_10m)
      
      wellnessMgsMat_5m = wellnessMgsMat_5m[,match(colnames(wellnessMgsMat_10m),colnames(wellnessMgsMat_5m))]
      wellnessMgsMat_5m[is.na(wellnessMgsMat_5m)] = 0 #20640 cases were NAs
      
      wellnessMgsMat_15m = mgs_med_vec_15m[match(rownames(wellnessMgsMat_10m), rownames(mgs_med_vec_15m)),]
      wellnessMgsMat_15m = wellnessMgsMat_15m[,match(colnames(wellnessMgsMat_10m),colnames(wellnessMgsMat_15m))]
      wellnessMgsMat_15m[is.na(wellnessMgsMat_15m)] = 0 # no NAs
      
      wellnessMgsMat_5m[wellnessMgsMat_5m<qCutOff] <- 0
      wellnessMgsMat_15m[wellnessMgsMat_15m<qCutOff] <- 0
      
      inflowOut_5m = getColonizationColonizingProbsNew(wellnessMetadata, wellnessMgsMat_5m)
      outflowOut_5m = getColonizationResistantProbsNew(wellnessMetadata, wellnessMgsMat_5m)
      
      inflowOut_15m = getColonizationColonizingProbsNew(wellnessMetadata, wellnessMgsMat_15m)
      outflowOut_15m = getColonizationResistantProbsNew(wellnessMetadata, wellnessMgsMat_15m)
      
      inflow_5m = inflowOut_5m[,1]
      outflow_5m = outflowOut_5m[,1]
      names(inflow_5m) = rownames(inflowOut_5m)
      names(outflow_5m) = rownames(outflowOut_5m)
      
      inflow_15m = inflowOut_15m[,1]
      outflow_15m = outflowOut_15m[,1]
      names(inflow_15m) = rownames(inflowOut_15m)
      names(outflow_15m) = rownames(outflowOut_15m)
      
      checkExistence = F
      if (checkExistence) {
        
        present_10m = apply(wellnessMgsMat, 1, any)
        present_5m = apply(wellnessMgsMat_5m, 1, any)
        present_15m = apply(wellnessMgsMat_15m, 1, any)
        
        table(present_10m & present_5m)
        table(present_10m & present_15m)
        table(present_5m & present_15m)
        getSpeciesName(names(present_10m[present_10m & !present_5m]), taxo)
        getSpeciesName(names(present_10m[present_10m & !present_15m]), taxo)
        
        
        sum(present_5m)
        sum(present_10m)
        sum(present_15m)
        
        
      }
      
      if(F) {
        plot(inflow_5m[present_10m & present_5m], inflow[present_10m & present_5m])
        plot(inflow_15m[present_10m & present_15m], inflow[present_10m & present_15m])
        
        plot(outflow_5m[present_10m & present_5m], outflow[present_10m & present_5m])
        plot(outflow_15m[present_10m & present_15m], outflow[present_10m & present_15m])
        
        
        cor.test(inflow_5m[present_10m & present_5m], inflow[present_10m & present_5m])
        cor.test(inflow_15m[present_10m & present_15m], inflow[present_10m & present_15m])
        
        cor.test(outflow_5m[present_10m & present_5m], outflow[present_10m & present_5m])
        cor.test(outflow_15m[present_10m & present_15m], outflow[present_10m & present_15m])
        
        cor.test(inflow_5m, inflow)
        cor.test(inflow_15m, inflow)
        
        cor.test(outflow_5m, outflow)
        cor.test(outflow_15m, outflow)
      }
      
    }
    
    checkSeqDepthAndFiltering = F
    if (checkSeqDepthAndFiltering) {
      
      wellnessMgs5mRData = "J:\\Deposit\\Project\\2018_microbiome_atlas\\gctTables\\wellness.final.mgs.med.vec.5M.RData"
      wellnessMgs15mRData = "J:\\Deposit\\Project\\2018_microbiome_atlas\\gctTables\\wellness.final.mgs.med.vec.15M.RData"
      load(wellnessMgs5mRData) #mgs_med_vec_5m
      load(wellnessMgs15mRData) #mgs_med_vec_15m
      
      #wellnessMgsMat
      wellnessMgsMat_10m = wellnessMgsMat
      wellnessMgsMat_5m = mgs_med_vec_5m[match(rownames(wellnessMgsMat_10m), rownames(mgs_med_vec_5m)),]
      rownames(wellnessMgsMat_5m) = rownames(wellnessMgsMat_10m)
      
      wellnessMgsMat_5m = wellnessMgsMat_5m[,match(colnames(wellnessMgsMat_10m),colnames(wellnessMgsMat_5m))]
      wellnessMgsMat_5m[is.na(wellnessMgsMat_5m)] = 0 #20640 cases were NAs
      
      
      wellnessMgsMean5mRData = "J:\\Deposit\\Project\\2018_microbiome_atlas\\gctTables\\wellness.final.mgs.mean.vec.5M.RData"
      wellnessMgsMean10mRData = "J:\\Deposit\\Project\\2018_microbiome_atlas\\gctTables\\wellness.final.mgs.mean.vec.10M.RData"
      wellnessMgsMean15mRData = "J:\\Deposit\\Project\\2018_microbiome_atlas\\gctTables\\wellness.final.mgs.mean.vec.15M.RData"
      
      load(wellnessMgsMean5mRData) #mgs_mean_vec_5m
      load(wellnessMgsMean10mRData) #mgs_mean_vec_10m
      load(wellnessMgsMean15mRData) #mgs_mean_vec_15m
      
      #wellnessMgsMat
      wellnessMgsMeanMat_5m = mgs_mean_vec_5m[match(rownames(wellnessMgsMat), rownames(mgs_mean_vec_5m)),]
      rownames(wellnessMgsMeanMat_5m) = rownames(wellnessMgsMat)
      
      wellnessMgsMeanMat_5m = wellnessMgsMeanMat_5m[,match(colnames(wellnessMgsMat),colnames(wellnessMgsMeanMat_5m))]
      wellnessMgsMeanMat_5m[is.na(wellnessMgsMeanMat_5m)] = 0 #no NAs
      
      wellnessMgsMeanMat_10m = mgs_mean_vec_10m[match(rownames(wellnessMgsMat), rownames(mgs_mean_vec_10m)),]
      rownames(wellnessMgsMeanMat_10m) = rownames(wellnessMgsMat)
      
      wellnessMgsMeanMat_10m = wellnessMgsMeanMat_10m[,match(colnames(wellnessMgsMat),colnames(wellnessMgsMeanMat_10m))]
      wellnessMgsMeanMat_10m[is.na(wellnessMgsMeanMat_10m)] = 0 #no NAs
      
      wellnessMgsMeanMat_15m = mgs_mean_vec_15m[match(rownames(wellnessMgsMat), rownames(mgs_mean_vec_15m)),]
      rownames(wellnessMgsMeanMat_15m) = rownames(wellnessMgsMat)
      
      wellnessMgsMeanMat_15m = wellnessMgsMeanMat_15m[,match(colnames(wellnessMgsMat),colnames(wellnessMgsMeanMat_15m))]
      wellnessMgsMeanMat_15m[is.na(wellnessMgsMeanMat_15m)] = 0 #no NAs
      
      
      #####
      wellnessMgsMeanMat_5m[wellnessMgsMeanMat_5m<qCutOff] <- 0
      wellnessMgsMeanMat_10m[wellnessMgsMeanMat_10m<qCutOff] <- 0
      wellnessMgsMeanMat_15m[wellnessMgsMeanMat_15m<qCutOff] <- 0
      
      inflowOut_mean_5m = getColonizationColonizingProbsNew(wellnessMetadata, wellnessMgsMeanMat_5m)
      outflowOut_mean_5m = getColonizationResistantProbsNew(wellnessMetadata, wellnessMgsMeanMat_5m)
      
      inflowOut_mean_10m = getColonizationColonizingProbsNew(wellnessMetadata, wellnessMgsMeanMat_10m)
      outflowOut_mean_10m = getColonizationResistantProbsNew(wellnessMetadata, wellnessMgsMeanMat_10m)
      
      inflowOut_mean_15m = getColonizationColonizingProbsNew(wellnessMetadata, wellnessMgsMeanMat_15m)
      outflowOut_mean_15m = getColonizationResistantProbsNew(wellnessMetadata, wellnessMgsMeanMat_15m)
      
      
      inflow_mean_5m = inflowOut_mean_5m[,1]
      outflow_mean_5m = outflowOut_mean_5m[,1]
      names(inflow_mean_5m) = rownames(inflowOut_mean_5m)
      names(outflow_mean_5m) = rownames(outflowOut_mean_5m)
      
      inflow_mean_10m = inflowOut_mean_10m[,1]
      outflow_mean_10m = outflowOut_mean_10m[,1]
      names(inflow_mean_10m) = rownames(inflowOut_mean_10m)
      names(outflow_mean_10m) = rownames(outflowOut_mean_10m)
      
      inflow_mean_15m = inflowOut_mean_15m[,1]
      outflow_mean_15m = outflowOut_mean_15m[,1]
      names(inflow_mean_15m) = rownames(inflowOut_mean_15m)
      names(outflow_mean_15m) = rownames(outflowOut_mean_15m)
      
      plot(inflow_5m, inflow_mean_5m)
      plot(inflow, inflow_mean_10m)
      plot(inflow_15m, inflow_mean_15m)
      
      plot(outflow_5m, outflow_mean_5m)
      plot(outflow, outflow_mean_10m)
      plot(outflow_15m, outflow_mean_15m)
      
      
      cor.test(inflow_5m, inflow_mean_5m)
      cor.test(inflow, inflow_mean_10m)
      cor.test(inflow_15m, inflow_mean_15m)
      
      cor.test(outflow_5m, outflow_mean_5m)
      cor.test(outflow, outflow_mean_10m)
      cor.test(outflow_15m, outflow_mean_15m)
      
      
      allInflowTab = data.frame(inflow_5m = inflow_5m, 
                                inflow_10m = inflow, 
                                inflow_15m = inflow_15m, 
                                stringsAsFactors = F)
      
      allOutflowTab = data.frame(outflow_5m = outflow_5m, 
                                 outflow_10m = outflow, 
                                 outflow_15m = outflow_15m, 
                                 stringsAsFactors = F)
      
    }
    
    checkEnrichment = T
    if (checkEnrichment) {
      outflowSpecies = names(outflow[outflow>=0.3])
      inflowSpecies = names(inflow[inflow>=0.3])
      sharedSpecies = inflowSpecies[inflowSpecies %in% outflowSpecies]
      commensalSpecies = names(inflow)[names(inflow) %in% c(inflowSpecies, outflowSpecies, sharedSpecies)]
      
      
      outflowAllSpecies = names(outflow[outflow>=0.3])
      inflowAllSpecies = names(inflow[inflow>=0.3])
      
      
      #outflowSpecies = names(outflow[outflow>=0.5])
      #inflowSpecies = names(inflow[inflow>=0.5])
      
      #outflowSpecies = outflowSpecies[!outflowSpecies %in% inflowSpecies]
      #inflowSpecies = inflowSpecies[!inflowSpecies %in% outflowSpecies]
      
      outflowSpecies = outflowAllSpecies[!outflowAllSpecies %in% inflowAllSpecies]
      inflowSpecies = inflowAllSpecies[!inflowAllSpecies %in% outflowAllSpecies]
      inflowSpecies = inflowSpecies[inflowSpecies!="msp_0076"]
      table(inflowSpecies %in% outflowSpecies)
      
      #save(inflowSpecies, outflowSpecies, file="inflow.outflow.species.RData")
      #getwd()
      
      getOverlapMode = F
      if (getOverlapMode){
        overlapInflowSpecies = inflowSpecies[inflowSpecies %in% dinamicInflowSpecies]
        overlapOutflowSpecies = outflowSpecies[outflowSpecies %in% dinamicOutflowSpecies]
        
        overlapInflowSpecies = overlapInflowSpecies[overlapInflowSpecies %in% americanInflowSpecies]
        overlapOutflowSpecies = overlapOutflowSpecies[overlapOutflowSpecies %in% americanOutflowSpecies]
        
        combinedOverlapSpecies = c(overlapInflowSpecies, overlapOutflowSpecies)
        markovStatTabTemp= data.frame(inflow = inflowOut[,1],
                                      inflowUpper = inflowOut[,2],
                                      inflowLower = inflowOut[,3],
                                      inflowSE = inflowOut[,4],
                                      outflow = outflowOut[,1],
                                      outflowUpper = outflowOut[,2],
                                      outflowLower = outflowOut[,3],
                                      outflowSE = outflowOut[,4],
                                      prevalence=freqs_wellness, 
                                      meanAbd=wellnessMgsMeans,
                                      species = getSpeciesName(rownames(inflowOut),taxo),
                                      stringsAsFactors = F)
        
        markovStatTabTemp = markovStatTabTemp[markovStatTabTemp$prevalence>0.1,]
        markovStatTabTemp = markovStatTabTemp[rownames(markovStatTabTemp) %in% combinedOverlapSpecies,  ]
        
        write.table(markovStatTabTemp, "markovStatTabTempAllOverlap.txt", sep="\t")
        
      }
      
      reloadMgsMatNonImputd = T
      if (reloadMgsMatImputed) {
        
        ######
        mgsMat = mergeMatUpdated
        metaTab = basicMetaMapUpdated
        
        wellnessSubjs = (wellnessMetadata$metadata.ID)
        missingSubjs1 = names(table(wellnessSubjs)[table(wellnessSubjs)==3])
        exceptionSubjs = c("X3452","X3834","X3918","X4215")
        missingSubjs2 = c("X3913","X3535","X3016")
        missingSubjs1 = missingSubjs1[!missingSubjs1 %in% exceptionSubjs]
        
        wellnessMetadata = wellnessMetadata[!wellnessMetadata$metadata.ID %in% missingSubjs1, ]
        wellnessMetadata = wellnessMetadata[!wellnessMetadata$metadata.ID %in% missingSubjs2, ]
        wellnessMetadata = wellnessMetadata[!wellnessMetadata$metadata.ID %in% exceptionSubjs, ]
        wellnessMgsMat2 = mergeMatUpdated[,wellnessMetadata$sample.ID]
        #wellnessMgsMat = mergeMatUpdated[,wellnessMetadata$sample.ID]
        
        wellnessSamples = wellnessMetadata$sample.ID
        wellnessUniqSubjs = unique(wellnessMetadata$metadata.ID)
        
        scaledWellnessMgsMat = t(scale(t(wellnessMgsMat2)))
        wellnessClinicalMatrix2 = wellnessClinicalMatrix
        bmi = wellnessClinicalMatrix["anthropometry_weight",] / ( wellnessClinicalMatrix["anthropometry_height",]/100)^2
        wellnessClinicalMatrix2 = rbind(wellnessClinicalMatrix2, bmi)
        #scaledWellnessClinicalMatrix = t(scale(t(wellnessClinicalMatrix)))
        scaledWellnessClinicalMatrix = t(scale(t(wellnessClinicalMatrix2)))
        scaledBmi = scaledWellnessClinicalMatrix["bmi",]
        
        scaledInflowMeans = colSums(scaledWellnessMgsMat[rownames(scaledWellnessMgsMat)%in%inflowSpecies,])
        scaledInflowMeans = scaledInflowMeans / sqrt(length(inflowSpecies))
        
        scaledOutflowMeans = colSums(scaledWellnessMgsMat[rownames(scaledWellnessMgsMat)%in%outflowSpecies,])
        scaledOutflowMeans = scaledOutflowMeans / sqrt(length(outflowSpecies))
        
        wellnessMetadata$scaledInflow = scaledInflowMeans
        wellnessMetadata$scaledOutflow = scaledOutflowMeans
        
        wellnessSeqTab$scaledInflow_bf = scaledInflowMeans[match(wellnessSeqTab$before,names(scaledInflowMeans))]
        wellnessSeqTab$scaledInflow_af = scaledInflowMeans[match(wellnessSeqTab$after,names(scaledInflowMeans))]
        
        wellnessSeqTab$scaledOutflow_bf = scaledOutflowMeans[match(wellnessSeqTab$before,names(scaledOutflowMeans))]
        wellnessSeqTab$scaledOutflow_af = scaledOutflowMeans[match(wellnessSeqTab$after,names(scaledOutflowMeans))]
        
        wellnessSeqTab$scaledBmi_bf = scaledBmi[match(wellnessSeqTab$before,names(scaledBmi))]
        wellnessSeqTab$scaledBmi_af = scaledBmi[match(wellnessSeqTab$after,names(scaledBmi))]
        
        wellnessSeqTab$outflowDiff = wellnessSeqTab$scaledOutflow_af - wellnessSeqTab$scaledOutflow_bf
        wellnessSeqTab$inflowDiff = wellnessSeqTab$scaledInflow_af - wellnessSeqTab$scaledInflow_bf
        wellnessSeqTab$bmiDiff = wellnessSeqTab$scaledBmi_af - wellnessSeqTab$scaledBmi_bf
        
        wellnessSeqTab$outflowSum = wellnessSeqTab$scaledOutflow_af + wellnessSeqTab$scaledOutflow_bf
        wellnessSeqTab$inflowSum = wellnessSeqTab$scaledInflow_af + wellnessSeqTab$scaledInflow_bf
        wellnessSeqTab$bmiSum = wellnessSeqTab$scaledBmi_af + wellnessSeqTab$scaledBmi_bf
        
        wellnessSeqTab$richnessLFC = log2(wellnessSeqTab$GeneRichness_af / wellnessSeqTab$GeneRichness_bf)
        
        
        
        ###PCoA analysis
        pcoaMode = F
        if (pcoaMode) {
          
          highOutflowSamples = wellnessSeqTab$before[wellnessSeqTab$scaledOutflow_bf>2]
          highOutflowSubjects = wellnessSeqTab$subj[wellnessSeqTab$scaledOutflow_bf>2]
          highOutflowSamplesSelected = highOutflowSamples[!duplicated(highOutflowSubjects)]
          highOutflowSamplesSelectedAfter = wellnessSeqTab$after[match(highOutflowSamplesSelected, wellnessSeqTab$before)]
          
          highInflowSamples = wellnessSeqTab$before[wellnessSeqTab$scaledInflow_bf>2]
          highInflowSubjects = wellnessSeqTab$subj[wellnessSeqTab$scaledInflow_bf>2]
          highInflowSamplesSelected = highInflowSamples[!duplicated(highInflowSubjects)]
          highInflowSamplesSelectedAfter = wellnessSeqTab$after[match(highInflowSamplesSelected, wellnessSeqTab$before)]
          
          
          
          wellnessSeqTab$scaledOutflow_bf[1:5]
          
          wellnessMgsDist=vegdist(t(wellnessMgsMat2), method="bray")
          wellnessPcoaOut=pcoa(wellnessMgsDist)
          wellnessPcoaMat = wellnessPcoaOut$vectors 
          wellnessPcoaTab = data.frame(wellnessPcoaMat[,1:2],
                                       subj=gsub("_v[1-4]","", rownames(wellnessPcoaMat)), 
                                       sample=rownames(wellnessPcoaMat), 
                                       inflow = scaledInflowMeans[match(rownames(wellnessPcoaMat),names(scaledInflowMeans))],
                                       outflow = scaledOutflowMeans[match(rownames(wellnessPcoaMat),names(scaledOutflowMeans))],
                                       stringsAsFactors = F)
          wellnessPcoaDist = as.matrix(dist(wellnessPcoaMat[,1:2]))
          
          distList = list(inflow = wellnessPcoaDist[cbind(highInflowSamplesSelected, highInflowSamplesSelectedAfter)],
                          outflow = wellnessPcoaDist[cbind(highOutflowSamplesSelected, highOutflowSamplesSelectedAfter)])
          boxplot(distList)
          
          
          distInOutAbdTab = data.frame(distance = wellnessPcoaDist[cbind(c(highInflowSamplesSelected, highOutflowSamplesSelected), c(highInflowSamplesSelectedAfter, highOutflowSamplesSelectedAfter))],
                                       scaledOutflow = scaledOutflowMeans[match(c(highInflowSamplesSelected, highOutflowSamplesSelected),names(scaledOutflowMeans))],
                                       scaledInflow = scaledInflowMeans[match(c(highInflowSamplesSelected, highOutflowSamplesSelected),names(scaledInflowMeans))])
          
          plot(distInOutAbdTab$scaledInflow, distInOutAbdTab$distance, pch=16,col="#80808088")
          abline(lm(distance ~ scaledInflow, distInOutAbdTab), lwd=2)
          cor.test(distInOutAbdTab$distance, distInOutAbdTab$scaledInflow)
          
          plot(distInOutAbdTab$scaledOutflow, distInOutAbdTab$distance, pch=16,col="#80808088",
               xaxt="n",yaxt="n",
               xlab="",ylab="")
          abline(lm(distance ~ scaledOutflow, distInOutAbdTab), lwd=2,col="red")
          cor.test(distInOutAbdTab$distance, distInOutAbdTab$scaledOutflow)
          
          
          wellnessPcoaTab$class = "before"
          wellnessPcoaTab = wellnessPcoaTab[wellnessPcoaTab$sample %in% c(, highOutflowSamplesSelectedAfter), ]
          wellnessPcoaTab$class[wellnessPcoaTab$sample %in% highOutflowSamplesSelectedAfter] = "after"
          
          ggplot(wellnessPcoaTab, aes(x=Axis.1 , y=Axis.2, fill=outflow)) + geom_line(aes(group=subj)) + 
            geom_point(size=3, shape=21, colour="#00000055") + xlab("") + ylab("")  +
            scale_fill_gradient2(low="gray", mid="white", high="red", limits = c(-15,15))+
            theme(panel.grid.major = element_blank(), 
                  panel.grid.minor = element_blank(), #legend.position = "none",
                  text = element_text(size = 15),
                  panel.background = element_rect(fill = "white", 
                                                  colour = "black", 
                                                  size = 0.5, 
                                                  linetype = "solid")) 
          
          ggplot(wellnessPcoaTab, aes(x=Axis.1 , y=Axis.2, fill=outflow)) + geom_line(aes(group=subj)) + 
            geom_point(size=3, shape=21, colour="#00000055") + xlab("") + ylab("")  +
            scale_fill_gradient(low="gray",high="red")+
            theme(panel.grid.major = element_blank(), 
                  panel.grid.minor = element_blank(), #legend.position = "none",
                  text = element_text(size = 15),
                  panel.background = element_rect(fill = "white", 
                                                  colour = "black", 
                                                  size = 0.5, 
                                                  linetype = "solid")) 
          
          
          wellnessPcoaTab$class = "before"
          wellnessPcoaTab = wellnessPcoaTab[wellnessPcoaTab$sample %in% c(highInflowSamplesSelected, highInflowSamplesSelectedAfter), ]
          wellnessPcoaTab$class[wellnessPcoaTab$sample %in% highInflowSamplesSelectedAfter] = "after"
          
          ggplot(wellnessPcoaTab, aes(x=Axis.1 , y=Axis.2, fill=outflow)) + geom_line(aes(group=subj)) + 
            geom_point(size=3, shape=21, colour="#00000055") + xlab("") + ylab("")  +
            scale_fill_gradient2(low="gray", mid="white", high="red", limits = c(-15,15))+
            theme(panel.grid.major = element_blank(), 
                  panel.grid.minor = element_blank(), #legend.position = "none",
                  text = element_text(size = 15),
                  panel.background = element_rect(fill = "white", 
                                                  colour = "black", 
                                                  size = 0.5, 
                                                  linetype = "solid")) 
          
          ggplot(wellnessPcoaTab, aes(x=Axis.1 , y=Axis.2, fill=subj)) + geom_line() + 
            geom_point(size=3, shape=21, colour="#00000055") + xlab("") + ylab("")  +
            theme(panel.grid.major = element_blank(), 
                  panel.grid.minor = element_blank(), #legend.position = "none",
                  text = element_text(size = 15),
                  panel.background = element_rect(fill = "white", 
                                                  colour = "black", 
                                                  size = 0.5, 
                                                  linetype = "solid")) 
          
          
          
          allOutflowSelected = c(highOutflowSamplesSelected, highOutflowSamplesSelectedAfter)
          allInflowSelected = c(highInflowSamplesSelected, highInflowSamplesSelectedAfter)
          
          wellnessPcoaTab$class = "before"
          wellnessPcoaTab = wellnessPcoaTab[wellnessPcoaTab$sample %in% c(highInflowSamplesSelected, highOutflowSamplesSelected, highInflowSamplesSelectedAfter, highOutflowSamplesSelectedAfter), ]
          wellnessPcoaTab$class[wellnessPcoaTab$sample %in% c(highInflowSamplesSelectedAfter, highOutflowSamplesSelectedAfter)] = "after"
          wellnessPcoaTab$type = "inflow"
          wellnessPcoaTab$type[wellnessPcoaTab$sample %in% c(highOutflowSamplesSelected, highOutflowSamplesSelectedAfter)] = "outflow"
          
          ggplot(wellnessPcoaTab, aes(x=Axis.1 , y=Axis.2, fill=outflow)) + geom_line(aes(group=subj)) + 
            geom_point(size=3, shape=21, colour="#00000055") + xlab("") + ylab("")  +
            scale_fill_gradient2(low="gray", mid="white", high="red", limits = c(-15,15))+
            theme(panel.grid.major = element_blank(), 
                  panel.grid.minor = element_blank(), #legend.position = "none",
                  text = element_text(size = 15),
                  panel.background = element_rect(fill = "white", 
                                                  colour = "black", 
                                                  size = 0.5, 
                                                  linetype = "solid")) 
          
          wellnessPcoaTab$class = "before"
          wellnessPcoaTab = wellnessPcoaTab[wellnessPcoaTab$sample %in% c(highOutflowSamplesSelected, highOutflowSamplesSelectedAfter), ]
          wellnessPcoaTab$class[wellnessPcoaTab$sample %in% highOutflowSamplesSelectedAfter] = "after"
          
          ggplot(wellnessPcoaTab, aes(x=Axis.1 , y=Axis.2, fill=outflow)) + geom_line(aes(group=subj)) + 
            geom_point(size=3, shape=21, colour="#00000055") + xlab("") + ylab("")  +
            scale_fill_gradient2(low="gray", mid="white", high="red", limits = c(-15,15))+
            theme(panel.grid.major = element_blank(), 
                  panel.grid.minor = element_blank(), #legend.position = "none",
                  text = element_text(size = 15),
                  panel.background = element_rect(fill = "white", 
                                                  colour = "black", 
                                                  size = 0.5, 
                                                  linetype = "solid")) 
          
          boxOutGrid = list(rowMeans(allGridMat[rownames(allGridMat)%in%outflowSpecies,allInflowSelected],na.rm=T),
                            rowMeans(allGridMat[rownames(allGridMat)%in%outflowSpecies,allOutflowSelected],na.rm=T))
          
          boxInGrid = list(rowMeans(allGridMat[rownames(allGridMat)%in%inflowSpecies,allInflowSelected],na.rm=T),
                           rowMeans(allGridMat[rownames(allGridMat)%in%inflowSpecies,allOutflowSelected],na.rm=T))
          
          
          boxInflowEnrichedGrid = list(rowMeans(allGridMat[rownames(allGridMat)%in%inflowSpecies,allInflowSelected],na.rm=T),
                                       rowMeans(allGridMat[rownames(allGridMat)%in%outflowSpecies,allInflowSelected],na.rm=T))
          
          boxOutflowEnrichedGrid = list(rowMeans(allGridMat[rownames(allGridMat)%in%inflowSpecies,allOutflowSelected],na.rm=T),
                                        rowMeans(allGridMat[rownames(allGridMat)%in%outflowSpecies,allOutflowSelected],na.rm=T))
          
          
          boxInflowEnrichedGridMelt = melt(boxInflowEnrichedGrid)
          boxOutflowEnrichedGridMelt = melt(boxOutflowEnrichedGrid)
          
          boxInflowEnrichedGridMelt$L1 = factor(boxInflowEnrichedGridMelt$L1)
          boxOutflowEnrichedGridMelt$L1 = factor(boxOutflowEnrichedGridMelt$L1)
          
          
          ggplot(boxInflowEnrichedGridMelt, aes(x=L1,y=value)) + geom_boxplot() + theme_bw() 
          ggplot(boxOutflowEnrichedGridMelt, aes(x=L1,y=value)) + geom_boxplot() + theme_bw()
          
          #save(boxInflowEnrichedGridMelt, boxOutflowEnrichedGridMelt, file="inflow.outflow.grid.RData")
          if (F) {
            load("C://Data//comparative.analysis.healthy.sweden//inflow.outflow.grid.RData")
            require(ggplot2)
            ggplot(boxInflowEnrichedGridMelt, aes(x=L1,y=value)) + geom_boxplot() + theme_bw() 
            ggplot(boxOutflowEnrichedGridMelt, aes(x=L1,y=value)) + geom_boxplot() + theme_bw()
            
            
          }
          boxplot(boxOutGrid, ylim=c(1,2))    
          boxplot(boxInGrid, ylim=c(1,2))  
          
          boxplot(boxInflowEnrichedGrid, ylim=c(1,2))    
          boxplot(boxOutflowEnrichedGrid, ylim=c(1,2))    
          
          
          wilcox.test(boxInGrid[[1]], boxInGrid[[2]])
          wilcox.test(boxInGrid[[1]], boxOutGrid[[1]], alternative = "greater")
          wilcox.test(boxInGrid[[2]], boxOutGrid[[2]], alternative = "greater")
          
          
          
          ggplot(wellnessPcoaTab, aes(x=Axis.1 , y=Axis.2, fill=type)) + geom_line(aes(group=subj)) + 
            geom_point(size=3, shape=21) + xlab("") + ylab("")  + 
            #stat_ellipse(aes(x=Axis.1, y=Axis.2,color=type, group=type),type = "norm") +
            theme(panel.grid.major = element_blank(), 
                  panel.grid.minor = element_blank(), #legend.position = "none",
                  text = element_text(size = 15),
                  panel.background = element_rect(fill = "white", 
                                                  colour = "black", 
                                                  size = 0.5, 
                                                  linetype = "solid")) 
          
          # wellnessPcoaTab$class = "low"
          # wellnessPcoaTab$class[wellnessPcoaTab$subj %in% subjWithHighOutflow] = "high"
          # 
          
          wellnessPcoaTab$class = "others"
          wellnessPcoaTab$class[wellnessPcoaTab$sample %in% wellnessSeqTab[wellnessSeqTab$scaledInflow_bf >= 2, "before"]] = "before"
          wellnessPcoaTab$class[wellnessPcoaTab$sample %in% wellnessSeqTab[wellnessSeqTab$scaledInflow_bf >= 2, "after"]] = "after"
          
          adonis(wellnessMgsDist~wellnessPcoaTab$class )
          
          
          inflow.before = wellnessPcoaTab$sample[wellnessPcoaTab$sample %in% wellnessSeqTab[wellnessSeqTab$scaledInflow_bf >= 2, "before"]] 
          inflow.after = wellnessPcoaTab$sample[wellnessPcoaTab$sample %in% wellnessSeqTab[wellnessSeqTab$scaledInflow_bf >= 2, "after"]] 
          
          
          
          
          ggplot(wellnessPcoaTab, aes(x=Axis.1 , y=Axis.2, fill=class)) + 
            geom_point(size=3, shape=21, colour="#00000055") + xlab("") + ylab("")+
            stat_ellipse(aes(x=Axis.1, y=Axis.2,color=class, group=class),type = "norm") +
            theme(panel.grid.major = element_blank(), 
                  panel.grid.minor = element_blank(),
                  legend.position = "none",
                  text = element_text(size = 15),
                  panel.background = element_rect(fill = "white", 
                                                  colour = "black", 
                                                  size = 0.5, 
                                                  linetype = "solid")) 
          
          
          ggplot(wellnessPcoaTab[wellnessPcoaTab$sample %in% c(inflow.before, inflow.after),], aes(x=Axis.1 , y=Axis.2, fill=class)) + 
            geom_point(size=3, shape=21, colour="#00000055") + xlab("") + ylab("")+
            stat_ellipse(aes(x=Axis.1, y=Axis.2,color=class, group=class),type = "norm")
          
          
          #scale_fill_gradient2(midpoint = 0.2, low="#FFD70044", mid = "white", high = "purple") +
          theme(panel.grid.major = element_blank(), 
                panel.grid.minor = element_blank(),
                legend.position = "none",
                text = element_text(size = 15),
                panel.background = element_rect(fill = "white", 
                                                colour = "black", 
                                                size = 0.5, 
                                                linetype = "solid")) 
          
          ggplot(wellnessPcoaTab[wellnessPcoaTab$class=="low",], aes(x=Axis.1 , y=Axis.2, fill=subj)) + 
            geom_point(size=3, shape=21, colour="#00000055") + xlab("") + ylab("")+
            #scale_fill_gradient2(midpoint = 0.2, low="#FFD70044", mid = "white", high = "purple") +
            theme(panel.grid.major = element_blank(), 
                  panel.grid.minor = element_blank(),
                  legend.position = "none",
                  text = element_text(size = 15),
                  panel.background = element_rect(fill = "white", 
                                                  colour = "black", 
                                                  size = 0.5, 
                                                  linetype = "solid")) 
          
        }
        
        jaccMode = F
        if (jaccMode){
          mergeMatUpdatedJaccRData = "J:\\Deposit\\Project\\2018_microbiome_atlas\\microbiome.atlas.Rproj//mergeMatUpdatedJaccard.RData"
          load(mergeMatUpdatedJaccRData) #mergeMatUpdatedJacc
          
          wellnessJaccDist = beta.pair(ifelse(t(wellnessMgsMat2)>0,1,0), index.family="jaccard")
          ###
          ### wellnessJaccDist[1] --> turnover partition
          ### wellnessJaccDist[2] --> nestedness partition
          ### wellnessJaccDist[3] --> all beta diversity
          
          wellnessSubjBetaDispersions = betadisper(wellnessJaccDist[[3]],factor(wellnessSubjects))
          
          
          wellnessSamplesBySubjList = split(wellnessMetadata$sample.ID, wellnessMetadata$metadata.ID)
          wellnessJaccBySubj = do.call(c,lapply(wellnessSamplesBySubjList, function(x) {
            currMat = mergeMatUpdatedJacc[x,x]
            return(mean(currMat[lower.tri(currMat)]))
          }))
          
          ### based on outflow ###
          subjWithHighOutflow = unique(wellnessSeqTab$subj[(wellnessSeqTab$outflowSum) >= 2])
          subjWithLowOutflow = unique(wellnessSeqTab$subj[(wellnessSeqTab$outflowSum) <= -2])
          #jaccByDifferentOutflowList = list(highOutflow=wellnessJaccBySubj[names(wellnessJaccBySubj) %in% subjWithHighOutflow], 
          #                                  lowOutflow=wellnessJaccBySubj[!names(wellnessJaccBySubj) %in% subjWithHighOutflow])
          
          jaccByDifferentOutflowList = list(highOutflow=wellnessJaccBySubj[names(wellnessJaccBySubj) %in% subjWithHighOutflow], 
                                            lowOutflow=wellnessJaccBySubj[names(wellnessJaccBySubj) %in% subjWithLowOutflow])
          t.test(jaccByDifferentOutflowList$highOutflow, jaccByDifferentOutflowList$lowOutflow)
          wilcox.test(jaccByDifferentOutflowList$highOutflow, jaccByDifferentOutflowList$lowOutflow)
          wilcox.test(jaccByDifferentOutflowList$highOutflow, jaccByDifferentOutflowList$lowOutflow, alternative = "less")
          
          length(jaccByDifferentOutflowList$highOutflow) #n=31
          length(jaccByDifferentOutflowList$lowOutflow)  #n=37
          
          boxplot(jaccByDifferentOutflowList)
          
          
          jaccByDifferentOutflowMelt = melt(jaccByDifferentOutflowList)
          # boxplot(jaccByDifferentOutflowList)
          # plot(density(jaccByDifferentOutflowList$highOutflow))
          # lines(density(jaccByDifferentOutflowList$lowOutflow))
          
          
          ggplot(jaccByDifferentOutflowMelt, aes(x=L1, y=value)) + 
            geom_boxplot() + xlab("") + ylab("")+
            theme(panel.grid.major = element_blank(), 
                  panel.grid.minor = element_blank(),
                  legend.position = "none",
                  text = element_text(size = 15),
                  #axis.text.x = element_blank(),
                  panel.background = element_rect(fill = "white", 
                                                  colour = "black", 
                                                  size = 0.5, 
                                                  linetype = "solid"))
          
          
          
          ### based on inflow ###
          subjWithHighInflow = unique(wellnessSeqTab$subj[(wellnessSeqTab$inflowSum) >= 2])
          subjWithLowInflow = unique(wellnessSeqTab$subj[(wellnessSeqTab$inflowSum) <= -2])
          jaccByDifferentInflowList = list(highInflow=wellnessJaccBySubj[names(wellnessJaccBySubj) %in% subjWithHighInflow], 
                                           lowInflow=wellnessJaccBySubj[names(wellnessJaccBySubj) %in% subjWithLowInflow])
          
          t.test(jaccByDifferentInflowList$highInflow, jaccByDifferentInflowList$lowInflow)
          wilcox.test(jaccByDifferentInflowList$highInflow, jaccByDifferentInflowList$lowInflow)
          boxplot(jaccByDifferentInflowList)
          
          plot(density(jaccByDifferentInflowList$highInflow))
          lines(density(jaccByDifferentInflowList$lowInflow))
          
        }
      }
      richness = wellnessMetadata$GeneRichness[match(colnames(wellnessMgsMat), wellnessMetadata$sample.ID)]
      
      plot(richness, scaledOutflowMeans)
      
      
      deltaMode = T
      if (deltaMode) {
        
        
        plot(wellnessSeqTab$scaledOutflow_bf, 
             wellnessSeqTab$scaledOutflow_af)
        abline(a=0,b=1)
        
        plot(richnessLFC ~ inflowDiff, data=wellnessSeqTab)
        abline(lm(richnessLFC ~ inflowDiff, data=wellnessSeqTab), col="red")
        cor.test(wellnessSeqTab$richnessLFC, wellnessSeqTab$inflowDiff)
        cor.test(wellnessSeqTab$richnessLFC, wellnessSeqTab$inflowDiff, method="spearman") #
        
        plot(richnessLFC ~ outflowDiff, data=wellnessSeqTab)
        abline(lm(richnessLFC ~ outflowDiff, data=wellnessSeqTab), col="red")
        cor.test(wellnessSeqTab$richnessLFC, wellnessSeqTab$outflowDiff) #, method="spearman"
        cor.test(wellnessSeqTab$richnessLFC, wellnessSeqTab$outflowDiff, method="spearman") #
        
        
        plot(bmiDiff ~ inflowDiff, data=wellnessSeqTab)
        abline(lm(bmiDiff ~ inflowDiff, data=wellnessSeqTab), col="red")
        cor.test(wellnessSeqTab$bmiDiff, wellnessSeqTab$inflowDiff) #, method="spearman"
        
        plot(bmiDiff ~ outflowDiff, data=wellnessSeqTab)
        abline(lm(bmiDiff ~ outflowDiff, data=wellnessSeqTab), col="red")
        cor.test(wellnessSeqTab$bmiDiff, wellnessSeqTab$outflowDiff) #, method="spearman"
        
        
        plot(richnessLFC ~ bmiDiff, data=wellnessSeqTab)
        abline(lm(richnessLFC ~ bmiDiff, data=wellnessSeqTab), col="red")
        cor.test(wellnessSeqTab$richnessLFC, wellnessSeqTab$bmiDiff) # , method="spearman"
        
        
        plot(inflowDiff ~ outflowDiff, data=wellnessSeqTab)
        abline(lm(inflowDiff ~ outflowDiff, data=wellnessSeqTab), col="red")
        cor.test(wellnessSeqTab$inflowDiff, wellnessSeqTab$outflowDiff) #, method="spearman"
        cor.test(wellnessSeqTab$inflowDiff, wellnessSeqTab$outflowDiff, method="spearman") #
        
        ggplot(wellnessSeqTab, aes(x=outflowDiff, y=inflowDiff, fill=inflowDiff)) + 
          geom_point(size=3, shape=16, colour="#00000055") + xlab("") + ylab("")+
          #scale_fill_gradient2(midpoint = 0.2, low="#FFD70044", mid = "white", high = "purple") +
          geom_smooth(method = "lm",  se = T, colour="red")+
          #geom_hline(yintercept=6)+
          #geom_abline(slope=1, intercept = 0)+
          theme(panel.grid.major = element_blank(), 
                panel.grid.minor = element_blank(),
                legend.position = "none",
                text = element_text(size = 15),
                panel.background = element_rect(fill = "white", 
                                                colour = "black", 
                                                size = 0.5, 
                                                linetype = "solid"))
        
        ggplot(wellnessSeqTab, aes(x=outflowSum, y=outflowDiff)) + 
          geom_point(size=3, shape=21, 
                     fill=ifelse(abs(wellnessSeqTab$outflowDiff)>2, "#ff000066", "#00000022" ), 
                     colour=ifelse(abs(wellnessSeqTab$outflowDiff)>2, "#00000088", "#00000000" )) + 
          xlab("") + ylab("")+
          #scale_fill_gradient2(midpoint = 0.2, low="#FFD70044", mid = "white", high = "purple") +
          #geom_smooth(method = "lm",  se = T, colour="red")+
          geom_hline(yintercept=c(-2,2), linetype = "dashed")+
          geom_vline(xintercept=2, linetype = "dashed")+
          #geom_abline(slope=1, intercept = 0)+
          theme(panel.grid.major = element_blank(), 
                panel.grid.minor = element_blank(),
                legend.position = "none",
                text = element_text(size = 15),
                panel.background = element_rect(fill = "white", 
                                                colour = "black", 
                                                size = 0.5, 
                                                linetype = "solid"))
        
        ggplot(wellnessSeqTab, aes(x=inflowDiff, y=richnessLFC, fill=richnessLFC)) + 
          geom_point(size=3, shape=16, colour="#00000055") + xlab("") + ylab("")+
          #scale_fill_gradient2(midpoint = 0.2, low="#FFD70044", mid = "white", high = "purple") +
          geom_smooth(method = "lm",  se = T, colour="red")+
          #geom_hline(yintercept=6)+
          #geom_abline(slope=1, intercept = 0)+
          theme(panel.grid.major = element_blank(), 
                panel.grid.minor = element_blank(),
                legend.position = "none",
                text = element_text(size = 15),
                panel.background = element_rect(fill = "white", 
                                                colour = "black", 
                                                size = 0.5, 
                                                linetype = "solid"))
        
        
        
        
        
        ggplot(wellnessSeqTab, aes(x=outflowDiff, y=richnessLFC, fill=richnessLFC)) + 
          geom_point(size=3, shape=16, colour="#00000055") + xlab("") + ylab("")+
          #scale_fill_gradient2(midpoint = 0.2, low="#FFD70044", mid = "white", high = "purple") +
          geom_smooth(method = "lm",  se = T, colour="red")+
          #geom_hline(yintercept=6)+
          #geom_abline(slope=1, intercept = 0)+
          theme(panel.grid.major = element_blank(), 
                panel.grid.minor = element_blank(),
                legend.position = "none",
                text = element_text(size = 15),
                panel.background = element_rect(fill = "white", 
                                                colour = "black", 
                                                size = 0.5, 
                                                linetype = "solid"))
        
        ggplot(wellnessSeqTab, aes(x=bmiDiff, y=richnessLFC, fill=richnessLFC)) + 
          geom_point(size=3, shape=16, colour="#00000055") + xlab("") + ylab("")+
          #scale_fill_gradient2(midpoint = 0.2, low="#FFD70044", mid = "white", high = "purple") +
          geom_smooth(method = "lm",  se = T, colour="red")+
          #geom_hline(yintercept=6)+
          #geom_abline(slope=1, intercept = 0)+
          theme(panel.grid.major = element_blank(), 
                panel.grid.minor = element_blank(),
                legend.position = "none",
                text = element_text(size = 15),
                panel.background = element_rect(fill = "white", 
                                                colour = "black", 
                                                size = 0.5, 
                                                linetype = "solid"))
        
        par(mar=c(.1,2,.1,1))
        par(mfrow=c(2,1))
        outSum = (wellnessSeqTab$scaledOutflow_bf+wellnessSeqTab$scaledOutflow_af)
        outDiff = (wellnessSeqTab$scaledOutflow_af-wellnessSeqTab$scaledOutflow_bf)
        plot(outSum, outDiff, 
             xaxt="n", yaxt="n",
             xlim=c(-10,25), ylim=c(-10,10), 
             xlab="", ylab="")
        abline(h=c(-2,2), lty=2)
        abline(v=2, lty=3)
        
        inSum = (wellnessSeqTab$scaledInflow_bf+wellnessSeqTab$scaledInflow_af)
        inDiff = (wellnessSeqTab$scaledInflow_af-wellnessSeqTab$scaledInflow_bf)
        plot(inSum, inDiff, 
             xaxt="n", yaxt="n",
             xlim=c(-10,25), ylim=c(-10,10), 
             xlab="", ylab="")
        abline(h=c(-2,2), lty=2)
        abline(v=2, lty=3)
        axis(1, at=seq(-10,25,5), as.character(seq(-10,25,5)), cex.axis=.7, font=1, tck=.01)
        
        
        cor.test(outDiff, inDiff)
        cor.test(outSum, inSum)
        
        plot(outDiff, inDiff)
        plot(outSum, inSum)
        
        plot(wellnessSeqTab$scaledInflow_bf, 
             wellnessSeqTab$scaledInflow_af)
        abline(a=0,b=1)
        
        cor.test(wellnessSeqTab$scaledOutflow_bf, 
                 wellnessSeqTab$scaledOutflow_af)
        
        cor.test(wellnessSeqTab$scaledInflow_bf, 
                 wellnessSeqTab$scaledInflow_af)
        
        
        plot(wellnessSeqTab$scaledOutflow_bf, 
             wellnessSeqTab$GeneRichness_bf)
        
        
        plot(scaledOutflowMeans, scaledWellnessMgsMat["msp_0030",])
        
        
        scaledInOutflowTab = data.frame(In = scaledInflowMeans, 
                                        Out = scaledOutflowMeans,
                                        Richness = richness)
        
        
        
        
        #### associating clinical parameters ####
        
        scaledWellnessSigMgsMat = scaledWellnessMgsMat[ sigClinMsps,]
        scaledOutflowMeans[1:5]
        table(names(scaledOutflowMeans)== colnames(scaledWellnessSigMgsMat))
        
        combinedScaledWellnessSigMgsMatWithInOutflow = rbind(scaledOutflowMeans,
                                                             scaledInflowMeans,
                                                             scaledWellnessSigMgsMat)
        
        combinedScaledWellnessSigMgsCorMatWithInOutflow = cor(t(combinedScaledWellnessSigMgsMatWithInOutflow))
        
        
        clinChangeMat = apply(scaledWellnessClinicalMatrix, 1, function(currRow){
          beforeFeature =  currRow[match(wellnessSeqTab$before, colnames(scaledWellnessClinicalMatrix))]
          afterFeature =  currRow[match(wellnessSeqTab$after, colnames(scaledWellnessClinicalMatrix))]
          diffFeature = afterFeature - beforeFeature
          return(diffFeature)
        })
        
        mgsChangeMat = apply(scaledWellnessMgsMat, 1, function(currRow){
          beforeFeature =  currRow[match(wellnessSeqTab$before, colnames(scaledWellnessMgsMat))]
          afterFeature =  currRow[match(wellnessSeqTab$after, colnames(scaledWellnessMgsMat))]
          diffFeature = afterFeature - beforeFeature
          return(diffFeature)
        })
        
        flowClinAssocMat = cbind(deltaInflow=wellnessSeqTab$inflowDiff,
                                 deltaOutflow=wellnessSeqTab$outflowDiff,
                                 clinChangeMat)
        flowClinCorMat = cor(flowClinAssocMat, use="pairwise.complete.obs")
        
        
        mgsClinAssocMat = cbind(clinChangeMat,
                                mgsChangeMat)
        
        mgsClinCorMat = cor(mgsClinAssocMat, use="pairwise.complete.obs")
        
        mgsClinCorMelt = melt(mgsClinCorMat)
        mgsClinCorMelt = mgsClinCorMelt[mgsClinCorMelt$Var1 %in% colnames(clinChangeMat), ]
        mgsClinCorMelt = mgsClinCorMelt[mgsClinCorMelt$Var2 %in% colnames(mgsChangeMat), ]
        mgsClinCorMelt$species = getSpeciesName(mgsClinCorMelt$Var2, taxo)
        View(mgsClinCorMelt)
        
        ### 
        
      }
      
      binningMode = F
      if (binningMode) {
        outflowTab = data.frame(richness=richness,
                                outflow=scaledOutflowMeans,
                                stringsAsFactors = F)
        
        richnessBins = getRichnessBinByOutflowMean(outflowTab)
        boxplot(richnessBins)
        
      }
      
      compareInflowOutflow = T
      if (compareInflowOutflow) {
        ggplot(scaledInOutflowTab, aes(x=Out, y=In, fill=In)) + 
          geom_point(size=3, shape=16, colour="#00000055") + xlab("") + ylab("")+
          #scale_fill_gradient2(midpoint = 0.2, low="#FFD70044", mid = "white", high = "purple") +
          geom_smooth(method = "lm",  se = F, colour="red")+
          #geom_hline(yintercept=6)+
          #geom_abline(slope=1, intercept = 0)+
          theme(panel.grid.major = element_blank(), 
                panel.grid.minor = element_blank(),
                legend.position = "none",
                text = element_text(size = 15),
                panel.background = element_rect(fill = "white", 
                                                colour = "black", 
                                                size = 0.5, 
                                                linetype = "solid"))
        
        clinicalMatrix = clinicalMatrix[match(wellnessMetadata$sample.ID, rownames(clinicalMatrix)),]
        
        tt=cbind(scaledOutflowMeans,
                 scaledInflowMeans,
                 richness,
                 clinicalMatrix)
        out = cor(tt, use = "pairwise.complete.obs")[1,]
        out2 = cor(tt, use = "pairwise.complete.obs")[2,]
        out3 = cor(tt, use = "pairwise.complete.obs")[3,]
        
        plot(scaledInflowMeans, scaledOutflowMeans)
        abline(lm(scaledOutflowMeans ~ scaledInflowMeans), col="red")
        cor.test(scaledInflowMeans, scaledOutflowMeans)
        cor.test(scaledInflowMeans, scaledOutflowMeans, method = "spearman")
        
        # cor.test(wellnessMetadata$GeneRichness, scaledOutflowMeans)
        # cor.test(wellnessMetadata$GeneRichness, scaledInflowMeans)
        
        
        plot(wellnessMetadata$GeneRichness, scaledOutflowMeans)
        plot(wellnessMetadata$GeneRichness, scaledInflowMeans)
        
        
        plot(density(scaledOutflowMeans))
        #plot(richnessFCofWellnessNew)
        
        quantile(scaledOutflowMeans, 0.05)
        quantile(scaledOutflowMeans, 0.95)
        
        quantile(scaledInflowMeans, 0.05)
        quantile(scaledInflowMeans, 0.95)
        
        highOutflowSamples = names(scaledOutflowMeans[scaledOutflowMeans>=as.numeric(quantile(scaledOutflowMeans, 0.9))])
        lowOutflowSamples = names(scaledOutflowMeans[scaledOutflowMeans<=as.numeric(quantile(scaledOutflowMeans, 0.1))])
        #otherSamples = names(scaledOutflowMeans[scaledOutflowMeans<5])
        
        highOutflowFC = increasedPairTab2[increasedPairTab2$before %in% highOutflowSamples, "fc"]
        lowOutflowFC = increasedPairTab2[increasedPairTab2$before %in% lowOutflowSamples, "fc"]
        
        absLog2FC = list(high=abs(log2(highOutflowFC)), 
                         low=abs(log2(lowOutflowFC)))
        
        absLog2FCMelt = melt(absLog2FC)
        
        ggplot(absLog2FCMelt, aes(x=L1, y=value)) +  
          scale_fill_manual(values=c("#ffd70088","#d3d3d377")) +
          geom_boxplot(aes(fill=L1)) +
          xlab("") +
          ylab("") + 
          theme(panel.grid.major = element_blank(), 
                panel.grid.minor = element_blank(),
                legend.position = "none",
                panel.background = element_rect(fill = "white", 
                                                colour = "black", 
                                                size = 0.5, 
                                                linetype = "solid"))
        
        
        
        highInflowSamples = names(scaledInflowMeans[scaledInflowMeans>=as.numeric(quantile(scaledInflowMeans, 0.9))])
        lowInflowSamples = names(scaledInflowMeans[scaledInflowMeans<=as.numeric(quantile(scaledInflowMeans, 0.1))])
        
        
        richnessOutflowList = (list(highOutflow=wellnessMetadata$GeneRichness[wellnessMetadata$sample.ID %in% highOutflowSamples],
                                    lowOutflow = wellnessMetadata$GeneRichness[wellnessMetadata$sample.ID %in% lowOutflowSamples]))
        
        richnessInflowList = (list(highInflow=wellnessMetadata$GeneRichness[wellnessMetadata$sample.ID %in% highInflowSamples],
                                   lowInflow = wellnessMetadata$GeneRichness[wellnessMetadata$sample.ID %in% lowInflowSamples]))
        
        richnessOutflowMelt = melt(richnessOutflowList)
        
        testList = list(highOutflow = scaledOutflowMeans[highOutflowSamples],
                        lowOutflow = scaledOutflowMeans[lowOutflowSamples])
        boxplot(testList)
        
        
        
        boxplot(richnessOutflowList)
        boxplot(richnessInflowList)
        
        
        plot(density(richnessList$highOutflow))
        lines(density(richnessList$lowOutflow),col="red")
        
        t.test(richnessOutflowList$highOutflow, richnessOutflowList$lowOutflow )
        t.test(richnessInflowList$highInflow, richnessInflowList$lowInflow )
        
        outflowDominant = names(scaledOutflowMeans[scaledOutflowMeans>5])
        
        out = apply(clinicalMatrix, 2, function(x) {
          highFeatures = x[rownames(clinicalMatrix) %in% highOutflowSamples]
          lowFeatures = x[rownames(clinicalMatrix) %in% lowOutflowSamples]
          return(c(mean(highFeatures)/mean(lowFeatures), t.test(highFeatures, lowFeatures)$p.value))
          
        })
        
        out = apply(clinicalMatrix, 2, function(x) {
          highFeatures = x[rownames(clinicalMatrix) %in% outflowDominant]
          lowFeatures = x[!rownames(clinicalMatrix) %in% outflowDominant]
          return(c(mean(highFeatures)-mean(lowFeatures), t.test(highFeatures, lowFeatures)$p.value))
          #return(c((mean(highFeatures)+1e-10)/(mean(lowFeatures)+1e-10), t.test(highFeatures, lowFeatures)$p.value))
          
        })
        
        
        out = t(out)
        colnames(out) = c("fold change", "pvalue")
        View(out)
        table(inflow[unique(allStatMeltSel$msp)]>0.3)
      }
      
      
    }
    
    if (F) {
      boxplot(list(allSpecies=inflowOut, oss=inflowOut[ossSpecies]))
      boxplot(list(allSpecies=outflowOut, oss=outflowOut[ossSpecies]), ylab="outflow")
      t.test(outflowOut, outflowOut[ossSpecies])
    }
    
    length(inflow)
    length(outflow)
    
    #statesMat = getColonizationSteadyStates(wellnessMetadata, wellnessMgsMat)
    #stationarity = getColonizationStationary(wellnessMetadata, wellnessMgsMat)
    
    colonizedRatios = (inflow+1e-4)/(outflow+1e-4)
    colonizedRatios[inflow==0 & outflow==0] = NA
    
    targetMsps = names(inflow)
    wellnessMgsPrevalence = apply(wellnessMgsMat, 1, function(currRow)mean(currRow>0))
    wellnessMgsPrevalence = wellnessMgsPrevalence[match(targetMsps, names(wellnessMgsPrevalence))]
    wellnessMgsMeans = rowMeans(wellnessMgsMat)
    wellnessMgsMeans = wellnessMgsMeans[match(targetMsps, names(wellnessMgsMeans))]
    freqs_wellness = freqs_wellness[match(targetMsps, names(freqs_wellness))]
    
    freqs_western = freqs_western[match(targetMsps, names(freqs_western))]
    freqs_nonWestern = freqs_nonWestern[match(targetMsps, names(freqs_nonWestern))]
    freqs_normal = freqs_normal[match(targetMsps, names(freqs_normal))]
    freqs_disease = freqs_disease[match(targetMsps, names(freqs_disease))]
    
    
    #mgsVisitCountFractions = mgsVisitCountFractions[match(targetMsps, names(mgsVisitCountFractions))]
    #mgsVisitCountFractions
    
    markovStatTabTemp= data.frame(inflow = inflowOut[,1],
                                  inflowUpper = inflowOut[,2],
                                  inflowLower = inflowOut[,3],
                                  inflowSE = inflowOut[,4],
                                  outflow = outflowOut[,1],
                                  outflowUpper = outflowOut[,2],
                                  outflowLower = outflowOut[,3],
                                  outflowSE = outflowOut[,4],
                                  prevalence=freqs_wellness, 
                                  meanAbd=wellnessMgsMeans,
                                  species = getSpeciesName(rownames(inflowOut),taxo),
                                  stringsAsFactors = F)
    
    markovStatTabTemp = markovStatTabTemp[markovStatTabTemp$prevalence>0.1,]
    markovStatTabTempInflow = markovStatTabTemp[markovStatTabTemp$inflow>=0.3 & markovStatTabTemp$outflow < 0.3, ]
    markovStatTabTempInflow = markovStatTabTempInflow[rownames(markovStatTabTempInflow)!="msp_0076",]
    markovStatTabTempInflow = markovStatTabTempInflow[!grepl("unclassified", markovStatTabTempInflow$species),]
    
    #"C:/Data/comparative.analysis.healthy.sweden"
    
    markovStatTabTempOutflow = markovStatTabTemp[markovStatTabTemp$outflow>=0.3 & markovStatTabTemp$inflow < 0.3, ]
    markovStatTabTempOutflow = markovStatTabTempOutflow[!grepl("unclassified", markovStatTabTempOutflow$species),]
    
    write.table(markovStatTabTempInflow, "markovStatTabTempInflow.txt", sep="\t")
    write.table(markovStatTabTempOutflow, "markovStatTabTempOutflow.txt", sep="\t")
    
    markovStatTab = data.frame(inflow = inflowOut[,1],
                               inflow_5m = inflow_5m,
                               inflow_15m = inflow_15m,
                               inflowUpper = inflowOut[,2],
                               inflowLower = inflowOut[,3],
                               inflowSE = inflowOut[,4],
                               outflow = outflowOut[,1],
                               outflow_5m = outflow_5m, 
                               outflow_15m = outflow_15m,
                               outflowUpper = outflowOut[,2],
                               outflowLower = outflowOut[,3],
                               outflowSE = outflowOut[,4],
                               ratio=colonizedRatios,
                               freq.wellness=freqs_wellness, 
                               freq.normal=freqs_normal,
                               freq.disease=freqs_disease,
                               freq.western=freqs_western, 
                               freq.nonWestern=freqs_nonWestern, 
                               
                               meanAbd=wellnessMgsMeans,
                               surv.rate.wellness.6mon = survivalStats$surv.rates[match(rownames(inflowOut), survivalStats$mgs)],
                               surv.rate.american.6mon = survivalStats_american_6mon$surv.rates[match(rownames(inflowOut), survivalStats_american_6mon$mgs)],
                               surv.rate.dinamic.1mon = survivalStats_DINAMIC_1mon$surv.rates[match(rownames(inflowOut), survivalStats_DINAMIC_1mon$mgs)],
                               
                               #freq=mgsVisitCountFractions,
                               type=rep("transient",length(wellnessMgsPrevalence)),
                               species = getSpeciesName(rownames(inflowOut),taxo),
                               stringsAsFactors = F)
    
    markovStatTab= data.frame(inflow = inflowOut[,1],
                              inflow_5m = inflow_5m,
                              inflow_15m = inflow_15m,
                              inflowUpper = inflowOut[,2],
                              inflowLower = inflowOut[,3],
                              inflowSE = inflowOut[,4],
                              outflow = outflowOut[,1],
                              outflow_5m = outflow_5m, 
                              outflow_15m = outflow_15m,
                              outflowUpper = outflowOut[,2],
                              outflowLower = outflowOut[,3],
                              outflowSE = outflowOut[,4],
                              ratio=colonizedRatios,
                              prevalence=freqs_wellness, 
                              meanAbd=wellnessMgsMeans,
                              #surv.rate = survivalStats$surv.rates[match(rownames(inflowOut), survivalStats$mgs)],
                              
                              #freq=mgsVisitCountFractions,
                              type=rep("transient",length(wellnessMgsPrevalence)),
                              species = getSpeciesName(rownames(inflowOut),taxo),
                              stringsAsFactors = F)
    
    markovStatTab$type[markovStatTab$freq.wellness <= 0.1] = "spontaneous"
    markovStatTab$type[markovStatTab$freq.wellness >= 0.9] = "colonized"
    markovStatTab = markovStatTab[rownames(markovStatTab) != "msp_0076",] #blautia wexelerae was present all samples, so we couldn't feed this species into model
    
    #markovStatTabFinalFile = "C://Data//comparative.analysis.healthy.sweden//markovStatTab.final.3.txt"
    #markovStatTabFinalFile = "C://Data//comparative.analysis.healthy.sweden//markovStatTab.final.4.txt"
    #markovStatTabFinalFile = "C://Data//comparative.analysis.healthy.sweden//markovStatTab.final.5.txt"
    markovStatTabFinalFile = "C://Data//comparative.analysis.healthy.sweden//markovStatTab.final.6.2020.0803.txt"
    #write.table(markovStatTab, markovStatTabFinalFile, sep="\t")
    
    
    ggplot(markovStatTab, aes(x=outflow, y=inflow, fill=inflow)) + 
      scale_fill_gradient2(midpoint = 0.2, low="#FFD70044", mid = "white", high = "purple") +
      geom_point(size=3, shape=21, colour="#d3d3d366")+ xlab("") + ylab("")+
      geom_vline(xintercept = 0.3, colour="gray",linetype="dashed")+
      geom_hline(yintercept = 0.3, colour="gray",linetype="dashed")+
      #geom_abline(slope=1, intercept = 0)+
      theme(panel.grid.major = element_blank(), 
            panel.grid.minor = element_blank(),
            legend.position = "none",
            text = element_text(size = 15),
            panel.background = element_rect(fill = "white", 
                                            colour = "black", 
                                            size = 0.5, 
                                            linetype = "solid"))
    
    ggplot(markovStatTab[markovStatTab$type=="transient",], aes(x=outflow, y=inflow, fill=inflow)) + 
      scale_fill_gradient2(midpoint = 0.2, low="#FFD70044", mid = "white", high = "purple") +
      geom_point(size=2, shape=21, colour="black")+ xlab("") + ylab("")+
      geom_vline(xintercept = 0.3, colour="gray",linetype="dashed")+
      geom_hline(yintercept = 0.3, colour="gray",linetype="dashed")+
      #geom_abline(slope=1, intercept = 0)+
      theme(panel.grid.major = element_blank(), 
            panel.grid.minor = element_blank(),
            legend.position = "none",
            text = element_text(size = 15),
            panel.background = element_rect(fill = "white", 
                                            colour = "black", 
                                            size = 0.5, 
                                            linetype = "solid"))
    if (F) {
      markovStatTabTransient = markovStatTab[rownames(markovStatTab) %in% transient,]
      
      abdCompTab = data.frame(markovStatTabTransient$ratio, 
                              markovStatTabTransient$meanAbd,
                              msp = rownames(markovStatTabTransient),
                              species = getSpeciesName(rownames(markovStatTabTransient), taxo),
                              checked=NA,
                              row.names =  rownames(markovStatTabTransient),
                              stringsAsFactors = F)
      
      checkMsps = c("msp_0291","msp_0414", "msp_0910", "msp_0199", "msp_0333", "msp_0044", "msp_0086", "msp_0263", "msp_0071", "msp_0166", "msp_0893", "msp_0007", "msp_0833", "msp_0005", "msp_0025")
      abdCompTab$checked[abdCompTab$msp %in% checkMsps] = T
      
      cols = rep("#80808055", dim(abdCompTab)[1])
      names(cols) = abdCompTab$msp
      cols[checkMsps]  = "#ff000099"
      
      plot(markovStatTabTransient$ratio, 
           markovStatTabTransient$meanAbd, 
           col=cols,pch=16,xlab="inflow/outflow ratio",ylab="Mean abundance",
           log="xy")
      
    }
    
    checkFunctionContrast = F
    if (checkFunctionContrast) {
      
      loadFunction = T
      if (loadFunction) {
        funcMatRData = "j://Deposit/Project/2018_microbiome_atlas/functional.annotation/funcMat.20190808.RData"
        load(funcMatRData)
        
        loadAnnots = T
        if (loadAnnots) { # seven types of functional/phenotypic annotations
          
          secretionClusters = c("8","120","425","250")
          mucinCazymes = c("cazy.GT27", "cazy.GT14", "cazy.GT11", "cazy.GT10", "cazy.GH89", "cazy.GH84", "cazy.GH33", "cazy.GH20", "cazy.GH29", "cazy.GH18", "cazy.GH123", "cazy.GH109", "cazy.CBM51", "cazy.CBM50", "cazy.CBM32")
          storageCazymes = c("cazy.CBM34", "cazy.CBM48", "cazy.CBM20", "cazy.GH13", "cazy.GH133", "cazy.GH32", "cazy.GH36", "cazy.GH37", "cazy.GH77", "cazy.GH57", "cazy.GT101", "cazy.GT26", "cazy.GT3", "cazy.GT35", "cazy.GT5")
          pectinCazymes = c("cazy.CBM77","cazy.CE8","cazy.GH105","cazy.GH28", "cazy.PL1")
          
          
          ### virulence factors ###
          vfMatRData = "j://Deposit/Project/2018_microbiome_atlas/functional.annotation/PATRIC/vfBestMat.RData"
          load(vfMatRData) #vfMatDigit
          vfTerms = colnames(vfMatDigit)
          patricVfDescTabFile = "J://Deposit/Project/2018_microbiome_atlas/functional.annotation/PATRIC/PATRIC_sp_gene.txt"
          patricVfDescTab = read.table(patricVfDescTabFile, header=T, sep="\t", stringsAsFactors = F)
          
          ### JGI phenotype ###
          jgiMatRData = "j://Deposit/Project/2018_microbiome_atlas/functional.annotation/jgiMat.RData"
          load(jgiMatRData)
          jgiTerms = colnames(jgiMat)
          
          ### reactions ###
          gemRData = "J://Deposit/Project/2018_microbiome_atlas//atlas.GEM.model.related/GEM.from.Reza.amazing/absentPresentInModels.RData"
          load(gemRData)
          reactionTerms = colnames(gemMat)
          
          ### antibiotic resistance ###
          mustardMatRData = "j://Deposit/Project/2018_microbiome_atlas/functional.annotation/mustard.mat.RData"
          load(mustardMatRData)
          arTerms = colnames(mustardMat)
          
          ### antismash - secondary metabolism ###
          antismashMatRData = "J://Deposit/Project/2018_microbiome_atlas/functional.annotation/msp1992.igc2.antismashMat.RData"
          load(antismashMatRData)
          antismashMat[1:3,1:3]
          antismashMelt = melt(antismashMat)
          head(antismashMelt)
          antismashMelt = antismashMelt[antismashMelt$value!=0,]
          dim(antismashMelt)
          
          writeMode = F
          if (writeMode) write.table(antismashMelt, "antismash.list.txt", sep="\t")
          antismashTerms = colnames(antismashMat)
          
          ### Cazy  ### 
          newGutCazyMatRData = "J://Deposit/Project/2018_microbiome_atlas/functional.annotation/CAZY.INRA/igc2.new.cazy.mat.RData"
          load(newGutCazyMatRData)
          colnames(cazyMat) = paste("cazy", colnames(cazyMat), sep=".")
          cazyTerms = colnames(cazyMat)
          
          ### PFAM ###
          igc2PfamMatRData = "j://Deposit/Project/2018_microbiome_atlas/functional.annotation/PFAM/igc2PfamMat.RData"
          load(igc2PfamMatRData)
          pfamTerms = colnames(pfamMat)
          
          ### KEGG ###
          gutKoBestMatRData = "j://Deposit/Project/2018_microbiome_atlas/functional.annotation/gutKoBestMat.1990.RData"
          load(gutKoBestMatRData)
          koTerms = colnames(gutKoBestMat)
          
          ### KEGG pathway ###
          # already loaded from above scripts
          # data: bacteriaKoElementList
          
          ### KEGG modules ###
          keggModuleDescTabRData = "J://Deposit/Project/2018_microbiome_atlas/functional.annotation/KEGG module/moduleDescTab.RData"
          koModuleListRData = "J://Deposit/Project/2018_microbiome_atlas/functional.annotation/KEGG module/koModuleList.20190219.RData"
          load(keggModuleDescTabRData) #moduleDescTab
          microbeModuleTab = moduleDescTab[moduleDescTab$isBacteriaModule|moduleDescTab$isArchaeaModule,]
          nonMicrobeModuleTab = moduleDescTab[!(moduleDescTab$isBacteriaModule|moduleDescTab$isArchaeaModule),]
          nonMicrobeModuleTab2 = moduleDescTab[grepl("eukaryotes",moduleDescTab$moduleName),]
          
          microbeModules = paste(microbeModuleTab$moduleId, microbeModuleTab$moduleName, sep = ":")
          nonMicrobeModules = paste(nonMicrobeModuleTab$moduleId, nonMicrobeModuleTab$moduleName, sep = ":")
          nonMicrobeModules2 = paste(nonMicrobeModuleTab2$moduleId, nonMicrobeModuleTab2$moduleName, sep = ":")
          nonMicrobeModules = unique(c(nonMicrobeModules,nonMicrobeModules2))
          load(koModuleListRData)
          
        }
        
      }
      
      transientInSpecies = rownames(markovStatTabTransient[markovStatTabTransient$ratio>1,])
      transientOutSpecies = rownames(markovStatTabTransient[markovStatTabTransient$ratio<=1,])
      
      inSpecies = c(colonized, transientInSpecies)
      outSpecies = c(passerby, transientOutSpecies)
      
      inOutFuncScores = getVarScoreMatTwo(funcMat, cazyTerms, inSpecies, outSpecies)
      selected = selectVarScoreMatTwo(inOutFuncScores)
      View(selected[selected$term %in% mucinCazymes,])
      View(selected[selected$term %in% storageCazymes,])
      inOutFuncScores = getVarScoreMatTwo(funcMat, arTerms, inSpecies, outSpecies)
      selectedAr = selectVarScoreMatTwo(inOutFuncScores)
      
      inOutFuncScores = getVarScoreMatTwo(funcMat, pfamTerms, inSpecies, outSpecies)
      selectedPfam = selectVarScoreMatTwo(inOutFuncScores)
      View(selectedPfam[selectedPfam$term %in% mgePfamTerms,])
      
      
    }
    
    checkFlowsWithAbundancePrevalenceBins = F
    if (checkFlowsWithAbundancePrevalenceBins) {
      
      manualCutFlows <- function(markovStatTab, target="inflow", by="prevalence", breaks =c(0,0.2, 0.4, 0.6, 0.8,1)) {
        cuts = cut(markovStatTab[,by], breaks=breaks)
        cutLevels = levels(cuts)
        targetBins = split(markovStatTab[,target], cuts)
        targetBins = targetBins[cutLevels]
        return(targetBins)
      }
      getBinTabForMarkov <- function(markovStatTab, by="meanAbd", target="inflow") {
        abdMin = min(markovStatTab[,by])
        abd02 = quantile(markovStatTab[,by], 0.25)
        abd04 = quantile(markovStatTab[,by], 0.5)
        abd06 = quantile(markovStatTab[,by], 0.75)
        abdMax = max(markovStatTab[,by])
        
        print(abdMin)
        print(abd02)
        print(abd04)
        print(abd06)
        print(abdMax)
        
        markovTab00 = getMarkovStatFrom(markovStatTab, abdMin, abd02, by, target)
        markovTab02 = getMarkovStatFrom(markovStatTab, abd02, abd04, by, target)
        markovTab04 = getMarkovStatFrom(markovStatTab, abd04, abd06, by, target)
        markovTab06 = getMarkovStatFrom(markovStatTab, abd06, abdMax, by, target)
        
        markovListByAbd = list(per25=markovTab00,
                               per50=markovTab02,
                               per75=markovTab04,
                               per100=markovTab06)
        
        return(markovListByAbd)
      }
      
      prevalenceMode = T
      if (prevalenceMode) {
        inflowBins = manualCutFlows(markovStatTab, target="inflow" )
        boxplot(inflowBins)
        
        outflowBins = manualCutFlows(markovStatTab, target="outflow")
        boxplot(outflowBins)
      }
      
      abundanceMode = T
      if (abundanceMode) {
        par(mar=c(5,4,4,1))
        
        inflowBins = getBinTabForMarkov(markovStatTab, by="meanAbd", target="inflow" )
        boxplot(inflowBins,las=2)
        t.test(inflowBins$per25, inflowBins$per50)
        t.test(inflowBins$per50, inflowBins$per75)
        t.test(inflowBins$per75, inflowBins$per100)
        
        outflowBins = getBinTabForMarkov(markovStatTab, by="meanAbd", target="outflow")
        boxplot(outflowBins,las=2)
        t.test(outflowBins$per25, outflowBins$per50)
        t.test(outflowBins$per50, outflowBins$per75)
        t.test(outflowBins$per75, outflowBins$per100)
        
        table(markovStatTab$ratio>1)
        ggplot(markovStatTab, aes(x=outflow, y=inflow, fill=inflow)) + 
          scale_fill_gradient2(midpoint = 0.2, low="#FFD70044", mid = "white", high = "purple") +
          geom_point(size=2, shape=21, colour="black")+ xlab("") + ylab("")+
          #geom_vline(xintercept = 0.5)+geom_hline(yintercept = 0.5)+
          geom_abline(slope=1, intercept = 0)+
          theme(panel.grid.major = element_blank(), 
                panel.grid.minor = element_blank(),
                legend.position = "none",
                panel.background = element_rect(fill = "white", 
                                                colour = "black", 
                                                size = 0.5, 
                                                linetype = "solid"))
        
      }
      
    }
    
    pieChartMode = F
    if (pieChartMode) {
      iniR = 0.3
      pie(1, radius=0.01, init.angle=90, col=c('white'), border = NA, labels='')
      
      floating.pie(0,0, c(33, 475, 905), #c(16, 370, 1571), 
                   radius=3*iniR, 
                   startpos=pi/2, 
                   col = c(col_vector[3:4],"gray"),
                   border="black")
      floating.pie(0,0,1,  
                   radius=1.5*iniR, 
                   startpos=pi/2, 
                   col="white",
                   border="black")
      floating.pie(0,0,1,  
                   radius=1.49*iniR, 
                   startpos=pi/2, 
                   col="white",
                   border=NA)
      
      
    }
    
  }
  
  inOutFlowFuncCheckModeOld = F
  if (inOutFlowFuncCheckModeOld) {
    
    funcMatRData = "j://Deposit/Project/2018_microbiome_atlas/functional.annotation/funcMat.20190808.RData"
    
    selectMode = T
    if (selectMode) {
      inflowSpecies = rownames(markovStatTab[markovStatTab$inflow>=0.25,])
      outflowSpecies = rownames(markovStatTab[markovStatTab$outflow>=0.25,])
      length(inflowSpecies)
      length(outflowSpecies)
    }
    
    loadFuncMode = T
    if (loadFuncMode) {
      load(funcMatRData)
      
      funcMatTranspose = t(funcMat)
      # hgcFuncMat = funcMatTranspose[,colnames(funcMatTranspose) %in% hgcSpecies]
      # lgcFuncMat = funcMatTranspose[,colnames(funcMatTranspose) %in% lgcSpecies]
      inFuncMat = funcMatTranspose[,colnames(funcMatTranspose) %in% inflowSpecies]
      outFuncMat = funcMatTranspose[,colnames(funcMatTranspose) %in% outflowSpecies]
      
      colonizingFuncMat = funcMatTranspose[,colnames(funcMatTranspose) %in% colonizingSpecies]
      resistantFuncMat = funcMatTranspose[,colnames(funcMatTranspose) %in% resistantSpecies]
      
    }
    
    checkInOutFuncDiff = T
    if (checkInOutFuncDiff) {
      
      funcs = rownames(funcMatTranspose)
      pvalues = sapply(funcs, function(currFunc) {
        inCurrFunc = inFuncMat[currFunc,]
        outCurrFunc = outFuncMat[currFunc,]
        
        inPresent = sum(inCurrFunc==1)
        inAbsent = sum(inCurrFunc==0)
        
        outPresent = sum(outCurrFunc==1)
        outAbsent = sum(outCurrFunc==0)
        
        contigTab = cbind(Inflow=c(present=inPresent, absent=inAbsent),
                          Outflow=c(present=outPresent, absent=outAbsent))
        pvalue=chisq.test(contigTab)$p.value
        return(pvalue)
      }, simplify = F)
      pvalues = unlist(pvalues)
      
      oddsRatios = sapply(funcs, function(currFunc) {
        inCurrFunc = inFuncMat[currFunc,]
        outCurrFunc = outFuncMat[currFunc,]
        
        inPresent = sum(inCurrFunc==1)
        inAbsent = sum(inCurrFunc==0)
        
        outPresent = sum(outCurrFunc==1)
        outAbsent = sum(outCurrFunc==0)
        
        oddsRatio = (inPresent*outAbsent)/(inAbsent*outPresent)
        if (outPresent*inAbsent == 0) return(Inf) 
        return(oddsRatio)
      }, simplify = F)
      oddsRatios = unlist(oddsRatios)
      
      chisqStatTab = data.frame(term=names(pvalues),
                                pvalue=pvalues,
                                oddsRatio=oddsRatios,
                                logP = -log10(pvalues),
                                logOdds = log(oddsRatios))
      
      #View(chisqStatTab[grepl("cazy",rownames(chisqStatTab)),])
      #View(chisqStatTab[chisqStatTab$pvalue<0.05,])
      View(chisqStatTab[mgePfamTerms,])
    }
    
    checkColonizingResistantFuncDiff = T
    if (checkColonizingResistantFuncDiff) {
      funcs = rownames(funcMatTranspose)
      pvalues = sapply(funcs, function(currFunc) {
        colCurrFunc = colonizingFuncMat[currFunc,]
        resCurrFunc = resistantFuncMat[currFunc,]
        
        colPresent = sum(colCurrFunc==1)
        colAbsent = sum(colCurrFunc==0)
        
        resPresent = sum(resCurrFunc==1)
        resAbsent = sum(resCurrFunc==0)
        
        contigTab = cbind(Inflow=c(present=colPresent, absent=colAbsent),
                          Outflow=c(present=resPresent, absent=resAbsent))
        pvalue=chisq.test(contigTab)$p.value
        return(pvalue)
      }, simplify = F)
      pvalues = unlist(pvalues)
      
      oddsRatios = sapply(funcs, function(currFunc) {
        colCurrFunc = colonizingFuncMat[currFunc,]
        resCurrFunc = resistantFuncMat[currFunc,]
        
        colPresent = sum(colCurrFunc==1)
        colAbsent = sum(colCurrFunc==0)
        
        resPresent = sum(resCurrFunc==1)
        resAbsent = sum(resCurrFunc==0)
        
        oddsRatio = (colPresent*resAbsent)/(colAbsent*resPresent)
        if (resPresent*colAbsent == 0) return(Inf) 
        return(oddsRatio)
      }, simplify = F)
      oddsRatios = unlist(oddsRatios)
      
      chisqStatTab = data.frame(term=names(pvalues),
                                pvalue=pvalues,
                                oddsRatio=oddsRatios,
                                logP = -log10(pvalues),
                                logOdds = log(oddsRatios))
      
      #View(chisqStatTab[grepl("cazy",rownames(chisqStatTab)),])
      #View(chisqStatTab[chisqStatTab$pvalue<0.05,])
      View(chisqStatTab[mgePfamTerms,])
    } 
    
    
  }
  
  checkOverallFlowMode = T
  if (checkOverallFlowMode) {
    
    loadData = F
    if (loadData) {
      wellnessMetadata = basicMetaMapUpdated[basicMetaMapUpdated$dataset.ID=="id28",]
      
      mgsMat = mergeMatUpdated
      metaTab = basicMetaMapUpdated
      
      wellnessMetadata = metaTab[metaTab$dataset.ID=="id28",]
      wellnessSubjs = (wellnessMetadata$metadata.ID)
      missingSubjs1 = names(table(wellnessSubjs)[table(wellnessSubjs)==3])
      exceptionSubjs = c("X3452","X3834","X3918","X4215")
      missingSubjs2 = c("X3913","X3535","X3016")
      missingSubjs1 = missingSubjs1[!missingSubjs1 %in% exceptionSubjs]
      
      wellnessMetadata = wellnessMetadata[!wellnessMetadata$metadata.ID %in% missingSubjs1, ]
      wellnessMetadata = wellnessMetadata[!wellnessMetadata$metadata.ID %in% missingSubjs2, ]
      wellnessMetadata = wellnessMetadata[!wellnessMetadata$metadata.ID %in% exceptionSubjs, ]
      wellnessMgsMat = mergeMatUpdated[,match(wellnessMetadata$sample.ID, colnames(mergeMatUpdated))]
      wellnessMgsMatDigit = (wellnessMgsMat>0) * 1
      wellnessMgsMeans = rowMeans(wellnessMgsMat)
      wellnessSamples = wellnessMetadata$sample.ID
      wellnessSubjects = gsub("_v[1-4]","",wellnessSamples)
      wellnessVisits = gsub("X[0-9]+_","",wellnessSamples)
      wellnessSamplesBySubjects = split(wellnessSamples, wellnessSubjects)
      wellnessUniqSubjs = unique(wellnessMetadata$metadata.ID)
      wellnessClinicalMatrix = t(clinicalMatrix)[,match(wellnessSamples, rownames(clinicalMatrix))]
      
    }
    
    checkCoefficientOfVariation = T
    if (checkCoefficientOfVariation) {
      freqs_wellness = rowMeans(wellnessMgsMat>0)
      
      quantile(freqs_wellness[freqs_wellness>0.1 & freqs_wellness<0.9], 0.333)
      quantile(freqs_wellness[freqs_wellness>0.1 & freqs_wellness<0.9], 0.666)
      
    }
    
    loadGeoSignal = T
    if (loadGeoSignal) {
      
      specificGroupCountriesMeltRData = "J://Deposit//Project//2018_microbiome_atlas//atlas.geography//specificGroupCountriesMelt.RData"
      specificGroupMgsByCountriesRData = "J://Deposit//Project//2018_microbiome_atlas//atlas.geography//specificGroupMgsByCountries.RData"
      specificGroupMgsAllRData = "J://Deposit//Project//2018_microbiome_atlas//atlas.geography//specificGroupMgsAll.RData"
      specificGroupMgsCountsByCountriesRData = "J://Deposit//Project//2018_microbiome_atlas//atlas.geography//specificGroupMgsCountsByCountries.RData" 
      
      load(specificGroupCountriesMeltRData)
      load(specificGroupMgsByCountriesRData)
      load(specificGroupMgsAllRData)
      load(specificGroupMgsCountsByCountriesRData)
      
      swedSpecies = specificGroupMgsByCountries$Sweden
      peruSpecies = specificGroupMgsByCountries$Peru
      japanSpecies = specificGroupMgsByCountries$Japan
      spainSpecies = specificGroupMgsByCountries$Spain
      
      wellnessSwedenMgsMatDigit = wellnessMgsMatDigit[swedSpecies,]
      wellnessJapanMgsMatDigit = wellnessMgsMatDigit[japanSpecies,]
      wellnessSpainMgsMatDigit = wellnessMgsMatDigit[spainSpecies,]
      wellnessPeruMgsMatDigit = wellnessMgsMatDigit[peruSpecies,]
      
      
    }
    
    ### 86 subjects ###
    if (T) {
      seqList = list(c("v1","v2"),
                     c("v2","v3"),
                     c("v3","v4"),
                     c("v1","v3"),
                     c("v1","v4"),
                     
                     c("v2","v1"),
                     c("v3","v2"),
                     c("v4","v3"))
      
      seqList = list(c("v1","v2"),
                     c("v2","v3"),
                     c("v3","v4"))#
      
      seqList = list(c("v2","v1"),
                     c("v3","v2"),
                     c("v4","v3"))#
      
      
      
    }
    
    checkCommonInEachSeason = T
    if (checkCommonInEachSeason) {
      seqs = c("v1","v2","v3","v4")
      targetDigit = wellnessMgsMatDigit 
      
      checkCommonFraction <- function(seqs, targetDigit, cutRatio=0.9) {
        commonMat = t(sapply(seqs, function(currSeq) {
          cutNum = length(wellnessUniqSubjs)*cutRatio
          subjs_per_visit = paste(wellnessUniqSubjs, currSeq, sep = "_")
          visitSum=rowSums(targetDigit[,subjs_per_visit])
          commonSpecies = names(visitSum[visitSum >= cutNum])
          inflowFraction = mean(commonSpecies %in% inflowSpecies)
          outflowFraction = mean(commonSpecies %in% outflowSpecies)
          return(c(inflowFraction, outflowFraction, 1-(inflowFraction+outflowFraction)))
        }))
        colnames(commonMat) = c("inflowSpecies", "outflowSpecies","others")
        commonMat = commonMat*100
        return(commonMat)
      }
      checkCommonFraction(seqs, targetDigit, 0.9)
      checkCommonFraction(seqs, targetDigit, 0.5)
      
      
      checkCommonList = sapply(seqs, function(currSeq) {
        subjs_per_visit = paste(wellnessUniqSubjs, currSeq, sep = "_")
        visitSum=rowSums(targetDigit[,subjs_per_visit])
        print(sort(visitSum))[1:5]
        return(visitSum)
      })
      
      
    }
    
    checkCommonAcrossSeason = T
    if (checkCommonAcrossSeason) {
      seqs = c("v1","v2","v3","v4")
      targetDigit = wellnessMgsMatDigit 
      
      checkCommonAcrossSeasonFraction <- function(seqs, targetDigit, cutNum=3) {
        commonMat = t(sapply(wellnessUniqSubjs, function(currSubj){
          subjs_over_visits = paste(currSubj, seqs, sep = "_")
          visitSum=rowSums(targetDigit[,subjs_over_visits])
          commonSpecies = names(visitSum[visitSum >= cutNum])
          inflowFraction = mean(commonSpecies %in% inflowSpecies)
          outflowFraction = mean(commonSpecies %in% outflowSpecies)
          return(c(inflowFraction, outflowFraction, 1-(inflowFraction+outflowFraction)))
        }))
        colnames(commonMat) = c("inflowSpecies", "outflowSpecies","others")
        commonMat = commonMat*100
        return(commonMat)
      }
      checkRareAcrossSeasonFraction <- function(seqs, targetDigit, cutNum=2) {
        rareMat = t(sapply(wellnessUniqSubjs, function(currSubj){
          subjs_over_visits = paste(currSubj, seqs, sep = "_")
          visitSum=rowSums(targetDigit[,subjs_over_visits])
          rareSpecies = names(visitSum[visitSum <= cutNum])
          inflowFraction = mean(rareSpecies %in% inflowSpecies)
          outflowFraction = mean(rareSpecies %in% outflowSpecies)
          return(c(inflowFraction, outflowFraction, 1-(inflowFraction+outflowFraction)))
        }))
        colnames(rareMat) = c("inflowSpecies", "outflowSpecies","others")
        rareMat = rareMat*100
        return(rareMat)
      }
      
      commonMat = checkCommonAcrossSeasonFraction(seqs, targetDigit, 4)
      rareMat = checkRareAcrossSeasonFraction(seqs, targetDigit, 1)
      par(mfrow=c(2,1))
      par(mar=c(2,4,2,1))
      barplot(colMeans(commonMat), col=c("purple","gold","gray"), las=1, main="common")
      barplot(colMeans(rareMat), col=c("purple","gold","gray"), las=1, main="unique")
      
      boxplot(commonMat)
      boxplot(rareMat)
      
      
      
      checkCommonList = sapply(seqs, function(currSeq) {
        subjs_per_visit = paste(wellnessUniqSubjs, currSeq, sep = "_")
        visitSum=rowSums(targetDigit[,subjs_per_visit])
        print(sort(visitSum))[1:5]
        return(visitSum)
      })
      
      
    }
    
    wellnessMgsMatDigit_outflowSpecies = wellnessMgsMatDigit[rownames(wellnessMgsMatDigit) %in% outflowSpecies,]
    wellnessMgsMatDigit_inflowSpecies =  wellnessMgsMatDigit[rownames(wellnessMgsMatDigit) %in% inflowSpecies,]
    
    seqList = list(c("v1","v2"),
                   c("v2","v3"),
                   c("v3","v4"))#
    
    targetDigit = wellnessMgsMatDigit_outflowSpecies
    targetDigit = wellnessMgsMatDigit_inflowSpecies
    targetDigit = wellnessMgsMatDigit 
    
    seqFractionList = lapply(seqList, function(currSeq) {
      
      fractions = sapply(wellnessUniqSubjs, function(currSubj) {
        subj_visit = paste(currSubj, currSeq, sep = "_")
        if (!all(subj_visit %in% colnames(wellnessMgsMat))) return(NA)
        visit1=targetDigit[,subj_visit[1]]
        visit2=targetDigit[,subj_visit[2]]
        visit1=visit1[visit1==1]
        
        fraction = mean(visit2[names(visit1)])
        return(fraction)
      })
      
      return(fractions)
      
      
    })
    names(seqFractionList) = c("v1-v2", "v2-v3", "v3-v4")
    
    seqList = list(c("v2","v1"),
                   c("v3","v2"),
                   c("v4","v3"))#
    
    targetDigit = wellnessMgsMatDigit_inflowSpecies
    targetDigit = wellnessMgsMatDigit 
    targetDigit = wellnessMgsMatDigit_outflowSpecies
    
    
    seqFractionList = lapply(seqList, function(currSeq) {
      
      fractions = sapply(wellnessUniqSubjs, function(currSubj) {
        subj_visit = paste(currSubj, currSeq, sep = "_")
        if (!all(subj_visit %in% colnames(wellnessMgsMat))) return(NA)
        visit1=targetDigit[,subj_visit[1]]
        visit2=targetDigit[,subj_visit[2]]
        visit1=visit1[visit1==1]
        
        fraction = mean(visit2[names(visit1)])
        return(fraction)
      })
      
      return(fractions)
      
      
    })
    
    names(seqFractionList) = c("v2-v1", "v3-v2", "v4-v3")
    #seqFractionList=lapply(seqFractionList, function(x)100*x)
    
    seqList = list(c("v1","v2"),
                   c("v2","v3"),
                   c("v3","v4"))#
    
    checkInflowFractionList = lapply(seqList, function(currSeq) {
      
      fractions = sapply(wellnessUniqSubjs, function(currSubj) {
        subj_visit = paste(currSubj, currSeq, sep = "_")
        if (!all(subj_visit %in% colnames(wellnessMgsMat))) return(NA)
        visit1=targetDigit[,subj_visit[1]]
        visit2=targetDigit[,subj_visit[2]]
        visit1=visit1[visit1==1]
        
        visit1Species = names(visit1)
        visit2Species = names(visit2[visit2==1])
        visit12Species = visit1Species[visit1Species %in% visit2Species]
        visit1_2species = visit1Species[!visit1Species %in% visit2Species]
        #print(visit1_2species)
        return(mean(visit1_2species %in% inflowSpecies))
        fraction = mean(visit2[names(visit1)])
        return(fraction)
      })
      
      return(fractions)
      
      
    })
    
    checkOutflowFractionList = lapply(seqList, function(currSeq) {
      
      fractions = sapply(wellnessUniqSubjs, function(currSubj) {
        subj_visit = paste(currSubj, currSeq, sep = "_")
        if (!all(subj_visit %in% colnames(wellnessMgsMat))) return(NA)
        visit1=targetDigit[,subj_visit[1]]
        visit2=targetDigit[,subj_visit[2]]
        visit1=visit1[visit1==1]
        
        visit1Species = names(visit1)
        visit2Species = names(visit2[visit2==1])
        visit12Species = visit1Species[visit1Species %in% visit2Species]
        visit1_2species = visit1Species[!visit1Species %in% visit2Species]
        return(mean(visit1_2species %in% outflowSpecies))
        
        fraction = mean(visit2[names(visit1)])
        return(fraction)
      })
      
      return(fractions)
      
      
    })
    
    checkInflowSharedFractionList = lapply(seqList, function(currSeq) {
      
      fractions = sapply(wellnessUniqSubjs, function(currSubj) {
        subj_visit = paste(currSubj, currSeq, sep = "_")
        if (!all(subj_visit %in% colnames(wellnessMgsMat))) return(NA)
        visit1=targetDigit[,subj_visit[1]]
        visit2=targetDigit[,subj_visit[2]]
        visit1=visit1[visit1==1]
        
        visit1Species = names(visit1)
        visit2Species = names(visit2[visit2==1])
        visit12Species = visit1Species[visit1Species %in% visit2Species]
        visit1_2species = visit1Species[!visit1Species %in% visit2Species]
        return(mean(visit12Species %in% inflowSpecies))
        fraction = mean(visit2[names(visit1)])
        return(fraction)
      })
      
      return(fractions)
      
      
    })
    
    checkOutflowSharedFractionList = lapply(seqList, function(currSeq) {
      
      fractions = sapply(wellnessUniqSubjs, function(currSubj) {
        subj_visit = paste(currSubj, currSeq, sep = "_")
        if (!all(subj_visit %in% colnames(wellnessMgsMat))) return(NA)
        visit1=targetDigit[,subj_visit[1]]
        visit2=targetDigit[,subj_visit[2]]
        visit1=visit1[visit1==1]
        
        visit1Species = names(visit1)
        visit2Species = names(visit2[visit2==1])
        visit12Species = visit1Species[visit1Species %in% visit2Species]
        visit1_2species = visit1Species[!visit1Species %in% visit2Species]
        return(mean(visit12Species %in% outflowSpecies))
        
        fraction = mean(visit2[names(visit1)])
        return(fraction)
      })
      
      return(fractions)
      
      
    })
    
    barplotFraction <- function(inflowFractionList, outflowFractionList, ind) {
      meanInflowFraction = unlist(lapply(inflowFractionList, mean))
      meanOutflowFraction = unlist(lapply(outflowFractionList, mean))
      barplot(c(meanInflowFraction[ind],meanOutflowFraction[ind], 1-(meanInflowFraction[ind]+ meanOutflowFraction[ind]))*100, 
              col=c("purple","gold","gray"),las=1, 
              ylim=c(0,60))
    }
    barplotFraction(checkInflowFractionList, checkOutflowFractionList, 1)
    barplotFraction(checkInflowSharedFractionList, checkOutflowSharedFractionList, 2)
    seqList = list(c("v2","v1"),
                   c("v3","v2"),
                   c("v4","v3"))#
    
    checkInflowFractionList = lapply(seqList, function(currSeq) {
      
      fractions = sapply(wellnessUniqSubjs, function(currSubj) {
        subj_visit = paste(currSubj, currSeq, sep = "_")
        if (!all(subj_visit %in% colnames(wellnessMgsMat))) return(NA)
        visit1=targetDigit[,subj_visit[1]]
        visit2=targetDigit[,subj_visit[2]]
        visit1=visit1[visit1==1]
        
        visit1Species = names(visit1)
        visit2Species = names(visit2[visit2==1])
        visit12Species = visit1Species[visit1Species %in% visit2Species]
        visit1_2species = visit1Species[!visit1Species %in% visit2Species]
        print(visit1_2species)
        return(mean(visit1_2species %in% inflowSpecies))
        fraction = mean(visit2[names(visit1)])
        return(fraction)
      })
      
      return(fractions)
      
      
    })
    
    checkOutflowFractionList = lapply(seqList, function(currSeq) {
      
      fractions = sapply(wellnessUniqSubjs, function(currSubj) {
        subj_visit = paste(currSubj, currSeq, sep = "_")
        if (!all(subj_visit %in% colnames(wellnessMgsMat))) return(NA)
        visit1=targetDigit[,subj_visit[1]]
        visit2=targetDigit[,subj_visit[2]]
        visit1=visit1[visit1==1]
        
        visit1Species = names(visit1)
        visit2Species = names(visit2[visit2==1])
        visit12Species = visit1Species[visit1Species %in% visit2Species]
        visit1_2species = visit1Species[!visit1Species %in% visit2Species]
        return(mean(visit1_2species %in% outflowSpecies))
        
        fraction = mean(visit2[names(visit1)])
        return(fraction)
      })
      
      return(fractions)
      
      
    })
    
    barplotFraction(checkInflowFractionList, checkOutflowFractionList, 1)
    
    seqFractionTab = rbind(data.frame(individual=names(seqFractionList$`v1-v2`), ratio=seqFractionList$`v1-v2`, time="v1 -> v2"),
                           data.frame(individual=names(seqFractionList$`v2-v3`), ratio=seqFractionList$`v2-v3`, time="v2 -> v3"),
                           data.frame(individual=names(seqFractionList$`v3-v4`), ratio=seqFractionList$`v3-v4`, time="v3 -> v4"))
    
    seqFractionTab2 = rbind(data.frame(individual=names(seqFractionList$`v2-v1`), ratio=seqFractionList$`v2-v1`, time="v2 -> v1"),
                            data.frame(individual=names(seqFractionList$`v3-v2`), ratio=seqFractionList$`v3-v2`, time="v3 -> v2"),
                            data.frame(individual=names(seqFractionList$`v4-v3`), ratio=seqFractionList$`v4-v3`, time="v4 -> v3"))
    seqFractionTabAll = rbind(seqFractionTab, seqFractionTab2)
    write.table(seqFractionTabAll, "individual.changes.for.fig1a.txt", sep="\t")
    
    
    boxplot(rev(seqFractionList), 
            horizontal=T, las=1, 
            xlab="Retained species (%)",
            col=colorRampPalette(c("white","skyblue"))(3))
    
    
    lapply(seqFractionList, mean)
    lapply(seqFractionList, function(x) 1-mean(x))
    
    seqSwedenFractionList = lapply(seqList, function(currSeq) {
      
      fractions = sapply(wellnessUniqSubjs, function(currSubj) {
        subj_visit = paste(currSubj[1], currSeq, sep = "_")
        if (!all(subj_visit %in% colnames(wellnessMgsMat))) return(NA)
        visit1=wellnessSwedenMgsMatDigit[,subj_visit[1]]
        visit2=wellnessSwedenMgsMatDigit[,subj_visit[2]]
        visit1=visit1[visit1==1]
        
        fraction = mean(visit2[names(visit1)])
        return(fraction)
      })
      
      return(fractions)
      
      
    })
    seqSpainFractionList = lapply(seqList, function(currSeq) {
      
      fractions = sapply(wellnessUniqSubjs, function(currSubj) {
        subj_visit = paste(currSubj[1], currSeq, sep = "_")
        if (!all(subj_visit %in% colnames(wellnessMgsMat))) return(NA)
        visit1=wellnessSpainMgsMatDigit[,subj_visit[1]]
        visit2=wellnessSpainMgsMatDigit[,subj_visit[2]]
        visit1=visit1[visit1==1]
        
        fraction = mean(visit2[names(visit1)])
        return(fraction)
      })
      
      return(fractions)
      
      
    })
    seqJapanFractionList = lapply(seqList, function(currSeq) {
      
      fractions = sapply(wellnessUniqSubjs, function(currSubj) {
        subj_visit = paste(currSubj[1], currSeq, sep = "_")
        if (!all(subj_visit %in% colnames(wellnessMgsMat))) return(NA)
        visit1=wellnessJapanMgsMatDigit[,subj_visit[1]]
        visit2=wellnessJapanMgsMatDigit[,subj_visit[2]]
        visit1=visit1[visit1==1]
        
        fraction = mean(visit2[names(visit1)])
        return(fraction)
      })
      
      return(fractions)
      
      
    })
    seqPeruFractionList = lapply(seqList, function(currSeq) {
      
      fractions = sapply(wellnessUniqSubjs, function(currSubj) {
        subj_visit = paste(currSubj[1], currSeq, sep = "_")
        if (!all(subj_visit %in% colnames(wellnessMgsMat))) return(NA)
        visit1=wellnessPeruMgsMatDigit[,subj_visit[1]]
        visit2=wellnessPeruMgsMatDigit[,subj_visit[2]]
        visit1=visit1[visit1==1]
        
        fraction = mean(visit2[names(visit1)])
        return(fraction)
      })
      
      return(fractions)
      
      
    })
    
    lapply(seqSwedenFractionList, mean)
    lapply(seqJapanFractionList, mean)
    lapply(seqPeruFractionList, mean)
    lapply(seqSpainFractionList, mean)
    
    
    seqFractionBySubjList = sapply(wellnessUniqSubjs, function(currSubj) {
      
      fractions = unlist(lapply(seqList, function(currSeq) {
        subj_visit = paste(currSubj, currSeq, sep = "_")
        if (!all(subj_visit %in% colnames(wellnessMgsMat))) return(NA)
        visit1=wellnessMgsMatDigit[,subj_visit[1]]
        visit2=wellnessMgsMatDigit[,subj_visit[2]]
        visit1=visit1[visit1==1]
        fraction = mean(visit2[names(visit1)])
        return(fraction)
      }))
      return(fractions)
      
    }, simplify = F)
    seqFractionBySubjMat = do.call(rbind, seqFractionBySubjList)
    colMeans(seqFractionBySubjMat)
    
    seqPeruFractionBySubjList = sapply(wellnessUniqSubjs, function(currSubj) {
      
      fractions = unlist(lapply(seqList, function(currSeq) {
        subj_visit = paste(currSubj, currSeq, sep = "_")
        if (!all(subj_visit %in% colnames(wellnessMgsMat))) return(NA)
        visit1=wellnessPeruMgsMatDigit[,subj_visit[1]]
        visit2=wellnessPeruMgsMatDigit[,subj_visit[2]]
        visit1=visit1[visit1==1]
        fraction = mean(visit2[names(visit1)])
        return(fraction)
      }))
      return(fractions)
      
    }, simplify = F)
    seqPeruFractionBySubjMat = do.call(rbind, seqPeruFractionBySubjList)
    
    
    seqSwedenFractionBySubjList = sapply(wellnessUniqSubjs, function(currSubj) {
      
      fractions = unlist(lapply(seqList, function(currSeq) {
        subj_visit = paste(currSubj, currSeq, sep = "_")
        if (!all(subj_visit %in% colnames(wellnessMgsMat))) return(NA)
        visit1=wellnessSwedenMgsMatDigit[,subj_visit[1]]
        visit2=wellnessSwedenMgsMatDigit[,subj_visit[2]]
        visit1=visit1[visit1==1]
        fraction = mean(visit2[names(visit1)])
        return(fraction)
      }))
      return(fractions)
      
    }, simplify = F)
    seqSwedenFractionBySubjMat = do.call(rbind, seqSwedenFractionBySubjList)
    colMeans(seqSwedenFractionBySubjMat)
    
    View(seqFractionBySubjMat)
    
    
    matplot(t(seqFractionBySubjMat), type = c("b"), pch=1 ) #plot
    matplot(t(seqSwedenFractionBySubjMat), type = c("b"), pch=1 ) #plot
    
    
    clusterPlot <- function(subjMat1, subjMat2) {
      subjMat1 = seqPeruFractionBySubjMat
      subjMat2 = seqSwedenFractionBySubjMat
      subjMelt1 = melt(subjMat1)
      subjMelt2 = melt(subjMat2)
      subjMelt1$type = "1"
      subjMelt2$type = "2"
      subjMeltAll = rbind(subjMelt1, 
                          subjMelt2)
      
      data_summary <- function(data, varname, groupnames){
        require(plyr)
        summary_func <- function(x, col){
          c(mean = mean(x[[col]], na.rm=TRUE),
            sd = sd(x[[col]], na.rm=TRUE))
        }
        data_sum<-ddply(data, groupnames, .fun=summary_func,
                        varname)
        data_sum <- rename(data_sum, c("mean" = varname))
        return(data_sum)
      }
      
      df3 <- data_summary(subjMeltAll, varname="value", 
                          groupnames=c("Var2", "type"))
      #df3$value = df3$value*100
      ggplot(df3, aes(x=Var2, y=value, group=type, color=type))  + 
        geom_errorbar(mapping=aes(ymin=value - sd, ymax=value + sd), width=.1, 
                      position=position_dodge(0.05)) +
        geom_line() + geom_point() + 
        ylab("Preserved (%)") + xlab("")+
        ylim(c(0.5,1.1))+
        scale_color_brewer(palette="Paired")+
        theme(panel.grid.major = element_blank(), 
              panel.grid.minor = element_blank(),
              legend.position = "none",axis.ticks.x = element_blank(),
              axis.text.x = element_blank(),
              panel.background = element_rect(fill = "white", 
                                              colour = "black", 
                                              size = 0.5, 
                                              linetype = "solid"))
      
      stat_summary(geom = "line", fun.y = mean, col="gold") +
        stat_summary(geom = "ribbon", fun.data = mean_sdl, alpha = 0.3) + 
        stat_summary(data=subjMelt2, geom = "line", fun.y = mean, col="blue") +
        stat_summary(data=subjMelt2, geom = "ribbon", fun.data = mean_sdl, alpha = 0.3) + 
        xlab("") +
        ylab("") + ylim(c(0.7,1.25)) + 
        theme(panel.grid.major = element_blank(), 
              panel.grid.minor = element_blank(),
              legend.position = "none",
              axis.text.x = element_text(angle = 60, hjust=1),
              panel.background = element_rect(fill = "white", 
                                              colour = "black", 
                                              size = 0.5, 
                                              linetype = "solid"))
      
      
    }
    
  }
  
  plotSankey = F
  if (plotSankey) {
    Sys.setenv(JAVA_HOME="C:\\Program Files\\Java\\jdk1.8.0_181")
    require(networkD3)
    fsSize=12
    qual_col_pals = brewer.pal.info[brewer.pal.info$category == 'qual',]
    col_vector = unlist(mapply(brewer.pal, qual_col_pals$maxcolors, rownames(qual_col_pals)))
    
    Nodes = data.frame(name = c("V1", "N2", "M1", "V2", "N3", "M2", "V3", "N4", "M3", "V4"),
                       group= c("V", "N", "M", "V", "N", "M", "V", "N", "M", "V"),
                       stringsAsFactors = F)
    
    interactions = data.frame(source = c("V1", "V1", "N2", "V2", "V2", "N3", "V3", "V3", "N4"),
                              target = c("M1", "V2", "V2", "M2", "V3", "V3", "M3", "V4", "V4"),
                              linkGroup =  c("M","T","N","M","T","N","M","T","N"),
                              value = c(12.1, 87.9, 13.1, 13.8, 85.5, 15.6, 15.6, 84.5, 14.9 ),
                              stringsAsFactors=FALSE)
    
    
    
    Links = data.frame(interactions)
    Links$sourceID = match(Links$source, Nodes$name)-1
    Links$targetID = match(Links$target, Nodes$name)-1
    
    elementCols = col_vector[1:length(Nodes$name)]
    elements = as.character(Nodes$name)
    color_scale <- data.frame(
      range = elementCols,
      domain = elements, 
      nodes = as.character(Nodes$name),
      stringsAsFactors = FALSE
    )
    
    sankeyNetwork(Links = Links, 
                  Nodes = Nodes, 
                  width=400,
                  height=300,
                  Source = "sourceID", 
                  Target = "targetID", 
                  Value = "value", 
                  NodeGroup = "group", 
                  NodeID = "name", 
                  nodePadding = 40,
                  nodeWidth = 30,
                  iterations = 1,
                  fontFamily = "arial", fontSize = fsSize, colourScale = JS(
                    sprintf(
                      'd3.scaleOrdinal()  
                      .domain(%s)
                      .range(%s)
                      ',
                      jsonlite::toJSON(color_scale$domain),
                      jsonlite::toJSON(color_scale$range)
                    )
                  ))
  }
  
  outlierCheck = F
  if (outlierCheck) {
    getSummaryStat<- function(scores) {
      
      outVec = c(min=min(scores),
                 max=max(scores),
                 median=median(scores),
                 iqr = quantile(scores, 0.75) - quantile(scores, 0.25))
      names(outVec) = c("min","max","median","IQR")
      return(outVec)
    }
    checkOutLiers <- function(scores) {
      summaryStat = getSummaryStat(scores)
      upperBound = summaryStat["median"] + (summaryStat["IQR"])/2
      lowerBound = summaryStat["median"] - (summaryStat["IQR"])/2
      
      inds = scores > upperBound | scores < lowerBound
      return(inds)
    }
    
    checkOutLiers(1:10)
    boxplot(1:10)
    
    
    wellnessMgsMat[1:3,1:3]
    
    summaryMat = t(apply(wellnessMgsMat, 1, function(currRow){
      return(getSummaryStat(currRow[currRow!=0]))
    }))
  }
  
  checkRichnessByVisit = T
  if (checkRichnessByVisit) {
    richnessByVisits = split(wellnessMetadata$GeneRichness, wellnessMetadata$subtype)
    richnessByVisitsMelt = melt(richnessByVisits)
    richnessByVisitsMelt$L1 = factor(richnessByVisitsMelt$L1, levels=rev(c("v1","v2","v3","v4")))
    ggplot(richnessByVisitsMelt, aes(x=L1, y=value)) + geom_boxplot(aes(colour = L1)) + 
      geom_jitter(aes(colour = L1),size=0.3,width=0.25) + 
      scale_colour_manual(values = colorRampPalette(c("#d3d3d3","#808080"))(4))+
      ylab("Gene richness") + xlab("") + 
      theme(panel.grid.major = element_blank(), 
            panel.grid.minor = element_blank(),
            legend.position = "none",
            axis.text.x = element_text(angle = 30, hjust=1),
            panel.background = element_rect(fill = "white", 
                                            colour = "black", 
                                            size = 0.5, 
                                            linetype = "solid")) +
      coord_flip()
    
    t.test(richnessByVisits$v1, richnessByVisits$v2)
    t.test(richnessByVisits$v2, richnessByVisits$v3)
    t.test(richnessByVisits$v3, richnessByVisits$v4)
    
  }
  
  newKmMode = T
  if (newKmMode) {
    
    decreaseCut = 0.9106159
    increaseCut = 1.096916
    
    if (F) {
      wellnessMetadata = basicMetaMapUpdated[basicMetaMapUpdated$dataset.ID=="id28",]
      mgsMat = mergeMatUpdated
      metaTab = basicMetaMapUpdated
    }
    
    reloadMgsMatImputed = T
    if (reloadMgsMatImputed) {
      
      ######
      mgsMat = mergeMatUpdated
      metaTab = basicMetaMapUpdated
      
      wellnessSubjs = (wellnessMetadata$metadata.ID)
      missingSubjs1 = names(table(wellnessSubjs)[table(wellnessSubjs)==3])
      exceptionSubjs = c("X3452","X3834","X3918","X4215")
      missingSubjs2 = c("X3913","X3535","X3016")
      missingSubjs1 = missingSubjs1[!missingSubjs1 %in% exceptionSubjs]
      
      wellnessMetadata = wellnessMetadata[!wellnessMetadata$metadata.ID %in% missingSubjs1, ]
      wellnessMetadata = wellnessMetadata[!wellnessMetadata$metadata.ID %in% missingSubjs2, ]
      wellnessMetadata = wellnessMetadata[!wellnessMetadata$metadata.ID %in% exceptionSubjs, ]
      wellnessMgsMat = mergeMatUpdatedImputed[,wellnessMetadata$sample.ID]
      #wellnessMgsMat = mergeMatUpdated[,wellnessMetadata$sample.ID]
      
      wellnessSamples = wellnessMetadata$sample.ID
      wellnessUniqSubjs = unique(wellnessMetadata$metadata.ID)
      
      
      
    }
    
    reloadMgsMatNonImputd = F
    if (reloadMgsMatImputed) {
      
      ######
      mgsMat = mergeMatUpdated
      metaTab = basicMetaMapUpdated
      
      wellnessSubjs = (wellnessMetadata$metadata.ID)
      missingSubjs1 = names(table(wellnessSubjs)[table(wellnessSubjs)==3])
      exceptionSubjs = c("X3452","X3834","X3918","X4215")
      missingSubjs2 = c("X3913","X3535","X3016")
      missingSubjs1 = missingSubjs1[!missingSubjs1 %in% exceptionSubjs]
      
      wellnessMetadata = wellnessMetadata[!wellnessMetadata$metadata.ID %in% missingSubjs1, ]
      wellnessMetadata = wellnessMetadata[!wellnessMetadata$metadata.ID %in% missingSubjs2, ]
      wellnessMetadata = wellnessMetadata[!wellnessMetadata$metadata.ID %in% exceptionSubjs, ]
      wellnessMgsMat = mergeMatUpdated[,wellnessMetadata$sample.ID]
      #wellnessMgsMat = mergeMatUpdated[,wellnessMetadata$sample.ID]
      wellnessMgsMat = wellnessMgsMat[rowSums(wellnessMgsMat)!=0, ]
      wellnessSamples = wellnessMetadata$sample.ID
      wellnessUniqSubjs = unique(wellnessMetadata$metadata.ID)
      
      scaledWellnessMgsMat = t(scale(t(wellnessMgsMat)))
      
    }
    
    generateMode = F
    if (generateMode) {
      
      getPairSeqs <- function(lenSeqs) {
        lenInds = lenSeqs-1
        pairList = sapply(1:lenInds, function(x) c(x, x+1), simplify = F)
        pairTab = do.call(rbind, pairList)
        return(pairTab)
      }
      getSurvVecByMgs <- function(mgsVec, mgsName, sampleNames) {
        
        lenAll = length(mgsVec)
        lenInd = length(mgsVec)-1
        seqs = 1:lenInd
        interval=3
        survTime = 0
        survEvent = "None"
        
        #####
        #check fist or second visit whether mgs exists
        
        mgsCutDigit <- mgsVec > 0
        presentInds = which(mgsCutDigit)
        if (length(presentInds)==0) {
          survTime = NA
          survEvent = "None"
          return(list(time=survTime, event=survEvent, sample=sampleNames[1], abd = 0, mgs=mgsName))
        }
        startInd = min(presentInds)
        lastInd = startInd
        for (ind in seq_along(presentInds)) {
          if (ind == length(presentInds)) break
          currInd = presentInds[ind]
          nextInd = presentInds[ind+1]
          diff = nextInd - currInd
          if (diff == 1) lastInd = nextInd
        }
        
        if (length(presentInds)!= 0) {
          survTime= (lastInd+1 - startInd)*interval
          survEvent = "Absent"
          if (lastInd == lenAll) {
            survTime = 9
            survEvent = "None"
          }
        }
        res=list(time=survTime, event=survEvent, sample=sampleNames[startInd], abd = mgsVec[startInd], mgs=mgsName)
        #print(res)
        return(res)
        
      }
      getSurvivalRates <- function(metaTab, mgsMat) {
        #wellness data
        #decrease
        #increase
        #mgsMat = mergeMatUpdated
        loadData = F
        if (loadData) {
          wellnessMetadata = basicMetaMapUpdated[basicMetaMapUpdated$dataset.ID=="id28",]
          
          mgsMat = mergeMatUpdated
          metaTab = basicMetaMapUpdated
          
          wellnessMetadata = metaTab[metaTab$dataset.ID=="id28",]
          wellnessSubjs = (wellnessMetadata$metadata.ID)
          missingSubjs1 = names(table(wellnessSubjs)[table(wellnessSubjs)==3])
          exceptionSubjs = c("X3452","X3834","X3918","X4215")
          missingSubjs2 = c("X3913","X3535","X3016")
          missingSubjs1 = missingSubjs1[!missingSubjs1 %in% exceptionSubjs]
          
          wellnessMetadata = wellnessMetadata[!wellnessMetadata$metadata.ID %in% missingSubjs1, ]
          wellnessMetadata = wellnessMetadata[!wellnessMetadata$metadata.ID %in% missingSubjs2, ]
          wellnessMetadata = wellnessMetadata[!wellnessMetadata$metadata.ID %in% exceptionSubjs, ]
          wellnessMgsMat = mergeMatUpdated[,match(wellnessMetadata$sample.ID, colnames(mergeMatUpdated))]
          wellnessMgsMatDigit = (wellnessMgsMat>0) * 1
          wellnessMgsMeans = rowMeans(wellnessMgsMat)
          wellnessSamples = wellnessMetadata$sample.ID
          wellnessSubjects = gsub("_v[1-4]","",wellnessSamples)
          wellnessVisits = gsub("X[0-9]+_","",wellnessSamples)
          wellnessSamplesBySubjects = split(wellnessSamples, wellnessSubjects)
          wellnessUniqSubjs = unique(wellnessMetadata$metadata.ID)
          wellnessClinicalMatrix = t(clinicalMatrix)[,match(wellnessSamples, rownames(clinicalMatrix))]
          
        }
        
        reloadMgsMatImputed = T
        if (reloadMgsMatImputed) {
          
          ######
          #mgsMat = mergeMatUpdated
          #metaTab = basicMetaMapUpdated
          
          wellnessSubjs = (wellnessMetadata$metadata.ID)
          missingSubjs1 = names(table(wellnessSubjs)[table(wellnessSubjs)==3])
          exceptionSubjs = c("X3452","X3834","X3918","X4215")
          missingSubjs2 = c("X3913","X3535","X3016")
          missingSubjs1 = missingSubjs1[!missingSubjs1 %in% exceptionSubjs]
          
          wellnessMetadata = wellnessMetadata[!wellnessMetadata$metadata.ID %in% missingSubjs1, ]
          wellnessMetadata = wellnessMetadata[!wellnessMetadata$metadata.ID %in% missingSubjs2, ]
          wellnessMetadata = wellnessMetadata[!wellnessMetadata$metadata.ID %in% exceptionSubjs, ]
          wellnessMgsMat = mgsMat[,wellnessMetadata$sample.ID]
          #wellnessMgsMat = mergeMatUpdated[,wellnessMetadata$sample.ID]
          
          wellnessSamples = wellnessMetadata$sample.ID
          wellnessUniqSubjs = unique(wellnessMetadata$metadata.ID)
          
        }
        
        wellnessMgsPrevalence = apply(wellnessMgsMat, 1, function(currRow){
          mean(currRow>0)
        })
        wellnessMgsMeans = rowMeans(wellnessMgsMat)
        
        wellnessSurvList = do.call(c,sapply(wellnessUniqSubjs, function(subj) {
          currSamples = wellnessMetadata$sample.ID[wellnessMetadata$metadata.ID==subj]
          currSamples = currSamples[order(currSamples)]
          currMgsMat = wellnessMgsMat[,currSamples]
          currMgsMat = currMgsMat[rowMeans(currMgsMat)!=0,]
          currSubSurvVecThroughs = (sapply(rownames(currMgsMat), function(currMgs){
            currSubSurvVecThrough = getSurvVecByMgs(currMgsMat[currMgs,], currMgs, currSamples)
            return(currSubSurvVecThrough)
          }, simplify = F))
          
          return(currSubSurvVecThroughs) 
          
        }, simplify = F))
        
        survTab = data.frame(time=do.call(c,lapply(wellnessSurvList, function(x) x$time)),
                             status=do.call(c,lapply(wellnessSurvList, function(x) x$event)),
                             sample=do.call(c,lapply(wellnessSurvList, function(x) x$sample)),
                             abd=do.call(c,lapply(wellnessSurvList, function(x) x$abd)),
                             mgs=do.call(c,lapply(wellnessSurvList, function(x) x$mgs)),
                             stringsAsFactors = F)
        survTab$logAbd = log10(survTab$abd)
        survivalRates = unlist(sapply(unique(survTab$mgs), function(currMsp) {
          currMspTab = survTab[survTab$mgs == currMsp,]
          if (F) {
            library("survminer")
            currMsp = "msp_0069"
            currMsp = "msp_0041"
            currMsp = "msp_0005"
            currMsp = "msp_0148c"
            currMsp = "msp_0742"
            
            
            currMspTab = survTab[survTab$mgs == currMsp,]
            currMspTab$status <- currMspTab$status == "Absent"
            currSurvFit=survfit(survival::Surv(time, status) ~ 1, data=currMspTab)
            ggsurvplot(currSurvFit, break.time.by = 3, color = "#2E9FDF")
            
          }
          halfYear=6
          halfYearSummary = summary(survfit(survival::Surv(currMspTab$time, currMspTab$status == "Absent") ~ 1), times=halfYear)
          halfYearSurvival = halfYearSummary$surv
          return(halfYearSurvival)
        }))
        
        survivalStats = data.frame(surv.rates=survivalRates[match(rownames(wellnessMgsMat), names(survivalRates))],
                                   mgs.abd = wellnessMgsMeans[match(rownames(wellnessMgsMat), names(wellnessMgsMeans))],
                                   freq=wellnessMgsPrevalence,
                                   mgs=rownames(wellnessMgsMat),
                                   species=getSpeciesName(rownames(wellnessMgsMat),taxo))
        
        return(survivalStats)
      }
      
      
      #survivalStats = getSurvivalRates(metaTab, mgsMat, decreaseCut)
      survivalStats = getSurvivalRates(basicMetaMapUpdated, mergeMatUpdatedImputed)
      survivalStatsRData = "C:\\Data\\comparative.analysis.healthy.sweden\\survivalStats.2020.Feb.7.RData"
      save(survivalStats, file=survivalStatsRData)
      
      
      table(is.na(survivalStats$surv.rates))
      survivalStats = survivalStats[!is.na(survivalStats$surv.rates),]
      hist(survivalStats$surv.rates)
      plot(density(survivalStats$surv.rates))
      
      View(survivalStats[survivalStats$surv.rates < 0.5,])
      View(survivalStats[survivalStats$mgs.abd > 1.749531e-07,])
      
      manualCutSurvRates <- function(survivalStats, by="freq", breaks =c(0,0.2, 0.4, 0.6, 0.8,1)) {
        cuts = cut(survivalStats[,by], breaks=breaks)
        cutLevels = levels(cuts)
        survRatesBins = split(survivalStats$surv.rates, cuts)
        survRatesBins = survRatesBins[cutLevels]
        return(survRatesBins)
      }
      binOut = manualCutSurvRates(survivalStats, "freq", c(0,0.3333, 0.6666, 1))
      binOut = manualCutSurvRates(survivalStats, "freq", c(0,0.25, 0.5, 0.75, 1))
      boxplot(binOut, las=2, col=colorRampPalette(c("white","skyblue"), alpha=T)(5))
      
      binOut = manualCutSurvRates(survivalStats, "mgs.abd", c(0,0.25, 0.5, 0.75, 1))
      boxplot(binOut, las=2, col=colorRampPalette(c("white","skyblue"), alpha=T)(5))
      
      
      survBins = getBinTabForSurvival2(survivalStats,"mgs.abd")
      boxplot(survBins, las=2, col=colorRampPalette(c("white","skyblue"), alpha=T)(5))
      
      
      #install.packages("rbin")
      #survivalStatsNewRData = "C://Data//comparative.analysis.healthy.sweden//survivalStats.new.RData"
      #save(survivalStats, file=survivalStatsNewRData)
      
      survivalStatsNewRData = "C://Data//comparative.analysis.healthy.sweden//survivalStats.new.2019.02.25.RData"
      #save(survivalStats, file=survivalStatsNewRData)
      
      plot(survivalStats$freq, survivalStats$surv.rates, log="x")
      
      
      
      
      boxplot(survBins)
      checkRichnessChangeBySubject <- function(wellnessMetadata, subj) {
        subjVisits = wellnessMetadata$subtype[wellnessMetadata$metadata.ID==subj]
        subjVisitNumbs = getNumbersFromWellnessVisits(subjVisits)
        subjRichness = wellnessMetadata$GeneRichness[wellnessMetadata$metadata.ID==subj]
        names(subjRichness) = subjVisitNumbs
        subjRichness = subjRichness[order(names(subjRichness))]
        subjFCs = getFoldChangeSeqs(subjRichness)
        names(subjFCs) = paste(subj, "_v", names(subjFCs), sep="")
        return(subjFCs)
      }
      getFoldChangeSeqs <- function(richnessVec) {
        lenInd=length(richnessVec) - 1 
        fcVec = sapply(1:lenInd,function(ind){
          return(richnessVec[ind+1]/richnessVec[ind])
        })
        print(lenInd)
        names(fcVec) = names(richnessVec)[1:lenInd]
        return(fcVec)
      }
      getSurvVec <- function(subjFCs, cut, mode = "decrease") {
        fcCutDigit = NULL
        interval=3
        survTime = 0
        survEvent = "None"
        firstSample = names(subjFCs)[1]
        
        if (mode == "decrease") {
          fcCutDigit <- subjFCs <= cut
          eventInds = which(fcCutDigit)
          if (length(eventInds)!= 0) {
            eventInds = min(eventInds)
            survTime = interval*eventInds
            survEvent="decrease"
          }
        }
        if (mode == "increase") {
          fcCutDigit <- subjFCs >= cut
          eventInds = which(fcCutDigit)
          if (length(eventInds)!= 0) {
            eventInds = min(eventInds)
            survTime = interval*eventInds
            survEvent="increase"
          }
        }
        return(list(time=survTime, event=survEvent, sample=firstSample))
      } 
      is.sequential <- function(x){
        all(diff(x) == diff(x)[1])
      }
      getSurvVecThrough <- function(subjFCs, cut, mode = "decrease") {
        lenAll = length(subjFCs)
        lenInd = length(subjFCs)-1
        seqs = 1:lenInd
        fcCutDigit = NULL
        interval=3
        survTime = 0
        survEvent = "None"
        
        results = sapply(seqs, function(currInd) {
          currSubjFCs = subjFCs[currInd:lenAll]
          result = getSurvVec(currSubjFCs, cut, mode)
        }, simplify = F)
        
        return(results)
        
        
        
      } 
      getSurvivalKMData <- function(metaTab, mgsMat, dCut, mode="decrease") {
        #wellness data
        #decrease
        #increase
        
        wellnessMetadata = metaTab[metaTab$dataset.ID=="id28",]
        wellnessSubjs = (wellnessMetadata$metadata.ID)
        missingSubjs1 = names(table(wellnessSubjs)[table(wellnessSubjs)==3])
        exceptionSubjs = c("X3452","X3834","X3918","X4215")
        missingSubjs2 = c("X3913","X3535","X3016")
        missingSubjs1 = missingSubjs1[!missingSubjs1 %in% exceptionSubjs]
        
        wellnessMetadata = wellnessMetadata[!wellnessMetadata$metadata.ID %in% missingSubjs1, ]
        wellnessMetadata = wellnessMetadata[!wellnessMetadata$metadata.ID %in% missingSubjs2, ]
        wellnessMgsMat = mergeMatUpdated[,wellnessMetadata$sample.ID]
        wellnessSamples = wellnessMetadata$sample.ID
        wellnessUniqSubjs = unique(wellnessMetadata$metadata.ID)
        wellnessUniqSubjVisit1 = paste(wellnessUniqSubjs, "_v1", sep="")
        wellnessMgsV1Mat = wellnessMgsMat[,wellnessUniqSubjVisit1]
        
        wellnessSurvList = do.call(c,sapply(wellnessUniqSubjs, function(subj) {
          currSubjFCs = checkRichnessChangeBySubject(wellnessMetadata, subj)
          #print(length(currSubjFCs))
          #currSubSurvVec = getSurvVec(currSubjFCs, dCut, mode)
          currSubSurvVecThrough = getSurvVecThrough(currSubjFCs, dCut, mode)
          
          return(currSubSurvVecThrough) 
          
        }, simplify = F))
        
        survTab = data.frame(time=do.call(c,lapply(wellnessSurvList, function(x) x$time)),
                             status=do.call(c,lapply(wellnessSurvList, function(x) x$event)),
                             sample=do.call(c,lapply(wellnessSurvList, function(x) x$sample)),
                             stringsAsFactors = F)
        
        require(survival)
        #getStatisticsPerEach
        
        presenceMode = T
        if (presenceMode) {
          
          wellnessMgsMatchedMat = mergeMatUpdated[,match(survTab$sample,colnames(mergeMatUpdated))]
          allMsps = rownames(wellnessMgsMatchedMat)
          
          survPvalues = sapply(allMsps, function(currMsp) {
            currMspSubjDigit = wellnessMgsMatchedMat[currMsp,]>0
            currMspSubjDigit = currMspSubjDigit * 1
            currSuvTab = survTab
            currSuvTab$msp = currMsp
            currSuvTab$presence = currMspSubjDigit
            lenTypes = length(table(currSuvTab$presence))
            if (lenTypes==1) return(1)
            km_fit <- survfit(Surv(currSuvTab$time, currSuvTab$status == mode) ~ currSuvTab$presence)
            #plot(km_fit)
            #cox_fit <- coxph(Surv(currSuvTab$time, currSuvTab$status == mode) ~ currSuvTab$presence)
            
            diff_out <- survdiff(Surv(currSuvTab$time, currSuvTab$status == mode) ~ currSuvTab$presence)
            pval = getChiPval(diff_out$chisq,1)
            oeRatio=diff_out$obs / diff_out$exp
            names(oeRatio) = names(diff_out$n)
            presenceRatio=oeRatio["currSuvTab$presence=1"]
            if (is.nan(presenceRatio)) return(1)
            if (presenceRatio>1) pval = -1 * pval
            return(pval)
            #based
          })
          
          
        }
        
        abundanceMode = T
        if (abundanceMode) {
          wellnessMgsMatchedMat = mergeMatUpdated[,match(survTab$sample,colnames(mergeMatUpdated))]
          allMsps = rownames(wellnessMgsMatchedMat)
          
          
          coxPvalues = sapply(allMsps, function(currMsp) {
            currMspSubjAbd = wellnessMgsMatchedMat[currMsp,]
            if (sum(currMspSubjAbd) == 0 ) return(1)
            currSuvTab = survTab
            currSuvTab$msp = currMsp
            currSuvTab$abundance = currMspSubjAbd
            
            #plot(km_fit)
            cox_fit <- coxph(Surv(currSuvTab$time, currSuvTab$status == mode) ~ currSuvTab$abundance)
            cox_summary = summary(cox_fit)
            pvalue = cox_summary$coefficients[5]
            if (is.na(cox_summary$coefficients[1])) return(1)
            if (cox_summary$coefficients[1] < 0) return( -1 * pvalue) 
            #based
          }, simplify = F)
          coxPvalues = unlist(coxPvalues)
        }
        
        plot(coxPvalues, survivalStats$surv.rates[match(names(coxPvalues),as.character(survivalStats$mgs))], xlim(c, 0.))
        View(survivalStats[survivalStats$mgs %in% names(coxPvalues[coxPvalues>0]), ])
      }
      
      #survivalStats = getSurvivalKMData(metaTab, mgsMat, decreaseCut)
      
    }
    
    loadMode = T
    if (loadMode) {
      survivalStatsNewRData = "C://Data//comparative.analysis.healthy.sweden//survivalStats.new.RData"
      load(survivalStatsNewRData)
    }
    
    checkWithOutflow = T
    if (checkWithOutflow) {
      plot(survivalStats$surv.rates, outflow[match(rownames(survivalStats), names(outflow))],
           xlab="retaining prob",
           ylab="outflow")
      cor.test(survivalStats$surv.rates, outflow[match(rownames(survivalStats), names(outflow))])
      #-0.877; p-value < 2.2e-16
      
    }
    
  }
  
  markovMode = F
  if (markovMode) {
    
    require(markovchain)
    require(diagram)
    
    wellnessMetadata = basicMetaMapUpdated[basicMetaMapUpdated$dataset.ID=="id28",]
    
    mgsMat = mergeMatUpdated
    metaTab = basicMetaMapUpdated
    
    getColonicationProbByMsp <- function(metaTab, mgsMat, currMgs) {
      wellnessMetadata = metaTab[metaTab$dataset.ID=="id28",]
      wellnessSubjs = (wellnessMetadata$metadata.ID)
      missingSubjs1 = names(table(wellnessSubjs)[table(wellnessSubjs)==3])
      exceptionSubjs = c("X3452","X3834","X3918","X4215")
      missingSubjs2 = c("X3913","X3535","X3016")
      missingSubjs1 = missingSubjs1[!missingSubjs1 %in% exceptionSubjs]
      
      wellnessMetadata = wellnessMetadata[!wellnessMetadata$metadata.ID %in% missingSubjs1, ]
      wellnessMetadata = wellnessMetadata[!wellnessMetadata$metadata.ID %in% missingSubjs2, ]
      wellnessMgsMat = mergeMatUpdated[,wellnessMetadata$sample.ID]
      wellnessMgsMeans = rowMeans(wellnessMgsMat)
      wellnessSamples = wellnessMetadata$sample.ID
      wellnessUniqSubjs = unique(wellnessMetadata$metadata.ID)
      
      markovSeqOfMsp = (sapply(wellnessUniqSubjs, function(subj) {
        currSamples = wellnessMetadata$sample.ID[wellnessMetadata$metadata.ID==subj]
        currSamples = currSamples[order(currSamples)]
        currMgsMat = wellnessMgsMat[,currSamples]
        currMgsSubjVec = (currMgsMat[currMgs,] >0)*1
        currMgsSubjVec = as.character(currMgsSubjVec)
        
      }))
      
      myMarkovFit<-markovchainFit(data=markovSeqOfMsp,confidencelevel = .9,method = "mle")
      myMarkovMat = myMarkovFit$estimate
      print(dim(myMarkovMat))
      if(dim(myMarkovMat)[1]==1)print(myMarkovMat)
      if (dim(myMarkovMat)[1]==1)return(c(prob_10=NA, prob_01=NA))
      
      return(c(prob_10=myMarkovMat[2,1], prob_01=myMarkovMat[1,2]))
      
    }
    getColonizationProbs <- function(metaTab, mgsMat) {
      wellnessMetadata = metaTab[metaTab$dataset.ID=="id28",]
      wellnessSubjs = (wellnessMetadata$metadata.ID)
      missingSubjs1 = names(table(wellnessSubjs)[table(wellnessSubjs)==3])
      exceptionSubjs = c("X3452","X3834","X3918","X4215")
      missingSubjs2 = c("X3913","X3535","X3016")
      missingSubjs1 = missingSubjs1[!missingSubjs1 %in% exceptionSubjs]
      
      wellnessMetadata = wellnessMetadata[!wellnessMetadata$metadata.ID %in% missingSubjs1, ]
      wellnessMetadata = wellnessMetadata[!wellnessMetadata$metadata.ID %in% missingSubjs2, ]
      wellnessMgsMat = mergeMatUpdated[,wellnessMetadata$sample.ID]
      wellnessMgsMeans = rowMeans(wellnessMgsMat)
      wellnessSamples = wellnessMetadata$sample.ID
      wellnessUniqSubjs = unique(wellnessMetadata$metadata.ID)
      
      markovProbMatByMsp = do.call(rbind, sapply(rownames(mgsMat), function(currMgs){
        
        markovSeqOfMsp = (sapply(wellnessUniqSubjs, function(subj) {
          currSamples = wellnessMetadata$sample.ID[wellnessMetadata$metadata.ID==subj]
          currSamples = currSamples[order(currSamples)]
          currMgsMat = wellnessMgsMat[,currSamples]
          currMgsSubjVec = (currMgsMat[currMgs,] >0)*1
          currMgsSubjVec = as.character(currMgsSubjVec)
          
        }))
        
        myMarkovFit<-markovchainFit(data=markovSeqOfMsp,confidencelevel = .9,method = "mle")
        myMarkovMat = myMarkovFit$estimate@transitionMatrix
        
        print(dim(myMarkovMat))
        if(dim(myMarkovMat)==1)print(myMarkovMat)
        if (dim(myMarkovMat)==1)return(c(prob_10=NA, prob_01=NA))
        
        return(c(prob_10=myMarkovMat[2,1], prob_01=myMarkovMat[1,2]))
      }, simplify = F))
      
      return(markovProbMatByMsp)
      
    }
    getColonizationOddRatios <- function(metaTab, mgsMat) {
      wellnessMetadata = metaTab[metaTab$dataset.ID=="id28",]
      wellnessSubjs = (wellnessMetadata$metadata.ID)
      missingSubjs1 = names(table(wellnessSubjs)[table(wellnessSubjs)==3])
      exceptionSubjs = c("X3452","X3834","X3918","X4215")
      missingSubjs2 = c("X3913","X3535","X3016")
      missingSubjs1 = missingSubjs1[!missingSubjs1 %in% exceptionSubjs]
      
      wellnessMetadata = wellnessMetadata[!wellnessMetadata$metadata.ID %in% missingSubjs1, ]
      wellnessMetadata = wellnessMetadata[!wellnessMetadata$metadata.ID %in% missingSubjs2, ]
      wellnessMgsMat = mergeMatUpdated[,wellnessMetadata$sample.ID]
      wellnessMgsMeans = rowMeans(wellnessMgsMat)
      wellnessSamples = wellnessMetadata$sample.ID
      wellnessUniqSubjs = unique(wellnessMetadata$metadata.ID)
      
      markovOddRatiosByMsp = do.call(c, sapply(rownames(mgsMat), function(currMgs){
        
        markovSeqOfMsp = (sapply(wellnessUniqSubjs, function(subj) {
          currSamples = wellnessMetadata$sample.ID[wellnessMetadata$metadata.ID==subj]
          currSamples = currSamples[order(currSamples)]
          currMgsMat = wellnessMgsMat[,currSamples]
          currMgsSubjVec = (currMgsMat[currMgs,] >0)*1
          currMgsSubjVec = as.character(currMgsSubjVec)
          
        }))
        
        myMarkovFit<-markovchainFit(data=markovSeqOfMsp,confidencelevel = .9,method = "mle")
        myMarkovMat = myMarkovFit$estimate@transitionMatrix
        
        if(dim(myMarkovMat)==1) {
          rname = rownames(myMarkovMat)
          if (rname == "0") return(-Inf)
          if (rname == "1") return(Inf)
        }
        else {
          colonized = myMarkovMat["0","1"] 
          resistant = myMarkovMat["1","0"]
          oddRatio = colonized/ resistant
          return(oddRatio)
        }
        
      }, simplify = F))
      names(markovOddRatiosByMsp) = rownames(mgsMat)
      return(markovOddRatiosByMsp)
      
    }
    getColonizationColonizingProbs <- function(metaTab, mgsMat) {
      wellnessMetadata = metaTab[metaTab$dataset.ID=="id28",]
      wellnessSubjs = (wellnessMetadata$metadata.ID)
      missingSubjs1 = names(table(wellnessSubjs)[table(wellnessSubjs)==3])
      exceptionSubjs = c("X3452","X3834","X3918","X4215")
      missingSubjs2 = c("X3913","X3535","X3016")
      missingSubjs1 = missingSubjs1[!missingSubjs1 %in% exceptionSubjs]
      
      wellnessMetadata = wellnessMetadata[!wellnessMetadata$metadata.ID %in% missingSubjs1, ]
      wellnessMetadata = wellnessMetadata[!wellnessMetadata$metadata.ID %in% missingSubjs2, ]
      wellnessMgsMat = mergeMatUpdated[,wellnessMetadata$sample.ID]
      wellnessMgsMeans = rowMeans(wellnessMgsMat)
      wellnessSamples = wellnessMetadata$sample.ID
      wellnessUniqSubjs = unique(wellnessMetadata$metadata.ID)
      
      scoresByMsp = do.call(c, sapply(rownames(mgsMat), function(currMgs){
        
        markovSeqOfMsp = (sapply(wellnessUniqSubjs, function(subj) {
          currSamples = wellnessMetadata$sample.ID[wellnessMetadata$metadata.ID==subj]
          currSamples = currSamples[order(currSamples)]
          currMgsMat = wellnessMgsMat[,currSamples]
          currMgsSubjVec = (currMgsMat[currMgs,] >0)*1
          currMgsSubjVec = as.character(currMgsSubjVec)
          
        }))
        
        myMarkovFit<-markovchainFit(data=markovSeqOfMsp,confidencelevel = .9,method = "mle")
        myMarkovMat = myMarkovFit$estimate@transitionMatrix
        
        if(dim(myMarkovMat)==1) {
          rname = rownames(myMarkovMat)
          #if (rname == "0") return(0)
          #if (rname == "1") return(Inf)
          if (rname == "0") return(0)
          if (rname == "1") return(1)
        }
        else {
          colonized = myMarkovMat["0","1"] 
          preserved = myMarkovMat["1","1"] 
          relativeRisk = colonized/ preserved
          #return(relativeRisk)
          return(colonized)
        }
        
      }, simplify = F))
      names(scoresByMsp) = rownames(mgsMat)
      return(scoresByMsp)
      
    }
    getColonizationResistantProbs <- function(metaTab, mgsMat) {
      wellnessMetadata = metaTab[metaTab$dataset.ID=="id28",]
      wellnessSubjs = (wellnessMetadata$metadata.ID)
      missingSubjs1 = names(table(wellnessSubjs)[table(wellnessSubjs)==3])
      exceptionSubjs = c("X3452","X3834","X3918","X4215")
      missingSubjs2 = c("X3913","X3535","X3016")
      missingSubjs1 = missingSubjs1[!missingSubjs1 %in% exceptionSubjs]
      
      wellnessMetadata = wellnessMetadata[!wellnessMetadata$metadata.ID %in% missingSubjs1, ]
      wellnessMetadata = wellnessMetadata[!wellnessMetadata$metadata.ID %in% missingSubjs2, ]
      wellnessMgsMat = mergeMatUpdated[,wellnessMetadata$sample.ID]
      wellnessMgsMeans = rowMeans(wellnessMgsMat)
      wellnessSamples = wellnessMetadata$sample.ID
      wellnessUniqSubjs = unique(wellnessMetadata$metadata.ID)
      
      scoresByMsp = do.call(c, sapply(rownames(mgsMat), function(currMgs){
        
        markovSeqOfMsp = (sapply(wellnessUniqSubjs, function(subj) {
          currSamples = wellnessMetadata$sample.ID[wellnessMetadata$metadata.ID==subj]
          currSamples = currSamples[order(currSamples)]
          currMgsMat = wellnessMgsMat[,currSamples]
          currMgsSubjVec = (currMgsMat[currMgs,] >0)*1
          currMgsSubjVec = as.character(currMgsSubjVec)
          
        }))
        
        myMarkovFit<-markovchainFit(data=markovSeqOfMsp,confidencelevel = .9,method = "mle")
        myMarkovMat = myMarkovFit$estimate@transitionMatrix
        
        if(dim(myMarkovMat)==1) {
          rname = rownames(myMarkovMat)
          #if (rname == "0") return(Inf)
          #if (rname == "1") return(0)
          if (rname == "0") return(1)
          if (rname == "1") return(0)
        }
        else {
          preserved = myMarkovMat["0","0"] 
          resistant = myMarkovMat["1","0"] 
          #relativeRisk = resistant/ preserved
          #return(relativeRisk)
          return(resistant)
          
        }
        
      }, simplify = F))
      names(scoresByMsp) = rownames(mgsMat)
      return(scoresByMsp)
      
    }
    
    colonizingRatios = getColonizationColonizingProbs(metaTab, mgsMat)
    resistantRatios = getColonizationResistantProbs(metaTab, mgsMat)
    #colonizedRatios = getColonizationOddRatios(metaTab, mgsMat)
    colonizedRatios = (colonizingRatios+1e-4)/(resistantRatios+1e-4)
    colonizedRatios[colonizingRatios==0 & resistantRatios==0] = NA
    
    markovStatTab= data.frame(inflow = colonizingRatios,
                              outflow = resistantRatios, 
                              ratio=colonizedRatios,
                              species = getSpeciesName(names(colonizingRatios),taxo))
    
    
    ggplot(markovStatTab, aes(x=outflow, y=inflow, fill=inflow)) + 
      scale_fill_gradient2(midpoint = 0.2, low="#FFD70044", mid = "white", high = "purple") +
      geom_point(size=2, shape=21, colour="black")+ xlab("") + ylab("")+
      geom_vline(xintercept = 0.5)+geom_hline(yintercept = 0.5)+
      theme(panel.grid.major = element_blank(), 
            panel.grid.minor = element_blank(),
            legend.position = "none",
            panel.background = element_rect(fill = "white", 
                                            colour = "black", 
                                            size = 0.5, 
                                            linetype = "solid"))
    table(markovStatTab$inflow>0.5)
    table(markovStatTab$outflow>0.5)
    table(markovStatTab$outflow==1)
    
    inflowSpecies = rownames(markovStatTab[markovStatTab$inflow>=0.25,])
    #outflowSpecies = rownames(markovStatTab[markovStatTab$outflow>=0.5,])
    outflowSpecies = rownames(markovStatTab[markovStatTab$outflow==1,])
    length(inflowSpecies)
    length(outflowSpecies)
    
    table(colonizedRatios>1)
    
    colonizingSpecies = names(colonizedRatios[colonizedRatios>1])
    resistantSpecies = names(colonizedRatios[colonizedRatios<1])
    
    
    iniR = 0.2
    pie(1, radius=0.01, init.angle=90, col=c('white'), border = NA, labels='')
    
    floating.pie(0,0,c(232,1590),  
                 radius=3*iniR, 
                 startpos=pi/2, 
                 col = c("purple","gray"),
                 border="black")
    floating.pie(0,0,1,  
                 radius=2*iniR, 
                 startpos=pi/2, 
                 col="white",
                 border="black")
    floating.pie(0,0,1,  
                 radius=1.99*iniR, 
                 startpos=pi/2, 
                 col="white",
                 border=NA)
    
  }
  
  kmMode = F
  if (kmMode) {
    wellnessMetadata = basicMetaMapUpdated[basicMetaMapUpdated$dataset.ID=="id28",]
    
    mgsMat = mergeMatUpdated
    metaTab = basicMetaMapUpdated
    
    decreaseCut = 0.9106159
    increaseCut = 1.096916
    
    generateMode = F
    if (generateMode) {
      
      getPairSeqs <- function(lenSeqs) {
        lenInds = lenSeqs-1
        pairList = sapply(1:lenInds, function(x) c(x, x+1), simplify = F)
        pairTab = do.call(rbind, pairList)
        return(pairTab)
      }
      getSurvVecByMgs <- function(mgsVec, mgsName, sampleNames) {
        
        lenAll = length(mgsVec)
        lenInd = length(mgsVec)-1
        seqs = 1:lenInd
        interval=3
        survTime = 0
        survEvent = "None"
        
        #####
        #check fist or second visit whether mgs exists
        
        mgsCutDigit <- mgsVec > 0
        presentInds = which(mgsCutDigit)
        if (length(presentInds)==0) {
          survTime = NA
          survEvent = "None"
          return(list(time=survTime, event=survEvent, sample=sampleNames[1], abd = 0, mgs=mgsName))
        }
        startInd = min(presentInds)
        lastInd = startInd
        for (ind in seq_along(presentInds)) {
          if (ind == length(presentInds)) break
          currInd = presentInds[ind]
          nextInd = presentInds[ind+1]
          diff = nextInd - currInd
          if (diff == 1) lastInd = nextInd
        }
        
        if (length(presentInds)!= 0) {
          survTime= (lastInd+1 - startInd)*interval
          survEvent = "Absent"
          if (lastInd == lenAll) survEvent = "None"
        }
        res=list(time=survTime, event=survEvent, sample=sampleNames[startInd], abd = mgsVec[startInd], mgs=mgsName)
        #print(res)
        return(res)
        
      }
      getSurvivalRates <- function(metaTab, mgsMat, dCut, mode="decrease") {
        #wellness data
        #decrease
        #increase
        #mgsMat = mergeMatUpdated
        
        wellnessMetadata = metaTab[metaTab$dataset.ID=="id28",]
        wellnessSubjs = (wellnessMetadata$metadata.ID)
        missingSubjs1 = names(table(wellnessSubjs)[table(wellnessSubjs)==3])
        exceptionSubjs = c("X3452","X3834","X3918","X4215")
        missingSubjs2 = c("X3913","X3535","X3016")
        missingSubjs1 = missingSubjs1[!missingSubjs1 %in% exceptionSubjs]
        
        wellnessMetadata = wellnessMetadata[!wellnessMetadata$metadata.ID %in% missingSubjs1, ]
        wellnessMetadata = wellnessMetadata[!wellnessMetadata$metadata.ID %in% missingSubjs2, ]
        wellnessMgsMat = mgsMat[,wellnessMetadata$sample.ID]
        wellnessMgsPrevalence = apply(wellnessMgsMat, 1, function(currRow){
          mean(currRow>0)
        })
        wellnessMgsMeans = rowMeans(wellnessMgsMat)
        wellnessSamples = wellnessMetadata$sample.ID
        wellnessUniqSubjs = unique(wellnessMetadata$metadata.ID)
        
        wellnessSurvList = do.call(c,sapply(wellnessUniqSubjs, function(subj) {
          currSamples = wellnessMetadata$sample.ID[wellnessMetadata$metadata.ID==subj]
          currSamples = currSamples[order(currSamples)]
          currMgsMat = wellnessMgsMat[,currSamples]
          currMgsMat = currMgsMat[rowMeans(currMgsMat)!=0,]
          currSubSurvVecThroughs = (sapply(rownames(currMgsMat), function(currMgs){
            currSubSurvVecThrough = getSurvVecByMgs(currMgsMat[currMgs,], currMgs, currSamples)
            return(currSubSurvVecThrough)
          }, simplify = F))
          
          return(currSubSurvVecThroughs) 
          
        }, simplify = F))
        
        survTab = data.frame(time=do.call(c,lapply(wellnessSurvList, function(x) x$time)),
                             status=do.call(c,lapply(wellnessSurvList, function(x) x$event)),
                             sample=do.call(c,lapply(wellnessSurvList, function(x) x$sample)),
                             abd=do.call(c,lapply(wellnessSurvList, function(x) x$abd)),
                             mgs=do.call(c,lapply(wellnessSurvList, function(x) x$mgs)),
                             stringsAsFactors = F)
        survTab$logAbd = log10(survTab$abd)
        survivalRates = unlist(sapply(unique(survTab$mgs), function(currMsp) {
          currMspTab = survTab[survTab$mgs == currMsp,]
          if (F) {
            library("survminer")
            currMsp = "msp_0148c"
            currMspTab = survTab[survTab$mgs == currMsp,]
            currMspTab$status <- currMspTab$status == "Absent"
            currSurvFit=survfit(survival::Surv(time, status) ~ 1, data=currMspTab)
            ggsurvplot(currSurvFit, color = "#2E9FDF") 
            
            
            #cox_fit <- coxph(Surv(time, status) ~ logAbd, data=currMspTab)
            #cox_summary = summary(cox_fit)
            #res.zph <- cox.zph(cox_fit)
            #plot(res.zph)
            
          }
          #survival::survfit(survival::Surv(currMspTab$time, currMspTab$status == "Absent") ~ 1)
          halfYear=6
          halfYearSummary = summary(survfit(survival::Surv(currMspTab$time, currMspTab$status == "Absent") ~ 1), times=halfYear)
          halfYearSurvival = halfYearSummary$surv
          return(halfYearSurvival)
        }))
        
        survivalStats = data.frame(surv.rates=survivalRates[match(rownames(wellnessMgsMat), names(survivalRates))],
                                   mgs.abd = wellnessMgsMeans[match(rownames(wellnessMgsMat), names(wellnessMgsMeans))],
                                   freq=wellnessMgsPrevalence,
                                   mgs=rownames(wellnessMgsMat),
                                   species=getSpeciesName(rownames(wellnessMgsMat),taxo))
        
        # 
        # survivalStats = data.frame(surv.rates=survivalRates,
        #                            mgs.abd = wellnessMgsMeans[match(names(survivalRates), names(wellnessMgsMeans))],
        #                            mgs=names(survivalRates),
        #                            species=getSpeciesName(names(survivalRates),taxo))
        # 
        
        
        
        # survivalGssOssList = list(gss_surv=survivalRates[taxo$MSP[taxo$type_sh=="gss"]],
        #                           oss_surv=survivalRates[taxo$MSP[taxo$type_sh=="oss"]])
        # boxplot(survivalGssOssList)
        
        checkMode = F
        if (checkMode) {
          if(T) plot(survivalStats$mgs.abd,survivalStats$surv.rates, log="x", 
                     col="#80808088",pch=16,las=1,
                     xlab="MGS Abd.", ylab="Species-retaining prob.")
          View(survivalStats[survivalStats$surv.rates<0.7 & survivalStats$mgs.abd>1e-6,])
          prevotellaSpecies = taxo$MSP[taxo$genus=="Prevotella"]
          bacteroidesSpecies = taxo$MSP[taxo$genus=="Bacteroides"]
          taretSpecies = taxo$MSP[taxo$genus=="Bifidobacterium"]
          
          View(survivalStats[rownames(survivalStats) %in% taretSpecies,])
          View(survivalStats[rownames(survivalStats) %in% hgcSpecies,])
          hgcLgcSurvivalRateList = list(HGC=survivalStats$surv.rates[rownames(survivalStats) %in% hgcSpecies],
                                        LGC=survivalStats$surv.rates[rownames(survivalStats) %in% lgcSpecies],
                                        all=survivalStats$surv.rates)
          boxplot(hgcLgcSurvivalRateList)
          
          View(survivalStats[rownames(survivalStats) %in% prevotellaSpecies,])
          View(survivalStats[rownames(survivalStats) %in% bacteroidesSpecies,])
          prevotellaSRates = survivalStats$surv.rates[rownames(survivalStats) %in% prevotellaSpecies]
          sort(prevotellaSRates)
        }
        
        return(survivalStats)
      }
      
      
      survivalStats = getSurvivalRates(metaTab, mgsMat, decreaseCut)
      View(survivalStats[survivalStats$freq!=0,])
      
      swedSpecies = specificGroupMgsByCountries$Sweden
      denSpecies = specificGroupMgsByCountries$Denmark
      spaSpecies = specificGroupMgsByCountries$Spain
      
      tradSpecies = unique(unlist(specificGroupMgsByCountries[traditionalCountries]))
      indSpecies = unique(unlist(specificGroupMgsByCountries[industrializedCountries]))
      
      traditionalCountries %in% names(specificGroupMgsByCountries)
      industrializedCountries %in% names(specificGroupMgsByCountries)
      
      tradIndSurvList = list(Traditional=survivalStats$surv.rates[survivalStats$mgs %in% tradSpecies],
                             Industrial=survivalStats$surv.rates[survivalStats$mgs %in% indSpecies],
                             Spain=survivalStats$surv.rates[survivalStats$mgs %in% spaSpecies],
                             Denmark=survivalStats$surv.rates[survivalStats$mgs %in% denSpecies],
                             Sweden=survivalStats$surv.rates[survivalStats$mgs %in% swedSpecies])
      t.test(tradIndSurvList$Sweden, tradIndSurvList$Denmark)
      tradIndColonizedList = list(Traditional=colonizedRatios[names(colonizedRatios) %in% tradSpecies],
                                  Industrial=colonizedRatios[names(colonizedRatios) %in% indSpecies],
                                  Spain=colonizedRatios[names(colonizedRatios) %in% spaSpecies],
                                  Denmark=colonizedRatios[names(colonizedRatios) %in% denSpecies],
                                  Sweden=colonizedRatios[names(colonizedRatios) %in% swedSpecies])
      
      
      boxplot(tradIndSurvList)
      boxplot(tradIndColonizedList,  las=2)
      
      t.test(tradIndSurvList$Traditional, tradIndSurvList$Industrial)
      survListByAbds = getBinTabForSurvival(survivalStats)
      survMeltByAbds = melt(survListByAbds)
      t.test(survListByAbds$per20, 
             survListByAbds$per40)
      t.test(survListByAbds$per40, 
             survListByAbds$per60)
      t.test(survListByAbds$per60, 
             survListByAbds$per80)
      
      boxplot(survListByAbds)
      blueCol = rgb(5/255,48/255,97/255,0.5)
      
      survMeltByAbds$gradient = as.numeric(gsub("per","",survMeltByAbds$L1))
      survMeltByAbds$L1=factor(survMeltByAbds$L1, levels=c("per20","per40","per60","per80","per100"))
      
      ggplot(survMeltByAbds, aes(x=L1, y=value)) +  
        scale_fill_gradient2(low="white",high=blueCol) +
        geom_boxplot(aes(fill=gradient)) +
        xlab("") +
        ylab("") + 
        theme(panel.grid.major = element_blank(), 
              panel.grid.minor = element_blank(),
              legend.position = "none",
              axis.text.x = element_text(angle = 60, hjust=1),
              panel.background = element_rect(fill = "white", 
                                              colour = "black", 
                                              size = 0.5, 
                                              linetype = "solid"))
      
      tradIndSurvTab = melt(tradIndSurvList)
      tradIndSurvTab$L1 = factor(tradIndSurvTab$L1, levels = c("Traditional","Industrial","Spain","Denmark","Sweden"))
      ggplot(tradIndSurvTab, aes(x=L1, y=value, fill=L1)) +  
        scale_fill_manual(values=c(tradCol, eurCol, eurCol, eurCol, blueCol)) +
        geom_boxplot() +
        xlab("") +
        ylab("") + 
        theme(panel.grid.major = element_blank(), 
              panel.grid.minor = element_blank(),
              legend.position = "none",
              axis.text.x = element_text(angle = 60, hjust=1),
              panel.background = element_rect(fill = "white", 
                                              colour = "black", 
                                              size = 0.5, 
                                              linetype = "solid"))
      
      
      
      checkRichnessChangeBySubject <- function(wellnessMetadata, subj) {
        subjVisits = wellnessMetadata$subtype[wellnessMetadata$metadata.ID==subj]
        subjVisitNumbs = getNumbersFromWellnessVisits(subjVisits)
        subjRichness = wellnessMetadata$GeneRichness[wellnessMetadata$metadata.ID==subj]
        names(subjRichness) = subjVisitNumbs
        subjRichness = subjRichness[order(names(subjRichness))]
        subjFCs = getFoldChangeSeqs(subjRichness)
        names(subjFCs) = paste(subj, "_v", names(subjFCs), sep="")
        return(subjFCs)
      }
      getFoldChangeSeqs <- function(richnessVec) {
        lenInd=length(richnessVec) - 1 
        fcVec = sapply(1:lenInd,function(ind){
          return(richnessVec[ind+1]/richnessVec[ind])
        })
        print(lenInd)
        names(fcVec) = names(richnessVec)[1:lenInd]
        return(fcVec)
      }
      getSurvVec <- function(subjFCs, cut, mode = "decrease") {
        fcCutDigit = NULL
        interval=3
        survTime = 0
        survEvent = "None"
        firstSample = names(subjFCs)[1]
        
        if (mode == "decrease") {
          fcCutDigit <- subjFCs <= cut
          eventInds = which(fcCutDigit)
          if (length(eventInds)!= 0) {
            eventInds = min(eventInds)
            survTime = interval*eventInds
            survEvent="decrease"
          }
        }
        if (mode == "increase") {
          fcCutDigit <- subjFCs >= cut
          eventInds = which(fcCutDigit)
          if (length(eventInds)!= 0) {
            eventInds = min(eventInds)
            survTime = interval*eventInds
            survEvent="increase"
          }
        }
        return(list(time=survTime, event=survEvent, sample=firstSample))
      } 
      is.sequential <- function(x){
        all(diff(x) == diff(x)[1])
      }
      getSurvVecThrough <- function(subjFCs, cut, mode = "decrease") {
        lenAll = length(subjFCs)
        lenInd = length(subjFCs)-1
        seqs = 1:lenInd
        fcCutDigit = NULL
        interval=3
        survTime = 0
        survEvent = "None"
        
        results = sapply(seqs, function(currInd) {
          currSubjFCs = subjFCs[currInd:lenAll]
          result = getSurvVec(currSubjFCs, cut, mode)
        }, simplify = F)
        
        return(results)
        
        
        
      } 
      getSurvivalKMData <- function(metaTab, mgsMat, dCut, mode="decrease") {
        #wellness data
        #decrease
        #increase
        
        wellnessMetadata = metaTab[metaTab$dataset.ID=="id28",]
        wellnessSubjs = (wellnessMetadata$metadata.ID)
        missingSubjs1 = names(table(wellnessSubjs)[table(wellnessSubjs)==3])
        exceptionSubjs = c("X3452","X3834","X3918","X4215")
        missingSubjs2 = c("X3913","X3535","X3016")
        missingSubjs1 = missingSubjs1[!missingSubjs1 %in% exceptionSubjs]
        
        wellnessMetadata = wellnessMetadata[!wellnessMetadata$metadata.ID %in% missingSubjs1, ]
        wellnessMetadata = wellnessMetadata[!wellnessMetadata$metadata.ID %in% missingSubjs2, ]
        wellnessMgsMat = mergeMatUpdated[,wellnessMetadata$sample.ID]
        wellnessSamples = wellnessMetadata$sample.ID
        wellnessUniqSubjs = unique(wellnessMetadata$metadata.ID)
        wellnessUniqSubjVisit1 = paste(wellnessUniqSubjs, "_v1", sep="")
        wellnessMgsV1Mat = wellnessMgsMat[,wellnessUniqSubjVisit1]
        
        wellnessSurvList = do.call(c,sapply(wellnessUniqSubjs, function(subj) {
          currSubjFCs = checkRichnessChangeBySubject(wellnessMetadata, subj)
          #print(length(currSubjFCs))
          #currSubSurvVec = getSurvVec(currSubjFCs, dCut, mode)
          currSubSurvVecThrough = getSurvVecThrough(currSubjFCs, dCut, mode)
          
          return(currSubSurvVecThrough) 
          
        }, simplify = F))
        
        survTab = data.frame(time=do.call(c,lapply(wellnessSurvList, function(x) x$time)),
                             status=do.call(c,lapply(wellnessSurvList, function(x) x$event)),
                             sample=do.call(c,lapply(wellnessSurvList, function(x) x$sample)),
                             stringsAsFactors = F)
        
        require(survival)
        #getStatisticsPerEach
        
        presenceMode = T
        if (presenceMode) {
          
          wellnessMgsMatchedMat = mergeMatUpdated[,match(survTab$sample,colnames(mergeMatUpdated))]
          allMsps = rownames(wellnessMgsMatchedMat)
          
          survPvalues = sapply(allMsps, function(currMsp) {
            currMspSubjDigit = wellnessMgsMatchedMat[currMsp,]>0
            currMspSubjDigit = currMspSubjDigit * 1
            currSuvTab = survTab
            currSuvTab$msp = currMsp
            currSuvTab$presence = currMspSubjDigit
            lenTypes = length(table(currSuvTab$presence))
            if (lenTypes==1) return(1)
            km_fit <- survfit(Surv(currSuvTab$time, currSuvTab$status == mode) ~ currSuvTab$presence)
            #plot(km_fit)
            #cox_fit <- coxph(Surv(currSuvTab$time, currSuvTab$status == mode) ~ currSuvTab$presence)
            
            diff_out <- survdiff(Surv(currSuvTab$time, currSuvTab$status == mode) ~ currSuvTab$presence)
            pval = getChiPval(diff_out$chisq,1)
            oeRatio=diff_out$obs / diff_out$exp
            names(oeRatio) = names(diff_out$n)
            presenceRatio=oeRatio["currSuvTab$presence=1"]
            if (is.nan(presenceRatio)) return(1)
            if (presenceRatio>1) pval = -1 * pval
            return(pval)
            #based
          })
          
          
        }
        
        abundanceMode = T
        if (abundanceMode) {
          wellnessMgsMatchedMat = mergeMatUpdated[,match(survTab$sample,colnames(mergeMatUpdated))]
          allMsps = rownames(wellnessMgsMatchedMat)
          
          
          coxPvalues = sapply(allMsps, function(currMsp) {
            currMspSubjAbd = wellnessMgsMatchedMat[currMsp,]
            if (sum(currMspSubjAbd) == 0 ) return(1)
            currSuvTab = survTab
            currSuvTab$msp = currMsp
            currSuvTab$abundance = currMspSubjAbd
            
            #plot(km_fit)
            cox_fit <- coxph(Surv(currSuvTab$time, currSuvTab$status == mode) ~ currSuvTab$abundance)
            cox_summary = summary(cox_fit)
            pvalue = cox_summary$coefficients[5]
            if (is.na(cox_summary$coefficients[1])) return(1)
            if (cox_summary$coefficients[1] < 0) return( -1 * pvalue) 
            #based
          })
          
        }
        
      }
      
      survivalStats = getSurvivalKMData(metaTab, mgsMat, decreaseCut)
      survivalStatsRData = "C://Data//comparative.analysis.healthy.sweden//survivalStats.RData"
      save(survivalStats, file=survivalStatsRData)
      
    }
    
    loadMode = T
    if (loadMode) {
      survivalStatsRData = "C://Data//comparative.analysis.healthy.sweden//survivalStats.RData"
      load(survivalStatsRData)
    }
    
    colonizingResistantSurvList = list(colonizing=survivalStats$surv.rates[survivalStats$mgs %in% colonizingSpecies],
                                       resistant=survivalStats$surv.rates[survivalStats$mgs %in% resistantSpecies])
    t.test(colonizingResistantSurvList$colonizing, colonizingResistantSurvList$resistant)
    colonizingResistantSurvTab = melt(colonizingResistantSurvList)
    boxplot(colonizingResistantSurvList)
    
    ggplot(colonizingResistantSurvTab, aes(x=L1, y=value, fill=L1)) +  
      scale_fill_manual(values=c("#6b4695","gray")) +
      geom_boxplot() +
      xlab("") +
      ylab("") + 
      theme(panel.grid.major = element_blank(), 
            panel.grid.minor = element_blank(),
            legend.position = "none",
            axis.text.x = element_text(angle = 60, hjust=1),
            panel.background = element_rect(fill = "white", 
                                            colour = "black", 
                                            size = 0.5, 
                                            linetype = "solid"))
    
  }
  
  bothMarkovKmMode= F
  if (bothMarkovKmMode) {
    wellnessMetadata = basicMetaMapUpdated[basicMetaMapUpdated$dataset.ID=="id28",]
    
    mgsMat = mergeMatUpdated
    metaTab = basicMetaMapUpdated
    
    decreaseCut = 0.9106159
    increaseCut = 1.096916
    
    survivalStats = getSurvivalRates(metaTab, mgsMat, decreaseCut)
    
    colonizingRatios = getColonizationColonizingProbs(metaTab, mgsMat)
    resistantRatios = getColonizationResistantProbs(metaTab, mgsMat)
    colonizedRatios = (colonizingRatios+1e-4)/(resistantRatios+1e-4)
    colonizedRatios[colonizingRatios==0 & resistantRatios==0] = NA
    markovStatTab= data.frame(inflow = colonizingRatios,
                              outflow = resistantRatios, 
                              ratio=colonizedRatios,
                              species = getSpeciesName(names(colonizingRatios),taxo))
    
    table(rownames(survivalStats)==rownames(markovStatTab)) 
    tempDynamicsResultTab = cbind(markovStatTab[,1:3], survivalStats)
    tempDynamicsResultFile = "tempDynamicsResultTab.for.reza.20191022.txt"
    write.table(tempDynamicsResultTab, tempDynamicsResultFile, sep="\t", quote = F)
    
    
  }
  
  wellnessMode = T
  if (wellnessMode) {
    
    if (F) {
      wellnessMetadata = basicMetaMapUpdated[basicMetaMapUpdated$dataset.ID=="id28",]
      exSubjs = c("X3913","X3535","X3016")
      wellnessMetadata = wellnessMetadata[!wellnessMetadata$metadata.ID %in% exSubjs,]
      
      richnessByVisits = split(wellnessMetadata$GeneRichness, wellnessMetadata$subtype)
      richnessByVisitsMelt = melt(richnessByVisits)
      ggplot(richnessByVisitsMelt, aes(x=L1, y=value)) + geom_boxplot() + 
        geom_jitter(aes(colour = L1),size=0.3,width=0.25) + theme_bw() +
        ylab("Gene richness") + xlab("") +
        
        
        wellnessSubjs = unique(wellnessMetadata$metadata.ID)
      wellnessSamples = wellnessMetadata$sample.ID
      wellnessMgsMat = mergeMatUpdated[,wellnessSamples]
      
      sort(wellnessMgsMat["msp_0213",],decreasing = T)[1:10]
      
    }
    
    regressGeneRichnessByVisits = sapply(wellnessSubjs, function(subj) {
      currMetadata = wellnessMetadata[wellnessMetadata$metadata.ID == subj,]
      currVisits = currMetadata$subtype
      currVisits = gsub("v","",currVisits)
      currVisits = as.numeric(currVisits)
      currGeneRichness = currMetadata$GeneRichness
      
      currStat = data.frame(visit=currVisits, geneRichness = currGeneRichness)
      
      ### statistics ###
      lmOut = lm(geneRichness ~ visit, data=currStat)
      lmSummary = summary(lmOut)
      print(subj)
      print(lmSummary)
      pvalue = lmSummary$coefficients[2,4]
      coeff = lmSummary$coefficients[2,1]
      
      return(c(pvalue,coeff))
    })
    corrGeneRichnessByVisits = sapply(wellnessSubjs, function(subj) {
      currMetadata = wellnessMetadata[wellnessMetadata$metadata.ID == subj,]
      currVisits = currMetadata$subtype
      currVisits = gsub("v","",currVisits)
      currVisits = as.numeric(currVisits)
      currGeneRichness = currMetadata$GeneRichness
      corOut = cor.test(currVisits, currGeneRichness)
      pvalue = corOut$p.value
      coeff = corOut$estimate
      
      return(c(pvalue,coeff))
    })
    cvGeneRichnessByVisits = sapply(wellnessSubjs, function(subj) {
      currMetadata = wellnessMetadata[wellnessMetadata$metadata.ID == subj,]
      currVisits = currMetadata$subtype
      currVisits = gsub("v","",currVisits)
      currVisits = as.numeric(currVisits)
      currGeneRichness = currMetadata$GeneRichness
      currSd = sd(currGeneRichness)
      currMean = mean(currGeneRichness)
      currCv = currSd/currMean
      return(currCv) 
    })
    meanGeneRichnessByVisits = sapply(wellnessSubjs, function(subj) {
      currMetadata = wellnessMetadata[wellnessMetadata$metadata.ID == subj,]
      currVisits = currMetadata$subtype
      currVisits = gsub("v","",currVisits)
      currVisits = as.numeric(currVisits)
      currGeneRichness = currMetadata$GeneRichness
      currMean = mean(currGeneRichness)
      return(currMean) 
    })
    
    richnessFCofWellness = do.call(c,sapply(wellnessSubjs, function(subj) {
      currMetadata = wellnessMetadata[wellnessMetadata$metadata.ID == subj,]
      currVisits = currMetadata$subtype
      currVisits = gsub("v","",currVisits)
      currVisits = as.numeric(currVisits)
      currGeneRichness = currMetadata$GeneRichness
      names(currGeneRichness) = currVisits
      currGeneRichness = currGeneRichness[order(names(currGeneRichness))]
      lenRichness = length(currGeneRichness)
      iterations  = lenRichness - 1
      
      seqTab = rbind(c(1,2),
                     c(2,3),
                     c(3,4))
      seqNameTab = apply(seqTab, 1:2, function(x) paste(subj, "_v", x, sep=""))
      seqNames = apply(seqNameTab, 1, function(x) paste(x, collapse = "."))
      
      selInds = seqTab[,1] %in% currVisits & seqTab[,2] %in% currVisits
      
      seqTab = seqTab[selInds,]
      seqNameTab = seqNameTab[selInds,]
      seqNames = seqNames[selInds]
      
      if (class(seqTab)=="numeric") {
        foldChanges = currGeneRichness[names(currGeneRichness)==seqTab[2]]/currGeneRichness[names(currGeneRichness)==seqTab[1]]
      }
      if (class(seqTab)=="matrix") {
        foldChanges = apply(seqTab, 1, function(currSeq){
          after =currSeq[2]
          before=currSeq[1]
          foldChange = currGeneRichness[names(currGeneRichness)==after]/currGeneRichness[names(currGeneRichness)==before]
          return(foldChange)
        })
      }
      names(foldChanges)=seqNames
      return(foldChanges)
    }, simplify = F))
    
    outflowDiffofWellness = do.call(c,sapply(wellnessSubjs, function(subj) {
      currMetadata = wellnessMetadata[wellnessMetadata$metadata.ID == subj,]
      currVisits = currMetadata$subtype
      currVisits = gsub("v","",currVisits)
      currVisits = as.numeric(currVisits)
      currOutflow = currMetadata$scaledOutflow
      names(currOutflow) = currVisits
      currOutflow = currOutflow[order(names(currOutflow))]
      lenRichness = length(currOutflow)
      iterations  = lenRichness - 1
      
      seqTab = rbind(c(1,2),
                     c(2,3),
                     c(3,4))
      seqNameTab = apply(seqTab, 1:2, function(x) paste(subj, "_v", x, sep=""))
      seqNames = apply(seqNameTab, 1, function(x) paste(x, collapse = "."))
      
      selInds = seqTab[,1] %in% currVisits & seqTab[,2] %in% currVisits
      
      seqTab = seqTab[selInds,]
      seqNameTab = seqNameTab[selInds,]
      seqNames = seqNames[selInds]
      
      if (class(seqTab)=="numeric") {
        #foldChanges = currOutflow[names(currOutflow)==seqTab[2]]/currOutflow[names(currOutflow)==seqTab[1]]
        foldChanges = currOutflow[names(currOutflow)==seqTab[2]] - currOutflow[names(currOutflow)==seqTab[1]]
        
      }
      if (class(seqTab)=="matrix") {
        foldChanges = apply(seqTab, 1, function(currSeq){
          after =currSeq[2]
          before=currSeq[1]
          #foldChange = currOutflow[names(currOutflow)==after]/currOutflow[names(currOutflow)==before]
          foldChange = currOutflow[names(currOutflow)==after] - currOutflow[names(currOutflow)==before]
          return(foldChange)
        })
      }
      names(foldChanges)=seqNames
      return(foldChanges)
    }, simplify = F))
    
    inflowDiffofWellness = do.call(c,sapply(wellnessSubjs, function(subj) {
      currMetadata = wellnessMetadata[wellnessMetadata$metadata.ID == subj,]
      currVisits = currMetadata$subtype
      currVisits = gsub("v","",currVisits)
      currVisits = as.numeric(currVisits)
      currInflow = currMetadata$scaledInflow
      names(currInflow) = currVisits
      currInflow = currInflow[order(names(currInflow))]
      lenRichness = length(currInflow)
      iterations  = lenRichness - 1
      
      seqTab = rbind(c(1,2),
                     c(2,3),
                     c(3,4))
      seqNameTab = apply(seqTab, 1:2, function(x) paste(subj, "_v", x, sep=""))
      seqNames = apply(seqNameTab, 1, function(x) paste(x, collapse = "."))
      
      selInds = seqTab[,1] %in% currVisits & seqTab[,2] %in% currVisits
      
      seqTab = seqTab[selInds,]
      seqNameTab = seqNameTab[selInds,]
      seqNames = seqNames[selInds]
      
      if (class(seqTab)=="numeric") {
        #foldChanges = currInflow[names(currInflow)==seqTab[2]]/currInflow[names(currInflow)==seqTab[1]]
        foldChanges = currInflow[names(currInflow)==seqTab[2]] - currInflow[names(currInflow)==seqTab[1]]
        
      }
      if (class(seqTab)=="matrix") {
        foldChanges = apply(seqTab, 1, function(currSeq){
          after =currSeq[2]
          before=currSeq[1]
          #foldChange = currInflow[names(currInflow)==after]/currInflow[names(currInflow)==before]
          foldChange = currInflow[names(currInflow)==after] - currInflow[names(currInflow)==before]
          return(foldChange)
        })
      }
      names(foldChanges)=seqNames
      return(foldChanges)
    }, simplify = F))
    
    #quantiles when all subjects used
    #decreaseCut = 0.9106159
    #increaseCut = 1.096916
    
    #quantiles when 86 subjects used
    decreaseCut = 0.9126689
    increaseCut = 1.096939
    
    par(mar=c(1,1,1,1))
    plot(density(richnessFCofWellness), xaxt = "n", yaxt="n", main="", ylab="", xlab = "", las=1)
    
    par(mar=c(5,4,4,1)) 
    abline(v=decreaseCut, lty=3, col="gray")
    abline(v=increaseCut, lty=3, col="gray")
    
    getRichnessNamesByBeforeVisit <- function(richnessFCofWellness) {
      newNames = unlist(sapply(names(richnessFCofWellness), function(nn){
        strsplitted = strsplit(nn, split = "\\.")[[1]]
        before = strsplitted[2]
        after = strsplitted[3]
        return(before)
      }, simplify = F))
      names(richnessFCofWellness) = newNames
      return(richnessFCofWellness)
    }
    #richnessFCofWellnessNew = getRichnessNamesByBeforeVisit(richnessFCofWellness)
    getPairOfRichnessDecrease <- function(richnessFCofWellness, cut=0.1) {
      dCut = as.numeric(quantile(richnessFCofWellness, cut))
      print(dCut)
      richnessCutFCs = richnessFCofWellness[richnessFCofWellness<=dCut]
      pairList = sapply(names(richnessCutFCs), function(nn){
        strsplitted = strsplit(nn, split = "\\.")[[1]]
        before = strsplitted[2]
        after = strsplitted[3]
        return(c(before=before,after=after))
      }, simplify = F)
      pairMat = do.call(rbind,pairList)
      rownames(pairMat) = NULL
      pairTab = data.frame(pairMat,
                           fc = richnessCutFCs)
      rownames(pairTab) = NULL
      return(pairTab)
    }
    getPairOfRichnessIncrease <- function(richnessFCofWellness, cut=0.1) {
      dCut = as.numeric(quantile(richnessFCofWellness, 1-cut))
      print(dCut)
      richnessCutFCs = richnessFCofWellness[richnessFCofWellness>=dCut]
      pairList = sapply(names(richnessCutFCs), function(nn){
        strsplitted = strsplit(nn, split = "\\.")[[1]]
        before = strsplitted[2]
        after = strsplitted[3]
        return(c(before=before,after=after))
      }, simplify = F)
      pairMat = do.call(rbind,pairList)
      rownames(pairMat) = NULL
      pairTab = data.frame(pairMat,
                           fc = richnessCutFCs)
      rownames(pairTab) = NULL
      return(pairTab)
    }
    
    
    sortMgsMatByPairs <- function(mgsMat, pairTab) {
      beforeMat = mgsMat[,match(pairTab[,1], colnames(mgsMat))]
      afterMat = mgsMat[,match(pairTab[,2], colnames(mgsMat))]
      sortedMat = cbind(beforeMat, afterMat)
      return(sortedMat)
    }
    getWilcoxStatByPairs <-function(mgsMat, pairTab, taxo, ImputeVal=10) {
      beforeMat = mgsMat[,match(pairTab[,1], colnames(mgsMat))]
      afterMat = mgsMat[,match(pairTab[,2], colnames(mgsMat))]
      caseMeans = rowMeans(afterMat)
      contMeans = rowMeans(beforeMat)
      foldChange = caseMeans / contMeans
      
      dataMat = cbind(afterMat, beforeMat)
      dataLab = c(rep("after",dim(afterMat)[2]),
                  rep("before",dim(beforeMat)[2]))
      
      currStats = testRelations(dataMat, dataLab, "wilcoxon", paired=T)
      currStats$msp = rownames(currStats)
      currStats$species = getSpeciesName(currStats$msp, taxo)
      currStats$foldChange = foldChange
      currStats$after = caseMeans
      currStats$before = contMeans
      currStats$p[is.nan(currStats$p)]=1
      currStats$q[is.nan(currStats$q)]=1
      currStats$foldChange[is.nan(currStats$foldChange)]=1
      currStats$l2fc=log2(currStats$foldChange)
      currStats$l2fc[currStats$l2fc==Inf]=ImputeVal
      currStats$l2fc[currStats$l2fc==-Inf]=-ImputeVal
      
      return(currStats)
      
    }
    getWilcoxClinicalStatByPairs <-function(clinMat, pairTab, ImputeVal=10) {
      beforeMat = clinMat[,match(pairTab[,1], colnames(clinMat))]
      afterMat = clinMat[,match(pairTab[,2], colnames(clinMat))]
      caseMeans = rowMeans(afterMat)
      contMeans = rowMeans(beforeMat)
      foldChange = caseMeans / contMeans
      
      dataMat = cbind(afterMat, beforeMat)
      dataLab = c(rep("after",dim(afterMat)[2]),
                  rep("before",dim(beforeMat)[2]))
      
      pvalues = sapply(rownames(clinMat), function(currFeature){
        wilcoxOut = wilcox.test(afterMat[currFeature,], beforeMat[currFeature,], paired = F)
        return(wilcoxOut$p.value)
      })
      pvalues[is.nan(pvalues)]=1
      currStats = data.frame(pvalue=pvalues)
      #currStats = testRelations(dataMat, dataLab, "wilcoxon", paired=T)
      currStats$feature = rownames(clinMat)
      currStats$foldChange = foldChange
      currStats$after = caseMeans
      currStats$before = contMeans
      currStats$p[is.nan(currStats$p)]=1
      currStats$foldChange[is.nan(currStats$foldChange)]=1
      currStats$l2fc=log2(currStats$foldChange)
      currStats$l2fc[currStats$l2fc==Inf]=ImputeVal
      currStats$l2fc[currStats$l2fc==-Inf]=-ImputeVal
      
      return(currStats)
      
    }
    
    #decreasedPairTab = getPairOfRichnessDecrease(richnessFCofWellness, 0.1)
    #increasedPairTab = getPairOfRichnessIncrease(richnessFCofWellness, 0.1)
    
    decreasedPairTab = getPairOfRichnessDecrease(richnessFCofWellness, 1)
    increasedPairTab = getPairOfRichnessIncrease(richnessFCofWellness, 1)
    decreasedPairTab2 = unique(decreasedPairTab)
    increasedPairTab2 = unique(increasedPairTab)
    
    decreasedPairTabOutflow = getPairOfRichnessDecrease(outflowDiffofWellness, 1)
    increasedPairTabOutflow = getPairOfRichnessIncrease(outflowDiffofWellness, 1)
    decreasedPairTabOutflow = unique(decreasedPairTabOutflow)
    increasedPairTabOutflow = unique(increasedPairTabOutflow)
    increasedPairTabOutflow$subj = gsub("_v[1-4]","",increasedPairTabOutflow$before)
    increasedPairTabOutflow$outflow = wellnessMetadata$scaledOutflow[match(increasedPairTabOutflow$before, wellnessMetadata$sample.ID)]
    plot(increasedPairTabOutflow$outflow,increasedPairTabOutflow$fc, xlab="outflow enrichment", ylab="after - before (outflow)")
    abline(lm(increasedPairTabOutflow$fc~increasedPairTabOutflow$outflow ), col="red")
    cor.test(increasedPairTabOutflow$outflow,increasedPairTabOutflow$fc)
    
    decreasedPairTabInflow = getPairOfRichnessDecrease(inflowDiffofWellness, 1)
    increasedPairTabInflow = getPairOfRichnessIncrease(inflowDiffofWellness, 1)
    decreasedPairTabInflow = unique(decreasedPairTabInflow)
    increasedPairTabInflow = unique(increasedPairTabInflow)
    
    
    
    
    wellnessMgsWilcoxStatsDecreasedPair = getWilcoxStatByPairs(wellnessMgsMat, decreasedPairTab, taxo)
    wellnessMgsWilcoxStatsIncreasedPair = getWilcoxStatByPairs(wellnessMgsMat, increasedPairTab, taxo)
    
    
    wellnessClinicalWilcoxStatsDecreasedPair = getWilcoxClinicalStatByPairs(t(clinicalMatrix), decreasedPairTab)
    wellnessClinicalWilcoxStatsIncreasedPair = getWilcoxClinicalStatByPairs(t(clinicalMatrix), increasedPairTab)
    
    
    plotLfcMode = T
    if (plotLfcMode) {
      welnessMgsLfcDecreaseIncrease = data.frame(lfc_decrease = wellnessMgsWilcoxStatsDecreasedPair$l2fc,
                                                 lfc_increase = wellnessMgsWilcoxStatsIncreasedPair$l2fc, 
                                                 species = wellnessMgsWilcoxStatsDecreasedPair$species)
      
      rownames(welnessMgsLfcDecreaseIncrease) = wellnessMgsWilcoxStatsDecreasedPair$msp
      welnessMgsLfcDecreaseIncrease$lfc_decrease[welnessMgsLfcDecreaseIncrease$lfc_decrease==10] = NA
      welnessMgsLfcDecreaseIncrease$lfc_decrease[welnessMgsLfcDecreaseIncrease$lfc_decrease== -10] = NA
      welnessMgsLfcDecreaseIncrease$lfc_increase[welnessMgsLfcDecreaseIncrease$lfc_increase==10] = NA
      welnessMgsLfcDecreaseIncrease$lfc_increase[welnessMgsLfcDecreaseIncrease$lfc_increase== -10] = NA
      naInds = is.na(welnessMgsLfcDecreaseIncrease$lfc_decrease) | is.na(welnessMgsLfcDecreaseIncrease$lfc_increase)
      welnessMgsLfcDecreaseIncrease = welnessMgsLfcDecreaseIncrease[!naInds,]
      cor.test(welnessMgsLfcDecreaseIncrease$lfc_decrease, welnessMgsLfcDecreaseIncrease$lfc_increase)
      
      ggplot(welnessMgsLfcDecreaseIncrease, aes(x=lfc_decrease, y=lfc_increase, fill=lfc_increase)) + 
        geom_point(size=2, shape=21, colour="black")  +
        xlab("") + ylab("")+
        geom_hline(yintercept = 0) + geom_vline(xintercept = 0)+
        geom_smooth(method='lm',formula=y~x, se=FALSE, color="red")+
        scale_fill_gradient2(midpoint = 0.2, low="#FFD70044", mid = "white", high = "purple") +
        theme(panel.grid.major = element_blank(), 
              panel.grid.minor = element_blank(),
              legend.position = "none",
              panel.background = element_rect(fill = "white", 
                                              colour = "black", 
                                              size = 0.5, 
                                              linetype = "solid"))
      
    }
    
    
    wellnessSamples[grepl("3920",wellnessSamples)]
    wellnessSamples[grepl("3512",wellnessSamples)]
    wellnessSamples[grepl("3834",wellnessSamples)]
    wellnessSamples[grepl("3119",wellnessSamples)]
    wellnessSamples[grepl("3003",wellnessSamples)]
    
    
    ab3885Lfc = wellnessMgsMat[,"X3885_v4"]/ wellnessMgsMat[,"X3885_v1"]
    ab3512Lfc = wellnessMgsMat[,"X3512_v4"]/ wellnessMgsMat[,"X3512_v1"]
    ab3512Lfc = log2(ab3512Lfc)
    ab3920Lfc = wellnessMgsMat[,"X3920_v4"]/ wellnessMgsMat[,"X3920_v1"]
    ab3920Lfc = log2(ab3920Lfc)
    ab3834Lfc = wellnessMgsMat[,"X3834_v3"]/ wellnessMgsMat[,"X3834_v1"]
    ab3834Lfc = log2(ab3834Lfc)
    ab3119Lfc = wellnessMgsMat[,"X3119_v43"]/ wellnessMgsMat[,"X3119_v1"]
    ab3119Lfc = log2(ab3119Lfc)
    
    ab3003Lfc = wellnessMgsMat[,"X3003_v4"]/ wellnessMgsMat[,"X3003_v1"]
    ab3003Lfc = log2(ab3003Lfc)
    
    plot(ab3003Lfc[names(ab3003Lfc) %in% bothSpecies], 
         volcano.stats.mgs.normal$lfc[names(ab3003Lfc) %in% bothSpecies])
    abline(h=0,v=0)
    
    plot(ab3119Lfc[names(ab3119Lfc) %in% bothSpecies], 
         volcano.stats.mgs.normal$lfc[names(ab3119Lfc) %in% bothSpecies])
    abline(h=0,v=0)
    
    plot(ab3834Lfc[names(ab3834Lfc) %in% bothSpecies], 
         volcano.stats.mgs.normal$lfc[names(ab3834Lfc) %in% bothSpecies])
    abline(h=0,v=0)
    
    plot(ab3920Lfc[names(ab3920Lfc) %in% bothSpecies], 
         volcano.stats.mgs.normal$lfc[names(ab3920Lfc) %in% bothSpecies])
    abline(h=0,v=0)
    plot(ab3512Lfc[names(ab3512Lfc) %in% bothSpecies], 
         volcano.stats.mgs.normal$lfc[names(ab3512Lfc) %in% bothSpecies])
    abline(h=0,v=0)
    
    abSubjPair = data.frame(before=c("X3885_v1","X2881_v1","X3162_v1","X3163_v1","X3210_v2","X3814_v3","X3829_v1","X4028_v3","X4135_v1","X4151_v1"),
                            after =c("X3885_v4","X2881_v2","X3162_v2","X3163_v2","X3210_v3","X3814_v4","X3829_v2","X4028_v4","X4135_v2","X4151_v2"),
                            stringsAsFactors = F)
    abSubjPairNew = data.frame(before=c("X3885_v1","X2881_v1","X3162_v1","X3163_v1","X3210_v2","X3829_v1","X4135_v1","X4151_v1"),
                               after =c("X3885_v3","X2881_v3","X3162_v3","X3163_v3","X3210_v4","X3829_v3","X4135_v3","X4151_v3"),
                               stringsAsFactors = F)
    
    
    abSubjMgsWilcoxDecreasedPair = getWilcoxStatByPairs(wellnessMgsMat, abSubjPair, taxo)
    abSubjNewMgsWilcoxDecreasedPair = getWilcoxStatByPairs(wellnessMgsMat, abSubjPairNew, taxo)
    
    ff = data.frame(ab=abSubjMgsWilcoxDecreasedPair$l2fc, normal=volcano.stats.mgs.normal$lfc)
    rownames(ff) = rownames(abSubjMgsWilcoxDecreasedPair)
    plot(ff$ab[rownames(ff) %in% bothSpecies], ff$normal[rownames(ff) %in% bothSpecies])
    
    plot(abSubjMgsWilcoxDecreasedPair$l2fc, volcano.stats.mgs.normal$lfc)
    plot(abSubjNewMgsWilcoxDecreasedPair$l2fc, volcano.stats.mgs.normal$lfc)
    
    plot((wellnessMgsWilcoxStatsDecreasedPair$l2fc), 
         (wellnessMgsWilcoxStatsIncreasedPair$l2fc), 
         xlab="Log2FC of decresed gene richness",
         ylab="Log2FC of increased gene richness", 
         main="corr=-0.361, P<1e-15")
    abline(a=0,b=-1, col="red")
    cor.test((wellnessMgsWilcoxStatsDecreasedPair$l2fc), 
             (wellnessMgsWilcoxStatsIncreasedPair$l2fc)) 
    
    
    hist(richnessFCofWellness, main="", xlab="fold change")
    hist(log2(richnessFCofWellness), main="", xlab="log2 FC")
    
    plot(cvGeneRichnessByVisits,meanGeneRichnessByVisits, xlim=c(0,0.1))
    
    regressGeneRichnessByVisits = t(regressGeneRichnessByVisits)
    corrGeneRichnessByVisits = t(corrGeneRichnessByVisits)
    target = "X3809"
    target = "X3917"
    target = "X3665"
    target = "X2809"
    target = "X3834"
    target = "X3920"
    target = "X3860"
    
    plot(getNumbersFromWellnessVisits(wellnessMetadata$subtype[wellnessMetadata$metadata.ID==target]), 
         wellnessMetadata$GeneRichness[wellnessMetadata$metadata.ID==target],
         ylab="gene richness", xlab="visit",
         main=target)
    
    hist(wellnessMetadata$GeneRichness)
    
    require(PairedData)
    pd <- paired(before, after)
    plot(pd, type = "profile") + theme_bw()
    
    targetSubjs = c("X3814", "X3920", "X3707", "X3918", "X4037", "X3614")
    wellnessMetadata2 = wellnessMetadata[wellnessMetadata$metadata.ID %in% targetSubjs, ]
    
    getPairedWellness <- function(wellnessMetadata, refVisit, targetVisit) {
      refVisitMetadata = wellnessMetadata[wellnessMetadata$subtype == refVisit, ]
      targetVisitMetadata = wellnessMetadata[wellnessMetadata$subtype == targetVisit, ]
      
      overlappedSubjs = refVisitMetadata$metadata.ID[refVisitMetadata$metadata.ID %in% targetVisitMetadata$metadata.ID]
      refVisitMetadata = refVisitMetadata[match(overlappedSubjs, refVisitMetadata$metadata.ID),]
      targetVisitMetadata = targetVisitMetadata[match(overlappedSubjs, targetVisitMetadata$metadata.ID),]
      
      return(list(ref = refVisitMetadata$GeneRichness,
                  target=targetVisitMetadata$GeneRichness,
                  subjs = refVisitMetadata$metadata.ID))
      
      
    }
    
    showPairedPlotFrom <- function(wellnessMetadata, refVisit, targetVisit) {
      outList = getPairedWellness(wellnessMetadata, refVisit, targetVisit)
      outList = getPairedWellness(wellnessMetadata, refVisit, targetVisit)
      before = outList$ref
      after = outList$target
      pd <- paired(before, after)
      plot(pd, type = "profile") + theme_bw()
    }
    getPairedWellness(wellnessMetadata2, "v2", "v3")
    showPairedPlotFrom(wellnessMetadata2, "v1", "v2")
    
    
    x3809Samples =wellnessMetadata$sample.ID[wellnessMetadata$metadata.ID=="X3809"]
    x3809MgsMat = mergeMatUpdated[,x3809Samples]
    x3809MgsTab = data.frame(species = getSpeciesName(rownames(x3809MgsMat),taxo),
                             x3809MgsMat)
    
    x3917Samples =wellnessMetadata$sample.ID[wellnessMetadata$metadata.ID=="X3917"]
    x3917MgsMat = mergeMatUpdated[,x3917Samples]
    x3917MgsTab = data.frame(species = getSpeciesName(rownames(x3917MgsMat),taxo),
                             x3917MgsMat)
    
    x3920Samples =wellnessMetadata$sample.ID[wellnessMetadata$metadata.ID=="X3920"]
    x3920MgsMat = mergeMatUpdated[,x3920Samples]
    x3920MgsTab = data.frame(species = getSpeciesName(rownames(x3920MgsMat),taxo),
                             x3920MgsMat)
    
    testMat = rbind(c(1,3,4),x3920MgsMat)
    testCor = cor(t(testMat))
    
    
    
  }
  
  treatmentMode = F
  if (treatmentMode) {
    antibioticsMetadata = basicMetaMapUpdated[basicMetaMapUpdated$dataset.ID=="id51",]
    t2dMetforminMetadata = basicMetaMapUpdated[basicMetaMapUpdated$dataset.ID=="id53",]
    rccImmuneThMetadata = basicMetaMapUpdated[basicMetaMapUpdated$dataset.ID=="id41",]
    nsclcImmuneThMetadata = basicMetaMapUpdated[basicMetaMapUpdated$dataset.ID=="id26",]
    antibioticsMetadataSubjs = getAbSubjs(antibioticsMetadata$metadata.ID)
    #antibioticsMetadata$metadata.ID = antibioticsMetadataSubjs
    antibioticsMetadata$subtype = gsub("Dag","", antibioticsMetadata$subtype)
    antibioticsMetadata$subtype = gsub("opt","", antibioticsMetadata$subtype)
    antibioticsMetadata$subtype = as.numeric(antibioticsMetadata$subtype)
    
    abGeneRichnessByVisit = split(antibioticsMetadata$GeneRichness, antibioticsMetadata$subtype)
    orderedVisits = c("Dag0", "Dag4", "Dag8", "Dag42", "Dag180")
    boxplot(abGeneRichnessByVisit[orderedVisits])
    
    getPairSeqsForAb <- function(seqs) {
      lenSeqs = length(seqs)
      lenInds = lenSeqs-1
      pairList = sapply(1:lenInds, function(x) seqs[c(x, x+1)] , simplify = F)
      pairTab = do.call(rbind, pairList)
      return(pairTab)
    }
    
    findAbPairs <- function(abMetadata, seqs) {
      seqTab = getPairSeqsForAb( seqs )
      subjs = getAbSubjs(abMetadata$metadata.ID)
      uniqSubjs = unique(subjs)
      
      pairTab = do.call(rbind,sapply(uniqSubjs, function(subj){
        currSubjTab = abMetadata[subjs == subj,]
        out = apply(seqTab, 1, function(currRow){
          before = currRow[1]
          after = currRow[2]
          
          if (all(currRow %in% currSubjTab$subtype)) {
            beforeSample = currSubjTab$sample.ID[currSubjTab$subtype==before]
            afterSample = currSubjTab$sample.ID[currSubjTab$subtype==after]
            return(c(subj, beforeSample, afterSample))
          }
          return(c(subj, NA,NA))
          
        })
        out = t(out)
        return(out)
      }, simplify = F))
      
      pairTab = pairTab[!is.na(pairTab[,2]),]
      pairTab = data.frame(pairTab, stringsAsFactors = F)
      pairTab$bfMetaId = abMetadata$metadata.ID[match(pairTab[,2],antibioticsMetadata$sample.ID)]
      pairTab$afMetaId = abMetadata$metadata.ID[match(pairTab[,3],antibioticsMetadata$sample.ID)]
      colnames(pairTab) = c("subject","before.sample","after.sample",
                            "before.meta","after.meta")
      pairTab$before.richness = abMetadata$GeneRichness[match(pairTab[,2],antibioticsMetadata$sample.ID)]
      pairTab$after.richness = abMetadata$GeneRichness[match(pairTab[,3],antibioticsMetadata$sample.ID)]
      
      return(pairTab)
      
    }
    abPairTabDecrease1 = findAbPairs(antibioticsMetadata, c(0,4))
    abPairTabDecrease2 = findAbPairs(antibioticsMetadata, c(0,8))
    abPairTabDecrease3 = findAbPairs(antibioticsMetadata, c(0,42))
    abPairTabDecrease4 = findAbPairs(antibioticsMetadata, c(0,180))
    
    abPairTabIncrease = findAbPairs(antibioticsMetadata, c(4,8,42,180))
    # abPairTabDecrease$after.richness/abPairTabDecrease$before.richness
    # abPairTabIncrease$after.richness/abPairTabIncrease$before.richness
    # 
    getWilcoxStatByAbPairs <-function(mgsMat, abPairTab, taxo, ImputeVal=10) {
      beforeMat = mgsMat[,match(abPairTab[,2], colnames(mgsMat))]
      afterMat = mgsMat[,match(abPairTab[,3], colnames(mgsMat))]
      caseMeans = rowMeans(afterMat)
      contMeans = rowMeans(beforeMat)
      foldChange = caseMeans / contMeans
      
      dataMat = cbind(afterMat, beforeMat)
      dataLab = c(rep("after",dim(afterMat)[2]),
                  rep("before",dim(beforeMat)[2]))
      
      currStats = testRelations(dataMat, dataLab, "wilcoxon")
      currStats$msp = rownames(currStats)
      currStats$species = getSpeciesName(currStats$msp, taxo)
      currStats$foldChange = foldChange
      currStats$after = caseMeans
      currStats$before = contMeans
      currStats$p[is.nan(currStats$p)]=1
      currStats$q[is.nan(currStats$q)]=1
      currStats$foldChange[is.nan(currStats$foldChange)]=1
      currStats$l2fc=log2(currStats$foldChange)
      currStats$l2fc[currStats$l2fc==Inf]=ImputeVal
      currStats$l2fc[currStats$l2fc==-Inf]=-ImputeVal
      
      return(currStats)
      
    }
    
    wilcoxStatsDecrease1 = getWilcoxStatByAbPairs(mergeMatUpdated, abPairTabDecrease1, taxo, 15)
    wilcoxStatsDecrease2 = getWilcoxStatByAbPairs(mergeMatUpdated, abPairTabDecrease2, taxo, 15)
    wilcoxStatsDecrease3 = getWilcoxStatByAbPairs(mergeMatUpdated, abPairTabDecrease3, taxo, 15)
    wilcoxStatsDecrease4 = getWilcoxStatByAbPairs(mergeMatUpdated, abPairTabDecrease4, taxo, 15)
    
    writeMode = F
    if (writeMode) {
      write.table(wilcoxStatsDecrease1, 
                  "C://Data//comparative.analysis.healthy.sweden//wilcoxStatsDecrease.day0.day4.txt", 
                  sep="\t")
      write.table(wilcoxStatsDecrease4, 
                  "C://Data//comparative.analysis.healthy.sweden//wilcoxStatsDecrease.day0.day180.txt", 
                  sep="\t")
      
      write.table(abPairTabDecrease1, 
                  "C://Data//comparative.analysis.healthy.sweden//antibiotics.treated.samples.day0.day4.txt", 
                  sep="\t")
      write.table(abPairTabDecrease4, 
                  "C://Data//comparative.analysis.healthy.sweden//antibiotics.treated.samples.day0.day180.txt", 
                  sep="\t")
      
      
    }
    
    
    wilcoxStatsIncrease = getWilcoxStatByAbPairs(mergeMatUpdated, abPairTabIncrease, taxo)
    butyrateMsps = butryCluster332$msps
    
    lfcMat = cbind(wilcoxStatsDecrease1$l2fc,
                   wilcoxStatsDecrease2$l2fc,
                   wilcoxStatsDecrease3$l2fc,
                   wilcoxStatsDecrease4$l2fc)
    rownames(lfcMat) = wilcoxStatsDecrease1$msp
    View(lfcMat[rownames(lfcMat) %in% butyrateMsps,])
    boxplot(lfcMat[rownames(lfcMat) %in% butyrateMsps,])
    hist(wilcoxStatsIncrease$l2fc[wilcoxStatsIncrease$msp %in% butyrateMsps])
    
    plot(density(lfcMat[rownames(lfcMat) %in% butyrateMsps,][,1]))
    lines(density(lfcMat[rownames(lfcMat) %in% butyrateMsps,][,2]), col="red")
    lines(density(lfcMat[rownames(lfcMat) %in% butyrateMsps,][,3]), col="pink")
    lines(density(lfcMat[rownames(lfcMat) %in% butyrateMsps,][,4]), col="gold")
    
    
    hist(lfcMat[rownames(lfcMat) %in% butyrateMsps,][,2])
    hist(lfcMat[rownames(lfcMat) %in% butyrateMsps,][,3])
    hist(lfcMat[rownames(lfcMat) %in% butyrateMsps,][,4])
    
    lfcMat2 = rbind(c(1,2,3,4),
                    lfcMat)
    rownames(lfcMat2) = c("seq", wilcoxStatsDecrease1$msp)
    pairs(lfcMat)
    cor.test(lfcMat[,1],lfcMat[,4])
    corMat = cor(t(lfcMat2))
    
    sort(corMat["seq",],decreasing = T)[1:5]
    sort(corMat["seq",],decreasing = F)[1:5] 
    
    
    
    #### global statistics ####
    
    
    
    
    volcano.stats.mgs.all = read.table(statsMgsForAllSamples, sep="\t", stringsAsFactors = F)
    volcano.stats.mgs.normal = read.table(statsMgsForNormSamples, sep="\t", stringsAsFactors = F)
    volcano.stats.mgs.disease = read.table(statsMgsForDiseaseSamples, sep="\t", stringsAsFactors = F)
    volcano.stats.mgs.industrial = read.table(statsMgsForIndustrialSamples, sep="\t", stringsAsFactors = F)
    volcano.stats.mgs.traditional = read.table(statsMgsForTraditionalSamples, sep="\t", stringsAsFactors = F)
    
    plot(ab3885Lfc[names(ab3885Lfc) %in% bothSpecies], volcano.stats.mgs.normal$lfc[names(ab3885Lfc) %in% bothSpecies])
    abline(h=0,v=0)
    lfcMat3 = cbind(shortAb = -1 *wilcoxStatsDecrease1$l2fc,
                    longAb = -1 * wilcoxStatsDecrease4$l2fc, 
                    normal = volcano.stats.mgs.normal$lfc,
                    disease = volcano.stats.mgs.disease$lfc, 
                    wellness = -1 * wellnessMgsWilcoxStatsDecreasedPair$l2fc)
    #trad = volcano.stats.mgs.traditional$lfc,
    lfcMat4 = lfcMat3[bothSpecies,]
    
    plot(lfcMat4[,"normal"],lfcMat4[,"disease"])
    abline(lm(lfcMat4[,"disease"]~lfcMat4[,"normal"]),col="red")
    
    plot(lfcMat4[,"normal"],lfcMat4[,"wellness"])
    abline(lm(lfcMat4[,"wellness"]~lfcMat4[,"normal"]),col="red")
    
    plot(lfcMat4[,"normal"],lfcMat4[,"shortAb"])
    abline(lm(lfcMat4[,"shortAb"]~lfcMat4[,"normal"]),col="red")
    
    plot(lfcMat4[,"normal"],lfcMat4[,"longAb"])
    abline(lm(lfcMat4[,"longAb"]~lfcMat4[,"normal"]),col="red")
    
    rownames(lfcMat3) = wilcoxStatsDecrease1$msp
    cor(lfcMat3[bothSpecies,])
    
    panel.cor <- function(x, y, digits = 2, prefix = "", cex.cor, ...)
    {
      usr <- par("usr"); on.exit(par(usr))
      par(usr = c(0, 1, 0, 1))
      r <- abs(cor(x, y, method="spearman", use="pairwise.complete.obs"))
      pvalue = (cor.test(x, y, method="spearman", use="pairwise.complete.obs")$p.value)
      
      txt <- format(c(r, 0.123456789), digits = digits)[1]
      txt <- paste0(prefix, txt)
      rect(usr[1], usr[3], usr[2], usr[4], 
           col = "#000088", 
           border = NA)
      if (pvalue <= 0.05) txt <- paste(txt, "*", sep="")
      if(missing(cex.cor)) cex.cor <- 0.8/strwidth(txt)
      text(0.5, 0.5, txt, cex = cex.cor, col="white" )
    }
    
    my_line <- function(x,y,...){
      points(x,y,...)
      abline(h=0, v=0,...)
      reg <- function(x, y, col) abline(lm(y~x), col=col) 
      ok <- is.finite(x) & is.finite(y)
      if (any(ok)) reg(x[ok], y[ok], "red")
    }
    
    lfcMat3Imp = lfcMat3[bothSpecies,]
    lfcMat3Imp[lfcMat3Imp==0]=NA
    lfcMat3Imp[lfcMat3Imp>=10]=NA
    lfcMat3Imp[lfcMat3Imp<=-10]=NA
    
    pairs(lfcMat3Imp, col= "#80808088", 
          lower.panel = my_line, 
          upper.panel = panel.cor)
    
  }
  
}

checkMappingTable = F
if (checkMappingTable) {
  mappingFile = "J:\\Deposit\\Project\\2018_microbiome_atlas\\gctTables\\csvFiles\\merged.csv"
  mappingTab = read.csv(mappingFile,  stringsAsFactors = F)
  table(grepl("_v",mappingTab$sample))
  
  wellnessInds = grepl("_v",mappingTab$sample)
  wellnessIds = mappingTab$sample[wellnessInds]
  wellnessIds = sapply(wellnessIds, function(x) paste("X", x, sep=""))
  mappingTab$sample[wellnessInds] = wellnessIds
  
  table(basicMetaMapUpdated$sample.ID  %in% mappingTab$sample)
  mappingTab = mappingTab[match(basicMetaMapUpdated$sample.ID, mappingTab$sample),]
  #wellnessMappingTab = mappingTab[mappingTab$sample %in% wellnessMetadata$sample.ID, ] 
  
  table(mappingTab$sample)
  samplesByGeo = split(basicMetaMapUpdated$sample.ID, basicMetaMapUpdated$Geography)
  mappingRatesByGeo = lapply(samplesByGeo, function(samples) {
    return(mappingTab$X.mapped_read_count[mappingTab$sample %in% samples])
  })
  
  mappedReadsByGeo = lapply(samplesByGeo, function(samples) {
    return(mappingTab$mapped_read_count[mappingTab$sample %in% samples])
  })
  
  
  
  ### taking only normal 
  mappingRatesByGeo = lapply(newContByGeoListNew, function(samples) {
    return(mappingTab$X.mapped_read_count[mappingTab$sample %in% samples])
  })
  mappingRatesByGeo = mappingRatesByGeo[names(mappingRatesByGeo)!="Finland"]
  medianMappingRatesByGeo = unlist(lapply(mappingRatesByGeo, median))
  medianMappingRatesByGeo = sort(medianMappingRatesByGeo)
  
  par(mar=c(5,7,4,1))
  boxplot(mappingRatesByGeo[match(names(medianMappingRatesByGeo), names(mappingRatesByGeo))],
          horizontal=T, las=1)
  
  
  
  
  mappedReadsByGeo = lapply(newContByGeoListNew, function(samples) {
    return(mappingTab$mapped_read_count[mappingTab$sample %in% samples])
  })
  mappedReadsByGeo = mappedReadsByGeo[names(mappedReadsByGeo)!="Finland"]
  medianMappedReadsByGeo = unlist(lapply(mappedReadsByGeo, median))
  medianMappedReadsByGeo = sort(medianMappedReadsByGeo)
  
  par(mar=c(5,7,4,1))
  boxplot(mappedReadsByGeo[match(names(medianMappedReadsByGeo), names(mappingRatesByGeo))],
          horizontal=T, las=1, log="x")
  
}

### very important part ###
diseaseSignatureCompare = T
if (diseaseSignatureCompare) {
  
  effectSizeMode = T
  if (effectSizeMode) {
    
    loadPair = T
    if (loadPair) {
      diseaseWithCountryPairTab = rbind(c("Atherosclerosis_SW_id1", "Sweden"),
                                        c("CVD_CN_id10", "China"),
                                        c("GDM_CN_id12", "China"),
                                        c("T2D_CN_id2", "China"),
                                        c("T1D_FI_id21", "Finland"),
                                        c("Cirrhosis_UK_id27", "UK"),
                                        c("CFS_US_id34", "US"),
                                        c("NAFLD_IT_id40", "Italy"),
                                        c("NAFLD_ES_id40", "Spain"),
                                        c("CRC_JP_id45", "Japan"),
                                        c("CRC_IT_id46", "Italy"),
                                        c("Behcet_CN_id52", "China"),
                                        c("CD_CN_id6", "China"),
                                        c("Ankylosing_CN_id9", "China"),
                                        c("CRC_US_id11", "US"),
                                        c("T2D_SW_id14", "Sweden"),
                                        c("IGT_SW_id14", "Sweden"),
                                        c("Obesity_DK_id16", "Denmark"),
                                        c("IBD_ES_id17", "Spain"),
                                        c("Cirrhosis_CN_id20", "China"),
                                        c("Obesity_DK_id25", "Denmark"),
                                        c("NSCLC_FR_id26", ""), 
                                        c("Melanoma_US_id3", "US"),
                                        c("RA_CN_id31", "China"),
                                        c("T1D_LU_id35", "Luxembourg"),
                                        c("RCC_FR_id41", ""),
                                        c("CRC_DE_id44", "Germany"),
                                        c("T2D_ES_id53", "Spain"),
                                        c("NAFLD_US_id7", "US"),
                                        c("PD_DE_id8", "Germany"))
      
      diseaseWithCountryPairTab2 = rbind(c("Atherosclerosis_SW_id1", "Sweden"),
                                         c("CVD_CN_id10", "China"),
                                         c("GDM_CN_id12", "China"),
                                         c("T2D_CN_id2", "China"),
                                         c("T1D_FI_id21", "Finland"),
                                         c("Cirrhosis_UK_id27", "UK"),
                                         c("CFS_US_id34", "US"),
                                         c("NAFLD_IT_id40", "Italy"),
                                         c("NAFLD_ES_id40", "Spain"),
                                         c("CRC_JP_id45", "Japan"),
                                         c("CRC_IT_id46", "Italy"),
                                         c("Behcet_CN_id52", "China"),
                                         c("CD_CN_id6", "China"),
                                         c("Ankylosing_CN_id9", "China"),
                                         c("CRC_US_id11", "US"),
                                         c("T2D_SW_id14", "Sweden"),
                                         c("IGT_SW_id14", "Sweden"),
                                         c("Obesity_DK_id16", "Denmark"),
                                         c("IBD_ES_id17", "Spain"),
                                         c("Cirrhosis_CN_id20", "China"),
                                         c("Obesity_DK_id25", "Denmark"),
                                         c("NSCLC_FR_id26", "Germany"), 
                                         c("Melanoma_US_id3", "US"),
                                         c("RA_CN_id31", "China"),
                                         c("T1D_LU_id35", "Luxembourg"),
                                         c("RCC_FR_id41", "Germany"),
                                         c("CRC_DE_id44", "Germany"),
                                         c("T2D_ES_id53", "Spain"),
                                         c("NAFLD_US_id7", "US"),
                                         c("PD_DE_id8", "Germany"))
      
      diseaseWithEtPairTab = rbind(c("Atherosclerosis_SW_id1", "etFirmicutes"),
                                   c("CVD_CN_id10", "etBacteroides"),
                                   c("GDM_CN_id12", "etBacteroides"),
                                   c("T2D_CN_id2", "etBacteroides"),
                                   c("T1D_FI_id21", "etFirmicutes"),
                                   c("Cirrhosis_UK_id27", "etFirmicutes"),
                                   c("CFS_US_id34", "etBacteroides"),
                                   c("NAFLD_IT_id40", "etFirmicutes"),
                                   c("NAFLD_ES_id40", "etFirmicutes"),
                                   c("CRC_JP_id45", "etBacteroides"),
                                   c("CRC_IT_id46", "etFirmicutes"),
                                   c("Behcet_CN_id52", "etBacteroides"),
                                   c("CD_CN_id6", "etBacteroides"),
                                   c("Ankylosing_CN_id9", "etBacteroides"),
                                   c("CRC_US_id11", "etBacteroides"),
                                   c("T2D_SW_id14", "etFirmicutes"),
                                   c("IGT_SW_id14", "etFirmicutes"),
                                   c("Obesity_DK_id16", "etFirmicutes"),
                                   c("IBD_ES_id17", "etFirmicutes"),
                                   c("Cirrhosis_CN_id20", "etBacteroides"),
                                   c("Obesity_DK_id25", "etFirmicutes"),
                                   c("NSCLC_FR_id26", "etFirmicutes"), 
                                   c("Melanoma_US_id3", "etBacteroides"),
                                   c("RA_CN_id31", "etBacteroides"),
                                   c("T1D_LU_id35", "etFirmicutes"),
                                   c("RCC_FR_id41", "etFirmicutes"),
                                   c("CRC_DE_id44", "etFirmicutes"),
                                   c("T2D_ES_id53", "etFirmicutes"),
                                   c("NAFLD_US_id7", "etBacteroides"),
                                   c("PD_DE_id8", "etFirmicutes"))
      
      diseaseWithMatchedPairTab = rbind(c("Atherosclerosis_SW_id1", "Sweden_id1"),
                                        c("CVD_CN_id10", "China_id10"),
                                        c("GDM_CN_id12", "China_id12"),
                                        c("T2D_CN_id2", "China_id2"),
                                        c("T1D_FI_id21", "Finland_id21"),
                                        c("Cirrhosis_UK_id27", ""),
                                        c("CFS_US_id34", "US_id34"),
                                        c("NAFLD_IT_id40", ""),
                                        c("NAFLD_ES_id40", ""),
                                        c("CRC_JP_id45", "Japan_id45"),
                                        c("CRC_IT_id46", "Italy_id46"),
                                        c("Behcet_CN_id52", "China_id52"),
                                        c("CD_CN_id6", ""),
                                        c("Ankylosing_CN_id9", "China_id9"),
                                        c("CRC_US_id11", "US_id11"),
                                        c("T2D_SW_id14", "Sweden_id14"),
                                        c("IGT_SW_id14", "Sweden_id14"),
                                        c("Obesity_DK_id16", "Denmark_id16"),
                                        c("IBD_ES_id17", "Spain_id17"),
                                        c("Cirrhosis_CN_id20", "China_id20"),
                                        c("Obesity_DK_id25", "Denmark_id25"),
                                        c("NSCLC_FR_id26", ""), 
                                        c("Melanoma_US_id3", ""),
                                        c("RA_CN_id31", "China_id31"),
                                        c("T1D_LU_id35", "Luxembourg_id35"),
                                        c("RCC_FR_id41", ""),
                                        c("CRC_DE_id44", "Germany_id44"),
                                        c("T2D_ES_id53", ""),
                                        c("NAFLD_US_id7", ""),
                                        c("PD_DE_id8", "Germany_id8"))
      diseaseWithCountryPairTab = data.frame(diseaseWithCountryPairTab, stringsAsFactors = F)
      diseaseWithMatchedPairTab = data.frame(diseaseWithMatchedPairTab, stringsAsFactors = F)
      diseaseWithEtPairTab = data.frame(diseaseWithEtPairTab, stringsAsFactors = F)
      
      colnames(diseaseWithCountryPairTab) = c("reference","target")
      colnames(diseaseWithMatchedPairTab) = c("reference","target")
      colnames(diseaseWithEtPairTab) = c("reference","target")
      
    }
    
    loadSurvivalSigMode = T
    if (loadSurvivalSigMode) {
      
      #survivalStatsRData = "C://Data//comparative.analysis.healthy.sweden//survivalStats.RData"
      #load(survivalStatsRData) #survivalStats
      
      survivalStatsNewRData = "C://Data//comparative.analysis.healthy.sweden//survivalStats.new.2019.02.25.RData"
      load(survivalStatsNewRData) #survivalStats
      
    }
    
    loadDiseaseSigNew = T
    if (loadDiseaseSigNew) {
      allPairMeltRData = "C://Data//comparative.analysis.healthy.sweden//allPairMelt.disease.signatures.RData"
      load(allPairMeltRData)
      allPairDownMeltRData = "C://Data//comparative.analysis.healthy.sweden//allPairDownMelt.disease.signatures.RData"
      load(allPairDownMeltRData)
      
      cohortOrderFile = "C://Data//comparative.analysis.healthy.sweden//disease.cohort.order.txt"
      cohortOrders = read.table(cohortOrderFile, stringsAsFactors = F, header=F)[,]
      names(cohortOrders)=1:30
      
      allPairMeltImputed = allPairMelt
      allPairMeltImputed = allPairMeltImputed[!allPairMeltImputed$disease_cohort %in% c("NSCLC_FR_id26", "RCC_FR_id41"),] #, "T1D_FI_id21"
      allPairMeltImputed$effect_size[allPairMeltImputed$effect_size < 0] = 0
      allPairMeltImputed$effect_size[allPairMeltImputed$effect_size > 1] = 1
      allPairMeltImputed$effect_size[is.na(allPairMeltImputed$effect_size)] = 0
      
      manhattanMode = T
      if (manhattanMode) {
        uniqMsp = unique(allPairMeltImputed$msp) 
        uniqMspId = 1:length(uniqMsp)
        names(uniqMspId) =uniqMsp
        
        allPairMeltImputed$CHR = names(cohortOrders)[match(allPairMeltImputed$disease_cohort, cohortOrders)]
        allPairMeltImputed$CHR = as.numeric(allPairMeltImputed$CHR)
        #allPairMeltImputed$CHR = as.numeric(gsub(".*_id", "", allPairMeltImputed$disease_cohort))
        allPairMeltImputed$BP = uniqMspId[match(allPairMeltImputed$msp, names(uniqMspId))]
        allPairMeltImputed$P = allPairMeltImputed$effect_size
        allPairMeltImputed$SNP = allPairMeltImputed$msp
        
      }
      allPairMeltImputedCountry = allPairMeltImputed[allPairMeltImputed$control=="country",]
      allPairMeltImputedCountrySig = allPairMeltImputedCountry[allPairMeltImputedCountry$effect_size>=0.3,]
      allPairMeltImputedCountrySigSpecies = unique(allPairMeltImputedCountrySig$msp)
      
      allPairMeltImputedGeoCluster = allPairMeltImputed[allPairMeltImputed$control=="cluster",]
      allPairMeltImputedGeoClusterSig = allPairMeltImputedGeoCluster[allPairMeltImputedGeoCluster$effect_size>=0.3,]
      allPairMeltImputedGeoClusterSigSpecies = unique(allPairMeltImputedGeoClusterSig$msp)
      
      allPairDownMeltImputed = allPairDownMelt
      allPairDownMeltImputed = allPairDownMeltImputed[!allPairDownMeltImputed$disease_cohort %in% c("NSCLC_FR_id26", "RCC_FR_id41"),]#, "T1D_FI_id21"
      allPairDownMeltImputed$effect_size[allPairDownMeltImputed$effect_size < 0] = 0
      allPairDownMeltImputed$effect_size[allPairDownMeltImputed$effect_size > 1] = 1
      allPairDownMeltImputed$effect_size[is.na(allPairDownMeltImputed$effect_size)] = 0
      
      manhattanMode = T
      if (manhattanMode) {
        uniqMsp = unique(allPairDownMeltImputed$msp) 
        uniqMspId = 1:length(uniqMsp)
        names(uniqMspId) =uniqMsp
        allPairDownMeltImputed$CHR = names(cohortOrders)[match(allPairDownMeltImputed$disease_cohort, cohortOrders)]
        allPairDownMeltImputed$CHR = as.numeric(allPairDownMeltImputed$CHR)
        #allPairDownMeltImputed$CHR = as.numeric(gsub(".*_id", "", allPairDownMeltImputed$disease_cohort))
        allPairDownMeltImputed$BP = uniqMspId[match(allPairDownMeltImputed$msp, names(uniqMspId))]
        allPairDownMeltImputed$P = allPairDownMeltImputed$effect_size
        allPairDownMeltImputed$SNP = allPairDownMeltImputed$msp
      }
      allPairDownMeltImputedCountry = allPairDownMeltImputed[allPairDownMeltImputed$control=="country",]
      allPairDownMeltImputedCountrySig = allPairDownMeltImputedCountry[allPairDownMeltImputedCountry$effect_size>=0.3,]
      allPairDownMeltImputedCountrySigSpecies = unique(allPairDownMeltImputedCountrySig$msp)
      
      allPairDownMeltImputedGeoCluster = allPairDownMeltImputed[allPairDownMeltImputed$control=="cluster",]
      allPairDownMeltImputedGeoClusterSig = allPairDownMeltImputedGeoCluster[allPairDownMeltImputedGeoCluster$effect_size>=0.3,]
      allPairDownMeltImputedGeoClusterSigSpecies = unique(allPairDownMeltImputedGeoClusterSig$msp)
      
      depletedStats = c(length(allPairDownMeltImputedCountrySigSpecies),
                        length(allPairDownMeltImputedGeoClusterSigSpecies))
      
      enrichedStats = c(length(allPairMeltImputedCountrySigSpecies),
                        length(allPairMeltImputedGeoClusterSigSpecies))
      
      barplot(depletedStats, col=c("#E9E3F0", "#FEE9D5"), las=1)
      barplot(enrichedStats, col=c("#E9E3F0", "#FEE9D5"), las=1)
      
      
      
      writeMode = F
      if (writeMode) {
        allPairMeltImputedCountrySigFile = "C://Data//comparative.analysis.healthy.sweden//allPairMeltImputedCountrySig.20191230.txt"
        allPairDownMeltImputedCountrySigFile = "C://Data//comparative.analysis.healthy.sweden//allPairDownMeltImputedCountrySig.20191230.txt"
        
        write.table(allPairMeltImputedCountrySig, allPairMeltImputedCountrySigFile, sep="\t")
        write.table(allPairDownMeltImputedCountrySig, allPairDownMeltImputedCountrySigFile, sep="\t")
      }
      
      
      markovStatTab$numEnrichedDepletedCohort = sapply(rownames(markovStatTab), function(currMsp){
        if (any(enrichedCountSpecies %in% currMsp)) {
          currTab = allPairMeltImputedCountrySig[allPairMeltImputedCountrySig$msp == currMsp, ]
          numEnriched = sum(currTab$P >= 0.3)
          return(numEnriched)  
        }
        if (any(depletedCountSpecies %in% currMsp)) {
          currTab = allPairDownMeltImputedCountrySig[allPairDownMeltImputedCountrySig$msp == currMsp, ]
          numDepleted = sum(currTab$P >= 0.3)
          return(numDepleted)
          
        }
        return("")
        
      })
      
      addEnrichedDepltedToInflowOutflowFile = "C://Data//comparative.analysis.healthy.sweden//enriched.depleted.stats.to.inflow.outflow.txt"
      #write.table(markovStatTab, addEnrichedDepltedToInflowOutflowFile, sep="\t")
      
      allPairDownMeltImputedCountryMat = acast(allPairDownMeltImputedCountry, msp ~ disease_cohort, value.var="effect_size")
      allPairMeltImputedCountryMat = acast(allPairMeltImputedCountry, msp ~ disease_cohort, value.var="effect_size")
      
      allPairDownMeltImputedCountryMat = (-1)*allPairDownMeltImputedCountryMat
      allPairSummaryCountryMat = sapply(1:28, function(currInd) {
        currTab = data.frame(enriched = allPairMeltImputedCountryMat[,currInd],
                             depleted = allPairDownMeltImputedCountryMat[,currInd],
                             stringsAsFactors = F)  
        bestValues = apply(currTab, 1, function(currRow){
          if (currRow[1] > abs(currRow[2])) return(currRow[1])
          else return(currRow[2])
        })
        return(bestValues)
      })
      colnames(allPairSummaryCountryMat) = colnames(allPairDownMeltImputedCountryMat)
      allPairSummaryCountryTab = data.frame(msp = rownames(allPairSummaryCountryMat),
                                            species = getSpeciesName(rownames(allPairSummaryCountryMat), taxo),
                                            allPairSummaryCountryMat,
                                            stringsAsFactors = F) 
      
      write.table(allPairSummaryCountryTab, file="C://Data//comparative.analysis.healthy.sweden//allPairSummaryCountryTab.txt", sep="\t" )
      
      
    }
    
    checkMGWASofSingleSpecies = T
    if (checkMGWASofSingleSpecies) {
      
      loadCols = T
      if (loadCols) {
        cancerCol = "#bb0e3d"
        diabetesCol="#044486"
        liverCol="#00a384"
        ibdCol="#5d0e8a"
        metabolicCol="#b81570"
        inflammationCol="#c9716b"
        parkinsonCol = "#4b5f7a"
        
        cohortCols = c(rep(cancerCol, 5), 
                       rep(diabetesCol, 7),
                       rep(liverCol, 5), 
                       rep(ibdCol,2), 
                       rep(metabolicCol, 4), 
                       rep(inflammationCol, 4), 
                       parkinsonCol)
      }
      
      targetMsp = "msp_1114" #allisonella histaminiformans
      targetMsp = "msp_0009" #clostridium bolteae
      targetMsp = "msp_0313" #v. parvula
      targetMsp = "msp_1339" # g. urolithinfaciens
      targetMsp = "msp_1006" # Lactobacillus fermentum
      targetMsp = "msp_1234" # p. micra
      
      targetMsp = "msp_0025" # A. muciniphila
      targetMsp = "msp_0087" # L. pectinoschiza
      targetMsp = "msp_0028" #   
      targetMsp = "msp_0742" #   
      
      upMgwasRes = allPairMeltImputedCountry[allPairMeltImputedCountry$msp==targetMsp,]
      downMgwasRes = allPairDownMeltImputedCountry[allPairDownMeltImputedCountry$msp==targetMsp, ]
      
      require(qqman)
      targetMain = "a. histaminiformans"
      targetMain = "v. parvula"
      targetMain = "c. bolteae"
      targetMain = "g. urolithinfaciens"
      targetMain = "l. fermentum"
      targetMain = "p. micra"
      targetMain = "p. goldsteinii"
      
      targetMain = "A. muciniphila"
      targetMain = "L. pectinoschiza"
      targetMain = "G. urolithinfaciens"
      targetMain = "Azospirillum sp. 51_20"
      targetMain = "msp_1641"
      targetMain = "s. parasanguinis"
      
      
      manhattan(upMgwasRes, genomewideline=0.5, suggestiveline = 0.3, ylim = c(0, 1.1), col = cohortCols, logp=F, main=targetMain)
      manhattan(downMgwasRes, genomewideline=0.5, suggestiveline = 0.3, ylim = c(0, 1.1), col = cohortCols, logp=F, main=targetMain)
      
    }
    
    checkUniqCommonMode = T
    if(checkUniqCommonMode) {
      diseaseUpSpecies = unique(allPairMeltImputedCountrySig$msp)
      diseaseDownSpecies = unique(allPairDownMeltImputedCountrySig$msp)
      diseaseSigList = list(up=diseaseUpSpecies, down=diseaseDownSpecies)
      
      table(diseaseUpSpecies %in% diseaseDownSpecies)
      table(diseaseDownSpecies %in% diseaseUpSpecies)
      
      # diseaseSurvivalSigList = list(Enriched=survivalStats$surv.rates[survivalStats$mgs %in% names(upCounts[upCounts>=6]) ], 
      #                               Depleted=survivalStats$surv.rates[survivalStats$mgs %in% names(downCounts[downCounts>=6])],
      #                               All=survivalStats$surv.rates)
      # 
      # diseaseSurvivalSigList = list(up=survivalStats$surv.rates[survivalStats$mgs %in% diseaseUpSpecies ], 
      #                               down=survivalStats$surv.rates[survivalStats$mgs %in% diseaseDownSpecies ],
      #                               all=survivalStats$surv.rates)
      #
      
      if (T) {
        swedSpecies = specificGroupMgsByCountries$Sweden
        spainSpecies = specificGroupMgsByCountries$Spain
        denSpecies = specificGroupMgsByCountries$Denmark
        tradSpecies = unique(unlist(specificGroupMgsByCountries[traditionalCountries]))
        indSpecies = unique(unlist(specificGroupMgsByCountries[industrializedCountries]))
      }
      
      #### general vs unique species ####
      upCounts = getCountsThroughTable(allPairMeltImputedCountrySig$msp) 
      downCounts = getCountsThroughTable(allPairDownMeltImputedCountrySig$msp)
      
      alls = unique(c(names(downCounts), names(upCounts)))
      downCountsAll = downCounts[match(alls, names(downCounts))]
      upCountsAll = upCounts[match(alls, names(upCounts))]
      upCountsAll[is.na(upCountsAll)] = 0
      downCountsAll[is.na(downCountsAll)] = 0
      
      countsTab = data.frame(up=upCountsAll, 
                             down=downCountsAll, 
                             msp=alls, 
                             species = getSpeciesName(alls, taxo),
                             stringsAsFactors = F)
      
      countsTab$sum = countsTab$up + countsTab$down
      #countsTab$diff =  countsTab$down - countsTab$up
      countsTab$diff =  countsTab$up - countsTab$down
      
      industrialMgsMat = mergeMatUpdated[,industrializedSamples]
      industrialMgsMeans = rowMeans(industrialMgsMat)
      
      statsMgsForNormSamples = "J:\\Deposit\\Project\\2018_microbiome_atlas\\atlas.volcanot.out\\richness.volcano.stats.mgs.for.normal.txt"
      statsMgsForNorm = read.table(statsMgsForNormSamples, sep="\t", header=T, stringsAsFactors = F)
      
      #countsTab$abd = industrialMgsMeans[match(countsTab$msp, names(industrialMgsMeans))]*10^11 + 1
      countsTab$richness_lfc = statsMgsForNorm$lfc[match(countsTab$msp,rownames(statsMgsForNorm))]
      countsTab$label = as.character(countsTab$species)
      countsTab$gss = NA
      countsTab$oss = NA
      countsTab$gutOnly = NA
      countsTab$oralOnly = NA
      countsTab$gss[countsTab$msp %in% gssSpecies] = T
      countsTab$oss[countsTab$msp %in% ossSpecies] = T
      countsTab$gutOnly[countsTab$msp %in% gutOnlySpecies] = T
      countsTab$oralOnly[countsTab$msp %in% oralOnlySpecies] = T
      countsTab$nonWestern = NA
      countsTab$industrial = NA
      countsTab$nonWestern[countsTab$msp %in% tradSpecies] = T
      countsTab$industrial[countsTab$msp %in% indSpecies] = T
      
      countsTab$inflow = inflow[match( countsTab$msp, names(inflow))]
      countsTab$outflow = outflow[match( countsTab$msp, names(outflow))]
      
      
      coutrySpecificSpeciesFile="J:\\Deposit\\Project\\2018_microbiome_atlas\\atlas.geography\\coutrySpecificSpecies.tab.txt"
      coutrySpecificSpeciesTab = read.table(coutrySpecificSpeciesFile, sep="\t", header=T,stringsAsFactors = F)
      
      countsTab$enrichedCountry = "-"
      countsTab$enrichedCountry = unlist(sapply(countsTab$msp, function(x) {
        enriched = coutrySpecificSpeciesTab$enrichedCountries[coutrySpecificSpeciesTab$msp==x]
        if (length(enriched)==0) return(NA)
        enriched = as.character(enriched)
        return(enriched)
      }))
      
      countsTab$NumEnrichedCountry = 0
      countsTab$NumEnrichedCountry = unlist(sapply(countsTab$msp, function(x) {
        enriched = coutrySpecificSpeciesTab$enrichedCountries[coutrySpecificSpeciesTab$msp==x]
        if (length(enriched)==0) return(0)
        if (enriched=="") return(0)
        enrichedCountries = strsplit(enriched, split="\\;")[[1]]
        enriched = length(enrichedCountries)
        return(enriched)
      }))
      
      
      #write.table(countsTab, "C:\\Data\\comparative.analysis.healthy.sweden\\frequency.disease.species.table.s9.txt", sep="\t")
      diffCut=2
      sumCut=3
      #dim(countsTab[countsTab$sum>=sumCut & (countsTab$diff >= diffCut | countsTab$diff <= -1 * diffCut),] )
      #View(countsTab[countsTab$sum>=sumCut & (countsTab$diff >= diffCut | countsTab$diff <= -1 * diffCut),] )
      sigCountsTab = (countsTab[countsTab$sum>=sumCut & (countsTab$diff >= diffCut | countsTab$diff <= -1 * diffCut),] )
      #write.table(sigCountsTab, "C:\\Data\\comparative.analysis.healthy.sweden\\frequency.disease.species.significant.table.s9.txt", sep="\t")
      
      #countsTab$abd[countsTab$abd == -Inf] = 0
      #plot(upCountsAll, downCountsAll)
      #plot(countsTab$diff, countsTab$sum)
      
      enrichedCountSpecies= countsTab$msp[countsTab$sum>=sumCut & countsTab$diff >= diffCut]
      depletedCountSpecies= countsTab$msp[countsTab$sum>=sumCut & countsTab$diff <= -1 * diffCut]
      bothSigSpecies = unique(c(enrichedCountSpecies, depletedCountSpecies))
      View(taxo[depletedCountSpecies,])
      View(taxo[enrichedCountSpecies,])
      
      
      markovStatTab$commonEnrichedDepleted = ""
      markovStatTab$commonEnrichedDepleted[rownames(markovStatTab)%in%enrichedCountSpecies] = "Common enriched"
      markovStatTab$commonEnrichedDepleted[rownames(markovStatTab)%in%depletedCountSpecies] = "Common depleted"
      
      
      if (F) wellnessMgsMat_5m_presence = apply(wellnessMgsMat_5m, 1, any)
      
      checkPathwayFraction = T
      if (checkPathwayFraction) {
        funcMatRData = "j://Deposit/Project/2018_microbiome_atlas/functional.annotation/funcMat.20190808.RData"
        load(funcMatRData)
        
        vfMatRData = "j://Deposit/Project/2018_microbiome_atlas/functional.annotation/PATRIC/vfBestMat.RData"
        load(vfMatRData) #vfMatDigit
        vfTerms = colnames(vfMatDigit)
        
        mucinCazymes = c("cazy.GT27", "cazy.GT14", "cazy.GT11", "cazy.GT10", "cazy.GH89", "cazy.GH84", "cazy.GH33", "cazy.GH20", "cazy.GH29", "cazy.GH18", "cazy.GH123", "cazy.GH109", "cazy.CBM51", "cazy.CBM50", "cazy.CBM32")
        storageCazymes = c("cazy.CBM34", "cazy.CBM48", "cazy.CBM20", "cazy.GH13", "cazy.GH133", "cazy.GH32", "cazy.GH36", "cazy.GH37", "cazy.GH77", "cazy.GH57", "cazy.GT101", "cazy.GT26", "cazy.GT3", "cazy.GT35", "cazy.GT5")
        
        btexFile = "C:\\Data\\comparative.analysis.healthy.sweden\\inflow.outflow.associated.functions.BTEXterms.txt"
        btexTab = read.delim(btexFile, sep="\t", stringsAsFactors = F, header=F)
        btexList = split(btexTab[,1], btexTab[,2])
        currFuncList = list(MGE=mgePfamTerms,
                            VF=vfTerms , 
                            Storage=storageCazymes,
                            Nutrient=carbLipidTransportKOs,
                            vitamin=vitaminKOs,
                            aa=aminoAcidKOs,
                            pts=PTS,
                            efflux=drugEffluxKOs)
        currFuncList = append(currFuncList, btexList) 
        currFuncMelt = melt(currFuncList)
        currFuncMelt[,1] = as.character(currFuncMelt[,1])
        currFuncMelt[,2] = as.character(currFuncMelt[,2])
        
        
        chisqTab = do.call(rbind, sapply(currFuncMelt[,1], function(currFunc){
          enrichedYes = sum(funcMat[enrichedCountSpecies,colnames(funcMat) == currFunc])
          enrichedNo = length(enrichedCountSpecies) - enrichedYes
          depletedYes = sum(funcMat[depletedCountSpecies,colnames(funcMat) == currFunc])
          depletedNo   = length(depletedCountSpecies) - depletedYes
          contigency = cbind(c(enrichedYes, depletedYes),
                             c(enrichedNo, depletedNo))
          #print(contigency)
          chisqOut = chisq.test(contigency)
          oddRatio = (enrichedYes*depletedNo)/(enrichedNo*depletedYes)
          return(c(chisqOut$p.value, oddRatio))
        }, simplify = F))
        currFuncMelt = data.frame(currFuncMelt, chisqTab)
        colnames(currFuncMelt) = c("term", "class", "pvalue", "oddRatio")
        currFuncMelt$minusLog10P = -log10(currFuncMelt$pvalue)
        currFuncMelt$lfc = log2(currFuncMelt$oddRatio)
        currFuncMelt$lfc[currFuncMelt$lfc == Inf] = 5
        currFuncMelt$lfc[currFuncMelt$lfc == -Inf] = -5
        
        plot(currFuncMelt$lfc, currFuncMelt$minusLog10P, 
             las=1, xlab="LOD", ylab="-log10 p-value")
        abline(h=-log10(0.05), col="red")
        abline(v=c(-3,3),col="black")
        
        
        chisqAllTab = sapply(colnames(funcMat),  function(currFunc){
          enrichedYes = sum(funcMat[enrichedCountSpecies,colnames(funcMat) == currFunc])
          enrichedNo = length(enrichedCountSpecies) - enrichedYes
          depletedYes = sum(funcMat[depletedCountSpecies,colnames(funcMat) == currFunc])
          depletedNo   = length(depletedCountSpecies) - depletedYes
          contigency = cbind(c(enrichedYes, depletedYes),
                             c(enrichedNo, depletedNo))
          #print(contigency)
          chisqOut = chisq.test(contigency)
          oddRatio = (enrichedYes*depletedNo)/(enrichedNo*depletedYes)
          return(c(chisqOut$p.value, oddRatio))
        })
        chisqAllTab=t(chisqAllTab)
        colnames(chisqAllTab) = c("pvalue", "oddratio")
        chisqAllTab = chisqAllTab[!is.nan(chisqAllTab[,"pvalue"]),]
        
        
        View(chisqAllTab[chisqAllTab[,"pvalue"]<0.05, ])
        
        table(chisqAllTab[,"pvalue"]<0.05 & chisqAllTab[,"oddratio"] > 1)
        table(chisqAllTab[,"pvalue"]<0.05 & chisqAllTab[,"oddratio"] < 1)
        
        enrichedTab = chisqAllTab[chisqAllTab[,"pvalue"]<0.05 & chisqAllTab[,"oddratio"] > 1,]
        depletedTab = chisqAllTab[chisqAllTab[,"pvalue"]<0.05 & chisqAllTab[,"oddratio"] < 1,]
        
        View(enrichedTab)
        View(depletedTab)
        
        depletedTab = data.frame(depletedTab, term=rownames(depletedTab))
        enrichedTab = data.frame(enrichedTab, term=rownames(enrichedTab))
        
        lOut = lapply(currFuncList, function(funcs){
          
          chisqPvalues = sapply(funcs, function(currFunc){
            enrichedYes = sum(funcMat[enrichedCountSpecies,colnames(funcMat) == currFunc])
            enrichedNo = length(enrichedCountSpecies) - enrichedYes
            depletedYes = sum(funcMat[depletedCountSpecies,colnames(funcMat) == currFunc])
            depletedNo   = length(depletedCountSpecies) - depletedYes
            contigency = cbind(c(enrichedYes, depletedYes),
                               c(enrichedNo, depletedNo))
            #print(contigency)
            chisqOut = chisq.test(contigency)
            oddRatio = (enrichedYes*depletedNo)/(enrichedNo*depletedYes)
            return(chisqOut$p.value)
            
          })
          return(sort(chisqPvalues[!is.nan(chisqPvalues)]))
          
          # out=((rowSums(funcMat[enrichedCountSpecies,colnames(funcMat) %in% funcs])/length(funcs)+0.1)/(
          #   rowSums(funcMat[depletedCountSpecies,colnames(funcMat) %in% funcs])/length(funcs)+0.1)
          #   )
          # return(log10(out))
        })
        lOut
        boxplot(lOut)
        
        
      }
      
      
      
      depletedCountSpeciesTab = data.frame(msp=depletedCountSpecies, 
                                           type="depleted")
      enrichedCountSpeciesTab = data.frame(msp=enrichedCountSpecies, 
                                           type="enriched")
      bothCountSpeciesTab = rbind(enrichedCountSpeciesTab, depletedCountSpeciesTab)
      bothCountSpeciesTab$species = getSpeciesName(bothCountSpeciesTab$msp, taxo)
      bothCountSpeciesFile = "C://Data//comparative.analysis.healthy.sweden//enrichedDepletedAcrossDiseases.txt"
      #write.table(bothCountSpeciesTab, bothCountSpeciesFile, sep="\t")
      
      checkVolcanoPlot = T
      if (checkVolcanoPlot) {
        require(ggrepel)
        set.seed(1)
        #pdf(file = "enriched.depleted.volcano.pdf", width=3.5, height=2.5)
        ggplot(countsTab, aes(x=diff, y=sum)) + #, label=label
          geom_jitter(shape = 21, colour = "black", width = 0.6, height = 0.6, size=2, aes(fill = diff)) + 
          #scale_fill_gradient2(low = "blue", mid="white",  high = "red") +
          scale_fill_gradient2(low = "red", mid="white",  high = "blue") +
          
          geom_hline(yintercept = 3, colour="gray") +
          geom_vline(xintercept = 2, colour="gray") +
          geom_vline(xintercept = -2, colour="gray") +
          geom_abline(slope=1)+
          geom_abline(slope=-1) +
          xlab("|Enrichment| - |Depletion|") + ylab("|Enrichment| + |Depletion|")+
          theme(panel.grid.major = element_blank(), 
                panel.grid.minor = element_blank(),
                legend.position = "none",
                panel.background = element_rect(fill = "white", 
                                                colour = "black", 
                                                size = 0.5, 
                                                linetype = "solid"))
      }
      
      
      checkSpeciesRetainingProbablity = T
      if (checkSpeciesRetainingProbablity) {
        diseaseSurvivalSigList = list(Enriched=survivalStats$surv.rates[survivalStats$mgs %in% enrichedCountSpecies ], 
                                      Depleted=survivalStats$surv.rates[survivalStats$mgs %in% depletedCountSpecies],
                                      All=survivalStats$surv.rates[rownames(survivalStats) %in% names(inflow)])
        
        
        wilcox.test(diseaseSurvivalSigList$All, diseaseSurvivalSigList$Enriched, alternative = "greater")
        wilcox.test(diseaseSurvivalSigList$All, diseaseSurvivalSigList$Depleted, alternative = "less")
        wilcox.test(diseaseSurvivalSigList$Depleted, diseaseSurvivalSigList$Enriched, alternative = "greater")
        
        t.test(diseaseSurvivalSigList$All, diseaseSurvivalSigList$Depleted)
        t.test(diseaseSurvivalSigList$Depleted, diseaseSurvivalSigList$Enriched)
        
        
        diseaseSurvivalSigMelt = melt(diseaseSurvivalSigList)
        #pdf(file = "enriched.depleted.gut.retaining.probability.pdf", width=2.2, height=3)
        diseaseSurvivalSigMelt = diseaseSurvivalSigMelt[diseaseSurvivalSigMelt$L1 != "All",]
        diseaseSurvivalSigMelt$L1 = factor(diseaseSurvivalSigMelt$L1, levels = c("Enriched","Depleted"))
        
        ggplot(diseaseSurvivalSigMelt, aes(x=L1, y=value, fill=L1)) +  
          geom_boxplot() + 
          scale_fill_manual(values=c("#0000ff55", "#FF000055")) +
          xlab("") + ylab("") + 
          theme(panel.grid.major = element_blank(), 
                panel.grid.minor = element_blank(),
                legend.position = "none",
                axis.ticks.y = element_line(size = 1),
                axis.text.x = element_text(face="bold", color="#993333", 
                                           size=14, angle=45),
                axis.text.y = element_text(face="bold", color="#993333", 
                                           size=14, angle=45),
                panel.background = element_rect(fill = "white", 
                                                colour = "black", 
                                                size = 0.5, 
                                                linetype = "solid"))
        
        
        # diseaseSurvivalSigList = list(Enriched=survivalStats$surv.rates[survivalStats$mgs %in% enrichedCountSpecies ], 
        #                               Depleted=survivalStats$surv.rates[survivalStats$mgs %in% depletedCountSpecies],
        #                               All=survivalStats$surv.rates)
        # 
        # diseaseSurvivalSigMelt = melt(diseaseSurvivalSigList)
        # #pdf(file = "enriched.depleted.gut.retaining.probability.pdf", width=2.2, height=3)
        # ggplot(diseaseSurvivalSigMelt, aes(x=L1, y=value, fill=L1)) +  
        #   scale_fill_manual(values=c("#80808077","#FF000077", "#0000ff77")) +
        #   geom_boxplot() +
        #   xlab("") +
        #   ylab("") + 
        #   theme(panel.grid.major = element_blank(), 
        #         panel.grid.minor = element_blank(),
        #         legend.position = "none",
        #         axis.text.x = element_text(angle = 60, hjust=1),
        #         panel.background = element_rect(fill = "white", 
        #                                         colour = "black", 
        #                                         size = 0.5, 
        #                                         linetype = "solid"))
        # 
      }
      
      checkInOutFlowMode = T
      if (checkInOutFlowMode) {
        diseaseInflowSigList = list(enriched=inflow[names(inflow) %in% enrichedCountSpecies ], 
                                    depleted=inflow[names(inflow) %in% depletedCountSpecies],
                                    all=inflow)
        
        diseaseOutflowSigList = list(enriched=outflow[names(outflow) %in% enrichedCountSpecies ], 
                                     depleted=outflow[names(outflow) %in% depletedCountSpecies],
                                     all=outflow)
        
        
        diseaseInflowSigList_5m = list(enriched=inflow_5m[names(inflow_5m) %in% enrichedCountSpecies ], 
                                       depleted=inflow_5m[names(inflow_5m) %in% depletedCountSpecies],
                                       all=inflow_5m)
        
        diseaseOutflowSigList_5m = list(enriched=outflow_5m[names(outflow_5m) %in% enrichedCountSpecies ], 
                                        depleted=outflow_5m[names(outflow_5m) %in% depletedCountSpecies],
                                        all=outflow_5m)
        
        diseaseInflowSigList_15m = list(enriched=inflow_15m[names(inflow_15m) %in% enrichedCountSpecies ], 
                                        depleted=inflow_15m[names(inflow_15m) %in% depletedCountSpecies],
                                        all=inflow_15m)
        
        diseaseOutflowSigList_15m = list(enriched=outflow_15m[names(outflow_15m) %in% enrichedCountSpecies ], 
                                         depleted=outflow_15m[names(outflow_15m) %in% depletedCountSpecies],
                                         all=outflow_15m)
        
        
        
        ###
        wilcox.test(diseaseInflowSigList$all, diseaseInflowSigList$depleted, alternative = "less")
        wilcox.test(diseaseInflowSigList$all, diseaseInflowSigList$enriched, alternative = "less")
        wilcox.test(diseaseInflowSigList$depleted, diseaseInflowSigList$enriched, alternative = "greater")
        
        wilcox.test(diseaseOutflowSigList$all, diseaseOutflowSigList$depleted, alternative = "greater")
        wilcox.test(diseaseOutflowSigList$all, diseaseOutflowSigList$enriched, alternative = "less")
        wilcox.test(diseaseOutflowSigList$depleted, diseaseOutflowSigList$enriched, alternative = "less")
        
        wilcox.test(diseaseInflowSigList_5m$all, diseaseInflowSigList_5m$depleted, alternative = "less")
        wilcox.test(diseaseInflowSigList_5m$all, diseaseInflowSigList_5m$enriched, alternative = "less")
        wilcox.test(diseaseInflowSigList_5m$depleted, diseaseInflowSigList_5m$enriched, alternative = "greater")
        
        wilcox.test(diseaseOutflowSigList_5m$all, diseaseOutflowSigList_5m$depleted, alternative = "greater")
        wilcox.test(diseaseOutflowSigList_5m$all, diseaseOutflowSigList_5m$enriched, alternative = "less")
        wilcox.test(diseaseOutflowSigList_5m$depleted, diseaseOutflowSigList_5m$enriched, alternative = "less")
        
        wilcox.test(diseaseInflowSigList_15m$all, diseaseInflowSigList_15m$depleted, alternative = "less")
        wilcox.test(diseaseInflowSigList_15m$all, diseaseInflowSigList_15m$enriched, alternative = "less")
        wilcox.test(diseaseInflowSigList_15m$depleted, diseaseInflowSigList_15m$enriched, alternative = "greater")
        
        wilcox.test(diseaseOutflowSigList_15m$all, diseaseOutflowSigList_15m$depleted, alternative = "greater")
        wilcox.test(diseaseOutflowSigList_15m$all, diseaseOutflowSigList_15m$enriched, alternative = "less")
        wilcox.test(diseaseOutflowSigList_15m$depleted, diseaseOutflowSigList_15m$enriched, alternative = "less")
        
        #####
        t.test(diseaseInflowSigList$all, diseaseInflowSigList$depleted, alternative = "less")
        t.test(diseaseInflowSigList$all, diseaseInflowSigList$enriched, alternative = "less")
        t.test(diseaseInflowSigList$depleted, diseaseInflowSigList$enriched, alternative = "greater")
        
        t.test(diseaseOutflowSigList$all, diseaseOutflowSigList$depleted, alternative = "greater")
        t.test(diseaseOutflowSigList$all, diseaseOutflowSigList$enriched, alternative = "less")
        t.test(diseaseOutflowSigList$depleted, diseaseOutflowSigList$enriched, alternative = "less")
        
        t.test(diseaseInflowSigList_5m$all, diseaseInflowSigList_5m$depleted, alternative = "less")
        t.test(diseaseInflowSigList_5m$all, diseaseInflowSigList_5m$enriched, alternative = "less")
        t.test(diseaseInflowSigList_5m$depleted, diseaseInflowSigList_5m$enriched, alternative = "greater")
        
        t.test(diseaseOutflowSigList_5m$all, diseaseOutflowSigList_5m$depleted, alternative = "greater")
        t.test(diseaseOutflowSigList_5m$all, diseaseOutflowSigList_5m$enriched, alternative = "less")
        t.test(diseaseOutflowSigList_5m$depleted, diseaseOutflowSigList_5m$enriched, alternative = "less")
        
        t.test(diseaseInflowSigList_15m$all, diseaseInflowSigList_15m$depleted, alternative = "less")
        t.test(diseaseInflowSigList_15m$all, diseaseInflowSigList_15m$enriched, alternative = "less")
        t.test(diseaseInflowSigList_15m$depleted, diseaseInflowSigList_15m$enriched, alternative = "greater")
        
        t.test(diseaseOutflowSigList_15m$all, diseaseOutflowSigList_15m$depleted, alternative = "greater")
        t.test(diseaseOutflowSigList_15m$all, diseaseOutflowSigList_15m$enriched, alternative = "less")
        t.test(diseaseOutflowSigList_15m$depleted, diseaseOutflowSigList_15m$enriched, alternative = "less")
        
        
        #####
        t.test(diseaseInflowSigList$all, diseaseInflowSigList$depleted)
        t.test(diseaseInflowSigList$depleted, diseaseInflowSigList$enriched)
        
        wilcox.test(diseaseInflowSigList_5m$all, diseaseInflowSigList_5m$depleted)
        wilcox.test(diseaseInflowSigList_5m$all, diseaseInflowSigList_5m$enriched)
        wilcox.test(diseaseInflowSigList_5m$depleted, diseaseInflowSigList_5m$enriched)
        
        t.test(diseaseInflowSigList_5m$all, diseaseInflowSigList_5m$depleted)
        t.test(diseaseInflowSigList_5m$depleted, diseaseInflowSigList_5m$enriched)
        
        
        diseaseInflowSigMelt = melt(diseaseInflowSigList)
        diseaseOutflowSigMelt = melt(diseaseOutflowSigList)
        
        diseaseInflowSigMelt = melt(diseaseInflowSigList_5m)
        diseaseOutflowSigMelt = melt(diseaseOutflowSigList_5m)
        
        diseaseInflowSigMelt = melt(diseaseInflowSigList_15m)
        diseaseOutflowSigMelt = melt(diseaseOutflowSigList_15m)
        
        #pdf(file = "enriched.depleted.gut.retaining.probability.pdf", width=2.2, height=3)
        ggplot(diseaseInflowSigMelt, aes(x=L1, y=value, fill=L1)) +  
          scale_fill_manual(values=c("#80808077","#FF000077", "#0000ff77")) +
          geom_boxplot() +
          xlab("") +
          ylab("inflow") + 
          theme(panel.grid.major = element_blank(), 
                panel.grid.minor = element_blank(),
                legend.position = "none",
                axis.text.x = element_text(angle = 60, hjust=1),
                panel.background = element_rect(fill = "white", 
                                                colour = "black", 
                                                size = 0.5, 
                                                linetype = "solid"))
        
        ggplot(diseaseOutflowSigMelt, aes(x=L1, y=value, fill=L1)) +  
          scale_fill_manual(values=c("#80808077","#FF000077", "#0000ff77")) +
          geom_boxplot() +
          xlab("") +
          ylab("outflow") + 
          theme(panel.grid.major = element_blank(), 
                panel.grid.minor = element_blank(),
                legend.position = "none",
                axis.text.x = element_text(angle = 60, hjust=1),
                panel.background = element_rect(fill = "white", 
                                                colour = "black", 
                                                size = 0.5, 
                                                linetype = "solid"))
      }
      
      checkOssFraction = T
      if (checkOssFraction) {
        
        lenEnriched = length(enrichedCountSpecies)
        lenDepleted = length(depletedCountSpecies)
        lenAll = dim(gutTaxo)[1]
        
        lenOssDepleted = sum( countsTab[countsTab$msp %in% depletedCountSpecies,"oss"], na.rm=T) 
        lenOssEnriched = sum( countsTab[countsTab$msp %in% enrichedCountSpecies,"oss"], na.rm=T) 
        lenOssAll = sum( gutTaxo$type_sh=="oss")
        
        
        ossFractionList = list(Enriched=lenOssEnriched*100/lenEnriched,
                               Depleted=lenOssDepleted*100/lenDepleted,
                               All=lenOssAll*100/lenAll)
        ossFractionMelt = melt(ossFractionList)
        ggplot(ossFractionMelt, aes(x=L1, y=value, fill=L1)) +  
          scale_fill_manual(values=c("#80808077","#FF000077", "#0000ff77")) +
          geom_bar(stat = "identity", colour="black") +
          xlab("") +
          ylab("Oral orient species (%)") + 
          ylim(c(0,50))+
          theme(panel.grid.major = element_blank(), 
                panel.grid.minor = element_blank(),
                legend.position = "none",
                axis.text.x = element_text(angle = 60, hjust=1),
                panel.background = element_rect(fill = "white", 
                                                colour = "black", 
                                                size = 0.5, 
                                                linetype = "solid"))
        
        
      }
      
      checkIndFraction = F
      if (checkOssFraction) {
        
        lenEnriched = length(enrichedCountSpecies)
        lenDepleted = length(depletedCountSpecies)
        lenAll = dim(gutTaxo)[1]
        
        lenIndDepleted = sum( countsTab[countsTab$msp %in% depletedCountSpecies,"industrial"], na.rm=T) 
        lenIndEnriched = sum( countsTab[countsTab$msp %in% enrichedCountSpecies,"industrial"], na.rm=T) 
        lenIndAll = length(specificGroupMgsAll)
        
        
        ossFractionList = list(Enriched=lenOssEnriched*100/lenEnriched,
                               Depleted=lenOssDepleted*100/lenDepleted,
                               All=lenOssAll*100/lenAll)
        ossFractionMelt = melt(ossFractionList)
        ggplot(ossFractionMelt, aes(x=L1, y=value, fill=L1)) +  
          scale_fill_manual(values=c("#80808077","#FF000077", "#0000ff77")) +
          geom_bar(stat = "identity", colour="black") +
          xlab("") +
          ylab("Oral orient species (%)") + 
          ylim(c(0,50))+
          theme(panel.grid.major = element_blank(), 
                panel.grid.minor = element_blank(),
                legend.position = "none",
                axis.text.x = element_text(angle = 60, hjust=1),
                panel.background = element_rect(fill = "white", 
                                                colour = "black", 
                                                size = 0.5, 
                                                linetype = "solid"))
        
        
      }
      
    }
    
    getDiseaseNetSignature = T
    if (getDiseaseNetSignature) {
      load(allPairMeltRData)
      allPairDownMeltRData = "C://Data//comparative.analysis.healthy.sweden//allPairDownMelt.disease.signatures.RData"
      load(allPairDownMeltRData)
      
      cohortOrderFile = "C://Data//comparative.analysis.healthy.sweden//disease.cohort.order.txt"
      cohortOrders = read.table(cohortOrderFile, stringsAsFactors = F, header=F)[,]
      names(cohortOrders)=1:30
      
      allPairMeltImputed = allPairMelt
      allPairMeltImputed = allPairMeltImputed[!allPairMeltImputed$disease_cohort %in% c("NSCLC_FR_id26", "RCC_FR_id41"),] #, "T1D_FI_id21"
      allPairMeltImputed$effect_size[allPairMeltImputed$effect_size < 0] = 0
      allPairMeltImputed$effect_size[allPairMeltImputed$effect_size > 1] = 1
      allPairMeltImputed$effect_size[is.na(allPairMeltImputed$effect_size)] = 0
      
      manhattanMode = T
      if (manhattanMode) {
        uniqMsp = unique(allPairMeltImputed$msp) 
        uniqMspId = 1:length(uniqMsp)
        names(uniqMspId) =uniqMsp
        
        allPairMeltImputed$CHR = names(cohortOrders)[match(allPairMeltImputed$disease_cohort, cohortOrders)]
        allPairMeltImputed$CHR = as.numeric(allPairMeltImputed$CHR)
        #allPairMeltImputed$CHR = as.numeric(gsub(".*_id", "", allPairMeltImputed$disease_cohort))
        allPairMeltImputed$BP = uniqMspId[match(allPairMeltImputed$msp, names(uniqMspId))]
        allPairMeltImputed$P = allPairMeltImputed$effect_size
        allPairMeltImputed$SNP = allPairMeltImputed$msp
        
      }
      allPairMeltImputedCountry = allPairMeltImputed[allPairMeltImputed$control=="country",]
      allPairMeltImputedCountrySig = allPairMeltImputedCountry[allPairMeltImputedCountry$effect_size>=0.3,]
      
      allPairDownMeltImputed = allPairDownMelt
      allPairDownMeltImputed = allPairDownMeltImputed[!allPairDownMeltImputed$disease_cohort %in% c("NSCLC_FR_id26", "RCC_FR_id41"),]#, "T1D_FI_id21"
      allPairDownMeltImputed$effect_size[allPairDownMeltImputed$effect_size < 0] = 0
      allPairDownMeltImputed$effect_size[allPairDownMeltImputed$effect_size > 1] = 1
      allPairDownMeltImputed$effect_size[is.na(allPairDownMeltImputed$effect_size)] = 0
      
      manhattanMode = T
      if (manhattanMode) {
        uniqMsp = unique(allPairDownMeltImputed$msp) 
        uniqMspId = 1:length(uniqMsp)
        names(uniqMspId) =uniqMsp
        allPairDownMeltImputed$CHR = names(cohortOrders)[match(allPairDownMeltImputed$disease_cohort, cohortOrders)]
        allPairDownMeltImputed$CHR = as.numeric(allPairDownMeltImputed$CHR)
        #allPairDownMeltImputed$CHR = as.numeric(gsub(".*_id", "", allPairDownMeltImputed$disease_cohort))
        allPairDownMeltImputed$BP = uniqMspId[match(allPairDownMeltImputed$msp, names(uniqMspId))]
        allPairDownMeltImputed$P = allPairDownMeltImputed$effect_size
        allPairDownMeltImputed$SNP = allPairDownMeltImputed$msp
      }
      allPairDownMeltImputedCountry = allPairDownMeltImputed[allPairDownMeltImputed$control=="country",]
      allPairDownMeltImputedCountrySig = allPairDownMeltImputedCountry[allPairDownMeltImputedCountry$effect_size>=0.3,]
      
      cohortDiseaseMappingFile = "C://Data/comparative.analysis.healthy.sweden//cohort.disease.mapping.txt"
      cohortDiseaseMapping = read.table(cohortDiseaseMappingFile, sep="\t", stringsAsFactors = F, header=T, row.names = NULL)
      colnames(cohortDiseaseMapping) = c("Cohort", "Disease")
      
      allPairDownMeltImputedCountry$disease = cohortDiseaseMapping$Disease[match(allPairDownMeltImputedCountry$disease_cohort, cohortDiseaseMapping$Cohort)]
      allPairMeltImputedCountry$disease = cohortDiseaseMapping$Disease[match(allPairMeltImputedCountry$disease_cohort, cohortDiseaseMapping$Cohort)]
      
      selectedColumns = c("msp", "disease", "effect_size")
      allPairDownMeltImputedCountrySimple = allPairDownMeltImputedCountry[,selectedColumns]
      allPairMeltImputedCountrySimple = allPairMeltImputedCountry[,selectedColumns]
      
      allSummaizedDiseaseUp = aggregate(. ~ msp + disease, allPairMeltImputedCountrySimple, mean)
      allSummaizedDiseaseDown = aggregate(. ~ msp + disease, allPairDownMeltImputedCountrySimple, mean)
      
      allSummaizedDiseaseUpSel = allSummaizedDiseaseUp[allSummaizedDiseaseUp$effect_size>=0.3,]
      allSummaizedDiseaseDownSel = allSummaizedDiseaseDown[allSummaizedDiseaseDown$effect_size>=0.3,]
      allSummaizedDiseaseUpSel$species = getSpeciesName(allSummaizedDiseaseUpSel$msp, taxo)
      allSummaizedDiseaseDownSel$species = getSpeciesName(allSummaizedDiseaseDownSel$msp, taxo)
      
      allSummaizedDiseaseUpSel$GssOss = "gut"
      allSummaizedDiseaseDownSel$GssOss = "gut"
      allSummaizedDiseaseUpSel$GssOss[allSummaizedDiseaseUpSel$msp %in% ossSpecies] = "oss"
      allSummaizedDiseaseDownSel$GssOss[allSummaizedDiseaseDownSel$msp %in% ossSpecies] = "oss"
      
      loadHgcLgcSpecies = T
      if (loadHgcLgcSpecies) {
        adjpCut = 1e-3
        statsMgsForNormSamples = "J:\\Deposit\\Project\\2018_microbiome_atlas\\atlas.volcanot.out\\richness.volcano.stats.mgs.for.normal.txt"
        statsMgsForNorm = read.table(statsMgsForNormSamples, sep="\t", header=T, stringsAsFactors = F)
        statsMgsForNormSel = statsMgsForNorm[statsMgsForNorm$qvalue <= adjpCut,]
        statsMgsForNormSelHgc = statsMgsForNormSel[statsMgsForNormSel$lfc > 2,]
        statsMgsForNormSelLgc = statsMgsForNormSel[statsMgsForNormSel$lfc < -2,]
        
        hgcSpecies = rownames(statsMgsForNormSelHgc)
        lgcSpecies = rownames(statsMgsForNormSelLgc)
      }
      
      allSummaizedDiseaseUpSel$richness = "none"
      allSummaizedDiseaseDownSel$richness = "none"
      allSummaizedDiseaseUpSel$richness[allSummaizedDiseaseUpSel$msp %in% hgcSpecies] = "hgc"
      allSummaizedDiseaseUpSel$richness[allSummaizedDiseaseUpSel$msp %in% lgcSpecies] = "lgc"
      allSummaizedDiseaseDownSel$richness[allSummaizedDiseaseDownSel$msp %in% hgcSpecies] = "hgc"
      allSummaizedDiseaseDownSel$richness[allSummaizedDiseaseDownSel$msp %in% lgcSpecies] = "lgc"
      
      getDiseaseNetwork <- function(diseaseTab) {
        uniqDisease = unique(diseaseTab$disease)
        overlapMat = do.call(rbind, sapply(uniqDisease, function(diseaseI){
          mspI = diseaseTab$msp[diseaseTab$disease==diseaseI]
          overlapVecs = unlist(sapply(uniqDisease, function(diseaseJ){
            
            mspJ = diseaseTab$msp[diseaseTab$disease==diseaseJ]
            return(sum(mspI %in% mspJ))
            
          }, simplify = F))
          return(overlapVecs)
        }, simplify = F)) 
        overlapMelt = melt(overlapMat)
        overlapMelt = overlapMelt[overlapMelt$Var1 != overlapMelt$Var2,]
        overlapMelt = overlapMelt[overlapMelt$value!=0,]
        colnames(overlapMelt) = c("source","target","value")
        overlapMelt$source = as.character(overlapMelt$source)
        overlapMelt$target = as.character(overlapMelt$target)
        return(overlapMelt)
      }
      
      enrichedSpeciesNet = getDiseaseNetwork(allSummaizedDiseaseUpSel)
      depletedSpeciesNet = getDiseaseNetwork(allSummaizedDiseaseDownSel)
      allSpeciesNet = getDiseaseNetwork(rbind(allSummaizedDiseaseDownSel, 
                                              allSummaizedDiseaseUpSel))
      
      write.table(enrichedSpeciesNet, "enrichedSpeciesNet.txt", sep="\t", row.names = F, quote=F)
      write.table(depletedSpeciesNet, "depletedSpeciesNet.txt", sep="\t", row.names = F, quote=F)
      write.table(allSpeciesNet, "allSpeciesNet.txt", sep="\t", row.names = F, quote=F)
      
      circosOutMode = F
      if (circosOutMode) {
        set.seed(1)
        chordDiagram(enrichedSpeciesNet)
        set.seed(1)
        chordDiagram(enrichedSpeciesNet, annotationTrack = "grid", preAllocateTracks = 1) #, grid.col = circosGridCols
        circos.trackPlotRegion(track.index = 1,  panel.fun = function(x, y) {
          xlim = get.cell.meta.data("xlim")
          ylim = get.cell.meta.data("ylim")
          sector.name = get.cell.meta.data("sector.index")
          circos.text(mean(xlim), ylim[1] + .1, cex = 0.8,  sector.name, facing = "clockwise", niceFacing = TRUE, adj = c(0, 0.5))
          circos.axis(h = "top", labels = F, major.tick = F, 
                      sector.index = sector.name, track.index = 2)
        }, bg.border = NA)
      }
      
    }
    
    loadDiseaseSigMode = F
    if (loadDiseaseSigMode) {
      allPairMeltRData = "C://Data//comparative.analysis.healthy.sweden//allPairMelt.disease.signatures.RData"
      load(allPairMeltRData)
      allPairDownMeltRData = "C://Data//comparative.analysis.healthy.sweden//allPairDownMelt.disease.signatures.RData"
      load(allPairDownMeltRData)
      
      
      cohortOrderFile = "C://Data//comparative.analysis.healthy.sweden//disease.cohort.order.txt"
      cohortOrders = read.table(cohortOrderFile, stringsAsFactors = F, header=F)[,]
      names(cohortOrders)=1:30
      
      allPairMeltImputed = allPairMelt
      allPairMeltImputed = allPairMeltImputed[!allPairMeltImputed$disease_cohort %in% c("NSCLC_FR_id26", "RCC_FR_id41", "T1D_FI_id21"),]
      allPairMeltImputed$effect_size[allPairMeltImputed$effect_size < 0] = 0
      allPairMeltImputed$effect_size[allPairMeltImputed$effect_size > 1] = 1
      allPairMeltImputed$effect_size[is.na(allPairMeltImputed$effect_size)] = 0
      
      manhattanMode = T
      if (manhattanMode) {
        uniqMsp = unique(allPairMeltImputed$msp) 
        uniqMspId = 1:length(uniqMsp)
        names(uniqMspId) =uniqMsp
        
        allPairMeltImputed$CHR = names(cohortOrders)[match(allPairMeltImputed$disease_cohort, cohortOrders)]
        allPairMeltImputed$CHR = as.numeric(allPairMeltImputed$CHR)
        #allPairMeltImputed$CHR = as.numeric(gsub(".*_id", "", allPairMeltImputed$disease_cohort))
        allPairMeltImputed$BP = uniqMspId[match(allPairMeltImputed$msp, names(uniqMspId))]
        allPairMeltImputed$P = allPairMeltImputed$effect_size
        allPairMeltImputed$SNP = allPairMeltImputed$msp
        
      }
      
      allPairMeltImputedMatched = allPairMeltImputed[allPairMeltImputed$control=="matched",]
      allPairMeltImputedCountry = allPairMeltImputed[allPairMeltImputed$control=="country",]
      allPairMeltImputedCluster = allPairMeltImputed[allPairMeltImputed$control=="cluster",]
      
      allPairMeltImputedMatchedSig = allPairMeltImputedMatched[allPairMeltImputedMatched$effect_size>=0.3,]
      allPairMeltImputedCountrySig = allPairMeltImputedCountry[allPairMeltImputedCountry$effect_size>=0.3,]
      allPairMeltImputedClusterSig = allPairMeltImputedCluster[allPairMeltImputedCluster$effect_size>=0.3,]
      
      
      mergeByDiseaseMode = T
      if (mergeByDiseaseMode) {
        
      }
      
      checkFromMatched = T
      if (checkFromMatched) {
        upMspMatchedByCohortList = split(allPairMeltImputedMatchedSig$msp, allPairMeltImputedMatchedSig$disease_cohort)
        upMspMatchedCohorts = names(upMspMatchedByCohortList)
        
        upMspEffectSizeCountry = sapply(upMspMatchedCohorts, function(currId) {
          currMsps = upMspMatchedByCohortList[[currId]]
          checkInd = allPairMeltImputedCountry$msp %in% currMsps & allPairMeltImputedCountry$disease_cohort == currId
          currMatchedTab = allPairMeltImputedCountry[checkInd,]
          return(currMatchedTab$effect_size)
        }, simplify = F )
        
        upMspEffectSizeMatched = sapply(upMspMatchedCohorts, function(currId) {
          currMsps = upMspMatchedByCohortList[[currId]]
          checkInd = allPairMeltImputedMatched$msp %in% currMsps & allPairMeltImputedMatched$disease_cohort == currId
          currMatchedTab = allPairMeltImputedMatched[checkInd,]
          return(currMatchedTab$effect_size)
        }, simplify = F )
        
        upMspEffectSizeCluster = sapply(upMspMatchedCohorts, function(currId) {
          currMsps = upMspMatchedByCohortList[[currId]]
          checkInd = allPairMeltImputedCluster$msp %in% currMsps & allPairMeltImputedCluster$disease_cohort == currId
          currMatchedTab = allPairMeltImputedCluster[checkInd,]
          return(currMatchedTab$effect_size)
        }, simplify = F )
        
        
        data1 = melt(upMspEffectSizeCountry)
        data2 = melt(upMspEffectSizeMatched)
        data3 = melt(upMspEffectSizeCluster)
        data1$type="country"
        data2$type="matched"
        data3$type="cluster"
        dataAll = rbind(data1, data2, data3)
        exMatched = c("Melanoma_US_id3",
                      "NAFLD_US_id7",
                      "Cirrhosis_UK_id27", 
                      "NAFLD_ES_id40", 
                      "NAFLD_IT_id40", 
                      "T2D_ES_id53", 
                      "T1D_FI_id21")
        dataAll = dataAll[!dataAll$L1 %in% exMatched, ]
        
        #dataAll$value[dataAll$L1 %in% exMatched & dataAll$type=="matched"] = NA
        
        ggplot(dataAll, aes(x=type, y=value)) + 
          geom_boxplot(aes(fill=L1)) +   
          facet_wrap(~L1) + 
          ylab("Effect size") +
          xlab("") + 
          theme(legend.position = "none", axis.text.x = element_text(angle = 60)) + 
          geom_hline(yintercept = 0.3, col="blue")
        
      }
      
      checkFromCountry = T
      if (checkFromCountry) {
        upMspCountryByCohortList = split(allPairMeltImputedCountrySig$msp, allPairMeltImputedCountrySig$disease_cohort)
        upMspCountryCohorts = names(upMspCountryByCohortList)
        
        upMspEffectSizeCountry = sapply(upMspCountryCohorts, function(currId) {
          currMsps = upMspCountryByCohortList[[currId]]
          checkInd = allPairMeltImputedCountry$msp %in% currMsps & allPairMeltImputedCountry$disease_cohort == currId
          currMatchedTab = allPairMeltImputedCountry[checkInd,]
          return(currMatchedTab$effect_size)
        }, simplify = F )
        
        upMspEffectSizeMatched = sapply(upMspCountryCohorts, function(currId) {
          currMsps = upMspCountryByCohortList[[currId]]
          checkInd = allPairMeltImputedMatched$msp %in% currMsps & allPairMeltImputedMatched$disease_cohort == currId
          currMatchedTab = allPairMeltImputedMatched[checkInd,]
          return(currMatchedTab$effect_size)
        }, simplify = F )
        
        upMspEffectSizeCluster = sapply(upMspCountryCohorts, function(currId) {
          currMsps = upMspCountryByCohortList[[currId]]
          checkInd = allPairMeltImputedCluster$msp %in% currMsps & allPairMeltImputedCluster$disease_cohort == currId
          currMatchedTab = allPairMeltImputedCluster[checkInd,]
          return(currMatchedTab$effect_size)
        }, simplify = F )
        
        data1 = melt(upMspEffectSizeCountry)
        data2 = melt(upMspEffectSizeMatched)
        data3 = melt(upMspEffectSizeCluster)
        data1$type="country"
        data2$type="matched"
        data3$type="cluster"
        dataAll = rbind(data1, data2, data3)
        exMatched = c("Melanoma_US_id3",
                      "NAFLD_US_id7",
                      "Cirrhosis_UK_id27", 
                      "NAFLD_ES_id40", 
                      "NAFLD_IT_id40", 
                      "T2D_ES_id53", 
                      "T1D_FI_id21")
        dataAll = dataAll[!dataAll$L1 %in% exMatched, ]
        
        #dataAll$value[dataAll$L1 %in% exMatched & dataAll$type=="matched"] = NA
        
        ggplot(dataAll, aes(x=type, y=value)) + 
          geom_boxplot(aes(fill=L1)) +   
          facet_wrap(~L1) + 
          ylab("Effect size") +
          xlab("") + 
          theme(legend.position = "none", axis.text.x = element_text(angle = 60)) + 
          geom_hline(yintercept = 0.3, col="blue")
        
      }
      
      checkFromCluster = T
      if (checkFromCluster) {
        upMspClusterByCohortList = split(allPairMeltImputedClusterSig$msp, allPairMeltImputedClusterSig$disease_cohort)
        upMspClusterCohorts = names(upMspClusterByCohortList)
        
        upMspEffectSizeCountry = sapply(upMspClusterCohorts, function(currId) {
          currMsps = upMspClusterByCohortList[[currId]]
          checkInd = allPairMeltImputedCountry$msp %in% currMsps & allPairMeltImputedCountry$disease_cohort == currId
          currMatchedTab = allPairMeltImputedCountry[checkInd,]
          return(currMatchedTab$effect_size)
        }, simplify = F )
        
        upMspEffectSizeMatched = sapply(upMspClusterCohorts, function(currId) {
          currMsps = upMspClusterByCohortList[[currId]]
          checkInd = allPairMeltImputedMatched$msp %in% currMsps & allPairMeltImputedMatched$disease_cohort == currId
          currMatchedTab = allPairMeltImputedMatched[checkInd,]
          return(currMatchedTab$effect_size)
        }, simplify = F )
        
        upMspEffectSizeCluster = sapply(upMspClusterCohorts, function(currId) {
          currMsps = upMspClusterByCohortList[[currId]]
          checkInd = allPairMeltImputedCluster$msp %in% currMsps & allPairMeltImputedCluster$disease_cohort == currId
          currMatchedTab = allPairMeltImputedCluster[checkInd,]
          return(currMatchedTab$effect_size)
        }, simplify = F )
        
        data1 = melt(upMspEffectSizeCountry)
        data2 = melt(upMspEffectSizeMatched)
        data3 = melt(upMspEffectSizeCluster)
        data1$type="country"
        data2$type="matched"
        data3$type="cluster"
        dataAll = rbind(data1, data2, data3)
        exMatched = c("Melanoma_US_id3",
                      "NAFLD_US_id7",
                      "Cirrhosis_UK_id27", 
                      "NAFLD_ES_id40", 
                      "NAFLD_IT_id40", 
                      "T2D_ES_id53", 
                      "T1D_FI_id21")
        dataAll = dataAll[!dataAll$L1 %in% exMatched, ]
        
        #dataAll$value[dataAll$L1 %in% exMatched & dataAll$type=="matched"] = NA
        
        ggplot(dataAll, aes(x=type, y=value)) + 
          geom_boxplot(aes(fill=L1)) +   
          facet_wrap(~L1) + 
          ylab("Effect size") +
          xlab("") + 
          theme(legend.position = "none", axis.text.x = element_text(angle = 60)) + 
          geom_hline(yintercept = 0.3, col="blue")
        
      }
      
      checkNumberOfEnrichedDepleted = T
      if (checkNumberOfEnrichedDepleted) {
        esCut = 0.3
        allPairMeltSel = allPairMelt[!is.na(allPairMelt$effect_size),]
        allPairMeltSel = allPairMeltSel[allPairMeltSel$effect_size>=esCut,]
        
        allPairMeltSelMatched = allPairMeltSel[allPairMeltSel$control=="matched",]
        allPairMeltSelCountry = allPairMeltSel[allPairMeltSel$control=="country",]
        allPairMeltSelCluster = allPairMeltSel[allPairMeltSel$control=="cluster",]
        
        allPairDownMeltSel = allPairDownMelt[!is.na(allPairDownMelt$effect_size),]
        allPairDownMeltSel = allPairDownMeltSel[allPairDownMeltSel$effect_size>=esCut,]
        
        allPairDownMeltSelMatched = allPairDownMeltSel[allPairDownMeltSel$control=="matched",]
        allPairDownMeltSelCountry = allPairDownMeltSel[allPairDownMeltSel$control=="country",]
        allPairDownMeltSelCluster = allPairDownMeltSel[allPairDownMeltSel$control=="cluster",]
        
        
        allCohorts = unique(c(allPairMeltSelCountry$disease_cohort, 
                              allPairDownMeltSelCountry$disease_cohort))
        
        numList = sapply(allCohorts, function(currCohort){
          currUpStats = allPairMeltSelCountry[allPairMeltSelCountry$disease_cohort == currCohort,]
          currDownStats = allPairDownMeltSelCountry[allPairDownMeltSelCountry$disease_cohort == currCohort,]
          
          currUpNum = dim(currUpStats)[1]
          currDownNum = dim(currDownStats)[1]
          
          return(c(up=currUpNum, down=currDownNum))
          
        }, simplify = F)
        
        numMat = do.call(rbind, numList)
        moreUps = apply(numMat, 1, function(currRow) currRow[1] > currRow[2])
        moreDowns = !moreUps
        View(numMat[moreUps,])
        View(numMat[moreDowns,])
        
        numTab = as.data.frame(numMat)
        numTab$cohort = rownames(numMat)
        numTab$ratio = (numTab$up+0.1)  / (numTab$down + 0.1)
        numTab = numTab[order(numTab$ratio, decreasing = T),]
        numTab$cohort = factor(numTab$cohort, levels = numTab$cohort)
        numTab$total = numTab$down + numTab$up
        cohort_new = gsub("_id[0-9]+", "", numTab$cohort)
        numTab$region = sapply(cohort_new, function(x) strsplit(x, split = "_")[[1]][2])
        numTab$name = NA
        numTab$name[numTab$cohort == "Melanoma_US_id3"] = "Melanoma"
        numTab$name[numTab$cohort == "CVD_CN_id10"] = "CVD"
        numTab$name[numTab$cohort == "Cirrhosis_CN_id20"] = "Cirrhosis" 
        numTab$name[numTab$cohort == "CRC_IT_id46"] = "CRC" 
        numTab$name[numTab$cohort == "IBD_ES_id17"] = "IBD"
        numTab$name[numTab$cohort == "NAFLD_ES_id40"] = "NAFLD"
        numTab$name[numTab$cohort == "NAFLD_IT_id40"] = "NAFLD"
        numTab$name[numTab$cohort == "NAFLD_US_id7"] = "NAFLD"
        numTab$name[numTab$cohort == "T1D_LU_id35"] = "T1D"
        numTab$name[numTab$cohort == "T2D_ES_id53"] = "T2D"
        numTab$name[numTab$cohort == "T1D_FI_id21"] = "T1D"
        numTab$name[numTab$cohort == "Cirrhosis_UK_id27"] = "Cirrhosis"
        numTab$name[numTab$cohort == "CRC_DE_id44"] = "CRC" 
        numTab$name[numTab$cohort == "CRC_JP_id45"] = "CRC" 
        numTab$name[numTab$cohort == "CRC_US_id11"] = "CRC" 
        numTab$name[numTab$cohort == "IGT_SW_id14"] = "IGT" 
        numTab$name[numTab$cohort == "T2D_SW_id14"] = "T2D" 
        numTab$name[numTab$cohort == "PD_DE_id8"] = "PD" 
        numTab$name[numTab$cohort == "Atherosclerosis_SW_id1"] = "Athero." 
        numTab$name[numTab$cohort == "CFS_US_id34"] = "CFS" 
        numTab$name[numTab$cohort == "RA_CN_id31"] = "RA" 
        numTab$name[numTab$cohort == "Behcet_CN_id52"] = "Behcet" 
        numTab$name[numTab$cohort == "GDM_CN_id12"] = "GDM" 
        
        # plot(numTab$down, numTab$up)
        # abline(a=0,b=1)
        # 
        
        ggplot(numTab, aes(x=down, y=up, size=total, label=name)) + geom_point(aes(colour=region)) +
          geom_abline(slope = 1, intercept = 0) + 
          geom_text_repel(size=3, fontface="italic")+
          xlab("number of depleted species") + 
          ylab("number of enriched species") + 
          scale_colour_manual(values=col_vector[1:10]) +
          theme(panel.grid.major = element_blank(), 
                panel.grid.minor = element_blank(),
                panel.background = element_rect(fill = "white", 
                                                colour = "black", 
                                                size = 0.5, 
                                                linetype = "solid"))
        
        
        
      }
      
      justCheck=F
      if (justCheck) {
        table(allPairMeltImputedMatched$msp == allPairMeltImputedCountry$msp)
        table(allPairMeltImputedMatched$disease_cohort == allPairMeltImputedCountry$disease_cohort)
        table(allPairMeltImputedMatched$effect_size == allPairMeltImputedCountry$effect_size)
        
        table(allPairMeltImputedMatched$msp == allPairMeltImputedCluster$msp)
        table(allPairMeltImputedMatched$disease_cohort == allPairMeltImputedCluster$disease_cohort)
        table(allPairMeltImputedMatched$effect_size == allPairMeltImputedCluster$effect_size)
      }
      
      allPairDownMeltImputed = allPairDownMelt
      allPairDownMeltImputed = allPairDownMeltImputed[!allPairDownMeltImputed$disease_cohort %in% c("NSCLC_FR_id26", "RCC_FR_id41", "T1D_FI_id21"),]
      allPairDownMeltImputed$effect_size[allPairDownMeltImputed$effect_size < 0] = 0
      allPairDownMeltImputed$effect_size[allPairDownMeltImputed$effect_size > 1] = 1
      allPairDownMeltImputed$effect_size[is.na(allPairDownMeltImputed$effect_size)] = 0
      
      manhattanMode = T
      if (manhattanMode) {
        uniqMsp = unique(allPairDownMeltImputed$msp) 
        uniqMspId = 1:length(uniqMsp)
        names(uniqMspId) =uniqMsp
        allPairDownMeltImputed$CHR = names(cohortOrders)[match(allPairDownMeltImputed$disease_cohort, cohortOrders)]
        allPairDownMeltImputed$CHR = as.numeric(allPairDownMeltImputed$CHR)
        #allPairDownMeltImputed$CHR = as.numeric(gsub(".*_id", "", allPairDownMeltImputed$disease_cohort))
        allPairDownMeltImputed$BP = uniqMspId[match(allPairDownMeltImputed$msp, names(uniqMspId))]
        allPairDownMeltImputed$P = allPairDownMeltImputed$effect_size
        allPairDownMeltImputed$SNP = allPairDownMeltImputed$msp
      }
      
      allPairDownMeltImputedMatched = allPairDownMeltImputed[allPairDownMeltImputed$control=="matched",]
      allPairDownMeltImputedCountry = allPairDownMeltImputed[allPairDownMeltImputed$control=="country",]
      allPairDownMeltImputedCluster = allPairDownMeltImputed[allPairDownMeltImputed$control=="cluster",]
      
      loadCols = T
      if (loadCols) {
        cancerCol = "#bb0e3d"
        diabetesCol="#044486"
        liverCol="#00a384"
        ibdCol="#5d0e8a"
        metabolicCol="#b81570"
        inflammationCol="#c9716b"
        parkinsonCol = "#4b5f7a"
        
        cohortCols = c(rep(cancerCol, 5), 
                       rep(diabetesCol, 7),
                       rep(liverCol, 5), 
                       rep(ibdCol,2), 
                       rep(metabolicCol, 4), 
                       rep(inflammationCol, 4), 
                       parkinsonCol)
      }
      
      require(qqman)
      #manhattan(allPairMeltImputedCountry, genomewideline=0.5, suggestiveline = 0.3, ylim = c(0, 1.1), col = c("blue4", "orange3"), logp=F)
      #manhattan(allPairDownMeltImputedCountry, genomewideline=0.5, suggestiveline = 0.3, ylim = c(0, 1.1), col = c("blue4", "orange3"), logp=F)
      manhattan(allPairMeltImputedCountry, genomewideline=0.5, suggestiveline = 0.3, ylim = c(0, 1.1), col = cohortCols, logp=F)
      manhattan(allPairDownMeltImputedCountry, genomewideline=0.5, suggestiveline = 0.3, ylim = c(0, 1.1), col = cohortCols, logp=F)
      
      
      plotScatterBy <- function(pairMelt, cohort, ref, target) {
        allPairDownMeltImputedMatched
        
      }
      
      
      
      View(allPairMeltImputedCountry[allPairMeltImputedCountry$effect_size>=0.5,])
      View(allPairDownMeltImputedCountry[allPairDownMeltImputedCountry$effect_size>=0.5,])
      
      
      allPairMeltMatched = allPairMelt[allPairMelt$control=="matched",]
      allPairMeltCountry = allPairMelt[allPairMelt$control=="country",]
      allPairMeltCluster = allPairMelt[allPairMelt$control=="cluster",]
      
      
      
      
      esCut = 0.3
      allPairMeltSel = allPairMelt[allPairMelt$effect_size>=esCut,]
      allPairMeltSel = allPairMeltSel[!allPairMeltSel$disease_cohort %in% c("NSCLC_FR_id26", "RCC_FR_id41"),]
      allPairMeltSel = allPairMeltSel[!is.na(allPairMeltSel$effect_size),]
      allPairMeltSel = allPairMeltSel[allPairMeltSel$disease_cohort != "T1D_FI_id21",]
      
      #atherosclerosis SWEDEN --> largest
      #Behcet_CN_id52 --> smallest
      
      
      allPairMeltSelMatched = allPairMeltSel[allPairMeltSel$control=="matched",]
      allPairMeltSelCountry = allPairMeltSel[allPairMeltSel$control=="country",]
      allPairMeltSelCluster = allPairMeltSel[allPairMeltSel$control=="cluster",]
      
      allPairMeltSelMatchedMsps = unique(allPairMeltSelMatched$msp)
      allPairMeltSelCountryMsps = unique(allPairMeltSelCountry$msp)
      allPairMeltSelClusterMsps = unique(allPairMeltSelCluster$msp)
      
      diseaseUpSpecies = unique(allPairMeltSel$msp)
      
      allPairDownMeltSel = allPairDownMelt[allPairDownMelt$effect_size>=esCut,]
      allPairDownMeltSel = allPairDownMeltSel[!is.na(allPairDownMeltSel$effect_size),]
      
      downSpeciesByDiseaseList = split(allPairDownMeltSel$msp, allPairDownMeltSel$disease_cohort)
      downSpeciesByDiseaseList = lapply(downSpeciesByDiseaseList, unique)
      downSpeciesCountsByDiseases = unlist(lapply(downSpeciesByDiseaseList, length))
      
      sort(downSpeciesCountsByDiseases)
      
      #atherosclerosis SWEDEN --> largest
      #Behcet_CN_id52 --> smallest
      
      
      diseaseDownSpecies = unique(allPairDownMeltSel$msp)
      
      allPairDownMeltSelMatched = allPairDownMeltSel[allPairDownMeltSel$control=="matched",]
      allPairDownMeltSelCountry = allPairDownMeltSel[allPairDownMeltSel$control=="country",]
      allPairDownMeltSelCluster = allPairDownMeltSel[allPairDownMeltSel$control=="cluster",]
      
      allPairDownMeltSelMatchedMsps = unique(allPairDownMeltSelMatched$msp)
      allPairDownMeltSelCountryMsps = unique(allPairDownMeltSelCountry$msp)
      allPairDownMeltSelClusterMsps = unique(allPairDownMeltSelCluster$msp)
      
      diseaseSigList = list(up=diseaseUpSpecies, down=diseaseDownSpecies)
      
      names(downCounts[downCounts>=6])
      downCounts[downCounts>=6]
      
      diseaseSurvivalSigList = list(Enriched=survivalStats$surv.rates[survivalStats$mgs %in% names(upCounts[upCounts>=6]) ], 
                                    Depleted=survivalStats$surv.rates[survivalStats$mgs %in% names(downCounts[downCounts>=6])],
                                    All=survivalStats$surv.rates)
      
      diseaseSurvivalSigList = list(up=survivalStats$surv.rates[survivalStats$mgs %in% diseaseUpSpecies ], 
                                    down=survivalStats$surv.rates[survivalStats$mgs %in% diseaseDownSpecies ],
                                    all=survivalStats$surv.rates)
      boxplot(diseaseSurvivalSigList, las=2)
      
      t.test(diseaseSurvivalSigList$up, diseaseSurvivalSigList$down)
      t.test(diseaseSurvivalSigList$down, diseaseSurvivalSigList$all)
      t.test(diseaseSurvivalSigList$up, diseaseSurvivalSigList$all)
      
      
      checkOverlap(allPairMeltSelMatchedMsps, allPairMeltSelCountryMsps)
      checkOverlap(allPairMeltSelCountryMsps, allPairMeltSelClusterMsps)
      checkOverlap(allPairMeltSelMatchedMsps, allPairMeltSelClusterMsps)
      
      
      checkOverlap(allPairDownMeltSelMatchedMsps, allPairDownMeltSelCountryMsps)
      checkOverlap(allPairDownMeltSelCountryMsps, allPairDownMeltSelClusterMsps)
      checkOverlap(allPairDownMeltSelMatchedMsps, allPairDownMeltSelClusterMsps)      
      
      #### general vs unique species ####
      upCounts = (table(allPairMeltSel$msp))
      upCounts = as.numeric(upCounts)
      names(upCounts) = names(table(allPairMeltSel$msp))
      
      downCounts = (table(allPairDownMeltSel$msp))
      downCounts = as.numeric(downCounts)
      names(downCounts) = names(table(allPairDownMeltSel$msp))
      
      
      alls = unique(c(names(downCounts), names(upCounts)))
      downCountsAll = downCounts[match(alls, names(downCounts))]
      upCountsAll = upCounts[match(alls, names(upCounts))]
      upCountsAll[is.na(upCountsAll)] = 0
      downCountsAll[is.na(downCountsAll)] = 0
      
      countsTab = (data.frame(up=upCountsAll, down=downCountsAll, msp=alls, species = getSpeciesName(alls, taxo)))
      countsTab$sum = countsTab$up + countsTab$down
      countsTab$diff =  countsTab$down - countsTab$up
      
      industrialMgsMat = mergeMatUpdated[,industrializedSamples]
      industrialMgsMeans = rowMeans(industrialMgsMat)
      
      statsMgsForNormSamples = "J:\\Deposit\\Project\\2018_microbiome_atlas\\atlas.volcanot.out\\richness.volcano.stats.mgs.for.normal.txt"
      statsMgsForNorm = read.table(statsMgsForNormSamples, sep="\t", header=T, stringsAsFactors = F)
      
      countsTab$abd = industrialMgsMeans[match(countsTab$msp, names(industrialMgsMeans))]*10^11 + 1
      countsTab$richness_lfc = statsMgsForNorm$lfc[match(countsTab$msp,rownames(statsMgsForNorm))]
      countsTab$label = as.character(countsTab$species)
      
      showSpecies = c("Parabacteroides goldsteinii",
                      "Odoribacter splanchnicus",
                      "Lachnospira pectinoschiza",
                      "Bacteroides xylanisolvens",
                      "Anaerostipes hadrus 1",
                      "Faecalibacterium prausnitzii 9",
                      "Streptococcus parasanguinis",
                      "[Clostridium] bolteae",
                      "Parvimonas micra",
                      "Peptostreptococcus stomatis",
                      "Ruthenibacterium lactatiformans",
                      "Streptococcus anginosus",
                      "Streptococcus vestibularis",
                      "[Ruminococcus] gnavus",
                      "Lactobacillus salivarius",
                      "Lactobacillus fermentum")
      
      countsTab$label[!countsTab$label %in% showSpecies] = ""
      countsTab$label[countsTab$label=="Parabacteroides goldsteinii"] = "P. goldsteinii"
      countsTab$label[countsTab$label=="Odoribacter splanchnicus"] = "O. splanchnicus"
      countsTab$label[countsTab$label=="Lachnospira pectinoschiza"] = "L. pectinoschiza"
      countsTab$label[countsTab$label=="Bacteroides xylanisolvens"] = "B. xylanisolvens"
      countsTab$label[countsTab$label=="Anaerostipes hadrus 1"] = "A. hadrus"
      countsTab$label[countsTab$label=="Faecalibacterium prausnitzii 9"] = "F. prausnitzii"
      countsTab$label[countsTab$label=="Streptococcus parasanguinis"] = "S. parasanguinis"
      countsTab$label[countsTab$label=="[Clostridium] bolteae"] = "C. bolteae"
      countsTab$label[countsTab$label=="Parvimonas micra"] = "P. micra"
      countsTab$label[countsTab$label=="Peptostreptococcus stomatis"] = "P. stomatis"
      countsTab$label[countsTab$label=="Ruthenibacterium lactatiformans"] = "R. lactatiformans"
      countsTab$label[countsTab$label=="Streptococcus anginosus"] = "S. anginosus"
      countsTab$label[countsTab$label=="Streptococcus vestibularis"] = "S. vestibularis"
      countsTab$label[countsTab$label=="[Ruminococcus] gnavus"] = "R. gnavus"
      countsTab$label[countsTab$label=="Lactobacillus salivarius"] = "L. salivarius"
      countsTab$label[countsTab$label=="Lactobacillus fermentum"] = "L. fermentum"
      View(countsTab[countsTab$sum>=5,])
      View(countsTab[countsTab$sum>=5 & countsTab$diff <3& countsTab$diff > -3,])
      
      #countsTab$abd[countsTab$abd == -Inf] = 0
      plot(upCountsAll, downCountsAll)
      plot(countsTab$diff, countsTab$sum)
      
      depletedCountSpecies = countsTab$msp[countsTab$sum>=5&countsTab$diff > 2]
      enrichedCountSpecies = countsTab$msp[countsTab$sum>=5&countsTab$diff < -2]
      depletedCountSpeciesTab = data.frame(msp=depletedCountSpecies, 
                                           type="depleted")
      enrichedCountSpeciesTab = data.frame(msp=enrichedCountSpecies, 
                                           type="enriched")
      bothCountSpeciesTab = rbind(enrichedCountSpeciesTab, depletedCountSpeciesTab)
      bothCountSpeciesTab$species = getSpeciesName(bothCountSpeciesTab$msp, taxo)
      # write.table(bothCountSpeciesTab, "enrichedDepletedAcrossDiseases.txt", sep="\t")
      
      checkVfSecond = T
      if (checkVfSecond) {
        vfMatRData = "j://Deposit/Project/2018_microbiome_atlas/functional.annotation/PATRIC/vfBestMat.RData"
        load(vfMatRData) #vfMatDigit
        
        depletedVfMat = vfMatDigit[depletedCountSpecies,]
        enrichedVfMat = vfMatDigit[enrichedCountSpecies,]
        
        depletedVfMelt = melt(depletedVfMat)
        enrichedVfMelt = melt(enrichedVfMat)
        depletedVfMelt = depletedVfMelt[depletedVfMelt$value!=0,]
        enrichedVfMelt = enrichedVfMelt[enrichedVfMelt$value!=0,]
        
        
        depletedVfs = rowSums(depletedVfMat)
        enrichedVfs = rowSums(enrichedVfMat)
        
        boxplot(list(depleted=depletedVfs, 
                     enriched=enrichedVfs),log="y")
        
        antismashMatRData = "J://Deposit/Project/2018_microbiome_atlas/functional.annotation/msp1992.igc2.antismashMat.RData"
        load(antismashMatRData)
        
        depletedSecMat = antismashMat[rownames(antismashMat)%in%depletedCountSpecies,]
        enrichedSecMat = antismashMat[rownames(antismashMat)%in%enrichedCountSpecies,]
        depletedSecs = rowSums(depletedSecMat)
        enrichedSecs = rowSums(enrichedSecMat)
        
        boxplot(list(depleted=depletedSecs, 
                     enriched=enrichedSecs),log="y")
        
        
      }
      
      
      diseaseSurvivalSigList = list(Enriched=survivalStats$surv.rates[survivalStats$mgs %in% enrichedCountSpecies ], 
                                    Depleted=survivalStats$surv.rates[survivalStats$mgs %in% depletedCountSpecies],
                                    All=survivalStats$surv.rates)
      
      diseaseSurvivalSigMelt = melt(diseaseSurvivalSigList)
      #pdf(file = "enriched.depleted.gut.retaining.probability.pdf", width=2.2, height=3)
      ggplot(diseaseSurvivalSigMelt, aes(x=L1, y=value, fill=L1)) +  
        scale_fill_manual(values=c("#80808077","#FF000077", "#0000ff77")) +
        geom_boxplot() +
        xlab("") +
        ylab("") + 
        theme(panel.grid.major = element_blank(), 
              panel.grid.minor = element_blank(),
              legend.position = "none",
              axis.text.x = element_text(angle = 60, hjust=1),
              panel.background = element_rect(fill = "white", 
                                              colour = "black", 
                                              size = 0.5, 
                                              linetype = "solid"))
      #dev.off()
      
      
      boxplot(diseaseSurvivalSigList)
      
      
      
      require(ggrepel)
      set.seed(1)
      #pdf(file = "enriched.depleted.volcano.pdf", width=3.5, height=2.5)
      ggplot(countsTab, aes(x=diff, y=sum, label=label)) + 
        geom_jitter(shape = 21, colour = "black", width = 0.6, height = 0.6, size=2, aes(fill = diff)) + 
        scale_fill_gradient2(low = "blue", mid="white",  high = "red") +
        geom_hline(yintercept = 5, colour="gray") +
        geom_vline(xintercept = 3, colour="gray") +
        geom_vline(xintercept = -3, colour="gray") +
        geom_abline(slope=1)+
        geom_abline(slope=-1) +
        xlab("|Depletion| - |Enrichment|") + ylab("|Depletion| + |Enrichment|")+
        theme(panel.grid.major = element_blank(), 
              panel.grid.minor = element_blank(),
              legend.position = "none",
              panel.background = element_rect(fill = "white", 
                                              colour = "black", 
                                              size = 0.5, 
                                              linetype = "solid"))
      #dev.off()
      #geom_text_repel()
      
      upSpeciesByDiseaseList = split(allPairMeltSel$msp, allPairMeltSel$disease_cohort)
      upSpeciesByDiseaseList = lapply(upSpeciesByDiseaseList, unique)
      upSpeciesCountsByDiseases = unlist(lapply(upSpeciesByDiseaseList, length))
      
      sort(upSpeciesCountsByDiseases)
      overlapDiseases = names(upSpeciesCountsByDiseases)[names(upSpeciesCountsByDiseases) %in% names(downSpeciesCountsByDiseases)]
      plot(upSpeciesCountsByDiseases[match(overlapDiseases, names(upSpeciesCountsByDiseases))],
           downSpeciesCountsByDiseases[match(overlapDiseases, names(downSpeciesCountsByDiseases))], 
           xlab="up", ylab="down")
      
    }
    
    loadDiseaseSigUpdated = F
    if (loadDiseaseSigUpdated) {
      
      allPairMeltRData = "C://Data//comparative.analysis.healthy.sweden//allPairMelt.disease.signatures.RData"
      load(allPairMeltRData)
      allPairDownMeltRData = "C://Data//comparative.analysis.healthy.sweden//allPairDownMelt.disease.signatures.RData"
      load(allPairDownMeltRData)
      
      cohortOrderFile = "C://Data//comparative.analysis.healthy.sweden//disease.cohort.order.txt"
      cohortOrders = read.table(cohortOrderFile, stringsAsFactors = F, header=F)[,]
      names(cohortOrders)=1:30
      
      allMsps = rownames(mergeMatUpdated)
      
      allPairUpMeltImputed = allPairMelt
      allPairUpMeltImputed = allPairUpMeltImputed[!allPairUpMeltImputed$disease_cohort %in% c("NSCLC_FR_id26", "RCC_FR_id41"),]
      allPairUpMeltImputed$effect_size[allPairUpMeltImputed$effect_size < 0] = 0
      allPairUpMeltImputed$effect_size[allPairUpMeltImputed$effect_size > 1] = 1
      allPairUpMeltImputed$effect_size[is.na(allPairUpMeltImputed$effect_size)] = 0
      
      allPairDownMeltImputed = allPairDownMelt
      allPairDownMeltImputed = allPairDownMeltImputed[!allPairDownMeltImputed$disease_cohort %in% c("NSCLC_FR_id26", "RCC_FR_id41"),]
      allPairDownMeltImputed$effect_size[allPairDownMeltImputed$effect_size < 0] = 0
      allPairDownMeltImputed$effect_size[allPairDownMeltImputed$effect_size > 1] = 1
      allPairDownMeltImputed$effect_size[is.na(allPairDownMeltImputed$effect_size)] = 0
      
      manhattanMode = T
      if (manhattanMode) {
        
        ### enriched species ###
        uniqMsp = unique(allPairUpMeltImputed$msp) 
        uniqMspId = 1:length(uniqMsp)
        names(uniqMspId) =uniqMsp
        
        allPairUpMeltImputed$CHR = names(cohortOrders)[match(allPairUpMeltImputed$disease_cohort, cohortOrders)]
        allPairUpMeltImputed$CHR = as.numeric(allPairUpMeltImputed$CHR)
        #allPairUpMeltImputed$CHR = as.numeric(gsub(".*_id", "", allPairUpMeltImputed$disease_cohort))
        allPairUpMeltImputed$BP = uniqMspId[match(allPairUpMeltImputed$msp, names(uniqMspId))]
        allPairUpMeltImputed$P = allPairUpMeltImputed$effect_size
        allPairUpMeltImputed$SNP = allPairUpMeltImputed$msp
        
        ### depleted species ###
        uniqMsp = unique(allPairDownMeltImputed$msp) 
        uniqMspId = 1:length(uniqMsp)
        names(uniqMspId) =uniqMsp
        allPairDownMeltImputed$CHR = names(cohortOrders)[match(allPairDownMeltImputed$disease_cohort, cohortOrders)]
        allPairDownMeltImputed$CHR = as.numeric(allPairDownMeltImputed$CHR)
        #allPairDownMeltImputed$CHR = as.numeric(gsub(".*_id", "", allPairDownMeltImputed$disease_cohort))
        allPairDownMeltImputed$BP = uniqMspId[match(allPairDownMeltImputed$msp, names(uniqMspId))]
        allPairDownMeltImputed$P = allPairDownMeltImputed$effect_size
        allPairDownMeltImputed$SNP = allPairDownMeltImputed$msp
        
      }
      
      allPairUpMeltImputedMatched = allPairUpMeltImputed[allPairUpMeltImputed$control=="matched",]
      allPairUpMeltImputedCountry = allPairUpMeltImputed[allPairUpMeltImputed$control=="country",]
      allPairUpMeltImputedCluster = allPairUpMeltImputed[allPairUpMeltImputed$control=="cluster",]
      
      allPairUpMeltImputedMatchedSig = allPairUpMeltImputedMatched[allPairUpMeltImputedMatched$effect_size>=0.3,]
      allPairUpMeltImputedCountrySig = allPairUpMeltImputedCountry[allPairUpMeltImputedCountry$effect_size>=0.3,]
      allPairUpMeltImputedClusterSig = allPairUpMeltImputedCluster[allPairUpMeltImputedCluster$effect_size>=0.3,]
      
      allPairDownMeltImputedMatched = allPairDownMeltImputed[allPairDownMeltImputed$control=="matched",]
      allPairDownMeltImputedCountry = allPairDownMeltImputed[allPairDownMeltImputed$control=="country",]
      allPairDownMeltImputedCluster = allPairDownMeltImputed[allPairDownMeltImputed$control=="cluster",]
      
      allPairDownMeltImputedMatchedSig = allPairDownMeltImputedMatched[allPairDownMeltImputedMatched$effect_size>=0.3,]
      allPairDownMeltImputedCountrySig = allPairDownMeltImputedCountry[allPairDownMeltImputedCountry$effect_size>=0.3,]
      allPairDownMeltImputedClusterSig = allPairDownMeltImputedCluster[allPairDownMeltImputedCluster$effect_size>=0.3,]
      
      plotManhattanMode = F
      if (plotManhattanMode) {
        require(qqman)
        manhattan(allPairUpMeltImputedCountry, genomewideline=0.5, suggestiveline = 0.3, ylim = c(0, 1.1), col = c("blue4", "orange3"), logp=F)
        manhattan(allPairDownMeltImputedCountry, genomewideline=0.5, suggestiveline = 0.3, ylim = c(0, 1.1), col = c("blue4", "orange3"), logp=F)
        
      }
      
      ids = c(2,4,8,9,11,14,15,20,21,22,23,24,25,26,27,28,29,30)
      ids = as.character(ids)
      
      cohortsWithAllTypeCohorts = cohortOrders[ids]
      
      enrichedIds = c(2,4,9,11,15,21,24,25,28,29,30)
      enrichedIds = as.character(enrichedIds)
      
      depletedIds = c(4,9,11,21)
      depletedIds = as.character(depletedIds)
      
      cohortsWithEnriched = cohortOrders[enrichedIds]
      cohortsWithDepleted = cohortOrders[depletedIds]
      
      checkDifferentControls = T
      if (checkDifferentControls) {
        ### all enriched species from matched controls
        enrichedFromMatchedControlList = sapply(cohortsWithAllTypeCohorts, function(currCohort){
          upSpeciesMatched = allPairUpMeltImputedMatchedSig$msp[allPairUpMeltImputedMatchedSig$disease_cohort == currCohort]
          return(upSpeciesMatched)
          
        }, simplify = F)
        enrichedFromMatchedControls = unique(unlist(enrichedFromMatchedControlList))
        
        ### all enriched species from country
        enrichedFromCountryList = sapply(cohortsWithAllTypeCohorts, function(currCohort){
          upSpeciesCountry = allPairUpMeltImputedCountrySig$msp[allPairUpMeltImputedCountrySig$disease_cohort == currCohort]
          return(upSpeciesCountry)
          
        }, simplify = F)
        enrichedFromCountrySamples = unique(unlist(enrichedFromCountryList))
        
        ### all enriched species from country cluster
        enrichedFromClusterList = sapply(cohortsWithAllTypeCohorts, function(currCohort){
          upSpeciesCluster = allPairUpMeltImputedClusterSig$msp[allPairUpMeltImputedClusterSig$disease_cohort == currCohort]
          return(upSpeciesCluster)
          
        }, simplify = F)
        enrichedFromClusterSamples = unique(unlist(enrichedFromClusterList))
        
        ### check statistics ###
        length(enrichedFromMatchedControls)
        length(enrichedFromCountrySamples)
        length(enrichedFromClusterSamples)
        
        allEnriched = unique(c(enrichedFromMatchedControls, enrichedFromCountrySamples, enrichedFromClusterSamples))
        table(enrichedFromMatchedControls %in% enrichedFromCountrySamples)
        table(enrichedFromMatchedControls %in% enrichedFromClusterSamples)
        table(enrichedFromCountrySamples %in% enrichedFromClusterSamples)
        tt = enrichedFromMatchedControls[enrichedFromMatchedControls %in% enrichedFromCountrySamples]
        table(tt %in% enrichedFromClusterSamples)
        
        ### all depleted sepcies from matched controls ###
        depletedFromMatchedControlList = sapply(cohortsWithAllTypeCohorts, function(currCohort){
          downSpeciesMatched = allPairDownMeltImputedMatchedSig$msp[allPairDownMeltImputedMatchedSig$disease_cohort == currCohort]
          
        }, simplify = F)
        depletedFromMatchedControls = unique(unlist(depletedFromMatchedControlList))
        
        ### all depleted sepcies from country ###
        depletedFromCountryList = sapply(cohortsWithAllTypeCohorts, function(currCohort){
          downSpeciesCountry = allPairDownMeltImputedCountrySig$msp[allPairDownMeltImputedCountrySig$disease_cohort == currCohort]
          return(downSpeciesCountry)
        }, simplify = F)
        depletedFromCountrySamples = unique(unlist(depletedFromCountryList))
        
        ### all depleted sepcies from country cluster ###
        depletedFromClusterList = sapply(cohortsWithAllTypeCohorts, function(currCohort){
          downSpeciesCluster = allPairDownMeltImputedClusterSig$msp[allPairDownMeltImputedClusterSig$disease_cohort == currCohort]
          return(downSpeciesCluster)
        }, simplify = F)
        depletedFromClusterSamples = unique(unlist(depletedFromClusterList))
        
        
        ### check statistics ###  
        length(depletedFromMatchedControls)
        length(depletedFromCountrySamples)
        length(depletedFromClusterSamples)
        
        allDepleted = unique(c(depletedFromMatchedControls, depletedFromCountrySamples, depletedFromClusterSamples))
        table(depletedFromMatchedControls %in% depletedFromCountrySamples)
        table(depletedFromMatchedControls %in% depletedFromClusterSamples)
        table(depletedFromCountrySamples %in% depletedFromClusterSamples)
        tt = depletedFromMatchedControls[depletedFromMatchedControls %in% depletedFromCountrySamples]
        table(tt %in% depletedFromClusterSamples)
        
        
        allOverlap = allEnriched[allEnriched %in% allDepleted]
        sort(getSpeciesName(allOverlap, taxo))
        
        
        ### plots ###
        barplot(c(30,18),las=2, col=col_vector[c(8,4)])
        
        barplot(c(217,67,5),las=2, col=col_vector[1:3])
        barplot(c(260,124,104),las=2, col=col_vector[1:3])
        
        
        ### enriched species 
        sapply(cohortsWithEnriched, function(currCohort){
          upSpeciesMatched = allPairUpMeltImputedMatchedSig$msp[allPairUpMeltImputedMatchedSig$disease_cohort == currCohort]
          upSpeciesCountry = allPairUpMeltImputedCountrySig$msp[allPairUpMeltImputedCountrySig$disease_cohort == currCohort]
          upSpeciesCluster = allPairUpMeltImputedClusterSig$msp[allPairUpMeltImputedClusterSig$disease_cohort == currCohort]
          jaccMaCo = getJaccFrom(upSpeciesMatched, upSpeciesCountry)
          jaccCoCl = getJaccFrom(upSpeciesCountry, upSpeciesCluster)
          jaccMaCl = getJaccFrom(upSpeciesMatched, upSpeciesCluster)
          
          pvalMaCo = getHyperP(upSpeciesMatched, upSpeciesCountry, allMsps, F)
          pvalCoCl = getHyperP(upSpeciesCountry, upSpeciesCluster, allMsps, F)
          pvalMaCl = getHyperP(upSpeciesMatched, upSpeciesCluster, allMsps, F)
          
          print(currCohort)
          print(c(length(upSpeciesMatched), 
                  length(upSpeciesCountry), 
                  length(upSpeciesCluster),
                  pvalMaCo, pvalCoCl, pvalMaCl,
                  jaccMaCo, jaccCoCl, jaccMaCl))
          
        }, simplify = F)
        
        sapply(cohortsWithDepleted, function(currCohort){
          downSpeciesMatched = allPairDownMeltImputedMatchedSig$msp[allPairDownMeltImputedMatchedSig$disease_cohort == currCohort]
          downSpeciesCountry = allPairDownMeltImputedCountrySig$msp[allPairDownMeltImputedCountrySig$disease_cohort == currCohort]
          downSpeciesCluster = allPairDownMeltImputedClusterSig$msp[allPairDownMeltImputedClusterSig$disease_cohort == currCohort]
          jaccMaCo = getJaccFrom(downSpeciesMatched, downSpeciesCountry)
          jaccCoCl = getJaccFrom(downSpeciesCountry, downSpeciesCluster)
          jaccMaCl = getJaccFrom(downSpeciesMatched, downSpeciesCluster)
          
          print(currCohort)
          print(c(length(downSpeciesMatched), 
                  length(downSpeciesCountry), 
                  length(downSpeciesCluster),
                  jaccMaCo, jaccCoCl, jaccMaCl))
          
        }, simplify = F)
        
        sapply(cohortsWithEnriched, function(currCohort){
          upSpeciesMatched = allPairUpMeltImputedMatchedSig$msp[allPairUpMeltImputedMatchedSig$disease_cohort == currCohort]
          upSpeciesCountry = allPairUpMeltImputedCountrySig$msp[allPairUpMeltImputedCountrySig$disease_cohort == currCohort]
          upSpeciesCluster = allPairUpMeltImputedClusterSig$msp[allPairUpMeltImputedClusterSig$disease_cohort == currCohort]
          
          barOut = c(length(upSpeciesMatched),length(upSpeciesCountry),length(upSpeciesCluster))
          barplot(barOut,
                  main=currCohort)
          
        }, simplify = F)
        
        sapply(cohortsWithDepleted, function(currCohort){
          downSpeciesMatched = allPairDownMeltImputedMatchedSig$msp[allPairDownMeltImputedMatchedSig$disease_cohort == currCohort]
          downSpeciesCountry = allPairDownMeltImputedCountrySig$msp[allPairDownMeltImputedCountrySig$disease_cohort == currCohort]
          downSpeciesCluster = allPairDownMeltImputedClusterSig$msp[allPairDownMeltImputedClusterSig$disease_cohort == currCohort]
          
          barOut = c(length(downSpeciesMatched),length(downSpeciesCountry),length(downSpeciesCluster))
          barplot(barOut,
                  main=currCohort)
          
          
        }, simplify = F)
        
        sapply(cohortsWithEnriched, function(currCohort){
          
          upSpeciesCountry = allPairUpMeltImputedCountrySig$msp[allPairUpMeltImputedCountrySig$disease_cohort == currCohort]
          
          currMatched = allPairUpMeltImputedMatched[allPairUpMeltImputedMatched$disease_cohort == currCohort,]
          currCountry = allPairUpMeltImputedCountry[allPairUpMeltImputedCountry$disease_cohort == currCohort,]
          currCluster = allPairUpMeltImputedCluster[allPairUpMeltImputedCluster$disease_cohort == currCohort,]
          
          esFromMatched = currMatched$effect_size[match(upSpeciesCountry, currMatched$msp)]
          esFromCountry = currCountry$effect_size[match(upSpeciesCountry, currCountry$msp)]
          esFromCluster = currCluster$effect_size[match(upSpeciesCountry, currCluster$msp)]
          
          effectSizeMat = cbind(country=esFromCountry,
                                matched=esFromMatched,
                                cluster=esFromCluster) 
          
          effectSizeList = list(country=esFromCountry,
                                matched=esFromMatched, 
                                cluster=esFromCluster)
          #pairs(effectSizeMat)
          boxplot(effectSizeList, main=paste("Enriched",currCohort))
          print(cor(effectSizeMat))
          
          
        }, simplify = F)
        
        sapply(cohortsWithDepleted, function(currCohort){
          downSpeciesCountry = allPairDownMeltImputedCountrySig$msp[allPairDownMeltImputedCountrySig$disease_cohort == currCohort]
          
          currMatched = allPairDownMeltImputedMatched[allPairDownMeltImputedMatched$disease_cohort == currCohort,]
          currCountry = allPairDownMeltImputedCountry[allPairDownMeltImputedCountry$disease_cohort == currCohort,]
          currCluster = allPairDownMeltImputedCluster[allPairDownMeltImputedCluster$disease_cohort == currCohort,]
          
          esFromMatched = currMatched$effect_size[match(downSpeciesCountry, currMatched$msp)]
          esFromCountry = currCountry$effect_size[match(downSpeciesCountry, currCountry$msp)]
          esFromCluster = currCluster$effect_size[match(downSpeciesCountry, currCluster$msp)]
          
          effectSizeMat = cbind(country=esFromCountry,
                                matched=esFromMatched,
                                cluster=esFromCluster) 
          
          effectSizeList = list(country=esFromCountry,
                                matched=esFromMatched, 
                                cluster=esFromCluster)
          #pairs(effectSizeMat)
          boxplot(effectSizeList, main=paste("Depleted",currCohort))
          print(cor(effectSizeMat))
          
          
        }, simplify = F)
        
      }
      
      generalVsUniqueMode = T
      if (generalVsUniqueMode) {
        upSignalTab = rbind(allPairUpMeltImputedCountrySig, 
                            allPairUpMeltImputedMatchedSig, 
                            allPairUpMeltImputedClusterSig)####
        
        
        upCounts = (table(upSignalTab$msp))
        upCounts = as.numeric(upCounts)
        names(upCounts) = names(table(allPairUpMeltImputedCountrySig$msp))
        
        downSignalTab = rbind(allPairDownMeltImputedCountrySig, 
                              allPairDownMeltImputedMatchedSig,
                              allPairDownMeltImputedClusterSig)
        
        downCounts = (table(downSignalTab$msp))
        downCounts = as.numeric(downCounts)
        names(downCounts) = names(table(allPairDownMeltImputedCountrySig$msp))
        
        
        alls = unique(c(names(downCounts), names(upCounts)))
        downCountsAll = downCounts[match(alls, names(downCounts))]
        upCountsAll = upCounts[match(alls, names(upCounts))]
        upCountsAll[is.na(upCountsAll)] = 0
        downCountsAll[is.na(downCountsAll)] = 0
        
        countsTab = data.frame(up=upCountsAll, 
                               down=downCountsAll, 
                               msp=alls, 
                               species = getSpeciesName(alls, taxo),
                               stringsAsFactors = F)
        
        countsTab$sum = countsTab$up + countsTab$down
        #countsTab$diff =  countsTab$down - countsTab$up
        countsTab$diff =countsTab$up - countsTab$down
        countsTab$label = as.character(countsTab$species)
        
        loadMode = T
        if (loadMode) {
          industrialMgsMat = mergeMatUpdated[,industrializedSamples]
          industrialMgsMeans = rowMeans(industrialMgsMat)
          
          statsMgsForNormSamples = "J:\\Deposit\\Project\\2018_microbiome_atlas\\atlas.volcanot.out\\richness.volcano.stats.mgs.for.normal.txt"
          statsMgsForNorm = read.table(statsMgsForNormSamples, sep="\t", header=T, stringsAsFactors = F)
          
          countsTab$abd = industrialMgsMeans[match(countsTab$msp, names(industrialMgsMeans))]*10^11 + 1
          countsTab$richness_lfc = statsMgsForNorm$lfc[match(countsTab$msp,rownames(statsMgsForNorm))]
        }
        
        
        
        #countsTab$abd[countsTab$abd == -Inf] = 0
        plot(upCountsAll, downCountsAll)
        plot(countsTab$diff, countsTab$sum)
        
        depletedCountSpecies = countsTab$msp[countsTab$sum>5&countsTab$diff <= 3]
        enrichedCountSpecies = countsTab$msp[countsTab$sum>5&countsTab$diff >= -3]
        #depletedCountSpecies = countsTab$msp[countsTab$sum>=4 & countsTab$diff < -3]
        #enrichedCountSpecies = countsTab$msp[countsTab$sum>=4 & countsTab$diff > 3]
        
        depletedCountSpeciesTab = data.frame(msp=depletedCountSpecies, 
                                             type="depleted")
        enrichedCountSpeciesTab = data.frame(msp=enrichedCountSpecies, 
                                             type="enriched")
        bothCountSpeciesTab = rbind(enrichedCountSpeciesTab, depletedCountSpeciesTab)
        bothCountSpeciesTab$species = getSpeciesName(bothCountSpeciesTab$msp, taxo)
        View(bothCountSpeciesTab)
        #write.table(bothCountSpeciesTab, "enrichedDepletedAcrossDiseases.NEW.txt", sep="\t")
        
        checkVfSecond = F
        if (checkVfSecond) {
          vfMatRData = "j://Deposit/Project/2018_microbiome_atlas/functional.annotation/PATRIC/vfBestMat.RData"
          load(vfMatRData) #vfMatDigit
          
          depletedVfMat = vfMatDigit[depletedCountSpecies,]
          enrichedVfMat = vfMatDigit[enrichedCountSpecies,]
          
          depletedVfMelt = melt(depletedVfMat)
          enrichedVfMelt = melt(enrichedVfMat)
          depletedVfMelt = depletedVfMelt[depletedVfMelt$value!=0,]
          enrichedVfMelt = enrichedVfMelt[enrichedVfMelt$value!=0,]
          
          
          depletedVfs = rowSums(depletedVfMat)
          enrichedVfs = rowSums(enrichedVfMat)
          
          boxplot(list(depleted=depletedVfs, 
                       enriched=enrichedVfs),log="y")
          
          antismashMatRData = "J://Deposit/Project/2018_microbiome_atlas/functional.annotation/msp1992.igc2.antismashMat.RData"
          load(antismashMatRData)
          
          depletedSecMat = antismashMat[rownames(antismashMat)%in%depletedCountSpecies,]
          enrichedSecMat = antismashMat[rownames(antismashMat)%in%enrichedCountSpecies,]
          depletedSecs = rowSums(depletedSecMat)
          enrichedSecs = rowSums(enrichedSecMat)
          
          boxplot(list(depleted=depletedSecs, 
                       enriched=enrichedSecs),log="y")
          
          
        }
        
        
        diseaseSurvivalSigList = list(Enriched=survivalStats$surv.rates[survivalStats$mgs %in% enrichedCountSpecies ], 
                                      Depleted=survivalStats$surv.rates[survivalStats$mgs %in% depletedCountSpecies],
                                      All=survivalStats$surv.rates)
        t.test(diseaseSurvivalSigList$All, diseaseSurvivalSigList$Depleted)
        t.test(diseaseSurvivalSigList$All, diseaseSurvivalSigList$Enriched)
        t.test(diseaseSurvivalSigList$Enriched, diseaseSurvivalSigList$Depleted)
        
        diseaseSurvivalSigMelt = melt(diseaseSurvivalSigList)
        
        #pdf(file = "enriched.depleted.gut.retaining.probability.pdf", width=2.2, height=3)
        
        ggplot(diseaseSurvivalSigMelt, aes(x=L1, y=value, fill=L1)) +  
          scale_fill_manual(values=c("#80808077","#FF000077", "#0000ff77")) +
          geom_boxplot() +
          xlab("") +
          ylab("") + 
          theme(panel.grid.major = element_blank(), 
                panel.grid.minor = element_blank(),
                legend.position = "none",
                axis.text.x = element_text(angle = 60, hjust=1),
                panel.background = element_rect(fill = "white", 
                                                colour = "black", 
                                                size = 0.5, 
                                                linetype = "solid"))
        #dev.off()
        
        
        #boxplot(diseaseSurvivalSigList)
        
        
        
        require(ggrepel)
        set.seed(1)
        #pdf(file = "enriched.depleted.volcano.pdf", width=3.5, height=2.5)
        ggplot(countsTab, aes(x=diff, y=sum, label=label)) + 
          geom_jitter(shape = 21, colour = "black", width = 0.6, height = 0.6, size=2, aes(fill = diff)) + 
          scale_fill_gradient2(low = "blue", mid="white",  high = "red") +
          geom_hline(yintercept = 5, colour="gray") +
          geom_vline(xintercept = 3, colour="gray") +
          geom_vline(xintercept = -3, colour="gray") +
          geom_abline(slope=1)+
          geom_abline(slope=-1) +
          xlab("|Depletion| - |Enrichment|") + ylab("|Depletion| + |Enrichment|")+
          theme(panel.grid.major = element_blank(), 
                panel.grid.minor = element_blank(),
                legend.position = "none",
                panel.background = element_rect(fill = "white", 
                                                colour = "black", 
                                                size = 0.5, 
                                                linetype = "solid"))
        #dev.off()
        #geom_text_repel()
        
        upSpeciesByDiseaseList = split(allPairMeltSel$msp, allPairMeltSel$disease_cohort)
        upSpeciesByDiseaseList = lapply(upSpeciesByDiseaseList, unique)
        upSpeciesCountsByDiseases = unlist(lapply(upSpeciesByDiseaseList, length))
        
        sort(upSpeciesCountsByDiseases)
        overlapDiseases = names(upSpeciesCountsByDiseases)[names(upSpeciesCountsByDiseases) %in% names(downSpeciesCountsByDiseases)]
        plot(upSpeciesCountsByDiseases[match(overlapDiseases, names(upSpeciesCountsByDiseases))],
             downSpeciesCountsByDiseases[match(overlapDiseases, names(downSpeciesCountsByDiseases))], 
             xlab="up", ylab="down")
        
        
        
      }
      ###
    }
    
    
    ###
    diseaseUniqSubjectList
    mixedUniqCountryForMetaList
    mixedUniqSubjectList
    ###
    
    generateAllMode = F
    if (generateAllMode) {
      
      #load(effectSizeTabRData)
      
      if (F) {
        effectSizePairUpTabRData = "J:\\Deposit\\Project\\2018_microbiome_atlas\\effectSizeTab.all.comparison.2019.09.03.RData"
        load(effectSizePairUpTabRData)
        esTabMatched = effectSizeTab
        
        effectSizeGeoUpTabRData = "J:\\Deposit\\Project\\2018_microbiome_atlas\\effectSizeTab.all.comparison.2019.09.03.GEO.RData"
        load(effectSizeGeoUpTabRData)
        esTabAll = effectSizeTab
        #names(normUniqCountryList)
        
      }
      
      checkFoodBornOutbreak = T
      if (checkFoodBornOutbreak) {
        aSamples = newPerturbCaseList$foodborne_US
        aSampleLen = length(aSamples)
        
        dSamples = normalSamplesByCountries$US
        dSampleLen = length(dSamples)
        
        targetMat = mergeMatUpdated[,aSamples]
        refMat = mergeMatUpdated[,dSamples]
        
        testClasses = c(rep("ref", dSampleLen),
                        rep("target", aSampleLen))
        testMat = cbind(refMat, targetMat)
        testMat = testMat[rowMeans(testMat)!=0,]
        upEffectSizes = apply(testMat, 1, function(currRow){
          effectSize =getEffectSize(currRow[testClasses=="ref"], #getEffectSize
                                    currRow[testClasses=="target"])
          return(effectSize)
        })
        
        downEffectSizes = apply(testMat, 1, function(currRow){
          effectSize =getEffectSizeDown(currRow[testClasses=="ref"],
                                        currRow[testClasses=="target"])
          return(effectSize)
        })
        
        foodbornOutbreak = (data.frame(up=upEffectSizes, 
                                       down=downEffectSizes,
                                       species = getSpeciesName(names(upEffectSizes),taxo)))
        
      }
      
      generateUpMode = F
      if (generateUpMode) {
        #effectSizeTab
        esTabAll = do.call(rbind, sapply(seq_along(diseaseUniqSubjectList), function(dInd){
          
          dSamples = diseaseUniqSubjectList[[dInd]]
          dSampleLen = length(dSamples)
          refName = names(diseaseUniqSubjectList)[dInd]
          refMat = mergeMatUpdated[,dSamples]
          
          statsTab <- do.call(rbind, sapply(seq_along(mixedUniqCountryForMetaList), function(aInd) {
            aSamples = mixedUniqCountryForMetaList[[aInd]]
            aSampleLen = length(aSamples)
            targetName = names(mixedUniqCountryForMetaList)[aInd]
            targetMat = mergeMatUpdated[,aSamples]
            
            
            testClasses = c(rep("ref", dSampleLen),
                            rep("target", aSampleLen))
            testMat = cbind(refMat, targetMat)
            testMat = testMat[rowMeans(testMat)!=0,]
            effectSizes = apply(testMat, 1, function(currRow){
              effectSize =getEffectSize(currRow[testClasses=="ref"], #getEffectSize
                                        currRow[testClasses=="target"])
              return(effectSize)
            })
            
            #msp ID
            #species name
            #ref
            #target
            #effect size
            
            esTab = data.frame(msp=rownames(testMat), 
                               species = getSpeciesName(rownames(testMat), taxo),
                               reference=refName,
                               target = targetName,
                               effectSize= effectSizes)
            return(esTab)
            
            
          }, USE.NAMES = T, simplify = F))
          
          
          
        }, USE.NAMES = T, simplify = F))
        esTabMatched = do.call(rbind, sapply(seq_along(diseaseUniqSubjectList), function(dInd){
          
          dSamples = diseaseUniqSubjectList[[dInd]]
          dSampleLen = length(dSamples)
          refName = names(diseaseUniqSubjectList)[dInd]
          refMat = mergeMatUpdated[,dSamples]
          
          statsTab <- do.call(rbind, sapply(seq_along(mixedUniqSubjectList), function(aInd) {
            aSamples = mixedUniqSubjectList[[aInd]]
            aSampleLen = length(aSamples)
            targetName = names(mixedUniqSubjectList)[aInd]
            targetMat = mergeMatUpdated[,aSamples]
            
            testClasses = c(rep("ref", dSampleLen),
                            rep("target", aSampleLen))
            testMat = cbind(refMat, targetMat)
            testMat = testMat[rowMeans(testMat)!=0,]
            effectSizes = apply(testMat, 1, function(currRow){
              effectSize =getEffectSize(currRow[testClasses=="ref"],
                                        currRow[testClasses=="target"])
              return(effectSize)
            })
            
            esTab = data.frame(msp=rownames(testMat), 
                               species = getSpeciesName(rownames(testMat), taxo),
                               reference=refName,
                               target = targetName,
                               effectSize= effectSizes)
            return(esTab)
            
            
          }, USE.NAMES = T, simplify = F))
        }, USE.NAMES = T, simplify = F))
        
        effectSizePairUpTabRData = "J:\\Deposit\\Project\\2018_microbiome_atlas\\effectSizeTab.all.comparison.2019.09.03.RData"
        effectSizeGeoUpTabRData = "J:\\Deposit\\Project\\2018_microbiome_atlas\\effectSizeTab.all.comparison.2019.09.03.GEO.RData"
        save(esTabMatched, file=effectSizePairUpTabRData)
        save(esTabAll, file=effectSizeGeoUpTabRData)
        
        getPairStats <- function(esTabTarget, diseaseWithMatchedPairTab) {
          references = diseaseWithMatchedPairTab[,1]
          species = unique(esTabTarget$msp)
          
          resultMat = do.call(cbind, lapply(as.list(species), function(sp){
            currTab = esTabTarget[esTabTarget$msp == sp,]
            
            esList = do.call(c,lapply(as.list(references), function(ref) {
              target = diseaseWithMatchedPairTab[diseaseWithMatchedPairTab$reference == ref,"target"]
              if (target == "") return(NA)
              effectSize = currTab[currTab$reference == ref & currTab$target == target,"effectSize"]
              if (length(effectSize)==0) return(NA)
              return(effectSize)
            }))
            return(esList)
          }))
          rownames(resultMat) = references
          
          colnames(resultMat) = species
          resultMat = t(resultMat)
          return(resultMat)
        }
        matchedPairStat = getPairStats(esTabMatched, diseaseWithMatchedPairTab)
        matchedPairStatRData = "C://Data//comparative.analysis.healthy.sweden//matchedPairStat.RData"
        matchedPairStat[matchedPairStat == -Inf ] = NA
        save(matchedPairStat, file=matchedPairStatRData)
        
        countryPairStat = getPairStats(esTabAll, diseaseWithCountryPairTab)
        countryPairStatRData = "C://Data//comparative.analysis.healthy.sweden//countryPairStat.RData"
        countryPairStat[countryPairStat == -Inf ] = NA
        save(countryPairStat, file=countryPairStatRData)
        
        etTypePairStat = getPairStats(esTabAll, diseaseWithEtPairTab)
        etTypePairStatRData = "C://Data//comparative.analysis.healthy.sweden//etTypePairStat.RData"
        etTypePairStat[etTypePairStat= -Inf ] = NA
        save(etTypePairStat, file=etTypePairStatRData)
        rowMaxs(etTypePairStat)
        
        summarizeEffectSizeTab <- function(matchedPairStat, countryPairStat, etTypePairStat) {
          
          allSpecies = sort(unique(c(rownames(countryPairStat),
                                     rownames(matchedPairStat),
                                     rownames(etTypePairStat))))
          
          matchedPairStat = matchedPairStat[match(allSpecies, rownames(matchedPairStat)),]
          countryPairStat = countryPairStat[match(allSpecies, rownames(countryPairStat)),]
          etTypePairStat = etTypePairStat[match(allSpecies, rownames(etTypePairStat)),]
          
          checkMode = F
          if (checkMode) {
            dim(matchedPairStat)
            dim(countryPairStat)
            dim(etTypePairStat)
            
            
            table(rownames(matchedPairStat)==rownames(countryPairStat))
            table(rownames(countryPairStat)==rownames(etTypePairStat))
            
            table(colnames(matchedPairStat)==colnames(countryPairStat))
            table(colnames(countryPairStat)==colnames(etTypePairStat))
            
          }
          
          
          matchedPairMelt = melt(matchedPairStat)
          countryPairMelt = melt(countryPairStat)
          etTypePairMelt = melt(etTypePairStat)
          colnames(matchedPairMelt) = c("msp","disease_cohort","effect_size")
          colnames(countryPairMelt) = c("msp","disease_cohort","effect_size")
          colnames(etTypePairMelt) = c("msp","disease_cohort","effect_size")
          
          matchedPairMelt$msp = as.character(matchedPairMelt$msp)
          countryPairMelt$msp = as.character(countryPairMelt$msp)
          etTypePairMelt$msp = as.character(etTypePairMelt$msp)
          
          matchedPairMelt$disease_cohort = as.character(matchedPairMelt$disease_cohort)
          countryPairMelt$disease_cohort = as.character(countryPairMelt$disease_cohort)
          etTypePairMelt$disease_cohort = as.character(etTypePairMelt$disease_cohort)
          
          matchedPairMelt$control = "matched"
          countryPairMelt$control = "country"
          etTypePairMelt$control = "cluster"
          speciesNames = getSpeciesName(allSpecies, taxo)
          
          matchedPairMelt$species = speciesNames
          countryPairMelt$species = speciesNames
          etTypePairMelt$species = speciesNames
          
          allPairMelt = rbind(matchedPairMelt, 
                              countryPairMelt, 
                              etTypePairMelt)
          allPairMelt = allPairMelt[order(allPairMelt$msp),]
          
          return(allPairMelt)
          
          
        }
        allPairMelt = summarizeEffectSizeTab(matchedPairStat, countryPairStat, etTypePairStat)
        
        
        allPairMeltSel = allPairMelt[!is.na(allPairMelt$effect_size),]
        allPairMeltSel = allPairMeltSel[allPairMeltSel$effect_size>=0.3,]
        sort(table(allPairMeltSel$msp),decreasing = T)[1:10]
        
        
        
      }
      
      generateDownMode = T
      if (generateDownMode) {
        
        effectSizePairDownTab = do.call(rbind, sapply(seq_along(diseaseUniqSubjectList), function(dInd){
          
          dSamples = diseaseUniqSubjectList[[dInd]]
          dSampleLen = length(dSamples)
          refName = names(diseaseUniqSubjectList)[dInd]
          refMat = mergeMatUpdated[,dSamples]
          #mixedUniqCountryForMetaList
          
          statsTab <- do.call(rbind, sapply(seq_along(mixedUniqSubjectList), function(aInd) {
            aSamples = mixedUniqSubjectList[[aInd]] #mixedUniqCountryForMetaList
            aSampleLen = length(aSamples)
            targetName = names(mixedUniqSubjectList)[aInd] #mixedUniqCountryForMetaList
            targetMat = mergeMatUpdated[,aSamples]
            
            testClasses = c(rep("ref", dSampleLen),
                            rep("target", aSampleLen))
            
            testMat = cbind(refMat, targetMat)
            testMat = testMat[rowMeans(testMat)!=0,]
            effectSizes = apply(testMat, 1, function(currRow){
              effectSize =getEffectSizeDown(currRow[testClasses=="ref"],
                                            currRow[testClasses=="target"])
              return(effectSize)
            })
            
            esTab = data.frame(msp=rownames(testMat), 
                               species = getSpeciesName(rownames(testMat), taxo),
                               reference=refName,
                               target = targetName,
                               effectSize= effectSizes)
            return(esTab)
            
          }, USE.NAMES = T, simplify = F))
        }, USE.NAMES = T, simplify = F))
        print(dim(effectSizePairDownTab))
        print(head(effectSizePairDownTab))
        
        effectSizePairDownTabRData = "J:\\Deposit\\Project\\2018_microbiome_atlas\\effectSizeTab.down.all.comparison.2019.09.09.RData"
        save(effectSizePairDownTab, file=effectSizePairDownTabRData)
        
        effectSizePairDownTabSel =effectSizePairDownTab[effectSizePairDownTab$effectSize>=0.5,]
        
        effectSizeGeoDownTab = do.call(rbind, sapply(seq_along(diseaseUniqSubjectList), function(dInd){
          
          dSamples = diseaseUniqSubjectList[[dInd]]
          dSampleLen = length(dSamples)
          refName = names(diseaseUniqSubjectList)[dInd]
          refMat = mergeMatUpdated[,dSamples]
          #mixedUniqCountryForMetaList
          
          statsTab <- do.call(rbind, sapply(seq_along(mixedUniqCountryForMetaList), function(aInd) {
            aSamples = mixedUniqCountryForMetaList[[aInd]]
            aSampleLen = length(aSamples)
            targetName = names(mixedUniqCountryForMetaList)[aInd]
            targetMat = mergeMatUpdated[,aSamples]
            
            testClasses = c(rep("ref", dSampleLen),
                            rep("target", aSampleLen))
            
            testMat = cbind(refMat, targetMat)
            testMat = testMat[rowMeans(testMat)!=0,]
            effectSizes = apply(testMat, 1, function(currRow){
              effectSize =getEffectSizeDown(currRow[testClasses=="ref"],
                                            currRow[testClasses=="target"])
              return(effectSize)
            })
            
            esTab = data.frame(msp=rownames(testMat), 
                               species = getSpeciesName(rownames(testMat), taxo),
                               reference=refName,
                               target = targetName,
                               effectSize= effectSizes)
            return(esTab)
            
          }, USE.NAMES = T, simplify = F))
        }, USE.NAMES = T, simplify = F))
        
        print(dim(effectSizeGeoDownTab))
        print(head(effectSizeGeoDownTab))
        
        effectSizeGeoDownTabRData = "J:\\Deposit\\Project\\2018_microbiome_atlas\\effectSizeTab.down.all.comparison.GEO.2019.09.09.RData"
        save(effectSizeGeoDownTab, file=effectSizeGeoDownTabRData)
        
        getPairStats <- function(esTabTarget, diseaseWithMatchedPairTab) {
          references = diseaseWithMatchedPairTab[,1]
          species = unique(esTabTarget$msp)
          
          resultMat = do.call(cbind, lapply(as.list(species), function(sp){
            currTab = esTabTarget[esTabTarget$msp == sp,]
            
            esList = do.call(c,lapply(as.list(references), function(ref) {
              target = diseaseWithMatchedPairTab[diseaseWithMatchedPairTab$reference == ref,"target"]
              if (target == "") return(NA)
              effectSize = currTab[currTab$reference == ref & currTab$target == target,"effectSize"]
              if (length(effectSize)==0) return(NA)
              return(effectSize)
            }))
            return(esList)
          }))
          rownames(resultMat) = references
          
          colnames(resultMat) = species
          resultMat = t(resultMat)
          return(resultMat)
        }
        
        matchedPairDownStat = getPairStats(effectSizePairDownTab, diseaseWithMatchedPairTab)
        matchedPairDownStatRData = "C://Data//comparative.analysis.healthy.sweden//matchedPairDownStat.RData"
        matchedPairDownStat[matchedPairDownStat == -Inf ] = NA
        save(matchedPairDownStat, file=matchedPairDownStatRData)
        
        countryPairDownStat = getPairStats(effectSizeGeoDownTab, diseaseWithCountryPairTab)
        countryPairDownStatRData = "C://Data//comparative.analysis.healthy.sweden//countryPairDownStat.RData"
        countryPairDownStat[countryPairDownStat == -Inf ] = NA
        save(countryPairDownStat, file=countryPairDownStatRData)
        
        etTypePairDownStat = getPairStats(effectSizeGeoDownTab, diseaseWithEtPairTab)
        etTypePairDownStatRData = "C://Data//comparative.analysis.healthy.sweden//etTypePairDownStat.RData"
        etTypePairDownStat[etTypePairDownStat= -Inf ] = NA
        save(etTypePairDownStat, file=etTypePairDownStatRData)
        rowMaxs(etTypePairDownStat)
        
        summarizeEffectSizeTab <- function(matchedPairStat, countryPairStat, etTypePairStat) {
          
          allSpecies = sort(unique(c(rownames(countryPairStat),
                                     rownames(matchedPairStat),
                                     rownames(etTypePairStat))))
          
          matchedPairStat = matchedPairStat[match(allSpecies, rownames(matchedPairStat)),]
          countryPairStat = countryPairStat[match(allSpecies, rownames(countryPairStat)),]
          etTypePairStat = etTypePairStat[match(allSpecies, rownames(etTypePairStat)),]
          
          checkMode = F
          if (checkMode) {
            dim(matchedPairStat)
            dim(countryPairStat)
            dim(etTypePairStat)
            
            
            table(rownames(matchedPairStat)==rownames(countryPairStat))
            table(rownames(countryPairStat)==rownames(etTypePairStat))
            
            table(colnames(matchedPairStat)==colnames(countryPairStat))
            table(colnames(countryPairStat)==colnames(etTypePairStat))
            
          }
          
          
          matchedPairMelt = melt(matchedPairStat)
          countryPairMelt = melt(countryPairStat)
          etTypePairMelt = melt(etTypePairStat)
          colnames(matchedPairMelt) = c("msp","disease_cohort","effect_size")
          colnames(countryPairMelt) = c("msp","disease_cohort","effect_size")
          colnames(etTypePairMelt) = c("msp","disease_cohort","effect_size")
          
          matchedPairMelt$msp = as.character(matchedPairMelt$msp)
          countryPairMelt$msp = as.character(countryPairMelt$msp)
          etTypePairMelt$msp = as.character(etTypePairMelt$msp)
          
          matchedPairMelt$disease_cohort = as.character(matchedPairMelt$disease_cohort)
          countryPairMelt$disease_cohort = as.character(countryPairMelt$disease_cohort)
          etTypePairMelt$disease_cohort = as.character(etTypePairMelt$disease_cohort)
          
          matchedPairMelt$control = "matched"
          countryPairMelt$control = "country"
          etTypePairMelt$control = "cluster"
          speciesNames = getSpeciesName(allSpecies, taxo)
          
          matchedPairMelt$species = speciesNames
          countryPairMelt$species = speciesNames
          etTypePairMelt$species = speciesNames
          
          allPairMelt = rbind(matchedPairMelt, 
                              countryPairMelt, 
                              etTypePairMelt)
          allPairMelt = allPairMelt[order(allPairMelt$msp),]
          
          return(allPairMelt)
          
          
        }
        allPairDownMelt = summarizeEffectSizeTab(matchedPairDownStat, countryPairDownStat, etTypePairDownStat)
        allPairDownMeltRData = "C://Data//comparative.analysis.healthy.sweden//allPairDownMelt.disease.signatures.RData"
        save(allPairDownMelt, file=allPairDownMeltRData)
        
        allPairDownMeltSel = allPairDownMelt[allPairDownMelt$effect_size>=0.5,]
        allPairDownMeltSel = allPairDownMeltSel[!is.na(allPairDownMeltSel$effect_size),]
        downMsps = unique(allPairDownMeltSel$msp)
        length(downMsps)
        
        
        pathoBionts=colSums(mergeMatUpdated[diseaseMsps,])
        commensals=colSums(mergeMatUpdated[downMsps,])
        
        commPatho = pathoBionts/commensals
        
        commPatho = sort(commPatho, decreasing = T)
        View(basicMetaMapUpdated[names(commPatho[1:20]),])
        
      }
      
      # effectSizeTab$msp = as.character(effectSizeTab$msp)
      # effectSizeTab$species = as.character(effectSizeTab$species)
      # effectSizeTab$reference = as.character(effectSizeTab$reference)
      # effectSizeTab$target = as.character(effectSizeTab$target)
      # save(effectSizeTab, file=effectSizeTabRData)
      
      exCohorts = c(names(newNonWesternSets), "Thailand_id43", names(diseaseUniqSubjectList))
      
      # allPairMeltRData = "C://Data//comparative.analysis.healthy.sweden//allPairMelt.disease.signatures.RData"
      # save(allPairMelt, file=allPairMeltRData)
      
      # allPairDownMeltRData = "C://Data//comparative.analysis.healthy.sweden//allPairDownMelt.disease.signatures.RData"
      # save(allPairDownMelt, file=allPairDownMeltRData)
      
      
      #### visualization ???? ####
      # 
      # View(effectSizeTab[!effectSizeTab$target %in% exCohorts,])
      # dSignatureMsps = findSignatureSpecies(effectSizeTab, exCohorts, 0.5)
      # boxplot(countTabRanks[dSignatureMsps])
      # View(data.frame(countTabRanks[match(dSignatureMsps, names(countTabRanks))], dSignatureMsps, getSpeciesName(dSignatureMsps, taxo)))
      # 
      # 
    }
    
    generateGenusMode = F
    if (generateGenusMode) {
      effectSizeGenusTabRData = "J:\\Deposit\\Project\\2018_microbiome_atlas\\effectSizeTab.genus.all.comparison.2019.09.09.RData"
      effectSizeGenusGeoTabRData = "J:\\Deposit\\Project\\2018_microbiome_atlas\\effectSizeTab.genus.all.comparison.2019.09.09.GEO.RData"
      
    }
    
    funcDiseaseAssociationMode = F
    if (funcDiseaseAssociationMode) {
      allPairMelt = allPairMelt[!allPairMelt$msp %in% suppressMsps2,]
      allPairMeltSel = allPairMelt[!is.na(allPairMelt$effect_size),]
      allPairMeltSel = allPairMeltSel[allPairMeltSel$effect_size>=0.5,]
      sort(table(allPairMeltSel$msp),decreasing = T)[1:10]
      
      
      
      allPairSimple = allPairMeltSel[,c("msp", "disease_cohort")]
      allPairSimple = unique(allPairSimple)
      table(duplicated(allPairSimple))
      
      
      diseaseMsps = unique(allPairMeltSel$msp)
      
      
      mspByDiseaseCohorts = split(allPairMeltSel$msp, allPairMeltSel$disease_cohort)
      mspByDiseaseCohorts = lapply(mspByDiseaseCohorts, unique)
      speciesByDiseaseCohorts = lapply(mspByDiseaseCohorts, function(x) getSpeciesName(x, gutTaxo))
      
      funcMatRData = "j://Deposit/Project/2018_microbiome_atlas/functional.annotation/funcMat.20190808.RData"
      load(funcMatRData)
      
      diseaseMergeMatUpdated = mergeMatUpdated[diseaseMsps,]
      diseaseMergeColsums = colSums(diseaseMergeMatUpdated)
      
      assocFuncMsps = lapply(as.list(names(funcModuleSel)), function(currInd){
        out = grepMspsByFuncModules(funcModuleSel, funcMat, gutTaxo, currInd )
        return(out$msp)
      })
      
      names(assocFuncMsps) = names(funcModuleSel)
      assocFuncMspNums = do.call(c, lapply(assocFuncMsps, length))
      mspCut = 10
      assocFuncMsps = assocFuncMsps[assocFuncMspNums>=mspCut]
      
      funcDiseaseEnrichMat = getEnrichMentGsByGs(assocFuncMsps, mspByDiseaseCohorts)
      funcDiseaseEnrichMelt = melt(funcDiseaseEnrichMat)
      funcDiseaseEnrichSel = funcDiseaseEnrichMelt[funcDiseaseEnrichMelt$value<=1e-5,]
      
      View(funcDiseaseEnrichSel)
      sort(table(funcDiseaseEnrichSel$Var2))
    }
    
  }
  
  densityClustMode = F
  if (densityClustMode) {
    #pcoaGenusMat
    
    etfCountriesUpdated = c("UK","Germany","Italy","Spain","Sweden","Denmark","Finland","Luxembourg","France")
    etpCountriesUpdated = c("Peru","Mongo","Thai","Fiji","Madagascar","Tanzania","India") #,"Fiji","India"
    etbCountriesUpdated = c("China","Japan","US")
    
    etfSamples = do.call(c,normalSamplesByCountries[etfCountriesUpdated])
    etpSamples = do.call(c,normalSamplesByCountries[etpCountriesUpdated])
    etbSamples = do.call(c,normalSamplesByCountries[etbCountriesUpdated])
    
    showDensityClustOutput <- function(abdMat, targetSamples, ...) {
      targetMat = abdMat[targetSamples,]
      set.seed(1)
      genusClust = densityClust(dist(targetMat))
      decisionTab = data.frame(delta=genusClust$delta,
                               rho=genusClust$rho)
      
      plot(genusClust)
      View(decisionTab)
      
      out = list(...)
      if (length(out)!=0) {
        print(names(out))
        rho = out[[1]]
        delta = out[[2]]
        set.seed(1)
        genusClust <- findClusters(genusClust, rho=rho, delta=delta)
        plot(genusClust)
        
        ccc = genusClust$clusters
        plot(targetMat[,1],targetMat[,2],col=ccc)
      }
    }
    
    showDensityClustOutput(pcoaGenusMat2, etfSamples, rho=5, delta=0.6)
    showDensityClustOutput(pcoaGenusMat2, etbSamples, rho=5, delta=0.6)
    showDensityClustOutput(pcoaGenusMat2, etpSamples, rho=5, delta=0.6)
    
    pcoaGenusMat2 = pcoaGenusMat[,1:2]
    genusClust = densityClust(dist(pcoaGenusMat2))
    decisionTab = data.frame(delta=genusClust$delta,
                             rho=genusClust$rho)
    
    plot(genusClust)
    genusClust <- findClusters(genusClust, rho=5, delta=0.12)
    
    ccc = densityClust::clusters(genusClust)
    cc = genusClust$clusters
    table(cc[genusClust$halo])
    cc[genusClust$halo] <- 10
    table(cc)
    cols = rep("#C0C0C022",dim(pcoaGenusMat2)[1])
    cols[cc==1]="red"
    cols[cc==2]="blue"
    
    plot(pcoaGenusMat2[,1],pcoaGenusMat2[,2],col=cc)
    
    plotTSNE(genusClust)
    
    plotMDS(genusClust)
    
    
    
    
    table(genusClust$clusters)
    
  }
  
  dbscanMode = F
  if (dbscanMode) {
    dd = dist(pcoaGenusMat)
    ddMat = as.matrix(dd)
    ddMaxs = sapply(1:5224, function(ind){
      currVec = ddMat[ind,]
      currVec = currVec[-ind]
      return(max(currVec))
    })
    ddMins = sapply(1:5224, function(ind){
      currVec = ddMat[ind,]
      currVec = currVec[-ind]
      return(min(currVec))
    })
    
    nn <- frNN(pcoaGenusMat, eps=.4)
    hist(sapply(adjacencylist(nn), length),
         xlab = "k", main="Number of Neighbors",
         sub = paste("Neighborhood size eps =", nn$eps))
    
    i <- 20
    nn$id[[i]]
    nn$dist[[i]]
    plot(pcoaGenusMat, pch=16,
         col = ifelse(1:nrow(pcoaGenusMat) %in% nn$id[[i]], "#ff240077", "#c0c0c011"))
    
    
    etfCountriesUpdated = c("UK","Germany","Italy","Spain","Sweden","Denmark","Finland","Luxembourg","France")
    etpCountriesUpdated = c("Peru","Mongo","Thai","Fiji","Madagascar","Tanzania","India") #,"Fiji","India"
    etbCountriesUpdated = c("China","Japan","US")
    
    etfSamples = do.call(c,normalSamplesByCountries[etfCountriesUpdated])
    etpSamples = do.call(c,normalSamplesByCountries[etpCountriesUpdated])
    etbSamples = do.call(c,normalSamplesByCountries[etbCountriesUpdated])
    length(etfSamples)
    length(etpSamples)
    length(etbSamples)
    
    pcoaGenusEtfMat = pcoaGenusMat[etfSamples,]
    pcoaGenusEtpMat = pcoaGenusMat[etpSamples,]
    pcoaGenusEtbMat = pcoaGenusMat[etbSamples,]
    
    mPts=10
    eps=0.45
    etfKeyClusterSamples = getKeyClusterFromFirst(pcoaGenusEtfMat, eps, mPts)
    #etbKeyClusterSamples = getKeyClusterFromFirst(pcoaGenusEtbMat, eps, mPts)
    
    mPts=3
    eps=0.3
    #etpKeyClusterSamples = getKeyClusterFromFirst(pcoaGenusEtpMat, eps, mPts)
    
    
    pcoaCols = rep("#c0c0c011", dim(pcoaGenusMat)[1])
    names(pcoaCols) = rownames(pcoaGenusMat)
    pcoaCols[etfKeyClusterSamples] = "#00bfff88"
    #pcoaCols[etbKeyClusterSamples] = "#80008077"
    #pcoaCols[etpKeyClusterSamples] = "#ff240077"
    plot(pcoaGenusMat[,1], pcoaGenusMat[,2], col=pcoaCols, pch=16) #, xlim=c(-0.55,0.4), ylim=c(-0.3,0.7)
    
    BIC <- mclustBIC((pcoaGenusEtbMat[,1:2]))
    mod1 <- Mclust((pcoaGenusEtbMat[,1:2]), x = BIC)
    summary(mod1, parameters = TRUE)
    plot(mod1, what = "classification")
    
    checkParameters <- function(pcoaGenusMat, targetSamples, ptsSeqs=1:50, epsSeqs=seq(0,1, 0.05)) {
      distMat = as.matrix(dist(pcoaGenusMat))
    }
    ### finding parameter that separates normal and diseased samples
  }
  
  if (F) {
    colVec = rep("gray", 5224)
    names(colVec) = rownames(pcoaGenusMat)
    colVec[newWesternIds] = "red"
    colVec[newDiseaseIds] = "blue"
    
    classes = rep("", 5224)
    names(classes) = rownames(pcoaGenusMat)
    classes[newWesternIds] = "normal"
    classes[newDiseaseIds] = "disease"
    
  }
  
  fuzzyClusterMode = F
  if (fuzzyClusterMode) {
    require(ppclust)
    pcoaGenusMat2 = pcoaGenusMat
    pcoaGenusMat2 = pcoaGenusMat2[rownames(pcoaGenusMat2) %in% c(newDiseaseIds, newwes),]
    res.fcm <- fcm(pcoaGenusMat[,1:2], centers=3)
    res.membership=as.data.frame(res.fcm$u)
    res.max = rowMaxs(as.matrix(res.membership))
    names(res.max) = rownames(res.membership)
    hist(res.max)
    
    looseSamples = names(res.max[res.max>=0.9])
    
    table(looseSamples %in% newDiseaseIds)
  }
  
  permanovaMode = T
  if (permanovaMode) {
    adonis(adonisFeatMat ~ adonisClass)
    
    colMeans(adonisMat[adonisMat$Class=="normal",1:2])
    colMeans(adonisMat[adonisMat$Class=="disease",1:2])
    
    colVec = rep("gray", 5224)
    names(colVec) = rownames(pcoaGenusMat)
    colVec[newWesternIds] = "red"
    colVec[newDiseaseIds] = "blue"
    
    plot(pcoaGenusMat[,1], 
         pcoaGenusMat[,2], col=colVec)
    
  }
  
  checkDominancy = F
  if (checkDominancy) {
    plotAllCumSums <- function(relAbdMat) {
      cumSumMat = apply(relAbdMat, 2, function(currCol) {
        names(currCol) = rownames(relAbdMat)
        currCol = sort(currCol, decreasing = T)
        currCol = cumsum(currCol)
        
        return(currCol)
        
      })
      return(cumSumMat)
    }
    cumSumMat = plotAllCumSums(mergeMatRelAbd)
    matplot(1:1977, cumSumMat, type = "l", log="x") 
    
    
    coli_vulgatus = mergeMatRelAbd["msp_0005",]  / mergeMatRelAbd["msp_0069",] 
    coli_vulgatus[mergeMatRelAbd["msp_0069",] == 0 & mergeMatRelAbd["msp_0069",] == 0 ] = 1
    coli_vulgatus = log10(coli_vulgatus)
    coli_vulgatus[coli_vulgatus == -Inf] = -5
    
    coli_vulgatus_List = list(disease=coli_vulgatus[newDiseaseIds],
                              western=coli_vulgatus[newWesternIds],
                              normal=coli_vulgatus[newNormalAllIds],
                              traditional=coli_vulgatus[newNonWesternIds],
                              hgc=coli_vulgatus[hgcSamples],
                              lgc=coli_vulgatus[lgcSamples])
    
    boxplot(coli_vulgatus_List, las=1)
    abline(h=0)
    
    extractDominantSpeciesList <- function(relAbdMat, cutOff = 0.9) {
      speciesList = apply(relAbdMat, 2, function(currCol) {
        names(currCol) = rownames(relAbdMat)
        currCol = sort(currCol, decreasing = T)
        currCol = cumsum(currCol)
        currInd = min(which(currCol >= cutOff))
        return(names(currCol)[1:currInd])
        
      })
    }
    
    dominantSpeciesList = extractDominantSpeciesList(mergeMatRelAbd, 0.9)
    numDominantSpecies = do.call(c, lapply(dominantSpeciesList, length))
    
    length(unique(do.call(c,dominantSpeciesList[names(dominantSpeciesList) %in% newDiseaseIds])))
    length(unique(do.call(c,dominantSpeciesList[names(dominantSpeciesList) %in% newWesternIds])))
    length(unique(do.call(c,dominantSpeciesList[names(dominantSpeciesList) %in% newNonWesternIds])))
    
    diseaseCounts = sort(table((do.call(c,dominantSpeciesList[names(dominantSpeciesList) %in% newDiseaseIds]))),decreasing = T)
    industrialCounts = sort(table((do.call(c,dominantSpeciesList[names(dominantSpeciesList) %in% newWesternIds]))),decreasing = T)
    traditionalCounts = sort(table((do.call(c,dominantSpeciesList[names(dominantSpeciesList) %in% newNonWesternIds]))),decreasing = T)
    
    
    unions = unique(c(names(diseaseCounts),names(industrialCounts),names(traditionalCounts)))
    countTab = data.frame(dd = diseaseCounts[unions],
                          ii = industrialCounts[unions],
                          tt = traditionalCounts[unions])
    countTab = countTab[order(countTab$ii.Freq, decreasing = T),]
    countTab$ranks = 1:1881
    plot(countTab$dd.Freq, countTab$ii.Freq)
    
    countTabRanks = countTab$ranks
    names(countTabRanks) = countTab$ii.Var1
    
    ranksOfDominantSpeciesBySamples <- do.call(c,lapply(dominantSpeciesList, function(species){
      currRanks = countTab$ranks[countTab$ii.Var1 %in% species]
      return(median(currRanks))
    }))
    
    ranksByCohorts = lapply(mixedUniqSubjectList, function(subjs) {
      return(ranksOfDominantSpeciesBySamples[subjs])
    })
    medianRanksByCohorts = do.call(c,lapply(ranksByCohorts, median))
    par(mar=c(5,10,4,1))
    boxplot(ranksByCohorts[names(sort(medianRanksByCohorts))], cex.axis=0.7,
            horizontal=T,las=2)
    
    par(mar=c(5,4,4,1))
    plot(ranksOfDominantSpeciesBySamples, basicMetaMapUpdated$GeneRichness[match(names(ranksOfDominantSpeciesBySamples), basicMetaMapUpdated$sample.ID)], log="xy", xlab="med ranks", ylab="richness")
    
    par(mar=c(5,7,4,1))
    numDominantList = list(disease=numDominantSpecies[newDiseaseIds],
                           western=numDominantSpecies[newWesternIds],
                           normal=numDominantSpecies[newNormalAllIds],
                           traditional=numDominantSpecies[newNonWesternIds],
                           hgc=numDominantSpecies[hgcSamples],
                           lgc=numDominantSpecies[lgcSamples])
    
    boxplot(numDominantList, las=1, horizontal=T)
    
    t.test(numDominantList$western, numDominantList$disease)
    
  }
  
}

### very important part ###
geoMetaAnalysisMode = T
if (geoMetaAnalysisMode) {
  
  rezaGeoMode = F # 2019-09-27
  if (rezaGeoMode) {
    
    statsMgsForTradVsOthersFile = "J:\\Deposit\\Project\\2018_microbiome_atlas\\atlas.geo.difference.stats\\geo.volcano.stats.mgs.trad.vs.others.txt"
    statsMgsForTradVsWesternFile = "J:\\Deposit\\Project\\2018_microbiome_atlas\\atlas.geo.difference.stats\\geo.volcano.stats.mgs.trad.vs.western.txt"
    sampleFile = "J:\\Deposit\\Project\\2018_microbiome_atlas\\atlas.geo.difference.stats\\geo.diff.simple.wilcoxon.output.samples.20190927.txt"
    
    volcano.stats.mgs.trad.others = getVolcanoStat(mergeMatUpdated, industrializedSamples, traditionalSamples, taxo, speciesMode = T)
    volcano.stats.mgs.trad.western = getVolcanoStat(mergeMatUpdated, westernSamples, traditionalSamples, taxo, speciesMode = T)
    
    s1 = data.frame(sample=industrializedSamples, type="others")
    s2 = data.frame(sample=westernSamples, type="western")
    s3 = data.frame(sample=traditionalSamples, type="traditional")
    sFinal = rbind(s1,s2,s3)
    
    write.table(volcano.stats.mgs.trad.others, file=statsMgsForTradVsOthersFile, sep="\t", quote = F)
    write.table(volcano.stats.mgs.trad.western, file=statsMgsForTradVsWesternFile, sep="\t", quote = F)
    write.table(sFinal, file=sampleFile, sep="\t")
    
    
    
  }
  
  loadPair = T
  if (loadPair) {
    diseaseWithCountryPairTab = rbind(c("Atherosclerosis_SW_id1", "Sweden"),
                                      c("CVD_CN_id10", "China"),
                                      c("GDM_CN_id12", "China"),
                                      c("T2D_CN_id2", "China"),
                                      c("T1D_FI_id21", "Finland"),
                                      c("Cirrhosis_UK_id27", "UK"),
                                      c("CFS_US_id34", "US"),
                                      c("NAFLD_IT_id40", "Italy"),
                                      c("NAFLD_ES_id40", "Spain"),
                                      c("CRC_JP_id45", "Japan"),
                                      c("CRC_IT_id46", "Italy"),
                                      c("Behcet_CN_id52", "China"),
                                      c("CD_CN_id6", "China"),
                                      c("Ankylosing_CN_id9", "China"),
                                      c("CRC_US_id11", "US"),
                                      c("T2D_SW_id14", "Sweden"),
                                      c("IGT_SW_id14", "Sweden"),
                                      c("Obesity_DK_id16", "Denmark"),
                                      c("IBD_ES_id17", "Spain"),
                                      c("Cirrhosis_CN_id20", "China"),
                                      c("Obesity_DK_id25", "Denmark"),
                                      c("NSCLC_FR_id26", ""), 
                                      c("Melanoma_US_id3", "US"),
                                      c("RA_CN_id31", "China"),
                                      c("T1D_LU_id35", "Luxembourg"),
                                      c("RCC_FR_id41", ""),
                                      c("CRC_DE_id44", "Germany"),
                                      c("T2D_ES_id53", "Spain"),
                                      c("NAFLD_US_id7", "US"),
                                      c("PD_DE_id8", "Germany"))
    
    diseaseWithCountryPairTab2 = rbind(c("Atherosclerosis_SW_id1", "Sweden"),
                                       c("CVD_CN_id10", "China"),
                                       c("GDM_CN_id12", "China"),
                                       c("T2D_CN_id2", "China"),
                                       c("T1D_FI_id21", "Finland"),
                                       c("Cirrhosis_UK_id27", "UK"),
                                       c("CFS_US_id34", "US"),
                                       c("NAFLD_IT_id40", "Italy"),
                                       c("NAFLD_ES_id40", "Spain"),
                                       c("CRC_JP_id45", "Japan"),
                                       c("CRC_IT_id46", "Italy"),
                                       c("Behcet_CN_id52", "China"),
                                       c("CD_CN_id6", "China"),
                                       c("Ankylosing_CN_id9", "China"),
                                       c("CRC_US_id11", "US"),
                                       c("T2D_SW_id14", "Sweden"),
                                       c("IGT_SW_id14", "Sweden"),
                                       c("Obesity_DK_id16", "Denmark"),
                                       c("IBD_ES_id17", "Spain"),
                                       c("Cirrhosis_CN_id20", "China"),
                                       c("Obesity_DK_id25", "Denmark"),
                                       c("NSCLC_FR_id26", "Germany"), 
                                       c("Melanoma_US_id3", "US"),
                                       c("RA_CN_id31", "China"),
                                       c("T1D_LU_id35", "Luxembourg"),
                                       c("RCC_FR_id41", "Germany"),
                                       c("CRC_DE_id44", "Germany"),
                                       c("T2D_ES_id53", "Spain"),
                                       c("NAFLD_US_id7", "US"),
                                       c("PD_DE_id8", "Germany"))
    
    diseaseWithEtPairTab = rbind(c("Atherosclerosis_SW_id1", "etFirmicutes"),
                                 c("CVD_CN_id10", "etBacteroides"),
                                 c("GDM_CN_id12", "etBacteroides"),
                                 c("T2D_CN_id2", "etBacteroides"),
                                 c("T1D_FI_id21", "etFirmicutes"),
                                 c("Cirrhosis_UK_id27", "etFirmicutes"),
                                 c("CFS_US_id34", "etBacteroides"),
                                 c("NAFLD_IT_id40", "etFirmicutes"),
                                 c("NAFLD_ES_id40", "etFirmicutes"),
                                 c("CRC_JP_id45", "etBacteroides"),
                                 c("CRC_IT_id46", "etFirmicutes"),
                                 c("Behcet_CN_id52", "etBacteroides"),
                                 c("CD_CN_id6", "etBacteroides"),
                                 c("Ankylosing_CN_id9", "etBacteroides"),
                                 c("CRC_US_id11", "etBacteroides"),
                                 c("T2D_SW_id14", "etFirmicutes"),
                                 c("IGT_SW_id14", "etFirmicutes"),
                                 c("Obesity_DK_id16", "etFirmicutes"),
                                 c("IBD_ES_id17", "etFirmicutes"),
                                 c("Cirrhosis_CN_id20", "etBacteroides"),
                                 c("Obesity_DK_id25", "etFirmicutes"),
                                 c("NSCLC_FR_id26", "etFirmicutes"), 
                                 c("Melanoma_US_id3", "etBacteroides"),
                                 c("RA_CN_id31", "etBacteroides"),
                                 c("T1D_LU_id35", "etFirmicutes"),
                                 c("RCC_FR_id41", "etFirmicutes"),
                                 c("CRC_DE_id44", "etFirmicutes"),
                                 c("T2D_ES_id53", "etFirmicutes"),
                                 c("NAFLD_US_id7", "etBacteroides"),
                                 c("PD_DE_id8", "etFirmicutes"))
    
    diseaseWithMatchedPairTab = rbind(c("Atherosclerosis_SW_id1", "Sweden_id1"),
                                      c("CVD_CN_id10", "China_id10"),
                                      c("GDM_CN_id12", "China_id12"),
                                      c("T2D_CN_id2", "China_id2"),
                                      c("T1D_FI_id21", "Finland_id21"),
                                      c("Cirrhosis_UK_id27", ""),
                                      c("CFS_US_id34", "US_id34"),
                                      c("NAFLD_IT_id40", ""),
                                      c("NAFLD_ES_id40", ""),
                                      c("CRC_JP_id45", "Japan_id45"),
                                      c("CRC_IT_id46", "Italy_id46"),
                                      c("Behcet_CN_id52", "China_id52"),
                                      c("CD_CN_id6", ""),
                                      c("Ankylosing_CN_id9", "China_id9"),
                                      c("CRC_US_id11", "US_id11"),
                                      c("T2D_SW_id14", "Sweden_id14"),
                                      c("IGT_SW_id14", "Sweden_id14"),
                                      c("Obesity_DK_id16", "Denmark_id16"),
                                      c("IBD_ES_id17", "Spain_id17"),
                                      c("Cirrhosis_CN_id20", "China_id20"),
                                      c("Obesity_DK_id25", "Denmark_id25"),
                                      c("NSCLC_FR_id26", ""), 
                                      c("Melanoma_US_id3", ""),
                                      c("RA_CN_id31", "China_id31"),
                                      c("T1D_LU_id35", "Luxembourg_id35"),
                                      c("RCC_FR_id41", ""),
                                      c("CRC_DE_id44", "Germany_id44"),
                                      c("T2D_ES_id53", ""),
                                      c("NAFLD_US_id7", ""),
                                      c("PD_DE_id8", "Germany_id8"))
    diseaseWithCountryPairTab = data.frame(diseaseWithCountryPairTab, stringsAsFactors = F)
    diseaseWithMatchedPairTab = data.frame(diseaseWithMatchedPairTab, stringsAsFactors = F)
    diseaseWithEtPairTab = data.frame(diseaseWithEtPairTab, stringsAsFactors = F)
    
    colnames(diseaseWithCountryPairTab) = c("reference","target")
    colnames(diseaseWithMatchedPairTab) = c("reference","target")
    colnames(diseaseWithEtPairTab) = c("reference","target")
    
  }
  
  loadDiseaseSigMode = T
  if (loadDiseaseSigMode) {
    
    allPairMeltRData = "C://Data//comparative.analysis.healthy.sweden//allPairMelt.disease.signatures.RData"
    load(allPairMeltRData)
    allPairDownMeltRData = "C://Data//comparative.analysis.healthy.sweden//allPairDownMelt.disease.signatures.RData"
    load(allPairDownMeltRData)
    
    cohortOrderFile = "C://Data//comparative.analysis.healthy.sweden//disease.cohort.order.txt"
    cohortOrders = read.table(cohortOrderFile, stringsAsFactors = F, header=F)[,]
    names(cohortOrders)=1:30
    
    allPairMeltImputed = allPairMelt
    #allPairMeltImputed = allPairMeltImputed[!allPairMeltImputed$disease_cohort %in% c("NSCLC_FR_id26", "RCC_FR_id41", "T1D_FI_id21"),]
    allPairMeltImputed$effect_size[allPairMeltImputed$effect_size < 0] = 0
    allPairMeltImputed$effect_size[allPairMeltImputed$effect_size > 1] = 1
    allPairMeltImputed$effect_size[is.na(allPairMeltImputed$effect_size)] = 0
    
    manhattanMode = T
    if (manhattanMode) {
      uniqMsp = unique(allPairMeltImputed$msp) 
      uniqMspId = 1:length(uniqMsp)
      names(uniqMspId) =uniqMsp
      
      allPairMeltImputed$CHR = names(cohortOrders)[match(allPairMeltImputed$disease_cohort, cohortOrders)]
      allPairMeltImputed$CHR = as.numeric(allPairMeltImputed$CHR)
      #allPairMeltImputed$CHR = as.numeric(gsub(".*_id", "", allPairMeltImputed$disease_cohort))
      allPairMeltImputed$BP = uniqMspId[match(allPairMeltImputed$msp, names(uniqMspId))]
      allPairMeltImputed$P = allPairMeltImputed$effect_size
      allPairMeltImputed$SNP = allPairMeltImputed$msp
      
    }
    
    allPairMeltImputedMatched = allPairMeltImputed[allPairMeltImputed$control=="matched",]
    allPairMeltImputedCountry = allPairMeltImputed[allPairMeltImputed$control=="country",]
    allPairMeltImputedCluster = allPairMeltImputed[allPairMeltImputed$control=="cluster",]
    
    allPairMeltImputedMatchedSig = allPairMeltImputedMatched[allPairMeltImputedMatched$effect_size>=0.3,]
    allPairMeltImputedCountrySig = allPairMeltImputedCountry[allPairMeltImputedCountry$effect_size>=0.3,]
    allPairMeltImputedClusterSig = allPairMeltImputedCluster[allPairMeltImputedCluster$effect_size>=0.3,]
    
    
    allPairDownMeltImputed = allPairDownMelt
    #allPairDownMeltImputed = allPairDownMeltImputed[!allPairDownMeltImputed$disease_cohort %in% c("NSCLC_FR_id26", "RCC_FR_id41", "T1D_FI_id21"),]
    allPairDownMeltImputed$effect_size[allPairDownMeltImputed$effect_size < 0] = 0
    allPairDownMeltImputed$effect_size[allPairDownMeltImputed$effect_size > 1] = 1
    allPairDownMeltImputed$effect_size[is.na(allPairDownMeltImputed$effect_size)] = 0
    
    manhattanMode = T
    if (manhattanMode) {
      uniqMsp = unique(allPairDownMeltImputed$msp) 
      uniqMspId = 1:length(uniqMsp)
      names(uniqMspId) =uniqMsp
      allPairDownMeltImputed$CHR = names(cohortOrders)[match(allPairDownMeltImputed$disease_cohort, cohortOrders)]
      allPairDownMeltImputed$CHR = as.numeric(allPairDownMeltImputed$CHR)
      #allPairDownMeltImputed$CHR = as.numeric(gsub(".*_id", "", allPairDownMeltImputed$disease_cohort))
      allPairDownMeltImputed$BP = uniqMspId[match(allPairDownMeltImputed$msp, names(uniqMspId))]
      allPairDownMeltImputed$P = allPairDownMeltImputed$effect_size
      allPairDownMeltImputed$SNP = allPairDownMeltImputed$msp
    }
    
    allPairDownMeltImputedMatched = allPairDownMeltImputed[allPairDownMeltImputed$control=="matched",]
    allPairDownMeltImputedCountry = allPairDownMeltImputed[allPairDownMeltImputed$control=="country",]
    allPairDownMeltImputedCluster = allPairDownMeltImputed[allPairDownMeltImputed$control=="cluster",]
    
    allPairDownMeltImputedMatchedSig = allPairDownMeltImputedMatched[allPairDownMeltImputedMatched$effect_size>=0.3,]
    allPairDownMeltImputedCountrySig = allPairDownMeltImputedCountry[allPairDownMeltImputedCountry$effect_size>=0.3,]
    allPairDownMeltImputedClusterSig = allPairDownMeltImputedCluster[allPairDownMeltImputedCluster$effect_size>=0.3,]
    
    depletedMspByCohortsWCountryControl = split(allPairDownMeltImputedCountrySig$msp,
                                                allPairDownMeltImputedCountrySig$disease_cohort)
    
    enrichedMspByCohortsWCountryControl = split(allPairMeltImputedCountrySig$msp,
                                                allPairMeltImputedCountrySig$disease_cohort)
    
  }
  
  mgsMode = T
  if (mgsMode) {
    normUniqCountryListNew = normUniqCountryList[names(normUniqCountryList) != "Finland"]
    normUniqAllNew = unique(do.call(c, normUniqCountryListNew))
    
    #avgMgsGeoMat = getAvgAbdMatBy(mergeMatUpdated, normUniqCountryList)
    #avgGenusGeoMat = getAvgAbdMatBy(genusMatUpdated, normUniqCountryList)
    avgMgsGeoMat = getAvgAbdMatBy(mergeMatUpdated, normUniqCountryListNew)
    avgGenusGeoMat = getAvgAbdMatBy(genusMatUpdated, normUniqCountryListNew)
    
    normMgsMat = mergeMatUpdated[,normUniqAllNew]
    normGenusMat = genusMatUpdated[,normUniqAllNew]
    
    if (F) {
      tradMgsMat = mergeMatUpdated[,traditionalSamples]
      indMgsMat = mergeMatUpdated[,industrializedSamples]
      
      tradMgsMatHigh <- tradMgsMat >= 1e-6
      indMgsMatHigh  <- indMgsMat >= 1e-6
      
      tradMgsMatHighCounts = rowSums(tradMgsMatHigh)
      indMgsMatHighCounts = rowSums(indMgsMatHigh)
      
      industry=list(B.vulgatus=list(high=indMgsMatHighCounts["msp_0069"]/1864,
                                    low=(1864-indMgsMatHighCounts["msp_0069"])/1864),
                    P.copri=list(high=indMgsMatHighCounts["msp_0041"]/1864,
                                 low=(1864-indMgsMatHighCounts["msp_0041"])/1864))
      industryMelt = melt(industry)
      industryMelt$type="Industrial"
      
      ggplot(industryMelt, aes(y=value,x=L1)) + geom_bar(aes(fill=L2), stat="identity")
      
      traditional=list(B.vulgatus=list(high=tradMgsMatHighCounts["msp_0069"]/262,
                                       low=(262-tradMgsMatHighCounts["msp_0069"])/262),
                       P.copri=list(high=tradMgsMatHighCounts["msp_0041"]/262,
                                    low=(262-tradMgsMatHighCounts["msp_0041"])/262))
      traditionalMelt = melt(traditional)
      traditionalMelt$type = "Traditional"
      
      bothMelt = rbind(industryMelt, traditionalMelt)
      
      blueCol = rgb(5/255,48/255,97/255,0.5)
      
      ggplot(bothMelt, aes(y=value,x=L1)) + 
        geom_bar(aes(fill=L2),colour="black", stat="identity") + 
        facet_wrap(~type) + 
        xlab("") + ylab("") +
        scale_fill_manual(values=c(blueCol,"white")) +
        theme(panel.grid.major = element_blank(), 
              panel.grid.minor = element_blank(),
              axis.text.x = element_text(angle = 60, hjust=1),
              legend.position = "none",
              panel.background = element_rect(fill = "white", 
                                              colour = "black", 
                                              size = 0.5, 
                                              linetype = "solid")) 
      
      
      
      tradMgsRankMat = apply(tradMgsMat, 2, function(x) rownames(tradMgsMat)[order(x,decreasing = T)])
      rownames(tradMgsRankMat) = 1:1977
      
      
      indMgsRankMat = apply(indMgsMat, 2, function(x) rownames(indMgsMat)[order(x,decreasing = T)])
      rownames(indMgsRankMat) = 1:1977
      
      checkTopTradMgsRank = apply(tradMgsRankMat, 1, function(x) sum(x==1))
      checkTopIndMgsRank = apply(indMgsRankMat, 1, function(x) sum(x==1))
      
    }
    
    generateMode = F
    if (generateMode) {
      #mgsGeoMinEsOut = checkMinEffectSizeOfFeatures(normUniqCountryList, mergeMatUpdated)
      mgsGeoMinEsOut = checkMinEffectSizeOfFeatures(normUniqCountryListNew, mergeMatUpdated)
      mgsGeoMinEsOutRData = "C:/Data/comparative.analysis.healthy.sweden/mgsGeoMinEsOut.20190924.RData"
      save(mgsGeoMinEsOut, file=mgsGeoMinEsOutRData)
      
      #mgsGeoMaxEsOut = checkMaxEffectSizeOfFeatures(normUniqCountryList, mergeMatUpdated)
      #mgsGeoEs02Out = checkBestFeaturesInES(normUniqCountryList, mergeMatUpdated, esCut=0.2)
      mgsGeoEs03Out = checkBestFeaturesInES(normUniqCountryListNew, mergeMatUpdated, esCut=0.3)
      mgsGeoEs03OutRData = "C:/Data/comparative.analysis.healthy.sweden/mgsGeoEs03Out.20190924.RData"
      save(mgsGeoEs03Out, file=mgsGeoEs03OutRData)
      
    }
    
    generateWoWellness = F
    if (generateWoWellness) {
      wellnessMetadata = metaTab[metaTab$dataset.ID=="id28",]
      wellnessSamples = (wellnessMetadata$sample.ID)
      
      normUniqCountryListNew$Sweden = normUniqCountryListNew$Sweden[!normUniqCountryListNew$Sweden %in% wellnessSamples]
      mgsGeoMinEsOut = checkMinEffectSizeOfFeatures(normUniqCountryListNew, mergeMatUpdated)
      mgsGeoMinEsOutRData = "C:/Data/comparative.analysis.healthy.sweden/mgsGeoMinEsOut.wo.wellness.20200111.RData"
      save(mgsGeoMinEsOut, file=mgsGeoMinEsOutRData)
      
      mgsGeoEs03Out = checkBestFeaturesInES(normUniqCountryListNew, mergeMatUpdated, esCut=0.3)
      mgsGeoEs03OutRData = "C:/Data/comparative.analysis.healthy.sweden/mgsGeoEs03Out.wo.wellness.20200111.RData"
      save(mgsGeoEs03Out, file=mgsGeoEs03OutRData)
      
      
    }
    
    loadMode = T
    if (loadMode) {
      mgsGeoMinEsOutRData = "C:/Data/comparative.analysis.healthy.sweden/mgsGeoMinEsOut.20190924.RData"
      mgsGeoEs03OutRData = "C:/Data/comparative.analysis.healthy.sweden/mgsGeoEs03Out.20190924.RData"
      load(mgsGeoMinEsOutRData)
      load(mgsGeoEs03OutRData)
    }
    
    loadCompCohortPair = T
    if (loadCompCohortPair) {
      diseaseWithCountryPairTab = rbind(c("Atherosclerosis_SW_id1", "Sweden"),
                                        c("CVD_CN_id10", "China"),
                                        c("GDM_CN_id12", "China"),
                                        c("T2D_CN_id2", "China"),
                                        c("T1D_FI_id21", "Finland"),
                                        c("Cirrhosis_UK_id27", "UK"),
                                        c("CFS_US_id34", "US"),
                                        c("NAFLD_IT_id40", "Italy"),
                                        c("NAFLD_ES_id40", "Spain"),
                                        c("CRC_JP_id45", "Japan"),
                                        c("CRC_IT_id46", "Italy"),
                                        c("Behcet_CN_id52", "China"),
                                        c("CD_CN_id6", "China"),
                                        c("Ankylosing_CN_id9", "China"),
                                        c("CRC_US_id11", "US"),
                                        c("T2D_SW_id14", "Sweden"),
                                        c("IGT_SW_id14", "Sweden"),
                                        c("Obesity_DK_id16", "Denmark"),
                                        c("IBD_ES_id17", "Spain"),
                                        c("Cirrhosis_CN_id20", "China"),
                                        c("Obesity_DK_id25", "Denmark"),
                                        c("NSCLC_FR_id26", ""), 
                                        c("Melanoma_US_id3", "US"),
                                        c("RA_CN_id31", "China"),
                                        c("T1D_LU_id35", "Luxembourg"),
                                        c("RCC_FR_id41", ""),
                                        c("CRC_DE_id44", "Germany"),
                                        c("T2D_ES_id53", "Spain"),
                                        c("NAFLD_US_id7", "US"),
                                        c("PD_DE_id8", "Germany"))
      
      diseaseWithCountryPairTab2 = rbind(c("Atherosclerosis_SW_id1", "Sweden"),
                                         c("CVD_CN_id10", "China"),
                                         c("GDM_CN_id12", "China"),
                                         c("T2D_CN_id2", "China"),
                                         c("T1D_FI_id21", "Finland"),
                                         c("Cirrhosis_UK_id27", "UK"),
                                         c("CFS_US_id34", "US"),
                                         c("NAFLD_IT_id40", "Italy"),
                                         c("NAFLD_ES_id40", "Spain"),
                                         c("CRC_JP_id45", "Japan"),
                                         c("CRC_IT_id46", "Italy"),
                                         c("Behcet_CN_id52", "China"),
                                         c("CD_CN_id6", "China"),
                                         c("Ankylosing_CN_id9", "China"),
                                         c("CRC_US_id11", "US"),
                                         c("T2D_SW_id14", "Sweden"),
                                         c("IGT_SW_id14", "Sweden"),
                                         c("Obesity_DK_id16", "Denmark"),
                                         c("IBD_ES_id17", "Spain"),
                                         c("Cirrhosis_CN_id20", "China"),
                                         c("Obesity_DK_id25", "Denmark"),
                                         c("NSCLC_FR_id26", "Germany"), 
                                         c("Melanoma_US_id3", "US"),
                                         c("RA_CN_id31", "China"),
                                         c("T1D_LU_id35", "Luxembourg"),
                                         c("RCC_FR_id41", "Germany"),
                                         c("CRC_DE_id44", "Germany"),
                                         c("T2D_ES_id53", "Spain"),
                                         c("NAFLD_US_id7", "US"),
                                         c("PD_DE_id8", "Germany"))
      
      diseaseWithEtPairTab = rbind(c("Atherosclerosis_SW_id1", "etFirmicutes"),
                                   c("CVD_CN_id10", "etBacteroides"),
                                   c("GDM_CN_id12", "etBacteroides"),
                                   c("T2D_CN_id2", "etBacteroides"),
                                   c("T1D_FI_id21", "etFirmicutes"),
                                   c("Cirrhosis_UK_id27", "etFirmicutes"),
                                   c("CFS_US_id34", "etBacteroides"),
                                   c("NAFLD_IT_id40", "etFirmicutes"),
                                   c("NAFLD_ES_id40", "etFirmicutes"),
                                   c("CRC_JP_id45", "etBacteroides"),
                                   c("CRC_IT_id46", "etFirmicutes"),
                                   c("Behcet_CN_id52", "etBacteroides"),
                                   c("CD_CN_id6", "etBacteroides"),
                                   c("Ankylosing_CN_id9", "etBacteroides"),
                                   c("CRC_US_id11", "etBacteroides"),
                                   c("T2D_SW_id14", "etFirmicutes"),
                                   c("IGT_SW_id14", "etFirmicutes"),
                                   c("Obesity_DK_id16", "etFirmicutes"),
                                   c("IBD_ES_id17", "etFirmicutes"),
                                   c("Cirrhosis_CN_id20", "etBacteroides"),
                                   c("Obesity_DK_id25", "etFirmicutes"),
                                   c("NSCLC_FR_id26", "etFirmicutes"), 
                                   c("Melanoma_US_id3", "etBacteroides"),
                                   c("RA_CN_id31", "etBacteroides"),
                                   c("T1D_LU_id35", "etFirmicutes"),
                                   c("RCC_FR_id41", "etFirmicutes"),
                                   c("CRC_DE_id44", "etFirmicutes"),
                                   c("T2D_ES_id53", "etFirmicutes"),
                                   c("NAFLD_US_id7", "etBacteroides"),
                                   c("PD_DE_id8", "etFirmicutes"))
      
      diseaseWithMatchedPairTab = rbind(c("Atherosclerosis_SW_id1", "Sweden_id1"),
                                        c("CVD_CN_id10", "China_id10"),
                                        c("GDM_CN_id12", "China_id12"),
                                        c("T2D_CN_id2", "China_id2"),
                                        c("T1D_FI_id21", "Finland_id21"),
                                        c("Cirrhosis_UK_id27", ""),
                                        c("CFS_US_id34", "US_id34"),
                                        c("NAFLD_IT_id40", ""),
                                        c("NAFLD_ES_id40", ""),
                                        c("CRC_JP_id45", "Japan_id45"),
                                        c("CRC_IT_id46", "Italy_id46"),
                                        c("Behcet_CN_id52", "China_id52"),
                                        c("CD_CN_id6", ""),
                                        c("Ankylosing_CN_id9", "China_id9"),
                                        c("CRC_US_id11", "US_id11"),
                                        c("T2D_SW_id14", "Sweden_id14"),
                                        c("IGT_SW_id14", "Sweden_id14"),
                                        c("Obesity_DK_id16", "Denmark_id16"),
                                        c("IBD_ES_id17", "Spain_id17"),
                                        c("Cirrhosis_CN_id20", "China_id20"),
                                        c("Obesity_DK_id25", "Denmark_id25"),
                                        c("NSCLC_FR_id26", ""), 
                                        c("Melanoma_US_id3", ""),
                                        c("RA_CN_id31", "China_id31"),
                                        c("T1D_LU_id35", "Luxembourg_id35"),
                                        c("RCC_FR_id41", ""),
                                        c("CRC_DE_id44", "Germany_id44"),
                                        c("T2D_ES_id53", ""),
                                        c("NAFLD_US_id7", ""),
                                        c("PD_DE_id8", "Germany_id8"))
      diseaseWithCountryPairTab = data.frame(diseaseWithCountryPairTab, stringsAsFactors = F)
      diseaseWithMatchedPairTab = data.frame(diseaseWithMatchedPairTab, stringsAsFactors = F)
      diseaseWithEtPairTab = data.frame(diseaseWithEtPairTab, stringsAsFactors = F)
      
      colnames(diseaseWithCountryPairTab) = c("reference","target")
      colnames(diseaseWithMatchedPairTab) = c("reference","target")
      colnames(diseaseWithEtPairTab) = c("reference","target")
      
    }
    
    checkSignalMode = T
    if (checkSignalMode) {
      
      specificMode = T
      if (specificMode) {
        esCut = 0.3
        maxMgsGeoMinEs = rowMaxs(mgsGeoMinEsOut)
        names(maxMgsGeoMinEs)=rownames(mgsGeoMinEsOut)
        
        mgsGeoMinEsCountries = colnames(mgsGeoMinEsOut)
        maxCountries = apply(mgsGeoMinEsOut, 1, function(currRow){
          maxEs = max(currRow)
          maxInd=which.max(currRow)
          if (maxEs>=esCut)return(mgsGeoMinEsCountries[maxInd])
          return(NA)
        })
        names(maxCountries) = rownames(mgsGeoMinEsOut)
        maxCountries=maxCountries[!is.na(maxCountries)]
        maxCountriesTab = data.frame(value=maxCountries, L1=names(maxCountries), stringsAsFactors = F)
        specificMgsByCountries = split(names(maxCountries), maxCountries)
        specificMgsAll = do.call(c, specificMgsByCountries)
        getSpeciesName(specificMgsAll, taxo)
        
        
        choseTopPerCountries <- apply(mgsGeoMinEsOut, 2, function(currCol) {
          ind=which.max(currCol)
          return(rownames(mgsGeoMinEsOut)[ind])
        })
        
      }
      
      groupMode = T
      if (groupMode) {
        #numCut = 9
        numCut = 6
        mgsGeoEs03Countries = colnames(mgsGeoEs03Out)
        groupCountries = apply(mgsGeoEs03Out, 1, function(currRow){
          currCountries = mgsGeoEs03Countries[currRow>=numCut]
          if (length(currCountries)==0) return(NA)
          return(currCountries)
        })
        groupCountries = groupCountries[unlist(lapply(groupCountries, function(x)!all(is.na(x))))]
        groupCountriesMelt = melt(groupCountries)
        groupCountriesMelt$value=as.character(groupCountriesMelt$value)
        groupCountriesMelt$L1=as.character(groupCountriesMelt$L1)
        groupMgsByCountries = split(groupCountriesMelt$L1, groupCountriesMelt$value)
        
        if (specificMode==T) {
          specificGroupCountriesMelt = rbind(maxCountriesTab, groupCountriesMelt)  
          specificGroupMgsByCountries = split(specificGroupCountriesMelt$L1, specificGroupCountriesMelt$value)
          specificGroupMgsAll = unique(do.call(c,specificGroupMgsByCountries))
          specificGroupMgsCountsByCountries = unlist(lapply(specificGroupMgsByCountries, length))
          
          getHyperP(bothCountSpeciesTab$msp[bothCountSpeciesTab$type=="enriched"], unique(do.call(c,specificGroupMgsByCountries[industrializedCountries])), rownames(wellnessMgsMat), debug = T)
          getHyperP(bothCountSpeciesTab$msp[bothCountSpeciesTab$type=="depleted"], unique(do.call(c,specificGroupMgsByCountries[industrializedCountries])), rownames(wellnessMgsMat), debug = T)
          
          saveMode = F
          if (saveMode) {
            specificGroupCountriesMeltRData = "J://Deposit//Project//2018_microbiome_atlas//atlas.geography//specificGroupCountriesMelt.RData"
            specificGroupMgsByCountriesRData = "J://Deposit//Project//2018_microbiome_atlas//atlas.geography//specificGroupMgsByCountries.RData"
            specificGroupMgsAllRData = "J://Deposit//Project//2018_microbiome_atlas//atlas.geography//specificGroupMgsAll.RData"
            specificGroupMgsCountsByCountriesRData = "J://Deposit//Project//2018_microbiome_atlas//atlas.geography//specificGroupMgsCountsByCountries.RData" 
            
            save(specificGroupCountriesMelt, file=specificGroupCountriesMeltRData)
            save(specificGroupMgsByCountries, file=specificGroupMgsByCountriesRData)
            save(specificGroupMgsAll, file=specificGroupMgsAllRData)
            save(specificGroupMgsCountsByCountries, file=specificGroupMgsCountsByCountriesRData)
            
          }
          
          makeTable = F
          if (makeTable) {
            uniqMsps = unique(specificGroupCountriesMelt$L1)
            outCountries = sapply(uniqMsps, function(currMsp) {
              currCountries = specificGroupCountriesMelt$value[specificGroupCountriesMelt$L1 == currMsp]
              currCountries = unique(currCountries)
              currCountriesStr = paste(currCountries, collapse = ";")
              return(currCountriesStr)
            },simplify = F)
            
            coutrySpecificSpeciesTab = data.frame(msp=names(outCountries),
                                                  enrichedCountries=unlist(outCountries),
                                                  species=getSpeciesName(names(outCountries), taxo))
            
            coutrySpecificSpeciesFile="J:\\Deposit\\Project\\2018_microbiome_atlas\\atlas.geography\\coutrySpecificSpecies.tab.txt"
            write.table(coutrySpecificSpeciesTab,
                        coutrySpecificSpeciesFile, sep="\t", row.names = F, quote = F)
          }
          
          checkMode = F
          if (checkMode) {
            par(mar=c(10,4,4,1))
            sortedCounts = sort(specificGroupMgsCountsByCountries,decreasing = T)
            sortedCols = rep("",length(sortedCounts))
            names(sortedCols) = names(sortedCounts)
            sortedCols[europeanCountries] = eurCol
            sortedCols[traditionalCountries] = tradCol
            sortedCols[c("China","US","Japan")] = cjpCol
            barplot(sortedCounts, col=sortedCols, las=2, cex.names = 0.7)
            
            seq1 = c("Peru","Tanzania","Mongo","Madagascar","Thailand","Fiji","India")
            seq2 = c("Spain","Italy","Luxembourg","Sweden","Germany","Denmark","UK")
            seq3 = c("Japan","US","China")
            
            tradVec = c(255,127,0,200)/255
            tradCol2 = rgb(tradVec[1],tradVec[2],tradVec[3],tradVec[4])
            
            cjpVec = c(211,211,211,200)/255
            cjpCol2 = rgb(cjpVec[1],cjpVec[2],cjpVec[3],cjpVec[4])
            
            eurVec = c(117,112,179,200)/255
            eurCol2 = rgb(eurVec[1],eurVec[2],eurVec[3],eurVec[4])
            
            sortedCountsNew = c(sortedCounts[seq1], sortedCounts[seq2])
            sortedColsNew = c(rep(tradCol2, length(seq1)), rep(eurCol2, length(seq2)))
            par(mar=c(10,3,2,1))
            barplot(sortedCountsNew, col=sortedColsNew, las=2, cex.names = 0.9, ylim=c(0,400), yaxt = "n")
            axis(2, at = c(0,100,200,300,400), las = 2)
            
            par(mar=c(10,3,2,1))
            barplot(sortedCounts[seq3], col=cjpCol2, las=2, cex.names = 0.9, ylim=c(0,30), yaxt="n")
            axis(2, at = c(0,10,20,30), las = 2)
            
            barplot(sortedCounts[seq1], col=tradCol, las=2, cex.names = 0.7)
            barplot(sortedCounts[seq2], col=eurCol, las=2, cex.names = 0.7)
            barplot(sortedCounts[seq3], col=cjpCol, las=2, cex.names = 0.7)
            
            ###
            iniR = 0.2
            pie(1, radius=0.01, init.angle=90, col=c('white'), border = NA, labels='')
            
            floating.pie(0,0,c(782,1195),  
                         radius=3*iniR, 
                         startpos=pi/2, 
                         col = c("gold","gray"),
                         border="black")
            floating.pie(0,0,1,  
                         radius=2*iniR, 
                         startpos=pi/2, 
                         col="white",
                         border="black")
            floating.pie(0,0,1,  
                         radius=1.99*iniR, 
                         startpos=pi/2, 
                         col="white",
                         border=NA)
            ###
            
            
            
            #general vs unique
            mpsGeoCounts = table(as.character(specificGroupCountriesMelt$L1))
            mspGeoCountTab = data.frame(msp=names(mpsGeoCounts),
                                        geo_counts=as.numeric(mpsGeoCounts),
                                        max_ES =maxMgsGeoMinEs[match(names(mpsGeoCounts),names(maxMgsGeoMinEs))] )
            plot(mspGeoCountTab$geo_counts,
                 mspGeoCountTab$max_ES)
            hist(mpsGeoCounts, breaks=5, main="", col=colorRampPalette(c("white","gray","blue"))(7),
                 xlab="",ylab="",las=1)
            mpsGeoCounts[mpsGeoCounts>=5]
            
            selInd = mpsGeoCounts>=5
            specificGroupCountriesFreq = specificGroupCountriesMelt[specificGroupCountriesMelt$L1 %in% names(mpsGeoCounts[selInd]),]
            specificGroupCountriesFreq$count=1
            geoFreqMat = acast(specificGroupCountriesFreq, value ~ L1, value.var = "count")
            geoFreqMat = t(geoFreqMat)
            geoFreqMat[geoFreqMat>1]=1
            
            
            plotTileHeatmap <- function(abdMat, taxo) {
              
              abdMatNorm = t(scale(t(abdMat)))
              hr =  heatmap.2(abdMatNorm )
              dev.off()
              hrRows= rownames(abdMatNorm)[hr$rowInd]
              hrCols= colnames(abdMatNorm)[hr$colInd]
              colNews <- c(colorRampPalette(c("#e7f0fa", "#c9e2f6", "#95cbee", "#0099dc", "#4ab04a", "#ffd73e"))(10),
                           colorRampPalette(c("#eec73a", "#e29421", "#e29421", "#f05336","#ce472e"), bias=2)(90))
              
              
              quantile_range <- quantile(abdMatNorm, probs = seq(0, 1, 0.2))
              color_palette <- colorRampPalette(c("#3794bf", "#FFFFFF", "#df8640"))(length(quantile_range) - 1)
              label_text <- rollapply(round(quantile_range, 2), width = 2, by = 1, FUN = function(i) paste(i, collapse = " : "))
              theme_change <- theme(
                legend.position = "none",
                plot.background = element_blank(),
                panel.grid.minor = element_blank(),
                panel.grid.major = element_blank(),
                panel.background = element_blank(),
                panel.border = element_blank(),
                axis.line = element_blank(),
                axis.ticks = element_blank(),
                axis.text.x = element_blank(),
                axis.text.y = element_blank(),
                axis.title.x = element_blank(),
                axis.title.y = element_blank()
              )
              mod_mat <- matrix(findInterval(abdMatNorm, quantile_range, all.inside = TRUE), nrow = nrow(abdMatNorm))
              rownames(mod_mat) = rownames(abdMatNorm)
              colnames(mod_mat) = colnames(abdMatNorm)
              
              mod_tab = melt(mod_mat)
              mod_tab = mod_tab[mod_tab$Var1 %in%hrRows[!grepl("unclassified", getSpeciesName(hrRows, taxo))], ]
              mod_tab$Var1 = factor(mod_tab$Var1, levels = hrRows[!grepl("unclassified", getSpeciesName(hrRows, taxo))])
              mod_tab$Var2 = factor(mod_tab$Var2, levels = hrCols)
              
              
              p <- ggplot(mod_tab, aes(x = Var2, y = Var1, fill = factor(value))) +
                geom_tile(color = "black") +
                scale_fill_manual(values = color_palette, name = "", labels = label_text) +
                theme_change
              plot(p)
              return(list(matrix=mod_mat, table=mod_tab))
              
            }
            
            geoSpecificAbdMat = avgMgsGeoMat[specificMgsAll,]
            modOut = plotTileHeatmap(geoSpecificAbdMat, taxo)
            
            geoAbdMat = avgMgsGeoMat[rownames(geoFreqMat),]
            modOut = plotTileHeatmap(geoAbdMat, taxo)
            
            plotTileHeatmapWoSorting <- function(abdMat, taxo) {
              
              abdMatNorm = t(scale(t(abdMat)))
              quantile_range <- quantile(abdMatNorm, probs = seq(0, 1, 0.2))
              
              colNews <- c(colorRampPalette(c("#e7f0fa", "#c9e2f6", "#95cbee", "#0099dc", "#4ab04a", "#ffd73e"))(10),
                           colorRampPalette(c("#eec73a", "#e29421", "#e29421", "#f05336","#ce472e"), bias=2)(90))
              
              
              color_palette <- colorRampPalette(c("#3794bf", "#FFFFFF", "#df8640"))(length(quantile_range) - 1)
              label_text <- rollapply(round(quantile_range, 2), width = 2, by = 1, FUN = function(i) paste(i, collapse = " : "))
              theme_change <- theme(
                legend.position = "none",
                plot.background = element_blank(),
                panel.grid.minor = element_blank(),
                panel.grid.major = element_blank(),
                panel.background = element_blank(),
                panel.border = element_blank(),
                axis.line = element_blank(),
                axis.ticks = element_blank(),
                axis.text.x = element_blank(),
                axis.text.y = element_blank(),
                axis.title.x = element_blank(),
                axis.title.y = element_blank()
              )
              mod_mat <- matrix(findInterval(abdMatNorm, quantile_range, all.inside = TRUE), nrow = nrow(abdMatNorm))
              rownames(mod_mat) = rownames(abdMatNorm)
              colnames(mod_mat) = colnames(abdMatNorm)
              
              mod_tab = melt(mod_mat)
              #mod_tab = mod_tab[mod_tab$Var1 %in%rownames(abdMatNorm)[!grepl("unclassified", getSpeciesName(rownames(abdMatNorm), taxo))], ]
              mod_tab$Var1 = factor(mod_tab$Var1, levels = rownames(abdMatNorm))
              mod_tab$Var2 = factor(mod_tab$Var2, levels = colnames(abdMatNorm))
              # mod_tab$Var1 = factor(mod_tab$Var1, levels = hrRows[!grepl("unclassified", getSpeciesName(hrRows, taxo))])
              # mod_tab$Var2 = factor(mod_tab$Var2, levels = hrCols)
              # 
              
              p <- ggplot(mod_tab, aes(x = Var2, y = Var1, fill = factor(value))) +
                geom_tile(color = "black") +
                scale_fill_manual(values = color_palette, name = "", labels = label_text) 
              #+ 
              #  theme_change
              plot(p)
              return(list(matrix=mod_mat, table=mod_tab))
              
            }
            
            newOrders = c("Sweden","Denmark","Germany","Luxembourg","Spain","Italy","UK",
                          "US","China","Japan",
                          "India","Fiji","Mongo","Tanzania","Peru","Madagascar","Thailand")
            choseTopPerCountries <- apply(mgsGeoMinEsOut, 2, function(currCol) {
              ind=which.max(currCol)
              return(rownames(mgsGeoMinEsOut)[ind])
            })
            choseTopPerCountries = choseTopPerCountries[newOrders[1:10]]
            #choseTopPerCountries = choseTopPerCountries[newOrders[8:10]]
            #choseTopPerCountries = choseTopPerCountries[newOrders[11:17]]
            
            geoSpecAbdMat = avgMgsGeoMat[match(choseTopPerCountries, rownames(avgMgsGeoMat)),]
            geoSpecAbdMat = geoSpecAbdMat[,match(names(choseTopPerCountries), colnames(geoSpecAbdMat))]
            modSpecOut = plotTileHeatmapWoSorting(geoSpecAbdMat, taxo)
            
            geoAbdMat2 = t(scale(t(geoAbdMat)))
            hr =  heatmap.2(geoAbdMat2)
            dev.off()
            hrRows= rownames(geoAbdMat2)[hr$rowInd]
            hrCols= colnames(geoAbdMat2)[hr$colInd]
            
            # heatmap.2(geoAbdMat2, trace="none",margins = c(5,20),
            #           key=F, lhei=c(1,20),lwid = c(1,20),
            #           cexCol = 0.6,cexRow = 0.6,
            #           sepcolor = "black",
            #           colsep=c(0, dim(geoFreqMat)[2]),
            #           rowsep=c(0, dim(geoFreqMat)[1]),
            #           sepwidth = c(0.01,0.01),
            #           col = colorRampPalette( c("white","gray","#0000ff66"))(256))
            # 
            
            colNews <- c(colorRampPalette(c("#e7f0fa", "#c9e2f6", "#95cbee", "#0099dc", "#4ab04a", "#ffd73e"))(10),
                         colorRampPalette(c("#eec73a", "#e29421", "#e29421", "#f05336","#ce472e"), bias=2)(90))
            
            library(ggplot2)
            library(reshape2)
            library(zoo)
            quantile_range <- quantile(geoAbdMat2, probs = seq(0, 1, 0.2))
            color_palette <- colorRampPalette(c("#3794bf", "#FFFFFF", "#df8640"))(length(quantile_range) - 1)
            label_text <- rollapply(round(quantile_range, 2), width = 2, by = 1, FUN = function(i) paste(i, collapse = " : "))
            theme_change <- theme(
              #legend.position = "none",
              plot.background = element_blank(),
              panel.grid.minor = element_blank(),
              panel.grid.major = element_blank(),
              panel.background = element_blank(),
              panel.border = element_blank(),
              axis.line = element_blank(),
              axis.ticks = element_blank(),
              axis.text.x = element_blank(),
              axis.text.y = element_blank(),
              axis.title.x = element_blank(),
              axis.title.y = element_blank()
            )
            mod_mat <- matrix(findInterval(geoAbdMat2, quantile_range, all.inside = TRUE), nrow = nrow(geoAbdMat2))
            rownames(mod_mat) = rownames(geoAbdMat2)
            colnames(mod_mat) = colnames(geoAbdMat2)
            
            mod_tab = melt(mod_mat)
            mod_tab = mod_tab[mod_tab$Var1 %in%hrRows[!grepl("unclassified", getSpeciesName(hrRows, taxo))], ]
            mod_tab$Var1 = factor(mod_tab$Var1, levels = hrRows[!grepl("unclassified", getSpeciesName(hrRows, taxo))])
            mod_tab$Var2 = factor(mod_tab$Var2, levels = hrCols)
            
            hrRows[!grepl("unclassified", getSpeciesName(hrRows, taxo))]
            
            ggplot(mod_tab, aes(x = Var2, y = Var1, fill = factor(value))) +
              geom_tile(color = "black") +
              scale_fill_manual(values = color_palette, name = "", labels = label_text) +
              theme_change
            
            geoAbdMat2Tab = melt(geoAbdMat2)
            
            ggplot(geoAbdMat2Tab, aes(y=Var1, x=Var2, fill=value)) + 
              geom_tile(colour="white", linewidth=2, 
                        width=.9, height=.9) + theme_minimal() +
              scale_fill_gradientn(colours=colNews, limits=c(-2, 4),
                                   breaks=seq(-2, 4, by=1), 
                                   na.value=rgb(246, 246, 246, max=255),
                                   labels=c("-2", "1", "0", "1", "2", "3", "4"),
                                   guide=guide_colourbar(ticks=T, nbin=50,
                                                         barheight=.5, label=T, 
                                                         barwidth=10)) +
              labs(x="", y="", fill="") 
            
            
            ggtitle("Measles") +
              theme(legend.position=c(.5, -.13),
                    legend.direction="horizontal",
                    legend.text=element_text(colour="grey20"),
                    plot.margin=grid::unit(c(.5,.5,1.5,.5), "cm"),
                    axis.text.y=element_text(size=6, family="Helvetica", 
                                             hjust=1),
                    axis.text.x=element_text(size=8),
                    axis.ticks.y=element_blank(),
                    panel.grid=element_blank(),
                    title=element_text(hjust=-.07, face="bold", vjust=1, 
                                       family="Helvetica"),
                    text=element_text(family="URWHelvetica"))
            
            
            
            #plotBarcode(geoAbdMat[rownames(geoAbdMat)[hr$rowInd], colnames(geoAbdMat)[hr$colInd]])
            
            oldHeatMap = F
            if (oldHeatMap) {
              rownames(geoFreqMat) = paste(getSpeciesName(rownames(geoFreqMat),taxo),
                                           rownames(geoFreqMat), sep=", ")
              unclassifiedInds = grepl("unclassified", rownames(geoFreqMat))
              geoFreqMat=geoFreqMat[!unclassifiedInds,]
              
              
              heatmap.2(geoFreqMat, trace="none",margins = c(5,20),
                        key=F, lhei=c(1,20),lwid = c(1,20),
                        cexCol = 0.6,cexRow = 0.6,
                        sepcolor = "black",
                        colsep=c(0, dim(geoFreqMat)[2]),
                        rowsep=c(0, dim(geoFreqMat)[1]),
                        sepwidth = c(0.01,0.01),
                        col = c("#80808011","#0000ff66"))
            }
            
            
          }
          
        }
        
        
      }
      
      checkGeoBetaDiversity = F
      if (checkGeoBetaDiversity) {
        newNormUniqMat = mergeMatUpdated[,normUniqAllNew]
        newNormUniqMat = newNormUniqMat[rowMeans(newNormUniqMat)!=0,]
        newNormUniqMat = newNormUniqMat[,colMeans(newNormUniqMat)!=0]
        
        geo.dist=vegdist(t(newNormUniqMat), method="bray")
        geo.group = rep("",length(normUniqAllNew))
        names(geo.group) = normUniqAllNew
        
        if (T) {
          geo.group[normUniqCountryListNew$China] = "China"
          geo.group[normUniqCountryListNew$Denmark] = "Denmark"
          geo.group[normUniqCountryListNew$Fiji] = "Fiji"
          geo.group[normUniqCountryListNew$Germany] = "Germany"
          geo.group[normUniqCountryListNew$India] = "India"
          geo.group[normUniqCountryListNew$Italy] = "Italy"
          geo.group[normUniqCountryListNew$Japan] = "Japan"
          geo.group[normUniqCountryListNew$Luxembourg] = "Luxembourg"
          geo.group[normUniqCountryListNew$Madagascar] = "Madagascar"
          geo.group[normUniqCountryListNew$Mongo] = "Mongo"
          geo.group[normUniqCountryListNew$Peru] = "Peru"
          geo.group[normUniqCountryListNew$Spain] = "Spain"
          geo.group[normUniqCountryListNew$Sweden] = "Sweden"
          geo.group[normUniqCountryListNew$Tanzania] = "Tanzania"
          geo.group[normUniqCountryListNew$Thailand] = "Thailand"
          geo.group[normUniqCountryListNew$UK] = "UK"
          geo.group[normUniqCountryListNew$US] = "US"
        }
        geo.group = geo.group[match(colnames(newNormUniqMat), names(geo.group))]
        set.seed(1)
        dispersion.out <- betadisper(geo.dist, group=geo.group)
        
        permutest(dispersion.out)
        plot(dispersion.out, hull=FALSE, ellipse=TRUE) ##sd ellipse
        
        diversity(t(newNormUniqMat))
      }
      
      checkGeoSignalLoss= T
      if (checkGeoSignalLoss) {
        ### checking loss of geographical signals
        unique(allPairDownMeltSel$disease_cohort)
        
        checkLossGeoFractions <- function(allPairDownMelt, specificGroupMgsByCountries, diseaseWithCountryPairTab) {
          
          #### down ####
          diseaseWithCountryPairTab2 = diseaseWithCountryPairTab[diseaseWithCountryPairTab$reference!="T1D_FI_id21",]
          checkDownEffectSizes = apply(diseaseWithCountryPairTab2, 1, function(currRow){
            refCohort = currRow[1]
            targetGeo = currRow[2]
            if (targetGeo == "") return(NULL)
            
            targetGeoMsps = specificGroupMgsByCountries[[targetGeo]]
            refPairDownMeltCountry = allPairDownMelt[allPairDownMelt$control=="country" & allPairDownMelt$disease_cohort== refCohort,]
            refPairDownMeltSigMsps = refPairDownMeltCountry$msp[refPairDownMeltCountry$effect_size>=0.3]
            tt=table(targetGeoMsps %in% refPairDownMeltSigMsps)
            tt2=table(refPairDownMeltSigMsps %in% targetGeoMsps)
            #print(tt)
            pariedEffectSizes = refPairDownMeltCountry$effect_size[match(targetGeoMsps,refPairDownMeltCountry$msp)]
            if (length(pariedEffectSizes)==0) return(NA)
            
            ksOut = ks.test(refPairDownMeltCountry$effect_size,
                            pariedEffectSizes, 
                            alternative = "greater")
            print(refCohort)
            print(c(tt["TRUE"]/length(targetGeoMsps),
                    tt2["TRUE"] /length(refPairDownMeltSigMsps),
                    ksOut$p.value))
            
            
            
            #return(ksOut$p.value)
            
            
            
            
            return(sum(targetGeoMsps %in% refPairDownMeltSigMsps)/length(targetGeoMsps))
          })
          names(checkDownEffectSizes) = diseaseWithCountryPairTab2[,1]
          return(checkDownEffectSizes)
          
          
        }
        checkLossGeoFractions2 <- function(allPairDownMelt, specificGroupMgsByCountries, diseaseWithCountryPairTab) {
          
          #### down ####
          diseaseWithCountryPairTab2 = diseaseWithCountryPairTab[diseaseWithCountryPairTab$reference!="T1D_FI_id21",]
          checkDownEffectSizes = apply(diseaseWithCountryPairTab2, 1, function(currRow){
            refCohort = currRow[1]
            targetGeo = currRow[2]
            if (targetGeo == "") return(NULL)
            
            targetGeoMsps = specificGroupMgsByCountries[[targetGeo]]
            refPairDownMeltCountry = allPairDownMelt[allPairDownMelt$control=="country" & allPairDownMelt$disease_cohort== refCohort,]
            refPairDownMeltSigMsps = refPairDownMeltCountry$msp[refPairDownMeltCountry$effect_size>=0.3]
            tt=table(targetGeoMsps %in% refPairDownMeltSigMsps)
            tt2=table(refPairDownMeltSigMsps %in% targetGeoMsps)
            #print(tt)
            pariedEffectSizes = refPairDownMeltCountry$effect_size[match(targetGeoMsps,refPairDownMeltCountry$msp)]
            if (length(pariedEffectSizes)==0) return(NA)
            
            ksOut = ks.test(refPairDownMeltCountry$effect_size,
                            pariedEffectSizes, 
                            alternative = "greater")
            print(refCohort)
            print(c(tt["TRUE"]/length(targetGeoMsps),
                    tt2["TRUE"] /length(refPairDownMeltSigMsps),
                    ksOut$p.value))
            
            
            
            #return(ksOut$p.value)
            
            
            
            
            return(sum(refPairDownMeltSigMsps %in% targetGeoMsps)/length(refPairDownMeltSigMsps))
          })
          names(checkDownEffectSizes) = diseaseWithCountryPairTab2[,1]
          return(checkDownEffectSizes)
          
          
        }
        
        lossGeoFraction = checkLossGeoFractions(allPairDownMeltImputed, specificGroupMgsByCountries, diseaseWithCountryPairTab)
        lossGeoFraction = unlist(lossGeoFraction)
        print(sort(lossGeoFraction))
        
        lossGeoFraction2 = checkLossGeoFractions2(allPairDownMeltImputed, specificGroupMgsByCountries, diseaseWithCountryPairTab)
        lossGeoFraction2 = unlist(lossGeoFraction2)
        print(sort(lossGeoFraction2))
        
        checkLossGeoSignals <- function(allPairDownMelt, specificGroupMgsByCountries, diseaseWithCountryPairTab) {
          
          #### down ####
          diseaseWithCountryPairTab2 = diseaseWithCountryPairTab[diseaseWithCountryPairTab$reference!="T1D_FI_id21",]
          checkDownEffectSizes = apply(diseaseWithCountryPairTab2, 1, function(currRow){
            print(currRow)
            refCohort = currRow[1]
            targetGeo = currRow[2]
            if (targetGeo == "") return(NULL)
            
            targetGeoMsps = specificGroupMgsByCountries[[targetGeo]]
            refPairDownMeltCountry = allPairDownMelt[allPairDownMelt$control=="country" & allPairDownMelt$disease_cohort== refCohort,]
            pariedEffectSizes = refPairDownMeltCountry$effect_size[match(targetGeoMsps,refPairDownMeltCountry$msp)]
            
            ksOut = ks.test(refPairDownMeltCountry$effect_size,
                            pariedEffectSizes, 
                            alternative = "greater")
            
            return(pariedEffectSizes)
          })
          names(checkDownEffectSizes) = diseaseWithCountryPairTab2[,1]
          
          if (F) {
            par(mar=c(5,12,4,1))
            boxplot(checkDownEffectSizes,horizontal=T,las=2, cex=0.5)
            par(mar=c(5,4,4,1))
            
            lapply(checkDownEffectSizes, median) 
            
          }
          
          #### up ####
          checkUpEffectSizes = apply(diseaseWithCountryPairTab2, 1, function(currRow){
            refCohort = currRow[1]
            targetGeo = currRow[2]
            if (targetGeo == "") return(NULL)
            
            targetGeoMsps = specificGroupMgsByCountries[[targetGeo]]
            refPairMeltCountry = allPairMelt[allPairMelt$control=="country" & allPairMelt$disease_cohort== refCohort,]
            pariedEffectSizes = refPairMeltCountry$effect_size[match(targetGeoMsps,refPairMeltCountry$msp)]
            return(pariedEffectSizes)
          })
          names(checkUpEffectSizes) = diseaseWithCountryPairTab2[,1]
          
          if (F) {
            par(mar=c(5,15,4,1))
            names(checkUpEffectSizes) = rep("",length(checkUpEffectSizes))
            boxplot(checkUpEffectSizes,horizontal=T,las=2, cex=0.5)
            par(mar=c(5,4,4,1))
            
            par(mfrow=c(1,3))
            par(mar=c(5,4,4,1))
            plot.new()
            par(mar=c(5,4,4,1))
            boxplot(checkDownEffectSizes,horizontal=T,las=2, cex=0.5)
            abline(v=0, col="red")
            par(mar=c(5,4,4,1))
            names(checkUpEffectSizes) = rep("",length(checkUpEffectSizes))
            boxplot(checkUpEffectSizes,horizontal=T,las=2, cex=0.5)
            abline(v=0, col="blue")
            
          }
          
          sigDownEffectSizes = apply(diseaseWithCountryPairTab2, 1, function(currRow){
            refCohort = currRow[1]
            targetGeo = currRow[2]
            if (targetGeo == "") return(NULL)
            
            targetGeoMsps = specificGroupMgsByCountries[[targetGeo]]
            refPairDownMeltCountry = allPairDownMelt[allPairDownMelt$control=="country" & allPairDownMelt$disease_cohort== refCohort,]
            pariedEffectSizes = refPairDownMeltCountry$effect_size[match(targetGeoMsps,refPairDownMeltCountry$msp)]
            if (length(pariedEffectSizes)==0) return(NA)
            
            ksOut = ks.test(refPairDownMeltCountry$effect_size,
                            pariedEffectSizes, 
                            alternative = "greater")
            
            return(ksOut$p.value)
          })
          names(sigDownEffectSizes) = diseaseWithCountryPairTab2[,1]
          sigDownEffectSizes = unlist(sigDownEffectSizes)
          sigUpEffectSizes = apply(diseaseWithCountryPairTab2, 1, function(currRow){
            refCohort = currRow[1]
            targetGeo = currRow[2]
            if (targetGeo == "") return(NULL)
            
            targetGeoMsps = specificGroupMgsByCountries[[targetGeo]]
            refPairMeltCountry = allPairMelt[allPairMelt$control=="country" & allPairMelt$disease_cohort== refCohort,]
            pariedEffectSizes = refPairMeltCountry$effect_size[match(targetGeoMsps,refPairMeltCountry$msp)]
            if (length(pariedEffectSizes)==0) return(NA)
            ksOut = ks.test(refPairMeltCountry$effect_size,
                            pariedEffectSizes, 
                            alternative = "greater")
            
            return(ksOut$p.value)
          })
          names(sigUpEffectSizes) = diseaseWithCountryPairTab2[,1]
          sigUpEffectSizes = unlist(sigUpEffectSizes)
          
          if(F) {
            plot(sigDownEffectSizes, sigUpEffectSizes, log="xy")
            abline(h=0.05, v=0.05)
            
            par(mar=c(15,4,4,1))
            barplot(sort(-log10(sigDownEffectSizes),decreasing = T), las=2, cex.names = 0.7)
            abline(h=-log10(0.05))
            
            par(mar=c(15,4,4,1))
            barplot(sort(-log10(sigUpEffectSizes),decreasing = T), las=2, cex.names = 0.7)
            abline(h=-log10(0.05))
            
            
          }
          return(list(up=sigUpEffectSizes,
                      down=sigDownEffectSizes))
          
        }
        geoSignalCheck=checkLossGeoSignals(allPairDownMeltImputed, specificGroupMgsByCountries, )
        geoSignalUp = data.frame(name=names(geoSignalCheck$up), 
                                 pvalue=geoSignalCheck$up, 
                                 logp=-log10(geoSignalCheck$up))
        geoSignalUp = geoSignalUp[order(geoSignalUp$pvalue),]
        geoSignalDown = data.frame(name=names(geoSignalCheck$down), 
                                   pvalue=geoSignalCheck$down, 
                                   logp=-log10(geoSignalCheck$down))
        geoSignalDown = geoSignalDown[order(geoSignalDown$pvalue),]
        geoSignalDown$name = factor(geoSignalDown$name, levels=geoSignalDown$name)
        geoSignalUp$name = factor(geoSignalUp$name, levels=geoSignalUp$name)
        geoSignalDown=geoSignalDown[geoSignalDown$pvalue<=0.01,]
        geoSignalUp=geoSignalUp[geoSignalUp$pvalue<=0.01,]
        
        blueCol = rgb(5/255,48/255,97/255,0.5)
        ggplot(geoSignalDown, aes(x=name, y=logp, fill=logp)) +
          geom_bar(stat="identity", colour="black") +
          xlab("")+ylab("log10 P-value")+
          scale_fill_gradient2(low="white",high=blueCol) +
          theme(panel.grid.major = element_blank(), 
                panel.grid.minor = element_blank(),
                axis.text.x = element_text(angle = 60, hjust=1),
                legend.position = "none",
                panel.background = element_rect(fill = "white", 
                                                colour = "black", 
                                                size = 0.5, 
                                                linetype = "solid")) 
        
        ggplot(geoSignalUp, aes(x=name, y=logp, fill=logp)) +
          geom_bar(stat="identity", colour="black") +
          xlab("")+ylab("log10 P-value")+
          scale_fill_gradient2(low="white",high=blueCol) +
          theme(panel.grid.major = element_blank(), 
                panel.grid.minor = element_blank(),
                axis.text.x = element_text(angle = 60, hjust=1),
                legend.position = "none",
                panel.background = element_rect(fill = "white", 
                                                colour = "black", 
                                                size = 0.5, 
                                                linetype = "solid")) 
        
        aggregatedGeoSignDown = data.frame(geoSignalDown[,c("name", "logp")],
                                           type="pvalue",
                                           stringsAsFactors = F)
        colnames(aggregatedGeoSignDown) = c("name","score","type")
        
        aggregatedGeoSignDown2 = data.frame(name=names(lossGeoFraction2),
                                            score=lossGeoFraction2*100,
                                            type="fraction",
                                            stringsAsFactors = F)
        aggregatedGeoSignDown2 = aggregatedGeoSignDown2[aggregatedGeoSignDown2$name %in% geoSignalDown$name, ]
        
        aggregatedGeoSignDown = rbind(aggregatedGeoSignDown, 
                                      aggregatedGeoSignDown2 )
        aggregatedGeoSignDown$score[is.nan(aggregatedGeoSignDown$score)] = 0
        aggregatedGeoSignDownMod = aggregatedGeoSignDown
        aggregatedGeoSignDownMod$score[aggregatedGeoSignDownMod$type == "fraction"] = (-1)*aggregatedGeoSignDownMod$score[aggregatedGeoSignDownMod$type == "fraction"]
        
        ggplot(aggregatedGeoSignDownMod, aes(x=name, y=score, fill="type" )) + 
          geom_bar(stat = "identity") + 
          geom_hline(yintercept =0) + 
          theme(axis.text.x = element_text(angle = 60, hjust=1),
                panel.grid.major = element_blank(), 
                panel.grid.minor = element_blank(),
                panel.background = element_rect(fill = "white", 
                                                colour = "black", 
                                                size = 0.5, 
                                                linetype = "solid"))
        
        if (F) {
          target="Atherosclerosis_SW_id1"
          table(allPairDownMeltSel$msp[allPairDownMeltSel$disease_cohort==target] %in% specificGroupMgsByCountries$Sweden)
          table(allPairMeltSel$msp[allPairMeltSel$disease_cohort==target] %in% specificGroupMgsByCountries$Sweden)
        }
        
        
        
        
      }
      
      if (F) {
        geoGroupSignals = apply(mgsGeoEs02Out, 1, function(currRow){
          
          cuts = 6:16
          trues = sapply(cuts, function(eachCut){
            currSum = sum(currRow>=eachCut)
            finalSum = eachCut + currSum
            if (finalSum >= 15) return(T)
            else return(F)
          })
          if (any(trues)) {
            finalCut = min(cuts[trues])
            finalAreas = colnames(mgsGeoEs02Out)[currRow>=finalCut]
            if (length(finalAreas)<2) return(NULL)
            return(finalAreas)
          }
          
          else return(NULL)
          
        })
        geoGroupSignals <- Filter(Negate(is.null), geoGroupSignals)
        length(geoGroupSignals)
        
        geoGroupMgss = names(geoGroupSignals)
        
        geoGroupSignalTab = melt(geoGroupSignals)
        geoGroupSignalTab$value = as.character(geoGroupSignalTab$value)
        geoGroupSignalTab$L1 = as.character(geoGroupSignalTab$L1)
        geoGroupSignalTab$count = 1
        geoGroupSignalTab = geoGroupSignalTab[!geoGroupSignalTab$L1 %in% geoSpecificMgs, ]
        
        geoGroupSignalsBy = split(geoGroupSignalTab$L1, geoGroupSignalTab$value)
        geoGroupSignalCounts = do.call(c,lapply(geoGroupSignalsBy, length))
        geoGroupSignalCounts = geoGroupSignalCounts[geoGroupSignalCounts!=0]
        par(mar=c(8,4,4,1))
        geoSignalCols = c("black", "red", "gold", "green")
        barplot(sort(geoGroupSignalCounts,decreasing = T),
                col = getColsOfCountry(names(sort(geoGroupSignalCounts,decreasing = T))), 
                las=2)
        
        
        
        
        geoGroupSignalMat = acast(geoGroupSignalTab, value ~ L1, value.var = "count")
        geoGroupSignalMat[is.na(geoGroupSignalMat)]=0
        heatmap.2(t(geoGroupSignalMat), trace="none", margins = c(7,7))
        
        barplot(sort(avgMgsGeoMat["msp_0141",],decreasing = T),las=2)
        
        if (F) {
          cutNumCountries = 6
          
          tests2 = apply(mgsGeoMaxEsOut, 1, function(currRow){
            currArea = colnames(mgsGeoMaxEsOut)[currRow>=0.2]
            return(length(currArea))
            
          })
          
          # geoGroupSignals = apply(mgsGeoEs03Out, 2, function(currCol){
          #   currMgs = rownames(mgsGeoMinEsOut)[currCol>=cutNumCountries & currCol < 17]
          #   currSpecies = gssOssTabNew$species[match(currMgs, gssOssTabNew$MSP)]
          #   currSpeciesMgs = paste(currSpecies, currMgs, sep=", ")
          #   return(currSpeciesMgs)
          # })
          
          tests = apply(mgsGeoEs03Out, 1, function(currRow){
            return(sum(currRow>=cutNumCountries ))
            #return(NULL)
          })
          sort(tests, decreasing = T)[1:10]
          
          geoGroupSignalsBy = apply(mgsGeoEs03Out, 1, function(currRow){
            currAreas = colnames(mgsGeoMinEsOut)[currRow>=cutNumCountries & currRow < 17]
            if (length(currAreas)>1) return(currAreas)
            #return(NULL)
          })
          geoGroupSignalsBy = geoGroupSignalsBy[!do.call(c,lapply(geoGroupSignalsBy, function(x)any(is.null(x))))]
          
          lapply(geoGroupSignalsBy, function(currMgs){
            
          })
          
          
          geoGroupSignalCounts = do.call(c,lapply(geoGroupSignals, length))
          geoGroupSignalCounts = geoGroupSignalCounts[geoGroupSignalCounts!=0]
          geoGroupSignalsBy$msp_0172
          par(mar=c(8,4,4,1))
          barplot(sort(geoGroupSignalCounts,decreasing = T),las=2)
          
        }
        
        geoSignalMgs = apply(mgsGeoMinEsOut, 2, function(currCol){
          #currMgs = rownames(mgsGeoMinEsOut)[currCol>=0.2]
          currMgs = rownames(mgsGeoMinEsOut)[currCol>=0.3]
          return(currMgs)
        })
        
        
        geoSignals = apply(mgsGeoMinEsOut, 2, function(currCol){
          currMgs = rownames(mgsGeoMinEsOut)[currCol>=0.2]
          currSpecies = gssOssTabNew$species[match(currMgs, gssOssTabNew$MSP)]
          currSpeciesMgs = paste(currSpecies, currMgs, sep=", ")
          return(currSpeciesMgs)
        })
        geoSpecificMgs = do.call(c, geoSignalMgs)
        
        geoSignalCounts = do.call(c,lapply(geoSignals, length))
        geoSignalCounts = geoSignalCounts[geoSignalCounts!=0]
        par(mar=c(8,4,4,1))
        geoSignalCols = c("black", "red", "gold", "green")
        barplot(sort(geoSignalCounts,decreasing = T),col = geoSignalCols, las=2)
        
        boxplot(list(specific=meanNormMgs[geoSpecificMgs],
                     group=meanNormMgs[geoGroupMgss], 
                     others=meanNormMgs[names(meanNormMgs) %in% c(geoSpecificMgs, geoGroupMgss)]), log="y" )
        
        
        table(geoGroupMgss %in% geoSpecificMgs)
        
        
        barplot(sort(avgMgsGeoMat["msp_0263",],decreasing = T),las=2)
        barplot(sort(avgMgsGeoMat["msp_0220",],decreasing = T),las=2)
        barplot(sort(avgMgsGeoMat["msp_0136",],decreasing = T),las=2)
        barplot(sort(avgMgsGeoMat["msp_0005",],decreasing = T),las=2)
        barplot(sort(avgMgsGeoMat["msp_0478",],decreasing = T),
                col=getColsOfCountry(names(sort(avgMgsGeoMat["msp_0478",],decreasing = T))),
                las=2)
        
        out1=showPerBarPlotByCountry(normUniqCountryList, "msp_0069", mergeMatUpdated)
        out2=showBoxPlotsByCountry(normUniqCountryList, "msp_0069", mergeMatUpdated)
        
        
        
        
        if (F) {
          geoSignalTab = melt(geoSignals)
          mgsGeoMinEsDigit = (mgsGeoMinEsOut>=0.2)*1
          mgsGeoMinEsDigit = mgsGeoMinEsDigit[,colSums(mgsGeoMinEsDigit)!=0]
          mgsGeoMinEsDigit = mgsGeoMinEsDigit[rowSums(mgsGeoMinEsDigit)!=0,]
          
          
          heatmap.2(mgsGeoMinEsDigit, trace="none",margins = c(5,5),
                    key=F,lhei=c(1,20),
                    cexCol = 0.6,cexRow = 0.6,
                    sepcolor = "black",
                    colsep=c(0, dim(mgsGeoMinEsDigit)[2]),
                    rowsep=c(0, dim(mgsGeoMinEsDigit)[1]),
                    sepwidth = c(0.01,0.01),
                    
                    col = colorRampPalette(c("white","blue"))(256))
          
          
          mgsGeoEs03Out = checkBestFeaturesInES(normUniqCountryList, mergeMatUpdated, esCut=0.3)
          
          targetOut = mgsGeoEs03Out
          maxCut=10
          esCutMat = targetOut[rowMaxs(targetOut)>=maxCut,]
          esCutMat[esCutMat<maxCut]=0
          esCutMat = esCutMat[, colSums(esCutMat)!=0]
          heatmap.2(esCutMat, trace="none",margins = c(5,18),
                    key=F,lhei=c(1,20),
                    cexCol = 0.6,cexRow = 0.6,
                    col = colorRampPalette(c("white","blue", "purple"))(256))
          #sepcolor = "black",
          #colsep=c(0,2,7,dim(esCutMat)[2]),
          #rowsep=c(0,5,20,47,dim(esCutMat)[1]),
          #sepwidth = c(0.01,0.01),
          
          
          mgsGeoEs05Out = checkBestFeaturesInES(normUniqCountryList, mergeMatUpdated, esCut=0.5)
          mgsGeoEs08Out = checkBestFeaturesInES(normUniqCountryList, mergeMatUpdated, esCut=0.8)
          View(mgsGeoEs08Out)
          View(t(mgsGeoEs05Out))
          mgsEs03Out = checkBestFeaturesInES(mixedUniqSubjectList, mergeMatUpdated, esCut=0.3)
          mgsEs08Out = checkBestFeaturesInES(mixedUniqSubjectList, mergeMatUpdated, esCut=0.8)
          
        }
        
      }
      
    }
    
    pieMgsMode = T
    if (pieMgsMode) {
      
      geoSpecificCol=rgb(56/255,108/255,176/255,0.5)
      groupEnrichedCol=rgb(117/255,112/255,179/255,0.5)
      hAbdCol=rgb(240/255,2/255,127/255,0.5)
      lAbdCol=rgb(230/255,171/255,2/255,0.5)
      mixedCol= rgb(102/255,102/255,102/255,0.5)
      
      meanNormMgs = rowMeans(normMgsMat)
      
      normHalfQ = quantile(meanNormMgs, 0.5)
      norm99Q = quantile(meanNormMgs, 0.99)
      
      lowNormMgs = names(meanNormMgs[meanNormMgs <= normHalfQ])
      highNormMgs = names(meanNormMgs[meanNormMgs >= norm99Q])
      
      table(lowNormMgs %in% geoSpecificMgs)
      table(lowNormMgs %in% geoGroupMgss)
      
      table(highNormMgs %in% geoSpecificMgs)
      table(highNormMgs %in% geoGroupMgss)
      
      table(geoGroupMgss %in% geoSpecificMgs)
      table(geoGroupMgss %in% c(geoSpecificMgs, lowNormMgs, highNormMgs))
      
      
      lowNormMgs = lowNormMgs[!lowNormMgs %in% c(geoSpecificMgs, geoGroupMgss)]
      highNormMgs = highNormMgs[!highNormMgs %in% c(geoSpecificMgs, geoGroupMgss)]
      
      
      
      mgsStats = c(179, 192, 14, 907, 685)
      names(mgsStats) = c("geography-specific", "group-enriched", "highly-abundant", "lowly-abundant", "mixed")
      
      mgsStatCols = c(geoSpecificCol, groupEnrichedCol, 
                      hAbdCol, lAbdCol, mixedCol)
      iniR = 0.2
      pie(1, radius=0.01, init.angle=90, col=c('white'), border = NA, labels='')
      
      floating.pie(0,0,mgsStats,  
                   radius=5*iniR, 
                   startpos=pi/2, 
                   col = mgsStatCols,
                   border="black")
      floating.pie(0,0,1,  
                   radius=4*iniR, 
                   startpos=pi/2, 
                   col="white",
                   border="black")
      floating.pie(0,0,1,  
                   radius=3.99*iniR, 
                   startpos=pi/2, 
                   col="white",
                   border=NA)
      
    }
    
    
    
  }
  
  checkInOutFlowMode = T
  if (checkInOutFlowMode) {
    
    industSpecies = unlist(specificGroupMgsByCountries[industrializedCountries])
    nonWesternSpecies = unlist(specificGroupMgsByCountries[traditionalCountries])
    
    inflowList = list(Sweden=inflow[specificGroupMgsByCountries$Sweden],
                      Denmark=inflow[specificGroupMgsByCountries$Denmark],
                      Spain=inflow[specificGroupMgsByCountries$Spain],
                      Westernized=inflow[industSpecies],
                      NonWesternized=inflow[nonWesternSpecies])
    
    outflowList = list(Sweden=outflow[specificGroupMgsByCountries$Sweden],
                       Denmark=outflow[specificGroupMgsByCountries$Denmark],
                       Spain=outflow[specificGroupMgsByCountries$Spain],
                       Westernized=outflow[industSpecies],
                       NonWesternized=outflow[nonWesternSpecies])
    
    inflowList_5m = list(Sweden=inflow_5m[specificGroupMgsByCountries$Sweden],
                         Denmark=inflow_5m[specificGroupMgsByCountries$Denmark],
                         Spain=inflow_5m[specificGroupMgsByCountries$Spain],
                         Westernized=inflow_5m[industSpecies],
                         NonWesternized=inflow_5m[nonWesternSpecies])
    
    outflowList_5m = list(Sweden=outflow_5m[specificGroupMgsByCountries$Sweden],
                          Denmark=outflow_5m[specificGroupMgsByCountries$Denmark],
                          Spain=outflow_5m[specificGroupMgsByCountries$Spain],
                          Westernized=outflow_5m[industSpecies],
                          NonWesternized=outflow_5m[nonWesternSpecies])
    
    inflowList_15m = list(Sweden=inflow_15m[specificGroupMgsByCountries$Sweden],
                          Denmark=inflow_15m[specificGroupMgsByCountries$Denmark],
                          Spain=inflow_15m[specificGroupMgsByCountries$Spain],
                          Westernized=inflow_15m[industSpecies],
                          NonWesternized=inflow_15m[nonWesternSpecies])
    
    outflowList_15m = list(Sweden=outflow_15m[specificGroupMgsByCountries$Sweden],
                           Denmark=outflow_15m[specificGroupMgsByCountries$Denmark],
                           Spain=outflow_15m[specificGroupMgsByCountries$Spain],
                           Westernized=outflow_15m[industSpecies],
                           NonWesternized=outflow_15m[nonWesternSpecies])
    
    
    survRates = survivalStats$surv.rates
    names(survRates) = rownames(survivalStats)
    retainingProbList = list(Sweden=survRates[specificGroupMgsByCountries$Sweden],
                             Denmark=survRates[specificGroupMgsByCountries$Denmark],
                             Spain=survRates[specificGroupMgsByCountries$Spain],
                             Westernized=survRates[industSpecies],
                             NonWesternized=survRates[nonWesternSpecies])
    
    wilcox.test(inflowList$Sweden, inflowList$Denmark, alternative = "greater")
    wilcox.test(inflowList$Sweden, inflowList$Spain, alternative = "greater")
    wilcox.test(inflowList$Sweden, inflowList$Westernized, alternative = "greater")
    wilcox.test(inflowList$Sweden, inflowList$NonWesternized, alternative = "greater")
    
    wilcox.test(outflowList$Sweden, outflowList$Denmark, alternative = "less")
    wilcox.test(outflowList$Sweden, outflowList$Spain, alternative = "less")
    wilcox.test(outflowList$Sweden, outflowList$Westernized, alternative = "less")
    wilcox.test(outflowList$Sweden, outflowList$NonWesternized, alternative = "less")
    
    wilcox.test(retainingProbList$Sweden, retainingProbList$Denmark, alternative = "greater")
    wilcox.test(retainingProbList$Sweden, retainingProbList$Spain, alternative = "greater")
    wilcox.test(retainingProbList$Sweden, retainingProbList$Westernized, alternative = "greater")
    wilcox.test(retainingProbList$Sweden, retainingProbList$NonWesternized, alternative = "greater")
    
    
    wilcox.test(inflowList_5m$Sweden, inflowList_5m$Denmark, alternative = "greater")
    wilcox.test(inflowList_5m$Sweden, inflowList_5m$Spain, alternative = "greater")
    wilcox.test(inflowList_5m$Sweden, inflowList_5m$Westernized, alternative = "greater")
    wilcox.test(inflowList_5m$Sweden, inflowList_5m$NonWesternized, alternative = "greater")
    
    wilcox.test(outflowList_5m$Sweden, outflowList_5m$Denmark, alternative = "less")
    wilcox.test(outflowList_5m$Sweden, outflowList_5m$Spain, alternative = "less")
    wilcox.test(outflowList_5m$Sweden, outflowList_5m$Westernized, alternative = "less")
    wilcox.test(outflowList_5m$Sweden, outflowList_5m$NonWesternized, alternative = "less")
    
    wilcox.test(inflowList_15m$Sweden, inflowList_15m$Denmark, alternative = "greater")
    wilcox.test(inflowList_15m$Sweden, inflowList_15m$Spain, alternative = "greater")
    wilcox.test(inflowList_15m$Sweden, inflowList_15m$Westernized, alternative = "greater")
    wilcox.test(inflowList_15m$Sweden, inflowList_15m$NonWesternized, alternative = "greater")
    
    wilcox.test(outflowList_15m$Sweden, outflowList_15m$Denmark, alternative = "less")
    wilcox.test(outflowList_15m$Sweden, outflowList_15m$Spain, alternative = "less")
    wilcox.test(outflowList_15m$Sweden, outflowList_15m$Westernized, alternative = "less")
    wilcox.test(outflowList_15m$Sweden, outflowList_15m$NonWesternized, alternative = "less")
    
    
    
    
    inflowMelt = melt(inflowList)
    inflowMelt$L1 = factor(inflowMelt$L1, levels = c("Sweden","Denmark","Spain","Westernized","NonWesternized"))
    
    inflowMelt2 = inflowMelt[inflowMelt$L1 %in% c("Sweden","Denmark","Spain"),]
    #inflow.part1.for.swedish.wellness.pdf
    
    inflowMelt3 = inflowMelt[inflowMelt$L1 %in% c("Westernized","NonWesternized"),]
    #inflow.part1.for.swedish.wellness.pdf
    
    inGG = ggplot(inflowMelt3, aes(x=L1, y=value, fill=L1)) + geom_boxplot() + 
      #scale_fill_manual(values = c(blueCol,  eurCol, eurCol, eurCol, tradCol)) +
      scale_fill_manual(values = c(eurCol, tradCol)) +
      
      xlab("") + ylab("Inflow") +
      
      theme(panel.grid.major = element_blank(), 
            panel.grid.minor = element_blank(),
            axis.text.x = element_text(angle = 60, hjust=1),
            #axis.text.y = element_blank(),
            axis.ticks = element_blank(),
            legend.position = "none",
            panel.background = element_rect(fill = "white", 
                                            colour = "black", 
                                            size = 0.5, 
                                            linetype = "solid")) 
    plot(inGG)
    inGG2 = ggplot(inflowMelt2, aes(x=L1, y=value, fill=L1)) + geom_boxplot() + 
      scale_fill_manual(values = c(blueCol,  eurCol, eurCol, eurCol, tradCol)) +
      
      xlab("") + ylab("") +
      
      theme(panel.grid.major = element_blank(), 
            panel.grid.minor = element_blank(),
            axis.text.x = element_blank(),
            axis.text.y = element_blank(),
            axis.ticks.x = element_blank(),
            legend.position = "none",
            panel.background = element_rect(fill = "white", 
                                            colour = "black", 
                                            size = 0.5, 
                                            linetype = "solid")) 
    
    inGG3 = ggplot(inflowMelt3, aes(x=L1, y=value, fill=L1)) + geom_boxplot() + 
      #scale_fill_manual(values = c(blueCol,  eurCol, eurCol, eurCol, tradCol)) +
      scale_fill_manual(values = c(eurCol, tradCol)) +
      
      xlab("") + ylab("") +
      
      theme(panel.grid.major = element_blank(), 
            panel.grid.minor = element_blank(),
            axis.text.x =  element_blank(),
            axis.text.y = element_blank(),
            axis.ticks = element_blank(),
            legend.position = "none",
            panel.background = element_rect(fill = "white", 
                                            colour = "black", 
                                            size = 0.5, 
                                            linetype = "solid")) 
    
    ggarrange(inGG2, inGG3, widths = c(2.5,2))
    
    outflowMelt = melt(outflowList)
    outflowMelt$L1 = factor(outflowMelt$L1, levels = c("Sweden","Denmark","Spain","Westernized","NonWesternized"))
    
    outflowMelt2 = outflowMelt[outflowMelt$L1 %in% c("Sweden","Denmark","Spain"),]
    #outflow.part1.for.swedish.wellness.pdf
    
    outflowMelt3 = outflowMelt[outflowMelt$L1 %in% c("Westernized","NonWesternized"),]
    #outflow.part2.for.swedish.wellness.pdf
    
    outGG = ggplot(outflowMelt3, aes(x=L1, y=value, fill=L1)) + geom_boxplot() + 
      #scale_fill_manual(values = c(blueCol,  eurCol, eurCol, eurCol, tradCol)) +
      scale_fill_manual(values = c(eurCol, tradCol)) +
      xlab("") + ylab("Outflow") +
      theme(panel.grid.major = element_blank(), 
            panel.grid.minor = element_blank(),
            axis.text.x = element_text(angle = 60, hjust=1),
            axis.ticks = element_blank(),
            legend.position = "none",
            panel.background = element_rect(fill = "white", 
                                            colour = "black", 
                                            size = 0.5, 
                                            linetype = "solid"))
    print(outGG)
    #ggarrange(inGG, outGG)
    
    outGG2 = ggplot(outflowMelt2, aes(x=L1, y=value, fill=L1)) + geom_boxplot() + 
      scale_fill_manual(values = c(blueCol,  eurCol, eurCol, eurCol, tradCol)) +
      
      xlab("") + ylab("") +
      
      theme(panel.grid.major = element_blank(), 
            panel.grid.minor = element_blank(),
            axis.text.x = element_blank(),
            axis.text.y = element_blank(),
            axis.ticks.x = element_blank(),
            legend.position = "none",
            panel.background = element_rect(fill = "white", 
                                            colour = "black", 
                                            size = 0.5, 
                                            linetype = "solid")) 
    
    outGG3 = ggplot(outflowMelt3, aes(x=L1, y=value, fill=L1)) + geom_boxplot() + 
      #scale_fill_manual(values = c(blueCol,  eurCol, eurCol, eurCol, tradCol)) +
      scale_fill_manual(values = c(eurCol, tradCol)) +
      
      xlab("") + ylab("") +
      
      theme(panel.grid.major = element_blank(), 
            panel.grid.minor = element_blank(),
            axis.text.x =  element_blank(),
            axis.text.y = element_blank(),
            axis.ticks = element_blank(),
            legend.position = "none",
            panel.background = element_rect(fill = "white", 
                                            colour = "black", 
                                            size = 0.5, 
                                            linetype = "solid")) 
    
    ggarrange(outGG2, outGG3, widths = c(2.5,2))
    
    par(mar=c(5,10,4,1))
    boxplot(rev(inflowList), horizontal = T,las=1)
    boxplot(rev(outflowList), horizontal = T,las=1)
    t.test(inflowList$Sweden, inflowList$Industrial)
    t.test(outflowList$Sweden, outflowList$Industrial)
    
    
    
  }
  
  genusMode = F
  if (genusMode) {
    genusGeoMinEsOut = checkMinEffectSizeOfFeatures(normUniqCountryList, genusMatUpdated)
    genusGeoMaxEsOut = checkMaxEffectSizeOfFeatures(normUniqCountryList, genusMatUpdated)
    genusGeoEs02Out = checkBestFeaturesInES(normUniqCountryList, genusMatUpdated, esCut=0.2)
    
    
    geoGenusSignals = apply(genusGeoMinEsOut, 2, function(currCol){
      currGenus = rownames(genusGeoMinEsOut)[currCol>=0.2]
      return(currGenus)
    })
    
    geoSpecificGenera = do.call(c,geoGenusSignals)
    
    geoGenusSignalCounts = do.call(c,lapply(geoGenusSignals, length))
    geoGenusSignalCounts = geoGenusSignalCounts[geoGenusSignalCounts!=0]
    par(mar=c(8,4,4,1))
    geoGenusSignalCols = c("black", "red", "gold", "blue", "blue","gold")
    barplot(sort(geoGenusSignalCounts,decreasing = T),col=geoGenusSignalCols, las=2)
    barplot(sort(avgGenusGeoMat["Sellimonas",],decreasing = T),col=getColsOfCountry(names(sort(avgGenusGeoMat["Holdemanella",],decreasing = T))),las=2)
    
    getColsOfCountry <- function(countryNames) {
      countryCols = c("gold", "gold", "blue", "red", "blue", "blue", "blue", "blue",
                      "blue", "blue", "blue", "blue", "green", "black", "gold", "gold", "red", 
                      "black", "gold")
      names(countryCols) = names(cohortIdByGeos)
      return(countryCols[match(countryNames, names(countryCols))])
    }
    
    pie(table(countryCols), col = names(table(countryCols)), 
        labels = c("Africa(2)", "Europe(9)", "Asia(5)", "Oceania(1)", "America(2)"))
    barplot(sort(avgGenusGeoMat["msp_0263",],decreasing = T),las=2)
    barplot(sort(avgGenusGeoMat["msp_0220",],decreasing = T),las=2)
    barplot(sort(avgGenusGeoMat["msp_0136",],decreasing = T),las=2)
    barplot(sort(avgGenusGeoMat["msp_1568",],decreasing = T),las=2)
    
    geoGroupGenusSignals = apply(genusGeoEs02Out, 1, function(currRow){
      
      cuts = 6:16
      trues = sapply(cuts, function(eachCut){
        currSum = sum(currRow>=eachCut)
        finalSum = eachCut + currSum
        if (finalSum >= 15) return(T)
        else return(F)
      })
      if (any(trues)) {
        finalCut = min(cuts[trues])
        finalAreas = colnames(genusGeoEs02Out)[currRow>=finalCut]
        if (length(finalAreas)<2) return(NULL)
        return(finalAreas)
      }
      
      else return(NULL)
      
    })
    geoGroupGenusSignals <- Filter(Negate(is.null), geoGroupGenusSignals)
    
    geoGroupGenera = names(geoGroupGenusSignals)
    table(geoGroupGenera %in% geoSpecificGenera)
    
    geoGroupGenusSignalTab = melt(geoGroupGenusSignals)
    geoGroupGenusSignalTab$value = as.character(geoGroupGenusSignalTab$value)
    geoGroupGenusSignalTab$L1 = as.character(geoGroupGenusSignalTab$L1)
    geoGroupGenusSignalTab$count = 1
    geoGroupGenusSignalTab = geoGroupGenusSignalTab[!geoGroupGenusSignalTab$L1 %in% geoSpecificGenera,  ]
    
    
    
    ###
    geoGroupGenusSignalsBy = split(geoGroupGenusSignalTab$L1, geoGroupGenusSignalTab$value)
    geoGroupGenusSignalCounts = do.call(c,lapply(geoGroupGenusSignalsBy, length))
    geoGroupGenusSignalCounts = geoGroupGenusSignalCounts[geoGroupGenusSignalCounts!=0]
    
    par(mar=c(8,5,4,1))
    geoSignalCols = c("black", "red", "gold", "green")
    barplot(sort(geoGroupGenusSignalCounts,decreasing = T),
            col = getColsOfCountry(names(sort(geoGroupGenusSignalCounts,decreasing = T))), 
            las=2)
    
    
    ###
    geoGroupGenusSignalMat = acast(geoGroupGenusSignalTab, value ~ L1, value.var = "count")
    geoGroupGenusSignalMat[is.na(geoGroupGenusSignalMat)]=0
    heatmap.2(t(geoGroupGenusSignalMat), trace="none", margins = c(7,12),key=F,
              col=colorRampPalette(c("white","blue")),
              lhei=c(1,20),
              cexCol = 0.6,cexRow = 0.6,
              sepcolor = "gray",
              colsep=seq(0,dim(geoGroupGenusSignalMat)[1]),
              rowsep=seq(0,dim(geoGroupGenusSignalMat)[2]),
              sepwidth = c(0.01,0.01))
    
    barplot(sort(avgGenusGeoMat["Collinsella",],decreasing = T),
            col=getColsOfCountry(names(sort(avgGenusGeoMat["Collinsella",],decreasing = T))),las=2)
    
    
    pieGenusMode = T
    if (pieGenusMode) {
      
      geoSpecificCol=rgb(56/255,108/255,176/255,0.5)
      groupEnrichedCol=rgb(117/255,112/255,179/255,0.5)
      hAbdCol=rgb(240/255,2/255,127/255,0.5)
      lAbdCol=rgb(230/255,171/255,2/255,0.5)
      mixedCol= rgb(102/255,102/255,102/255,0.5)
      
      
      meanNormGenus = rowMeans(normGenusMat)
      
      normHalfQ = quantile(meanNormGenus, 0.5)
      norm99Q = quantile(meanNormGenus, 0.99)
      
      lowNormGenus = names(meanNormGenus[meanNormGenus <= normHalfQ])
      highNormGenus = names(meanNormGenus[meanNormGenus >= norm99Q])
      
      table(lowNormGenus %in% geoSpecificGenera)
      table(lowNormGenus %in% geoGroupGenera)
      
      table(highNormGenus %in% geoSpecificGenera)
      table(highNormGenus %in% geoGroupGenera)
      
      table(geoGroupGenera %in% geoSpecificGenera)
      table(geoGroupGenera %in% c(geoSpecificGenera, lowNormGenus, highNormGenus))
      
      
      lowNormGenus = lowNormGenus[!lowNormMgs %in% c(geoSpecificGenera, geoGroupGenera)]
      highNormGenus = highNormGenus[!highNormGenus %in% c(geoSpecificGenera, geoGroupGenera)]
      
      genusStats = c(18, 38, 4, 158, 98)
      names(genusStats) = c("geography-specific", "group-enriched", "highly-abundant", "lowly-abundant", "mixed")
      
      genusStatCols = c(geoSpecificCol, groupEnrichedCol, 
                        hAbdCol, lAbdCol, mixedCol)
      iniR = 0.2
      pie(1, radius=0.01, init.angle=90, col=c('white'), border = NA, labels='')
      
      floating.pie(0,0,genusStats,  
                   radius=5*iniR, 
                   startpos=pi/2, 
                   col = genusStatCols,
                   border="black")
      floating.pie(0,0,1,  
                   radius=4*iniR, 
                   startpos=pi/2, 
                   col="white",
                   border="black")
      floating.pie(0,0,1,  
                   radius=3.99*iniR, 
                   startpos=pi/2, 
                   col="white",
                   border=NA)
    }
    
    if (F) {
      genusGeoMinEsOut = checkMinEffectSizeOfFeatures(normUniqCountryList, genusMatUpdated)
      genusGeoEs02Out = checkBestFeaturesInES(normUniqCountryList, genusMatUpdated, esCut=0.2)
      genusGeoEs03Out = checkBestFeaturesInES(normUniqCountryList, genusMatUpdated, esCut=0.3)
      genusGeoEs05Out = checkBestFeaturesInES(normUniqCountryList, genusMatUpdated, esCut=0.5)
      genusGeoEs08Out = checkBestFeaturesInES(normUniqCountryList, genusMatUpdated, esCut=0.8)
      
    }   
    
    if (F) {
      
      
      genusNormalEs03Out = checkBestFeaturesInES(normUniqSubjectList, genusMatUpdated, esCut=0.3)
      genusNormalEs05Out = checkBestFeaturesInES(normUniqSubjectList, genusMatUpdated, esCut=0.5)
      genusNormalEs08Out = checkBestFeaturesInES(normUniqSubjectList, genusMatUpdated, esCut=0.8)
      
      targetOut = mgsGeoEs02Out
      maxCut=12
      esCutMat = targetOut[rowMaxs(targetOut)>=maxCut,]
      #esCutMat[esCutMat<maxCut]=0
      esCutMat = esCutMat[, colSums(esCutMat)!=0]
      heatmap.2(esCutMat, trace="none",margins = c(5,18),
                key=F,lhei=c(1,20),
                cexCol = 0.6,cexRow = 0.6,
                sepcolor = "black",
                colsep=c(0,2,6,dim(esCutMat)[2]),
                rowsep=c(0,dim(esCutMat)[1]),
                sepwidth = c(0.01,0.01),
                col = colorRampPalette(c("white","blue", "purple"))(256))
      
    }
  }
  
  if (F) {
    freqs = rowSums(mergeMatUpdated > 0)
    avgs = rowMeans(mergeMatUpdated)
    
    cols = rep("#A9A9A944", length(freqs))
    names(cols) = names(freqs)
    cols[rownames(gssOssTabNew[gssOssTabNew$type_sh=="oss",])] = "#FF000088"
    cols[rownames(gssOssTabNew[gssOssTabNew$type_sh=="gss",])] = "#0000FF55"
    
    
    cols = rep("#A9A9A944", length(freqs))
    names(cols) = names(freqs)
    cols[diseaseOnlySpecies] = "#FF000088"
    
    
    plot(avgs, freqs, col=cols, pch=16, log='x')
    abline(v=quantile(avgs, 0.1))
    abline(v=quantile(avgs, 0.5))
    #abline(v=quantile(avgs, 0.9))
    abline(v=quantile(avgs, 0.99))
    tt=data.frame(avgs, freqs)
    View(tt)
    View(tt[rownames(gssOssTabNew[gssOssTabNew$type_sh=="oss",]),])
    
    lowerMgs = names(avgs[avgs <= quantile(avgs, 0.5)])
    lowerMgsPhyla = table(gssOssTabNew$phylum[gssOssTabNew$MSP %in% lowerMgs])
    989-557-141-98-69
    pieLowerMgsPhyla = c(557,141,98,69,124)
    names(pieLowerMgsPhyla) = c("Firmicutes","Bacteroidetes","Tenericutes","Proteobaceria","Others")
    pieLowerCols = c("#BEAED4", "#7FC97F","#B15928", "#666666",  "#F3F3F3")
    pie(pieLowerMgsPhyla, col = pieLowerCols, labels = c("","","","",""))
    
    higherMgs = names(avgs[avgs >= quantile(avgs, 0.99)])
    higherMgsPhyla = table(gssOssTabNew$phylum[gssOssTabNew$MSP %in% higherMgs])
    pieHigherMgsPhyla = c(11,8,1)
    names(pieHigherMgsPhyla) = c("Bacteroidetes","Firmicutes","Proteobaceria")
    pieHigherCols = c("#7FC97F", "#BEAED4", "#666666")
    pie(pieHigherMgsPhyla, col = pieHigherCols, labels = c("","",""))
    
    rgb(190/255,174/255,212/255)
    rgb(127/255,201/255,127/255)
    rgb(177/255,89/255,40/255)
    rgb(102/255,102/255,102/255)
    rgb(243/255,243/255,243/255)
    
    
    
    
    freqs = rowSums(mergeMatUpdated > 0)
    meds = rowMedians(mergeMatUpdated)
    plot(meds, freqs)
    View(data.frame(meds, freqs))
    
    
    freqs = rowSums(genusMatUpdated > 0)
    avgs = rowMeans(genusMatUpdated)
    plot(avgs, freqs, log='x')
    View(data.frame(avgs, freqs))
    
    
    freqs = rowSums(familyMatUpdated > 0)
    avgs = rowMeans(familyMatUpdated)
    plot(avgs, freqs, log='x')
    View(data.frame(avgs, freqs))
    genusEs08Out = checkBestFeaturesInES(mixedUniqSubjectList, genusMatUpdated, esCut=0.8)
    genusES08RData = "C://Data//comparative.analysis.healthy.sweden//genusEs08Out.RData"
    
    getwd()
    save(genusEs08Out, file=genusES08RData)
    load(genusES08RData)
    
    mergeMatUpdatedImp = mergeMatUpdated
    mergeMatUpdatedImp[mergeMatUpdatedImp==0] <- 1e-12
    mergeMatUpdatedImpLog = log10(mergeMatUpdatedImp)
    
    
  }
  
  avgMode = T
  if (avgMode) {
    avgMgsGeoMat = getAvgAbdMatBy(mergeMatUpdated, normUniqCountryList)
    avgGenusGeoMat = getAvgAbdMatBy(genusMatUpdated, normUniqCountryList)
    
    
    avgMgsDsMat = getAvgAbdMatBy(mergeMatUpdated, mixedUniqSubjectList)
    corMgsDsMat = cor(avgMgsDsMat)
    
    avgGenusDsMat = getAvgAbdMatBy(genusMatUpdated, mixedUniqSubjectList)
    corGenusDsMat = cor(avgGenusDsMat)
    
    
    heatmap.2(corMgsDsMat, 
              trace="none",margins = c(8,8),cexCol = 0.6,cexRow = 0.6,
              col = colorRampPalette(c("white","blue"))(256))
    
    
    avgGenusDsMat[avgGenusDsMat==0] <- 1e-12
    avgGenusDsMatLog = log10(avgGenusDsMat)
    heatmap.2(t(avgGenusDsMatLog[c("Sphaerochaeta","Blautia","Mitsuokella","Butyrivibrio"),]), 
              trace="none",margins = c(5,10),cexCol = 0.8,cexRow = 0.8,
              col = colorRampPalette(c("white","blue"))(256))
    
    
    require(momr)
    plotBarcode(avgGenusDsMat[c("Sphaerochaeta","Veillonella","Mitsuokella","Butyrivibrio"),])  
    
  }
  
}

### very important ###
trajectoryMode = T
if (trajectoryMode) {
  
  wholeDataTrajectoryRData = "j://Deposit/Project/2018_microbiome_atlas/trajectoryData/whole.trajectory.cds.RData"
  normDataTrajectoryRData = "j://Deposit/Project/2018_microbiome_atlas/trajectoryData/norm.trajectory.cds.RData"
  normDataTrajectory2RData = "j://Deposit/Project/2018_microbiome_atlas/trajectoryData/norm.trajectory2.cds.RData"
  exceptEurDataTrajectoryRData = "j://Deposit/Project/2018_microbiome_atlas/trajectoryData/except.eur.trajectory.cds.RData"
  exceptCjpDataTrajectoryRData = "j://Deposit/Project/2018_microbiome_atlas/trajectoryData/except.cjp.trajectory.cds.RData"
  exceptTradDataTrajectoryRData = "j://Deposit/Project/2018_microbiome_atlas/trajectoryData/except.trad.trajectory.cds.RData"
  
  prunedDataTrajectoryRData = "j://Deposit/Project/2018_microbiome_atlas/trajectoryData/pruned.trajectory.cds.RData"
  
  loadData = T
  if (loadData) {
    newNormalMetadata = basicMetaMapUpdated[basicMetaMapUpdated$sample.ID %in% newNormalAllIds, ]
    newNormalSamplesByGeo = split(newNormalMetadata$sample.ID, newNormalMetadata$Geography)
    newNormalSampleCountsByGeo = do.call(c, lapply(newNormalSamplesByGeo, length))
    
    cjpSamples = unlist(newNormalSamplesByGeo[c("China","Japan", "US")])
    eurSamples = unlist(newNormalSamplesByGeo[europeanCountries])
    eurFiSamples = unlist(newNormalSamplesByGeo[c(europeanCountries, "Finland")])
    trdSamples = unlist(newNormalSamplesByGeo[c(traditionalCountries,"Thai")])
  }
  
  normDataMode2 = F
  if (normDataMode2) {
    
    
    
    generateMode = T
    if (generateMode) {
      
      allNormMode = T
      if (allNormMode) {
        normAllMat = mergeMatUpdated*10^9
        normAllMat = normAllMat[, colnames(normAllMat) %in% newNormalAllIds_wo_FI]
        #min(normAllMat[normAllMat!=0]) --> 3.8194
        
        trajectorySampleTab = data.frame(Library=basicMetaMapUpdated$sample.ID,
                                         Well=basicMetaMapUpdated$dataset.ID,
                                         Hours=0,
                                         Media="NONE")
        
        rownames(trajectorySampleTab) = basicMetaMapUpdated$sample.ID
        trajectorySampleTab = trajectorySampleTab[rownames(trajectorySampleTab) %in% newNormalAllIds_wo_FI,]
        
        trajectoryTaxaTab = data.frame(gene_short_name = taxo$species, 
                                       biotype="protein_coding",
                                       num_cells_expressed = 1,
                                       use_for_ordering=FALSE)
        
        rownames(trajectoryTaxaTab) = rownames(taxo)
        trajectoryTaxaTab = trajectoryTaxaTab[match(rownames(normAllMat), rownames(taxo)),]
        
        pd <- new("AnnotatedDataFrame", data = trajectorySampleTab)
        fd <- new("AnnotatedDataFrame", data = trajectoryTaxaTab)
        
        cds <- newCellDataSet(normAllMat, 
                              phenoData = pd, 
                              featureData = fd,
                              lowerDetectionLimit = 0.1)
        
        cds <- estimateSizeFactors(cds)
        cds <- estimateDispersions(cds)
        
        cds <- reduceDimension(cds, 
                               max_components = 2,
                               method = 'DDRTree')
        
        cds <- orderCells(cds)
        
        save(cds, file=normDataTrajectory2RData)
      }
      
      
      
      
    }
    
    allNormMode  = T
    if (allNormMode) {
      getPercentGeoByState <- function(targetSamples, newNormalSamplesByGeo, newNormalSampleCountsByGeo) {
        
        
        currCounts = do.call(c, lapply(newNormalSamplesByGeo, function(currSamples) {
          counts = sum(currSamples %in% targetSamples )
          return(counts)
        }))
        
        return(currCounts*100/newNormalSampleCountsByGeo)
      }
      
      firstBranchSamples = unlist(samplesByState[c("1","2","9","8")])
      midBranchSamples = unlist(samplesByState[c("3")])
      lastBranchSamples = unlist(samplesByState[c("4","5","6","7")])
      
      
    }
  }
  
  normDataMode = T
  if (normDataMode) {
    
    pcaTsneMode = F
    if (pcaTsneMode) {
      
      loadData = T
      if (loadData) {
        newNormalMetadata = basicMetaMapUpdated[basicMetaMapUpdated$sample.ID %in% newNormalAllIds, ]
        newNormalSamplesByGeo = split(newNormalMetadata$sample.ID, newNormalMetadata$Geography)
        newNormalSampleCountsByGeo = do.call(c, lapply(newNormalSamplesByGeo, length))
        
        cjpSamples = unlist(newNormalSamplesByGeo[c("China","Japan", "US")])
        eurSamples = unlist(newNormalSamplesByGeo[europeanCountries])
        trdSamples = unlist(newNormalSamplesByGeo[traditionalCountries])
        
        length(cjpSamples)
        length(eurSamples)
        length(trdSamples)
        
        normAllMat = mergeMatUpdated*10^9
        normAllMat = normAllMat[, colnames(normAllMat) %in% newNormalAllIds]
      }
      
      dimReductionMode = T
      if (dimReductionMode) {
        normPcaOut = prcomp(t(normAllMat))
        eigs <- normPcaOut$sdev^2
        cumsum(eigs[1:20]) / sum(eigs)
        normPcaMat = normPcaOut$x[,1:11]
      }
      
      tsneMode = T
      if (tsneMode) {
        set.seed(1)
        tsneOut = Rtsne(normPcaMat, dims = 2, perplexity = 100)
        normTsneMat = tsneOut$Y
        rownames(normTsneMat) = rownames(normPcaMat)
        
        # tsneOut = Rtsne(t(normAllMat), dims = 2, perplexity = 50)
        # normTsneMat = tsneOut$Y
        # rownames(normTsneMat) = colnames(normAllMat)
        
        tsneCols = rep("gray",dim(normTsneMat)[1])
        names(tsneCols) = rownames(normTsneMat)
        tsneCols[cjpSamples] = cjpCol
        tsneCols[eurSamples] = eurCol
        tsneCols[trdSamples] = tradCol
        
        plot(normTsneMat, col=tsneCols, pch=16,xlab="tSNE1",ylab="tSNE2", xaxt="n",yaxt="n" )
      }
      
      
      
      
    }
    
    generateMode = F
    if (generateMode) {
      normAllMat = mergeMatUpdated*10^9
      normAllMat = normAllMat[, colnames(normAllMat) %in% newNormalAllIds]
      
      trajectorySampleTab = data.frame(Library=basicMetaMapUpdated$sample.ID,
                                       Well=basicMetaMapUpdated$dataset.ID,
                                       Hours=0,
                                       Media="NONE")
      
      rownames(trajectorySampleTab) = basicMetaMapUpdated$sample.ID
      trajectorySampleTab = trajectorySampleTab[rownames(trajectorySampleTab) %in% newNormalAllIds,]
      
      trajectoryTaxaTab = data.frame(gene_short_name = taxo$species, 
                                     biotype="protein_coding",
                                     num_cells_expressed = 1,
                                     use_for_ordering=FALSE)
      
      rownames(trajectoryTaxaTab) = rownames(taxo)
      trajectoryTaxaTab = trajectoryTaxaTab[match(rownames(normAllMat), rownames(taxo)),]
      
      pd <- new("AnnotatedDataFrame", data = trajectorySampleTab)
      fd <- new("AnnotatedDataFrame", data = trajectoryTaxaTab)
      
      cds <- newCellDataSet(normAllMat, 
                            phenoData = pd, 
                            featureData = fd,
                            lowerDetectionLimit = 0.1)
      
      cds <- estimateSizeFactors(cds)
      cds <- estimateDispersions(cds)
      
      cds <- reduceDimension(cds, 
                             max_components = 2,
                             method = 'DDRTree')
      
      cds <- orderCells(cds)
      
      save(cds, file=normDataTrajectoryRData)
      
    }
    
    loadMode = T
    if (loadMode) load(normDataTrajectoryRData)
    #save(cds, file=normDataTrajectoryRData)
    
    checkTrajectory = T
    if (checkTrajectory) {
      
      mergeMatUpdatedScaled = t(scale(t(mergeMatUpdatedScaled)))
      mergeMatInflowScaled  = colSums(mergeMatUpdatedScaled[rownames(mergeMatUpdatedScaled)%in%inflowSpecies,])
      mergeMatOutflowScaled  = colSums(mergeMatUpdatedScaled[rownames(mergeMatUpdatedScaled)%in%outflowSpecies,])
      
      basicMetaMapUpdated$inflowScaled = mergeMatInflowScaled[match(basicMetaMapUpdated$sample.ID , names(mergeMatInflowScaled))]
      basicMetaMapUpdated$outflowScaled = mergeMatOutflowScaled[match(basicMetaMapUpdated$sample.ID , names(mergeMatOutflowScaled))]
      
      inflowByEnterotype = split(basicMetaMapUpdated$inflowScaled, basicMetaMapUpdated$enteroType)
      outflowByEnterotype = split(basicMetaMapUpdated$outflowScaled, basicMetaMapUpdated$enteroType)
      
      par(mar=c(10,4,4,1))
      boxplot(inflowByEnterotype, las=2)
      boxplot(outflowByEnterotype, las=2)
      
      
      getTrajectoryData <- function(cds, newNormalMetadata) {
        requireNamespace("igraph")
        gene_short_name <- NA
        sample_name <- NA
        sample_state <- pData(cds)$State
        data_dim_1 <- NA
        data_dim_2 <- NA
        lib_info_with_pseudo <- pData(cds)
        x = 1; y = 2
        data_df <- t(monocle::reducedDimS(cds)) %>% as.data.frame() %>% 
          select_(data_dim_1 = x, data_dim_2 = y) %>% rownames_to_column("sample_name") %>% 
          mutate(sample_state) %>% left_join(lib_info_with_pseudo %>% 
                                               rownames_to_column("sample_name"), by = "sample_name")
        output = data_df[,c("data_dim_1","data_dim_2")]
        rownames(output) = data_df$sample_name
        output$geo = newNormalMetadata$Geography[match(rownames(output),newNormalMetadata$sample.ID)]
        return(output)
      }
      getAvgPointsByGeo <- function(trajectoryData, newNormalSamplesByGeo, plotMode = T, labelMode = T) {
        newNormalSamplesByGeo = newNormalSamplesByGeo[names(newNormalSamplesByGeo) !="Finland"]
        geoPoints = do.call(rbind,lapply(newNormalSamplesByGeo, function(currSamples){
          currData = trajectoryData[rownames(trajectoryData) %in% currSamples,]
          return(colMeans(currData[,1:2]))
        }))
        geoSds = do.call(c,lapply(newNormalSamplesByGeo, function(currSamples){
          currData = trajectoryData[rownames(trajectoryData) %in% currSamples,]
          print(length(currSamples))
          #print(dim(currData))
          #print(colSds(currData))
          currSds = c(sd(currData[,1]),sd(currData[,2]))
          currSqSds = sqrt(sum(currSds^2))*2
          return(currSqSds)
        }))
        geoPointTab = data.frame(geoPoints, geo=rownames(geoPoints), sd=geoSds)
        if (plotMode) {
          cols = c("Tanzania"=tradCol,"Peru"=tradCol,"Madagascar"=tradCol,"Mongo"=tradCol,"Thai"=tradCol,"Fiji"=tradCol, "India"=tradCol,"Finland"=eurCol,"China"=cjpCol,"Japan"=cjpCol,"US"=cjpCol,"Italy"=eurCol,"Spain"=eurCol,"Luxembourg"=eurCol,"Germany"=eurCol,"UK"=eurCol,"Denmark"=eurCol,"Sweden"=eurCol)
          
          if (labelMode) {
            p = ggplot(geoPointTab, aes(x=data_dim_1, y=data_dim_2, color=factor(geo), label=geo, size=sd)) + 
              geom_point() + 
              ggrepel::geom_text_repel(size=3.5, color="black") +
              scale_color_manual(values=cols) + 
              xlab("Component 1")+ylab("Component 2")+ 
              theme(panel.background = element_blank(),
                    axis.ticks = element_blank(),
                    axis.text.x = element_blank(),
                    strip.text.x = element_blank(),
                    axis.text.y = element_blank(),
                    strip.text.y = element_blank(),
                    legend.position="none", 
                    panel.border = element_rect(colour = "black", fill=NA, size=.5))
            print(p)
          }
          if (!labelMode) {
            p = ggplot(geoPointTab, aes(x=data_dim_1, y=data_dim_2, color=factor(geo), label=geo, size=sd)) + 
              geom_point() + 
              scale_color_manual(values=cols) + 
              xlab("Component 1")+ylab("Component 2")+ 
              theme(panel.background = element_blank(),
                    axis.ticks = element_blank(),
                    axis.text.x = element_blank(),
                    strip.text.x = element_blank(),
                    axis.text.y = element_blank(),
                    strip.text.y = element_blank(),
                    legend.position="none", 
                    panel.border = element_rect(colour = "black", fill=NA, size=.5))
            print(p)
          }
          
          
        }
        return(geoPointTab)
      }
      
      
      
      cols = c("Tanzania"=tradCol,"Peru"=tradCol,"Madagascar"=tradCol,"Mongo"=tradCol,"Thai"=tradCol,"Fiji"=tradCol, "India"=tradCol,"Finland"=eurCol,"China"=cjpCol,"Japan"=cjpCol,"US"=cjpCol,"Italy"=eurCol,"Spain"=eurCol,"Luxembourg"=eurCol,"Germany"=eurCol,"UK"=eurCol,"Denmark"=eurCol,"Sweden"=eurCol)
      
      trajectoryData = getTrajectoryData(cds, newNormalMetadata)
      
      getAvgPointsByGeo(trajectoryData, newNormalSamplesByGeo, T, F)
      
      trajectoryData2 = getTrajectoryData(cds, basicMetaMapUpdated)
      
      trajectoryData$Age = basicMetaMapUpdated$Age[match(rownames(trajectoryData), basicMetaMapUpdated$sample.ID)]
      trajectoryData$BMI = basicMetaMapUpdated$BMI[match(rownames(trajectoryData), basicMetaMapUpdated$sample.ID)]
      trajectoryData$Sequencer = basicMetaMapUpdated$Sequencer[match(rownames(trajectoryData), basicMetaMapUpdated$sample.ID)]
      trajectoryData$Gender = basicMetaMapUpdated$Gender[match(rownames(trajectoryData), basicMetaMapUpdated$sample.ID)]
      trajectoryData$Enterotype = basicMetaMapUpdated$enteroType[match(rownames(trajectoryData), basicMetaMapUpdated$sample.ID)]
      trajectoryData$inflow = mergeMatInflowScaled[match(rownames(trajectoryData), names(mergeMatInflowScaled))]
      trajectoryData$outflow = mergeMatOutflowScaled[match(rownames(trajectoryData), names(mergeMatOutflowScaled))]
      trajectoryData$richness = basicMetaMapUpdated$GeneRichness[match(rownames(trajectoryData), basicMetaMapUpdated$sample.ID)]
      
      
      ggplot(trajectoryData, aes(x=data_dim_1, y=data_dim_2, color=factor(geo), label=geo)) + 
        geom_point() + 
        scale_color_manual(values=cols) + 
        xlab("Component 1")+ylab("Component 2")+ 
        theme(panel.background = element_blank(),
              axis.ticks = element_blank(),
              axis.text.x = element_blank(),
              strip.text.x = element_blank(),
              axis.text.y = element_blank(),
              strip.text.y = element_blank(),
              legend.position="none", 
              panel.border = element_rect(colour = "black", fill=NA, size=.5))
      
      
      #trajectory.inflow.mapping
      ggplot(trajectoryData[!is.na(trajectoryData$richness),], aes(x=data_dim_1, y=data_dim_2, color=richness)) + 
        geom_point(alpha=0.5) + scale_colour_gradient2(low="blue",mid = "gray",high = "red", midpoint=6e5)+
        xlab("Component 1")+ylab("Component 2") + theme_bw()
      
      
      #trajectory.inflow.mapping
      ggplot(trajectoryData[!is.na(trajectoryData$inflow),], aes(x=data_dim_1, y=data_dim_2, color=inflow)) + 
        geom_point(alpha=0.5) + scale_colour_gradient2(low="blue",mid = "gray",high = "red", midpoint=20)+
        xlab("Component 1")+ylab("Component 2") + theme_bw()
      
      
      #trajectory.outflow.mapping
      ggplot(trajectoryData[!is.na(trajectoryData$outflow),], aes(x=data_dim_1, y=data_dim_2, color=outflow)) + 
        geom_point(alpha=0.5) + scale_colour_gradient2(low="blue",mid = "gray",high = "red", midpoint=100)+
        xlab("Component 1")+ylab("Component 2") + theme_bw()
      
      #trajectory.age.mapping
      ggplot(trajectoryData[!is.na(trajectoryData$Age),], aes(x=data_dim_1, y=data_dim_2, color=Age)) + 
        geom_point(alpha=0.5) + scale_colour_gradient2(low="blue",mid = "gray",high = "red", midpoint=50)+
        xlab("Component 1")+ylab("Component 2") + theme_bw()
      
      #trajectory.bmi.mapping
      ggplot(trajectoryData[!is.na(trajectoryData$BMI),], aes(x=data_dim_1, y=data_dim_2, color=BMI)) + 
        geom_point(alpha=0.5) + scale_colour_gradient2(low="blue",mid = "gray",high = "red", midpoint=20)+
        xlab("Component 1")+ylab("Component 2") + theme_bw()
      
      #trajectory.sequencer.mapping
      ggplot(trajectoryData[!is.na(trajectoryData$Sequencer),], aes(x=data_dim_1, y=data_dim_2, color=factor(Sequencer))) + 
        geom_point(alpha=0.3)+ scale_color_brewer(palette="Dark2") +
        xlab("Component 1")+ylab("Component 2") + theme_bw()
      
      
      #trajectory.sequencer.mapping
      ggplot(trajectoryData[!is.na(trajectoryData$Sequencer),], aes(x=data_dim_1, y=data_dim_2, color=factor(Sequencer))) + 
        geom_point(alpha=0.3)+ scale_color_brewer(palette="Dark2") +
        xlab("Component 1")+ylab("Component 2") + theme_bw()
      
      #trajectory.gender.mapping
      ggplot(trajectoryData[!is.na(trajectoryData$Gender),], aes(x=data_dim_1, y=data_dim_2, color=factor(Gender))) + 
        geom_point(alpha=0.3)+ scale_color_brewer(palette="Dark2") +
        xlab("Component 1")+ylab("Component 2") + theme_bw()
      
      #trajectory.enterotype.mapping
      ggplot(trajectoryData[!is.na(trajectoryData$Enterotype),], aes(x=data_dim_1, y=data_dim_2, color=factor(Enterotype))) + 
        geom_point(alpha=0.3)+ scale_color_brewer(palette="Dark2") +
        xlab("Component 1")+ylab("Component 2") + theme_bw()
      
      
      
      plot_cell_trajectory(cds, color_by = "Pseudotime")
      plot_cell_trajectory(cds, color_by = "State") +
        scale_color_manual(breaks = c("1", "2", "3","4","5"), 
                           values=c(tradCol, cjpCol, cjpCol, cjpCol, eurCol)) + 
        xlab("Component 1")+ylab("Component 2")+
        theme(axis.text.x=element_blank(), #legend.position = "none",
              
              axis.ticks.x=element_blank(),
              axis.text.y=element_blank(),
              axis.ticks.y=element_blank())
      
      
      brachInfo = pData(cds)
      samplesByState = split(as.character(brachInfo$Library), brachInfo$State)
      
      targetData = newNormalAllList$Finland_id21
      lapply(samplesByState, function(currSamples)any(currSamples %in% targetData))
      
      targetState = 5
      currStateInfo = (basicMetaMapUpdated[ basicMetaMapUpdated$sample.ID %in% rownames(brachInfo[brachInfo$State==targetState,]),])
      mean(currStateInfo$GeneRichness)
      sort(table(currStateInfo$Geography))
      
      newNormalMetadata = basicMetaMapUpdated[basicMetaMapUpdated$sample.ID %in% newNormalAllIds, ]
      newNormalSamplesByGeo = split(newNormalMetadata$sample.ID, newNormalMetadata$Geography)
      newNormalSampleCountsByGeo = do.call(c, lapply(newNormalSamplesByGeo, length))
      
      getPercentGeoByState <- function(targetSamples, newNormalSamplesByGeo, newNormalSampleCountsByGeo) {
        
        
        currCounts = do.call(c, lapply(newNormalSamplesByGeo, function(currSamples) {
          counts = sum(currSamples %in% targetSamples )
          return(counts)
        }))
        
        return(currCounts*100/newNormalSampleCountsByGeo)
      }
      
      firstBranchSamples = samplesByState[["1"]]
      midBranchSamples = unlist(samplesByState[c("2","3","4")])
      lastBranchSamples = samplesByState[["5"]]
      
      
      View(basicMetaMapUpdated[basicMetaMapUpdated$sample.ID %in% firstBranchSamples, ])
      View(basicMetaMapUpdated[basicMetaMapUpdated$sample.ID %in% midBranchSamples, ])
      
      
      firstBranchStat = getPercentGeoByState(firstBranchSamples,
                                             newNormalSamplesByGeo,
                                             newNormalSampleCountsByGeo)
      
      midBranchStat = getPercentGeoByState(midBranchSamples,#unlist(samplesByState[c("2","3","4")])
                                           newNormalSamplesByGeo,
                                           newNormalSampleCountsByGeo)
      
      lastBranchStat = getPercentGeoByState(lastBranchSamples,#samplesByState[["5"]]
                                            newNormalSamplesByGeo,
                                            newNormalSampleCountsByGeo)
      sort(getPercentGeoByState(samplesByState[["7"]],
                                newNormalSamplesByGeo,
                                newNormalSampleCountsByGeo))
      
      
      ordered0 = c("Tanzania","Madagascar","Peru","India","Fiji","Mongo","Thai",
                   "Japan","China","US", "Finland",
                   "Sweden","Denmark","UK","Spain","Germany","Luxembourg","Italy")
      
      ordered = c("Tanzania","Madagascar","Peru","India","Fiji","Mongo","Thai",
                  "Japan","China","US", 
                  "Sweden","Denmark","UK","Spain","Germany","Luxembourg","Italy")
      
      
      tradVec = c(255,127,0,120)/255
      tradCol = rgb(tradVec[1],tradVec[2],tradVec[3],tradVec[4])
      
      cjpVec = c(211,211,211,120)/255
      cjpCol = rgb(cjpVec[1],cjpVec[2],cjpVec[3],cjpVec[4])
      
      eurVec = c(117,112,179,120)/255
      eurCol = rgb(eurVec[1],eurVec[2],eurVec[3],eurVec[4])
      
      
      orderedCols = c(rep(tradCol, 7),
                      rep(cjpCol, 3), #4
                      rep(eurCol, 7))
      par(mfrow=c(4,1))
      par(mar=c(1,4,1,1))
      barplot(firstBranchStat[ordered], col=orderedCols, las=2, axisnames = F)
      par(mar=c(1,4,1,1))
      barplot(midBranchStat[ordered], col=orderedCols, las=2, axisnames = F)
      par(mar=c(1,4,1,1))
      barplot(lastBranchStat[ordered], col=orderedCols, las=2)
      #barplot 
      
      ### check age distrubtion by 
      
      agesByNewNormalGeo <- lapply(newNormalSamplesByGeo, function(currSamples){
        ages = basicMetaMapUpdated$Age[basicMetaMapUpdated$sample.ID %in% currSamples]
        return(ages)
      })
      
      par(mfrow=c(2,1))
      par(mar=c(3,4,1,1))
      boxplot(agesByNewNormalGeo[ordered0], las=2, ylab="Age")  
      
      generateGeoDiffStatForRezaPouyan = T
      if (generateGeoDiffStatForRezaPouyan) {
        firstBranchSamples = samplesByState[["1"]]
        midBranchSamples = unlist(samplesByState[c("2","3","4")])
        lastBranchSamples = samplesByState[["5"]]
        
        firstSampleTab = data.frame(sample=firstBranchSamples, type="TraditionalCluster")
        midSampleTab = data.frame(sample=midBranchSamples, type="UsChinaJapanCluster")
        lastSampleTab = data.frame(sample=lastBranchSamples, type="EuropeanCluster")
        allBranchSampleTab = rbind(firstSampleTab, 
                                   midSampleTab,
                                   lastSampleTab)
        write.table(allBranchSampleTab, "C://Data//comparative.analysis.healthy.sweden//countryClusterSampleTab.txt", sep="\t")
        
        volcano.stats.mgs.first.vs.mid = getVolcanoStat(mergeMatUpdated, midBranchSamples, firstBranchSamples, taxo, T)
        volcano.stats.mgs.mid.vs.last = getVolcanoStat(mergeMatUpdated, lastBranchSamples, midBranchSamples, taxo, T)
        volcano.stats.mgs.first.vs.last = getVolcanoStat(mergeMatUpdated, lastBranchSamples, firstBranchSamples, taxo, T)
        
        statsMgsForFirstVsMidNewFile = "J:\\Deposit\\Project\\2018_microbiome_atlas\\atlas.geo.difference.stats\\richness.volcano.stats.mgs.first.vs.mid.new.txt"
        statsMgsForMidVsLastNewFile = "J:\\Deposit\\Project\\2018_microbiome_atlas\\atlas.geo.difference.stats\\richness.volcano.stats.mgs.mid.vs.last.new.txt"
        statsMgsForFirstVsLastNewFile = "J:\\Deposit\\Project\\2018_microbiome_atlas\\atlas.geo.difference.stats\\richness.volcano.stats.mgs.first.vs.last.new.txt"
        
        write.table(volcano.stats.mgs.first.vs.mid, statsMgsForFirstVsMidNewFile, sep="\t")
        write.table(volcano.stats.mgs.mid.vs.last, statsMgsForMidVsLastNewFile, sep="\t")
        write.table(volcano.stats.mgs.first.vs.last, statsMgsForFirstVsLastNewFile, sep="\t")
        
      }
      
      generateDetailGeoDiffStatForReza = T
      if (generateDetailGeoDiffStatForReza) {
        
        newNormalMetadata = basicMetaMapUpdated[basicMetaMapUpdated$sample.ID %in% newNormalAllIds, ]
        newNormalSamplesByGeo = split(newNormalMetadata$sample.ID, newNormalMetadata$Geography)
        newNormalSampleCountsByGeo = do.call(c, lapply(newNormalSamplesByGeo, length))
        
        europeCountries = c("Denmark", "Germany", "Italy", "Japan", "Luxembourg", "Spain", "Sweden", "UK")
        tradCountries = c("Fiji", "India", "Madagascar", "Mongo", "Peru", "Tanzania", "Thai")
        geoCompList = list(CN = newNormalSamplesByGeo$China, 
                           EU = unlist(newNormalSamplesByGeo[europeCountries]), 
                           US = newNormalSamplesByGeo$US, 
                           JP = newNormalSamplesByGeo$Japan, 
                           TR = unlist(newNormalSamplesByGeo[tradCountries]))
        
        geoCompMelt = melt(geoCompList)
        colnames(geoCompMelt) = c("sample.ID","Country_code")
        
        geoCompMeltFile = "J:\\Deposit\\Project\\2018_microbiome_atlas\\atlas.geo.difference.stats//geo.diff.stats.samples.20190925.txt"
        write.table(geoCompMelt, geoCompMeltFile, sep="\t", quote = F, row.names = F)
        
        geoCompLen = length(geoCompList)
        geoSeqs = seq_along(geoCompList)
        
        resList = list()
        ind=1
        iSeqs = seq(1, geoCompLen-1)
        for (i in iSeqs) {
          iGeoSamples = geoCompList[[i]]
          iGeo = names(geoCompList)[i]
          
          jSeqs = seq(i+1, geoCompLen)
          for (j in jSeqs) {
            if (i!=j) 
              jGeoSamples = geoCompList[[j]]
            jGeo = names(geoCompList)[j]
            
            ijVolcanoStats = getVolcanoStat(mergeMatUpdated, jGeoSamples, iGeoSamples, taxo, T)
            colnames(ijVolcanoStats)= c("lfc_target_over_ref", "pvalue", "qvalue", "sig", "relAbd", "relAbd_target", "relAbd_reference", "label")
            ijVolcanoStats$msp = rownames(ijVolcanoStats)
            ijVolcanoStats$target = iGeo
            ijVolcanoStats$reference = jGeo
            resList[[ind]] = ijVolcanoStats
            ind = ind+1
          }
        }
        resTab = do.call(rbind, resList)
        geoCompWilcoxStatsFile = "J:\\Deposit\\Project\\2018_microbiome_atlas\\atlas.geo.difference.stats//geo.diff.stats.wilcoxon.output.20190925.txt"
        write.table(resTab, geoCompWilcoxStatsFile, sep="\t", quote = F, row.names = F)
        
      }
      
      generatePouyanStat = F
      if (generatePouyanStat) {
        
        firstSamples = newNormalSamplesByGeo[ordered[1:7]]
        midSamples = newNormalSamplesByGeo[ordered[8:10]]
        lastSamples = newNormalSamplesByGeo[ordered[11:17]]
        
        firstSamples = unlist(firstSamples)
        midSamples = unlist(midSamples)
        lastSamples = unlist(lastSamples)
        
        volcano.stats.mgs.first.vs.mid = getVolcanoStat(mergeMatUpdated, midSamples, firstSamples, taxo, T)
        volcano.stats.mgs.mid.vs.last = getVolcanoStat(mergeMatUpdated, lastSamples, midSamples, taxo, T)
        volcano.stats.mgs.first.vs.last = getVolcanoStat(mergeMatUpdated, lastSamples, firstSamples, taxo, T)
        
        
        statsMgsForFirstVsMidFile = "J:\\Deposit\\Project\\2018_microbiome_atlas\\atlas.geo.difference.stats\\richness.volcano.stats.mgs.first.vs.mid.txt"
        statsMgsForMidVsLastFile = "J:\\Deposit\\Project\\2018_microbiome_atlas\\atlas.geo.difference.stats\\richness.volcano.stats.mgs.mid.vs.last.txt"
        statsMgsForFirstVsLastFile = "J:\\Deposit\\Project\\2018_microbiome_atlas\\atlas.geo.difference.stats\\richness.volcano.stats.mgs.first.vs.last.txt"
        
        write.table(volcano.stats.mgs.first.vs.mid, statsMgsForFirstVsMidFile, sep="\t")
        write.table(volcano.stats.mgs.mid.vs.last, statsMgsForMidVsLastFile, sep="\t")
        write.table(volcano.stats.mgs.first.vs.last, statsMgsForFirstVsLastFile, sep="\t")
        
      }
    }
    
    checkWithRichnessSignature = T
    if (checkWithRichnessSignature) {
      
      loadRichnessNewCutMode = T
      if (loadRichnessNewCutMode) {
        adjpCut = 1e-3
        statsMgsForNormSamples = "J:\\Deposit\\Project\\2018_microbiome_atlas\\atlas.volcanot.out\\richness.volcano.stats.mgs.for.normal.txt"
        statsMgsForNorm = read.table(statsMgsForNormSamples, sep="\t", header=T, stringsAsFactors = F)
        statsMgsForNormSel = statsMgsForNorm[statsMgsForNorm$qvalue <= adjpCut,]
        # statsMgsForNormSelHgc = statsMgsForNormSel[statsMgsForNormSel$lfc > 0,]
        # statsMgsForNormSelLgc = statsMgsForNormSel[statsMgsForNormSel$lfc < -0,]
        statsMgsForNormSelHgc = statsMgsForNormSel[statsMgsForNormSel$lfc > 2,]
        statsMgsForNormSelLgc = statsMgsForNormSel[statsMgsForNormSel$lfc < -2,]
        
        hgcSpecies = rownames(statsMgsForNormSelHgc)
        lgcSpecies = rownames(statsMgsForNormSelLgc)
        
        length(hgcSpecies)
        length(lgcSpecies)
        
        if (F) {
          mergeScaleMat =  t(scale(t(mergeMatUpdated)))
          hgcSignatureSum = colSums(mergeScaleMat[hgcSpecies,])/sqrt(length(hgcSpecies))
          lgcSignatureSum = colSums(mergeScaleMat[lgcSpecies,])/sqrt(length(lgcSpecies))
          
          gssLen = sum(rownames(mergeScaleMat) %in% gssSpecies)
          ossLen = sum(rownames(mergeScaleMat) %in% ossSpecies)
          
          
          gssSignatureSum = colSums(mergeScaleMat[rownames(mergeScaleMat) %in% gssSpecies,])/sqrt(length(gssSpecies))
          ossSignatureSum = colSums(mergeScaleMat[rownames(mergeScaleMat) %in% ossSpecies,])/sqrt(length(ossSpecies))
          
          totalSignatures = hgcSignatureSum - lgcSignatureSum
          trajectoryData2$richnessSignature = totalSignatures[match(rownames(trajectoryData2), names(totalSignatures))]
          
          gutOralSignatures = ossSignatureSum - gssSignatureSum
          trajectoryData2$gutOralSignature = gutOralSignatures[match(rownames(trajectoryData2), names(gutOralSignatures))]
          
          oralSignature = ossSignatureSum
          trajectoryData2$oralSignature = oralSignature[match(rownames(trajectoryData2), names(oralSignature))]
          
          
        }
        normMgsMat = mergeMatUpdated[, colnames(mergeMatUpdated) %in% newNormalAllIds]
        normMgsMat = normMgsMat[rowSums(normMgsMat)!=0,]
        normMgsScaleMat = t(scale(t(normMgsMat)))
        
        
        hgcSpecies = hgcSpecies[hgcSpecies %in% rownames(normMgsScaleMat)]
        lgcSpecies = lgcSpecies[lgcSpecies %in% rownames(normMgsScaleMat)]
        length(hgcSpecies)
        length(lgcSpecies)
        
        
        hgcSignatureSum = colSums(normMgsScaleMat[hgcSpecies,])/sqrt(length(hgcSpecies))
        lgcSignatureSum = colSums(normMgsScaleMat[lgcSpecies,])/sqrt(length(lgcSpecies))
        
        gssLen = sum(rownames(normMgsScaleMat) %in% gssSpecies)
        ossLen = sum(rownames(normMgsScaleMat) %in% ossSpecies)
        
        
        gssSignatureSum = colSums(normMgsScaleMat[rownames(normMgsScaleMat) %in% gssSpecies,])/sqrt(length(gssSpecies))
        ossSignatureSum = colSums(normMgsScaleMat[rownames(normMgsScaleMat) %in% ossSpecies,])/sqrt(length(ossSpecies))
        
        totalSignatures = hgcSignatureSum - lgcSignatureSum
        trajectoryData$richnessSignature = totalSignatures[match(rownames(trajectoryData), names(totalSignatures))]
        
        gutOralSignatures = ossSignatureSum - gssSignatureSum
        trajectoryData$gutOralSignature = gutOralSignatures[match(rownames(trajectoryData), names(gutOralSignatures))]
        
        oralSignature = ossSignatureSum
        trajectoryData$oralSignature = oralSignature[match(rownames(trajectoryData), names(oralSignature))]
        
        library(scales)
        ggplot(trajectoryData, aes(x=data_dim_1, y=data_dim_2, color=richnessSignature, label=geo)) + 
          geom_point() + 
          scale_color_gradient2(low = "#2c528c88", mid = "#ffffff77",high = "#ff000088") + 
          xlab("Component 1")+ylab("Component 2")+ 
          theme(panel.background = element_blank(),
                axis.ticks = element_blank(),
                axis.text.x = element_blank(),
                strip.text.x = element_blank(),
                axis.text.y = element_blank(),
                strip.text.y = element_blank(), #legend.position="none", 
                panel.border = element_rect(colour = "black", fill=NA, size=.5))
        
        ggplot(trajectoryData, aes(x=data_dim_1, y=data_dim_2, color=gutOralSignatures, label=geo)) + 
          geom_point(aes(size=gutOralSignature)) + 
          scale_color_gradient2(low = "#2c528c33", mid = "#ffffff33", midpoint = 5,high = "#ff0000") + 
          #scale_color_gradient(low =  "#ffffff33", high = "#ff0000") + 
          xlab("Component 1")+ylab("Component 2")+ 
          theme(panel.background = element_blank(),
                axis.ticks = element_blank(),
                axis.text.x = element_blank(),
                strip.text.x = element_blank(),
                axis.text.y = element_blank(),
                strip.text.y = element_blank(), #legend.position="none", 
                panel.border = element_rect(colour = "black", fill=NA, size=.5))
        
        ggplot(trajectoryData, aes(x=data_dim_1, y=data_dim_2, color=gutOralSignatures, label=geo)) + 
          geom_point() + 
          #scale_color_gradient2(low = "#2c528c88", mid = "#ffffff77",high = "#ff000088") + 
          scale_color_gradient(low = "#ffffff77", high = "#ff0000") + 
          xlab("Component 1")+ylab("Component 2")+ 
          theme(panel.background = element_blank(),
                axis.ticks = element_blank(),
                axis.text.x = element_blank(),
                strip.text.x = element_blank(),
                axis.text.y = element_blank(),
                strip.text.y = element_blank(), #legend.position="none", 
                panel.border = element_rect(colour = "black", fill=NA, size=.5))
        
      }
      
    }
    
    checkEnteroType = F
    if (checkEnteroType) {
      stateEnteroTypeList = grepEnteroTypesFrom(samplesByState, basicMetaMapUpdated)
      getEntGeoPlot(stateEnteroTypeList)
    }
    
    diffMode = F
    if (diffMode) {
      diff_test_rest_RData = "j://Deposit/Project/2018_microbiome_atlas/trajectoryData//diff.norm.by.pseudotime.RData"
      diff_test_res <- differentialGeneTest(cds, fullModelFormulaStr = "~sm.ns(Pseudotime)")
      save(diff_test_res, file=diff_test_rest_RData)
      
      diffOut = diff_test_res[,c("gene_short_name", "pval", "qval")]
      diffOut = diffOut[order(diffOut[,"pval"]),]
      diffOutClassified = diffOut[!grepl("unclassified", diffOut$gene_short_name),]
      
      head(diffOutClassified)
      
      sig_genes <- rownames(diffOutClassified[1:5,])
      
      sig_genes = c("msp_0905c",
                    "msp_1169c",
                    "msp_0656",
                    "msp_0832",
                    "msp_0187",
                    "msp_0209",
                    "msp_0224")
      sig_genes = c("msp_1341",
                    "msp_0483",
                    "msp_0956",
                    "msp_0811",
                    "msp_0006",
                    "msp_0792")
      
      sig_genes = c("msp_0216",
                    "msp_0368c",
                    "msp_1458",
                    "msp_0983",
                    "msp_0197")
      
      
      sig_genes = c("msp_0174",
                    "msp_0458",
                    "msp_0108",
                    "msp_0005",
                    "msp_0501")
      
      sig_genes = c("msp_0053",
                    "msp_1268",
                    "msp_0916",
                    "msp_0170",
                    "msp_0025")
      
      sig_genes = c("msp_0042",
                    "msp_1088",
                    "msp_0220",
                    "msp_0041",
                    "msp_0338",
                    "msp_0313")
      
      
      
      cds_subset <- cds[sig_genes,]
      plot_genes_in_pseudotime(cds_subset, color_by = "State")
      
      
    }
    
  }
  
  exceptEurMode = T
  if (exceptEurMode) {
    
    generateMode = T
    if (generateMode) {
      exceptEurSamples = newNormalAllIds[!newNormalAllIds %in% eurFiSamples]
      normAllMat = mergeMatUpdated*10^9
      normAllMat = normAllMat[, colnames(normAllMat) %in% exceptEurSamples]
      
      trajectorySampleTab = data.frame(Library=basicMetaMapUpdated$sample.ID,
                                       Well=basicMetaMapUpdated$dataset.ID,
                                       Hours=0,
                                       Media="NONE")
      
      rownames(trajectorySampleTab) = basicMetaMapUpdated$sample.ID
      trajectorySampleTab = trajectorySampleTab[rownames(trajectorySampleTab) %in% exceptEurSamples,]
      
      trajectoryTaxaTab = data.frame(gene_short_name = taxo$species, 
                                     biotype="protein_coding",
                                     num_cells_expressed = 1,
                                     use_for_ordering=FALSE)
      
      rownames(trajectoryTaxaTab) = rownames(taxo)
      trajectoryTaxaTab = trajectoryTaxaTab[match(rownames(normAllMat), rownames(taxo)),]
      
      pd <- new("AnnotatedDataFrame", data = trajectorySampleTab)
      fd <- new("AnnotatedDataFrame", data = trajectoryTaxaTab)
      
      cds <- newCellDataSet(normAllMat, 
                            phenoData = pd, 
                            featureData = fd,
                            lowerDetectionLimit = 0.1)
      
      cds <- estimateSizeFactors(cds)
      cds <- estimateDispersions(cds)
      
      cds <- reduceDimension(cds, 
                             max_components = 2,
                             method = 'DDRTree')
      
      cds <- orderCells(cds)
      
      save(cds, file=exceptEurDataTrajectoryRData)
    }
    
    loadMode = T
    if (loadMode) {
      load(exceptEurDataTrajectoryRData)
    }
    
    checkTrajectory = T
    if (checkTrajectory) {
      plot_cell_trajectory(cds, color_by = "Pseudotime")
      plot_cell_trajectory(cds, color_by = "State")
      
      
      brachInfo = pData(cds)
      samplesByState = split(as.character(brachInfo$Library), brachInfo$State)
      
      newNormalMetadata = basicMetaMapUpdated[basicMetaMapUpdated$sample.ID %in% newNormalAllIds, ]
      newNormalSamplesByGeo = split(newNormalMetadata$sample.ID, newNormalMetadata$Geography)
      newNormalSampleCountsByGeo = do.call(c, lapply(newNormalSamplesByGeo, length))
      
      getPercentGeoByState <- function(targetSamples, newNormalSamplesByGeo, newNormalSampleCountsByGeo) {
        currCounts = do.call(c, lapply(newNormalSamplesByGeo, function(currSamples) {
          counts = sum(currSamples %in% targetSamples )
          return(counts)
        }))
        
        return(currCounts*100/newNormalSampleCountsByGeo)
      }
      
      geoByStateMat = do.call(rbind,lapply(samplesByState, function(currSamples) getPercentGeoByState(currSamples, newNormalSamplesByGeo, newNormalSampleCountsByGeo) ))
      View(geoByStateMat)
      
      barplotMode = T
      if (barplotMode) {
        firstSeqs = c("1","2","3","4","5","6","7")
        midSeqs = c("8")
        lastSeqs = c("9")
        firstBranchSamples = unlist(samplesByState[firstSeqs])
        midBranchSamples = unlist(samplesByState[midSeqs])
        lastBranchSamples = unlist(samplesByState[lastSeqs])
        
        firstBranchStat = getPercentGeoByState(firstBranchSamples,
                                               newNormalSamplesByGeo,
                                               newNormalSampleCountsByGeo)
        
        midBranchStat = getPercentGeoByState(midBranchSamples,
                                             newNormalSamplesByGeo,
                                             newNormalSampleCountsByGeo)
        
        lastBranchStat = getPercentGeoByState(lastBranchSamples,
                                              newNormalSamplesByGeo,
                                              newNormalSampleCountsByGeo)
        
        ordered = c("Tanzania","Madagascar","Peru","India","Fiji","Mongo","Thai",
                    "Japan","China","US", 
                    "Sweden","Denmark","UK","Spain","Germany","Luxembourg","Italy")
        
        
        tradVec = c(255,127,0,120)/255
        tradCol = rgb(tradVec[1],tradVec[2],tradVec[3],tradVec[4])
        
        cjpVec = c(211,211,211,120)/255
        cjpCol = rgb(cjpVec[1],cjpVec[2],cjpVec[3],cjpVec[4])
        
        eurVec = c(117,112,179,120)/255
        eurCol = rgb(eurVec[1],eurVec[2],eurVec[3],eurVec[4])
        
        
        orderedCols = c(rep(tradCol, 7),
                        rep(cjpCol, 3), #4
                        rep(eurCol, 7))
        par(mfrow=c(4,1))
        par(mar=c(1,4,1,1))
        barplot(firstBranchStat[ordered], col=orderedCols, las=2, axisnames = F)
        par(mar=c(1,4,1,1))
        barplot(midBranchStat[ordered], col=orderedCols, las=2, axisnames = F)
        par(mar=c(1,4,1,1))
        barplot(lastBranchStat[ordered], col=orderedCols, las=2)
        #barplot 
        
      }
      
    }
  }
  
  exceptCjpMode = T
  if (exceptCjpMode) {
    
    generateMode = F
    if (generateMode) {
      exceptCjpSamples = newNormalAllIds[!newNormalAllIds %in% cjpSamples]
      normAllMat = mergeMatUpdated*10^9
      normAllMat = normAllMat[, colnames(normAllMat) %in% exceptCjpSamples]
      
      trajectorySampleTab = data.frame(Library=basicMetaMapUpdated$sample.ID,
                                       Well=basicMetaMapUpdated$dataset.ID,
                                       Hours=0,
                                       Media="NONE")
      
      rownames(trajectorySampleTab) = basicMetaMapUpdated$sample.ID
      trajectorySampleTab = trajectorySampleTab[rownames(trajectorySampleTab) %in% exceptCjpSamples,]
      
      trajectoryTaxaTab = data.frame(gene_short_name = taxo$species, 
                                     biotype="protein_coding",
                                     num_cells_expressed = 1,
                                     use_for_ordering=FALSE)
      
      rownames(trajectoryTaxaTab) = rownames(taxo)
      trajectoryTaxaTab = trajectoryTaxaTab[match(rownames(normAllMat), rownames(taxo)),]
      
      pd <- new("AnnotatedDataFrame", data = trajectorySampleTab)
      fd <- new("AnnotatedDataFrame", data = trajectoryTaxaTab)
      
      cds <- newCellDataSet(normAllMat, 
                            phenoData = pd, 
                            featureData = fd,
                            lowerDetectionLimit = 0.1)
      
      cds <- estimateSizeFactors(cds)
      cds <- estimateDispersions(cds)
      
      cds <- reduceDimension(cds, 
                             max_components = 2,
                             method = 'DDRTree')
      
      cds <- orderCells(cds)
      
      save(cds, file=exceptCjpDataTrajectoryRData)
      
    }
    
    loadMode = T
    if (loadMode) {
      load(exceptCjpDataTrajectoryRData)
    }
    
    checkTrajectory = T
    if (checkTrajectory) {
      plot_cell_trajectory(cds, color_by = "Pseudotime")
      plot_cell_trajectory(cds, color_by = "State")
      
      brachInfo = pData(cds)
      samplesByState = split(as.character(brachInfo$Library), brachInfo$State)
      
      newNormalMetadata = basicMetaMapUpdated[basicMetaMapUpdated$sample.ID %in% newNormalAllIds, ]
      newNormalSamplesByGeo = split(newNormalMetadata$sample.ID, newNormalMetadata$Geography)
      newNormalSampleCountsByGeo = do.call(c, lapply(newNormalSamplesByGeo, length))
      
      getPercentGeoByState <- function(targetSamples, newNormalSamplesByGeo, newNormalSampleCountsByGeo) {
        currCounts = do.call(c, lapply(newNormalSamplesByGeo, function(currSamples) {
          counts = sum(currSamples %in% targetSamples )
          return(counts)
        }))
        
        return(currCounts*100/newNormalSampleCountsByGeo)
      }
      
      geoByStateMat = do.call(rbind,lapply(samplesByState, function(currSamples) getPercentGeoByState(currSamples, newNormalSamplesByGeo, newNormalSampleCountsByGeo) ))
      View(geoByStateMat)
      
      barplotMode = T
      if (barplotMode) {
        firstSeqs = c("1")
        midSeqs = c("2","3","4","5","6","7","8")
        lastSeqs = c("9","10","11")
        firstBranchSamples = unlist(samplesByState[firstSeqs])
        midBranchSamples = unlist(samplesByState[midSeqs])
        lastBranchSamples = unlist(samplesByState[lastSeqs])
        
        firstBranchStat = getPercentGeoByState(firstBranchSamples,
                                               newNormalSamplesByGeo,
                                               newNormalSampleCountsByGeo)
        
        midBranchStat = getPercentGeoByState(midBranchSamples,
                                             newNormalSamplesByGeo,
                                             newNormalSampleCountsByGeo)
        
        lastBranchStat = getPercentGeoByState(lastBranchSamples,
                                              newNormalSamplesByGeo,
                                              newNormalSampleCountsByGeo)
        
        ordered = c("Tanzania","Madagascar","Peru","India","Fiji","Mongo","Thai",
                    "Japan","China","US", 
                    "Sweden","Denmark","UK","Spain","Germany","Luxembourg","Italy")
        
        
        tradVec = c(255,127,0,120)/255
        tradCol = rgb(tradVec[1],tradVec[2],tradVec[3],tradVec[4])
        
        cjpVec = c(211,211,211,120)/255
        cjpCol = rgb(cjpVec[1],cjpVec[2],cjpVec[3],cjpVec[4])
        
        eurVec = c(117,112,179,120)/255
        eurCol = rgb(eurVec[1],eurVec[2],eurVec[3],eurVec[4])
        
        
        orderedCols = c(rep(tradCol, 7),
                        rep(cjpCol, 3), #4
                        rep(eurCol, 7))
        par(mfrow=c(4,1))
        par(mar=c(1,4,1,1))
        barplot(firstBranchStat[ordered], col=orderedCols, las=2, axisnames = F)
        par(mar=c(1,4,1,1))
        barplot(midBranchStat[ordered], col=orderedCols, las=2, axisnames = F)
        par(mar=c(1,4,1,1))
        barplot(lastBranchStat[ordered], col=orderedCols, las=2)
        #barplot 
        
      }
      
    }
  }
  
  exceptTradMode = T
  if (exceptTradMode) {
    
    generateMode = F
    if (generateMode) {
      exceptTrdSamples = newNormalAllIds[!newNormalAllIds %in% trdSamples]
      normAllMat = mergeMatUpdated*10^9
      normAllMat = normAllMat[, colnames(normAllMat) %in% exceptTrdSamples]
      
      trajectorySampleTab = data.frame(Library=basicMetaMapUpdated$sample.ID,
                                       Well=basicMetaMapUpdated$dataset.ID,
                                       Hours=0,
                                       Media="NONE")
      
      rownames(trajectorySampleTab) = basicMetaMapUpdated$sample.ID
      trajectorySampleTab = trajectorySampleTab[rownames(trajectorySampleTab) %in% exceptTrdSamples,]
      
      trajectoryTaxaTab = data.frame(gene_short_name = taxo$species, 
                                     biotype="protein_coding",
                                     num_cells_expressed = 1,
                                     use_for_ordering=FALSE)
      
      rownames(trajectoryTaxaTab) = rownames(taxo)
      trajectoryTaxaTab = trajectoryTaxaTab[match(rownames(normAllMat), rownames(taxo)),]
      
      pd <- new("AnnotatedDataFrame", data = trajectorySampleTab)
      fd <- new("AnnotatedDataFrame", data = trajectoryTaxaTab)
      
      cds <- newCellDataSet(normAllMat, 
                            phenoData = pd, 
                            featureData = fd,
                            lowerDetectionLimit = 0.1)
      
      cds <- estimateSizeFactors(cds)
      cds <- estimateDispersions(cds)
      
      cds <- reduceDimension(cds, 
                             max_components = 2,
                             method = 'DDRTree')
      
      cds <- orderCells(cds)
      
      save(cds, file=exceptTradDataTrajectoryRData)
      
    }
    
    loadMode = T
    if (loadMode) {
      load(exceptTradDataTrajectoryRData)
    }
    
    checkTrajectory = T
    if (checkTrajectory) {
      plot_cell_trajectory(cds, color_by = "Pseudotime")
      plot_cell_trajectory(cds, color_by = "State")
      
      brachInfo = pData(cds)
      samplesByState = split(as.character(brachInfo$Library), brachInfo$State)
      
      newNormalMetadata = basicMetaMapUpdated[basicMetaMapUpdated$sample.ID %in% newNormalAllIds, ]
      newNormalSamplesByGeo = split(newNormalMetadata$sample.ID, newNormalMetadata$Geography)
      newNormalSampleCountsByGeo = do.call(c, lapply(newNormalSamplesByGeo, length))
      
      getPercentGeoByState <- function(targetSamples, newNormalSamplesByGeo, newNormalSampleCountsByGeo) {
        currCounts = do.call(c, lapply(newNormalSamplesByGeo, function(currSamples) {
          counts = sum(currSamples %in% targetSamples )
          return(counts)
        }))
        
        return(currCounts*100/newNormalSampleCountsByGeo)
      }
      
      geoByStateMat = do.call(rbind,lapply(samplesByState, function(currSamples) getPercentGeoByState(currSamples, newNormalSamplesByGeo, newNormalSampleCountsByGeo) ))
      View(geoByStateMat)
      
      barplotMode = T
      if (barplotMode) {
        firstSeqs = c("7","8")
        midSeqs = c("2","3","4","5","6")
        lastSeqs = c("1","9")
        firstBranchSamples = unlist(samplesByState[firstSeqs])
        midBranchSamples = unlist(samplesByState[midSeqs])
        lastBranchSamples = unlist(samplesByState[lastSeqs])
        
        firstBranchStat = getPercentGeoByState(firstBranchSamples,
                                               newNormalSamplesByGeo,
                                               newNormalSampleCountsByGeo)
        
        midBranchStat = getPercentGeoByState(midBranchSamples,
                                             newNormalSamplesByGeo,
                                             newNormalSampleCountsByGeo)
        
        lastBranchStat = getPercentGeoByState(lastBranchSamples,
                                              newNormalSamplesByGeo,
                                              newNormalSampleCountsByGeo)
        
        ordered = c("Tanzania","Madagascar","Peru","India","Fiji","Mongo","Thai",
                    "Japan","China","US", 
                    "Sweden","Denmark","UK","Spain","Germany","Luxembourg","Italy")
        
        
        tradVec = c(255,127,0,120)/255
        tradCol = rgb(tradVec[1],tradVec[2],tradVec[3],tradVec[4])
        
        cjpVec = c(211,211,211,120)/255
        cjpCol = rgb(cjpVec[1],cjpVec[2],cjpVec[3],cjpVec[4])
        
        eurVec = c(117,112,179,120)/255
        eurCol = rgb(eurVec[1],eurVec[2],eurVec[3],eurVec[4])
        
        
        orderedCols = c(rep(tradCol, 7),
                        rep(cjpCol, 3), #4
                        rep(eurCol, 7))
        par(mfrow=c(4,1))
        par(mar=c(1,4,1,1))
        barplot(firstBranchStat[ordered], col=orderedCols, las=2, axisnames = F)
        par(mar=c(1,4,1,1))
        barplot(midBranchStat[ordered], col=orderedCols, las=2, axisnames = F)
        par(mar=c(1,4,1,1))
        barplot(lastBranchStat[ordered], col=orderedCols, las=2)
        #barplot 
        
      }
      
    }
    
  }
  
  wholeDataMode = F
  if (wholeDataMode) {
    
    generateMode = F
    if (generateMode) {
      trajectorySampleTab = data.frame(Library=basicMetaMapUpdated$sample.ID,
                                       Well=basicMetaMapUpdated$dataset.ID,
                                       Hours=0,
                                       Media="NONE")
      
      rownames(trajectorySampleTab) = basicMetaMapUpdated$sample.ID
      
      trajectoryTaxaTab = data.frame(gene_short_name = taxo$species, 
                                     biotype="protein_coding",
                                     num_cells_expressed = 1,
                                     use_for_ordering=FALSE)
      
      rownames(trajectoryTaxaTab) = rownames(taxo)
      trajectoryTaxaTab = trajectoryTaxaTab[match(rownames(mergeMatUpdated), rownames(taxo)),]
      
      pd <- new("AnnotatedDataFrame", data = trajectorySampleTab)
      fd <- new("AnnotatedDataFrame", data = trajectoryTaxaTab)
      cds <- newCellDataSet(mergeMatUpdated*10^9, 
                            phenoData = pd, 
                            featureData = fd,
                            lowerDetectionLimit = 0.1)
      
      cds <- estimateSizeFactors(cds)
      cds <- estimateDispersions(cds)
      #cds <- detectGenes(cds, min_expr = 0.1)
      #expressed_genes <- row.names(subset(fData(cds), num_cells_expressed >= 1))
      
      
      cds <- reduceDimension(cds, 
                             max_components = 2,
                             method = 'DDRTree')
      
      cds <- orderCells(cds)
    }
    
    loadMode = T
    if (loadMode) load(wholeDataTrajectoryRData)
    
    #save(cds, file=wholeDataTrajectoryRData)
    
    plot_cell_trajectory(cds, color_by = "Pseudotime")
    plot_cell_trajectory(cds, color_by = "State")
    
    
    plot_cell_trajectory(cds, color_by = "State") +
      facet_wrap(~State, nrow = 1)
    
    
    
    brachInfo = pData(cds)
    samplesByState = split(as.character(brachInfo$Library), brachInfo$State)
    babyData = basicMetaMapUpdated$sample.ID[basicMetaMapUpdated$dataset.ID=="id38"]
    
    targetData = diseaseUniqSubjectList$T2D_ES_id53
    lapply(samplesByState, function(currSamples)any(currSamples %in% targetData))
    
    targetState = 2
    currStateInfo = (basicMetaMapUpdated[ basicMetaMapUpdated$sample.ID %in% rownames(brachInfo[brachInfo$State==targetState,]),])
    mean(currStateInfo$GeneRichness)
    
    sort(table(currStateInfo$Geography))
    
    
    # plot_genes_branched_pseudotime(lung[lung_genes,],
    #                                branch_point = 1,
    #                                color_by = "Time",
    #                                ncol = 1)
    
  }
  
  prunedDataMode = T
  if (prunedDataMode) {
    
  }
  
}

checkDysbiosisGenus = F
if (checkDysbiosisGenus) {
  
  corrGenusMat = cor(genusMatUpdated)
  
  boxplot(as.numeric(corrGenusMat[newDiseaseIds, newWesternIds]))
  boxplot(as.numeric(corrGenusMat[newWesternIds, newWesternIds]))
  boxplot(as.numeric(corrGenusMat[newDiseaseIds, newDiseaseIds]))
  
  corrList = list(withinIndustrial = as.numeric(corrGenusMat[newWesternIds, newWesternIds]),
                  withinTraditional = as.numeric(corrGenusMat[newNonWesternIds, newNonWesternIds]),
                  withinDiseases = as.numeric(corrGenusMat[newDiseaseIds, newDiseaseIds]))
  
  
  mean(as.numeric(corrGenusMat[newDiseaseIds, newWesternIds]))
  mean(as.numeric(corrGenusMat[newWesternIds, newWesternIds]))
  mean(as.numeric(corrGenusMat[newWesternIds, newNonWesternIds]))
  mean(as.numeric(corrGenusMat[newNonWesternIds, newNonWesternIds]))
  mean(as.numeric(corrGenusMat[newDiseaseIds, newDiseaseIds]))
  
}

### very important ###
analyzeFunctionCluster = T
if (analyzeFunctionCluster) {
  
  # funcModNetNodeTableFile = "J:\\Deposit\\Project\\2018_microbiome_atlas\\functypeData\\funcModeNet.nodeTable.20190828.txt"
  # funcModNetAttachedNodeTableFile = "J:\\Deposit\\Project\\2018_microbiome_atlas\\functypeData\\funcModeNet.attached.nodeTable.20190902.txt"
  # funcModNetEdgeTableFile = "J:\\Deposit\\Project\\2018_microbiome_atlas\\functypeData\\funcModeNet.edgeTable.20190828.txt"
  # 
  # funcGemModNetNodeTableFile = "J:\\Deposit\\Project\\2018_microbiome_atlas\\functypeData\\funcGemModeNet.nodeTable.20190828.txt"
  # funcGemModNetAttachedNodeTableFile = "J:\\Deposit\\Project\\2018_microbiome_atlas\\functypeData\\funcGemModeNet.attached.nodeTable.20190828.txt"
  # funcGemModNetEdgeTableFile = "J:\\Deposit\\Project\\2018_microbiome_atlas\\functypeData\\funcGemModeNet.edgeTable.20190828.txt"
  # 
  funcMatRData = "j://Deposit/Project/2018_microbiome_atlas/functional.annotation/funcMat.20190808.RData"
  load(funcMatRData)
  
  funcModNetNodeTableFile = "J:\\Deposit\\Project\\2018_microbiome_atlas\\functional.annotation//funcModeNet.nodeTable.20191227.txt"
  funcModNetAttachedNodeTableFile = "J:\\Deposit\\Project\\2018_microbiome_atlas\\functional.annotation\\funcModeNet.attached.nodeTable.20191227.txt"
  funcModNetEdgeTableFile = "J:\\Deposit\\Project\\2018_microbiome_atlas\\functional.annotation\\funcModeNet.edgeTable.20191227.txt"
  
  #funcModulesRData = "J:\\Deposit\\Project\\2018_microbiome_atlas\\functypeData\\funcModules.20190828.RData"
  funcModulesRData = "J:\\Deposit\\Project\\2018_microbiome_atlas\\functional.annotation//funcModules.20191227.RData"
  #funcGemModulesRData = "J:\\Deposit\\Project\\2018_microbiome_atlas\\functypeData\\funcGemModules.20190828.RData"
  
  checkMaxNumberOfMatchedSpecies <- function(funcModules, funcMat, suppressMsps2) {
    funcMat = funcMat[ !rownames(funcMat) %in% suppressMsps2, ]
    maxNums = lapply(funcModules, function(items){
      if(length(items)==1) {
        currMat = funcMat[,colnames(funcMat)%in% items]
        rowCount = sum(currMat>0)
        return(rowCount)
        
      }
      currMat = funcMat[,colnames(funcMat)%in% items]
      rowCounts = rowSums(currMat>0) 
      return(max(rowCounts))
    })
    return(maxNums)
  }
  maxNums = checkMaxNumberOfMatchedSpecies(funcModules, funcMat, suppressMsps2)
  smallFcs = names(maxNums[maxNums<3])
  
  length(funcModules)
  
  loadAnnots = T
  if (loadAnnots) { # seven types of functional/phenotypic annotations
    
    ### virulence factors ###
    vfMatRData = "j://Deposit/Project/2018_microbiome_atlas/functional.annotation/PATRIC/vfBestMat.RData"
    load(vfMatRData) #vfMatDigit
    vfTerms = colnames(vfMatDigit)
    patricVfDescTabFile = "J://Deposit/Project/2018_microbiome_atlas/functional.annotation/PATRIC/PATRIC_sp_gene.txt"
    patricVfDescTab = read.table(patricVfDescTabFile, header=T, sep="\t", stringsAsFactors = F)
    
    ### JGI phenotype ###
    jgiMatRData = "j://Deposit/Project/2018_microbiome_atlas/functional.annotation/jgiMat.RData"
    load(jgiMatRData)
    jgiTerms = colnames(jgiMat)
    
    ### reactions ###
    gemRData = "J://Deposit/Project/2018_microbiome_atlas//atlas.GEM.model.related/GEM.from.Reza.amazing/absentPresentInModels.RData"
    load(gemRData)
    reactionTerms = colnames(gemMat)
    
    ### antibiotic resistance ###
    mustardMatRData = "j://Deposit/Project/2018_microbiome_atlas/functional.annotation/mustard.mat.RData"
    load(mustardMatRData)
    arTerms = colnames(mustardMat)
    
    ### antismash - secondary metabolism ###
    antismashMatRData = "J://Deposit/Project/2018_microbiome_atlas/functional.annotation/msp1992.igc2.antismashMat.RData"
    load(antismashMatRData)
    antismashMat[1:3,1:3]
    antismashMelt = melt(antismashMat)
    head(antismashMelt)
    antismashMelt = antismashMelt[antismashMelt$value!=0,]
    dim(antismashMelt)
    
    writeMode = F
    if (writeMode) write.table(antismashMelt, "antismash.list.txt", sep="\t")
    antismashTerms = colnames(antismashMat)
    
    ### Cazy  ### 
    newGutCazyMatRData = "J://Deposit/Project/2018_microbiome_atlas/functional.annotation/CAZY.INRA/igc2.new.cazy.mat.RData"
    load(newGutCazyMatRData)
    colnames(cazyMat) = paste("cazy", colnames(cazyMat), sep=".")
    cazyTerms = colnames(cazyMat)
    
    ### PFAM ###
    igc2PfamMatRData = "j://Deposit/Project/2018_microbiome_atlas/functional.annotation/PFAM/igc2PfamMat.RData"
    load(igc2PfamMatRData)
    pfamTerms = colnames(pfamMat)
    
    ### KEGG ###
    gutKoBestMatRData = "j://Deposit/Project/2018_microbiome_atlas/functional.annotation/gutKoBestMat.1990.RData"
    load(gutKoBestMatRData)
    koTerms = colnames(gutKoBestMat)
    
    ### KEGG pathway ###
    # already loaded from above scripts
    # data: bacteriaKoElementList
    
    ### KEGG modules ###
    keggModuleDescTabRData = "J://Deposit/Project/2018_microbiome_atlas/functional.annotation/KEGG module/moduleDescTab.RData"
    koModuleListRData = "J://Deposit/Project/2018_microbiome_atlas/functional.annotation/KEGG module/koModuleList.20190219.RData"
    load(keggModuleDescTabRData) #moduleDescTab
    load(koModuleListRData)
    microbeModuleTab = moduleDescTab[moduleDescTab$isBacteriaModule|moduleDescTab$isArchaeaModule,]
    nonMicrobeModuleTab = moduleDescTab[!(moduleDescTab$isBacteriaModule|moduleDescTab$isArchaeaModule),]
    nonMicrobeModuleTab2 = moduleDescTab[grepl("eukaryotes",moduleDescTab$moduleName),]
    
    microbeModules = paste(microbeModuleTab$moduleId, microbeModuleTab$moduleName, sep = ":")
    nonMicrobeModules = paste(nonMicrobeModuleTab$moduleId, nonMicrobeModuleTab$moduleName, sep = ":")
    nonMicrobeModules2 = paste(nonMicrobeModuleTab2$moduleId, nonMicrobeModuleTab2$moduleName, sep = ":")
    nonMicrobeModules = unique(c(nonMicrobeModules,nonMicrobeModules2))
    
    
    ### antismash ### spreaded
    antismashEnrichVec = getEnrichMentGsByList(funcModules, antismashTerms)
    antismashEnrichVec[antismashEnrichVec<=0.01]
    length(antismashEnrichVec[antismashEnrichVec<1])
    antismashClusters = names(antismashEnrichVec[antismashEnrichVec<1])
    
    ### virulence factors ###
    vfEnrichVec = getEnrichMentGsByList(funcModules, vfTerms)
    sort( vfEnrichVec[vfEnrichVec<=0.01])
    sort( vfEnrichVec[vfEnrichVec<1])
    length( sort( vfEnrichVec[vfEnrichVec<1]))
    vfClusters = names(vfEnrichVec[vfEnrichVec<1])
    
    ### antibiotic resistance ### spreaded
    arEnrichVec = getEnrichMentGsByList(funcModules, arTerms)
    sort( arEnrichVec[arEnrichVec<=0.01])
    length(sort( arEnrichVec[arEnrichVec<1]))
    arClusters = names(arEnrichVec[arEnrichVec<1])
    
    ### phenotype - JGI ### spreaded, but funcModules-149 with anaerobe, rod-shaped and non-sporulating
    jgiEnrichVec = getEnrichMentGsByList(funcModules, jgiTerms)
    sort( jgiEnrichVec[jgiEnrichVec<=0.01])
    length(sort( jgiEnrichVec[jgiEnrichVec<1]))
    jgiClusters = names(jgiEnrichVec[jgiEnrichVec<1])
    
    ### cazy ### mostly spreaded
    cazyEnrichVec = getEnrichMentGsByList(funcModules, cazyTerms)
    length(sort( cazyEnrichVec[cazyEnrichVec<1]))
    #sort( cazyEnrichVec[cazyEnrichVec<=0.01])
    cazyClusters = names(cazyEnrichVec[cazyEnrichVec<1])
    
    ### pfam ###
    pfamEnrichVec = getEnrichMentGsByList(funcModules, pfamTerms) # it contains "NA" values --> should be removed
    length(sort( pfamEnrichVec[pfamEnrichVec<1]))
    #sort( pfamEnrichVec[pfamEnrichVec<=0.01])
    pfamClusters = names(pfamEnrichVec[pfamEnrichVec<1])
    
    ### KEGG ###
    keggEnrichVec = getEnrichMentGsByList(funcModules, koTerms)
    length(sort( keggEnrichVec[keggEnrichVec<1]))
    #sort( keggEnrichVec[keggEnrichVec<=0.01])
    keggClusters = names(keggEnrichVec[keggEnrichVec<1])
    
    #pie(table(funcModuleNums==1))
    #median(funcModuleNums[funcModuleNums!=1])
    
    
    
  }
  
  loadFuncModule = T
  if (loadFuncModule) {
    
    load(funcModulesRData)
    #load(funcGemModulesRData)
    
    funcMatRData = "j://Deposit/Project/2018_microbiome_atlas/functional.annotation/funcMat.20190808.RData"
    load(funcMatRData)
    
    funcModuleNums = do.call(c, lapply(funcModules, length))
    funcModuleSel = funcModules[funcModuleNums>=3]
    signletons = names(funcModuleNums[funcModuleNums==1])
    singletons = names(funcModuleNums[funcModuleNums==1])
    
    table(funcModuleNums>1)
    table(funcModuleNums>=3)
    
    bileAcidKeggs =  c("K01442","K00076", "K23231", "K22604", "K22605", "K22606", "K22607", "K15868", "K15871", "K15869", "K15870", "K15872", "K15873", "K15874", "K07007")
    
    checkCountMode = F
    if (checkCountMode) {
      enrichedClusters = c(4153, 4105, 339,315, 20 , 18,16)
      names(enrichedClusters) = c("KEGG",
                                  "PFAM",
                                  "virulence",
                                  "CAZyme",
                                  "phenotype",
                                  "secondary met.", 
                                  "antibiotic res.")
      
      par(mar=c(10,4,4,1))
      barplot(enrichedClusters, las=2, log="y",
              col=colorRampPalette(c("gray","white", "blue"))(8))
    }
    
  }
  
  checkFunctionTableS4 = T
  if (checkFunctionTableS4) {
    tabS4FuncFile = "C:\\Data\\comparative.analysis.healthy.sweden\\functions.in.table.s4.txt"
    tabS4FuncList = read.table(tabS4FuncFile, sep="\t", stringsAsFactors = F, header=F)[,]
    tabS4ClassList = rep(NA, length(tabS4FuncList))
    
    tabS4ClassList[tabS4FuncList %in% vfTerms] = "Virluence(PATRIC)"
    tabS4ClassList[tabS4FuncList %in% jgiTerms] = "phenotype(JGI)"
    tabS4ClassList[tabS4FuncList %in% koTerms] = "KEGG orthology"
    tabS4ClassList[tabS4FuncList %in% pfamTerms] = "protein domain(PFAM)"
    tabS4ClassList[tabS4FuncList %in% cazyTerms] = "CAZyme"
    tabS4ClassList[tabS4FuncList %in% arTerms] = "AMR(Mustard)"
    tabS4ClassList[tabS4FuncList %in% antismashTerms] = "Secondary metabolites (Anti-SMASH)"
    # write.table(data.frame(func=tabS4FuncList, class=tabS4ClassList),
    #             file="functions.class.in.table.s4.txt", 
    #             sep="\t")
    # 
  }
  
  checkTableS10 = T
  if (checkTableS10) {
    clusterName = paste("CL-",names(funcModules),sep="")
    
    mspByModList =grepMspsAssociatedWithModules(funcModules = funcModules,
                                                funcMat = funcMat[!rownames(funcMat) %in% suppressMsps2, ], 
                                                cutRatio = 0.75)
    mspLenByModList = unlist(lapply(mspByModList,length))
    clusterMsps = unlist(lapply(mspByModList, function(x) paste(x, collapse = ";")))
    clusterSpecies = unlist(lapply(mspByModList, function(x) paste(getSpeciesName(x,taxo), collapse = ";")))
    
    clusterTerms = unlist(lapply(funcModules, function(x) paste(x, collapse = ";")))
    
    koCheck = unlist(lapply(funcModules, function(x) any(x %in% koTerms)))
    pfamCheck = unlist(lapply(funcModules, function(x) any(x %in% pfamTerms)))
    vfCheck = unlist(lapply(funcModules, function(x) any(x %in% vfTerms)))
    cazyCheck = unlist(lapply(funcModules, function(x) any(x %in% cazyTerms)))
    antismashCheck = unlist(lapply(funcModules, function(x) any(x %in% antismashTerms)))
    arCheck = unlist(lapply(funcModules, function(x) any(x %in% arTerms)))
    jgiCheck = unlist(lapply(funcModules, function(x) any(x %in% jgiTerms)))
    
    
    
    koTermsInMod = unlist(lapply(funcModules, function(x) {
      currTerms = x[x %in% koTerms]
      if (length(currTerms)==0) return("-")
      return(paste(currTerms, collapse = ";"))
    }))
    
    igc2PfamDescRData = "j://Deposit/Project/2018_microbiome_atlas/functional.annotation/PFAM/igc2PfamDesc.RData"
    load(igc2PfamDescRData)
    pfamDescTab = unique(igc2PfamDesc[,c("pfam_id","pfam_name", "pfam_desc")])
    
    getDescFromPfamTerms <- function(pfamTerms, pfamDescTab) {
      desc = pfamDescTab$pfam_desc[match(pfamTerms, pfamDescTab$pfam_name)]
      return(desc)
    }
    
    pfamTermsInMod = unlist(lapply(funcModules, function(x) {
      currTerms = x[x %in% pfamTerms ] 
      if (length(currTerms)==0) return("-")
      return(paste(currTerms, collapse = ";"))
    }))
    vfTermsInMod = unlist(lapply(funcModules, function(x) {
      currTerms = x[x %in% vfTerms]
      if (length(currTerms)==0) return("-")
      return(paste(currTerms, collapse = ";"))
    }))
    cazyTermsInMod = unlist(lapply(funcModules, function(x) {
      currTerms = x[x %in% cazyTerms]
      if (length(currTerms)==0) return("-")
      return(paste(currTerms, collapse = ";"))
    }))
    arTermsInMod = unlist(lapply(funcModules, function(x) {
      currTerms = x[x %in% arTerms]
      if (length(currTerms)==0) return("-")
      return(paste(currTerms, collapse = ";"))
    }))
    antismashTermsInMod = unlist(lapply(funcModules, function(x) {
      currTerms = x[x %in% antismashTerms]
      if (length(currTerms)==0) return("-")
      return(paste(currTerms, collapse = ";"))
    }))
    
    jgiTermsInMod = unlist(lapply(funcModules, function(x) {
      currTerms = x[x %in% jgiTerms]
      if (length(currTerms)==0) return("-")
      return(paste(currTerms, collapse = ";"))
    }))
    
    #koDesc
    #pfamDesc
    #vfProducts
    
    keggDescs = do.call(c, lapply(funcModules, function(x) {
      descs = getDescFromKoTerms(koTerms[koTerms %in% x], koDescMap)
      descs = descs[descs!=""]
      descs = descs[!is.na(descs)]
      str = getStrCatBy(descs)
      descs = unique(getItemsFromStr(str))
      out = getStrCatBy(descs)
      if (out=="") out = "-"
      return(out)
    }) )
    
    pfamDescs = do.call(c, lapply(funcModules, function(x) {
      descs = getDescFromPfamTerms(pfamTerms[pfamTerms %in% x], pfamDescTab)
      descs = descs[descs!=""]
      descs = descs[!is.na(descs)]
      str = getStrCatBy(descs)
      descs = unique(getItemsFromStr(str))
      out = getStrCatBy(descs)
      if (out=="") out = "-"
      return(out)
    }) )
    
    vfProducts = do.call(c, lapply(funcModules, function(x) {
      prods = getProductsFromVfTerms(vfTerms[vfTerms %in% x],patricVfDescTab)
      prods = prods[prods!=""]
      prods = prods[!is.na(prods)]
      str = getStrCatBy(prods)
      prods = unique(getItemsFromStr(str))
      out = getStrCatBy(prods)
      if (out=="") out = "-"
      return(out)
    }) )
    
    vfClasses = do.call(c, lapply(matchedFunc, function(x) {
      clss = getClassesFromVfTerms(vfTerms[vfTerms %in% x],patricVfDescTab)
      clss = clss[clss!=""]
      clss = clss[!is.na(clss)]
      str = getStrCatBy(clss)
      clss = unique(getItemsFromStr(str))
      out = getStrCatBy(clss)
      if (out=="") out = "-"
      return(out)
    }) )
    
    loadEnrichmentMode = T
    if (loadEnrichmentMode) {
      keggEnrichPathMatRData = "J:\\Deposit\\Project\\2018_microbiome_atlas\\functypeData\\keggEnrichPathMat.20200102.RData"
      #keggEnrichPathMatRData = "J:\\Deposit\\Project\\2018_microbiome_atlas\\functypeData\\keggEnrichPathMat.20190905.RData"
      keggEnrichModuleMatRData = "J:\\Deposit\\Project\\2018_microbiome_atlas\\functypeData\\keggEnrichModuleMat.20200102.RData"
      #keggEnrichModuleMatRData = "J:\\Deposit\\Project\\2018_microbiome_atlas\\functypeData\\keggEnrichModuleMat.20190905.RData"
      load(keggEnrichPathMatRData)
      load(keggEnrichModuleMatRData)
      
      keggEnrichPathMelt = melt(keggEnrichPathMat)
      keggEnrichPathMelt = keggEnrichPathMelt[keggEnrichPathMelt$value<=0.01,]
      keggEnrichPathMelt$Var1 = as.character(keggEnrichPathMelt$Var1)
      colnames(keggEnrichPathMelt) = c("pathway","funcModule","pvalue")
      exPaths = c("ko04626:Plant-pathogen interaction",
                  "ko05120:Epithelial cell signaling in Helicobacter pylori infection",
                  "ko04621:NOD-like receptor signaling pathway")
      keggEnrichPathMelt = keggEnrichPathMelt[!keggEnrichPathMelt$pathway %in% exPaths,]
      keggEnrichPathMelt$pathway[keggEnrichPathMelt$pathway=="ko05111:Vibrio cholerae pathogenic cycle"] <- "ko05111:Biofilm formation - Vibrio cholerae"
      
      keggEnrichModuleMelt = melt(keggEnrichModuleMat)
      keggEnrichModuleMelt = keggEnrichModuleMelt[keggEnrichModuleMelt$value<=0.01,]
      keggEnrichModuleMelt$Var1 = as.character(keggEnrichModuleMelt$Var1)
      keggEnrichModuleMelt = keggEnrichModuleMelt[keggEnrichModuleMelt$Var1 %in% microbeModules,]
      keggEnrichModuleMelt = keggEnrichModuleMelt[!keggEnrichModuleMelt$Var1 %in% nonMicrobeModules,]
      colnames(keggEnrichModuleMelt) = c("module","funcModule","pvalue")
    }
    
    keggEnModules = do.call(c, sapply(as.numeric(names(funcModules)), function(x) {
      currEnriched <- keggEnrichModuleMelt[keggEnrichModuleMelt$funcModule == x,]
      if (dim(currEnriched)[1] == 0 ) return("")
      return(getStrCatBy(currEnriched$module))
    }, simplify = F) )
    keggEnModSubSystems = do.call(c, sapply(as.numeric(names(funcModules)), function(x) {
      currEnriched <- keggEnrichModuleMelt[keggEnrichModuleMelt$funcModule == x,]
      if (dim(currEnriched)[1] == 0 ) return("")
      currModules = currEnriched$module
      currModuleIds = getIdsFromSplit(currModules)
      currSubsystems = moduleDescTab$subsystem[moduleDescTab$moduleId %in%currModuleIds]
      currSubsystems = unique(currSubsystems)
      return(getStrCatBy(currSubsystems))
    }, simplify = F) )
    keggEnPathways = do.call(c, sapply(as.numeric(names(funcModules)), function(x) {
      currEnriched <- keggEnrichPathMelt[keggEnrichPathMelt$funcModule == x,]
      if (dim(currEnriched)[1] == 0 ) return("")
      return(getStrCatBy(currEnriched$pathway))
    }, simplify = F) )
    
    normalClusters = names(funcModules)[!names(funcModules) %in% c(smallFcs, signletons)]
    normalClustersMod = paste("CL-",(normalClusters),sep="")
    View(normalClustersMod)
    
    tabS10 = data.frame(ID=clusterName,
                        size=funcModuleNums,
                        numEnrichedSpecies = mspLenByModList,
                        sigletons = ifelse(funcModuleNums==1, "Yes","No"), 
                        enrichedMsp = clusterMsps,
                        enrichedSpecies = clusterSpecies,
                        keggModules = keggEnModules,
                        keggSubsystems = keggEnModSubSystems,
                        keggPathwys = keggEnPathways,
                        koTerm = koTermsInMod,
                        pfamTerm = pfamTermsInMod,
                        virulenceTerm=vfTermsInMod,
                        cazyTerm=cazyTermsInMod,
                        amrTerm = arTermsInMod,
                        antismashTerm = antismashTermsInMod,
                        phenotypeTerm = jgiTermsInMod,
                        koTermDesc = keggDescs,
                        pfamTermDesc = pfamDescs,
                        virulenceDesc = vfProducts,
                        allTerms = clusterTerms)
    tabS10File = "C://Data//comparative.analysis.healthy.sweden//tabS10.txt"
    #write.table(tabS10, tabS10File, sep="\t", quote = F, row.names = F)
    
    tabS10File = "C://Data//comparative.analysis.healthy.sweden//tabS10.csv"
    #write.csv(tabS10, tabS10File, row.names = F)
    
  }
  
  generateMode = F
  if (generateMode) {
    ### KEGG pathways ###
    funcKoModules <- lapply(funcModules, function(x) x[x%in% koTerms])
    funcKoModules <- funcKoModules[do.call(c,lapply(funcKoModules, length))!=0]
    keggEnrichPathMat = getEnrichMentGsByGs(funcKoModules, bacteriaKoElementList)
    keggEnrichPathMelt = melt(keggEnrichPathMat)
    keggEnrichPathMelt = keggEnrichPathMelt[keggEnrichPathMelt$value<=0.05,]
    
    #keggEnrichPathMatRData = "J:\\Deposit\\Project\\2018_microbiome_atlas\\functypeData\\keggEnrichPathMat.20190905.RData"
    keggEnrichPathMatRData = "J:\\Deposit\\Project\\2018_microbiome_atlas\\functypeData\\keggEnrichPathMat.20200102.RData"
    save(keggEnrichPathMat, file=keggEnrichPathMatRData)
    
    ### KEGG modules ###
    funcKoModules <- lapply(funcModules, function(x) x[x%in% koTerms])
    funcKoModules <- funcKoModules[do.call(c,lapply(funcKoModules, length))!=0]
    keggEnrichModuleMat = getEnrichMentGsByGs(funcKoModules, koModuleList[names(koModuleList) %in% microbeModules])
    keggEnrichModuleMelt = melt(keggEnrichModuleMat)
    keggEnrichModuleMelt = keggEnrichModuleMelt[keggEnrichModuleMelt$value<=0.05,]
    
    #keggEnrichModuleMatRData = "J:\\Deposit\\Project\\2018_microbiome_atlas\\functypeData\\keggEnrichModuleMat.20190905.RData"
    keggEnrichModuleMatRData = "J:\\Deposit\\Project\\2018_microbiome_atlas\\functypeData\\keggEnrichModuleMat.20200102.RData"
    save(keggEnrichModuleMat, file=keggEnrichModuleMatRData)
    
  }
  
  
  
  attachMode = T
  if (attachMode) {
    # funcModules$`230`
    # 
    # outs <- do.call(c,lapply(funcModules, function(x) any(x %in% vfTerms)))
    # outTab = data.frame(names(funcModules),outs)
    # View(outTab[outTab[,2],])
    
    # termTest = do.call(c,lapply(funcModules, function(x)any(x%in%jgiTerms)))
    # termTest[termTest]
    # funcModules[["80"]]
    loadEnrichmentMode = T
    if (loadEnrichmentMode) {
      keggEnrichPathMatRData = "J:\\Deposit\\Project\\2018_microbiome_atlas\\functypeData\\keggEnrichPathMat.20200102.RData"
      #keggEnrichPathMatRData = "J:\\Deposit\\Project\\2018_microbiome_atlas\\functypeData\\keggEnrichPathMat.20190905.RData"
      keggEnrichModuleMatRData = "J:\\Deposit\\Project\\2018_microbiome_atlas\\functypeData\\keggEnrichModuleMat.20200102.RData"
      #keggEnrichModuleMatRData = "J:\\Deposit\\Project\\2018_microbiome_atlas\\functypeData\\keggEnrichModuleMat.20190905.RData"
      load(keggEnrichPathMatRData)
      load(keggEnrichModuleMatRData)
      
      keggEnrichPathMelt = melt(keggEnrichPathMat)
      keggEnrichPathMelt = keggEnrichPathMelt[keggEnrichPathMelt$value<=0.01,]
      keggEnrichPathMelt$Var1 = as.character(keggEnrichPathMelt$Var1)
      colnames(keggEnrichPathMelt) = c("pathway","funcModule","pvalue")
      exPaths = c("ko04626:Plant-pathogen interaction",
                  "ko05120:Epithelial cell signaling in Helicobacter pylori infection",
                  "ko04621:NOD-like receptor signaling pathway")
      keggEnrichPathMelt = keggEnrichPathMelt[!keggEnrichPathMelt$pathway %in% exPaths,]
      keggEnrichPathMelt$pathway[keggEnrichPathMelt$pathway=="ko05111:Vibrio cholerae pathogenic cycle"] <- "ko05111:Biofilm formation - Vibrio cholerae"
      
      keggEnrichModuleMelt = melt(keggEnrichModuleMat)
      keggEnrichModuleMelt = keggEnrichModuleMelt[keggEnrichModuleMelt$value<=0.01,]
      keggEnrichModuleMelt$Var1 = as.character(keggEnrichModuleMelt$Var1)
      keggEnrichModuleMelt = keggEnrichModuleMelt[keggEnrichModuleMelt$Var1 %in% microbeModules,]
      keggEnrichModuleMelt = keggEnrichModuleMelt[!keggEnrichModuleMelt$Var1 %in% nonMicrobeModules,]
      colnames(keggEnrichModuleMelt) = c("module","funcModule","pvalue")
    }
    
    ### attaching information ###
    funcModNetNodeTable = read.table(funcModNetNodeTableFile, stringsAsFactors = F, header=T, sep="\t")
    attachInfo <- function(funcModNetNodeTable,  funcModules,  antismashTerms, vfTerms, arTerms, jgiTerms, cazyTerms, pfamTerms, koTerms, koDescMap, moduleDescTab, patricVfDescTab) {
      
      
      nodes = as.character(funcModNetNodeTable$node)
      matchedFunc = funcModules[match(nodes, names(funcModules))]
      
      antismashCheck = do.call(c, lapply(matchedFunc, function(x) any(x %in% antismashTerms)))
      antismashCheck = antismashCheck*1
      antismashIds = do.call(c, lapply(matchedFunc, function(x) getStrCatBy(antismashTerms[antismashTerms %in% x])) )
      
      vfCheck = do.call(c, lapply(matchedFunc, function(x) any(x %in% vfTerms)))
      vfCheck = vfCheck*1
      vfIds = do.call(c, lapply(matchedFunc, function(x) getStrCatBy(vfTerms[vfTerms %in% x])) )
      vfGenes = do.call(c, lapply(matchedFunc, function(x) {
        
        genes = getGenesFromVfTerms(vfTerms[vfTerms %in% x],patricVfDescTab)
        genes = genes[genes!=""]
        genes = genes[!is.na(genes)]
        str = getStrCatBy(genes)
        genes = unique(getItemsFromStr(str))
        return(getStrCatBy(genes))
        
      }) )
      vfProducts = do.call(c, lapply(matchedFunc, function(x) {
        
        prods = getProductsFromVfTerms(vfTerms[vfTerms %in% x],patricVfDescTab)
        prods = prods[prods!=""]
        prods = prods[!is.na(prods)]
        str = getStrCatBy(prods)
        prods = unique(getItemsFromStr(str))
        return(getStrCatBy(prods))
        
      }) )
      vfClasses = do.call(c, lapply(matchedFunc, function(x) {
        
        clss = getClassesFromVfTerms(vfTerms[vfTerms %in% x],patricVfDescTab)
        clss = clss[clss!=""]
        clss = clss[!is.na(clss)]
        str = getStrCatBy(clss)
        clss = unique(getItemsFromStr(str))
        return(getStrCatBy( clss ))
        
      }) )
      
      arCheck = do.call(c, lapply(matchedFunc, function(x) any(x %in% arTerms)))
      arCheck = arCheck*1
      arIds = do.call(c, lapply(matchedFunc, function(x) getStrCatBy(arTerms[arTerms %in% x])) )
      
      jgiCheck = do.call(c, lapply(matchedFunc, function(x) any(x %in% jgiTerms)))
      jgiCheck = jgiCheck*1
      jgiIds = do.call(c, lapply(matchedFunc, function(x) getStrCatBy(jgiTerms[jgiTerms %in% x])) )
      
      cazyCheck = do.call(c, lapply(matchedFunc, function(x) any(x %in% cazyTerms)))
      cazyCheck = cazyCheck*1
      cazyIds = do.call(c, lapply(matchedFunc, function(x) {
        str = getStrCatBy(cazyTerms[cazyTerms %in% x])
        str = gsub("cazy.","",str)
        return(str)
      } ) )
      
      pfamCheck = do.call(c, lapply(matchedFunc, function(x) any(x %in% pfamTerms)))
      pfamCheck = pfamCheck*1
      pfamIds = do.call(c, lapply(matchedFunc, function(x) getStrCatBy(pfamTerms[pfamTerms %in% x])) )
      
      keggCheck = do.call(c, lapply(matchedFunc, function(x) any(x %in% koTerms)))
      keggCheck = keggCheck*1
      keggIds = do.call(c, lapply(matchedFunc, function(x) getStrCatBy(koTerms[koTerms %in% x])) )
      keggGenes = do.call(c, lapply(matchedFunc, function(x) {
        genes = getGeneFromKoTerms(koTerms[koTerms %in% x], koDescMap)
        genes = genes[genes!=""]
        genes = genes[!is.na(genes)]
        str = getStrCatBy(genes)
        genes = unique(getItemsFromStr(str))
        return(getStrCatBy(genes))
      }) )
      
      keggDescs = do.call(c, lapply(matchedFunc, function(x) {
        descs = getDescFromKoTerms(koTerms[koTerms %in% x], koDescMap)
        descs = descs[descs!=""]
        descs = descs[!is.na(descs)]
        str = getStrCatBy(descs)
        descs = unique(getItemsFromStr(str))
        return(getStrCatBy(descs))
      }) )
      
      ### KEGG modules and pathway
      keggEnModules = do.call(c, lapply(as.list(nodes), function(x) {
        currEnriched <- keggEnrichModuleMelt[keggEnrichModuleMelt$funcModule == x,]
        if (dim(currEnriched)[1] == 0 ) return("")
        return(getStrCatBy(currEnriched$module))
      }) )
      keggEnModSubSystems = do.call(c, lapply(as.list(nodes), function(x) {
        currEnriched <- keggEnrichModuleMelt[keggEnrichModuleMelt$funcModule == x,]
        if (dim(currEnriched)[1] == 0 ) return("")
        currModules = currEnriched$module
        currModuleIds = getIdsFromSplit(currModules)
        currSubsystems = moduleDescTab$subsystem[moduleDescTab$moduleId %in%currModuleIds]
        currSubsystems = unique(currSubsystems)
        return(getStrCatBy(currSubsystems))
      }) )
      keggEnPathways = do.call(c, lapply(as.list(nodes), function(x) {
        currEnriched <- keggEnrichPathMelt[keggEnrichPathMelt$funcModule == x,]
        if (dim(currEnriched)[1] == 0 ) return("")
        return(getStrCatBy(currEnriched$pathway))
      }) )
      funcModNetAttachNodeTable = funcModNetNodeTable
      funcModNetAttachNodeTable$hasAntismash = antismashCheck
      funcModNetAttachNodeTable$hasVirulence = vfCheck
      funcModNetAttachNodeTable$hasAR = arCheck
      funcModNetAttachNodeTable$hasJgi = jgiCheck
      funcModNetAttachNodeTable$hasCazy = cazyCheck
      funcModNetAttachNodeTable$hasPfam = pfamCheck
      funcModNetAttachNodeTable$hasKegg = keggCheck
      
      funcModNetAttachNodeTable$antismahsTerms = antismashIds
      funcModNetAttachNodeTable$virulenceTerms = vfIds
      funcModNetAttachNodeTable$virulenceGenes = vfGenes
      funcModNetAttachNodeTable$virulenceProducts = vfProducts
      funcModNetAttachNodeTable$virulenceClasses = vfClasses
      funcModNetAttachNodeTable$ARTerms = arIds
      funcModNetAttachNodeTable$jgiTerms = jgiIds
      funcModNetAttachNodeTable$cazyTerms = cazyIds
      funcModNetAttachNodeTable$pfamTerms = pfamIds
      funcModNetAttachNodeTable$keggTerms = keggIds
      funcModNetAttachNodeTable$keggDescs = keggDescs
      funcModNetAttachNodeTable$keggGenes = keggGenes
      funcModNetAttachNodeTable$keggModules = keggEnModules
      funcModNetAttachNodeTable$keggSubsystems = keggEnModSubSystems
      funcModNetAttachNodeTable$keggPathway = keggEnPathways
      
      return(funcModNetAttachNodeTable)
      
    }
    
    
    funcModNetAttachNodeTable = attachInfo(funcModNetNodeTable, 
                                           funcModules, 
                                           antismashTerms, 
                                           vfTerms, 
                                           arTerms, 
                                           jgiTerms,
                                           cazyTerms,
                                           pfamTerms, 
                                           koTerms,
                                           koDescMap,
                                           moduleDescTab,
                                           patricVfDescTab)
    
    
    
    
    writeMode = F
    if (writeMode) write.table(funcModNetAttachNodeTable, funcModNetAttachedNodeTableFile, row.names = F, quote = F, sep="\t")
    
  } 
  
  countMode = T
  if (countMode) {
    pfamCheck = unlist(lapply(funcModules, function(x) any(x %in% pfamTerms)))
    koCheck = unlist(lapply(funcModules, function(x) any(x %in% koTerms)))
    antismashCheck = unlist(lapply(funcModules, function(x) any(x %in% antismashTerms)))
    cazyCheck = unlist(lapply(funcModules, function(x) any(x %in% cazyTerms)))
    vfCheck = unlist(lapply(funcModules, function(x) any(x %in% vfTerms)))
    arCheck = unlist(lapply(funcModules, function(x) any(x %in% arTerms)))
    jgiCheck = unlist(lapply(funcModules, function(x) any(x %in% jgiTerms)))
    
    table(pfamCheck)
    table(koCheck)
    table(antismashCheck)
    table(cazyCheck)
    table(vfCheck)
    table(jgiCheck)
    
    pfamClusters = names(funcModules)[pfamCheck]
    koClusters = names(funcModules)[koCheck]
    cazyClusters = names(funcModules)[cazyCheck]
    antismashClusters = names(funcModules)[antismashCheck]
    vfClusters = names(funcModules)[vfCheck]
    jgiClusters = names(funcModules)[jgiCheck]
    arClusters = names(funcModules)[arCheck]
    
    table(pfamClusters %in% singletons)
    table(koClusters %in% singletons)
    table(cazyClusters %in% singletons)
    table(antismashClusters %in% singletons)
    table(vfClusters %in% singletons)
    table(jgiClusters %in% singletons)
    table(arClusters %in% singletons)
    
    fcStatTab = data.frame(stringsAsFactors=FALSE,
                           Number = c(2901, 1204, 2957, 1197, 153, 162, 9, 9, 249, 92, 14, 6, 14, 2),
                           Category = c("PFAM", "PFAM", "KEGG", "KEGG", "CAZy", "CAZy", "antiSMASH",
                                        "antiSMASH", "virulence", "virulence", "JGI-GOLD",
                                        "JGI-GOLD", "AMR", "AMR"),
                           Type = c("Singleton", "non-singleton", "Singleton", "non-singleton",
                                    "Singleton", "non-singleton", "Singleton",
                                    "non-singleton", "Singleton", "non-singleton", "Singleton",
                                    "non-singleton", "Singleton", "non-singleton")
    )
    fcStatTab$Category = factor(fcStatTab$Category, levels = c("KEGG", "PFAM", "virulence","CAZy", "JGI-GOLD","antiSMASH","AMR"))
    
    ggplot(fcStatTab[fcStatTab$Category%in%c("KEGG","PFAM","virulence", "CAZy"),], aes(x=Category, fill=Type, y=Number)) +geom_bar(stat="identity")
    ggplot(fcStatTab[!fcStatTab$Category%in%c("KEGG","PFAM","virulence","CAZy"),], aes(x=Category, fill=Type, y=Number)) +geom_bar(stat="identity")
    
    
  }
  
  keggModuleVsFunctionalCluster = T
  if (keggModuleVsFunctionalCluster) {
    
    funcJaccMatRData = "J://Deposit/Project/2018_microbiome_atlas//functional.annotation//funcJaccMat.20191227.RData"
    load(funcJaccMatRData)
    
    mucinCazymes = c("cazy.GT27", "cazy.GT14", "cazy.GT11", "cazy.GT10", "cazy.GH89", "cazy.GH84", "cazy.GH33", "cazy.GH20", "cazy.GH29", "cazy.GH18", "cazy.GH123", "cazy.GH109", "cazy.CBM51", "cazy.CBM50", "cazy.CBM32")
    storageCazymes = c("cazy.CBM34", "cazy.CBM48", "cazy.CBM20", "cazy.GH13", "cazy.GH133", "cazy.GH32", "cazy.GH36", "cazy.GH37", "cazy.GH77", "cazy.GH57", "cazy.GT101", "cazy.GT26", "cazy.GT3", "cazy.GT35", "cazy.GT5")
    pectinCazymes = c("cazy.CBM77","cazy.CE8","cazy.GH105","cazy.GH28", "cazy.PL1")
    
    koMicrobeModuleList = koModuleList[microbeModules]
    koMicrobeModuleNums = unlist(lapply(koMicrobeModuleList, length))
    koMicrobeModuleList = koMicrobeModuleList[koMicrobeModuleNums>=5]
    
    moduleJaccList = lapply(koMicrobeModuleList, function(kos) {
      currJaccMat = funcJaccMat[rownames(funcJaccMat)%in% kos,rownames(funcJaccMat)%in% kos]
      currJaccs = currJaccMat[lower.tri(currJaccMat)]
      return(currJaccs)
    })
    
    pathJaccList = lapply(bacteriaKoElementList, function(kos) {
      currJaccMat = funcJaccMat[rownames(funcJaccMat)%in% kos,rownames(funcJaccMat)%in% kos]
      currJaccs = currJaccMat[lower.tri(currJaccMat)]
      return(currJaccs)
    })
    
    fcJaccList = lapply(funcModules, function(kos) {
      currJaccMat = funcJaccMat[rownames(funcJaccMat)%in% kos,rownames(funcJaccMat)%in% kos]
      currJaccs = currJaccMat[lower.tri(currJaccMat)]
      return(currJaccs)
    })
    
    
    normalClusters = names(funcModules)[!names(funcModules) %in% c(smallFcs, signletons)]
    normFuncModules = funcModules[normalClusters]
    
    exModules = c("M00161:Photosystem II ", "M00163:Photosystem I ")
    
    moduleJaccMeans = unlist(lapply(moduleJaccList[!names(moduleJaccList) %in% exModules], mean))
    pathJaccMeans = unlist(lapply(pathJaccList, mean))
    fcJaccMeans = unlist(lapply(fcJaccList[normalClusters], mean))
    
    jaccCompList = list(module=moduleJaccMeans,
                        pathway=pathJaccMeans,
                        fc=fcJaccMeans)
    jaccCompMelt = melt(jaccCompList)
    jaccCompMelt[,2]=as.character(jaccCompMelt[,2])
    colnames(jaccCompMelt) = c("jacc","category")
    jaccCompMelt$category[jaccCompMelt$category=="fc"] = "Functional cluster"
    jaccCompMelt$category[jaccCompMelt$category=="module"] = "KEGG module"
    jaccCompMelt$category[jaccCompMelt$category=="pathway"] = "KEGG pathway"
    jaccCompMelt$category = factor(jaccCompMelt$category, levels = c("KEGG pathway", "KEGG module", "Functional cluster"))
    
    p=ggplot(jaccCompMelt) + geom_boxplot(aes(x=category,y=jacc)) + 
      scale_fill_manual(values = c("white", "white", "white")) +
      xlab("") + ylab("") +
      theme(panel.grid.major = element_blank(), 
            panel.grid.minor = element_blank(),
            axis.text.x = element_text(angle = 60, hjust=1),
            axis.ticks = element_blank(),
            legend.position = "none",
            panel.background = element_rect(fill = "white", 
                                            colour = "black", 
                                            size = 0.5, 
                                            linetype = "solid")) 
    #ggsave("jacc.kegg.functional.cluster.pdf", p, width=2.35, heigth=3.29, units = "in")
    
    
    boxplot(jaccCompList, las=1)
    
    
    jaccOutList = lapply(koMicrobeModuleList, function(kos) {
      currJaccMat = funcJaccMat[rownames(funcJaccMat)%in% kos,rownames(funcJaccMat)%in% kos]
      currJaccs = currJaccMat[lower.tri(currJaccMat)]
      return(currJaccs)
    })
    jaccOutNums = unlist(lapply(jaccOutList, length))
    jaccOutMedians = unlist(lapply(jaccOutList, median))
    jaccOutMeans = unlist(lapply(jaccOutList, mean))
    par(mar=c(5,20,4,1))
    boxplot(jaccOutList[names(sort(jaccOutMedians,T))[1:20]], cex=0.8, horizontal=T,las=1)
    
    hist(jaccOutMedians)
    hist(jaccOutMeans)
    
    targetModule = koMicrobeModuleList$`M00126:Tetrahydrofolate biosynthesis, GTP => THF `
    targetModule = koMicrobeModuleList$`M00161:Photosystem II `
    tt = unlist(lapply(funcModules, function(x) any(x %in% targetModule)))
    tt[tt]
    targetModule[targetModule %in% funcModules$`49`]
    
    boxplot(jaccOutList$`M00126:Tetrahydrofolate biosynthesis, GTP => THF `)
    
    View(jaccOutMedians)
    
    targetModule = koModuleList$`M00023:Tryptophan biosynthesis, chorismate => tryptophan `
    View(funcJaccMat[rownames(funcJaccMat) %in% targetModule, colnames(funcJaccMat) %in% targetModule])
    
    
    targetList = list(LCarnintine=funcModules$`143`,
                      Cholin=funcModules$`4071`,
                      TMAO=funcModules$`236`)
    
    targetList = list(LCarnintine=funcModules$`143`,
                      Butyrate=funcModules$`332`,
                      Acetogenesis=c(funcModules$`361`, funcModules$`354`),
                      #AcetateWL_2=funcModules$`354`,
                      BCAA=funcModules$`12`, # folate, BCAA
                      #Methan=funcModules$`20`,
                      mucin_sialidase=funcModules$`303`,
                      #Cholin=funcModules$`4071`,
                      TMAO=funcModules$`236`,
                      Propioante1=funcModules$`4`
                      #,Propioante2=funcModules$`10`
                      
    )
    targetNames = c("L-carnitine", "Butyrate", "Acetogenesis",  "BCAA", "Host-mucin(sialidase)", "Choline", "TMAO", "Propionate")
    
    # Bile1=funcModules$`12`,
    # Bile2=funcModules$`2908`,
    # Bile3=funcModules$`4534`
    
    mspByTargetModList =grepMspsAssociatedWithModules(funcModules = targetList,
                                                      funcMat = funcMat[!rownames(funcMat) %in% suppressMsps2, ], 
                                                      cutRatio = 0.75)
    
    
    
    
    
    
    
    getCircosInputForFunctionCluster <- function(mspByTargetModList, targetNames,  taxo, mode = NULL, cutRatio=0.75) {
      names(mspByTargetModList) = targetNames
      outTab = melt(mspByTargetModList)
      rownames(outTab) = paste(outTab[,1], outTab[,2])
      #return(outTab)
      outTab$species = getSpeciesName(outTab[,1], taxo)
      outTab$genus = getGenusName(outTab[,1], taxo)
      outTab$phylum = getPhylaName(outTab[,1], taxo)
      if (!is.null(mode)) {
        outTab = outTab[,c("L1",mode)] # genus, species, phylum
        colnames(outTab)=c("from","to")
        outTab = outTab[outTab$to != "unclassified",]
        outTab = outTab[!grepl("unclassified",outTab$to),]
      }
      
      return(outTab)
      
    }
    
    circosOut = getCircosInputForFunctionCluster(mspByTargetModList, targetNames, taxo , "phylum", 0.75)
    View(circosOut)
    if (F) {
      
      elements = (unique(c(circosOut$from, circosOut$to)))
      lenElements = length(elements)
      
      circosGridCols = (col_vector)[1:length(elements)]
      names(circosGridCols) = elements
      
      #circos.par(track.margin=c(0,0)) 
      set.seed(1)
      #circos.initialize(elements, xlim=c(0,1))
      #chordDiagram(circosOut)
      chordDiagram(circosOut, annotationTrack = "grid", preAllocateTracks = 1, grid.col = circosGridCols)
      circos.trackPlotRegion(track.index = 1,  panel.fun = function(x, y) {
        xlim = get.cell.meta.data("xlim")
        ylim = get.cell.meta.data("ylim")
        sector.name = get.cell.meta.data("sector.index")
        circos.text(mean(xlim), ylim[1] + .1, cex = 0.5,  sector.name, facing = "clockwise", niceFacing = TRUE, adj = c(0, 0.5))
        circos.axis(h = "top", labels = F, major.tick = F, 
                    sector.index = sector.name, track.index = 2)
      }, bg.border = NA)
      circos.clear() 
      
    }    
    
    
    funcModuleNums[1:5]
    nums = sort(mspByModNums[!names(mspByModNums) %in% signletons & !names(mspByModNums) %in% smallFcs],T)
    barplot(nums[nums>5])
    numStatTab = data.frame(clusterName = names(nums),
                            numSpecies=nums,
                            clusterSize=funcModuleNums[match(names(nums), names(funcModuleNums))],
                            inflow=(inflowsOfMspByModList)[match(names(nums), names(inflowsOfMspByModList))],
                            outflow=(outflowsOfMspByModList)[match(names(nums), names((outflowsOfMspByModList)))],
                            stringsAsFactors = F)
    numStatTab = numStatTab[numStatTab$numSpecies!=0, ]
    
    plot(numStatTab$inflow, numStatTab$numSpecies)
    plot(numStatTab$outflow, numStatTab$numSpecies)
    
    plot(numStatTab$inflow, numStatTab$clusterSize, log="y")
    plot(numStatTab$outflow, numStatTab$clusterSize, log="y")
    
    ggplot(numStatTab, aes(x=numSpecies, y=clusterSize)) + 
      geom_point(aes(fill=numSpecies, size=clusterSize), shape=21) +
      xlab("Number of enriched species") + ylab("Cluster size") +
      xlim(-10,650)+
      scale_size(range = c(2, 10))+ 
      coord_trans(y="log10")+
      #scale_fill_gradient2(mid="#ffffff88",  high = "#6a0dad88") +
      scale_fill_gradient2(mid="#ffffff88",  high = "#964B00AA") +
      theme(panel.grid.minor = element_line(colour="#80808044", size=0.5),
            panel.grid.major = element_line(colour="#80808044", size=0.5),
            legend.position = "none",
            panel.background = element_rect(fill = "white", 
                                            colour = "black", 
                                            size = 0.5, 
                                            linetype = "solid")) 
    
    inflowCut=quantile(inflow, 0.9)
    outflowCut=quantile(outflow, 0.9)
    
    ggplot(numStatTab, aes(x=numSpecies, y=clusterSize)) + 
      #geom_point(aes(fill=numSpecies, size=clusterSize), shape=21) +
      geom_point(aes(fill=inflow >= inflowCut, size=clusterSize), shape=21) +
      xlab("Number of enriched species") + ylab("Cluster size") +
      xlim(-10,650)+
      scale_size(range = c(2, 10))+ 
      coord_trans(y="log10")+
      #scale_fill_gradient2(mid="#ffffff88",  high = "#6a0dad88") +
      scale_fill_manual(name = 'inflow > top-10%', values = setNames(c('blue','#ffffff11'),c(T, F))) +
      theme(panel.grid.minor = element_line(colour="#80808044", size=0.5),
            panel.grid.major = element_line(colour="#80808044", size=0.5),
            legend.position = "none",
            panel.background = element_rect(fill = "white", 
                                            colour = "black", 
                                            size = 0.5, 
                                            linetype = "solid")) 
    
    ggplot(numStatTab, aes(x=numSpecies, y=clusterSize)) + 
      #geom_point(aes(fill=numSpecies, size=clusterSize), shape=21) +
      geom_point(aes(fill= outflow >= outflowCut, size=clusterSize), shape=21) +
      xlab("Number of enriched species") + ylab("Cluster size") +
      xlim(-10,650)+
      scale_size(range = c(2, 10))+ 
      coord_trans(y="log10")+
      #scale_fill_gradient2(mid="#ffffff88",  high = "#6a0dad88") +
      scale_fill_manual(name = 'outflow > top-10%', values = setNames(c('red','#ffffff11'),c(T, F))) +
      theme(panel.grid.minor = element_line(colour="#80808044", size=0.5),
            panel.grid.major = element_line(colour="#80808044", size=0.5),
            legend.position = "none",
            panel.background = element_rect(fill = "white", 
                                            colour = "black", 
                                            size = 0.5, 
                                            linetype = "solid")) 
    
    plot(nums, funcModuleNums[match(names(nums), names(funcModuleNums))], 
         xlab="number of enriched species", 
         ylab="cluster size", log="y")
    
    
  }
  
  hgclgcList = list(hgc=hgcSpecies, 
                    lgc=lgcSpecies)
  
  hgclgcFuncAssocMat = getEnrichMentGsByGs(hgclgcList, mspByModList)
  hgclgcFuncAssocTab = data.frame(name=rownames(hgclgcFuncAssocMat), hgclgcFuncAssocMat)
  
  hgclgcFuncAssocMat = getEnrichMentGsByGs(hgclgcList, mspByTargetModList)
  
  
  getFuncMspMapping = T
  if (getFuncMspMapping) {
    mspByModList =grepMspsAssociatedWithModules(funcModules = funcModules,
                                                funcMat = funcMat[!rownames(funcMat) %in% suppressMsps2, ], 
                                                cutRatio = 0.75)
    inflowOutflowMode = F
    if (inflowOutflowMode) {
      inflowsOfMspByModList = do.call(c, lapply(mspByModList[normalClusters], function(x) mean(inflow[names(inflow)%in% x])))
      outflowsOfMspByModList = do.call(c, lapply(mspByModList[normalClusters], function(x) mean(outflow[names(outflow)%in% x])))
      
    }
    
    checkInflowOutflowEnrichment = T
    if (checkInflowOutflowEnrichment) {
      
      inflowEnrichment = getEnrichMentGsByList(mspByModList, inflowSpecies)
      outflowEnrichment = getEnrichMentGsByList(mspByModList, outflowSpecies)
      enrichmentTab = data.frame(id = names(mspByModList),
                                 inflowP = inflowEnrichment, 
                                 outflowP = outflowEnrichment, 
                                 stringsAsFactors = F)
      enrichmentNormTab = enrichmentTab[enrichmentTab$id %in% normalClusters,]
      
      
    }
    
    inflowOutflowNormalClusters = data.frame(name=names(inflowsOfMspByModList),
                                             inflow=(inflowsOfMspByModList), 
                                             outflow=(outflowsOfMspByModList),
                                             stringsAsFactors = F)
    mspByModNums = do.call(c,lapply(mspByModList, length))
    
    checkSpecifcity<- function(msps, taxo, target) {
      targetTaxa = taxo[taxo$MSP %in% msps,target]
      targetTaxa = targetTaxa[targetTaxa!="unclassified"]
      targetTaxa = targetTaxa[!grepl("unclassified",targetTaxa)]
      targetTaxaUniq = unique(targetTaxa)
      if (length(targetTaxaUniq)==1) return(targetTaxaUniq)
      return(NA)
    }
    
    checkSpecificies = unlist(lapply(mspByModList, function(x) checkSpecifcity(x, taxo, "family")))
    sort(table(checkSpecificies))
    checkSpecificies = checkSpecificies[!is.na(checkSpecificies)]
    checkSpecificies[checkSpecificies=="Prevotellaceae"]
    checkSpecificies[checkSpecificies=="Spirochaetaceae"]
    checkSpecificies[checkSpecificies=="Enterococcaceae"]
    
    checkSpecificies = unlist(lapply(mspByModList, function(x) checkSpecifcity(x, taxo, "class")))
    sort(table(checkSpecificies))
    checkSpecificies = checkSpecificies[!is.na(checkSpecificies)]
    checkSpecificies[checkSpecificies=="Verrucomicrobiae"]
    
    
    checkSpecificies = unlist(lapply(mspByModList, function(x) checkSpecifcity(x, taxo, "genus")))
    table(checkSpecificies)
    checkSpecificies = checkSpecificies[!is.na(checkSpecificies)]
    checkSpecificies[checkSpecificies=="Prevotella"]
    checkSpecificies[checkSpecificies=="Bacteroides"]
    checkSpecificies[checkSpecificies=="Veillonella"]
    checkSpecificies[checkSpecificies=="Streptococcus"]
    checkSpecificies[checkSpecificies=="Sphaerochaeta"]
    
    #mspByModList =grepMspsAssociatedWithModules(funcModules = funcModules,funcMat = funcMat, cutRatio = 0.95)
    mspNamesByModList = lapply(mspByModList, function(x) getSpeciesName(x, taxo))
    
    bestMspByModList = grepBestMspsAssociatedWithModules(funcModules = funcModules, funcMat = funcMat)
    bestScoresMspByModList = grepBestScoresMspsAssociatedWithModules(funcModules = funcModules, funcMat = funcMat)
    bestScoresMspByModList = unlist(bestScoresMspByModList)
    bestMspNamesByModList = lapply(bestMspByModList, function(x) getSpeciesName(x, taxo))
    
    mspByModSelList =grepMspsAssociatedWithModules(funcModules = funcModuleSel,funcMat = funcMat, cutRatio = 0.75)
    mspBySecMetList = grepMspsAssociatedWithModules(funcModules = funcModules[antismashClusters],funcMat = funcMat, cutRatio = 0.75)
    
    mspByModMelt = melt(mspByModList)
    mspByModMelt$value = as.character(mspByModMelt$value)
    mspByModMelt$L1 = as.character(mspByModMelt$L1)
    moduleByMspList = split(mspByModMelt[,2], mspByModMelt[,1])
    moduleNumsByMsp = do.call(c, lapply(moduleByMspList, length))
    
    sort(moduleNumsByMsp)[1:5]
    sort(moduleNumsByMsp, T)[1:5]
    
    crisprModules = c("1440","20","251","315","339","411","415","443","445","48","57","633","90","93")
    competenceModules = c("101","105","12","128","14","383","606","63","64","8")
    
    
    if (F) {
      funcModules$`332`
      funcModules$`143`
      funcModules$`236`
      funcModules$`4071`
      funcModules$`3206`
      funcModules$`354`
      funcModules$`332`
      
      K08298 
      
      tt = unlist(lapply(funcModules, function(x) any(x=="K01687"))) # 
      
      tt = unlist(lapply(funcModules, function(x) any(x=="cazy.GH33"))) #303, sialidase
      
      tt = unlist(lapply(funcModules, function(x) any(x=="K11264"))) #propanoyl-coA
      tt = unlist(lapply(funcModules, function(x) any(x=="K18426"))) #propanoyl-coA
      
      tt = unlist(lapply(funcModules, function(x) any(x=="K01847"))) #propanoyl-coA
      tt = unlist(lapply(funcModules, function(x) any(x=="K01026"))) #propanoyl-coA
      
      tt = unlist(lapply(funcModules, function(x) any(x=="K13922"))) #propionaldehyde
      
      
      tt = unlist(lapply(funcModules, function(x) any(x=="K15023"))) #361, acetogen
      tt = unlist(lapply(funcModules, function(x) any(x=="K00198"))) #354, acetogen
      tt = unlist(lapply(funcModules, function(x) any(x=="K01067"))) #4, pyruvate acetate  
      
      tt = unlist(lapply(funcModules, function(x) any(x=="K08299"))) #143, L-carnitine degradation
      tt = unlist(lapply(funcModules, function(x) any(x=="K08298"))) #3206
      
      tt = unlist(lapply(funcModules, function(x) any(x=="K18277")))# None
      tt = unlist(lapply(funcModules, function(x) any(x=="K22444")))# None # l-carnitine --> TMA
      tt = unlist(lapply(funcModules, function(x) any(x=="K07811")))#236 # TMA --> TMAO
      
      tt = unlist(lapply(funcModules, function(x) any(x=="K20038"))) #4071 # choline trimethylamine-lyase; choline --> TMA
      tt = unlist(lapply(funcModules, function(x) any(x=="Rieske")))#1086 # carnitine monooxygenase motif; carnitine --> TMA  
      tt = unlist(lapply(funcModules, function(x) any(x=="Rieske_2")))#10 # carnitine monooxygenase motif; carnitine --> TMA  
      tt = unlist(lapply(funcModules, function(x) any(x=="Ring_hydroxyl_A")))#10 # carnitine monooxygenase motif; carnitine --> TMA  
      tt = unlist(lapply(funcModules, function(x) any(x=="K01034")))
      tt = unlist(lapply(funcModules, function(x) any(x=="K20811"))) #inulin, inulosucrase
      
      tt = unlist(lapply(funcModules, function(x) any(x=="K01667"))) #tryptophan --> indole
      tt = unlist(lapply(funcModules, function(x) any(x=="K01905"))) #  acetate
      names(tt[tt])
      
      tt = unlist(lapply(funcModules, function(x) any(x=="K00627"))) #  acetate
      names(tt[tt])
      
      tt = unlist(lapply(funcModules, function(x) any(x=="K20626"))) #  propionate lactate-pathway
      tt = unlist(lapply(funcModules, function(x) any(x=="K00248"))) #  propionate lactate-pathway
      
      K00837
      
      tt = unlist(lapply(funcModules, function(x) any(x %in% mucinCazymes))) #12, 117, 294, 303,  ...
      tt = unlist(lapply(funcModules, function(x) any(x %in% mgePfamTerms))) #10, 12, 48, 51, ...
      tt = unlist(lapply(funcModules, function(x) any(x %in% bileAcidKeggs))) #12, 2908, 4534
      
      bileAcidKeggs[bileAcidKeggs %in% funcModules[["12"]] ]
      bileAcidKeggs[bileAcidKeggs %in% funcModules[["2908"]] ]
      bileAcidKeggs[bileAcidKeggs %in% funcModules[["4534"]] ]
      
      names(tt[tt])
      funcModules[names(tt[tt])]
      
      sort(bestMspNamesByModList$`12`)
      sort(bestMspNamesByModList$`117`)
      
      sort(bestMspNamesByModList$`53`)
      sort(bestMspNamesByModList$`132`)
      
      sort(bestMspNamesByModList$`143`)
      sort(bestMspNamesByModList$`332`)
      sort(bestMspNamesByModList$`236`)
      sort(bestMspNamesByModList$`1086`)
      sort(bestMspNamesByModList$`4071`) # 
      sort(bestMspNamesByModList$`303`) # 
      sort(bestMspNamesByModList$`4479`) # 
      sort(bestMspNamesByModList$`2288`) # acetate
      sort(bestMspNamesByModList$`2283`) # indole
      sort(bestMspNamesByModList$`1321`)
      
      sort(mspNamesByModList$`12`) # mucin-degrader??
      sort(mspNamesByModList$`143`)
      sort(mspNamesByModList$`332`)
      sort(mspNamesByModList$`236`)
      sort(mspNamesByModList$`3206`)
      sort(mspNamesByModList$`4071`)
      sort(mspNamesByModList$`1086`)
      
      sort(mspNamesByModList$`494`)
      sort(mspNamesByModList$`4`)
      
      funcModules$`152`
      funcModules$`236`
      funcModules$`2283`
      funcModules$`2288`
      funcModules$`1322`
    }
    
    
    
  }
  
  associateModuleMspMode = T
  if (associateModuleMspMode) {
    
    
    loadPair = T
    if (loadPair) {
      diseaseWithCountryPairTab = rbind(c("Atherosclerosis_SW_id1", "Sweden"),
                                        c("CVD_CN_id10", "China"),
                                        c("GDM_CN_id12", "China"),
                                        c("T2D_CN_id2", "China"),
                                        c("T1D_FI_id21", "Finland"),
                                        c("Cirrhosis_UK_id27", "UK"),
                                        c("CFS_US_id34", "US"),
                                        c("NAFLD_IT_id40", "Italy"),
                                        c("NAFLD_ES_id40", "Spain"),
                                        c("CRC_JP_id45", "Japan"),
                                        c("CRC_IT_id46", "Italy"),
                                        c("Behcet_CN_id52", "China"),
                                        c("CD_CN_id6", "China"),
                                        c("Ankylosing_CN_id9", "China"),
                                        c("CRC_US_id11", "US"),
                                        c("T2D_SW_id14", "Sweden"),
                                        c("IGT_SW_id14", "Sweden"),
                                        c("Obesity_DK_id16", "Denmark"),
                                        c("IBD_ES_id17", "Spain"),
                                        c("Cirrhosis_CN_id20", "China"),
                                        c("Obesity_DK_id25", "Denmark"),
                                        c("NSCLC_FR_id26", ""), 
                                        c("Melanoma_US_id3", "US"),
                                        c("RA_CN_id31", "China"),
                                        c("T1D_LU_id35", "Luxembourg"),
                                        c("RCC_FR_id41", ""),
                                        c("CRC_DE_id44", "Germany"),
                                        c("T2D_ES_id53", "Spain"),
                                        c("NAFLD_US_id7", "US"),
                                        c("PD_DE_id8", "Germany"))
      
      diseaseWithCountryPairTab2 = rbind(c("Atherosclerosis_SW_id1", "Sweden"),
                                         c("CVD_CN_id10", "China"),
                                         c("GDM_CN_id12", "China"),
                                         c("T2D_CN_id2", "China"),
                                         c("T1D_FI_id21", "Finland"),
                                         c("Cirrhosis_UK_id27", "UK"),
                                         c("CFS_US_id34", "US"),
                                         c("NAFLD_IT_id40", "Italy"),
                                         c("NAFLD_ES_id40", "Spain"),
                                         c("CRC_JP_id45", "Japan"),
                                         c("CRC_IT_id46", "Italy"),
                                         c("Behcet_CN_id52", "China"),
                                         c("CD_CN_id6", "China"),
                                         c("Ankylosing_CN_id9", "China"),
                                         c("CRC_US_id11", "US"),
                                         c("T2D_SW_id14", "Sweden"),
                                         c("IGT_SW_id14", "Sweden"),
                                         c("Obesity_DK_id16", "Denmark"),
                                         c("IBD_ES_id17", "Spain"),
                                         c("Cirrhosis_CN_id20", "China"),
                                         c("Obesity_DK_id25", "Denmark"),
                                         c("NSCLC_FR_id26", "Germany"), 
                                         c("Melanoma_US_id3", "US"),
                                         c("RA_CN_id31", "China"),
                                         c("T1D_LU_id35", "Luxembourg"),
                                         c("RCC_FR_id41", "Germany"),
                                         c("CRC_DE_id44", "Germany"),
                                         c("T2D_ES_id53", "Spain"),
                                         c("NAFLD_US_id7", "US"),
                                         c("PD_DE_id8", "Germany"))
      
      diseaseWithEtPairTab = rbind(c("Atherosclerosis_SW_id1", "etFirmicutes"),
                                   c("CVD_CN_id10", "etBacteroides"),
                                   c("GDM_CN_id12", "etBacteroides"),
                                   c("T2D_CN_id2", "etBacteroides"),
                                   c("T1D_FI_id21", "etFirmicutes"),
                                   c("Cirrhosis_UK_id27", "etFirmicutes"),
                                   c("CFS_US_id34", "etBacteroides"),
                                   c("NAFLD_IT_id40", "etFirmicutes"),
                                   c("NAFLD_ES_id40", "etFirmicutes"),
                                   c("CRC_JP_id45", "etBacteroides"),
                                   c("CRC_IT_id46", "etFirmicutes"),
                                   c("Behcet_CN_id52", "etBacteroides"),
                                   c("CD_CN_id6", "etBacteroides"),
                                   c("Ankylosing_CN_id9", "etBacteroides"),
                                   c("CRC_US_id11", "etBacteroides"),
                                   c("T2D_SW_id14", "etFirmicutes"),
                                   c("IGT_SW_id14", "etFirmicutes"),
                                   c("Obesity_DK_id16", "etFirmicutes"),
                                   c("IBD_ES_id17", "etFirmicutes"),
                                   c("Cirrhosis_CN_id20", "etBacteroides"),
                                   c("Obesity_DK_id25", "etFirmicutes"),
                                   c("NSCLC_FR_id26", "etFirmicutes"), 
                                   c("Melanoma_US_id3", "etBacteroides"),
                                   c("RA_CN_id31", "etBacteroides"),
                                   c("T1D_LU_id35", "etFirmicutes"),
                                   c("RCC_FR_id41", "etFirmicutes"),
                                   c("CRC_DE_id44", "etFirmicutes"),
                                   c("T2D_ES_id53", "etFirmicutes"),
                                   c("NAFLD_US_id7", "etBacteroides"),
                                   c("PD_DE_id8", "etFirmicutes"))
      
      diseaseWithMatchedPairTab = rbind(c("Atherosclerosis_SW_id1", "Sweden_id1"),
                                        c("CVD_CN_id10", "China_id10"),
                                        c("GDM_CN_id12", "China_id12"),
                                        c("T2D_CN_id2", "China_id2"),
                                        c("T1D_FI_id21", "Finland_id21"),
                                        c("Cirrhosis_UK_id27", ""),
                                        c("CFS_US_id34", "US_id34"),
                                        c("NAFLD_IT_id40", ""),
                                        c("NAFLD_ES_id40", ""),
                                        c("CRC_JP_id45", "Japan_id45"),
                                        c("CRC_IT_id46", "Italy_id46"),
                                        c("Behcet_CN_id52", "China_id52"),
                                        c("CD_CN_id6", ""),
                                        c("Ankylosing_CN_id9", "China_id9"),
                                        c("CRC_US_id11", "US_id11"),
                                        c("T2D_SW_id14", "Sweden_id14"),
                                        c("IGT_SW_id14", "Sweden_id14"),
                                        c("Obesity_DK_id16", "Denmark_id16"),
                                        c("IBD_ES_id17", "Spain_id17"),
                                        c("Cirrhosis_CN_id20", "China_id20"),
                                        c("Obesity_DK_id25", "Denmark_id25"),
                                        c("NSCLC_FR_id26", ""), 
                                        c("Melanoma_US_id3", ""),
                                        c("RA_CN_id31", "China_id31"),
                                        c("T1D_LU_id35", "Luxembourg_id35"),
                                        c("RCC_FR_id41", ""),
                                        c("CRC_DE_id44", "Germany_id44"),
                                        c("T2D_ES_id53", ""),
                                        c("NAFLD_US_id7", ""),
                                        c("PD_DE_id8", "Germany_id8"))
      diseaseWithCountryPairTab = data.frame(diseaseWithCountryPairTab, stringsAsFactors = F)
      diseaseWithMatchedPairTab = data.frame(diseaseWithMatchedPairTab, stringsAsFactors = F)
      diseaseWithEtPairTab = data.frame(diseaseWithEtPairTab, stringsAsFactors = F)
      
      colnames(diseaseWithCountryPairTab) = c("reference","target")
      colnames(diseaseWithMatchedPairTab) = c("reference","target")
      colnames(diseaseWithEtPairTab) = c("reference","target")
      
    }
    
    loadDiseaseSigMode = T
    if (loadDiseaseSigMode) {
      
      allPairMeltRData = "C://Data//comparative.analysis.healthy.sweden//allPairMelt.disease.signatures.RData"
      load(allPairMeltRData)
      allPairDownMeltRData = "C://Data//comparative.analysis.healthy.sweden//allPairDownMelt.disease.signatures.RData"
      load(allPairDownMeltRData)
      
      cohortOrderFile = "C://Data//comparative.analysis.healthy.sweden//disease.cohort.order.txt"
      cohortOrders = read.table(cohortOrderFile, stringsAsFactors = F, header=F)[,]
      names(cohortOrders)=1:30
      
      allPairMeltImputed = allPairMelt
      #allPairMeltImputed = allPairMeltImputed[!allPairMeltImputed$disease_cohort %in% c("NSCLC_FR_id26", "RCC_FR_id41", "T1D_FI_id21"),]
      allPairMeltImputed$effect_size[allPairMeltImputed$effect_size < 0] = 0
      allPairMeltImputed$effect_size[allPairMeltImputed$effect_size > 1] = 1
      allPairMeltImputed$effect_size[is.na(allPairMeltImputed$effect_size)] = 0
      
      manhattanMode = T
      if (manhattanMode) {
        uniqMsp = unique(allPairMeltImputed$msp) 
        uniqMspId = 1:length(uniqMsp)
        names(uniqMspId) =uniqMsp
        
        allPairMeltImputed$CHR = names(cohortOrders)[match(allPairMeltImputed$disease_cohort, cohortOrders)]
        allPairMeltImputed$CHR = as.numeric(allPairMeltImputed$CHR)
        #allPairMeltImputed$CHR = as.numeric(gsub(".*_id", "", allPairMeltImputed$disease_cohort))
        allPairMeltImputed$BP = uniqMspId[match(allPairMeltImputed$msp, names(uniqMspId))]
        allPairMeltImputed$P = allPairMeltImputed$effect_size
        allPairMeltImputed$SNP = allPairMeltImputed$msp
        
      }
      
      allPairMeltImputedMatched = allPairMeltImputed[allPairMeltImputed$control=="matched",]
      allPairMeltImputedCountry = allPairMeltImputed[allPairMeltImputed$control=="country",]
      allPairMeltImputedCluster = allPairMeltImputed[allPairMeltImputed$control=="cluster",]
      
      allPairMeltImputedMatchedSig = allPairMeltImputedMatched[allPairMeltImputedMatched$effect_size>=0.3,]
      allPairMeltImputedCountrySig = allPairMeltImputedCountry[allPairMeltImputedCountry$effect_size>=0.3,]
      allPairMeltImputedClusterSig = allPairMeltImputedCluster[allPairMeltImputedCluster$effect_size>=0.3,]
      
      
      allPairDownMeltImputed = allPairDownMelt
      #allPairDownMeltImputed = allPairDownMeltImputed[!allPairDownMeltImputed$disease_cohort %in% c("NSCLC_FR_id26", "RCC_FR_id41", "T1D_FI_id21"),]
      allPairDownMeltImputed$effect_size[allPairDownMeltImputed$effect_size < 0] = 0
      allPairDownMeltImputed$effect_size[allPairDownMeltImputed$effect_size > 1] = 1
      allPairDownMeltImputed$effect_size[is.na(allPairDownMeltImputed$effect_size)] = 0
      
      manhattanMode = T
      if (manhattanMode) {
        uniqMsp = unique(allPairDownMeltImputed$msp) 
        uniqMspId = 1:length(uniqMsp)
        names(uniqMspId) =uniqMsp
        allPairDownMeltImputed$CHR = names(cohortOrders)[match(allPairDownMeltImputed$disease_cohort, cohortOrders)]
        allPairDownMeltImputed$CHR = as.numeric(allPairDownMeltImputed$CHR)
        #allPairDownMeltImputed$CHR = as.numeric(gsub(".*_id", "", allPairDownMeltImputed$disease_cohort))
        allPairDownMeltImputed$BP = uniqMspId[match(allPairDownMeltImputed$msp, names(uniqMspId))]
        allPairDownMeltImputed$P = allPairDownMeltImputed$effect_size
        allPairDownMeltImputed$SNP = allPairDownMeltImputed$msp
      }
      
      allPairDownMeltImputedMatched = allPairDownMeltImputed[allPairDownMeltImputed$control=="matched",]
      allPairDownMeltImputedCountry = allPairDownMeltImputed[allPairDownMeltImputed$control=="country",]
      allPairDownMeltImputedCluster = allPairDownMeltImputed[allPairDownMeltImputed$control=="cluster",]
      
      allPairDownMeltImputedMatchedSig = allPairDownMeltImputedMatched[allPairDownMeltImputedMatched$effect_size>=0.3,]
      allPairDownMeltImputedCountrySig = allPairDownMeltImputedCountry[allPairDownMeltImputedCountry$effect_size>=0.3,]
      allPairDownMeltImputedClusterSig = allPairDownMeltImputedCluster[allPairDownMeltImputedCluster$effect_size>=0.3,]
      
      depletedMspByCohortsWCountryControl = split(allPairDownMeltImputedCountrySig$msp,
                                                  allPairDownMeltImputedCountrySig$disease_cohort)
      
      enrichedMspByCohortsWCountryControl = split(allPairMeltImputedCountrySig$msp,
                                                  allPairMeltImputedCountrySig$disease_cohort)
      
    }
    
    loadDiseaseSignatures = F
    if (loadDiseaseSignatures) {
      
      allPairMeltRData = "C://Data//comparative.analysis.healthy.sweden//allPairMelt.disease.signatures.RData"
      load(allPairMeltRData)
      allPairDownMeltRData = "C://Data//comparative.analysis.healthy.sweden//allPairDownMelt.disease.signatures.RData"
      load(allPairDownMeltRData)
      
      esCut = 0.3
      
      allPairMeltSel = allPairMelt[allPairMelt$effect_size>=esCut,]
      allPairMeltSel = allPairMeltSel[!is.na(allPairMeltSel$effect_size),]
      #allPairMeltSel = allPairMeltSel[allPairMeltSel$disease_cohort != "T1D_FI_id21",]
      diseaseUpSpecies = unique(allPairMeltSel$msp)
      
      allPairDownMeltSel = allPairDownMelt[allPairDownMelt$effect_size>=esCut,]
      allPairDownMeltSel = allPairDownMeltSel[!is.na(allPairDownMeltSel$effect_size),]
      diseaseDownSpecies = unique(allPairDownMeltSel$msp)
      
      diseaseSigList = list(up=diseaseUpSpecies, down=diseaseDownSpecies)
    }
    
    checkAssociations = T
    if (checkAssociations) {
      
      withoutSelection = T
      if (withoutSelection) {
        pCut=1e-3
        enrichSigDownDiseaseFuncMat = getEnrichMentGsByGs(mspByModList, depletedMspByCohortsWCountryControl)
        enrichSigDownDiseaseFuncMelt = melt(enrichSigDownDiseaseFuncMat)
        enrichSigDownDiseaseFuncMeltSig=enrichSigDownDiseaseFuncMelt[enrichSigDownDiseaseFuncMelt$value<=pCut, ]
        funcDownDiseaseCounts = getVectorFromTable(table(enrichSigDownDiseaseFuncMeltSig$Var2))
        
        enrichSigUpDiseaseFuncMat = getEnrichMentGsByGs(mspByModList, enrichedMspByCohortsWCountryControl)
        enrichSigUpDiseaseFuncMelt = melt(enrichSigUpDiseaseFuncMat)
        enrichSigUpDiseaseFuncMeltSig=enrichSigUpDiseaseFuncMelt[enrichSigUpDiseaseFuncMelt$value<=pCut, ]
        funcUpDiseaseCounts = getVectorFromTable(table(enrichSigUpDiseaseFuncMeltSig$Var2))
        
        allFuncNames = unique(c(names(funcDownDiseaseCounts),
                                names(funcUpDiseaseCounts)))
        
        funcClusterStatTab = data.frame(name=allFuncNames,
                                        down=funcDownDiseaseCounts[match(allFuncNames, names(funcDownDiseaseCounts))],
                                        up=funcUpDiseaseCounts[match(allFuncNames, names(funcUpDiseaseCounts))],
                                        size=funcModuleNums[match(allFuncNames, names(funcModuleNums))],
                                        stringsAsFactors = F)
        
        funcClusterStatTab[is.na(funcClusterStatTab)]=0
        #funcClusterStatTab$diff = funcClusterStatTab$down - funcClusterStatTab$up
        funcClusterStatTab$diff =  funcClusterStatTab$up - funcClusterStatTab$down
        
        funcClusterStatTab$sum = funcClusterStatTab$up + funcClusterStatTab$down
        funcClusterStatTab$fcInFewSpecies = F
        funcClusterStatTab$fcInFewSpecies[funcClusterStatTab$name %in% smallFcs] = T
        
        signletons = names(funcModuleNums[funcModuleNums==1])
        funcClusterStatTab$singleton = F
        funcClusterStatTab$singleton[funcClusterStatTab$name %in% signletons] = T
        funcClusterStatTab2 = funcClusterStatTab
        funcClusterStatTab2 = funcClusterStatTab2[!funcClusterStatTab2$fcInFewSpecies,]
        funcClusterStatTab2 = funcClusterStatTab2[!funcClusterStatTab2$singleton,]
        funcClusterStatTab2$inflowEnrichmentP = enrichmentNormTab$inflowP[match(funcClusterStatTab2$name, enrichmentNormTab$id)]
        funcClusterStatTab2$outflowEnrichmentP = enrichmentNormTab$outflowP[match(funcClusterStatTab2$name, enrichmentNormTab$id)]
        funcClusterStatTab2$inflowSig = funcClusterStatTab2$inflowEnrichmentP <= 0.01
        funcClusterStatTab2$outflowSig = funcClusterStatTab2$outflowEnrichmentP <= 0.01
        
        
        
        set.seed(1)
        ggplot(funcClusterStatTab2, aes(x=diff, y=sum)) + 
          geom_jitter(shape = 21, colour = "#00000055", width = 0.5, height = 0.5, size=2, aes(fill = outflowSig)) + 
          scale_fill_manual(values=c("#80808033", "#6a0dad88")) +
          #geom_jitter(shape = 21, colour = "black", width = 0.5, height = 0.5, size=2, aes(fill = diff)) + 
          #scale_fill_gradient2(low = "blue", mid="white",  high = "red") +
          geom_hline(yintercept = 3, colour="gray") +
          geom_vline(xintercept = 2, colour="gray") +
          geom_vline(xintercept = -2, colour="gray") +
          geom_abline(slope=1)+
          geom_abline(slope=-1) +
          xlab("|Enrichment| - |Depletion|") + ylab("|Enrichment| + |Depletion|")+
          scale_y_continuous(breaks=c(2,4,6,8))+
          theme(panel.grid.major = element_blank(), 
                panel.grid.minor = element_blank(),
                legend.position = "none",
                panel.background = element_rect(fill = "white", 
                                                colour = "black", 
                                                size = 0.5, 
                                                linetype = "solid"))
        
        View(funcClusterStatTab[funcClusterStatTab$name %in% antismashClusters,])
        View(funcClusterStatTab[funcClusterStatTab$name %in% vfClusters,])
        View(funcClusterStatTab[funcClusterStatTab$name %in% cazyClusters,])
        
        table(funcClusterStatTab2$outflowEnrichmentP[funcClusterStatTab2$diff >= 2 & funcClusterStatTab2$sum >= 3] <=  1e-3)
        table(funcClusterStatTab2$inflowEnrichmentP[funcClusterStatTab2$diff >= 2& funcClusterStatTab2$sum >= 3] <= 1e-3)
        
        
        table(funcClusterStatTab2$outflowEnrichmentP[funcClusterStatTab2$diff <= -2 & funcClusterStatTab2$sum >= 3] <=  1e-3)
        table(funcClusterStatTab2$inflowEnrichmentP[funcClusterStatTab2$diff <= -2 & funcClusterStatTab2$sum >= 3] <= 1e-3)
        
        
        
        
        set.seed(1)
        ggplot(funcClusterStatTab2, aes(x=inflowEnrichmentP, y=outflowEnrichmentP)) + 
          geom_point(shape = 21, aes(fill=diff, size=size)) + 
          scale_fill_gradient2(low = "blue", mid="white",  high = "red") +
          scale_x_continuous(trans='log10') +scale_y_continuous(trans='log10') +
          scale_size(range = c(2, 10))+ 
          xlab("Inflow enrichment") + ylab("Outflow enrichment")+
          theme(panel.grid.major = element_blank(), 
                panel.grid.minor = element_blank(),
                legend.position = "none",
                panel.background = element_rect(fill = "white", 
                                                colour = "black", 
                                                size = 0.5, 
                                                linetype = "solid"))
        
        
        
        ggplot(numStatTab, aes(x=numSpecies, y=clusterSize)) + 
          geom_point(aes(fill=numSpecies, size=clusterSize), shape=21) +
          xlab("Number of enriched species") + ylab("Cluster size") +
          xlim(-10,650)+
          scale_size(range = c(2, 10))+ 
          coord_trans(y="log10")+
          #scale_fill_gradient2(mid="#ffffff88",  high = "#6a0dad88") +#964B00
          scale_fill_gradient2(mid="#ffffff88",  high = "#964B0088") +
          theme(panel.grid.minor = element_line(colour="#80808044", size=0.5),
                panel.grid.major = element_line(colour="#80808044", size=0.5),
                legend.position = "none",
                panel.background = element_rect(fill = "white", 
                                                colour = "black", 
                                                size = 0.5, 
                                                linetype = "solid")) 
        
        
        
      }
      
      withSelection = T
      if (withSelection) {
        pCut=1e-3
        enrichSigDownDiseaseFuncMat = getEnrichMentGsByGs(mspByModSelList, depletedMspByCohortsWCountryControl)
        enrichSigDownDiseaseFuncMat = enrichSigDownDiseaseFuncMat[,colnames(enrichSigDownDiseaseFuncMat) %in% names(funcModuleSel)]
        enrichSigDownDiseaseFuncMelt = melt(enrichSigDownDiseaseFuncMat)
        enrichSigDownDiseaseFuncMeltSig=enrichSigDownDiseaseFuncMelt[enrichSigDownDiseaseFuncMelt$value<=pCut, ]
        #lapply(funcModules[names(sort(table(enrichSigDownDiseaseFuncMeltSig$Var2), decreasing = T)[1:10])], function(x) getDescFromKoTerms(x, koDescMap))
        funcDownDiseaseCounts = getVectorFromTable(table(enrichSigDownDiseaseFuncMeltSig$Var2))
        
        enrichSigUpDiseaseFuncMat = getEnrichMentGsByGs(mspByModSelList, enrichedMspByCohortsWCountryControl)
        enrichSigUpDiseaseFuncMat = enrichSigUpDiseaseFuncMat[,colnames(enrichSigUpDiseaseFuncMat) %in% names(funcModuleSel)]
        enrichSigUpDiseaseFuncMelt = melt(enrichSigUpDiseaseFuncMat)
        enrichSigUpDiseaseFuncMeltSig=enrichSigUpDiseaseFuncMelt[enrichSigUpDiseaseFuncMelt$value<=pCut, ]
        funcUpDiseaseCounts = getVectorFromTable(table(enrichSigUpDiseaseFuncMeltSig$Var2))
        
        allFuncNames = unique(c(names(funcDownDiseaseCounts),
                                names(funcUpDiseaseCounts)))
        
        funcClusterStatTab = data.frame(name=allFuncNames,
                                        down=funcDownDiseaseCounts[match(allFuncNames, names(funcDownDiseaseCounts))],
                                        up=funcUpDiseaseCounts[match(allFuncNames, names(funcUpDiseaseCounts))])
        funcClusterStatTab[is.na(funcClusterStatTab)]=0
        funcClusterStatTab$diff = funcClusterStatTab$down - funcClusterStatTab$up
        funcClusterStatTab$sum = funcClusterStatTab$up + funcClusterStatTab$down
        
        set.seed(1)
        ggplot(funcClusterStatTab, aes(x=diff, y=sum)) + 
          geom_jitter(shape = 21, colour = "black", width = 0.5, height = 0.5, size=2, aes(fill = diff)) + 
          scale_fill_gradient2(low = "blue", mid="white",  high = "red") +
          geom_hline(yintercept = 3, colour="gray") +
          geom_vline(xintercept = 2, colour="gray") +
          geom_vline(xintercept = -2, colour="gray") +
          geom_abline(slope=1)+
          geom_abline(slope=-1) +
          xlab("|Depletion| - |Enrichment|") + ylab("|Depletion| + |Enrichment|")+
          theme(panel.grid.major = element_blank(), 
                panel.grid.minor = element_blank(),
                legend.position = "none",
                panel.background = element_rect(fill = "white", 
                                                colour = "black", 
                                                size = 0.5, 
                                                linetype = "solid"))
        
        #ggplot(funcClusterStatTab, aes(x=diff,y=sum)) + geom_point()
        
        sort(table(enrichSigUpDiseaseFuncMeltSig$Var2), decreasing = T)[1:10]
        hist(table(enrichSigUpDiseaseFuncMeltSig$Var2))
        lapply(funcModules[names(sort(table(enrichSigUpDiseaseFuncMeltSig$Var2), decreasing = T)[1:10])], function(x) getDescFromKoTerms(x, koDescMap))
        
      }
      
      ### secondary only ###
      
      pCut=1e-3
      enrichSigDownDiseaseFuncMat = getEnrichMentGsByGs(mspBySecMetList, depletedMspByCohortsWCountryControl)
      enrichSigDownDiseaseFuncMelt = melt(enrichSigDownDiseaseFuncMat)
      enrichSigDownDiseaseFuncMeltSig=enrichSigDownDiseaseFuncMelt[enrichSigDownDiseaseFuncMelt$value<=pCut, ]
      #lapply(funcModules[names(sort(table(enrichSigDownDiseaseFuncMeltSig$Var2), decreasing = T)[1:10])], function(x) getDescFromKoTerms(x, koDescMap))
      funcDownDiseaseCounts = getVectorFromTable(table(enrichSigDownDiseaseFuncMeltSig$Var2))
      
      enrichSigUpDiseaseFuncMat = getEnrichMentGsByGs(mspBySecMetList, enrichedMspByCohortsWCountryControl)
      enrichSigUpDiseaseFuncMelt = melt(enrichSigUpDiseaseFuncMat)
      enrichSigUpDiseaseFuncMeltSig=enrichSigUpDiseaseFuncMelt[enrichSigUpDiseaseFuncMelt$value<=pCut, ]
      funcUpDiseaseCounts = getVectorFromTable(table(enrichSigUpDiseaseFuncMeltSig$Var2))
      
      allFuncNames = unique(c(names(funcDownDiseaseCounts),
                              names(funcUpDiseaseCounts)))
      
      funcClusterStatTab = data.frame(name=allFuncNames,
                                      down=funcDownDiseaseCounts[match(allFuncNames, names(funcDownDiseaseCounts))],
                                      up=funcUpDiseaseCounts[match(allFuncNames, names(funcUpDiseaseCounts))])
      funcClusterStatTab[is.na(funcClusterStatTab)]=0
      funcClusterStatTab$diff = funcClusterStatTab$down - funcClusterStatTab$up
      funcClusterStatTab$sum = funcClusterStatTab$up + funcClusterStatTab$down
      
      ggplot(funcClusterStatTab, aes(x=diff, y=sum)) + 
        geom_jitter(shape = 21, colour = "black", width = 0.5, height = 0.5, size=2, aes(fill = diff)) + 
        scale_fill_gradient2(low = "blue", mid="white",  high = "red") +
        geom_hline(yintercept = 3, colour="gray") +
        geom_vline(xintercept = 2, colour="gray") +
        geom_vline(xintercept = -2, colour="gray") +
        geom_abline(slope=1)+
        geom_abline(slope=-1) +
        xlab("|Depletion| - |Enrichment|") + ylab("|Depletion| + |Enrichment|")+
        theme(panel.grid.major = element_blank(), 
              panel.grid.minor = element_blank(),
              legend.position = "none",
              panel.background = element_rect(fill = "white", 
                                              colour = "black", 
                                              size = 0.5, 
                                              linetype = "solid"))
      
      #ggplot(funcClusterStatTab, aes(x=diff,y=sum)) + geom_point()
      
      sort(table(enrichSigUpDiseaseFuncMeltSig$Var2), decreasing = T)[1:10]
      hist(table(enrichSigUpDiseaseFuncMeltSig$Var2))
      lapply(funcModules[names(sort(table(enrichSigUpDiseaseFuncMeltSig$Var2), decreasing = T)[1:10])], function(x) getDescFromKoTerms(x, koDescMap))
      
      
    }
    
    loadRichnessSignatures = T
    if (loadRichnessSignatures) {
      statsMgsForAllSamples = "J:\\Deposit\\Project\\2018_microbiome_atlas\\atlas.volcanot.out\\richness.volcano.stats.mgs.for.all.txt"
      statsMgsForNormSamples = "J:\\Deposit\\Project\\2018_microbiome_atlas\\atlas.volcanot.out\\richness.volcano.stats.mgs.for.normal.txt"
      statsMgsForDiseaseSamples = "J:\\Deposit\\Project\\2018_microbiome_atlas\\atlas.volcanot.out\\richness.volcano.stats.mgs.for.disease.txt"
      statsMgsForIndustrialSamples  = "J:\\Deposit\\Project\\2018_microbiome_atlas\\atlas.volcanot.out\\richness.volcano.stats.mgs.for.industrial.txt"
      statsMgsForTraditionalSamples  = "J:\\Deposit\\Project\\2018_microbiome_atlas\\atlas.volcanot.out\\richness.volcano.stats.mgs.for.traditional.txt"
      
      adjpCut = 1e-3
      fcCut = 1
      statsMgsForNorm = read.table(statsMgsForNormSamples, sep="\t", header=T, stringsAsFactors = F)
      statsMgsForNormSel = statsMgsForNorm[statsMgsForNorm$qvalue <= adjpCut,]
      #statsMgsForNormSelHgc = statsMgsForNormSel[statsMgsForNormSel$lfc > 0,]
      #statsMgsForNormSelLgc = statsMgsForNormSel[statsMgsForNormSel$lfc < 0,]
      statsMgsForNormSelHgc = statsMgsForNormSel[statsMgsForNormSel$lfc > fcCut,]
      statsMgsForNormSelLgc = statsMgsForNormSel[statsMgsForNormSel$lfc < (-1)*fcCut,]
      
      
      quantCut = 0.5
      hgcCut = quantile(statsMgsForNormSelHgc$relAbd_HGC, quantCut)
      lgcCut = quantile(statsMgsForNormSelLgc$relAbd_LGC, quantCut)
      
      hgcSpecies = rownames(statsMgsForNormSelHgc[statsMgsForNormSelHgc$relAbd_HGC >= hgcCut,])
      lgcSpecies = rownames(statsMgsForNormSelLgc[statsMgsForNormSelLgc$relAbd_LGC >= lgcCut,])
      
      richnessSpecies = c(hgcSpecies, lgcSpecies)
      
      ###
      statsMgsForNormSelHgcNew = statsMgsForNormSelHgc[order(statsMgsForNormSelHgc$qvalue),]
      statsMgsForNormSelLgcNew = statsMgsForNormSelLgc[order(statsMgsForNormSelLgc$qvalue),]
      hgcTop25Species = rownames(statsMgsForNormSelHgcNew[1:25,])
      lgcTop25Species = rownames(statsMgsForNormSelLgcNew[1:25,])
      
      richnessTop25Species = c(hgcTop25Species, lgcTop25Species)
      
      hgcLgcList = list(hgc=hgcSpecies, lgc=lgcSpecies)
      hgcLgcTop25List = list(hgc=hgcTop25Species, lgc=lgcTop25Species)
      
    }
    
    loadGeoSignatures = T
    if (loadGeoSignatures) {
      
      
      
    }
    
    tt = unlist(lapply(funcModules, function(x) any(x == "K03416")))
    tt[tt]
    
    
    getFuncClusterInfo = T
    if (getFuncClusterInfo) {
      secretionClusters = c("8","120","425","250")
      mucinCazymes = c("cazy.GT27", "cazy.GT14", "cazy.GT11", "cazy.GT10", "cazy.GH89", "cazy.GH84", "cazy.GH33", "cazy.GH20", "cazy.GH29", "cazy.GH18", "cazy.GH123", "cazy.GH109", "cazy.CBM51", "cazy.CBM50", "cazy.CBM32")
      storageCazymes = c("cazy.CBM34", "cazy.CBM48", "cazy.CBM20", "cazy.GH13", "cazy.GH133", "cazy.GH32", "cazy.GH36", "cazy.GH37", "cazy.GH77", "cazy.GH57", "cazy.GT101", "cazy.GT26", "cazy.GT3", "cazy.GT35", "cazy.GT5")
      pectinCazymes = c("cazy.CBM77","cazy.CE8","cazy.GH105","cazy.GH28", "cazy.PL1")
      
      mucinInd=do.call(c,lapply(funcModules, function(x) any(x %in% mucinCazymes)))
      mucinClusters = names(mucinInd[mucinInd])
      
      storageInd=do.call(c,lapply(funcModules, function(x) any(x %in% storageCazymes)))
      storageClusters = names(storageInd[storageInd])
      
      pectinInd=do.call(c,lapply(funcModules, function(x) any(x %in% pectinCazymes)))
      pectinClusters = names(pectinInd[pectinInd])
      
      mucinClusterInfo = lapply(funcModules[mucinClusters], function(terms){
        termStr = paste(terms[terms %in% mucinCazymes], collapse =";")
        return(termStr)
      })
      mucinClusterInfoList = unlist(mucinClusterInfo)
      
      storageClusterInfo = lapply(funcModules[storageClusters], function(terms){
        termStr = paste(terms[terms %in% storageCazymes], collapse =";")
        return(termStr)
      })
      storageClusterInfoList = unlist(storageClusterInfo)
      
      pectinClusterInfo = lapply(funcModules[pectinClusters], function(terms){
        termStr = paste(terms[terms %in% pectinCazymes], collapse =";")
        return(termStr)
      })
      pectinClusterInfoList = unlist(pectinClusterInfo)
      
      antismashClusterInfo = lapply(funcModules[antismashClusters], function(terms){
        termStr = paste(terms[terms %in% antismashTerms], collapse =";")
        return(termStr)
      })
      antismashClusterInfoList = unlist(antismashClusterInfo)
      
      vfClusterInfo = lapply(funcModules[vfClusters], function(terms){
        terms = terms[terms %in% vfTerms]
        if (length(terms)==0) return(NA)
        if (length(terms)!=0) {
          terms=getClassesFromVfTerms(terms, patricVfDescTab)
          terms=unique(terms)
        }
        #if (terms =="NA") return(NA)
        termStr = paste(terms, collapse =";")
        return(termStr)
      })
      vfClusterInfoList = unlist(vfClusterInfo)
      
      cazyClusterInfo = lapply(funcModules[cazyClusters], function(terms){
        termStr = paste(terms[terms %in% cazyTerms], collapse =";")
        return(termStr)
      })
      cazyClusterInfoList = unlist(cazyClusterInfo)
      
    }
    
    analyzHgcLgcMode = T
    if (analyzHgcLgcMode) {
      moduleHgcLgcEnrichMat = getEnrichMentGsByGs(mspByModList, hgcLgcList)
      moduleHgcLgcEnrichMelt = melt(moduleHgcLgcEnrichMat)
      moduleHgcLgcEnrichMelt$Var2 = as.character(moduleHgcLgcEnrichMelt$Var2)
      moduleHgcLgcEnrichMelt = moduleHgcLgcEnrichMelt[moduleHgcLgcEnrichMelt$value<=0.05,]
      
      attachClusterInfoMode = F
      if (attachClusterInfoMode) {
        antismashMatched = do.call(c,sapply(moduleHgcLgcEnrichMelt$Var2, function(ind) {
          if (any(antismashClusters == ind)) {
            return(antismashClusterInfoList[ind])
          }
          return(NA)
        }, simplify = F))
        moduleHgcLgcEnrichMelt$antismash = antismashMatched
        
        mucinMatched = do.call(c,sapply(moduleHgcLgcEnrichMelt$Var2, function(ind) {
          if (any(mucinClusters == ind)) {
            return(mucinClusterInfoList[ind])
          }
          return(NA)
        }, simplify = F))
        moduleHgcLgcEnrichMelt$mucin = mucinMatched
        
        storageMatched = do.call(c,sapply(moduleHgcLgcEnrichMelt$Var2, function(ind) {
          if (any(storageClusters == ind)) {
            return(storageClusterInfoList[ind])
          }
          return(NA)
        }, simplify = F))
        moduleHgcLgcEnrichMelt$storage = storageMatched
        
        pectinMatched = do.call(c,sapply(moduleHgcLgcEnrichMelt$Var2, function(ind) {
          if (any(pectinClusters == ind)) {
            return(pectinClusterInfoList[ind])
          }
          return(NA)
        }, simplify = F))
        moduleHgcLgcEnrichMelt$pectin = pectinMatched
        
        
        vfMatched = do.call(c,sapply(moduleHgcLgcEnrichMelt$Var2, function(ind) {
          if (any(vfClusters == ind)) {
            return(vfClusterInfoList[ind])
          }
          return(NA)
        }, simplify = F))
        moduleHgcLgcEnrichMelt$vf = vfMatched
        
        moduleHgcLgcEnrichMeltFile = "C://Data//comparative.analysis.healthy.sweden//moduleHgcLgcEnrichMelt.txt"
        write.table(moduleHgcLgcEnrichMelt, file=moduleHgcLgcEnrichMeltFile, sep="\t")
        
        rm(moduleHgcLgcEnrichMelt)
      }
      
      if (F) {
        
        #antismashClusters
        View(moduleHgcLgcEnrichMelt[moduleHgcLgcEnrichMelt$Var2 %in% antismashClusters,])
        View(moduleHgcLgcEnrichMelt[moduleHgcLgcEnrichMelt$Var2 %in% mucinClusters,])
        View(moduleHgcLgcEnrichMelt[moduleHgcLgcEnrichMelt$Var2 %in% secretionClusters,])
        
        tempCheck = moduleHgcLgcEnrichMelt[moduleHgcLgcEnrichMelt$Var2 %in% names(antismashEnrichVec[antismashEnrichVec<=0.1]), ]
        View(tempCheck)
        tempCheck$value = -log10(tempCheck$value)
        ggplot(tempCheck, aes(x=Var1, y=value, group=Var2)) +geom_bar(stat="identity")
        
      }
      
    }
    
    moduleDiseaseSigEnrichMat = getEnrichMentGsByGs(mspByModList, diseaseSigList)
    moduleDiseaseSigEnrichMelt = melt(moduleDiseaseSigEnrichMat)
    moduleDiseaseSigEnrichMelt = moduleDiseaseSigEnrichMelt[moduleDiseaseSigEnrichMelt$value<=0.05,]
    
    ### mucin-degrading ###
    if (F) {
      View(taxo[taxo$MSP %in% mspByModList$`518`, ]) # gt20
      View(taxo[taxo$MSP %in% mspByModList$`511`, ]) # gt10
      
      View(taxo[taxo$MSP %in% mspByModList$`1356`, ]) # gt27
      View(taxo[taxo$MSP %in% mspByModList$`435`, ]) # gt11
      
      View(taxo[taxo$MSP %in% mspByModList$`294`, ]) # gh89
      View(taxo[taxo$MSP %in% mspByModList$`508`, ]) # gh84
      View(taxo[taxo$MSP %in% mspByModList$`303`, ]) # gh33
    }
    
    vfClusters = names(sort(vfEnrichVec)[1:5])
    
    vfClusterMspList = mspByModList[vfClusters]
    vfClusterMspMelt = melt(vfClusterMspList)
    vfClusterMspMelt$count = 1
    vfClusterMspMat = acast(vfClusterMspMelt, value~ L1, value.var = "count")
    vfClusterMspMat[is.na(vfClusterMspMat)] = 0
    rownames(vfClusterMspMat) = getSpeciesName(rownames(vfClusterMspMat), taxo)
    dim(vfClusterMspMat)
    
    heatmap.2(vfClusterMspMat, trace="none")
    heatmap.2(vfClusterMspMat,
              margins = c(4,12), 
              key=F, lhei=c(z,20),
              trace="none", 
              cexRow = 0.7, cexCol = 0.9,
              col=colorRampPalette(c("white","red"))(256),
              sepwidth = c(0.01,0.01),
              sepcolor = "gray",
              colsep = 0:10,
              rowsep = 0:56,
              srtCol = 45)
    
    tt=do.call(c,lapply(funcModules, function(x) any(x %in% mucinCazymes)))
    
    mucinClusters = names(tt[tt])
    mucinClusterMspList = mspByModList[mucinClusters]
    mucinClusterMspMelt = melt(mucinClusterMspList)
    mucinClusterMspMelt$count = 1
    mucinClusterMspMat = acast(mucinClusterMspMelt, value~ L1, value.var = "count")
    dim(mucinClusterMspMat)
    mucinClusterMspMat[is.na(mucinClusterMspMat)] = 0
    mucinClusterMspMatRowCounts = rowSums(mucinClusterMspMat)
    mucinClusterMspMatRowCountsWithName = mucinClusterMspMatRowCounts
    names(mucinClusterMspMatRowCountsWithName) = getSpeciesName(names(mucinClusterMspMatRowCounts),taxo)
    mucinClusterMspMatRowCountsWithName = sort(mucinClusterMspMatRowCountsWithName)
    
    par(mar=c(5,25,4,1))
    barplot(rev(rev(mucinClusterMspMatRowCountsWithName[mucinClusterMspMatRowCountsWithName>=10])), horiz = T, las=1, cex.axis = 0.7, cex.names = 0.7)
    
    tt=do.call(c,lapply(funcModules, function(x) any(x %in% storageCazymes)))
    storageClusters = names(tt[tt])
    storageClusterMspList = mspByModList[storageClusters]
    storageClusterMspMelt = melt(storageClusterMspList)
    storageClusterMspMelt$count = 1
    storageClusterMspMat = acast(storageClusterMspMelt, value~ L1, value.var = "count")
    dim(storageClusterMspMat)
    storageClusterMspMat[is.na(storageClusterMspMat)] = 0
    storageClusterMspMatRowCounts = rowSums(storageClusterMspMat)
    storageClusterMspMatRowCountsWithName = storageClusterMspMatRowCounts
    names(storageClusterMspMatRowCountsWithName) = getSpeciesName(names(storageClusterMspMatRowCounts),taxo)
    storageClusterMspMatRowCountsWithName = sort(storageClusterMspMatRowCountsWithName)
    
    par(mar=c(5,25,4,1))
    barplot(rev(rev(storageClusterMspMatRowCountsWithName[storageClusterMspMatRowCountsWithName>=10])), horiz = T, las=1, cex.axis = 0.7, cex.names = 0.7)
    
    modLengths = do.call(c, lapply(funcModules, length))
    modLengths = sort(modLengths)
    
    
  }
  
  checkKoMode = F
  if (checkKoMode) {
    targetModules = findModulesBy(funcModules, c("K00132","K04072","K04073","K18366"))
    targetModules = findModulesBy(funcModules, c("K00100"))
    targetModules = findModulesBy(funcModules, c("K00634"))
    targetModules = findModulesBy(funcModules, c("K01034"))
    targetModules = findModulesBy(funcModules, c("K01035"))
    targetModules = findModulesBy(funcModules, c("K19709"))
    targetModules = findModulesBy(funcModules, c("K01034","K01035","K19709"))
    targetModules
    
    getDescFromKoTerms(funcKoModules[["132"]], koDescMap)
    unlist(funcKoModules[targetModules])
    
    
  }
  
  checkModuleMode = T
  if (checkModuleMode) {
    butryCluster132 = grepMspsByFuncModules(funcKoModules, gutKoBestMat, taxo, "132")
    View(butryCluster132$wholeData)
    
    
    butryCluster332 = grepMspsByFuncModules(funcKoModules, gutKoBestMat, taxo, "332")
    View(butryCluster332$wholeData)
    
    butryCluster621 = grepMspsByFuncModules(funcKoModules, gutKoBestMat, taxo, "621")
    View(butryCluster621$wholeData)
    
    butryCluster1241 = grepMspsByFuncModules(funcKoModules, gutKoBestMat, taxo, "1241")
    View(butryCluster1241$wholeData)
    
    butryCluster1539 = grepMspsByFuncModules(funcKoModules, gutKoBestMat, taxo, "1539")
    View(butryCluster1539$wholeData)
    
    butryCluster3404 = grepMspsByFuncModules(funcKoModules, gutKoBestMat, taxo, "3404")
    View(butryCluster3404$wholeData)
    
    
    butryCluster_132_1539_3404 = grepMspsByKOs(unlist(funcKoModules[targetModules]), gutKoBestMat, taxo)
    View(butryCluster_132_1539_3404$wholeData)
    
    macroCluster4930 = grepMspsByFuncModules(funcKoModules, gutKoBestMat, taxo, "4930")
    View(macroCluster4930$wholeData)
    
    macroCluster4969 = grepMspsByFuncModules(funcKoModules, gutKoBestMat, taxo, "4969")
    View(macroCluster4969$wholeData)
    
    
    
    acetateCluster361 = grepMspsByFuncModules(funcModules, gutKoBestMat, taxo, "361")
    acetateCluster354 = grepMspsByFuncModules(funcModules, gutKoBestMat, taxo, "354")
    
    propionateCluster551 = grepMspsByFuncModules(funcKoModules, gutKoBestMat, taxo, "551")
    propionateCluster620 = grepMspsByFuncModules(funcKoModules, gutKoBestMat, taxo, "620")
    propionateCluster609 = grepMspsByFuncModules(funcKoModules, gutKoBestMat, taxo, "609")
    propionateCluster100 = grepMspsByFuncModules(funcKoModules, gutKoBestMat, taxo, "100")
    
    check = checkFractionOfGenus(gutTaxo, butryCluster332$msps, T)
    check2 = checkFractionOfGenus(gutTaxo, acetateCluster354$msps, T)
    check3 = checkFractionOfGenus(gutTaxo, butryCluster1241$msps, T)
    
    zz =do.call(c, lapply(funcGemModules, function(x) any(x %in% funcModules[["332"]])))
    
    View(acetateCluster354$wholeData[acetateCluster354$msps,])   
  }
  
  associateSamplesWithCluster = T
  if (associateSamplesWithCluster) {
    
    generateMode = F
    if (generateMode) {
      funcMatRData = "j://Deposit/Project/2018_microbiome_atlas/functional.annotation/funcMat.20190808.RData"
      load(funcMatRData)
      
      if (F) {
        naTests = unlist(lapply(funcModules, function(x) any(x %in% "NA")))
        funcModules = funcModules[!names(funcModules) %in%c("1766","7749")]
        save(funcModules, file = funcModulesRData)
        
      }
      
      zMatUpdated = scale(t(mergeMatUpdated))
      zMatUpdated = t(zMatUpdated)
      
      getMspByModules <- function(funcModules, funcMat, cutRatio=0.75) {
        mspByModules = lapply(funcModules, function(currTerms){
          
          if (length(currTerms)<2) {
            currMspsByCurrTerms = funcMat[,colnames(funcMat) %in% currTerms]
          } else {
            currMat = funcMat[,colnames(funcMat) %in% currTerms]
            currMspsByCurrTerms = rowSums(currMat)
          }
          currMspsByCurrTerms = sort(currMspsByCurrTerms, decreasing = T)
          outMsps = currMspsByCurrTerms[currMspsByCurrTerms >= (length(currTerms)*cutRatio)]
          return(names(outMsps))
        })
        names(mspByModules) = names(funcModules)
        return(mspByModules)
      }
      
      
      mspByModules = getMspByModules(funcModules, funcMat)
      mspCountsByModules = do.call(c, lapply(mspByModules,length) )
      hist(mspCountsByModules)
      
      getBSCorrelationsByModules <- function(mspByModules, zMat, pCut=0.01) {
        bsCorrMat <- do.call(rbind, lapply(mspByModules, function(currMsps) {
          inclusion = rep(0,dim(zMat)[1])
          names(inclusion) = rownames(zMat)
          inclusion[names(inclusion) %in% currMsps] = 1 
          
          
          corrVec = apply(zMat, 2, function(currCol){
            if (sum(currCol)==0) return(0)
            if (sum(inclusion)==0) return(0)
            corrOut = cor.test(currCol, inclusion)
            rho = corrOut$estimate
            pvalue = corrOut$p.value
            if (pvalue > pCut) rho=0
            return(rho)
          })
          
          
          return(corrVec)
          
        }))
        rownames(bsCorrMat) = names(mspByModules)
        return(bsCorrMat)
      }
      bsCorrMat <- getBSCorrelationsByModules(mspByModules, zMatUpdated)
      bsCorrMatRData = "j://Deposit/Project/2018_microbiome_atlas//functypeData//bsCorrMat.RData"
      save(bsCorrMat, file=bsCorrMatRData)
    }
    
    loadMode = T
    if (loadMode) {
      
      bsCorrMatRData = "j://Deposit/Project/2018_microbiome_atlas//functypeData//bsCorrMat.RData"
      load(bsCorrMatRData)
      
      length(funcModules)
      dim(bsCorrMat)
      rownames(bsCorrMat) = names(funcModules)
      
      bsCorrMaxRows = rowMaxs(bsCorrMat)
      names(bsCorrMaxRows)= rownames(bsCorrMat)
      
      bsCorrMelt = melt(bsCorrMat)
      bsCorrMeltCut = bsCorrMelt[bsCorrMelt$value>=0.3,]
      bsCorrMeltCut$Var1 = as.character(bsCorrMeltCut$Var1)
      bsCorrMeltCut$Var2 = as.character(bsCorrMeltCut$Var2)
      sigSamplesByCluster = split(bsCorrMeltCut$Var2, bsCorrMeltCut$Var1)
      
      colVec = rep("#80808011", 5224)
      names(colVec) = rownames(pcoaGenusMat)
      colVec[(sigSamplesByCluster[["367"]])] = "#FF0000"
      
      plot(pcoaGenusMat[,1], 
           pcoaGenusMat[,2], col=colVec, pch=16)
      
      
      
    }
    
  }
  
  ### attach annotations
  
}

addEnteroType = T
if (addEnteroType) {
  
  loadEnteroLists = T
  if (loadEnteroLists) {
    
    
    babyList = list(baby.finland = newNormalList$Finland_id21,
                    baby.T1D = newDiseaseList$T1D_FI,
                    preterm = newPerturbCaseList$Preterm_US)
    
    
    newCaseContByGeoList= list(Sweden_D=c(newDiseaseList$Atherosclerosis_SE, 
                                          newDiseaseList$T2D_SE),
                               Sweden_N=c(newNormalList$Sweden_id28,
                                          newNormalList$Sweden_id1,
                                          newNormalList$Sweden_id14),
                               Denmark_D = c(newDiseaseList$Obesity_DK_16,
                                             newDiseaseList$Obesity_DK_25),
                               Denmark_N=c(newNormalList$Denmark_id16, 
                                           newNormalList$Denmark_id25),
                               UK_D=c(newDiseaseList$Cirrhosis_GB),
                               UK_N=c(newNormalList$UK_id29,
                                      newNormalList$UK_id39),
                               Spain_D=c(newDiseaseList$IBD_ES,
                                         newDiseaseList$T2D_ES),
                               Spain_N=c(newNormalList$Spain_id17),
                               Italy_D=c(newDiseaseList$NAFLD_IT,
                                         newDiseaseList$CRC_IT),
                               Italy_N=c(newNormalList$Italy_id37,
                                         newNormalList$Italy_id46),
                               Germany_D=c(newDiseaseList$CRC_DE,
                                           newDiseaseList$Parkinson_DE),
                               Germany_N=c(newNormalList$Germany_id44,
                                           newNormalList$Germany_id8),
                               Luxembourg_D=c(newDiseaseList$T1D_LU),
                               Luxembourg_N=c(newNormalList$Luxembourg_id35),
                               Japan_D=c(newDiseaseList$CRC_JP),
                               Japan_N=c(newNormalList$Japan_id45),
                               US_D=c(newDiseaseList$CRC_US,
                                      newDiseaseList$Melanoma_US,
                                      newDiseaseList$CFS_US,
                                      newDiseaseList$NAFLD_US),
                               US_N=c(newNormalList$US_id11,
                                      newNormalList$US_id32,
                                      newNormalList$US_id34,
                                      newNormalList$US_id36,
                                      newNormalList$US_id43),
                               China_D=c(newDiseaseList$CVD_CN,
                                         newDiseaseList$GDM_CN,
                                         newDiseaseList$T2D_CN,
                                         newDiseaseList$Cirrhosis_CN,
                                         newDiseaseList$Rheumatoid_CN,
                                         newDiseaseList$Crohn_CN,
                                         newDiseaseList$Ankylosing_CN,
                                         newDiseaseList$Behcet_CN),
                               China_N=c(newNormalList$China_id10,
                                         newNormalList$China_id12,
                                         newNormalList$China_id2,
                                         newNormalList$China_id20,
                                         newNormalList$China_id31,
                                         newNormalList$China_id52,
                                         newNormalList$China_id6,
                                         newNormalList$China_id9))
    
    newCaseContByEnteroList = grepEnteroTypesFrom(rev(newCaseContByGeoList), basicMetaMapUpdated) 
    targeCountry="Sweden"
    getEntTab(newCaseContByGeoList[paste(targeCountry, c("_N","_D"), sep="")])
    entGeoTab = getEntGeoPlot(newCaseContByEnteroList, F)
    chisq.test(entGeoTab[paste(targeCountry, c("_N","_D"), sep=""),])
    chisq.test(entGeoTab[c("Sweden_N","Sweden_D"),])
    
    newNormalEnteroList = grepEnteroTypesFrom(c(newNormalList,newNonWesternList), basicMetaMapUpdated) 
    newDiseaseEnteroList = grepEnteroTypesFrom(newDiseaseList, basicMetaMapUpdated) 
    newPerturbationEnteroList = grepEnteroTypesFrom(newPerturbCaseList, basicMetaMapUpdated) 
    babyEnteroList = grepEnteroTypesFrom(babyList, basicMetaMapUpdated) 
    
    newGeoEnteroList = grepEnteroTypesFrom(newGeoList, basicMetaMapUpdated) 
    newContEnteroList = grepEnteroTypesFrom(newContByEnteroList, basicMetaMapUpdated) 
    newCaseEnteroList = grepEnteroTypesFrom(newCaseByEnteroList, basicMetaMapUpdated) 
    diseaseContComp = list(ETB_norm = newContEnteroList$ETB,
                           ETB_case = newCaseEnteroList$ETB,
                           ETF_norm = newContEnteroList$ETF,
                           ETF_case = newCaseEnteroList$ETF)
  }
  
  
  overallCaseContComparison = T
  if (overallCaseContComparison) {
    entGeoTab = getEntGeoPlot(diseaseContComp)
  }
  
  normalByGeoComparison = T
  if (normalByGeoComparison) {
  }
  
  caseContByGeoComparison = F
  if (caseContByGeoComparison) {
    names(newGeoList)
    names(newDiseaseList)
  }
  
  
  thaiList = list(thai=normUniqCountryList$Thailand,
                  shortImmigrant=thaiShort,
                  longImmigrant=thaiLong,
                  us=normUniqCountryList$US)
  
  entGeoTab = getEntGeoPlot(grepEnteroTypesFrom(thaiList, basicMetaMapUpdated))
  
  entGeoTab = getEntGeoPlot(newContEnteroList)
  entGeoTab = getEntGeoPlot(newGeoEnteroList)
  entGeoTab = getEntGeoPlot(newGeoEnteroList) #[c(1,2,3,4,5,6,7,8,9,11,10,12:17)]
  entGeoTab = getEntGeoPlot(newGeoEnteroList) #[c(1,2,4)]
  entGeoTab = getEntGeoPlot(newNormalEnteroList)
  entGeoTab = getEntGeoPlot(newDiseaseEnteroList, T, c(10,4,4,1))
  entGeoTab = getEntGeoPlot(newPerturbationEnteroList)
  entGeoTab = getEntGeoPlot(babyEnteroList, ordered = F)
  
  etTypesByRichnessMode = T
  if (etTypesByRichnessMode) {
    diseaseMetaData = basicMetaMapUpdated[basicMetaMapUpdated$sample.ID %in% newDiseaseIds,]
    etfMetaData = basicMetaMapUpdated[basicMetaMapUpdated$sample.ID %in% newEtfCountriesSamples,]
    etbMetaData = basicMetaMapUpdated[basicMetaMapUpdated$sample.ID %in% newEtbCountriesSamples,]
    etpMetaData = basicMetaMapUpdated[basicMetaMapUpdated$sample.ID %in% newEtpCountriesSamples,]
    
    entTabByRichness = getBinTab(basicMetaMapUpdated)
    #entTabByRichness = getBinTab(diseaseMetaData)
    entTabByRichness = getBinTab(etfMetaData)
    
    etfCol = "#77e2c799"
    etbCol = "#bee24599"
    etpCol = "#f35b38"
    
    tradVec = c(255,127,0,120)/255
    tradVec = c(255,127,0,200)/255
    tradCol = rgb(tradVec[1],tradVec[2],tradVec[3],tradVec[4])
    
    cjpVec = c(211,211,211,120)/255
    cjpVec = c(211,211,211,200)/255
    cjpCol = rgb(cjpVec[1],cjpVec[2],cjpVec[3],cjpVec[4])
    
    eurVec = c(117,112,179,120)/255
    eurVec = c(117,112,179,200)/255
    eurCol = rgb(eurVec[1],eurVec[2],eurVec[3],eurVec[4])
    
    etpCol = tradCol
    etbCol = cjpCol
    etfCol = eurCol
    
    
    
    colnames(entTabByRichness) = c("ET-Firmicutes","ET-Bacteroides","ET-Prevotella")
    rownames(entTabByRichness) = NULL
    #par(mar=c(10,10,10,10))
    par(mar=c(5,4,4,1))
    barplot(t(entTabByRichness),
            col=c(etfCol,etbCol,etpCol), las=2)
    legend("topright", 
           bty="n",
           inset = c(1.2,0),xpd=T,cex=0.7,
           legend=c("ET-Firmicutes","ET-Bacteroides","ET-Prevotella"),
           fill=c(etfCol,etbCol,etpCol))
  }
  
  etTypesByBmiMode = T
  if (etTypesByBmiMode) {
    
    View(basicMetaMapUpdated)
    
    imputedInd <- basicMetaMapUpdated$Imputation=="Imputed"
    nonImputedInd <- basicMetaMapUpdated$Imputation=="None"
    bmiInds <- basicMetaMapUpdated$WhatImputed  %in% c("AGX","AXX","XXX")
    bmiNaInds <- is.na(basicMetaMapUpdated$BMI)
    
    noImputedMetaMap = basicMetaMapUpdated[!bmiNaInds & nonImputedInd,]
    bmiOnlyMetaMap = basicMetaMapUpdated[!bmiNaInds & bmiInds & imputedInd, ]
    
    bmiAllMetaMap = rbind(noImputedMetaMap, bmiOnlyMetaMap)
    
    bmiEtfAllMetaMap = bmiAllMetaMap[bmiAllMetaMap$sample.ID %in% etfSamples, ]
    bmiEtbAllMetaMap = bmiAllMetaMap[bmiAllMetaMap$sample.ID %in% etbSamples, ]
    bmiEtpAllMetaMap = bmiAllMetaMap[bmiAllMetaMap$sample.ID %in% etpSamples, ]
    
    entTabByBMI = getBinTab(bmiEtpAllMetaMap, by="BMI")
    
    colnames(entTabByBMI) = c("ET-Firmicutes","ET-Bacteroides","ET-Prevotella")
    rownames(entTabByBMI) = NULL
    
    par(mar=c(5,4,4,1))
    barplot(t(entTabByBMI),
            col=c("#77e2c799","#bee24599","#f35b38"), las=2)
    legend("topright", 
           bty="n",
           inset = c(1.2,0),xpd=T,cex=0.7,
           legend=c("ET-Firmicutes","ET-Bacteroides","ET-Prevotella"),
           fill=c("#77e2c799","#bee24599","#f35b38"))
    
    
  }
  
  
  
}

checkEnteroType = T
if (checkEnteroType) {
  
  etfCol = "#77e2c799"
  etbCol = "#bee24599"
  etpCol = "#f35b38"
  
  tradVec = c(255,127,0,120)/255
  #tradVec = c(255,127,0,200)/255
  tradCol = rgb(tradVec[1],tradVec[2],tradVec[3],tradVec[4])
  
  cjpVec = c(211,211,211,120)/255
  #cjpVec = c(211,211,211,200)/255
  cjpCol = rgb(cjpVec[1],cjpVec[2],cjpVec[3],cjpVec[4])
  
  eurVec = c(117,112,179,120)/255
  #eurVec = c(117,112,179,200)/255
  eurCol = rgb(eurVec[1],eurVec[2],eurVec[3],eurVec[4])
  
  etpCol = tradCol
  etbCol = cjpCol
  etfCol = eurCol
  
  
  ### MAIN FIGURE FOR ET-types by geography ###
  geographyWithEnterotypeMode = T
  if (geographyWithEnterotypeMode) {
    
    genusMode = T
    if (genusMode) {
      
      notIndiGeos = basicMetaMapUpdated$Geography[match(newNormalIds, basicMetaMapUpdated$sample.ID)]
      notIndiEnterotype = basicMetaMapUpdated$enteroType[match(newNormalIds, basicMetaMapUpdated$sample.ID)]
      
      notIndiEnterptypeByGeos = split(notIndiEnterotype,notIndiGeos)
      entGeoTab = do.call(rbind,lapply(notIndiEnterptypeByGeos,  function(enterotypes){
        entCounts = getEntCount(enterotypes)
        return(entCounts)
      }))
      colnames(entGeoTab) = c("ET-Firmicutes","ET-Bacteroides","ET-Prevotella")
      entGeoTab = sweep(entGeoTab, MARGIN=1, rowSums(entGeoTab)/100, FUN="/")
      entGeoTab = entGeoTab[order(entGeoTab[,1]),]
      par(mar=c(8,3,2,1))
      seqs = c("Finland","Japan","China","US",
               "Luxembourg","Germany","Italy","Spain","Denmark","UK","Sweden")
      barplot(t(entGeoTab[seqs,]), #2:11, #t(entGeoTab[c(2,4,3,5,6,7,8,9),])
              col=c(etfCol,etbCol,etpCol), las=2)
      
      indiGeos = basicMetaMapUpdated$Geography[match(newNonWesternIds, basicMetaMapUpdated$sample.ID)]
      indiEnterotype = basicMetaMapUpdated$enteroType[match(newNonWesternIds, basicMetaMapUpdated$sample.ID)]
      
      indiEnterptypeByGeos = split(indiEnterotype,indiGeos)
      indiEntGeoTab = do.call(rbind,lapply(indiEnterptypeByGeos,  function(enterotypes){
        entCounts = getEntCount(enterotypes)
        return(entCounts)
      }))
      colnames(indiEntGeoTab) = c("ET-Firmicutes","ET-Bacteroides","ET-Prevotella")
      indiEntGeoTab = sweep(indiEntGeoTab, MARGIN=1, rowSums(indiEntGeoTab)/100, FUN="/")
      indiEntGeoTab = indiEntGeoTab[order(indiEntGeoTab[,2]),]
      par(mar=c(8,3,2,1))
      barplot(t(indiEntGeoTab),
              col=c(etfCol,etbCol,etpCol), las=2)
      allEntGeoTab = rbind(indiEntGeoTab, entGeoTab[seqs,])
      par(mar=c(8,3,2,1))
      barplot(t(allEntGeoTab),
              col=c(etfCol,etbCol,etpCol), las=2)
      
    }
    
    mgsMode = F
    if (mgsMode) {
      notIndiGeos = basicMetaMapUpdated$Geography[match(newNormalIds, basicMetaMapUpdated$sample.ID)]
      notIndiEnterotype = basicMetaMapUpdated$enteroTypeM[match(newNormalIds, basicMetaMapUpdated$sample.ID)]
      
      notIndiEnterptypeByGeos = split(notIndiEnterotype,notIndiGeos)
      entGeoTab = do.call(rbind,lapply(notIndiEnterptypeByGeos,  function(enterotypes){
        entCounts = getEntCount(enterotypes)
        return(entCounts)
      }))
      colnames(entGeoTab) = c("ET-Firmicutes","ET-Bacteroides","ET-Prevotella")
      entGeoTab = sweep(entGeoTab, MARGIN=1, rowSums(entGeoTab)/100, FUN="/")
      entGeoTab = entGeoTab[order(entGeoTab[,1]),]
      par(mar=c(8,4,4,1))
      barplot(t(entGeoTab[2:11,]), #t(entGeoTab[c(2,4,3,5,6,7,8,9),])
              col=c("#77e2c799","#bee24599","#f35b38"), las=2)
      
      indiGeos = basicMetaMapUpdated$Geography[match(newNonWesternIds, basicMetaMapUpdated$sample.ID)]
      indiEnterotype = basicMetaMapUpdated$enteroTypeM[match(newNonWesternIds, basicMetaMapUpdated$sample.ID)]
      
      indiEnterptypeByGeos = split(indiEnterotype,indiGeos)
      indiEntGeoTab = do.call(rbind,lapply(indiEnterptypeByGeos,  function(enterotypes){
        entCounts = getEntCount(enterotypes)
        return(entCounts)
      }))
      colnames(indiEntGeoTab) = c("ET-Firmicutes","ET-Bacteroides","ET-Prevotella")
      indiEntGeoTab = sweep(indiEntGeoTab, MARGIN=1, rowSums(indiEntGeoTab)/100, FUN="/")
      indiEntGeoTab = indiEntGeoTab[order(indiEntGeoTab[,2]),]
      par(mar=c(8,4,4,1))
      barplot(t(indiEntGeoTab),
              col=c("#77e2c799","#bee24599","#f35b38"), las=2)
      
      
    }
    
  }
  
  checkNextTime = F
  if (checkNextTime) {
    
    ### SUPPLEMENTARY FIGURES ###
    geographyAndDiseaseByEnterotypeMode = F
    if (geographyAndDiseaseByEnterotypeMode) {
      
      etfNormalSamples = etFirmicutesCountriesSamples[etFirmicutesCountriesSamples  %in% normalIDs]
      etbNormalSamples = etBacteroidesCountriesSamples[etBacteroidesCountriesSamples  %in% normalIDs]
      etfDiseaseSamples = etFirmicutesCountriesSamples[etFirmicutesCountriesSamples  %in% caseIDs]
      etbDiseaseSamples = etBacteroidesCountriesSamples[etBacteroidesCountriesSamples  %in% caseIDs]
      
      etfNormalEnterotypes = basicMetaMap$enteroType[basicMetaMap$sample.ID %in% etfNormalSamples]
      etbNormalEnterotypes = basicMetaMap$enteroType[basicMetaMap$sample.ID %in% etbNormalSamples]
      etfDiseaseEnterotypes = basicMetaMap$enteroType[basicMetaMap$sample.ID %in% etfDiseaseSamples]
      etbDiseaseEnterotypes = basicMetaMap$enteroType[basicMetaMap$sample.ID %in% etbDiseaseSamples]
      
      etfNormalCounts = getEntCount(etfNormalEnterotypes)
      etbNormalCounts = getEntCount(etbNormalEnterotypes)
      etfDiseaseCounts = getEntCount(etfDiseaseEnterotypes)
      etbDiseaseCounts = getEntCount(etbDiseaseEnterotypes)
      
      etDiseaseTab = rbind(ETF_norm=etfNormalCounts,
                           ETF_case=etfDiseaseCounts,
                           ETB_norm=etbNormalCounts,
                           ETB_case=etbDiseaseCounts)
      
      colnames(etDiseaseTab) = c("ET-Firmicutes","ET-Bacteroides","ET-Prevotella")
      etDiseaseTab = sweep(etDiseaseTab, MARGIN=1, rowSums(etDiseaseTab)/100, FUN="/")
      
      barplot(t(etDiseaseTab),
              col=c("#77e2c799","#bee24599","#f35b38"), las=2)
      
    }
    
    ### MAIN FIGURE FOR ET-types by richness ###
    richnessBinningWithEnterotypeMode = T
    if (richnessBinningWithEnterotypeMode) {
      
      loadMode = T
      if (loadMode) {
        diseaseMetaMap = basicMetaMapUpdated[basicMetaMapUpdated$sample.ID %in% newDiseaseIds,]
        normalMetaMap = basicMetaMapUpdated[basicMetaMapUpdated$sample.ID %in% newNormalAllIds,]
        westernizedMetaMap = basicMetaMapUpdated[basicMetaMapUpdated$sample.ID %in% newWesternIds,]
        indiMetaMap = basicMetaMapUpdated[basicMetaMapUpdated$sample.ID %in% newNonWesternIds,]
        
        
      }
      histMode = F
      if (histMode) {
        
        plot(density(normalMetaMap$GeneRichness), 
             ylim=c(0,2.5e-6), xaxt="n", yaxt="n", main="", xlab="")
        lines(density(diseaseMetaMap$GeneRichness), col="red") 
      }
      
      
      etTypesByRichnessMode = T
      if (etTypesByRichnessMode) {
        
        allMode = T
        diseaseMode = T
        normalMode = T
        westernizedMode = T
        nonWesternizedMode = T
        if (allMode) {
          allEntTabByRichnes = getBinTab(basicMetaMapUpdated)
          
          colnames(allEntTabByRichnes) = c("ET-Firmicutes","ET-Bacteroides","ET-Prevotella")
          rownames(allEntTabByRichnes) = NULL
          #par(mar=c(10,10,10,10))
          par(mar=c(5,4,4,1))
          barplot(t(allEntTabByRichnes),
                  col=c(eurCol,cjpCol,tradCol), las=2)
        }
        if (diseaseMode) {
          diseaseEntTabByRichnes = getBinTab(diseaseMetaMap)
          
          colnames(diseaseEntTabByRichnes) = c("ET-Firmicutes","ET-Bacteroides","ET-Prevotella")
          rownames(diseaseEntTabByRichnes) = NULL
          #par(mar=c(10,10,10,10))
          par(mar=c(5,4,4,1))
          barplot(t(diseaseEntTabByRichnes),
                  col=c(eurCol,cjpCol,tradCol), las=2)
          # legend("topright", 
          #        bty="n",
          #        inset = c(1.2,0),xpd=T,cex=0.7,
          #        legend=c("ET-Firmicutes","ET-Bacteroides","ET-Prevotella"),
          #        fill=c("#77e2c799","#bee24599","#f35b38"))
          
        }
        if (normalMode) {
          normalEntTabByRichnes = getBinTab(normalMetaMap)
          
          colnames(normalEntTabByRichnes) = c("ET-Firmicutes","ET-Bacteroides","ET-Prevotella")
          rownames(normalEntTabByRichnes) = NULL
          #par(mar=c(10,10,10,10))
          par(mar=c(5,4,4,1))
          barplot(t(normalEntTabByRichnes),
                  col=c(eurCol,cjpCol,tradCol), las=2)
          # legend("topright", 
          #        bty="n",
          #        inset = c(1.2,0),xpd=T,cex=0.7,
          #        legend=c("ET-Firmicutes","ET-Bacteroides","ET-Prevotella"),
          #        fill=c("#77e2c799","#bee24599","#f35b38"))
        }
        if (westernizedMode) {
          westernizedEntTabByRichnes = getBinTab(westernizedMetaMap)
          
          colnames(westernizedEntTabByRichnes) = c("ET-Firmicutes","ET-Bacteroides","ET-Prevotella")
          rownames(westernizedEntTabByRichnes) = NULL
          #par(mar=c(10,10,10,10))
          par(mar=c(5,4,4,1))
          barplot(t(westernizedEntTabByRichnes),
                  col=c(eurCol,cjpCol,tradCol), las=2)
        }
        if (nonWesternizedMode) {
          indiEntTabByRichnes = getBinTab(indiMetaMap)
          
          colnames(indiEntTabByRichnes) = c("ET-Firmicutes","ET-Bacteroides","ET-Prevotella")
          rownames(indiEntTabByRichnes) = NULL
          #par(mar=c(10,10,10,10))
          par(mar=c(5,4,4,1))
          barplot(t(indiEntTabByRichnes),
                  col=c(eurCol,cjpCol,tradCol), las=2)
        }
        
        
      }
      
      etTypesByRichnessMode = T
      if (etTypesByRichnessMode) {
        
      }
      
      
      
      
      
      
    }
    
    checkOverallEnterotype = F
    if (checkOverallEnterotype) {
      
      ### BREIF CHECK OVERALL ENTEROTYPES ###
      enterotypeCount = do.call(rbind, lapply(diseaseSets, function(datasetId) {
        
        caseEnteroTypes = basicMetaMap[basicMetaMap$dataset.ID==datasetId & basicMetaMap$type=="Case", "enteroType"]
        contEnteroTypes = basicMetaMap[basicMetaMap$dataset.ID==datasetId & basicMetaMap$type=="Control", "enteroType"]
        
        caseEtf <- sum(caseEnteroTypes == "ET-Firmicutes")
        caseEtb <- sum(caseEnteroTypes == "ET-Bacteroides")
        caseEtp <- sum(caseEnteroTypes == "ET-Prevotella")
        
        contEtf <- sum(contEnteroTypes == "ET-Firmicutes")
        contEtb <- sum(contEnteroTypes == "ET-Bacteroides")
        contEtp <- sum(contEnteroTypes == "ET-Prevotella")
        
        resEnteroTypes = c(caseEtf, caseEtb, caseEtp, contEtf, contEtb, contEtp)
        return(resEnteroTypes)
        
      }))
      colnames(enterotypeCount) = c("CaseEtF", "CaseEtB", "CaseEtP", "ContEtF", "ContEtB", "ContEtP")
      rownames(enterotypeCount) =  diseaseSets
      
      out = apply(enterotypeCount, 1, function(currRow) {
        currTab = makeCurrTab(currRow)
        chisqOut = chisq.test(currTab)
        pvalue = chisqOut$p.value
        pvalue = ifelse(is.nan(pvalue), 1, pvalue)
        return(pvalue)
      })
      sort(out[out<0.05])
      diseaseEnteroTypes = basicMetaMap$enteroType[basicMetaMap$sample.ID %in% caseIDs2]
      normalEnteroTypes = basicMetaMap$enteroType[basicMetaMap$sample.ID %in% normalIDs]
      indiEnteroTypes = basicMetaMap$enteroType[basicMetaMap$sample.ID %in% indiIDs]
      
      overallEnteroTypes = list(disease=diseaseEnteroTypes, 
                                normal = normalEnteroTypes, 
                                indigenous = indiEnteroTypes)
      
      overallEnteroTypeCount = do.call(rbind, lapply(overallEnteroTypes, function(currTypes) {
        caseEtf <- sum(currTypes == "ET-Firmicutes")
        caseEtb <- sum(currTypes == "ET-Bacteroides")
        caseEtp <- sum(currTypes == "ET-Prevotella")
        
        resEnteroTypes = c(caseEtf, caseEtb, caseEtp)
        return(resEnteroTypes)
      }))
      colnames(overallEnteroTypeCount) = c("ET-Firmicutes", "ET-Bacteroides", "ET-Prevotella")
      overallEnteroTypeCount = t(overallEnteroTypeCount)
      overallEnteroTypeCount = sweep(overallEnteroTypeCount, MARGIN=2, FUN="/", STAT=colSums(overallEnteroTypeCount))
      
    }
    
    if (F) {
      diseaseDatasetIds = mgsInfo$ID[mgsInfo$Classification=="disease"]
    } 
    
    enterotypeByPerturbation = T
    if (enterotypeByPerturbation) {
      immuneDataSets = c("id26","id41")
      enterotypeCountImmune = do.call(rbind, lapply(immuneDataSets, function(datasetId) {
        
        caseEnteroTypes = basicMetaMap[basicMetaMap$dataset.ID==datasetId & tolower(basicMetaMap$subtype) == "after", "enteroType"]
        contEnteroTypes = basicMetaMap[basicMetaMap$dataset.ID==datasetId & tolower(basicMetaMap$subtype) == "before", "enteroType"]
        
        caseEtf <- sum(caseEnteroTypes == "ET-Firmicutes")
        caseEtb <- sum(caseEnteroTypes == "ET-Bacteroides")
        caseEtp <- sum(caseEnteroTypes == "ET-Prevotella")
        
        contEtf <- sum(contEnteroTypes == "ET-Firmicutes")
        contEtb <- sum(contEnteroTypes == "ET-Bacteroides")
        contEtp <- sum(contEnteroTypes == "ET-Prevotella")
        
        resEnteroTypes = c(caseEtf, caseEtb, caseEtp, contEtf, contEtb, contEtp)
        return(resEnteroTypes)
        
      }))
      colnames(enterotypeCountImmune) = c("CaseEtF", "CaseEtB", "CaseEtP", "ContEtF", "ContEtB", "ContEtP")
      rownames(enterotypeCountImmune) =  immuneDataSets
      
      outImmune = apply(enterotypeCountImmune, 1, function(currRow) {
        currTab = makeCurrTab(currRow)
        chisqOut = chisq.test(currTab)
        pvalue = chisqOut$p.value
        pvalue = ifelse(is.nan(pvalue), 1, pvalue)
        return(pvalue)
      })
      sort(outImmune[outImmune<0.05])
    }
    
  }
  
}

updatePcoaOut = T
if (updatePcoaOut) {
  
  genusMatUpdated = getTaxaSumMat(mergeMatUpdated, taxo, "genus")
  familyMatUpdated = getTaxaSumMat(mergeMatUpdated, taxo, "family")
  orderMatUpdated = getTaxaSumMat(mergeMatUpdated, taxo, "order")
  classMatUpdated = getTaxaSumMat(mergeMatUpdated, taxo, "class")
  phylumMatUpdated = getTaxaSumMat(mergeMatUpdated, taxo, "phylum")
  
  mgsCs = rowSums(mergeMatUpdated)
  genusCs = rowSums(genusMatUpdated)
  familyCs = rowSums(familyMatUpdated)
  orderCs = rowSums(orderMatUpdated)
  classCs = rowSums(classMatUpdated)
  phylumCs = rowSums(phylumMatUpdated)
  
  table(mgsCs==0)
  table(genusCs==0)
  table(familyCs==0)
  table(orderCs==0)
  table(classCs==0)
  table(phylumCs==0)
  
  vegOut=vegdist(t(mergeMatUpdated),"bray")
  vegGenusOut=vegdist(t(genusMatUpdated[genusCs!=0,]),"bray")
  vegFamilyOut=vegdist(t(familyMatUpdated[familyCs!=0,]),"bray")
  vegOrderOut=vegdist(t(orderMatUpdated),"bray")
  vegClassOut=vegdist(t(classMatUpdated),"bray")
  vegPhylumOut=vegdist(t(phylumMatUpdated),"bray")
  
  pcoaOut=pcoa(vegOut)
  pcoaGenusOut=pcoa(vegGenusOut)
  pcoaFamilyOut=pcoa(vegFamilyOut)
  pcoaOrderOut=pcoa(vegOrderOut)
  pcoaClassOut=pcoa(vegClassOut)
  pcoaPhylumOut=pcoa(vegPhylumOut) 
  
  pcoaAllUpdatedData = "C:/Data/comparative.analysis.healthy.sweden/pcoaAllData.updated.20190805.RData"
  #load(pcoaAllUpdatedData)
  
  
  save(pcoaOut, 
       pcoaGenusOut, 
       pcoaFamilyOut,
       pcoaOrderOut, 
       pcoaClassOut, 
       pcoaPhylumOut, 
       file=pcoaAllUpdatedData)
  
  pcoaGenusMat = pcoaGenusOut$vectors[,1:2]
  
  pcoaCols = rep("#c0c0c088", dim(pcoaGenusMat)[1])
  names(pcoaCols) = rownames(pcoaGenusMat)
  pcoaCols[newNormalIds] = "#00bfff88"
  pcoaCols[newNonWesternIds] = "#ff240077"
  #pcoaCols[allEtbKeySamples] = "#80008077"
  plot(pcoaGenusMat[,1], pcoaGenusMat[,2], col=pcoaCols, pch=16) #, xlim=c(-0.55,0.4), ylim=c(-0.3,0.7)
  
  pcoaCols = rep("#c0c0c088", dim(pcoaGenusMat)[1])
  names(pcoaCols) = rownames(pcoaGenusMat)
  #pcoaCols[normalIDs] = "#00bfff88"
  pcoaCols[caseList$Case.id48.US] = "#ff240077"
  #pcoaCols[allEtbKeySamples] = "#80008077"
  plot(pcoaGenusMat[,1], pcoaGenusMat[,2], col=pcoaCols, pch=16) #, xlim=c(-0.55,0.4), ylim=c(-0.3,0.7)
  
}

richnessBoxPlotMode = T
if (richnessBoxPlotMode) {
  
  loadSampleInfo = T
  if (loadSampleInfo) {
    
    newCaseContByGeoList = list(Sweden_D=c(newDiseaseList$Atherosclerosis_SE, 
                                           newDiseaseList$T2D_SE),
                                Sweden_N=c(newNormalList$Sweden_id28,
                                           newNormalList$Sweden_id1,
                                           newNormalList$Sweden_id14),
                                Denmark_D = c(newDiseaseList$Obesity_DK_16,
                                              newDiseaseList$Obesity_DK_25),
                                Denmark_N=c(newNormalList$Denmark_id16, 
                                            newNormalList$Denmark_id25),
                                UK_D=c(newDiseaseList$Cirrhosis_GB),
                                UK_N=c(newNormalList$UK_id29,
                                       newNormalList$UK_id39),
                                Spain_D=c(newDiseaseList$IBD_ES,
                                          newDiseaseList$T2D_ES),
                                Spain_N=c(newNormalList$Spain_id17),
                                Italy_D=c(newDiseaseList$NAFLD_IT,
                                          newDiseaseList$CRC_IT),
                                Italy_N=c(newNormalList$Italy_id37,
                                          newNormalList$Italy_id46),
                                Germany_D=c(newDiseaseList$CRC_DE,
                                            newDiseaseList$Parkinson_DE),
                                Germany_N=c(newNormalList$Germany_id44,
                                            newNormalList$Germany_id8),
                                Luxembourg_D=c(newDiseaseList$T1D_LU),
                                Luxembourg_N=c(newNormalList$Luxembourg_id35),
                                Japan_D=c(newDiseaseList$CRC_JP),
                                Japan_N=c(newNormalList$Japan_id45),
                                US_D=c(newDiseaseList$CRC_US,
                                       newDiseaseList$Melanoma_US,
                                       newDiseaseList$CFS_US,
                                       newDiseaseList$NAFLD_US),
                                US_N=c(newNormalList$US_id11,
                                       newNormalList$US_id32,
                                       newNormalList$US_id34,
                                       newNormalList$US_id36,
                                       newNormalList$US_id43),
                                China_D=c(newDiseaseList$CVD_CN,
                                          newDiseaseList$GDM_CN,
                                          newDiseaseList$T2D_CN,
                                          newDiseaseList$Cirrhosis_CN,
                                          newDiseaseList$Rheumatoid_CN,
                                          newDiseaseList$Crohn_CN,
                                          newDiseaseList$Ankylosing_CN,
                                          newDiseaseList$Behcet_CN),
                                China_N=c(newNormalList$China_id10,
                                          newNormalList$China_id12,
                                          newNormalList$China_id2,
                                          newNormalList$China_id20,
                                          newNormalList$China_id31,
                                          newNormalList$China_id52,
                                          newNormalList$China_id6,
                                          newNormalList$China_id9))
    
    newCaseContByGeoListNew = list(Sweden = list(Atherosclerosis= newDiseaseList$Atherosclerosis_SE,
                                                 T2D=newDiseaseList$T2D_SE,
                                                 IGT=diseaseUniqSubjectList$IGT_SW_id14,
                                                 Normal=c(newNormalList$Sweden_id28,
                                                          newNormalList$Sweden_id1,
                                                          newNormalList$Sweden_id14)),
                                   Denmark = list(Obesity=newDiseaseList$Obesity_DK_16,
                                                  Obesity=newDiseaseList$Obesity_DK_25,
                                                  Normal=c(newNormalList$Denmark_id16, 
                                                           newNormalList$Denmark_id25)),
                                   UK = list(Cirrhosis=newDiseaseList$Cirrhosis_GB,
                                             Normal = c(newNormalList$UK_id29,
                                                        newNormalList$UK_id39)),
                                   Spain = list(IBD=newDiseaseList$IBD_ES, 
                                                T2D=newDiseaseList$T2D_ES,
                                                Normal=newNormalList$Spain_id17),
                                   Italy = list(NAFLD=newDiseaseList$NAFLD_IT,
                                                CRC=newDiseaseList$CRC_IT,
                                                Normal=c(newNormalList$Italy_id37,
                                                         newNormalList$Italy_id46)),
                                   Germany = list(CRC=newDiseaseList$CRC_DE,
                                                  Parkinson=newDiseaseList$Parkinson_DE,
                                                  Normal=c(newNormalList$Germany_id44,
                                                           newNormalList$Germany_id8)),
                                   Finland = list(T1D = newDiseaseList$T1D_FI,
                                                  Normal = newNormalList$Finland_id21),
                                   Luxembourg = list(T1D=newDiseaseList$T1D_LU,
                                                     Normal=newNormalList$Luxembourg_id35),
                                   Japan = list(CRC=newDiseaseList$CRC_JP,
                                                Normal=newNormalList$Japan_id45),
                                   US = list(CRC=newDiseaseList$CRC_US,
                                             Melanoma=newDiseaseList$Melanoma_US,
                                             CFS=newDiseaseList$CFS_US,
                                             NAFLD=newDiseaseList$NAFLD_US,
                                             Normal=c(newNormalList$US_id11,
                                                      newNormalList$US_id32,
                                                      newNormalList$US_id34,
                                                      newNormalList$US_id36,
                                                      newNormalList$US_id43)),
                                   China = list(CVD=newDiseaseList$CVD_CN,
                                                GDM=newDiseaseList$GDM_CN,
                                                T2D=newDiseaseList$T2D_CN,
                                                Cirrhosis=newDiseaseList$Cirrhosis_CN,
                                                Rhematoid=newDiseaseList$Rheumatoid_CN,
                                                Crohn=newDiseaseList$Crohn_CN,
                                                Ankylosing=newDiseaseList$Ankylosing_CN,
                                                Behcet=newDiseaseList$Behcet_CN,
                                                Normal=c(newNormalList$China_id10,
                                                         newNormalList$China_id12,
                                                         newNormalList$China_id2,
                                                         newNormalList$China_id20,
                                                         newNormalList$China_id31,
                                                         newNormalList$China_id52,
                                                         newNormalList$China_id6,
                                                         newNormalList$China_id9)))
    
    newCaseContByGeoListNewWithFoodborne = list(Sweden = list(Atherosclerosis= newDiseaseList$Atherosclerosis_SE,
                                                              T2D=newDiseaseList$T2D_SE,
                                                              IGT=diseaseUniqSubjectList$IGT_SW_id14,
                                                              Normal=c(newNormalList$Sweden_id28,
                                                                       newNormalList$Sweden_id1,
                                                                       newNormalList$Sweden_id14)),
                                                Denmark = list(Obesity=newDiseaseList$Obesity_DK_16,
                                                               Obesity=newDiseaseList$Obesity_DK_25,
                                                               Antibiotics = newPerturbCaseList$antibiotics_DK,
                                                               Normal=c(newNormalList$Denmark_id16, 
                                                                        newNormalList$Denmark_id25)),
                                                UK = list(Cirrhosis=newDiseaseList$Cirrhosis_GB,
                                                          Normal = c(newNormalList$UK_id29,
                                                                     newNormalList$UK_id39)),
                                                Spain = list(IBD=newDiseaseList$IBD_ES, 
                                                             T2D=newDiseaseList$T2D_ES,
                                                             Normal=newNormalList$Spain_id17),
                                                Italy = list(NAFLD=newDiseaseList$NAFLD_IT,
                                                             CRC=newDiseaseList$CRC_IT,
                                                             Normal=c(newNormalList$Italy_id37,
                                                                      newNormalList$Italy_id46)),
                                                Germany = list(CRC=newDiseaseList$CRC_DE,
                                                               Parkinson=newDiseaseList$Parkinson_DE,
                                                               Normal=c(newNormalList$Germany_id44,
                                                                        newNormalList$Germany_id8)),
                                                Finland = list(T1D = newDiseaseList$T1D_FI,
                                                               Normal = newNormalList$Finland_id21),
                                                Luxembourg = list(T1D=newDiseaseList$T1D_LU,
                                                                  Normal=newNormalList$Luxembourg_id35),
                                                France = list(RCC=newDiseaseList$RCC_FR,
                                                              NSCLC=newDiseaseList$NSCLC_FR),
                                                
                                                Japan = list(CRC=newDiseaseList$CRC_JP,
                                                             Normal=newNormalList$Japan_id45),
                                                US = list(CRC=newDiseaseList$CRC_US,
                                                          Melanoma=newDiseaseList$Melanoma_US,
                                                          CFS=newDiseaseList$CFS_US,
                                                          NAFLD=newDiseaseList$NAFLD_US,
                                                          Food=newPerturbCaseList$foodborne_US,
                                                          Preterm=newPerturbCaseList$Preterm_US,
                                                          Normal=c(newNormalList$US_id11,
                                                                   newNormalList$US_id32,
                                                                   newNormalList$US_id34,
                                                                   newNormalList$US_id36,
                                                                   newNormalList$US_id43)),
                                                China = list(CVD=newDiseaseList$CVD_CN,
                                                             GDM=newDiseaseList$GDM_CN,
                                                             T2D=newDiseaseList$T2D_CN,
                                                             Cirrhosis=newDiseaseList$Cirrhosis_CN,
                                                             Rhematoid=newDiseaseList$Rheumatoid_CN,
                                                             Crohn=newDiseaseList$Crohn_CN,
                                                             Ankylosing=newDiseaseList$Ankylosing_CN,
                                                             Behcet=newDiseaseList$Behcet_CN,
                                                             Normal=c(newNormalList$China_id10,
                                                                      newNormalList$China_id12,
                                                                      newNormalList$China_id2,
                                                                      newNormalList$China_id20,
                                                                      newNormalList$China_id31,
                                                                      newNormalList$China_id52,
                                                                      newNormalList$China_id6,
                                                                      newNormalList$China_id9)))
    
    if (F) {
      newContByGeoListNew = list(Sweden = c(newNormalList$Sweden_id28,
                                            newNormalList$Sweden_id1,
                                            newNormalList$Sweden_id14),
                                 Denmark = c(newNormalList$Denmark_id16, 
                                             newNormalList$Denmark_id25),
                                 UK = c(newNormalList$UK_id29,
                                        newNormalList$UK_id39),
                                 Spain = newNormalList$Spain_id17,
                                 Italy = c(newNormalList$Italy_id37,
                                           newNormalList$Italy_id46),
                                 Germany = c(newNormalList$Germany_id44,
                                             newNormalList$Germany_id8),
                                 Finland = newNormalList$Finland_id21,
                                 Luxembourg = newNormalList$Luxembourg_id35,
                                 Japan = newNormalList$Japan_id45,
                                 US = c(newNormalList$US_id11,
                                        newNormalList$US_id32,
                                        newNormalList$US_id34,
                                        newNormalList$US_id36,
                                        newNormalList$US_id43),
                                 China = c(newNormalList$China_id10,
                                           newNormalList$China_id12,
                                           newNormalList$China_id2,
                                           newNormalList$China_id20,
                                           newNormalList$China_id31,
                                           newNormalList$China_id52,
                                           newNormalList$China_id6,
                                           newNormalList$China_id9),
                                 Peru=newNonWesternList$Peru_id32,
                                 Mongolia=newNonWesternList$Mongo_id33,
                                 Thailand = newNonWesternList$Thai_id43,
                                 Fiji = newNonWesternList$Fiji_id49,
                                 India = newNonWesternList$India_id50,
                                 Tanzania = newNonWesternList$Tanzania_id37,
                                 Madagascar = newNonWesternList$Madagascar_id42
      )
      
    }
    
    
    
  }
  
  newCirclePlotMode = T
  if (newCirclePlotMode) {
    #newCaseContRichnessNewByGeoList = grepGeneRichnesssFromNew(newCaseContByGeoListNew, basicMetaMapUpdated)
    newCaseContRichnessNewByGeoList = grepGeneRichnesssFromNew(newCaseContByGeoListNewWithFoodborne, basicMetaMapUpdated)
    #newCaseContRichnessNewByGeoList = grepGeneRichnesssFromNew(newContByGeoListNew, basicMetaMapUpdated)
    
    newCaseContMeanRichnessNewByGeoList = lapply(newCaseContRichnessNewByGeoList, function(cohortList){
      meanRichness = lapply(cohortList, function(currRichness){
        return(mean(currRichness))
      })
      return(meanRichness)
    })
    newCaseContMeanRichnessNewByGeoMelt = melt(newCaseContMeanRichnessNewByGeoList)
    
    newCaseContCvRichnessNewByGeoList = lapply(newCaseContRichnessNewByGeoList, function(cohortList){
      meanRichness = lapply(cohortList, function(currRichness){
        meanRichness = mean(currRichness)
        sdRichness = sd(currRichness)
        return(sdRichness/meanRichness)
      })
      return(meanRichness)
    })
    newCaseContCvRichnessNewByGeoMelt = melt(newCaseContCvRichnessNewByGeoList)
    
    newCaseContStatsRichnessNewByGeoTab = data.frame(Geo=newCaseContMeanRichnessNewByGeoMelt$L1, 
                                                     Cohort=newCaseContMeanRichnessNewByGeoMelt$L2,
                                                     Mean=newCaseContMeanRichnessNewByGeoMelt$value,
                                                     CV=newCaseContCvRichnessNewByGeoMelt$value)
    newCaseContStatsRichnessNewByGeoTab$Geo = factor(newCaseContStatsRichnessNewByGeoTab$Geo, 
                                                     levels=c("Denmark", 
                                                              "Sweden",
                                                              "Spain",
                                                              "Luxembourg",
                                                              "Germany",
                                                              "UK",
                                                              "Italy",
                                                              "France",
                                                              "US",
                                                              "China",
                                                              "Japan",
                                                              "Finland"))
    newCaseContStatsRichnessNewByGeoTab$Class = "Disease"
    newCaseContStatsRichnessNewByGeoTab$Class[newCaseContStatsRichnessNewByGeoTab$Cohort=="Normal"]="Normal"
    newCaseContStatsRichnessNewByGeoTab$Class = factor(newCaseContStatsRichnessNewByGeoTab$Class)
    disCol="#FFD70044"
    
    ggplot(newCaseContStatsRichnessNewByGeoTab, aes(x=Geo, y=Mean, size=CV)) +
      geom_point(aes(fill=factor(Class)), colour="black", shape=21) +
      xlab("") + ylab("") +
      scale_size(range = c(2, 10))+
      scale_fill_manual(values=c(Normal=eurCol, Disease=disCol)) +
      theme(panel.grid.major = element_line(colour="gray"), 
            axis.text.x = element_text(angle = 60, hjust=1),
            axis.ticks = element_blank(),
            legend.position = "none",
            panel.background = element_rect(fill = "white", 
                                            colour = "black", 
                                            size = 0.5, 
                                            linetype = "solid")) 
    
  }
  
  oldBoxplotMode = F
  if (oldBoxplotMode) {
    
    newCaseContRichnessByGeoList = grepGeneRichnesssFrom(newCaseContByGeoList, basicMetaMapUpdated)
    newCaseContRichnessByGeoTab = melt(newCaseContRichnessByGeoList)
    boxConSeqs = c("Denmark_N",
                   "Denmark_D",
                   "Sweden_N",
                   "Sweden_D",
                   "Spain_N",
                   "Spain_D",
                   "Luxembourg_N",
                   "Luxembourg_D",
                   "Germany_N",
                   "Germany_D",
                   "UK_N",
                   "UK_D",
                   "Italy_N",
                   "Italy_D",
                   "US_N",
                   "US_D",
                   "China_N",
                   "China_D",
                   "Japan_N",
                   "Japan_D")
    gsub(".*_","","Japan_N")
    
    disCol="#FFD70044"
    
    richColVec = c("Denmark_N"=eurCol,
                   "Denmark_D"=disCol,
                   "Sweden_N"=eurCol,
                   "Sweden_D"=disCol,
                   "Spain_N"=eurCol,
                   "Spain_D"=disCol,
                   "Germany_N"=eurCol,
                   "Germany_D"=disCol,
                   "Luxembourg_N"=eurCol,
                   "Luxembourg_D"=disCol,
                   "UK_N"=eurCol,
                   "UK_D"=disCol,
                   "Italy_N"=eurCol,
                   "Italy_D"=disCol,
                   "US_N"=cjpCol,
                   "US_D"=disCol,
                   "China_N"=cjpCol,
                   "China_D"=disCol,
                   "Japan_N"=cjpCol,
                   "Japan_D"=disCol)
    newCaseContRichnessByGeoTab$L1=factor(newCaseContRichnessByGeoTab$L1, levels = boxConSeqs)
    newCaseContRichnessByGeoTab$type=gsub(".*_","", newCaseContRichnessByGeoTab$L1)
    
    ggplot(newCaseContRichnessByGeoTab, aes(x=L1, y=value, fill=L1)) + geom_boxplot() + 
      scale_fill_manual(values = richColVec) +
      xlab("") + ylab("") +
      theme(panel.grid.major = element_blank(), 
            panel.grid.minor = element_blank(),
            axis.text.x = element_text(angle = 60, hjust=1),
            axis.ticks = element_blank(),
            legend.position = "none",
            panel.background = element_rect(fill = "white", 
                                            colour = "black", 
                                            size = 0.5, 
                                            linetype = "solid")) 
    
    t.test(newCaseContRichnessByGeoList$Denmark_D, newCaseContRichnessByGeoList$Denmark_N)
    t.test(newCaseContRichnessByGeoList$Sweden_D, newCaseContRichnessByGeoList$Sweden_N)
    t.test(newCaseContRichnessByGeoList$Spain_D, newCaseContRichnessByGeoList$Spain_N)
    t.test(newCaseContRichnessByGeoList$China_D, newCaseContRichnessByGeoList$China_N)
    t.test(newCaseContRichnessByGeoList$US_D, newCaseContRichnessByGeoList$US_N)
    t.test(newCaseContRichnessByGeoList$Japan_D, newCaseContRichnessByGeoList$Japan_N)
    t.test(newCaseContRichnessByGeoList$UK_D, newCaseContRichnessByGeoList$UK_N)
    t.test(newCaseContRichnessByGeoList$Italy_D, newCaseContRichnessByGeoList$Italy_N)
    t.test(newCaseContRichnessByGeoList$Germany_D, newCaseContRichnessByGeoList$Germany_N)
    t.test(newCaseContRichnessByGeoList$Luxembourg_D, newCaseContRichnessByGeoList$Luxembourg_N)
    
    par(mfrow=c(2,1))
    par(mar=c(4,4,2,1))
    boxplot(rev(newCaseContRichnessByGeoList), las=2)
    plot.new()
    
  }
  
  oldBoxPlotMode2 = F
  if (oldBoxPlotMode2) {
    orderedCountries = c("Tanzania","Madagascar","Peru","India","Fiji","Mongo","Thai",
                         "Japan","China","US", "Finlan",
                         "Sweden","Denmark","UK","Spain","Germany","Luxembourg","Italy")
    
    europeCountries = c("Denmark", "Germany", "Italy", "Luxembourg", "Spain", "Sweden", "UK")
    tradCountries = c("Fiji", "India", "Madagascar", "Mongo", "Peru", "Tanzania", "Thai")
    cjpCountries = c("US","China","Japan")
    
    
    newNormalMetadata = basicMetaMapUpdated[match(newNormalAllIds_wo_FI, basicMetaMapUpdated$sample.ID),]
    newNormalRichnessByGeoList = split(newNormalMetadata$GeneRichness, newNormalMetadata$Geography )
    newNormalRichnessByGeoTab = melt(newNormalRichnessByGeoList)
    newNormalRichnessByGeoTab$L1[newNormalRichnessByGeoTab$L1 %in% cjpCountries ] = "UsChinaJapan"
    newNormalRichnessByGeoTab$L1[newNormalRichnessByGeoTab$L1 %in% tradCountries ] = "Traditional"
    newNormalRichnessByGeoTab$L1[newNormalRichnessByGeoTab$L1 %in% europeCountries ] = "European"
    newNormalRichnessByGeoTab$L1 = factor(newNormalRichnessByGeoTab$L1, levels = c("European","UsChinaJapan","Traditional"))
    ggplot(newNormalRichnessByGeoTab, aes(x=L1, y=value, fill=L1)) + geom_boxplot() + scale_fill_brewer("Set1") +
      theme(legend.position = "none")
    ss = split(newNormalRichnessByGeoTab$value, newNormalRichnessByGeoTab$L1)
    t.test(ss$European, ss$Traditional)
  }
  
}

richnessVolcanoModeUpdated = T
if (richnessVolcanoModeUpdated) {
  
  afterUpdate = T
  if (afterUpdate) {
    statsGenusForAllSamples = "J:\\Deposit\\Project\\2018_microbiome_atlas\\atlas.volcanot.out\\richness.volcano.stats.genus.for.all.txt"
    statsGenusForNormSamples = "J:\\Deposit\\Project\\2018_microbiome_atlas\\atlas.volcanot.out\\richness.volcano.stats.genus.for.normal.txt"
    statsGenusForDiseaseSamples = "J:\\Deposit\\Project\\2018_microbiome_atlas\\atlas.volcanot.out\\richness.volcano.stats.genus.for.disease.txt"
    statsGenusForIndustrialSamples  = "J:\\Deposit\\Project\\2018_microbiome_atlas\\atlas.volcanot.out\\richness.volcano.stats.genus.for.industrial.txt"
    statsGenusForTraditionalSamples  = "J:\\Deposit\\Project\\2018_microbiome_atlas\\atlas.volcanot.out\\richness.volcano.stats.genus.for.traditional.txt"
    
    statsMgsForAllSamples = "J:\\Deposit\\Project\\2018_microbiome_atlas\\atlas.volcanot.out\\richness.volcano.stats.mgs.for.all.txt"
    statsMgsForNormSamples = "J:\\Deposit\\Project\\2018_microbiome_atlas\\atlas.volcanot.out\\richness.volcano.stats.mgs.for.normal.txt"
    statsMgsForDiseaseSamples = "J:\\Deposit\\Project\\2018_microbiome_atlas\\atlas.volcanot.out\\richness.volcano.stats.mgs.for.disease.txt"
    statsMgsForIndustrialSamples  = "J:\\Deposit\\Project\\2018_microbiome_atlas\\atlas.volcanot.out\\richness.volcano.stats.mgs.for.industrial.txt"
    statsMgsForTraditionalSamples  = "J:\\Deposit\\Project\\2018_microbiome_atlas\\atlas.volcanot.out\\richness.volcano.stats.mgs.for.traditional.txt"
    
    lCut=0.25
    hCut=0.75
    
    
    richnessHist = T
    if (richnessHist) {
      plot(density(basicMetaMapUpdated$GeneRichness),  main="")
      abline(v=hAllCutNumber, col="red")
      abline(v=lAllCutNumber, col="red")
      
    }
    
    richness = basicMetaMapUpdated$GeneRichness
    names(richness) = basicMetaMapUpdated$sample.ID
    
    normalRichness = richness[names(richness) %in% newNormalAllIds]
    diseaseRichness = richness[names(richness) %in% newDiseaseIds]
    industrialRichness = richness[names(richness) %in% newWesternIds]
    
    hHalfCutNumber=as.numeric(quantile(basicMetaMapUpdated$GeneRichness,0.5))
    
    hAllCutNumber=as.numeric(quantile(basicMetaMapUpdated$GeneRichness,hCut))
    lAllCutNumber=as.numeric(quantile(basicMetaMapUpdated$GeneRichness,lCut))
    
    print(hAllCutNumber) #727622.5
    print(lAllCutNumber) #441994.5
    
    hgcSamples = names(richness[richness>=hAllCutNumber])
    lgcSamples = names(richness[richness<=lAllCutNumber])
    
    
    
    
    hgcHalfSamples = names(richness[richness>=hHalfCutNumber])
    lgcHalfSamples = names(richness[richness<=hHalfCutNumber])
    
    sort(table(basicMetaMapUpdated$Geography[basicMetaMapUpdated$sample.ID %in% hgcSamples] ))
    sort(table(basicMetaMapUpdated$Geography[basicMetaMapUpdated$sample.ID %in% lgcSamples] ))
    
    par(mar=c(5,10,4,1))
    barplot(sort(table(basicMetaMapUpdated$Geography[basicMetaMapUpdated$sample.ID %in% lgcSamples] )),
            horiz = T,
            las=1)
    barplot(sort(table(basicMetaMapUpdated$Geography[basicMetaMapUpdated$sample.ID %in% hgcSamples] )),
            horiz = T,
            las=1)
    
    caseContFraction = (cbind(hgc=c(sum(hgcSamples %in% newDiseaseIds), sum(hgcSamples %in% newNormalAllIds)),
                              lgc=c(sum(lgcSamples %in% newDiseaseIds), sum(lgcSamples %in% newNormalAllIds))))
    barplot(cbind(HGR=caseContFraction[,1]/1299, LGR=caseContFraction[,2]/1238))
    
    
    
    length(newDiseaseIds)
    
    hgcPrevMat=mergeMatUpdated[rownames(mergeMatUpdated) %in% taxo$MSP[taxo$genus=="Prevotella"],hgcHalfSamples]
    lgcPrevMat=mergeMatUpdated[rownames(mergeMatUpdated) %in% taxo$MSP[taxo$genus=="Prevotella"],lgcHalfSamples]
    hgcPrevMat=hgcPrevMat[,colnames(hgcPrevMat) %in% basicMetaMapUpdated$sample.ID[basicMetaMapUpdated$enteroType=="ET-Prevotella"]]
    lgcPrevMat=lgcPrevMat[,colnames(lgcPrevMat) %in% basicMetaMapUpdated$sample.ID[basicMetaMapUpdated$enteroType=="ET-Prevotella"]]
    
    hgcPrevMeans = rowMeans(hgcPrevMat)
    lgcPrevMeans = rowMeans(lgcPrevMat)
    data.frame(hgcPrevMeans, lgcPrevMeans)
    
    
    plot(hgcPrevMeans, lgcPrevMeans, log="xy")
    abline(a=0,b=1)
    length(hgcSamples)
    length(lgcSamples)
    
    normalHgcSamples = newNormalAllIds[newNormalAllIds %in% hgcSamples]
    normalLgcSamples = newNormalAllIds[newNormalAllIds %in% lgcSamples]
    
    normHgcSampleTab = data.frame(sample=normalHgcSamples, type="HGC.normal")
    normLgcSampleTab = data.frame(sample=normalLgcSamples, type="LGC.normal")
    normSampleTab = rbind(normHgcSampleTab, normLgcSampleTab)
    
    writeMode = F
    if (writeMode) write.table(normSampleTab, file="C://Data//comparative.analysis.healthy.sweden//reza.HGC.LGC.normal.samples.txt", sep="\t")
    
    checkInflowOutflowMode = T
    if (checkInflowOutflowMode) {
      volcano.stats.mgs.all = getVolcanoStat(mergeMatUpdated, hgcSamples, lgcSamples, taxo, T)
      volcano.stats.mgs.normal = getVolcanoStat(mergeMatUpdated, normalHgcSamples, normalLgcSamples, taxo, T)
      adjpCut=1e-3
      volcano.stats.mgs.normal.sel = volcano.stats.mgs.normal[volcano.stats.mgs.normal$qvalue <= adjpCut,]
      volcano.stats.mgs.normal.sel.hgc = volcano.stats.mgs.normal.sel[volcano.stats.mgs.normal.sel$lfc > 2,]
      volcano.stats.mgs.normal.sel.lgc = volcano.stats.mgs.normal.sel[volcano.stats.mgs.normal.sel$lfc < -2,]
      
      hgcSpecies = rownames(volcano.stats.mgs.normal.sel.hgc)
      lgcSpecies = rownames(volcano.stats.mgs.normal.sel.lgc)
      
      
      hgcLgcInflowList=list(HGR=inflow[hgcSpecies], LGR=inflow[lgcSpecies])
      hgcLgcOutflowList=list(HGR=outflow[hgcSpecies], LGR=outflow[lgcSpecies])
      hgcLgcInflowMelt = melt(hgcLgcInflowList)
      hgcLgcOutflowMelt = melt(hgcLgcOutflowList)
      
      t.test(hgcLgcInflowList$HGR, hgcLgcInflowList$LGR)
      t.test(hgcLgcOutflowList$HGR, hgcLgcOutflowList$LGR)
      
      wilcox.test(hgcLgcInflowList$HGR, hgcLgcInflowList$LGR, alternative = "less")
      wilcox.test(hgcLgcOutflowList$HGR, hgcLgcOutflowList$LGR, alternative = "less")
      
      
      ggplot(hgcLgcInflowMelt) + geom_boxplot(aes(x=L1, y=value, fill=L1)) +
        scale_fill_manual(values = c("#ff000055","#2c528c55")) +
        xlab("") +
        ylab("") + 
        theme(panel.grid.major = element_blank(), 
              panel.grid.minor = element_blank(),
              legend.position = "none",
              axis.text.x = element_text(angle = 60, hjust=1),
              panel.background = element_rect(fill = "white", 
                                              colour = "black", 
                                              size = 0.5, 
                                              linetype = "solid"))
      
      ggplot(hgcLgcOutflowMelt) + geom_boxplot(aes(x=L1, y=value, fill=L1)) +
        scale_fill_manual(values = c("#ff000055","#2c528c55")) +
        xlab("") +
        ylab("") + 
        theme(panel.grid.major = element_blank(), 
              panel.grid.minor = element_blank(),
              legend.position = "none",
              axis.text.x = element_text(angle = 60, hjust=1),
              panel.background = element_rect(fill = "white", 
                                              colour = "black", 
                                              size = 0.5, 
                                              linetype = "solid"))
    }
    
    checkInflowOutflowFractionAbundance = T
    if (checkInflowOutflowFractionAbundance) {
      
      diseaseHgcSamples = newDiseaseIds[newDiseaseIds %in% hgcSamples]
      diseaseLgcSamples = newDiseaseIds[newDiseaseIds %in% lgcSamples]
      
      length(diseaseHgcSamples)
      length(diseaseLgcSamples)
      
      
      industrialHgcSamples = newWesternIds[newWesternIds %in% hgcSamples]
      industrialLgcSamples = newWesternIds[newWesternIds %in% lgcSamples]
      
      length(industrialHgcSamples)
      length(industrialLgcSamples)
      
      traditionalHgcSamples = newNonWesternIds[newNonWesternIds %in% hgcSamples]
      traditionalLgcSamples = newNonWesternIds[newNonWesternIds %in% lgcSamples]
      
      
      
      boxplot(list(hgc=richness[hgcSamples], lgc=richness[lgcSamples]))
      
      table(hgcSpecies %in% inflowSpecies)
      table(hgcSpecies %in% outflowSpecies)
      
      table(lgcSpecies %in% inflowSpecies)
      table(lgcSpecies %in% outflowSpecies)
      
      mergeMatUpdatedDigit = mergeMatUpdated != 0
      
      
      getSigFraction <- function(mgsMatDigit, targetSamples, targetSpecies,pCut=0.05) {
        targetMat = mgsMatDigit[,targetSamples]
        rowSpecies = rownames(targetMat)
        result = apply(targetMat, 2, function(currCol){
          currSpecies = rowSpecies[currCol != 0]
          overlapSpecies = currSpecies[currSpecies %in% targetSpecies]
          totalSpecies = unique(c(currSpecies, targetSpecies))
          hyperP = getHyperP(targetSpecies, currSpecies, rowSpecies,F)
          #jaccard = length(overlapSpecies)/length(currSpecies)
          return((hyperP))
        })
        
        return(mean(result < pCut))
      }
      
      inflow.outflow.sig.fraction.list = c(inflow.hgc.health = getSigFraction(mergeMatUpdatedDigit, industrialHgcSamples, inflowSpecies),
                                           outflow.hgc.health = getSigFraction(mergeMatUpdatedDigit, industrialHgcSamples, outflowSpecies),
                                           inflow.lgc.health = getSigFraction(mergeMatUpdatedDigit, industrialLgcSamples, inflowSpecies),
                                           outflow.lgc.health = getSigFraction(mergeMatUpdatedDigit, industrialLgcSamples, outflowSpecies),
                                           inflow.hgc.disease = getSigFraction(mergeMatUpdatedDigit, diseaseHgcSamples, inflowSpecies),
                                           outflow.hgc.disease = getSigFraction(mergeMatUpdatedDigit, diseaseHgcSamples, outflowSpecies),
                                           inflow.lgc.disease = getSigFraction(mergeMatUpdatedDigit, diseaseLgcSamples, inflowSpecies),
                                           outflow.lgc.disease = getSigFraction(mergeMatUpdatedDigit, diseaseLgcSamples, outflowSpecies))
      
      save(melted, file = "C:\\Data\\comparative.analysis.healthy.sweden\\melted.inflow.outflow.sig.fraction.0.05.RData")
      
      melted = melt(inflow.outflow.sig.fraction.list)
      melted$health = c(rep("health",4),rep("disease",4))
      melted$Richness = c(rep(c("HGC","HGC","LGC","LGC"),2))
      melted$in.outflow = c(rep(c("inflow","outflow"),4))
      melted$group = paste(melted$health, melted$Richness)
      melted$group = factor(melted$group)
      ggplot(melted, aes(x=group, y=value, fill=in.outflow)) + geom_bar(position="stack", stat="identity")
      
      par(mar=c(10,4,4,1))
      barplot(inflow.outflow.sig.fraction.list, ylab="fractions per inflow/outflow species", las=2)
      
      par(mar=c(7,4,4,1))
      barplot(inflow.outflow.sig.fraction.list[c(1,3)], ylab="fractions per inflow/outflow species", las=2)
      
      par(mar=c(7,4,4,1))
      barplot(inflow.outflow.sig.fraction.list[c(2,4)], ylab="fractions per inflow/outflow species", las=2)
      
      inflow.outflow.presence.list = list(inflow.hgc.health = colSums(mergeMatUpdatedDigit[inflowSpecies, industrialHgcSamples])/length(inflowSpecies),
                                          outflow.hgc.health = colSums(mergeMatUpdatedDigit[outflowSpecies, industrialHgcSamples])/length(outflowSpecies),
                                          inflow.lgc.health = colSums(mergeMatUpdatedDigit[inflowSpecies, industrialLgcSamples])/length(inflowSpecies),
                                          outflow.lgc.health = colSums(mergeMatUpdatedDigit[outflowSpecies, industrialLgcSamples])/length(outflowSpecies),
                                          inflow.hgc.disease = colSums(mergeMatUpdatedDigit[inflowSpecies, diseaseHgcSamples])/length(inflowSpecies),
                                          outflow.hgc.disease = colSums(mergeMatUpdatedDigit[outflowSpecies, diseaseHgcSamples])/length(outflowSpecies),
                                          inflow.lgc.disease = colSums(mergeMatUpdatedDigit[inflowSpecies, diseaseLgcSamples])/length(inflowSpecies),
                                          outflow.lgc.disease = colSums(mergeMatUpdatedDigit[outflowSpecies, diseaseLgcSamples])/length(outflowSpecies))
      
      par(mar=c(7,4,4,1))
      boxplot(inflow.outflow.presence.list, ylab="fractions per inflow/outflow species", las=2)
      
      mergeMatUpdatedRelAbd =  getRelAbd(mergeMatUpdated)
      
      inflow.outflow.abd.list = list(inflow.hgc.health = colSums(mergeMatUpdatedRelAbd[inflowSpecies, industrialHgcSamples]),
                                     outflow.hgc.health = colSums(mergeMatUpdatedRelAbd[outflowSpecies, industrialHgcSamples]),
                                     inflow.lgc.health = colSums(mergeMatUpdatedRelAbd[inflowSpecies, industrialLgcSamples]),
                                     outflow.lgc.health = colSums(mergeMatUpdatedRelAbd[outflowSpecies, industrialLgcSamples]),
                                     inflow.hgc.disease = colSums(mergeMatUpdatedRelAbd[inflowSpecies, diseaseHgcSamples]),
                                     outflow.hgc.disease = colSums(mergeMatUpdatedRelAbd[outflowSpecies, diseaseHgcSamples]),
                                     inflow.lgc.disease = colSums(mergeMatUpdatedRelAbd[inflowSpecies, diseaseLgcSamples]),
                                     outflow.lgc.disease = colSums(mergeMatUpdatedRelAbd[outflowSpecies, diseaseLgcSamples]))
      
      #inflow.outflow.abd.list = lapply(inflow.outflow.abd.list, function(x) log10(x + 1e-10))
      
      
      par(mar=c(10,4,4,1))
      boxplot(inflow.outflow.abd.list, ylab="log10 (cumulative sum abundance)", las=2)
      
      boxplot(inflow.outflow.abd.list[c(1,3,5,7)], ylab="log10 (cumulative sum abundance)", las=2)
      boxplot(inflow.outflow.abd.list[c(2,4,6,8)], ylab="log10 (cumulative sum abundance)", las=2)
      
      wilcox.test(inflow.outflow.abd.list$outflow.hgc.health, inflow.outflow.abd.list$outflow.lgc.health, alternative = "greater")
      wilcox.test(inflow.outflow.abd.list$outflow.hgc.health, inflow.outflow.abd.list$outflow.hgc.disease, alternative = "less")
      wilcox.test(inflow.outflow.abd.list$outflow.hgc.health, inflow.outflow.abd.list$outflow.lgc.disease, alternative = "less")
      wilcox.test(inflow.outflow.abd.list$outflow.hgc.disease, inflow.outflow.abd.list$outflow.lgc.disease, alternative = "less")
      
      wilcox.test(inflow.outflow.abd.list$inflow.hgc, inflow.outflow.abd.list$outflow.hgc)
      wilcox.test(inflow.outflow.abd.list$inflow.hgc, inflow.outflow.abd.list$inflow.lgc)
      wilcox.test(inflow.outflow.abd.list$inflow.lgc, inflow.outflow.abd.list$outflow.hgc)
      
      
      boxplot(inflow.outflow.abd.list[c(1,3)])
      boxplot(inflow.outflow.abd.list[c(2,4)])
      
      
    }
    
    length(normalHgcSamples)
    length(normalLgcSamples)
    
    diseaseHgcSamples = newDiseaseIds[newDiseaseIds %in% hgcSamples]
    diseaseLgcSamples = newDiseaseIds[newDiseaseIds %in% lgcSamples]
    
    length(diseaseHgcSamples)
    length(diseaseLgcSamples)
    
    
    industrialHgcSamples = newWesternIds[newWesternIds %in% hgcSamples]
    industrialLgcSamples = newWesternIds[newWesternIds %in% lgcSamples]
    
    length(industrialHgcSamples)
    length(industrialLgcSamples)
    
    traditionalHgcSamples = newNonWesternIds[newNonWesternIds %in% hgcSamples]
    traditionalLgcSamples = newNonWesternIds[newNonWesternIds %in% lgcSamples]
    
    length(traditionalHgcSamples)
    length(traditionalLgcSamples)
    
    
    normHgcSampleTab = data.frame(sample=normalHgcSamples, type="HGC.normal")
    normLgcSampleTab = data.frame(sample=normalLgcSamples, type="LGC.normal")
    diseaseHgcSampleTab = data.frame(sample=diseaseHgcSamples, type="HGC.disease")
    diseaseLgcSampleTab = data.frame(sample=diseaseLgcSamples, type="LGC.disease")
    industrialHgcSampleTab = data.frame(sample=industrialHgcSamples, type="HGC.industrial")
    industrialLgcSampleTab = data.frame(sample=industrialLgcSamples, type="LGC.industrial")
    traditionalHgcSampleTab = data.frame(sample=traditionalHgcSamples, type="HGC.traditional")
    traditionalLgcSampleTab = data.frame(sample=traditionalLgcSamples, type="LGC.traditional")
    allHgcSampleTab = data.frame(sample=traditionalHgcSamples, type="HGC.all")
    allLgcSampleTab = data.frame(sample=traditionalLgcSamples, type="LGC.all")
    
    
    normSampleTab = rbind(normHgcSampleTab, 
                          normLgcSampleTab,
                          diseaseHgcSampleTab,
                          diseaseLgcSampleTab,
                          industrialHgcSampleTab,
                          industrialLgcSampleTab,
                          traditionalHgcSampleTab,
                          traditionalLgcSampleTab,
                          allHgcSampleTab,
                          allLgcSampleTab)
    writeMode = F
    if (writeMode) write.table(normSampleTab, file="C://Data//comparative.analysis.healthy.sweden//reza.HGC.LGC.samples.all.types.txt", sep="\t")
    
    #newNormalAllIds
    #newDiseaseIds
    #newWesternIds = newNormalAllIds[!newNormalAllIds %in% newNonWesternIds]
    
    
    #and then volcano plots
    
    
    genusMode = F
    if (genusMode) {
      
      ### all modes ####
      
      
      volcano.stats.genus.all = getVolcanoStat(genusMatUpdated, hgcSamples, lgcSamples, taxo)
      volcano.stats.genus.normal = getVolcanoStat(genusMatUpdated, normalHgcSamples, normalLgcSamples, taxo)
      volcano.stats.genus.disease = getVolcanoStat(genusMatUpdated, diseaseHgcSamples, diseaseLgcSamples, taxo)
      volcano.stats.genus.industrial = getVolcanoStat(genusMatUpdated, industrialHgcSamples, industrialLgcSamples, taxo)
      volcano.stats.genus.traditional = getVolcanoStat(genusMatUpdated, traditionalHgcSamples, traditionalLgcSamples, taxo)
      
      write.table(volcano.stats.genus.all, statsGenusForAllSamples, sep="\t")
      write.table(volcano.stats.genus.normal, statsGenusForNormSamples, sep="\t")
      write.table(volcano.stats.genus.disease, statsGenusForDiseaseSamples, sep="\t")
      write.table(volcano.stats.genus.industrial, statsGenusForIndustrialSamples, sep="\t")
      write.table(volcano.stats.genus.traditional, statsGenusForTraditionalSamples, sep="\t")
      
      plotVolcano(volcano.stats.genus.all, F)
      
      pairs(cbind(all=volcano.stats.genus.all$lfc,
                  normal=volcano.stats.genus.normal$lfc,
                  disease=volcano.stats.genus.disease$lfc,
                  industrial=volcano.stats.genus.industrial$lfc))
      
      
    }
    
    mgsMode = T
    if (mgsMode) {
      
      generateMode = F
      if (generateMode) {
        volcano.stats.mgs.all = getVolcanoStat(mergeMatUpdated, hgcSamples, lgcSamples, taxo, T)
        volcano.stats.mgs.normal = getVolcanoStat(mergeMatUpdated, normalHgcSamples, normalLgcSamples, taxo, T)
        volcano.stats.mgs.disease = getVolcanoStat(mergeMatUpdated, diseaseHgcSamples, diseaseLgcSamples, taxo, T)
        volcano.stats.mgs.industrial = getVolcanoStat(mergeMatUpdated, industrialHgcSamples, industrialLgcSamples, taxo, T)
        volcano.stats.mgs.traditional = getVolcanoStat(mergeMatUpdated, traditionalHgcSamples, traditionalLgcSamples, taxo, T)
        
        write.table(volcano.stats.mgs.all, statsMgsForAllSamples, sep="\t")
        write.table(volcano.stats.mgs.normal, statsMgsForNormSamples, sep="\t")
        write.table(volcano.stats.mgs.disease, statsMgsForDiseaseSamples, sep="\t")
        write.table(volcano.stats.mgs.industrial, statsMgsForIndustrialSamples, sep="\t")
        write.table(volcano.stats.mgs.traditional, statsMgsForTraditionalSamples, sep="\t")
      }
      
      
      loadMode = T
      if (loadMode) {
        volcano.stats.mgs.all = read.table(statsMgsForAllSamples, sep="\t")
        volcano.stats.mgs.normal = read.table(statsMgsForNormSamples, sep="\t")
        volcano.stats.mgs.disease = read.table(statsMgsForDiseaseSamples, sep="\t")
        
      }
      
      
      
      plotVolcano(volcano.stats.mgs.all, F)
      
      pairs(cbind(all=volcano.stats.mgs.all$lfc,
                  normal=volcano.stats.mgs.normal$lfc,
                  disease=volcano.stats.mgs.disease$lfc,
                  industrial=volcano.stats.mgs.industrial$lfc))
      
      ### figure for scatter plot ### 
      
      normDiseaseScatterTab = data.frame(normal=volcano.stats.mgs.normal$lfc, 
                                         disease=volcano.stats.mgs.disease$lfc,
                                         stringsAsFactors = F)
      rownames(normDiseaseScatterTab)=rownames(volcano.stats.mgs.normal)
      normDiseaseScatterTab = normDiseaseScatterTab[bothSpecies,]
      
      normDiseaseScatterTab[normDiseaseScatterTab >= 15]=NA
      normDiseaseScatterTab[normDiseaseScatterTab <= -15]=NA
      
      ggplot(normDiseaseScatterTab, aes(x=disease, y=normal, fill=normal)) + 
        geom_point(size=2, shape=21, colour="black")  +
        xlab("") + ylab("")+
        geom_hline(yintercept = 0) + geom_vline(xintercept = 0)+
        geom_smooth(method='lm',formula=y~x, se=FALSE, color="red")+
        scale_fill_gradient2(midpoint = 0.2, low="#FFD70044", mid = "white", high = "purple") +
        theme(panel.grid.major = element_blank(), 
              panel.grid.minor = element_blank(),
              legend.position = "none",
              panel.background = element_rect(fill = "white", 
                                              colour = "black", 
                                              size = 0.5, 
                                              linetype = "solid"))
      
      normWellnessScatterTab = data.frame(normal=volcano.stats.mgs.normal$lfc, 
                                          wellness=- wellnessMgsWilcoxStatsDecreasedPair$l2fc,
                                          stringsAsFactors = F)
      rownames(normWellnessScatterTab)=rownames(volcano.stats.mgs.normal)
      normWellnessScatterTab = normWellnessScatterTab[bothSpecies,]
      
      normWellnessScatterTab[normWellnessScatterTab >= 10]=NA
      normWellnessScatterTab[normWellnessScatterTab <= -10]=NA
      normWellnessScatterTab[normWellnessScatterTab == 0]=NA
      
      
      ggplot(normWellnessScatterTab, aes(x=wellness, y=normal, fill=normal)) + 
        geom_point(size=2, shape=21, colour="black")  +
        xlab("") + ylab("")+ xlim(c(-3.5,6))+
        geom_smooth(method='lm',formula=y~x, se=FALSE, color="red")+
        geom_hline(yintercept = 0) + geom_vline(xintercept = 0)+
        scale_fill_gradient2(midpoint = 0.2, low="#FFD70044", mid = "white", high = "purple") +
        theme(panel.grid.major = element_blank(), 
              panel.grid.minor = element_blank(),
              legend.position = "none",
              panel.background = element_rect(fill = "white", 
                                              colour = "black", 
                                              size = 0.5, 
                                              linetype = "solid"))
      
      normAntibioticsScatterTab = data.frame(normal=volcano.stats.mgs.normal$lfc, 
                                             antibiotics=- wilcoxStatsDecrease4$l2fc,
                                             stringsAsFactors = F)
      
      
      rownames(normAntibioticsScatterTab)=rownames(volcano.stats.mgs.normal)
      normAntibioticsScatterTab = normAntibioticsScatterTab[bothSpecies,]
      
      normAntibioticsScatterTab[normAntibioticsScatterTab >= 10]=NA
      normAntibioticsScatterTab[normAntibioticsScatterTab <= -10]=NA
      normAntibioticsScatterTab[normAntibioticsScatterTab == 0]=NA
      
      
      ggplot(normAntibioticsScatterTab, aes(x=antibiotics, y=normal, fill=normal)) + 
        geom_point(size=2, shape=21, colour="black")  +
        xlab("") + ylab("")+ 
        geom_smooth(method='lm',formula=y~x, se=FALSE, color="red")+
        geom_hline(yintercept = 0) + geom_vline(xintercept = 0)+
        scale_fill_gradient2(midpoint = 0.2, low="#FFD70044", mid = "white", high = "purple") +
        theme(panel.grid.major = element_blank(), 
              panel.grid.minor = element_blank(),
              legend.position = "none",
              panel.background = element_rect(fill = "white", 
                                              colour = "black", 
                                              size = 0.5, 
                                              linetype = "solid"))
      
      normAntibioticsShortScatterTab = data.frame(normal=volcano.stats.mgs.normal$lfc, 
                                                  antibiotics=- wilcoxStatsDecrease1$l2fc,
                                                  stringsAsFactors = F)
      
      rownames(normAntibioticsShortScatterTab)=rownames(volcano.stats.mgs.normal)
      normAntibioticsShortScatterTab = normAntibioticsShortScatterTab[bothSpecies,]
      
      normAntibioticsShortScatterTab[normAntibioticsShortScatterTab >= 10]=NA
      normAntibioticsShortScatterTab[normAntibioticsShortScatterTab <= -10]=NA
      normAntibioticsShortScatterTab[normAntibioticsShortScatterTab == 0]=NA
      
      
      ggplot(normAntibioticsShortScatterTab, aes(x=antibiotics, y=normal, fill=normal)) + 
        geom_point(size=2, shape=21, colour="black")  +
        xlab("") + ylab("")+ 
        geom_smooth(method='lm',formula=y~x, se=FALSE, color="red")+
        geom_hline(yintercept = 0) + geom_vline(xintercept = 0)+
        scale_fill_gradient2(midpoint = 0.2, low="#FFD70044", mid = "white", high = "purple") +
        theme(panel.grid.major = element_blank(), 
              panel.grid.minor = element_blank(),
              legend.position = "none",
              panel.background = element_rect(fill = "white", 
                                              colour = "black", 
                                              size = 0.5, 
                                              linetype = "solid"))
      
      cor(normDiseaseScatterTab$normal, 
          normDiseaseScatterTab$disease, 
          method = "spearman", use="pairwise.complete.obs")
      
      cor(normWellnessScatterTab$normal, 
          normWellnessScatterTab$wellness, 
          method = "spearman", use="pairwise.complete.obs")
      
      cor(normAntibioticsScatterTab$normal, 
          normAntibioticsScatterTab$antibiotics, 
          method = "spearman", use="pairwise.complete.obs")
      
      cor(normAntibioticsShortScatterTab$normal, 
          normAntibioticsShortScatterTab$antibiotics, 
          method = "spearman", use="pairwise.complete.obs")
      
    }
    
    
  }
  
  woWellness = T
  if (woWellness) {
    
    lCut=0.25
    hCut=0.75
    
    basicMetaMapUpdatedWoWellness = basicMetaMapUpdated[!basicMetaMapUpdated$sample.ID %in% wellnessSamples, ]
    
    richness = basicMetaMapUpdatedWoWellness$GeneRichness
    names(richness) = basicMetaMapUpdatedWoWellness$sample.ID
    
    normalRichness = richness[names(richness) %in% newNormalAllIds]
    diseaseRichness = richness[names(richness) %in% newDiseaseIds]
    industrialRichness = richness[names(richness) %in% newWesternIds]
    
    hHalfCutNumber=as.numeric(quantile(basicMetaMapUpdatedWoWellness$GeneRichness,0.5))
    
    hAllCutNumber=as.numeric(quantile(basicMetaMapUpdatedWoWellness$GeneRichness,hCut))
    lAllCutNumber=as.numeric(quantile(basicMetaMapUpdatedWoWellness$GeneRichness,lCut))
    
    print(hAllCutNumber) #708843.2
    print(lAllCutNumber) #431824.5
    
    hgcSamples = names(richness[richness>=hAllCutNumber])
    lgcSamples = names(richness[richness<=lAllCutNumber])
    
    normalHgcSamples = newNormalAllIds[newNormalAllIds %in% hgcSamples]
    normalLgcSamples = newNormalAllIds[newNormalAllIds %in% lgcSamples]
    
    
    table(normalHgcSamples %in% wellnessSamples)
    table(normalLgcSamples %in% wellnessSamples)
    
    
    checkInflowOutflowMode = T
    if (checkInflowOutflowMode) {
      volcano.stats.mgs.all = getVolcanoStat(mergeMatUpdated, hgcSamples, lgcSamples, taxo, T)
      volcano.stats.mgs.normal = getVolcanoStat(mergeMatUpdated, normalHgcSamples, normalLgcSamples, taxo, T)
      adjpCut=1e-3
      volcano.stats.mgs.normal.sel = volcano.stats.mgs.normal[volcano.stats.mgs.normal$qvalue <= adjpCut,]
      volcano.stats.mgs.normal.sel.hgc = volcano.stats.mgs.normal.sel[volcano.stats.mgs.normal.sel$lfc > 2,]
      volcano.stats.mgs.normal.sel.lgc = volcano.stats.mgs.normal.sel[volcano.stats.mgs.normal.sel$lfc < -2,]
      
      hgcSpecies = rownames(volcano.stats.mgs.normal.sel.hgc)
      lgcSpecies = rownames(volcano.stats.mgs.normal.sel.lgc)
      
      
      hgcLgcInflowList=list(HGC=inflow[hgcSpecies], LGC=inflow[lgcSpecies])
      hgcLgcOutflowList=list(HGC=outflow[hgcSpecies], LGC=outflow[lgcSpecies])
      hgcLgcInflowMelt = melt(hgcLgcInflowList)
      hgcLgcOutflowMelt = melt(hgcLgcOutflowList)
      
      t.test(hgcLgcInflowList$HGC, hgcLgcInflowList$LGC)
      t.test(hgcLgcOutflowList$HGC, hgcLgcOutflowList$LGC)
      
      CPCOLS <- c("#493D8C91", "#FF735786")
      
      
      ggplot(hgcLgcInflowMelt) + geom_boxplot(aes(x=L1, y=value, fill=L1)) +
        scale_fill_manual(values = c("#ff000055","#2c528c55")) +
        xlab("") +
        ylab("Inflow") + 
        theme(panel.grid.major = element_blank(), 
              panel.grid.minor = element_blank(),
              legend.position = "none",
              axis.text.x = element_text(angle = 60, hjust=1),
              panel.background = element_rect(fill = "white", 
                                              colour = "black", 
                                              size = 0.5, 
                                              linetype = "solid"))
      
      ggplot(hgcLgcOutflowMelt) + geom_boxplot(aes(x=L1, y=value, fill=L1)) +
        scale_fill_manual(values = c("#ff000055","#2c528c55")) +
        xlab("") +
        ylab("Outflow") + 
        theme(panel.grid.major = element_blank(), 
              panel.grid.minor = element_blank(),
              legend.position = "none",
              axis.text.x = element_text(angle = 60, hjust=1),
              panel.background = element_rect(fill = "white", 
                                              colour = "black", 
                                              size = 0.5, 
                                              linetype = "solid"))
    }
    
    
    
  }
  
  beforeUpdate = F
  if (beforeUpdate) {
    normalMat = mergeMat[,colnames(mergeMat) %in% normalIDs]
    diseaseMat = mergeMat[,colnames(mergeMat) %in% caseIDs2]
    diseaseFamilyMat = getTaxaSumMat(diseaseMat,taxo,"family",T)
    
    indiMat = mergeMat[,colnames(mergeMat) %in% indiIDs]
    indiFamilyMat = getTaxaSumMat(indiMat,taxo,"family",T)
    
    targetMat = mergeMat[,colnames(mergeMat) %in% normalIDs]
    phylaMat = getTaxaSumMat(targetMat,taxo,"phylum",T)
    familyMat = getTaxaSumMat(targetMat,taxo,"family",T)
    
    #healthy controls by geometry
    notIndiRichness = basicMetaMap$GeneRichness[match(normalIDs, basicMetaMap$sample.ID)]
    names(notIndiRichness) = basicMetaMap$sample.ID[match(normalIDs, basicMetaMap$sample.ID)]
    
    notIndiAge = basicMetaMap$Age[match(normalIDs, basicMetaMap$sample.ID)]
    notIndiBMI = basicMetaMap$BMI[match(normalIDs, basicMetaMap$sample.ID)]
    notIndiGender = basicMetaMap$Gender[match(normalIDs, basicMetaMap$sample.ID)]
    notIndiGeos = basicMetaMap$Geography[match(normalIDs, basicMetaMap$sample.ID)]
    
    notIndiRichnessByGeos = split(notIndiRichness,notIndiGeos)
    notIndiAgeByGeos = split(notIndiAge,notIndiGeos)
    notIndiBmiByGeos = split(notIndiBMI,notIndiGeos)
    notIndiGenderByGeos = split(notIndiGender,notIndiGeos)
    
    notIndiRichGeoMedian = do.call(c,lapply(notIndiRichnessByGeos,median))
    notIndiRichOrdered = names(sort(notIndiRichGeoMedian))
    
    diseaseRichness = basicMetaMap$GeneRichness[match(caseIDs2, basicMetaMap$sample.ID)]
    names(diseaseRichness) = basicMetaMap$sample.ID[match(caseIDs2, basicMetaMap$sample.ID)]
    
    diseaseAge = basicMetaMap$Age[match(caseIDs2, basicMetaMap$sample.ID)]
    diseaseBMI = basicMetaMap$BMI[match(caseIDs2, basicMetaMap$sample.ID)]
    diseaseGender = basicMetaMap$Gender[match(caseIDs2, basicMetaMap$sample.ID)]
    diseaseGeos = basicMetaMap$Geography[match(caseIDs2, basicMetaMap$sample.ID)]
    diseaseDsIds = basicMetaMap$dataset.ID[match(caseIDs2, basicMetaMap$sample.ID)]
    diseaseDsNames = caseName2Tab$Case[match(diseaseDsIds, caseName2Tab$ID)]
    
    diseaseRichnessByDs = split(diseaseRichness,diseaseDsNames)
    diseaseAgeByDs = split(diseaseAge,diseaseDsNames)
    diseaseBmiByDs = split(diseaseBMI,diseaseDsNames)
    diseaseGenderByDs = split(diseaseGender,diseaseDsNames)
    
    diseaseRichDsMedian = do.call(c,lapply(diseaseRichnessByDs,median))
    diseaseRichOrdered = names(sort(diseaseRichDsMedian))
    
    indiRichness = basicMetaMap$GeneRichness[match(indiIDs, basicMetaMap$sample.ID)]
    names(indiRichness) = basicMetaMap$sample.ID[match(indiIDs, basicMetaMap$sample.ID)]
    
    indiAge = basicMetaMap$Age[match(indiIDs, basicMetaMap$sample.ID)]
    indiBMI = basicMetaMap$BMI[match(indiIDs, basicMetaMap$sample.ID)]
    indiGender = basicMetaMap$Gender[match(indiIDs, basicMetaMap$sample.ID)]
    indiGeos = basicMetaMap$Geography[match(indiIDs, basicMetaMap$sample.ID)]
    
    indiRichnessByGeos = split(indiRichness,indiGeos)
    indiAgeByGeos = split(indiAge,indiGeos)
    indiBmiByGeos = split(indiBMI,indiGeos)
    indiGenderByGeos = split(indiGender,indiGeos)
    
    indiRichGeoMedian = do.call(c,lapply(indiRichnessByGeos,median))
    indiRichOrdered = names(sort(indiRichGeoMedian))
    
    lCut=0.25
    hCut=0.75
    
    recapMode = T
    if (recapMode) {
      load("healthy.HGC.LGC.RData")
      load("diseased.HGC.LGC.RData")
      load("indi.HGC.LGC.RData")
      
      # statSpeciesTab
      # statDiseaseSpeciesTab
      # statIndiSpeciesTab
      # 
      statSpeciesTab$lfc[statSpeciesTab$lfc==15]=16
      statSpeciesTab$lfc[statSpeciesTab$lfc==-15]=-16
      
      statIndiSpeciesTab$lfc[statIndiSpeciesTab$lfc==15]=16
      statIndiSpeciesTab$lfc[statIndiSpeciesTab$lfc==-15]=-16
      
      lfcMat = cbind(healthy=statSpeciesTab$lfc,
                     diseased=statDiseaseSpeciesTab$lfc,
                     indigenous=statIndiSpeciesTab$lfc)
      pairs(lfcMat)
      
      
      cor(statSpeciesTab$lfc, statDiseaseSpeciesTab$lfc)
      cor(statSpeciesTab$lfc, statIndiSpeciesTab$lfc)
      
      plot(statSpeciesTab$lfc, statDiseaseSpeciesTab$lfc, 
           pch=16, col="#80808088",las=1, tck = 0.02,yaxt="n",xaxt="n",
           xlab="LFC, healthy", ylab="LFC, diseased", xlim=c(-13,13),ylim=c(-13,13))
      abline(a=0,b=1,col="red")
      grid(lty=3,col="#80808077")
      
      plot(statSpeciesTab$lfc, statIndiSpeciesTab$lfc, 
           pch=16, col="#80808088",las=1,tck = 0.02,yaxt="n",xaxt="n",
           xlab="LFC, healthy", ylab="LFC, indigenous", xlim=c(-13,13),ylim=c(-13,13))
      abline(a=0,b=1,col="red")
      grid(lty=3,col="#80808077")
      
    }
    
    healthyMode = T
    if (healthyMode) {
      
      
      healthyHcutNumber=as.numeric(quantile(notIndiRichness,hCut))
      healthyLcutNumber=as.numeric(quantile(notIndiRichness,lCut))
      
      speciesMode = T
      if (speciesMode) {
        specMean = rowMeans(normalMat)
        sPart = specMean/sum(specMean)
        
        hgcSamples = notIndiRichness[notIndiRichness > as.numeric(quantile(notIndiRichness,hCut))]
        lgcSamples = notIndiRichness[notIndiRichness < as.numeric(quantile(notIndiRichness,lCut))]
        
        hgcMat = normalMat[,names(hgcSamples)]
        lgcMat = normalMat[,names(lgcSamples)]
        
        lenSpecies = dim(hgcMat)[1]
        pValues = sapply(1:lenSpecies, function(ind) {
          mergedVec = c(hgcMat[ind,],lgcMat[ind,])
          mergedFac = c(rep("HGC",length(hgcMat[ind,])),
                        rep("LGC",length(lgcMat[ind,])))
          
          wilcoxOut = wilcox.test(mergedVec ~ mergedFac)
          return(wilcoxOut$p.value)
        })
        pValues[is.nan(pValues)]=1
        adjPs = p.adjust(pValues,method="BH")
        logAdjPs = -log10(adjPs)
        
        lfcs = sapply(1:lenSpecies,function(ind){
          hmeans = mean(hgcMat[ind,])
          lmeans = mean(lgcMat[ind,]) 
          return(hmeans/lmeans)
        })
        names(lfcs)=rownames(normalMat)
        lfcs[is.nan(lfcs)]=1
        lfcs[lfcs==0]=2^-15
        lfcs[is.infinite(lfcs)]=2^15
        lfcs = log2(lfcs)
        summary(lfcs)
        
        plot(lfcs, -log10(adjPs))
        abline(h=-log10(1e-10))
        abline(v=3)
        abline(v=-3)
        
        statSpeciesTab = data.frame(lfc=lfcs, sig=logAdjPs, relAbd=sPart, label=getSpeciesName(names(lfcs),taxo), stringsAsFactors = F)
        
        lfcInd = abs(statSpeciesTab$lfc)>2
        sigInd = statSpeciesTab$sig>-log10(1e-3)
        unclassifiedInd = !grepl("unclassified", names(lfcs))
        statSpeciesTabModLabel  = statSpeciesTab
        statSpeciesTabModLabel$label[!sigInd | !lfcInd | !unclassifiedInd] = ""
        #statTabModLabel$label[!sigInd | !lfcInd] = ""
        
        plotVolcano(statSpeciesTabModLabel, F)
        
        
      }
      
      familyMode = F
      if (familyMode) {
        
        familyMean = rowMeans(familyMat)
        fPart = familyMean/sum(familyMean)
        
        as.numeric(quantile(notIndiRichness,hCut))
        as.numeric(quantile(notIndiRichness,lCut))
        
        hgcSamples = notIndiRichness[notIndiRichness > as.numeric(quantile(notIndiRichness,hCut))]
        lgcSamples = notIndiRichness[notIndiRichness < as.numeric(quantile(notIndiRichness,lCut))]
        
        hgcFamilyMat = familyMat[,names(hgcSamples)]
        lgcFamilyMat = familyMat[,names(lgcSamples)]
        
        lenFamilies = dim(hgcFamilyMat)[1]
        pValues = sapply(1:lenFamilies, function(ind) {
          mergedVec = c(hgcFamilyMat[ind,],lgcFamilyMat[ind,])
          mergedFac = c(rep("HGC",length(hgcFamilyMat[ind,])),
                        rep("LGC",length(lgcFamilyMat[ind,])))
          
          wilcoxOut = wilcox.test(mergedVec ~ mergedFac)
          return(wilcoxOut$p.value)
        })
        pValues[is.nan(pValues)]=1
        adjPs = p.adjust(pValues,method="BH")
        logAdjPs = -log10(adjPs)
        
        lfcs = sapply(1:lenFamilies,function(ind){
          hmeans = mean(hgcFamilyMat[ind,])
          lmeans = mean(lgcFamilyMat[ind,]) 
          return(hmeans/lmeans)
        })
        names(lfcs)=rownames(familyMat)
        lfcs[is.nan(lfcs)]=1
        lfcs[lfcs==0]=2^-15
        lfcs[is.infinite(lfcs)]=2^15
        lfcs = log2(lfcs)
        summary(lfcs)
        
        plot(lfcs, -log10(adjPs))
        abline(h=-log10(1e-10))
        abline(v=3)
        abline(v=-3)
        
        statTab = data.frame(lfc=lfcs, sig=logAdjPs, relAbd=fPart, label=names(lfcs), stringsAsFactors = F)
        
        lfcInd = abs(statTab$lfc)>2
        sigInd = statTab$sig>-log10(1e-3)
        unclassifiedInd = !grepl("unclassified", names(lfcs))
        statTabModLabel  = statTab
        statTabModLabel$label[!sigInd | !lfcInd | !unclassifiedInd] = ""
        #statTabModLabel$label[!sigInd | !lfcInd] = ""
        
        
        
        
        plotVolcano(statTabModLabel, F)
        
      }
      
      save(statTab, statSpeciesTab, file="healthy.HGC.LGC.RData")
      
      
    }
    
    diseaseMode = T
    if (diseaseMode) {
      diseaseHcutNumber=as.numeric(quantile(diseaseRichness,hCut))
      diseaseLcutNumber=as.numeric(quantile(diseaseRichness,lCut))
      
      targetHCut = healthyHcutNumber
      targetLCut = healthyLcutNumber
      
      hgcDiseaseSamples = diseaseRichness[diseaseRichness > targetHCut]
      lgcDiseaseSamples = diseaseRichness[diseaseRichness < targetLCut]
      
      
      speciesMode = T
      if (speciesMode){
        specDiseaseMean = rowMeans(diseaseMat)
        sDiseasePart = specDiseaseMean/sum(specDiseaseMean)
        
        ################
        hgcDiseaseMat = diseaseMat[,names(hgcDiseaseSamples)]
        lgcDiseaseMat = diseaseMat[,names(lgcDiseaseSamples)]
        
        
        lenSpecies = dim(hgcDiseaseMat)[1]
        pValues = sapply(1:lenSpecies, function(ind) {
          mergedVec = c(hgcDiseaseMat[ind,],lgcDiseaseMat[ind,])
          mergedFac = c(rep("HGC",length(hgcDiseaseMat[ind,])),
                        rep("LGC",length(lgcDiseaseMat[ind,])))
          
          wilcoxOut = wilcox.test(mergedVec ~ mergedFac)
          return(wilcoxOut$p.value)
        })
        pValues[is.nan(pValues)]=1
        adjPs = p.adjust(pValues,method="BH")
        logAdjPs = -log10(adjPs)
        
        lfcs = sapply(1:lenSpecies,function(ind){
          hmeans = mean(hgcDiseaseMat[ind,])
          lmeans = mean(lgcDiseaseMat[ind,]) 
          return(hmeans/lmeans)
        })
        names(lfcs)=rownames(diseaseMat)
        lfcs[is.nan(lfcs)]=1
        lfcs[lfcs==0]=2^-16
        lfcs[is.infinite(lfcs)]=2^16
        lfcs = log2(lfcs)
        summary(lfcs)
        
        plot(lfcs, -log10(adjPs))
        abline(h=-log10(1e-10))
        abline(v=3)
        abline(v=-3)
        
        statDiseaseSpeciesTab = data.frame(lfc=lfcs, sig=logAdjPs, relAbd=sPart, label=getSpeciesName(names(lfcs),taxo), stringsAsFactors = F)
        
        lfcInd = abs(statDiseaseSpeciesTab$lfc)>2
        sigInd = statDiseaseSpeciesTab$sig>-log10(1e-3)
        unclassifiedInd = !grepl("unclassified", names(lfcs))
        statDiseaseSpeciesTabModLabel  = statDiseaseSpeciesTab
        statDiseaseSpeciesTabModLabel$label[!sigInd | !lfcInd | !unclassifiedInd] = ""
        #statTabModLabel$label[!sigInd | !lfcInd] = ""
        
        
        
        
        plotVolcano(statDiseaseSpeciesTabModLabel, F)
        
        ###############
      }
      
      familyMode = T
      if (familyMode) {
        diseaseFamilyMean = rowMeans(diseaseFamilyMat)
        fDiseasePart = diseaseFamilyMean/sum(diseaseFamilyMean)
        
        hgcDiseaseFamilyMat = diseaseFamilyMat[,names(hgcDiseaseSamples)]
        lgcDiseaseFamilyMat = diseaseFamilyMat[,names(lgcDiseaseSamples)]
        
        lenFamilies = dim(diseaseFamilyMat)[1]
        
        pValues = sapply(1:lenFamilies, function(ind) {
          mergedVec = c(hgcDiseaseFamilyMat[ind,],lgcDiseaseFamilyMat[ind,])
          mergedFac = c(rep("HGC",length(hgcDiseaseFamilyMat[ind,])),
                        rep("LGC",length(lgcDiseaseFamilyMat[ind,])))
          
          wilcoxOut = wilcox.test(mergedVec ~ mergedFac)
          return(wilcoxOut$p.value)
        })
        pValues[is.nan(pValues)]=1
        adjPs = p.adjust(pValues,method="BH")
        logAdjPs = -log10(adjPs)
        
        lfcs = sapply(1:lenFamilies,function(ind){
          hmeans = mean(hgcDiseaseFamilyMat[ind,])
          lmeans = mean(lgcDiseaseFamilyMat[ind,]) 
          return(hmeans/lmeans)
        })
        names(lfcs)=rownames(diseaseFamilyMat)
        lfcs[is.nan(lfcs)]=1
        lfcs[lfcs==0]=2^-15
        lfcs[is.infinite(lfcs)]=2^15
        lfcs = log2(lfcs)
        summary(lfcs)
        
        plot(lfcs, -log10(adjPs))
        abline(h=-log10(1e-10))
        abline(v=3)
        abline(v=-3)
        
        statDiseaseTab = data.frame(lfc=lfcs, sig=logAdjPs, relAbd=fPart, label=names(lfcs), stringsAsFactors = F)
        
        lfcInd = abs(statDiseaseTab$lfc)>2
        sigInd = statDiseaseTab$sig>-log10(1e-3)
        unclassifiedInd = !grepl("unclassified", names(lfcs))
        statDiseaseTabModLabel  = statDiseaseTab
        statDiseaseTabModLabel$label[!sigInd | !lfcInd | !unclassifiedInd] = ""
        #statTabModLabel$label[!sigInd | !lfcInd] = ""
        
        
        plotVolcano(statDiseaseTabModLabel, F)
        
      }
      
      save(statDiseaseSpeciesTab, statDiseaseSpeciesTab, file="diseased.HGC.LGC.RData")
      
    }
    
    indiMode = T
    if (indiMode) {
      indiHcutNumber=as.numeric(quantile(indiRichness,hCut))
      indiLcutNumber=as.numeric(quantile(indiRichness,lCut))
      
      targetHCut = healthyHcutNumber
      targetLCut = healthyLcutNumber
      
      hgcIndiSamples = indiRichness[indiRichness > targetHCut]
      lgcIndiSamples = indiRichness[indiRichness < targetLCut]
      
      speciesMode = T
      if (speciesMode) {
        specIndiMean = rowMeans(indiMat)
        sIndiPart = specIndiMean/sum(specIndiMean)
        
        ################
        hgcIndiMat = indiMat[,names(hgcIndiSamples)]
        lgcIndiMat = indiMat[,names(lgcIndiSamples)]
        
        
        lenSpecies = dim(hgcIndiMat)[1]
        pValues = sapply(1:lenSpecies, function(ind) {
          mergedVec = c(hgcIndiMat[ind,],lgcIndiMat[ind,])
          mergedFac = c(rep("HGC",length(hgcIndiMat[ind,])),
                        rep("LGC",length(lgcIndiMat[ind,])))
          
          wilcoxOut = wilcox.test(mergedVec ~ mergedFac)
          return(wilcoxOut$p.value)
        })
        pValues[is.nan(pValues)]=1
        adjPs = p.adjust(pValues,method="BH")
        logAdjPs = -log10(adjPs)
        
        lfcs = sapply(1:lenSpecies,function(ind){
          hmeans = mean(hgcIndiMat[ind,])
          lmeans = mean(lgcIndiMat[ind,]) 
          return(hmeans/lmeans)
        })
        names(lfcs)=rownames(indiMat)
        lfcs[is.nan(lfcs)]=1
        lfcs[lfcs==0]=2^-16
        lfcs[is.infinite(lfcs)]=2^16
        lfcs = log2(lfcs)
        summary(lfcs)
        
        plot(lfcs, -log10(adjPs))
        abline(h=-log10(1e-10))
        abline(v=3)
        abline(v=-3)
        
        statIndiSpeciesTab = data.frame(lfc=lfcs, sig=logAdjPs, relAbd=sPart, label=getSpeciesName(names(lfcs),taxo), stringsAsFactors = F)
        
        lfcInd = abs(statIndiSpeciesTab$lfc)>2
        sigInd = statIndiSpeciesTab$sig>-log10(1e-3)
        unclassifiedInd = !grepl("unclassified", names(lfcs))
        statIndiSpeciesTabModLabel  = statIndiSpeciesTab
        statIndiSpeciesTabModLabel$label[!sigInd | !lfcInd | !unclassifiedInd] = ""
        #statTabModLabel$label[!sigInd | !lfcInd] = ""
        
        plotVolcano(statIndiSpeciesTabModLabel, F)
        
        
      }
      
      familyMode = T
      if (familyMode) {
        indiFamilyMean = rowMeans(indiFamilyMat)
        fIndiPart = indiFamilyMean/sum(indiFamilyMean)
        
        hgcIndiFamilyMat = indiFamilyMat[,names(hgcIndiSamples)]
        lgcIndiFamilyMat = indiFamilyMat[,names(lgcIndiSamples)]
        
        lenFamilies = dim(indiFamilyMat)[1]
        
        pValues = sapply(1:lenFamilies, function(ind) {
          mergedVec = c(hgcIndiFamilyMat[ind,],lgcIndiFamilyMat[ind,])
          mergedFac = c(rep("HGC",length(hgcIndiFamilyMat[ind,])),
                        rep("LGC",length(lgcIndiFamilyMat[ind,])))
          
          wilcoxOut = wilcox.test(mergedVec ~ mergedFac)
          return(wilcoxOut$p.value)
        })
        pValues[is.nan(pValues)]=1
        adjPs = p.adjust(pValues,method="BH")
        logAdjPs = -log10(adjPs)
        
        lfcs = sapply(1:lenFamilies,function(ind){
          hmeans = mean(hgcIndiFamilyMat[ind,])
          lmeans = mean(lgcIndiFamilyMat[ind,]) 
          return(hmeans/lmeans)
        })
        names(lfcs)=rownames(indiFamilyMat)
        lfcs[is.nan(lfcs)]=1
        lfcs[lfcs==0]=2^-15
        lfcs[is.infinite(lfcs)]=2^15
        lfcs = log2(lfcs)
        summary(lfcs)
        
        plot(lfcs, -log10(adjPs))
        abline(h=-log10(1e-10))
        abline(v=3)
        abline(v=-3)
        
        statIndiTab = data.frame(lfc=lfcs, sig=logAdjPs, relAbd=fPart, label=names(lfcs), stringsAsFactors = F)
        
        lfcInd = abs(statIndiTab$lfc)>2
        sigInd = statIndiTab$sig>-log10(1e-3)
        unclassifiedInd = !grepl("unclassified", names(lfcs))
        statIndiTabModLabel  = statIndiTab
        statIndiTabModLabel$label[!sigInd | !lfcInd | !unclassifiedInd] = ""
        
        plotVolcano(statIndiTabModLabel, F)
        
        
      }
      save(statIndiTab, statIndiSpeciesTab, file="indi.HGC.LGC.RData")
      
      
    }
    #480735.3
    #767550.3
    #save.image("today.2019.03.11.RData")
    
    hgcDSamples = diseaseRichness[diseaseRichness > as.numeric(quantile(diseaseRichness,hCut))]
    lgcDSamples = diseaseRichness[diseaseRichness < as.numeric(quantile(diseaseRichness,lCut))]
    
  }
  
}

richnessFuncCheckMode = T
if (richnessFuncCheckMode) {
  statsMgsForAllSamples = "J:\\Deposit\\Project\\2018_microbiome_atlas\\atlas.volcanot.out\\richness.volcano.stats.mgs.for.all.txt"
  statsMgsForNormSamples = "J:\\Deposit\\Project\\2018_microbiome_atlas\\atlas.volcanot.out\\richness.volcano.stats.mgs.for.normal.txt"
  statsMgsForDiseaseSamples = "J:\\Deposit\\Project\\2018_microbiome_atlas\\atlas.volcanot.out\\richness.volcano.stats.mgs.for.disease.txt"
  statsMgsForIndustrialSamples  = "J:\\Deposit\\Project\\2018_microbiome_atlas\\atlas.volcanot.out\\richness.volcano.stats.mgs.for.industrial.txt"
  statsMgsForTraditionalSamples  = "J:\\Deposit\\Project\\2018_microbiome_atlas\\atlas.volcanot.out\\richness.volcano.stats.mgs.for.traditional.txt"
  funcMatRData = "j://Deposit/Project/2018_microbiome_atlas/functional.annotation/funcMat.20190808.RData"
  
  
  newCutMode = T
  if (newCutMode) {
    adjpCut = 1e-3
    statsMgsForNorm = read.table(statsMgsForNormSamples, sep="\t", header=T, stringsAsFactors = F)
    statsMgsForNormSel = statsMgsForNorm[statsMgsForNorm$qvalue <= adjpCut,]
    # statsMgsForNormSelHgc = statsMgsForNormSel[statsMgsForNormSel$lfc > 0,]
    # statsMgsForNormSelLgc = statsMgsForNormSel[statsMgsForNormSel$lfc < -0,]
    statsMgsForNormSelHgc = statsMgsForNormSel[statsMgsForNormSel$lfc > 2,]
    statsMgsForNormSelLgc = statsMgsForNormSel[statsMgsForNormSel$lfc < -2,]
    
    hgcSpecies = rownames(statsMgsForNormSelHgc)
    lgcSpecies = rownames(statsMgsForNormSelLgc)
    
    statsMgsForNormSelHgcNew = statsMgsForNormSelHgc[order(statsMgsForNormSelHgc$qvalue),]
    statsMgsForNormSelLgcNew = statsMgsForNormSelLgc[order(statsMgsForNormSelLgc$qvalue),]
    hgcTopSpecies = rownames(statsMgsForNormSelHgcNew[1:25,])
    lgcTopSpecies = rownames(statsMgsForNormSelLgcNew[1:25,])
    
    
  }
  
  loadFuncAnnot = T
  if (loadFuncAnnot) {
    ### virulence factors ###
    vfMatRData = "j://Deposit/Project/2018_microbiome_atlas/functional.annotation/PATRIC/vfBestMat.RData"
    load(vfMatRData) #vfMatDigit
    vfTerms = colnames(vfMatDigit)
    patricVfDescTabFile = "J://Deposit/Project/2018_microbiome_atlas/functional.annotation/PATRIC/PATRIC_sp_gene.txt"
    patricVfDescTab = read.table(patricVfDescTabFile, header=T, sep="\t", stringsAsFactors = F)
    
    ### JGI phenotype ###
    jgiMatRData = "j://Deposit/Project/2018_microbiome_atlas/functional.annotation/jgiMat.RData"
    load(jgiMatRData)
    jgiTerms = colnames(jgiMat)
    
    ### antibiotic resistance ###
    mustardMatRData = "j://Deposit/Project/2018_microbiome_atlas/functional.annotation/mustard.mat.RData"
    load(mustardMatRData)
    arTerms = colnames(mustardMat)
    
    ### antismash - secondary metabolism ###
    antismashMatRData = "J://Deposit/Project/2018_microbiome_atlas/functional.annotation/msp1992.igc2.antismashMat.RData"
    load(antismashMatRData)
    antismashMelt = melt(antismashMat)
    antismashMelt = antismashMelt[antismashMelt$value!=0,]
    antismashTerms = colnames(antismashMat)
    
    ### Cazy  ### 
    newGutCazyMatRData = "J://Deposit/Project/2018_microbiome_atlas/functional.annotation/CAZY.INRA/igc2.new.cazy.mat.RData"
    load(newGutCazyMatRData)
    colnames(cazyMat) = paste("cazy", colnames(cazyMat), sep=".")
    cazyTerms = colnames(cazyMat)
    
  }
  
  loadFuncMode = T
  if (loadFuncMode) {
    load(funcMatRData)
    
    funcMatTranspose = t(funcMat)
    # hgcFuncMat = funcMatTranspose[,colnames(funcMatTranspose) %in% hgcSpecies]
    # lgcFuncMat = funcMatTranspose[,colnames(funcMatTranspose) %in% lgcSpecies]
    hgcFuncMat = funcMatTranspose[,colnames(funcMatTranspose) %in% hgcSpecies]
    lgcFuncMat = funcMatTranspose[,colnames(funcMatTranspose) %in% lgcSpecies]
    
    funcs = rownames(funcMatTranspose)
    pvalues = sapply(funcs, function(currFunc) {
      hgcCurrFunc = hgcFuncMat[currFunc,]
      lgcCurrFunc = lgcFuncMat[currFunc,]
      
      hgcAbsent = sum(hgcCurrFunc==0)
      hgcPresent = sum(hgcCurrFunc==1)
      
      lgcAbsent = sum(lgcCurrFunc==0)
      lgcPresent = sum(lgcCurrFunc==1)
      
      contigTab = cbind(HGC=c(present=hgcPresent, absent=hgcAbsent),
                        LGC=c(present=lgcPresent, absent=lgcAbsent))
      pvalue=chisq.test(contigTab)$p.value
      return(pvalue)
    }, simplify = F)
    pvalues = unlist(pvalues)
    
    oddsRatios = sapply(funcs, function(currFunc) {
      hgcCurrFunc = hgcFuncMat[currFunc,]
      lgcCurrFunc = lgcFuncMat[currFunc,]
      
      hgcAbsent = sum(hgcCurrFunc==0)
      hgcPresent = sum(hgcCurrFunc==1)
      
      lgcAbsent = sum(lgcCurrFunc==0)
      lgcPresent = sum(lgcCurrFunc==1)
      
      oddsRatio = (hgcPresent*lgcAbsent)/(hgcAbsent*lgcPresent)
      if (lgcPresent*hgcAbsent == 0) return(Inf) 
      return(oddsRatio)
    }, simplify = F)
    oddsRatios = unlist(oddsRatios)
    
    chisqStatTab = data.frame(term=names(pvalues),
                              pvalue=pvalues,
                              oddsRatio=oddsRatios,
                              logP = -log10(pvalues),
                              logOdds = log(oddsRatios))
    
    # View(chisqStatTab[grepl("cazy",rownames(chisqStatTab)),])
    # View(chisqStatTab[chisqStatTab$pvalue<1e-5,])
    
    View(chisqStatTab[mgePfamTerms,])
    View(chisqStatTab[arTerms,])
    
  }
  
  #inflow
  hgcLgcInflowList=list(HGC=inflow[hgcSpecies], LGC=inflow[lgcSpecies])
  hgcLgcOutflowList=list(HGC=outflow[hgcSpecies], LGC=outflow[lgcSpecies])
  hgcLgcInflowMelt = melt(hgcLgcInflowList)
  hgcLgcOutflowMelt = melt(hgcLgcOutflowList)
  
  t.test(hgcLgcInflowList$HGC, hgcLgcInflowList$LGC)
  t.test(hgcLgcOutflowList$HGC, hgcLgcOutflowList$LGC)
  
  CPCOLS <- c("#493D8C91", "#FF735786")
  
  
  p = ggplot(hgcLgcInflowMelt) + geom_boxplot(aes(x=L1, y=value, fill=L1)) +
    scale_fill_manual(values = c("#ff000055","#2c528c55")) +
    xlab("") +
    ylab("Inflow") + 
    theme(panel.grid.major = element_blank(), 
          panel.grid.minor = element_blank(),
          legend.position = "none",
          axis.text.x = element_text(angle = 60, hjust=1),
          panel.background = element_rect(fill = "white", 
                                          colour = "black", 
                                          size = 0.5, 
                                          linetype = "solid"))
  ggsave("hgc.lgc.inflow.boxplot.pdf", p, width=1.56, height=2.92, units="in")
  
  p = ggplot(hgcLgcOutflowMelt) + geom_boxplot(aes(x=L1, y=value, fill=L1)) +
    scale_fill_manual(values = c("#ff000055","#2c528c55")) +
    xlab("") +
    ylab("Outflow") + 
    theme(panel.grid.major = element_blank(), 
          panel.grid.minor = element_blank(),
          legend.position = "none",
          axis.text.x = element_text(angle = 60, hjust=1),
          panel.background = element_rect(fill = "white", 
                                          colour = "black", 
                                          size = 0.5, 
                                          linetype = "solid"))
  ggsave("hgc.lgc.outflow.boxplot.pdf", p, width=1.56, height=2.92, units="in")
  
  
  boxplot(hgcLgcInflowList)
  boxplot(hgcLgcOutflowList)
  
  t.test(hgcLgcInflowList$hgc, hgcLgcInflowList$lgc)
  t.test(hgcLgcOutflowList$hgc, hgcLgcOutflowList$lgc)
  
  plot(inflow, statsMgsForNorm$lfc[match(names(inflow), rownames(statsMgsForNorm))] )
  plot(outflow, statsMgsForNorm$lfc[match(names(outflow), rownames(statsMgsForNorm))]  )
  
  cor.test(inflow, statsMgsForNorm$lfc[match(names(inflow), rownames(statsMgsForNorm))] , method="spearman" )
  cor.test(outflow, statsMgsForNorm$lfc[match(names(outflow), rownames(statsMgsForNorm))]  , method="spearman")
  
  
}

richnessSpeciesPCoaMode = F
if (richnessSpeciesPCoaMode) {
  statsMgsForAllSamples = "J:\\Deposit\\Project\\2018_microbiome_atlas\\atlas.volcanot.out\\richness.volcano.stats.mgs.for.all.txt"
  statsMgsForNormSamples = "J:\\Deposit\\Project\\2018_microbiome_atlas\\atlas.volcanot.out\\richness.volcano.stats.mgs.for.normal.txt"
  statsMgsForDiseaseSamples = "J:\\Deposit\\Project\\2018_microbiome_atlas\\atlas.volcanot.out\\richness.volcano.stats.mgs.for.disease.txt"
  statsMgsForIndustrialSamples  = "J:\\Deposit\\Project\\2018_microbiome_atlas\\atlas.volcanot.out\\richness.volcano.stats.mgs.for.industrial.txt"
  statsMgsForTraditionalSamples  = "J:\\Deposit\\Project\\2018_microbiome_atlas\\atlas.volcanot.out\\richness.volcano.stats.mgs.for.traditional.txt"
  
  prevCutMode = F
  if (prevCutMode) {
    adjpCut = 1e-3
    statsMgsForNorm = read.table(statsMgsForNormSamples, sep="\t", header=T, stringsAsFactors = F)
    statsMgsForNormSel = statsMgsForNorm[statsMgsForNorm$qvalue <= adjpCut,]
    statsMgsForNormSelHgc = statsMgsForNormSel[statsMgsForNormSel$lfc > 0,]
    statsMgsForNormSelLgc = statsMgsForNormSel[statsMgsForNormSel$lfc < 0,]
    # statsMgsForNormSelHgc = statsMgsForNormSel[statsMgsForNormSel$lfc > 2,]
    # statsMgsForNormSelLgc = statsMgsForNormSel[statsMgsForNormSel$lfc < -2,]
    
    if (F) {
      hgcCut = quantile(statsMgsForNormSelHgc$relAbd_HGC, 0.75)
      lgcCut = quantile(statsMgsForNormSelLgc$relAbd_LGC, 0.75)
      
      hgcSpecies = rownames(statsMgsForNormSelHgc[statsMgsForNormSelHgc$relAbd_HGC >= hgcCut,])
      lgcSpecies = rownames(statsMgsForNormSelLgc[statsMgsForNormSelLgc$relAbd_LGC >= lgcCut,])
      
      bothSpecies = c(hgcSpecies, lgcSpecies)
    }
    
    
    if (F) {
      hgcCut = quantile(statsMgsForNormSelHgc$qvalue, 0.1)
      lgcCut = quantile(statsMgsForNormSelLgc$qvalue, 0.1)
      
      hgcSpecies = rownames(statsMgsForNormSelHgc[statsMgsForNormSelHgc$qvalue <= hgcCut,])
      lgcSpecies = rownames(statsMgsForNormSelLgc[statsMgsForNormSelLgc$qvalue <= lgcCut,])
      
      bothSpecies = c(hgcSpecies, lgcSpecies)
    }
    
    
    sort(getSpeciesName(hgcSpecies, taxo))
    sort(getSpeciesName(lgcSpecies, taxo))
    
    ###
    statsMgsForNormSelHgcNew = statsMgsForNormSelHgc[order(statsMgsForNormSelHgc$qvalue),]
    statsMgsForNormSelLgcNew = statsMgsForNormSelLgc[order(statsMgsForNormSelLgc$qvalue),]
    hgcTopSpecies = rownames(statsMgsForNormSelHgcNew[1:25,])
    lgcTopSpecies = rownames(statsMgsForNormSelLgcNew[1:25,])
    
    
    getSpeciesName(hgcTopSpecies, taxo)
    getSpeciesName(lgcTopSpecies, taxo)
  }
  
  newCutMode = T
  if (newCutMode) {
    adjpCut = 1e-3
    statsMgsForNorm = read.table(statsMgsForNormSamples, sep="\t", header=T, stringsAsFactors = F)
    statsMgsForNormSel = statsMgsForNorm[statsMgsForNorm$qvalue <= adjpCut,]
    statsMgsForNormSelHgc = statsMgsForNormSel[statsMgsForNormSel$lfc > 2,]
    statsMgsForNormSelLgc = statsMgsForNormSel[statsMgsForNormSel$lfc < -2,]
    
    hgcSpecies = rownames(statsMgsForNormSelHgc)
    lgcSpecies = rownames(statsMgsForNormSelLgc)
    
  }
  
  
  
  diseaseUpSpecies = unique(allPairMeltSel$msp)
  
  allPairMeltRData = "C://Data//comparative.analysis.healthy.sweden//allPairMelt.disease.signatures.RData"
  load(allPairMeltRData)
  allPairMeltSel = allPairMelt[allPairMelt$effect_size>=0.5,]
  allPairMeltSel = allPairMeltSel[!is.na(allPairMeltSel$effect_size),]
  diseaseUpSpecies = unique(allPairMeltSel$msp)
  
  
  allPairDownMeltRData = "C://Data//comparative.analysis.healthy.sweden//allPairDownMelt.disease.signatures.RData"
  load(allPairDownMeltRData)
  allPairDownMeltSel = allPairDownMelt[allPairDownMelt$effect_size>=0.5,]
  allPairDownMeltSel = allPairDownMeltSel[!is.na(allPairDownMeltSel$effect_size),]
  diseaseDownSpecies = unique(allPairDownMeltSel$msp)
  
  richSpecMgsMat = mergeMatUpdated[diseaseUpSpecies,]
  richSpecMgsMat = richSpecMgsMat[rowMeans(richSpecMgsMat)!=0,]
  richSpecMgsMat = richSpecMgsMat[,colMeans(richSpecMgsMat)!=0]
  
  pcoaMgsMat = pcoaMat[,1:2]
  
  out = plotContourBoth(pcoaMgsMat, 
                        newCaseByEnteroList$ETF,
                        newContByEnteroList$ETF)
  
  
  par(mfrow=c(1,2))
  zD2 = plotContour(pcoaMgsMat[rownames(pcoaMgsMat) %in% newCaseByEnteroList$ETF,])
  zC2 = plotContour(pcoaMgsMat[rownames(pcoaMgsMat) %in%newContByEnteroList$ETF,])
  
  zAll = list(x=c(zD2$x, zC2$x),
              y=c(zD2$y, zC2$y),
              z=c(zD2$z, zC2$z))
  
  contourMat = pcoaMgsMat[rownames(pcoaMgsMat) %in% newCaseByEnteroList$ETF,]
  plot(contourMat, xlab="X label", ylab="Y label", col="#80808033", pch=19, cex=.4, xlim=c(-1,1),ylim=c(-1,1))
  
  contour(zC2, drawlabels=FALSE, nlevels=11, col=my.cols, add=TRUE)
  
  
  par(mfrow=c(1,2))
  zD2 = plotContour(pcoaMgsMat[rownames(pcoaMgsMat) %in% newCaseByEnteroList$ETB,])
  zC2 = plotContour(pcoaMgsMat[rownames(pcoaMgsMat) %in%newContByEnteroList$ETB,])
  
  par(mfrow=c(1,2))
  zD2 = plotContour(pcoaMgsMat[rownames(pcoaMgsMat) %in% newCaseByEnteroList$ETF,])
  zC2 = plotContour(pcoaMgsMat[rownames(pcoaMgsMat) %in%newContByEnteroList$ETF,])
  
  vegOut=vegdist(t(richSpecMgsMat),"bray")
  pcoaRichSpecMgsOut=pcoa(vegOut)
  pcoaRichSpecMgsMat= pcoaRichSpecMgsOut$vectors[,1:2]
  
  checkAnosim = F
  if (checkAnosim) {
    ### ETB-country
    etbSampels = unique(c(newCaseByEnteroList$ETB,
                          newContByEnteroList$ETB))
    
    etbClasses = rep("case",length(etbSampels))
    etbClasses[colnames(richSpecMgsMat[,etbSampels]) %in% newContByEnteroList$ETB] = "control"
    
    vegEtbOut=vegdist(t(richSpecMgsMat[,etbSampels]),"bray")
    pcoaEtbOut=pcoa(vegOut)
    anosimEtbOut = anosim(vegEtbOut, etbClasses) 
    # ANOSIM statistic R: 0.05184 
    # Significance: 0.001 
    
    
    ### ETF-country
    
  }
  
  out = plotContourBoth(pcoaRichSpecMgsMat, 
                        newCaseByEnteroList$ETF,
                        newContByEnteroList$ETF)
  out2 = plotContourBoth(pcoaRichSpecMgsMat, 
                         newCaseByEnteroList$ETB,
                         newContByEnteroList$ETB)
  
  
  
  
  
  
  
  
  
  par(mfrow=c(1,2))
  zD2 = plotContour(pcoaRichSpecMgsMat[rownames(pcoaRichSpecMgsMat) %in% newCaseByEnteroList$ETF,])
  zC2 = plotContour(pcoaRichSpecMgsMat[rownames(pcoaRichSpecMgsMat) %in%newContByEnteroList$ETF,])
  
  par(mfrow=c(1,2))
  zD = plotContour(pcoaRichSpecMgsMat[rownames(pcoaRichSpecMgsMat) %in% newCaseByEnteroList$ETB,])
  zC = plotContour(pcoaRichSpecMgsMat[rownames(pcoaRichSpecMgsMat) %in%newContByEnteroList$ETB,])
  
  par(mfrow=c(1,1))
  plotContour(pcoaRichSpecMgsMat)
  
  
  
  
}

checkIndJaccMode = T
if (checkIndJaccMode) {
  
  #same individual from wellness
  #T1D same individual
  #metformin
  #immunotherapy
  #antibiotics
  
  mergeMatUpdatedJaccRData = "J:\\Deposit\\Project\\2018_microbiome_atlas\\microbiome.atlas.Rproj//mergeMatUpdatedJaccard.RData"
  load(mergeMatUpdatedJaccRData) #mergeMatUpdatedJacc
  dim(mergeMatUpdatedJacc)
  
  
  loadSampleInfo = T
  if (loadSampleInfo) {
    
    
    getJaccPairList <- function(samples, basicMetadata, speciesJaccMat) {
      
      currBasicMetadata = basicMetadata[basicMetadata$sample.ID %in% samples, ]
      samplesByIndividualList = split(currBasicMetadata$sample.ID, currBasicMetadata$metadata.ID)
      
      jaccVecs = lapply(samplesByIndividualList, function(currSamples){
        if (length(currSamples)<=1) return(NA)
        currMat = speciesJaccMat[currSamples, currSamples]
        return(currMat[lower.tri(currMat)])
      })
      return(jaccVecs)
    }
    intraMat=mergeMatUpdatedJacc[!duplicated(basicMetaMapUpdated$metadata.ID),!duplicated(basicMetaMapUpdated$metadata.ID)]
    intraMat = intraMat[rownames(intraMat)%in% newNormalAllIds, colnames(intraMat)%in% newNormalAllIds ]
    allJaccs = intraMat[lower.tri(intraMat)]
    hist(allJaccs)
    
    wellnessSamples = basicMetaMapUpdated$sample.ID[basicMetaMapUpdated$dataset.ID=="id28"]
    wellnessJaccs = unlist(getJaccPairList(wellnessSamples, basicMetaMapUpdated, mergeMatUpdatedJacc))
    hist(wellnessJaccs)
    
    antibioticsSamples = basicMetaMapUpdated$sample.ID[basicMetaMapUpdated$dataset.ID=="id51"]
    if (F) {
      id51MetaIds = basicMetaMapUpdated$metadata.ID[basicMetaMapUpdated$dataset.ID=="id51"]
      id51MetaIds = gsub("_Dag0$","",id51MetaIds)
      id51MetaIds = gsub("_Dag42$","",id51MetaIds)
      id51MetaIds = gsub("_Dag4opt$","",id51MetaIds)
      id51MetaIds = gsub("_Dag180$","",id51MetaIds)
      id51MetaIds = gsub("_Dag4$","",id51MetaIds)
      id51MetaIds = gsub("_Dag8$","",id51MetaIds)
      id51MetaIds = gsub("_Dag8opt$","",id51MetaIds)
      basicMetaMapUpdated$metadata.ID[basicMetaMapUpdated$dataset.ID=="id51"] = id51MetaIds
      
    }
    
    antibioticsJaccs = unlist(getJaccPairList(antibioticsSamples, basicMetaMapUpdated, mergeMatUpdatedJacc))
    hist(antibioticsJaccs)
    
    t1dLuSamples = basicMetaMapUpdated$sample.ID[basicMetaMapUpdated$dataset.ID=="id35" 
                                                 & basicMetaMapUpdated$type == "Case"]
    t1dLuJaccs = unlist(getJaccPairList(t1dLuSamples, basicMetaMapUpdated, mergeMatUpdatedJacc))
    hist(t1dLuJaccs)
    
    contLuSamples = basicMetaMapUpdated$sample.ID[basicMetaMapUpdated$dataset.ID=="id35" 
                                                  & basicMetaMapUpdated$type == "Control"]
    contLuJaccs = unlist(getJaccPairList(contLuSamples, basicMetaMapUpdated, mergeMatUpdatedJacc))
    hist(contLuJaccs)
    
    ##update basicmetadata for metformin data
    id53MetaIds = basicMetaMapUpdated$metadata.ID[basicMetaMapUpdated$dataset.ID=="id53"]
    id53MetaIds = gsub("V[0-9]+$","",id53MetaIds)
    basicMetaMapUpdated$metadata.ID[basicMetaMapUpdated$dataset.ID=="id53"] = id53MetaIds
    
    exMetfominSamples = c("SRR5580332", "SRR5580333", "SRR5580334", "SRR5580335", "SRR5579968", "SRR5579973", "SRR5579974", "SRR5580118", "SRR5580113", "SRR5580120", "SRR5580121", "SRR5580272", "SRR5580392", "SRR5580393", "SRR5580259", "SRR5580260", "SRR5580394", "SRR5580395", "SRR5579978", "SRR5579979", "SRR5579980", "SRR5579983", "SRR5580316", "SRR5580317", "SRR5580318", "SRR5580320", "SRR5558211", "SRR5558212", "SRR5558213", "SRR5558214", "SRR5558297", "SRR5558298", "SRR5558304", "SRR5580103", "SRR5580313", "SRR5580314", "SRR5580327", "SRR5580039", "SRR5580040", "SRR5580045", "SRR5580046", "SRR5579956", "SRR5580397", "SRR5580399", "SRR5558182", "SRR5558183", "SRR5558191", "SRR5558192", "SRR5558092", "SRR5558093", "SRR5558094", "SRR5558157", "SRR5558158", "SRR5558219", "SRR5558220", "SRR5558169", "SRR5558178", "SRR5558179", "SRR5558181", "SRR5558190", "SRR5558309", "SRR5558313", "SRR5558330", "SRR5558331", "SRR5558061", "SRR5558062", "SRR5558063", "SRR5558068", "SRR5558069", "SRR5558194", "SRR5558195", "SRR5558196", "SRR5558206", "SRR5558102", "SRR5558107", "SRR5558108", "SRR5558232", "SRR5558071", "SRR5558072", "SRR5558073", "SRR5558154", "SRR5558284", "SRR5558386", "SRR5558387", "SRR5558388", "SRR5580390", "SRR5580391", "SRR5580392", "SRR5580393")
    metforminT2dSamples = basicMetaMapUpdated$sample.ID[basicMetaMapUpdated$dataset.ID=="id53" 
                                                        & basicMetaMapUpdated$subtype %in% c("M0","M2","M4")]
    metforminT2dSamples = metforminT2dSamples[!metforminT2dSamples %in% exMetfominSamples]
    
    metforminT2dJaccs = unlist(getJaccPairList(metforminT2dSamples, basicMetaMapUpdated, mergeMatUpdatedJacc))
    hist(metforminT2dJaccs)
    
    contT2dSamples = basicMetaMapUpdated$sample.ID[basicMetaMapUpdated$dataset.ID=="id53" 
                                                   & basicMetaMapUpdated$subtype %in% c("P0","P2","P4")]
    contT2dSamples = contT2dSamples[!contT2dSamples %in% exMetfominSamples]
    
    contT2dJaccs = unlist(getJaccPairList(contT2dSamples, basicMetaMapUpdated, mergeMatUpdatedJacc))
    hist(contT2dJaccs)
    
    exRccSamples = c("ERR2235250", "ERR2235262", "ERR2235267", "ERR2235270", "ERR2235273", "ERR2235286", "ERR2235292", "ERR2235295", "ERR2235299", "ERR2235308")
    
    rccSamples = basicMetaMapUpdated$sample.ID[basicMetaMapUpdated$dataset.ID=="id41"]
    rccSamples = rccSamples[!rccSamples %in% exRccSamples]
    if (F) {
      rccMetaIds = basicMetaMapUpdated$metadata.ID[basicMetaMapUpdated$dataset.ID=="id41"]
      rccMetaIds = gsub("_T[0-9]$","",rccMetaIds)
      
    }
    basicMetaMapUpdated$metadata.ID[basicMetaMapUpdated$dataset.ID=="id41"] = rccMetaIds
    
    rccJaccs = unlist(getJaccPairList(rccSamples, basicMetaMapUpdated, mergeMatUpdatedJacc))
    hist(rccJaccs)
    exNsclcSamples = c("ERR2213662", "ERR2213668", "ERR2213684", "ERR2213687", "ERR2213690", "ERR2213693", "ERR2213698", "ERR2213703", "ERR2213709", "ERR2213730", "ERR2213739", "ERR2213743")
    
    nsclcSamples = basicMetaMapUpdated$sample.ID[basicMetaMapUpdated$dataset.ID=="id26"]
    nsclcSamples = nsclcSamples[!nsclcSamples %in% exNsclcSamples]
    if (F) {
      nsclcMetaIds = basicMetaMapUpdated$metadata.ID[basicMetaMapUpdated$dataset.ID=="id26"]
      nsclcMetaIds = gsub("_T[0-9]$","",nsclcMetaIds)
    }
    basicMetaMapUpdated$metadata.ID[basicMetaMapUpdated$dataset.ID=="id26"] = nsclcMetaIds
    
    nsclcJaccs = unlist(getJaccPairList(nsclcSamples, basicMetaMapUpdated, mergeMatUpdatedJacc))
    hist(nsclcJaccs)
    
    jaccList = list(all=allJaccs[!is.na(allJaccs)], 
                    antibiotics = antibioticsJaccs[!is.na(antibioticsJaccs)],
                    wellness=wellnessJaccs[!is.na(wellnessJaccs)],
                    t1d_LU=t1dLuJaccs[!is.na(t1dLuJaccs)],
                    cont_LU=contLuJaccs[!is.na(contLuJaccs)],
                    metformin_T2D=metforminT2dJaccs[!is.na(metforminT2dJaccs)],
                    cont_T2D=contT2dJaccs[!is.na(contT2dJaccs)],
                    rcc_immunotherapy=rccJaccs[!is.na(rccJaccs)],
                    nsclc_immunotherapy=nsclcJaccs[!is.na(nsclcJaccs)])
    par(mar=c(10,4,4,1))
    boxplot(jaccList,las=2, ylab="Jaccard")
    jaccMeans = unlist(lapply(jaccList, mean))
    jaccSds = unlist(lapply(jaccList, sd))
    
    newNames = c("Inter-individual",
                 "Antibiotics (DK)",
                 "Intra-individual (ES)",
                 "Intra-individual (T1D, LU)",
                 "Intra-individual (LU)",
                 "Metformin (T2D, ES)",
                 "Intra-individual (T2D, ES)",
                 "Immunotherapy (RCC, FR)",
                 "Immunotherapy (NSCLC, FR)")
    
    orderedNames = c("Inter-individual",
                     "Intra-individual (LU)",
                     "Intra-individual (ES)",
                     "Intra-individual (T2D, ES)",
                     "Intra-individual (T1D, LU)",
                     "Antibiotics (DK)",
                     "Metformin (T2D, ES)",
                     "Immunotherapy (NSCLC, FR)",
                     "Immunotherapy (RCC, FR)")
    
    classes = c("Inter",
                "Perturb",
                "Intra.normal",
                "Intra.disease",
                "Intra.normal",
                "Perturb",
                "Intra.disease",
                "Perturb",
                "Perturb")
    jaccStatTab = data.frame(name=newNames,
                             mean=jaccMeans,
                             sd=jaccSds,
                             class=classes)
    
    jaccStatTab$name = factor(jaccStatTab$name, levels=orderedNames)
    
    ggplot(jaccStatTab, aes(x=name, y=mean, size=sd)) +
      geom_point(aes(fill=factor(class)), colour="black", shape=21) +
      geom_vline(xintercept = c(1.5, 3.5, 5.5))+
      xlab("") + ylab("Jaccard") +
      scale_size(range = c(2, 10))+
      ylim(c(0.1,0.8))+
      theme(panel.grid.major = element_line(colour="gray"), 
            axis.text.x = element_text(angle = 60, hjust=1),
            axis.ticks = element_blank(),
            legend.position = "none",
            panel.background = element_rect(fill = "white", 
                                            colour = "black", 
                                            size = 0.5, 
                                            linetype = "solid")) 
    
  }
}


####### UNUSED SOURCE CODES #######
checkPlotWorldMap  = F
if (checkPlotWorldMap) {
  library(rgdal)
  #download.file("http://thematicmapping.org/downloads/TM_WORLD_BORDERS_SIMPL-0.3.zip" , destfile="DATA/world_shape_file.zip")
  
  
  my_spdf <- readOGR( 
    dsn= path.expand("J://Deposit//Project//2018_microbiome_atlas//functional.annotation"), 
    layer="TM_WORLD_BORDERS_SIMPL-0.3",
    verbose=FALSE
  )
  
  library(broom)
  spdf_fortified <- tidy(my_spdf, region = "NAME")
  
  # Plot it
  library(ggplot2)
  ggplot() +
    geom_polygon(data = spdf_fortified, aes( x = long, y = lat, group = group), fill="#69b3a2", color="white") +
    theme_void() 
  
  par(mar=c(0,0,0,0))
  plot(my_spdf, col="#f2f2f2", bg="skyblue", lwd=0.25, border=0 )
  
}

checkDatasetInfoNew2 = F
if (checkDatasetInfoNew2) {
  
  datasets = unique(basicMetaMapUpdated$dataset.ID)
  
  checkDuplicates = F
  if (checkDuplicates) {
    for( ds in datasets ) {
      isDuplicated = duplicated(basicMetaMapUpdated[basicMetaMapUpdated$dataset.ID==ds,"metadata.ID"])
      print(ds)
      print(table(isDuplicated))
    }
    
  }
  
  checkScatterPlotMode = T 
  if (checkScatterPlotMode) {
    
    diseaseMgsMat = mergeMatUpdated[,diseaseUniqAll] 
    industrialMgsMat = mergeMatUpdated[,industrializedSamples]
    traditionalMgsMat = mergeMatUpdated[,traditionalSamples]
    
    diseaseMGsRowMs = rowMeans(diseaseMgsMat)
    industrialMGsRowMs = rowMeans(industrialMgsMat)
    traditionalMGsRowMs = rowMeans(traditionalMgsMat)
    
    industrialSpecies = names(industrialMGsRowMs[industrialMGsRowMs!=0])
    traditionalSpecies = names(traditionalMGsRowMs[traditionalMGsRowMs!=0])
    
    diseaseTop10 =  names(sort(diseaseMGsRowMs, decreasing = T)[1:10])
    industrialTop10 =  names(sort(industrialMGsRowMs, decreasing = T)[1:10])
    traditionalTop10 =  names(sort(traditionalMGsRowMs, decreasing = T)[1:10])
    
    tradOnlySpecies = traditionalTop10[!traditionalTop10 %in% industrialTop10]
    industOnlySpecies = industrialTop10[!industrialTop10 %in% traditionalTop10]
    commSpecies = traditionalTop10[traditionalTop10 %in% industrialTop10]
    
    
    freqs = rowSums(mergeMatUpdated > 0)
    avgs = rowMeans(mergeMatUpdated)
    maxs = rowMaxs(mergeMatUpdated)
    
    cols = rep("#A9A9A944", length(freqs))
    names(cols) = names(freqs)
    cols[tradOnlySpecies] = "#FF000088"
    cols[industOnlySpecies] = "#0000FF55"
    cols[commSpecies] = "#800080"
    
    
    plot(avgs, freqs, col=cols, pch=16, log='x', las=1, xlab="",ylab="")
    abline(v=quantile(avgs, 0.5))
    abline(v=quantile(avgs, 0.99))
    
    ###
    plot(maxs, avgs, log="xy")
    
    
    pairs(cbind(diseaseMGsRowMs,
                industrialMGsRowMs,
                traditionalMGsRowMs), log="xy")
    
    diseaseSpecies = rowSums(mergeMatUpdated[,diseaseUniqAll] >0)
    diseaseSpecies = names(diseaseSpecies[diseaseSpecies>0])
    
    normSpecies = rowSums(mergeMatUpdated[,normUniqAll] >0)
    normSpecies = names(normSpecies[normSpecies>0])
    
    
    industSpecies = rowSums(mergeMatUpdated[,industrializedSamples] >0)
    industSpecies = names(industSpecies[industSpecies>0])
    
    tradSpecies = rowSums(mergeMatUpdated[,traditionalSamples] >0)
    tradSpecies = names(tradSpecies[tradSpecies>0])
    
    tradOnlySpecies = tradSpecies[!tradSpecies %in% industSpecies]
    industOnlySpecies = industSpecies[!industSpecies %in% tradSpecies]
    diseaseOnlySpecies = diseaseSpecies[!diseaseSpecies %in% normSpecies]
    
    
  }
  
  getDiseaseSigByPlsDaMode = T
  if (getDiseaseSigByPlsDaMode) {
    
    genusMode = T
    if (genusMode) {
      
      diseaseGenusMat = genusMatUpdated[,diseaseUniqAll] 
      normalGenusMat = genusMatUpdated[,normUniqAll]
      industrialGenusMat = genusMatUpdated[,industrializedSamples]
      traditionalGenusMat = genusMatUpdated[,traditionalSamples]
      
      diseaseVipCollectList = list()
      for( ind in seq_along(cohortIdByDiseases)) {
        currDiseaseCohorts = cohortIdByDiseases[[ind]]
        print(names(cohortIdByDiseases)[ind])
        currDiseaseSamples = unique(do.call(c,diseaseUniqSubjectList[currDiseaseCohorts]))
        currDiseaseGenusMat = genusMatUpdated[,currDiseaseSamples]
        currDataMat = cbind(industrialGenusMat, currDiseaseGenusMat)
        
        currPlsClasses = c(rep("health", dim(industrialGenusMat)[2]),
                           rep("disease", dim(currDiseaseGenusMat)[2]))
        currPlsCols = c(rep("#A9A9A988",dim(industrialGenusMat)[2]), #health: gray
                        rep("#FF000088",dim(currDiseaseGenusMat)[2])) #disease: red
        
        currPlsDaOut = opls(t(currDataMat), currPlsClasses, plotL=F, predI=2)
        currVips  = getVipVn(currPlsDaOut)
        currVipsAboveTwo = names(currVips[currVips>=2])
        currVipList = getVipList(currVips, currDataMat, currPlsClasses)
        
        diseaseVipCollectList[[ind]] = currVipsAboveTwo
        plotMode = F
        if (plotMode) {
          fileName = paste(names(cohortIdByDiseases)[ind], "pls_da" , "pdf", sep=".")
          pdf(fileName, width = 2, height = 2)
          #plot(currPlsDaOut@scoreMN[,1:2], col=currPlsCols, pch=16,cex=1,las=1,xlab="",ylab="", xaxt="n",yaxt="n")
          s.class(currPlsDaOut@scoreMN[,1:2], fac = factor(currPlsCols), sub = names(cohortIdByDiseases)[ind],
                  cell = 2, axesell = FALSE, csta = 0, grid=F,addaxes = F,
                  col = unique(currPlsCols), clabel = FALSE)
          dev.off()
        }
      }
      names(diseaseVipCollectList) = names(cohortIdByDiseases)
      
      diseaseVipTab = melt(diseaseVipCollectList)
      diseaseVipTab$value = as.character(diseaseVipTab$value)
      diseaseVipTab$L1 = as.character(diseaseVipTab$L1)
      diseaseVipTab$count = 1
      diseaseVipMat = acast(diseaseVipTab, value ~ L1, value.var = "count")
      diseaseVipMat[is.na(diseaseVipMat)]=0
      colSums(diseaseVipMat)
      dim(diseaseVipMat)
      heatmap.2(diseaseVipMat, trace="none",
                col = colorRampPalette(c("white","red"))(256),
                margins = c(5,12),
                key=F,lhei=c(1,20),
                cexCol = 0.5,cexRow = 0.5,
                sepcolor = "gray",
                colsep=seq(0, dim(diseaseVipMat)[2]),
                rowsep=seq(0, dim(diseaseVipMat)[1]),
                sepwidth = c(0.01,0.01))
      
      
      allInputList=c(newDiseaseList, newNormalAllList)
      allInputList = newNormalAllList[c("US_id11",
                                        "US_id36",
                                        "US_id34",
                                        "US_id43")] 
      
      allInputList = newNormalAllList[c("UK_id29",
                                        "UK_id39")] 
      
      allInputList = newNormalAllList[c("Italy_id46",
                                        "Italy_id37")] 
      
      allInputList = newNormalAllList[c("Sweden_id28",
                                        "Sweden_id1",
                                        "Sweden_id14")] 
      
      allInputList = newNormalAllList[c("China_id10",
                                        "China_id12",
                                        "China_id2",
                                        "China_id20",
                                        "China_id31",
                                        "China_id6",
                                        "China_id9")] 
      
      allInputEnteroList = getEntGeoPlot(grepEnteroTypesFrom(allInputList, basicMetaMapUpdated))
      
      entGeoTab = getEntGeoPlot(allInputEnteroList)
      
      avgGenusDiseaseMat = getAvgAbdMatBy(genusMatUpdated, c(diseaseUniqList) )
      avgGenusDiseaseMeans = rowMeans(avgGenusDiseaseMat)
      topGenera = names(sort(avgGenusDiseaseMeans,decreasing = T)[1:108])
      
      avgGenusDiseaseMatSel = avgGenusDiseaseMat[rownames(diseaseVipMat),]
      avgGenusDiseaseMatSel2 = avgGenusDiseaseMat[topGenera,]
      
      cor(avgGenusDiseaseMatSel)
      cc = melt( cor(avgGenusDiseaseMatSel))
      cc = cc[cc$value>=0.99,]
      cc = cc[cc$Var1 != cc$Var2,]
      View(cc)
      
      heatmap.2(cor(avgGenusDiseaseMatSel),trace="none",key=F,
                col = colorRampPalette(c("blue","white","red"))(256))
      heatmap.2(cor(avgGenusDiseaseMat),trace="none",key=F,
                col = colorRampPalette(c("blue","white","red"))(256))
      
      
      set.seed(1)
      targetMat = avgGenusDiseaseMatSel2
      diseaseTsneOut = Rtsne(t(targetMat), dims=2, perplexity = 5)
      rownames(diseaseTsneOut$Y) = colnames(targetMat)
      plot(diseaseTsneOut$Y)
      
      tsneTab = data.frame(diseaseTsneOut$Y, type=rownames(diseaseTsneOut$Y))
      ggplot(tsneTab, aes(x=X1 , y=X2, label=type)) + 
        geom_point(aes(colour = type, alpha=.02)) + 
        theme(panel.background = element_blank(),
              axis.ticks = element_blank(),
              axis.text.x = element_blank(),
              strip.text.x = element_blank(),
              axis.text.y = element_blank(),
              strip.text.y = element_blank(),
              legend.position="none", 
              panel.border = element_rect(colour = "black", fill=NA, size=.5))  + 
        geom_text_repel(size=3, force=2) + xlab("") + ylab("")
      
    }
    
    mgsMode = T
    if (mgsMode) {
      
      countryMode = T
      if (countryMode) {
        
        
        diseaseMgsCountryVipCollectList = list()
        for (ind in 1:30) {
          currDiseaseCohortId = diseaseWithCountryPairTab2[ind,1]
          currCountryId = diseaseWithCountryPairTab2[ind,2]
          currDiseaseSamples = diseaseUniqSubjectList[[currDiseaseCohortId]]
          currContSamples = normUniqCountryList[[currCountryId]] 
          currDiseaseMat = mergeMatUpdated[,currDiseaseSamples]
          currContMat = mergeMatUpdated[,currContSamples]
          currDataMat = cbind(currContMat, currDiseaseMat)
          
          currPlsClasses = c(rep("health", dim(currContMat)[2]),
                             rep("disease", dim(currDiseaseMat)[2]))
          currPlsCols = c(rep("#A9A9A988",dim(currContMat)[2]), #health: gray
                          rep("#FF000088",dim(currDiseaseMat)[2])) #disease: red
          
          currPlsDaOut = opls(t(currDataMat), currPlsClasses, plotL=F, predI=2)
          currVips  = getVipVn(currPlsDaOut)
          currVipsAboveOne = (currVips[currVips>=1])
          currVipList = getVipList(currVips, currDataMat, currPlsClasses)
          diseaseMgsCountryVipCollectList[[ind]] = currVipsAboveOne
          
          
          plotMode = T
          if (plotMode) {
            currTopVipList = takeTopOfList(currVipList, 10)
            currTopVipList = labelVipList(currTopVipList, taxo)
            
            fileName = paste(currDiseaseCohortId, "country.MGS.VIP.top10" , "pdf", sep=".")
            pdf(fileName, width = 7, height = 5)
            plotVipList(currTopVipList, marVec = c(5,25,4,1))
            dev.off()
            
            
            fileName = paste(currDiseaseCohortId, "country.MGS.pls_da" , "pdf", sep=".")
            pdf(fileName, width = 2, height = 2)
            #plot(currPlsDaOut@scoreMN[,1:2], col=currPlsCols, pch=16,cex=1,las=1,xlab="",ylab="", xaxt="n",yaxt="n")
            s.class(currPlsDaOut@scoreMN[,1:2], fac = factor(currPlsCols), sub = currDiseaseCohortId,
                    cell = 2, axesell = FALSE, csta = 0, grid=F,addaxes = F,
                    col = unique(currPlsCols), clabel = FALSE)
            dev.off()
          }
          
          
          
        }
        names(diseaseMgsCountryVipCollectList) = diseaseWithCountryPairTab2[,1]
        
        diseaseMgsCountryVipCollectList2 = lapply(diseaseMgsCountryVipCollectList, names)
        diseaseMgsVipTab = melt(diseaseMgsCountryVipCollectList2)
        diseaseMgsVipTab$value = as.character(diseaseMgsVipTab$value)
        diseaseMgsVipTab$L1 = as.character(diseaseMgsVipTab$L1)
        diseaseMgsVipTab$count = 1
        diseaseMgsVipTab$species = sapply(diseaseMgsVipTab$value, function(x) getSpeciesName(x, taxo))
        
        
        
        
      }
      allCompMode = T
      if (allCompMode) {
        diseaseMat = mergeMatUpdated[,diseaseUniqAll] 
        normalMat = mergeMatUpdated[,normUniqAll]
        industrialMat = mergeMatUpdated[,industrializedSamples]
        traditionalMat = mergeMatUpdated[,traditionalSamples]
        
        diseaseMgsVipCollectList = list()
        for( ind in seq_along(cohortIdByDiseases)) {
          currDiseaseCohorts = cohortIdByDiseases[[ind]]
          print(names(cohortIdByDiseases)[ind])
          currDiseaseSamples = unique(do.call(c,diseaseUniqSubjectList[currDiseaseCohorts]))
          currDiseaseMat = mergeMatUpdated[,currDiseaseSamples]
          currDataMat = cbind(industrialMat, currDiseaseMat)
          
          currPlsClasses = c(rep("health", dim(industrialMat)[2]),
                             rep("disease", dim(currDiseaseMat)[2]))
          currPlsCols = c(rep("#A9A9A988",dim(industrialMat)[2]), #health: gray
                          rep("#FF000088",dim(currDiseaseMat)[2])) #disease: red
          
          currPlsDaOut = opls(t(currDataMat), currPlsClasses, plotL=F, predI=2)
          currVips  = getVipVn(currPlsDaOut)
          currVipsAboveTwo = names(currVips[currVips>=2])
          currVipList = getVipList(currVips, currDataMat, currPlsClasses)
          
          diseaseMgsVipCollectList[[ind]] = currVipsAboveTwo
          
          plotMode = T
          if (plotMode) {
            fileName = paste(names(cohortIdByDiseases)[ind], "MGS.pls_da" , "pdf", sep=".")
            pdf(fileName, width = 2, height = 2)
            #plot(currPlsDaOut@scoreMN[,1:2], col=currPlsCols, pch=16,cex=1,las=1,xlab="",ylab="", xaxt="n",yaxt="n")
            s.class(currPlsDaOut@scoreMN[,1:2], fac = factor(currPlsCols), sub = names(cohortIdByDiseases)[ind],
                    cell = 2, axesell = FALSE, csta = 0, grid=F,addaxes = F,
                    col = unique(currPlsCols), clabel = FALSE)
            dev.off()
          }
        }
        names(diseaseMgsVipCollectList) = names(cohortIdByDiseases)
        
        diseaseVipTab = melt(diseaseMgsVipCollectList)
        diseaseVipTab$value = as.character(diseaseVipTab$value)
        diseaseVipTab$L1 = as.character(diseaseVipTab$L1)
        diseaseVipTab$count = 1
        diseaseVipTab$species = sapply(diseaseVipTab$value, function(x) getSpeciesName(x, taxo))
        
        diseaseVipMat = acast(diseaseVipTab, value ~ L1, value.var = "count")
        diseaseVipMat[is.na(diseaseVipMat)]=0
        colSums(diseaseVipMat)
        dim(diseaseVipMat)
        
        
        heatmap.2(diseaseVipMat, trace="none",
                  col = colorRampPalette(c("white","red"))(256),
                  margins = c(5,12),
                  key=F,lhei=c(1,20),
                  cexCol = 0.5,cexRow = 0.5,
                  sepcolor = "gray",
                  colsep=seq(0, dim(diseaseVipMat)[2]),
                  rowsep=seq(0, dim(diseaseVipMat)[1]),
                  sepwidth = c(0.01,0.01))
        
      }
      
      
      allInputList=c(newDiseaseList, newNormalAllList)
      allInputList = newNormalAllList[c("US_id11",
                                        "US_id36",
                                        "US_id34",
                                        "US_id43")] 
      
      allInputList = newNormalAllList[c("UK_id29",
                                        "UK_id39")] 
      
      allInputList = newNormalAllList[c("Italy_id46",
                                        "Italy_id37")] 
      
      allInputList = newNormalAllList[c("Sweden_id28",
                                        "Sweden_id1",
                                        "Sweden_id14")] 
      
      allInputList = newNormalAllList[c("China_id10",
                                        "China_id12",
                                        "China_id2",
                                        "China_id20",
                                        "China_id31",
                                        "China_id6",
                                        "China_id9")] 
      
      allInputEnteroList = getEntGeoPlot(grepEnteroTypesFrom(allInputList, basicMetaMapUpdated))
      
      entGeoTab = getEntGeoPlot(allInputEnteroList)
      
      avgGenusDiseaseMat = getAvgAbdMatBy(genusMatUpdated, c(diseaseUniqList) )
      avgGenusDiseaseMeans = rowMeans(avgGenusDiseaseMat)
      topGenera = names(sort(avgGenusDiseaseMeans,decreasing = T)[1:108])
      
      avgGenusDiseaseMatSel = avgGenusDiseaseMat[rownames(diseaseVipMat),]
      avgGenusDiseaseMatSel2 = avgGenusDiseaseMat[topGenera,]
      
      cor(avgGenusDiseaseMatSel)
      cc = melt( cor(avgGenusDiseaseMatSel))
      cc = cc[cc$value>=0.99,]
      cc = cc[cc$Var1 != cc$Var2,]
      View(cc)
      
      heatmap.2(cor(avgGenusDiseaseMatSel),trace="none",key=F,
                col = colorRampPalette(c("blue","white","red"))(256))
      heatmap.2(cor(avgGenusDiseaseMat),trace="none",key=F,
                col = colorRampPalette(c("blue","white","red"))(256))
      
      
      set.seed(1)
      targetMat = avgGenusDiseaseMatSel2
      diseaseTsneOut = Rtsne(t(targetMat), dims=2, perplexity = 5)
      rownames(diseaseTsneOut$Y) = colnames(targetMat)
      plot(diseaseTsneOut$Y)
      
      tsneTab = data.frame(diseaseTsneOut$Y, type=rownames(diseaseTsneOut$Y))
      ggplot(tsneTab, aes(x=X1 , y=X2, label=type)) + 
        geom_point(aes(colour = type, alpha=.02)) + 
        theme(panel.background = element_blank(),
              axis.ticks = element_blank(),
              axis.text.x = element_blank(),
              strip.text.x = element_blank(),
              axis.text.y = element_blank(),
              strip.text.y = element_blank(),
              legend.position="none", 
              panel.border = element_rect(colour = "black", fill=NA, size=.5))  + 
        geom_text_repel(size=3, force=2) + xlab("") + ylab("")
      
    }
  }
  
  getDiseaseSigMode = F
  if (getDiseaseSigMode) {
    
    sampleTab = melt(mixedUniqSubjectList)
    colnames(sampleTab) = c("sample.ID", "cohort.desc")
    sampleTab$cohort.ID =  gsub(".*_id","id",sampleTab$cohort.desc)
    
    sampleTab$type = NA
    sampleTab$type[sampleTab$cohort.desc %in% names(diseaseUniqSubjectList)] = "disease"
    sampleTab$type[sampleTab$cohort.desc %in% names(normUniqSubjectList)] = "health"
    sampleTab$type = as.factor(sampleTab$type)
    
    sampleTab$geography = NA
    sapply(seq_along(cohortIdByGeos), function(id) {
      currCountry = names(cohortIdByGeos)[id]
      currCohorts = cohortIdByGeos[[id]]
      sampleTab$geography[sampleTab$cohort.desc %in% currCohorts] <<- currCountry
    })
    sampleTab$geography = as.factor(sampleTab$geography)
    
    sampleTab$sequencer = NA
    seqInfoTab = unique(basicMetaMapUpdated[,c("dataset.ID","Sequencer")])
    apply(seqInfoTab, 1, function(currRow){
      currId = currRow[1]
      currSeq = currRow[2]
      sampleTab$sequencer[sampleTab$cohort.ID == currId] <<- currSeq
    })
    
    sampleTab$Imputation =  basicMetaMapUpdated$Imputation[match(sampleTab$sample.ID, basicMetaMapUpdated$sample.ID)]
    sampleTab$WhatImputed =  basicMetaMapUpdated$WhatImputed[match(sampleTab$sample.ID, basicMetaMapUpdated$sample.ID)]
    sampleTab$Age = basicMetaMapUpdated$Age[match(sampleTab$sample.ID, basicMetaMapUpdated$sample.ID)]
    sampleTab$BMI = basicMetaMapUpdated$BMI[match(sampleTab$sample.ID, basicMetaMapUpdated$sample.ID)]
    
    matchedMgsTab = t(mergeMatUpdated[,match(sampleTab$sample.ID, colnames(mergeMatUpdated))])
    matchedGenusTab = t(genusMatUpdated[,match(sampleTab$sample.ID, colnames(genusMatUpdated))])
    matchedFamilyTab = t(familyMatUpdated[,match(sampleTab$sample.ID, colnames(familyMatUpdated))])
    
    matchedMgsTab = matchedMgsTab*10^9
    matchedGenusTab = matchedGenusTab*10^9
    matchedFamilyTab = matchedFamilyTab*10^9
    
    matchedMgsTab = matchedMgsTab+1
    matchedGenusTab = matchedGenusTab+1
    matchedFamilyTab = matchedFamilyTab+1
    
    matchedMgsTab = log10(matchedMgsTab)
    matchedGenusTab = log10(matchedGenusTab)
    matchedFamilyTab = log10(matchedFamilyTab)
    
    # 
    # inputMgsTab = cbind(sampleTab, matchedMgsTab)
    # inputGenusTab = cbind(sampleTab, matchedGenusTab)
    # inputFamilyTab = cbind(sampleTab, matchedFamilyTab)
    featureTab = matchedMgsTab
    
    lmerSignatureOut = lmerDiseaseSignature(sampleTab, featureTab)
    geoMat = lmerSignatureOut$geography
    diseaseMat = lmerSignatureOut$disease
    diseaseMat[is.na(diseaseMat[,1]),1] = 0
    diseaseMat[is.na(diseaseMat[,2]),2] = 1
    
    featureTab = matchedGenusTab
    lmerGenusSignatureOut = lmerDiseaseSignature(sampleTab, featureTab)
    geoGenusMat = lmerGenusSignatureOut$geography
    diseaseGenusMat = lmerGenusSignatureOut$disease
    
    genusCorrected = getCorrectedFeatureTab(lmerGenusSignatureOut, sampleTab, featureTab)
    
    getCorrectedFeatureTab <- function(signatureOut, sampleTab, featureTab) {
      featureMat = as.matrix(featureTab)
      geoMat = signatureOut$geography
      sampleTab = sampleTab[match(rownames(featureTab), sampleTab$sample.ID),]
      geoFeatureMat = featureMat
      geoFeatureMat = geoFeatureMat[,match(colnames(geoMat), colnames(geoFeatureMat))]
      rownames(geoFeatureMat) = sampleTab$geography
      
      
      invisible(sapply(rownames(geoMat), function(currGeo) {
        currGeoVec = geoMat[currGeo,]
        geoFeatureMat[rownames(geoFeatureMat) == currGeo] <<- geoFeatureMat[rownames(geoFeatureMat) == currGeo] - currGeoVec
      }))
      rownames(geoFeatureMat) = sampleTab$sample.ID
      
      set.seed(1)
      tsneOut = Rtsne(geoFeatureMat, dims=2, perplexity = 100)
      set.seed(1)
      tsneUncorrectedOut = Rtsne(featureMat, dims=2, perplexity = 100 )
      
      
      geoCols = rep("gray",3781)
      geoCols[sampleTab$geography=="France"]="red"
      geoCols[sampleTab$geography=="China"]="blue"
      
      plot(tsneOut$Y, col=geoCols)
      plot(tsneUncorrectedOut$Y, col=geoCols)
      
      avgMgsCorrectMat = getAvgAbdMatBy(t(geoFeatureMat), mixedUniqSubjectList )
      avgMgsUnCorrectMat = getAvgAbdMatBy(t(featureMat), mixedUniqSubjectList )
      
      
      heatmap.2(cor(avgMgsCorrectMat), 
                trace="none", 
                margins = c(7,7), 
                col = colorRampPalette(c("blue","white","red"))(256))
      
      
      heatmap.2(cor(avgMgsUnCorrectMat), 
                trace="none", 
                margins = c(7,7), 
                col = colorRampPalette(c("blue","white","red"))(256))
      
      
      set.seed(1)
      tsneCorrectedAvgOut = Rtsne(t(avgMgsCorrectMat), dims=2, perplexity = 10)
      set.seed(1)
      tsneUncorrectedAvgOut = Rtsne(t(avgMgsUnCorrectMat), dims=2, perplexity = 10 )
      
      dataCols = rep("#80808077", 65)
      dataCols[colnames(avgMgsCorrectMat)%in%cohortIdByGeos$Sweden]="blue" 
      
      plot(tsneCorrectedAvgOut$Y, col=dataCols)
      plot(tsneUncorrectedAvgOut$Y, col=dataCols)
      
      
      return(list(corrected=geoFeatureMat))
      
    }
    
    datasetID = "Ankylosing_CN_id9"
    
    ##### contructing input matrix ####
    # rows: 
    # all possible samples --> DONE
    
    # columns: 
    # sample ID
    # cohort ID
    # healthy/disease --> DONE
    # subtypes
    # cohorts
    # geography
    # sequencer
    # all MGSs
    # all genus
    # all family
    
    # additional variables
    # index indicating MGSs
    # index indicating genus
    # index indicating family
    
    
  }
  
  jaccNetMode = T
  if (jaccNetMode) {
    
    cohortWiseJaccNetMode = T
    if (cohortWiseJaccNetMode) {
      #wellness id28
      #kcl twin?? id39
      #luxembourg id35?? 
      #metformin treatment??? id53??
      
      ###
      
      mergeMatUpdatedJaccRData = "J:\\Deposit\\Project\\2018_microbiome_atlas\\microbiome.atlas.Rproj//mergeMatUpdatedJaccard.RData"
      load(mergeMatUpdatedJaccRData) #mergeMatUpdatedJacc
      
      wellnessSamples = basicMetaMapUpdated$sample.ID[basicMetaMapUpdated$dataset.ID=="id28"]
      wellnessSubjs = basicMetaMapUpdated$metadata.ID[basicMetaMapUpdated$dataset.ID=="id28"]
      wellnessVisits = basicMetaMapUpdated$subtype[basicMetaMapUpdated$dataset.ID=="id28"]
      
      kclTwinSamples = basicMetaMapUpdated$sample.ID[basicMetaMapUpdated$dataset.ID=="id39"] 
      # intereting six subjects were connected each other
      
      us43Samples = normUniqSubjectList$US_id43
      hmpSamples = normUniqSubjectList$US_id36
      
      # intereting six subjects were connected each other
      
      t1dLuSamples = basicMetaMapUpdated$sample.ID[basicMetaMapUpdated$dataset.ID=="id35" 
                                                   & basicMetaMapUpdated$type == "Case"]
      
      contLuSamples = basicMetaMapUpdated$sample.ID[basicMetaMapUpdated$dataset.ID=="id35" 
                                                    & basicMetaMapUpdated$type == "Control"]
      
      t1dContLuSamples = basicMetaMapUpdated$sample.ID[basicMetaMapUpdated$dataset.ID=="id35"]
      
      metforminT2dSamples = basicMetaMapUpdated$sample.ID[basicMetaMapUpdated$dataset.ID=="id53" 
                                                          & basicMetaMapUpdated$subtype %in% c("M0","M2","M4")]
      
      contT2dSamples = basicMetaMapUpdated$sample.ID[basicMetaMapUpdated$dataset.ID=="id53" 
                                                     & basicMetaMapUpdated$subtype %in% c("P0","P2","P4")]
      
      metforminContT2dSamples = basicMetaMapUpdated$sample.ID[basicMetaMapUpdated$dataset.ID=="id53" 
                                                              & basicMetaMapUpdated$subtype %in% c("M0","M2","M4","P0","P2","P4")]
      
      
      
      wellnessNodeFile = "wellness.jacc.net.nodeTable.txt"
      wellnessEdgeFile = "wellness.jacc.net.edgeTable.txt"
      wellnessOut = writeJaccNetToCytoscape(wellnessSamples, mergeMatUpdatedJacc, basicMetaMapUpdated, 
                                            wellnessNodeFile, wellnessEdgeFile)
      
      kclTwinNodeFile = "kcl.twin.jacc.net.nodeTable.txt"
      kclTwinEdgeFile = "kcl.twin.jacc.net.edgeTable.txt"
      kclTwinOut = writeJaccNetToCytoscape(kclTwinSamples, mergeMatUpdatedJacc, basicMetaMapUpdated, 
                                           kclTwinNodeFile, kclTwinEdgeFile)
      
      t1dLuNodeFile = "t1dLu.jacc.net.nodeTable.txt"
      t1dLuEdgeFile = "t1dLu.jacc.net.edgeTable.txt"
      t1dLuOut = writeJaccNetToCytoscape(t1dLuSamples, mergeMatUpdatedJacc, basicMetaMapUpdated, 
                                         t1dLuNodeFile, t1dLuEdgeFile)
      
      contLuNodeFile = "contLu.jacc.net.nodeTable.txt"
      contLuEdgeFile = "contLu.jacc.net.edgeTable.txt"
      contLuOut = writeJaccNetToCytoscape(contLuSamples, mergeMatUpdatedJacc, basicMetaMapUpdated, 
                                          contLuNodeFile, contLuEdgeFile)
      
      us43NodeFile = "us43.jacc.net.nodeTable.txt"
      us43EdgeFile = "us43.jacc.net.edgeTable.txt"
      us43Out = writeJaccNetToCytoscape(us43Samples, mergeMatUpdatedJacc, basicMetaMapUpdated, 
                                        us43NodeFile, us43EdgeFile)
      
      t1dContLuNodeFile = "t1dContLu.jacc.net.nodeTable.txt"
      t1dContLuEdgeFile = "t1dContLu.jacc.net.edgeTable.txt"
      t1dContLuOut = writeJaccNetToCytoscape(t1dContLuSamples, mergeMatUpdatedJacc, basicMetaMapUpdated, 
                                             t1dContLuNodeFile, t1dContLuEdgeFile)
      
      metContT2dNodeFile = "metContT2d.jacc.net.nodeTable.txt"
      metContT2dEdgeFile = "metContT2d.jacc.net.edgeTable.txt"
      metContT2dOut = writeJaccNetToCytoscape(metforminContT2dSamples, mergeMatUpdatedJacc, basicMetaMapUpdated, 
                                              metContT2dNodeFile, metContT2dEdgeFile)
      
      
      hmpNodeFile = "hmp.jacc.net.0.5.nodeTable.txt"
      hmpEdgeFile = "hmp.jacc.net.0.5.edgeTable.txt"
      hmpOut = writeJaccNetToCytoscape(hmpSamples, mergeMatUpdatedJacc, basicMetaMapUpdated, 
                                       hmpNodeFile, hmpEdgeFile)
      
      hmpNodeFile = "hmp.jacc.net.0.33.nodeTable.txt"
      hmpEdgeFile = "hmp.jacc.net.0.33.edgeTable.txt"
      hmpOut = writeJaccNetToCytoscape(hmpSamples, mergeMatUpdatedJacc, basicMetaMapUpdated, 
                                       hmpNodeFile, hmpEdgeFile, 0.33)
      
      hmpNodeFile = "hmp.jacc.net.0.4.nodeTable.txt"
      hmpEdgeFile = "hmp.jacc.net.0.4.edgeTable.txt"
      hmpOut = writeJaccNetToCytoscape(hmpSamples, mergeMatUpdatedJacc, basicMetaMapUpdated, 
                                       hmpNodeFile, hmpEdgeFile, 0.4)
      
      japanNodeFile = "japan.jacc.net.0.33.nodeTable.txt"
      japanEdgeFile = "japan.jacc.net.0.33.edgeTable.txt"
      japanOut = writeJaccNetToCytoscape(normUniqCountryList$Japan, mergeMatUpdatedJacc, basicMetaMapUpdated, 
                                         japanNodeFile, japanEdgeFile, 0.33)
      
      antibioticsNodeFile = "antibiotics.jacc.net.0.5.nodeTable.txt"
      antibioticsEdgeFile = "antibiotics.jacc.net.0.5.edgeTable.txt"
      antibioticsOut = writeJaccNetToCytoscape(basicMetaMapUpdated$sample.ID[basicMetaMapUpdated$dataset.ID=="id51"], mergeMatUpdatedJacc, basicMetaMapUpdated, 
                                               antibioticsNodeFile, antibioticsEdgeFile, 0.5)
      
    }
    
    
    geoJaccNetMode = T
    if (geoJaccNetMode) {
      mergeMatUpdatedJaccRData = "J:\\Deposit\\Project\\2018_microbiome_atlas\\microbiome.atlas.Rproj//mergeMatUpdatedJaccard.RData"
      load(mergeMatUpdatedJaccRData) #mergeMatUpdatedJacc
      mergeMatUpdatedJaccSelected = mergeMatUpdatedJacc[normUniqAll,normUniqAll]
      
      jacc05Mode = T
      if (jacc05Mode) {
        jaccNet = makeJaccNet(mergeMatUpdatedJaccSelected, cutJacc = 0.5)
        
        geoList = normUniqSubjectList
        geoClusterScores = getClusterScores(jaccNet, geoList)
        
        jaccClusterInfo = annotateModulesByCC(jaccNet, geoList)
        writeModuleNetworkAnnot(jaccClusterInfo, 
                                "C://Data//comparative.analysis.healthy.sweden//nodeTableJaccGeo.pruned.txt",
                                "C://Data//comparative.analysis.healthy.sweden//edgeTableJaccGeo.pruned.txt")
        
        
        geoMergedList = normUniqCountryList
        geoMergedClusterScores = getClusterScores(jaccNet, geoMergedList)
        
        jaccMergedClusterInfo = annotateModulesByCC(jaccNet, geoMergedList)
        writeModuleNetworkAnnot(jaccMergedClusterInfo, 
                                "C://Data//comparative.analysis.healthy.sweden//nodeTableJaccGeoMerged.pruned.txt",
                                "C://Data//comparative.analysis.healthy.sweden//edgeTableJaccGeoMerged.pruned.txt")
        
      }
      
      jacc033Mode = F
      if (jacc033Mode) {
        jaccNet = makeJaccNet(mergeMatUpdatedJaccSelected, cutJacc = 0.33)
        
        geoList = normUniqSubjectList
        geoClusterScores = getClusterScores(jaccNet, geoList)
        
        jaccClusterInfo = annotateModulesByCC(jaccNet, geoList)
        writeModuleNetworkAnnot(jaccClusterInfo, 
                                "C://Data//comparative.analysis.healthy.sweden//nodeTableJaccGeo.pruned.0.33.txt",
                                "C://Data//comparative.analysis.healthy.sweden//edgeTableJaccGeo.pruned.0.33.txt")
        
        
        geoMergedList = normUniqCountryList
        geoMergedClusterScores = getClusterScores(jaccNet, geoMergedList)
        
        jaccMergedClusterInfo = annotateModulesByCC(jaccNet, geoMergedList)
        writeModuleNetworkAnnot(jaccMergedClusterInfo, 
                                "C://Data//comparative.analysis.healthy.sweden//nodeTableJaccGeoMerged.pruned.0.33.txt",
                                "C://Data//comparative.analysis.healthy.sweden//edgeTableJaccGeoMerged.pruned.0.33.txt")
        
      }
      
    }
    
    geoGenusJaccNetMode = T
    if (geoGenusJaccNetMode) {
      genusMatUpdatedJaccRData = "J:\\Deposit\\Project\\2018_microbiome_atlas\\microbiome.atlas.Rproj//genusMatUpdatedJaccard.RData"
      load(genusMatUpdatedJaccRData)
      genusMatUpdatedJaccSelected = genusMatUpdatedJacc[normUniqAll, normUniqAll]
      
      jacc05Mode = T
      if (jacc05Mode) {
        
        genusCut = 0.33
        jaccNet = makeJaccNet(genusMatUpdatedJaccSelected, cutJacc = genusCut)
        
        geoList = normUniqSubjectList
        geoClusterScores = getClusterScores(jaccNet, geoList)
        
        jaccClusterInfo = annotateModulesByCC(jaccNet, geoList)
        writeModuleNetworkAnnot(jaccClusterInfo, 
                                "C://Data//comparative.analysis.healthy.sweden//nodeTableJaccGenusGeo.pruned.0.33.txt",
                                "C://Data//comparative.analysis.healthy.sweden//edgeTableJaccGenusGeo.pruned.0.33.txt")
        
        
        geoMergedList = normUniqCountryList
        geoMergedClusterScores = getClusterScores(jaccNet, geoMergedList)
        
        jaccMergedClusterInfo = annotateModulesByCC(jaccNet, geoMergedList)
        writeModuleNetworkAnnot(jaccMergedClusterInfo, 
                                "C://Data//comparative.analysis.healthy.sweden//nodeTableJaccGenusGeoMerged.pruned.txt",
                                "C://Data//comparative.analysis.healthy.sweden//edgeTableJaccGenusGeoMerged.pruned.txt")
        
        
        
      }
      
    }
    
    diseaseJaccNetMode = T
    if (diseaseJaccNetMode) {
      mergeMatUpdatedJaccRData = "J:\\Deposit\\Project\\2018_microbiome_atlas\\microbiome.atlas.Rproj//mergeMatUpdatedJaccard.RData"
      load(mergeMatUpdatedJaccRData) #mergeMatUpdatedJacc
      mergeMatUpdatedJaccSelected = mergeMatUpdatedJacc[diseaseUniqAll, diseaseUniqAll]
      
      jacc05Mode = T
      if (jacc05Mode) {
        jaccNet = makeJaccNet(mergeMatUpdatedJaccSelected, cutJacc = 0.5)
        
        diseaseList = diseaseUniqSubjectList
        diseaseClusterScores = getClusterScores(jaccNet, diseaseList)
        
        jaccDiseaseClusterInfo = annotateModulesByCC(jaccNet, diseaseList)
        writeModuleNetworkAnnot(jaccDiseaseClusterInfo, 
                                "C://Data//comparative.analysis.healthy.sweden//nodeTableJaccDisease.pruned2.txt",
                                "C://Data//comparative.analysis.healthy.sweden//edgeTableJaccDisease.pruned2.txt")
        
      }
      jacc033Mode = F
      if (jacc033Mode) {
        jaccNet = makeJaccNet(mergeMatUpdatedJaccSelected, cutJacc = 0.33)
        
        diseaseList = diseaseUniqSubjectList
        diseaseClusterScores = getClusterScores(jaccNet, diseaseList)
        
        jaccDiseaseClusterInfo = annotateModulesByCC(jaccNet, diseaseList)
        writeModuleNetworkAnnot(jaccDiseaseClusterInfo, 
                                "C://Data//comparative.analysis.healthy.sweden//nodeTableJaccDisease.pruned.0.33.txt",
                                "C://Data//comparative.analysis.healthy.sweden//edgeTableJaccDisease.pruned.0.33.txt")
        
      }
      
      
    }
    
    mixedJaccNetMode = F
    if (mixedJaccNetMode) {
      
      mergeMatUpdatedJaccRData = "J:\\Deposit\\Project\\2018_microbiome_atlas\\microbiome.atlas.Rproj//mergeMatUpdatedJaccard.RData"
      load(mergeMatUpdatedJaccRData) #mergeMatUpdatedJacc
      mergeMatUpdatedJaccSelected = mergeMatUpdatedJacc[mixedUniqAll, mixedUniqAll]
      
      jacc05Mode = T
      if (jacc05Mode) {
        jaccNet = makeJaccNet(mergeMatUpdatedJaccSelected, cutJacc = 0.5)
        
        mixedList = mixedUniqSubjectList
        mixedClusterScores = getClusterScores(jaccNet, mixedList)
        
        jaccDiseaseClusterInfo = annotateModulesByCC(jaccNet, mixedList)
        writeModuleNetworkAnnot(jaccDiseaseClusterInfo, 
                                "C://Data//comparative.analysis.healthy.sweden//nodeTableJaccMixed.pruned.txt",
                                "C://Data//comparative.analysis.healthy.sweden//edgeTableJaccMixed.pruned.txt")
        
      }
      
      jacc033Mode = T
      if (jacc033Mode) {
        jaccNet = makeJaccNet(mergeMatUpdatedJaccSelected, cutJacc = 0.4)
        
        mixedList = mixedUniqSubjectList
        mixedClusterScores = getClusterScores(jaccNet, mixedList)
        
        jaccDiseaseClusterInfo = annotateModulesByCC(jaccNet, mixedList)
        writeModuleNetworkAnnot(jaccDiseaseClusterInfo, 
                                "C://Data//comparative.analysis.healthy.sweden//nodeTableJaccMixed.pruned.0.4.txt",
                                "C://Data//comparative.analysis.healthy.sweden//edgeTableJaccMixed.pruned.0.4.txt")
        
        
      }
      
    }
  }
  
  
}

findFunctionCluster = F
if (findFunctionCluster) {
  
  loadFuncAnnot = T
  if (loadFuncAnnot) {
    
    gemFile = "J://Deposit/Project/2018_microbiome_atlas//atlas.GEM.model.related/GEM.from.Reza.amazing/absentPresentInModels.txt"
    gemRData = "J://Deposit/Project/2018_microbiome_atlas//atlas.GEM.model.related/GEM.from.Reza.amazing/absentPresentInModels.RData"
    
    if (F) {
      gemTab = read.delim(gemFile, sep="\t", stringsAsFactors = F, header = T)
      gemMat = as.matrix(gemTab[,-1])
      rownames(gemMat) = gemTab[,1]
      gemMat = t(gemMat)
      save(gemMat, file=gemRData)
    }
    
    antismashMatRData = "J://Deposit/Project/2018_microbiome_atlas/functional.annotation/msp1992.igc2.antismashMat.RData"
    vfMatRData = "j://Deposit/Project/2018_microbiome_atlas/functional.annotation/PATRIC/vfBestMat.RData"
    mustardMatRData = "j://Deposit/Project/2018_microbiome_atlas/functional.annotation/mustard.mat.RData"
    igc2PfamMatRData = "j://Deposit/Project/2018_microbiome_atlas/functional.annotation/PFAM/igc2PfamMat.RData"
    newGutCazyMatRData = "J://Deposit/Project/2018_microbiome_atlas/functional.annotation/CAZY.INRA/igc2.new.cazy.mat.RData"
    gutKoBestMatRData = "j://Deposit/Project/2018_microbiome_atlas/functional.annotation/gutKoBestMat.1990.RData"
    jgiMatRData = "j://Deposit/Project/2018_microbiome_atlas/functional.annotation/jgiMat.RData"
    funcMatRData = "j://Deposit/Project/2018_microbiome_atlas/functional.annotation/funcMat.20190808.RData"
    funcGemMatRData = "j://Deposit/Project/2018_microbiome_atlas/functional.annotation/funcGemMat.20190820.RData"
    
    igc2PfamDescRData = "j://Deposit/Project/2018_microbiome_atlas/functional.annotation/PFAM/igc2PfamDesc.RData"
    
    patricVfDescTabFile = "J://Deposit/Project/2018_microbiome_atlas/functional.annotation/PATRIC/PATRIC_sp_gene.txt"
    patricVfDescTab = read.table(patricVfDescTabFile, header=T, sep="\t", stringsAsFactors = F)
    gut1992MspVfBestDescMapRData = "j://Deposit/Project/2018_microbiome_atlas/igc2/gut1992MspVfBestDescMap.RData"
    
    keggModuleDescMatFile = "J://Deposit/Project/2018_microbiome_atlas/functional.annotation/KEGG module/moduleDescTab.RData"
    koModuleListRData = "J://Deposit/Project/2018_microbiome_atlas/functional.annotation/KEGG module/koModuleList.20190219.RData"
    
    koJaccMatRData = "J://Deposit/Project/2018_microbiome_atlas//functional.annotation//koJaccMat.RData"
    cazyJaccMatRData = "J://Deposit/Project/2018_microbiome_atlas//functional.annotation//cazyJaccMat.RData"
    pfamJaccMatRData = "J://Deposit/Project/2018_microbiome_atlas//functional.annotation//pfamJaccMat.RData"
    funcJaccMatRData = "J://Deposit/Project/2018_microbiome_atlas//functional.annotation//funcJaccMat.20191227.RData"
    funcGemJaccMatRData = "J://Deposit/Project/2018_microbiome_atlas//functional.annotation//funcGemJaccMat.20190820.RData"
    
    load(keggModuleDescMatFile)
    microbeModuleTab = moduleDescTab[moduleDescTab$isBacteriaModule|moduleDescTab$isArchaeaModule,]
    nonMicrobeModuleTab = moduleDescTab[!(moduleDescTab$isBacteriaModule|moduleDescTab$isArchaeaModule),]
    
    microbeModuleNames =  paste(microbeModuleTab$moduleId, microbeModuleTab$moduleName, sep=":")
    nonMicrobeModuleNames =  paste(nonMicrobeModuleTab$moduleId, nonMicrobeModuleTab$moduleName, sep=":")
    
    load(koModuleListRData)
    load(igc2PfamDescRData)
    uniqPfamNames = unique(igc2PfamDesc$pfam_name)
    uniqPfamDesc = unique(igc2PfamDesc$pfam_desc)
    compDescs = c(uniqPfamDesc[grep("competence",uniqPfamDesc)], uniqPfamDesc[grep("Competence",uniqPfamDesc)])
    compPfamNames = unique(igc2PfamDesc$pfam_name[igc2PfamDesc$pfam_desc %in% compDescs])
    
    
    load(gut1992MspVfBestDescMapRData)
    
    load(igc2PfamMatRData) #pfamMat
    load(newGutCazyMatRData) #cazyMat
    load(gutKoBestMatRData) #gutKoBestMat
    
    forNeeluMode = T
    if (forNeeluMode) {
      pfamMat = pfamMat[!rownames(pfamMat) %in% suppressMsps2,]
      pfamMelt = melt(pfamMat)
      pfamMelt = pfamMelt[pfamMelt$value==1,]
      pfamMelt = pfamMelt[,1:2]
      pfamMelt$Var1  = as.character(pfamMelt$Var1)
      pfamMelt$Var2  = as.character(pfamMelt$Var2)
      pfamMelt$species = getSpeciesName(pfamMelt$Var1, taxo)
      gutPfamMeltFile="C:\\Data\\comparative.analysis.healthy.sweden\\gutPfamMelt.for.neelu.txt"
      write.table(pfamMelt, gutPfamMeltFile, sep="\t", quote=F, row.names = F)
      
      igc2PfamDesc$species = getSpeciesName(igc2PfamDesc$msp, taxo)
      gutPfamDescFile = "C:\\Data\\comparative.analysis.healthy.sweden\\gutPfamDesc.for.neelu.txt"
      write.table(igc2PfamDesc, gutPfamDescFile, sep="\t", quote=F, row.names = F)
      
    }
    
    if (F) {
      
      cazyMelt = melt(cazyMat)
      cazyMelt = cazyMelt[cazyMelt$value!=0,]
      cazyMelt$species = getSpeciesName(as.character(cazyMelt$Var1), taxo)
      colnames(cazyMelt)[1:2] = c("MSP_ID","CAZY")
      cazyMelt = cazyMelt[!cazyMelt$MSP_ID %in% suppressMsps2, ]
      write.table(cazyMelt[,-3], 
                  "C://Data//comparative.analysis.healthy.sweden//gut.cazy.tab.for.neelu.txt",
                  row.names = F, quote = F,
                  sep="\t")
    }
    
    load(antismashMatRData) #antismashMat
    load(vfMatRData) #vfMatDigit
    load(mustardMatRData) #mustardMat --> antibiotic resistance
    load(jgiMatRData) #jgiMat
    
    checkPathMode = F
    if (checkPathMode) {
      load(koJaccMatRData)
      allKos = rownames(koJaccMat)
      
      quantile(koJaccMat[lower.tri(koJaccMat)], 0.5) #0
      quantile(koJaccMat[lower.tri(koJaccMat)], 0.90) #0.1111111
      quantile(koJaccMat[lower.tri(koJaccMat)], 0.95) #0.2377622
      quantile(koJaccMat[lower.tri(koJaccMat)], 0.99) #0.5909091 
      
      moduleLevelMode = T
      if (moduleLevelMode) {
        
        countKoModuleList <- do.call(c,lapply(koModuleList, function(kos){
          kos = kos[kos %in% allKos]
          return(length(kos))
        }))
        
        fractionKoModuleList <- do.call(c,lapply(koModuleList, function(kos){
          total = length(kos)
          kos = kos[kos %in% allKos]
          remained = length(kos)
          return(remained/total)
        }))
        
        koModuleCatalogStat = data.frame(total=do.call(c,lapply(koModuleList,length)),
                                         number=countKoModuleList, 
                                         fraction=fractionKoModuleList)
        
        nonSmallModules = rownames(koModuleCatalogStat[koModuleCatalogStat$total>=5,])
        
        coveredKoModules = rownames(koModuleCatalogStat[koModuleCatalogStat$total>=5&
                                                          koModuleCatalogStat$fraction>=0.5,])
        
        microbeModuleCoveredNames = microbeModuleNames[microbeModuleNames %in% nonSmallModules]
        nonMicrobeModuleCoveredNames = nonMicrobeModuleNames[nonMicrobeModuleNames %in% nonSmallModules]
        
        jaccKoModuleList <- lapply(koModuleList, function(kos){
          kos = kos[kos %in% allKos]
          if (length(kos)==0) return(NA)
          if (length(kos)==1) return(NA)
          currJaccMat = koJaccMat[kos,kos]
          return(currJaccMat[lower.tri(currJaccMat)])
        })
        
        jaccMeanKoModules = do.call(c,lapply(jaccKoModuleList, mean))
        koBoxList = list(microbe= jaccMeanKoModules[names(jaccMeanKoModules)%in% microbeModuleNames],
                         nonMicrobe= jaccMeanKoModules[names(jaccMeanKoModules) %in% nonMicrobeModuleNames])
        boxplot(rev(koBoxList),horizontal = T,las=2)
        
        koBoxList = list(microbe= jaccMeanKoModules[names(jaccMeanKoModules)%in% microbeModuleCoveredNames],
                         nonMicrobe= jaccMeanKoModules[names(jaccMeanKoModules) %in% nonMicrobeModuleCoveredNames])
        boxplot(rev(koBoxList),horizontal = T,las=2)
        
        jaccKoStats = 3
        
      }
      
      pathLevelMode = T
      if (pathLevelMode) {
        
      }
      
    }
    
    generateMode = F
    if (generateMode) {
      koJaccMat =  getJaccMat(gutKoBestMat)
      save(koJaccMat, file = koJaccMatRData)
      
      cazyJaccMat =  getJaccMat(cazyMat)
      save(cazyJaccMat, file = cazyJaccMatRData)
      
      pfamJaccMat =  getJaccMat(pfamMat)
      pfamJaccMat = pfamJaccMat[!rownames(pfamJaccMat) %in% suppressMsps,]
      save(pfamJaccMat, file = pfamJaccMatRData)
      
      colnames(cazyMat) = paste("cazy", colnames(cazyMat), sep=".")
      
      funcMat = mustardMat
      funcMat = mergeMatrix(funcMat, cazyMat)
      funcMat = mergeMatrix(funcMat, antismashMat)
      funcMat = mergeMatrix(funcMat, vfMatDigit)
      funcMat = mergeMatrix(funcMat, gutKoBestMat)
      funcMat = mergeMatrix(funcMat, pfamMat)
      funcMat = mergeMatrix(funcMat, jgiMat)
      funcMat = funcMat[!rownames(funcMat) %in% suppressMsps,]
      
      save(funcMat, file=funcMatRData)
      
      
      gemSpecies = rownames(gemMat)
      table(gemSpecies %in% rownames(funcMat))
      
      funcGemMat = funcMat[rownames(funcMat) %in% gemSpecies,]
      funcGemMat = mergeMatrix(funcGemMat, gemMat)
      save(funcGemMat, file=funcGemMatRData)
      
      funcJaccMat =  getJaccMat(funcMat)
      save(funcJaccMat, file = funcJaccMatRData)
      
    }
    
  }
  
  loadFuncJaccData = F
  if (loadFuncJaccData) {
    load(funcJaccMatRData)
    funcJaccTabSelRData = "J://Deposit/Project/2018_microbiome_atlas//functional.annotation//funcJaccTabSel.20190815.RData"
    load(funcJaccTabSelRData)
    View(funcJaccTabSel[is.na(funcJaccTabSel$Cat2),])
  }
  
  checkFuncJaccMat = F
  if (checkFuncJaccMat) {
    load(funcJaccMatRData)
    funcJaccNet = makeJaccNet(funcJaccMat, cutJacc = 0.9)
    funcJaccModList = makeModuleList(funcJaccNet)
    #annotateModulesByCC
    writeModuleNetworkAnnot(jaccClusterInfo, 
                            "C://Data//comparative.analysis.healthy.sweden//nodeTableJaccGeo.pruned.txt",
                            "C://Data//comparative.analysis.healthy.sweden//edgeTableJaccGeo.pruned.txt")
    
    
    
    makeTabMode = F
    if (makeTabMode) {
      require(reshape2)
      funcJaccTab = melt(funcJaccMat)
      jaccCut = 0.5
      funcJaccTabSel = funcJaccTab[funcJaccTab$value>=jaccCut,]
      funcJaccTabSel = funcJaccTabSel[funcJaccTabSel$Var1 != funcJaccTabSel$Var2,]
      funcJaccTabSel$Cat1 = NA
      funcJaccTabSel$Cat2 = NA
      funcJaccTabSel$Cat1[funcJaccTabSel$Var1 %in% colnames(pfamMat)] = "Pfam"
      funcJaccTabSel$Cat2[funcJaccTabSel$Var2 %in% colnames(pfamMat)] = "Pfam"
      funcJaccTabSel$Cat1[funcJaccTabSel$Var1 %in% colnames(cazyMat)] = "Cazy"
      funcJaccTabSel$Cat2[funcJaccTabSel$Var2 %in% colnames(cazyMat)] = "Cazy"
      funcJaccTabSel$Cat1[funcJaccTabSel$Var1 %in% colnames(gutKoBestMat)] = "KO"
      funcJaccTabSel$Cat2[funcJaccTabSel$Var2 %in% colnames(gutKoBestMat)] = "KO"
      funcJaccTabSel$Cat1[funcJaccTabSel$Var1 %in% colnames(antismashMat)] = "antiSMASH"
      funcJaccTabSel$Cat2[funcJaccTabSel$Var2 %in% colnames(antismashMat)] = "antiSMASH"
      funcJaccTabSel$Cat1[funcJaccTabSel$Var1 %in% colnames(vfMatDigit)] = "virulence"
      funcJaccTabSel$Cat2[funcJaccTabSel$Var2 %in% colnames(vfMatDigit)] = "virulence"
      funcJaccTabSel$Cat1[funcJaccTabSel$Var1 %in% colnames(mustardMat)] = "AR"
      funcJaccTabSel$Cat2[funcJaccTabSel$Var2 %in% colnames(mustardMat)] = "AR"
      funcJaccTabSel$Cat1[funcJaccTabSel$Var1 %in% colnames(jgiMat)] = "Phenotype"
      funcJaccTabSel$Cat2[funcJaccTabSel$Var2 %in% colnames(jgiMat)] = "Phenotype"
      funcJaccTabSel = funcJaccTabSel[!is.na(funcJaccTabSel$value),]
      
      funcJaccTabSelRData = "J://Deposit/Project/2018_microbiome_atlas//functional.annotation//funcJaccTabSel.20190815.RData"
      save(funcJaccTabSel, file=funcJaccTabSelRData)
      
    }
    
  }
  
}

makeFunctionCluster = F ### performed on UPPMAX rackham cluster
if (makeFunctionCluster) {
  
  source("~/rscripts/2019_funcType/funcType.analysis.functions.r")
  source("~/rscripts/2019_funcType/funcType.data.files.r")
  
  funcGemJaccNetMode = T
  if (funcGemJaccNetMode) {
    funcGemModNetNodeTableFile = "~/sunjae.mAtlas.func.annotation/funcGemModeNet.nodeTable.20190828.txt"
    funcGemModNetEdgeTableFile = "~/sunjae.mAtlas.func.annotation/funcGemModeNet.edgeTable.20190828.txt"
    
    load(funcGemJaccMatRData)
    funcGemJaccNet = makeJaccNet(funcGemJaccMat)
    save(funcGemJaccNet, file = funcGemJaccNetRData)
    
    funcGemModules = makeModuleList(funcGemJaccNet)
    save(funcGemModules, file = funcGemModulesRData)
    
    funcGemModuleAnnots = annotateModulesByCC(funcGemJaccNet, funcGemModules, 3)
    writeModuleNetworkAnnot(funcGemModuleAnnots, funcGemModNetNodeTableFile, funcGemModNetEdgeTableFile)
    save(funcGemModuleAnnots, file=funcGemModuleAnnotsRData)
    
  }
  
  rm(list=ls())
  
  source("~/rscripts/2019_funcType/funcType.analysis.functions.r")
  source("~/rscripts/2019_funcType/funcType.data.files.r")
  
  funcJaccNetMode = T
  if (funcJaccNetMode) {
    funcModNetNodeTableFile = "~/sunjae.mAtlas.func.annotation/funcModeNet.nodeTable.20190828.txt"
    funcModNetEdgeTableFile = "~/sunjae.mAtlas.func.annotation/funcModeNet.edgeTable.20190828.txt"
    
    load(funcJaccMatRData)
    funcJaccNet = makeJaccNet(funcJaccMat)
    save(funcJaccNet, file = funcJaccNetRData)
    
    funcModules = makeModuleList(funcJaccNet)
    save(funcModules, file = funcModulesRData)
    
    funcModuleAnnots = annotateModulesByCC(funcJaccNet, funcModules, 3)
    writeModuleNetworkAnnot(funcModuleAnnots, funcModNetNodeTableFile, funcModNetEdgeTableFile)
    save(funcModuleAnnots, file=funcModuleAnnotsRData)
    
  }
  
}

mainFig1PiePlotMode = F
if (mainFig1PiePlotMode) {
  
  diseaseMgsMat = mergeMatUpdated[,diseaseUniqAll] 
  normalMgsMat = mergeMatUpdated[,normUniqAll]
  industrialMgsMat = mergeMatUpdated[,industrializedSamples]
  traditionalMgsMat = mergeMatUpdated[,traditionalSamples]
  
  familyMode = T
  if (familyMode) {
    normalFamilyRelAbd = getRelAbdOfTaxa(industrialMgsMat, "family", taxo)
    diseaseFamilyRelAbd = getRelAbdOfTaxa(diseaseMgsMat, "family", taxo)
    indiFamilyRelAbd = getRelAbdOfTaxa(traditionalMgsMat, "family", taxo)
    
    normalFamilySortedNames = names(sort(normalFamilyRelAbd,decreasing = T))
    normalFamilySortedNames = normalFamilySortedNames[!grepl("unclassified",normalFamilySortedNames)]
    diseaseFamilySortedNames = names(sort(diseaseFamilyRelAbd,decreasing = T))
    diseaseFamilySortedNames = diseaseFamilySortedNames[!grepl("unclassified",diseaseFamilySortedNames)]
    indiFamilySortedNames = names(sort(indiFamilyRelAbd,decreasing = T))
    indiFamilySortedNames = indiFamilySortedNames[!grepl("unclassified",indiFamilySortedNames)]
    
    normalTop20 = normalFamilySortedNames[1:20]
    diseaseTop20 = diseaseFamilySortedNames[1:20]
    indiTop20 = indiFamilySortedNames[1:20]
    
    mergedTop20 = unique(c(normalTop20, diseaseTop20, indiTop20))
    
    normalFamilyTop20Imputed = getRelAbdWithOthersImputed(normalFamilyRelAbd, mergedTop20)
    diseaseFamilyTop20Imputed = getRelAbdWithOthersImputed(diseaseFamilyRelAbd, mergedTop20)
    indiFamilyTop20Imputed = getRelAbdWithOthersImputed(indiFamilyRelAbd, mergedTop20)
    
    mergedList = names(sort(normalFamilyTop20Imputed,decreasing = T))
    lenMerged = length(mergedList)
    
    normalFamilyTop20Imputed = normalFamilyTop20Imputed[match(mergedList, names(normalFamilyTop20Imputed))]
    diseaseFamilyTop20Imputed = diseaseFamilyTop20Imputed[match(mergedList, names(diseaseFamilyTop20Imputed))]
    indiFamilyTop20Imputed = indiFamilyTop20Imputed[match(mergedList, names(indiFamilyTop20Imputed))]
    
    normalFamilyTop20Imputed = c(normalFamilyTop20Imputed[names(normalFamilyTop20Imputed)!="others"],
                                 normalFamilyTop20Imputed[names(normalFamilyTop20Imputed)=="others"])
    diseaseFamilyTop20Imputed = c(diseaseFamilyTop20Imputed[names(diseaseFamilyTop20Imputed)!="others"],
                                  diseaseFamilyTop20Imputed[names(diseaseFamilyTop20Imputed)=="others"])
    indiFamilyTop20Imputed = c(indiFamilyTop20Imputed[names(indiFamilyTop20Imputed)!="others"],
                               indiFamilyTop20Imputed[names(indiFamilyTop20Imputed)=="others"])
    
    piePlotMode = T
    if (piePlotMode) {
      col_vector = unlist(mapply(brewer.pal, qual_col_pals$maxcolors, rownames(qual_col_pals)))
      familyCols = c(col_vector[1:(lenMerged-1)],"lightgray")
      familyCols = col_vector[1:(lenMerged)] 
      
      diseaseTarget = diseaseFamilyTop20Imputed
      normalTarget = normalFamilyTop20Imputed
      indiTarget = indiFamilyTop20Imputed
      
      iniR = 0.2
      pie(1, radius=0.01, init.angle=90, col=c('white'), border = NA, labels='')
      floating.pie(0,0,diseaseTarget ,
                   radius=5*iniR, 
                   startpos=pi/2, 
                   col = familyCols,
                   border="black")
      
      floating.pie(0,0,normalTarget ,  
                   radius=4*iniR, 
                   startpos=pi/2, 
                   col = familyCols,
                   border="black")
      
      floating.pie(0,0,indiTarget,  
                   radius=3*iniR, 
                   startpos=pi/2, 
                   col = familyCols,
                   border="black")
      floating.pie(0,0,1,  
                   radius=2*iniR, 
                   startpos=pi/2, 
                   col="white",
                   border="black")
      floating.pie(0,0,1,  
                   radius=1.99*iniR, 
                   startpos=pi/2, 
                   col="white",
                   border=NA)
      
      
    }
  }
}

pcoaCheckMode = F
if (pcoaCheckMode) {
  
  pcoaEnteroCheck = T
  if (pcoaEnteroCheck) {
    
  }
  
  simpleCheck = F
  if (simpleCheck) {
    pcoaGenusMat= pcoaGenusOut$vectors[,1:2]
    
    pcoaCols = rep("#c0c0c033", dim(pcoaGenusMat)[1])
    names(pcoaCols) = rownames(pcoaGenusMat)
    
    barplot(c(1,2),col = c("#00bfff","#ff2400"))
    
    pcoaCols[normUniqCountryList$Thailand] = "blue"
    pcoaCols[thaiShort] = "#00bfff"
    pcoaCols[thaiLong] = "#ff2400"
    plot(pcoaGenusMat[,1], pcoaGenusMat[,2], col=pcoaCols, pch=16, xaxt="n", yaxt="n", xlab="", ylab="") #, xlim=c(-0.55,0.4), ylim=c(-0.3,0.7)
    
    pcoaCols = rep("#c0c0c033", dim(pcoaGenusMat)[1])
    names(pcoaCols) = rownames(pcoaGenusMat)
    
    pcoaCols[normUniqCountryList$Thailand] = "blue"
    pcoaCols[thaiShort] = "#00bfff"
    pcoaCols[thaiLong] = "#ff2400"
    plot(pcoaGenusMat[,1], pcoaGenusMat[,2], col=pcoaCols, pch=16, xaxt="n", yaxt="n", xlab="", ylab="") #, xlim=c(-0.55,0.4), ylim=c(-0.3,0.7)
    
    beforeAntibiotics=basicMetaMapUpdated$sample.ID[basicMetaMapUpdated$dataset.ID=="id51" & basicMetaMapUpdated$subtype=="Dag0"]
    afterAntibiotics=basicMetaMapUpdated$sample.ID[basicMetaMapUpdated$dataset.ID=="id51" & basicMetaMapUpdated$subtype=="Dag4"]
    afterAntibiotics2=basicMetaMapUpdated$sample.ID[basicMetaMapUpdated$dataset.ID=="id51" & basicMetaMapUpdated$subtype=="Dag8"]
    afterAntibiotics3=basicMetaMapUpdated$sample.ID[basicMetaMapUpdated$dataset.ID=="id51" & basicMetaMapUpdated$subtype=="Dag180"]
    
    pcoaCols = rep("#c0c0c033", dim(pcoaGenusMat)[1])
    names(pcoaCols) = rownames(pcoaGenusMat)
    
    pcoaCols[afterAntibiotics] = "blue"
    pcoaCols[beforeAntibiotics] = "#00bfff"
    pcoaCols[afterAntibiotics2] = "purple"
    pcoaCols[afterAntibiotics3] = "black"
    plot(pcoaGenusMat[,1], pcoaGenusMat[,2], col=pcoaCols, pch=16, xaxt="n", yaxt="n", xlab="", ylab="") #, xlim=c(-0.55,0.4), ylim=c(-0.3,0.7)
    
    poisonedSamples =  basicMetaMapUpdated$sample.ID[basicMetaMapUpdated$dataset.ID=="id48" ]
    
    pcoaCols = rep("#c0c0c033", dim(pcoaGenusMat)[1])
    names(pcoaCols) = rownames(pcoaGenusMat)
    
    pcoaCols[poisonedSamples] = "blue"
    pcoaCols[normUniqCountryList$US] = "#00bfff"
    
    plot(pcoaGenusMat[,1], pcoaGenusMat[,2], col=pcoaCols, pch=16, xaxt="n", yaxt="n", xlab="", ylab="") #, xlim=c(-0.55,0.4), ylim=c(-0.3,0.7)
    
    
    beforeMetformin= basicMetaMapUpdated$sample.ID[basicMetaMapUpdated$dataset.ID=="id53" & basicMetaMapUpdated$subtype=="M0"]
    afterMetformin= basicMetaMapUpdated$sample.ID[basicMetaMapUpdated$dataset.ID=="id53" & basicMetaMapUpdated$subtype=="M4"]
    
    pcoaCols = rep("#c0c0c033", dim(pcoaGenusMat)[1])
    names(pcoaCols) = rownames(pcoaGenusMat)
    
    pcoaCols[afterMetformin] = "blue"
    pcoaCols[beforeMetformin] = "#00bfff"
    plot(pcoaGenusMat[,1], pcoaGenusMat[,2], col=pcoaCols, pch=16, xaxt="n", yaxt="n", xlab="", ylab="") #, xlim=c(-0.55,0.4), ylim=c(-0.3,0.7)
    
    renalBefore = basicMetaMapUpdated$sample.ID[basicMetaMapUpdated$dataset.ID=="id41" & 
                                                  basicMetaMapUpdated$type=="Case" &
                                                  basicMetaMapUpdated$subtype=="before"]
    renalAfter = basicMetaMapUpdated$sample.ID[basicMetaMapUpdated$dataset.ID=="id41" & 
                                                 basicMetaMapUpdated$type=="Case" &
                                                 basicMetaMapUpdated$subtype=="after"]
    pcoaCols = rep("#c0c0c033", dim(pcoaGenusMat)[1])
    names(pcoaCols) = rownames(pcoaGenusMat)
    
    pcoaCols[renalAfter] = "blue"
    pcoaCols[renalBefore] = "#00bfff"
    plot(pcoaGenusMat[,1], pcoaGenusMat[,2], col=pcoaCols, pch=16, xaxt="n", yaxt="n", xlab="", ylab="") #, xlim=c(-0.55,0.4), ylim=c(-0.3,0.7)
    
    lungBefore = basicMetaMapUpdated$sample.ID[basicMetaMapUpdated$dataset.ID=="id26" & 
                                                 basicMetaMapUpdated$type=="Case" &
                                                 basicMetaMapUpdated$subtype=="Before"]
    lungAfter = basicMetaMapUpdated$sample.ID[basicMetaMapUpdated$dataset.ID=="id26" & 
                                                basicMetaMapUpdated$type=="Case" &
                                                basicMetaMapUpdated$subtype=="after"]
    pcoaCols = rep("#c0c0c022", dim(pcoaGenusMat)[1])
    names(pcoaCols) = rownames(pcoaGenusMat)
    
    pcoaCols[lungAfter] = "blue"
    pcoaCols[lungBefore] = "#00bfff"
    plot(pcoaGenusMat[,1], pcoaGenusMat[,2], col=pcoaCols, pch=16, xaxt="n", yaxt="n", xlab="", ylab="") #, xlim=c(-0.55,0.4), ylim=c(-0.3,0.7)
    
  }
  
}

updateResidualRichness = F
if (updateResidualRichness) {
  
  noImputeMetaMapUpdated = basicMetaMapUpdated[!is.na(basicMetaMapUpdated$Age) & !is.na(basicMetaMapUpdated$BMI),]
  fitGeneRichnessUpdated = lmer(GeneRichness ~ Age + BMI + (1|Sequencer), data=noImputeMetaMapUpdated, na.action = na.exclude)
  noImputeMetaMapUpdated$ResidualRichness = noImputeMetaMapUpdated$GeneRichness - predict(fitGeneRichnessUpdated)
  
  basicMetaMapUpdated$ResidualRichness = NA
  basicMetaMapUpdated$ResidualRichness[match(noImputeMetaMapUpdated$sample.ID, basicMetaMapUpdated$sample.ID)] = noImputeMetaMapUpdated$ResidualRichness
  
  basicMetaCleanRDataUpdated2 = "C:\\Data/comparative.analysis.healthy.sweden/all.basic.clean.metadata.behcet.20190805.RData"
  save(basicMetaMapUpdated, file=basicMetaCleanRDataUpdated2)
  
}

checkMgsStat = F
if (checkMgsStat) {
  
  out = apply(mergeMatUpdated, 2, function(currCol){
    names(currCol) = rownames(mergeMatUpdated)
    currCol = sort(currCol, decreasing = T)
    return(names(currCol[1:10]))
  })
  outVec = as.vector(out)
  uniqTop10Species = unique(outVec)
  
  topSpeciesMergeMat = mergeMatUpdated[uniqTop10Species,]
  #corMat = cor(topSpeciesMergeMat)
  avgMixedTopMat = getAvgAbdMatBy(topSpeciesMergeMat, mixedUniqSubjectList)
  corMxiedTopMat = cor(avgMixedTopMat)
  heatmap.2(corMxiedTopMat, trace="none",
            col = colorRampPalette(c("white","red"))(256),
            margins = c(7,7),key=F)
  
  set.seed(1)
  targetMat = avgMixedTopMat
  targetTsneOut = Rtsne(t(targetMat), dims=2, perplexity = 5)
  rownames(targetTsneOut$Y) = colnames(targetMat)
  plot(targetTsneOut$Y)
  
  tsneTab = data.frame(targetTsneOut$Y, type=rownames(targetTsneOut$Y))
  ggplot(tsneTab, aes(x=X1 , y=X2, label=type)) + 
    geom_point(aes(colour = type, alpha=.02)) + 
    theme(panel.background = element_blank(),
          axis.ticks = element_blank(),
          axis.text.x = element_blank(),
          strip.text.x = element_blank(),
          axis.text.y = element_blank(),
          strip.text.y = element_blank(),
          legend.position="none", 
          panel.border = element_rect(colour = "black", fill=NA, size=.5))  + 
    geom_text_repel(size=3, force=2) + xlab("") + ylab("")
  
  # ,cexCol = 0.5,cexRow = 0.5,
  # sepcolor = "gray",
  # colsep=seq(0, dim(diseaseVipMat)[2]),
  # rowsep=seq(0, dim(diseaseVipMat)[1]),
  # sepwidth = c(0.01,0.01)
}

checkJaccForCooccurence = F
if (checkJaccForCooccurence) {
  
  etfCountriesUpdated = c("UK","Germany","Italy","Spain","Sweden","Denmark","Finland","Luxembourg","France")
  etpCountriesUpdated = c("Peru","Mongo","Thai","Fiji","Madagascar","Tanzania","India") #,"Fiji","India"
  etbCountriesUpdated = c("China","Japan","US")
  
  etfSamples = do.call(c,normalSamplesByCountries[etfCountriesUpdated])
  etpSamples = do.call(c,normalSamplesByCountries[etpCountriesUpdated])
  etbSamples = do.call(c,normalSamplesByCountries[etbCountriesUpdated])
  
  etfMat = mergeMatUpdated[,etfSamples]
  etpMat = mergeMatUpdated[,etpSamples]
  etbMat = mergeMatUpdated[,etbSamples]
  etfMat = etfMat[rowMeans(etfMat)!=0,]
  etpMat = etpMat[rowMeans(etpMat)!=0,]
  etbMat = etbMat[rowMeans(etbMat)!=0,]
  
  industrialMat = mergeMatUpdated[,newWesternIds]
  industrialMat = industrialMat[rowMeans(industrialMat)!=0,]
  traditionalMat = mergeMatUpdated[,newNonWesternIds]
  traditionalMat = traditionalMat[rowMeans(traditionalMat)!=0,]
  diseaseMat = mergeMatUpdated[,newDiseaseIds]
  diseaseMat = diseaseMat[rowMeans(diseaseMat)!=0,]
  
  mgsJaccMat = 1-jaccard(t(mergeMatUpdated>0))
  mgsJaccIndustrialMat = 1-jaccard(t(industrialMat>0))
  mgsJaccTraditionalMat = 1-jaccard(t(traditionalMat>0))
  mgsJaccDiseaseMat = 1-jaccard(t(diseaseMat>0))
  mgsJaccEtFMat = 1-jaccard(t(etfMat>0))
  mgsJaccEtBMat = 1-jaccard(t(etbMat>0))
  mgsJaccEtPMat = 1-jaccard(t(etpMat>0))
  
  mgsJaccNet = makeJaccNet(mgsJaccMat, cutJacc = 0.75)
  mgsJaccIndustrialNet = makeJaccNet(mgsJaccIndustrialMat, cutJacc = 0.75)
  mgsJaccTraditionalNet = makeJaccNet(mgsJaccTraditionalMat, cutJacc = 0.75)
  mgsJaccDiseaseNet = makeJaccNet(mgsJaccDiseaseMat, cutJacc = 0.75)
  
  
  #### all samples ####
  mgsJaccModule = makeModuleList(mgsJaccNet)
  mgsJaccAnnot = annotateModulesByCC(mgsJaccNet, mgsJaccModule, cutCluster = 3)
  
  writeModuleNetworkAnnot(mgsJaccAnnot, 
                          "mgs.jacc.0.75.nodeTable.txt",
                          "mgs.jacc.0.75.edgeTable.txt")
  
  #### industrial samples ####
  mgsJaccIndustrialModule = makeModuleList(mgsJaccIndustrialNet)
  mgsJaccIndustrialAnnot = annotateModulesByCC(mgsJaccIndustrialNet, mgsJaccIndustrialModule, cutCluster = 3)
  
  writeModuleNetworkAnnot(mgsJaccIndustrialAnnot, 
                          "mgs.jacc.0.75.industrial.nodeTable.txt",
                          "mgs.jacc.0.75.industrial.edgeTable.txt")
  
  #### traditional samples ####
  mgsJaccTraditionalModule = makeModuleList(mgsJaccTraditionalNet)
  mgsJaccTraditionalAnnot = annotateModulesByCC(mgsJaccTraditionalNet, mgsJaccTraditionalModule, cutCluster = 3)
  
  writeModuleNetworkAnnot(mgsJaccTraditionalAnnot, 
                          "mgs.jacc.0.75.traditional.nodeTable.txt",
                          "mgs.jacc.0.75.traditional.edgeTable.txt")
  
  #### diseased samples ####
  mgsJaccDiseaseModule = makeModuleList(mgsJaccDiseaseNet)
  mgsJaccDiseaseAnnot = annotateModulesByCC(mgsJaccDiseaseNet, mgsJaccDiseaseModule, cutCluster = 3)
  
  writeModuleNetworkAnnot(mgsJaccDiseaseAnnot, 
                          "mgs.jacc.0.75.disease.nodeTable.txt",
                          "mgs.jacc.0.75.disease.edgeTable.txt")
  
  #### ET-firmiuctes samples ####
  mgsJaccEtFNet = makeJaccNet(mgsJaccEtFMat, cutJacc = 0.5)
  mgsJaccEtFModule = makeModuleList(mgsJaccEtFNet)
  mgsJaccEtFAnnot = annotateModulesByCC(mgsJaccEtFNet, mgsJaccEtFModule, cutCluster = 3)
  
  writeModuleNetworkAnnot(mgsJaccEtFAnnot, 
                          "mgs.jacc.0.5.EtF.nodeTable.txt",
                          "mgs.jacc.0.5.EtF.edgeTable.txt")
  
  #### ET-bacteroides samples ####
  mgsJaccEtBNet = makeJaccNet(mgsJaccEtBMat, cutJacc = 0.5)
  mgsJaccEtBModule = makeModuleList(mgsJaccEtBNet)
  mgsJaccEtBAnnot = annotateModulesByCC(mgsJaccEtBNet, mgsJaccEtBModule, cutCluster = 3)
  
  writeModuleNetworkAnnot(mgsJaccEtBAnnot, 
                          "mgs.jacc.0.5.EtB.nodeTable.txt",
                          "mgs.jacc.0.5.EtB.edgeTable.txt")
  
  #### ET-prevotella samples ####
  mgsJaccEtPNet = makeJaccNet(mgsJaccEtPMat, cutJacc = 0.5)
  mgsJaccEtPModule = makeModuleList(mgsJaccEtPNet)
  mgsJaccEtPAnnot = annotateModulesByCC(mgsJaccEtPNet, mgsJaccEtPModule, cutCluster = 3)
  
  writeModuleNetworkAnnot(mgsJaccEtPAnnot, 
                          "mgs.jacc.0.5.EtP.nodeTable.txt",
                          "mgs.jacc.0.5.EtP.edgeTable.txt")
  
  #### check ####
  targetList = mgsJaccModule$`2`
  targetList = mgsJaccIndustrialModule$`9`
  targetList = mgsJaccTraditionalModule$`29`
  sort(taxo$species[taxo$MSP %in% targetList])
  
  lapply(mgsJaccIndustrialModule[mgsJaccIndustrialAnnot$nodeTable$node], function(x) sort(taxo$species[taxo$MSP %in% x]) )
  lapply(mgsJaccTraditionalModule[mgsJaccTraditionalAnnot$nodeTable$node], function(x) sort(taxo$species[taxo$MSP %in% x]) )
  lapply(mgsJaccDiseaseModule[mgsJaccDiseaseAnnot$nodeTable$node], function(x) sort(taxo$species[taxo$MSP %in% x]) )
  
  table(mgsJaccDiseaseModule$`1` %in% mgsJaccIndustrialModule$`1`)
  table(mgsJaccDiseaseModule$`1` %in% mgsJaccIndustrialModule$`2`)
  
  statCheck = F
  if (statCheck) {
    mm = mgsJaccMat[lower.tri(mgsJaccMat)]
    mgsJaccMaxs = sapply(1:1977, function(ind){
      currVec = mgsJaccMat[ind,]
      currVec = currVec[-ind]
      return(max(currVec))
    })
    plot(density(mgsJaccMaxs))
    median(mgsJaccMaxs)
    quantile(mgsJaccMaxs,0.999)
    quantile(mgsJaccMaxs,0.001)
    
    quantile(mgsJaccMaxs,0.99)
    quantile(mgsJaccMaxs,0.01)
    
    plot(density(mm[mm!=0]), log="x")
  }
  
}

loadCoSymptomNetwork = F
if (loadCoSymptomNetwork) {
  coSymptomNetFile = "J:\\Deposit\\Project\\2018_microbiome_atlas\\atlas.disease.network.data\\co-symptom-network.txt"
  coSymtpomTab = read.delim(coSymptomNetFile, stringsAsFactors = F, sep="\t")
  
  targetDiseaseFile = "J:\\Deposit\\Project\\2018_microbiome_atlas\\atlas.disease.network.data\\co-symptom-network-diseaseNameMapping.txt"
  targetDiseaseTab = read.delim(targetDiseaseFile, stringsAsFactors = F, sep="\t")
  
  cut05Mode = F
  if (cut05Mode) {
    coSymtpomTabSel = coSymtpomTab[coSymtpomTab$MeSH.Disease.Term %in% targetDiseaseTab$Diseases.in.symptom.disease.network &
                                     coSymtpomTab$MeSH.Disease.Term.1 %in% targetDiseaseTab$Diseases.in.symptom.disease.network &
                                     coSymtpomTab$symptom.similarity.score >= 0.5, ] 
    
    dim(coSymtpomTabSel)  
    
    coSymptomNetSelFile = "J:\\Deposit\\Project\\2018_microbiome_atlas\\atlas.disease.network.data\\co-symptom-network.selected.atlas.txt"
    write.table(coSymtpomTabSel, file=coSymptomNetSelFile, sep="\t", quote = F, row.names = F)
    
  }
  
  cut033Mode = T
  if (cut033Mode) {
    coSymtpomTabSel = coSymtpomTab[coSymtpomTab$MeSH.Disease.Term %in% targetDiseaseTab$Diseases.in.symptom.disease.network &
                                     coSymtpomTab$MeSH.Disease.Term.1 %in% targetDiseaseTab$Diseases.in.symptom.disease.network &
                                     coSymtpomTab$symptom.similarity.score >= 1/3, ] 
    
    dim(coSymtpomTabSel)  
    
    coSymptomNetSelFile = "J:\\Deposit\\Project\\2018_microbiome_atlas\\atlas.disease.network.data\\co-symptom-network.selected.atlas.033.txt"
    write.table(coSymtpomTabSel, file=coSymptomNetSelFile, sep="\t", quote = F, row.names = F)
    
  }
}

loadCoMorbidGeneticNetwork = F
if (loadCoMorbidGeneticNetwork) {
  coNetworkFile = "J:\\Deposit\\Project\\2018_microbiome_atlas\\atlas.disease.network.data\\co-host-genetic-network.txt"
  coNetworkTab = read.delim(coNetworkFile, sep="\t", header=T, stringsAsFactors = F)
  coNetworkNameMapFile = "J://Deposit//Project//2018_microbiome_atlas//atlas.disease.network.data/co-host-genetic-network-diseaseNameMapping.txt"
  coNetworkNameTab = read.delim(coNetworkNameMapFile, sep="\t", header=T, stringsAsFactors = F)
  
  coNetworkTab$X..ICD1. = gsub("\\[","",coNetworkTab$X..ICD1.)
  coNetworkTab$X..ICD1. = gsub("\\]","",coNetworkTab$X..ICD1.)
  coNetworkTab$X.Name1. = gsub("\\[","",coNetworkTab$X.Name1.)
  coNetworkTab$X.Name1. = gsub("\\]","",coNetworkTab$X.Name1.)
  coNetworkTab$X.ICD2.  = gsub("\\[","",coNetworkTab$X.ICD2.)
  coNetworkTab$X.ICD2.  = gsub("\\]","",coNetworkTab$X.ICD2.)
  coNetworkTab$X.Name2. = gsub("\\[","",coNetworkTab$X.Name2.)
  coNetworkTab$X.Name2. = gsub("\\]","",coNetworkTab$X.Name2.)
  
  targetDiseaseNames = coNetworkNameTab$Diseases.in.host.genetic.disease.network
  
  table(coNetworkTab$X.Name1. %in% coNetworkNameTab$Diseases.in.host.genetic.disease.network)
  coNetworkNameTab$Diseases.in.host.genetic.disease.network[!coNetworkNameTab$Diseases.in.host.genetic.disease.network  %in%  coNetworkNameTab$X.Name1.]
  
  coNetworkSel =  coNetworkTab[coNetworkTab$X.Name1. %in% targetDiseaseNames
                               & coNetworkTab$X.Name2. %in% targetDiseaseNames,]
  
  coMorbidNetSel  = coNetworkSel[coNetworkSel$RHO>0,]
  coGeneticNetSel = coNetworkSel[coNetworkSel$NG>0,]
  
}

### pouyan data ###
compModeForPouyan = F
if (compModeForPouyan) {
  dsIds = c("id9", "id1","id10","id34",
            "id11","id44","id46","id45",
            "id6","id12","id17","id20",
            "id16","id25","id8","id31",
            "id21","id35","id2","id14", 
            "id52")
  
  statOut = list()
  for (dsId in dsIds) {
    
    currCaseIds = basicMetaMapUpdated$sample.ID[basicMetaMapUpdated$dataset.ID==dsId & basicMetaMapUpdated$type == "Case"]
    currContIds = basicMetaMapUpdated$sample.ID[basicMetaMapUpdated$dataset.ID==dsId & basicMetaMapUpdated$type == "Control"]
    currSamples = c(currCaseIds, currContIds)
    print(table(currCaseIds %in% currContIds))
    print(paste(dsId, length(currCaseIds), length(currContIds)))
    
    currCaseMat = mergeMatUpdated[,colnames(mergeMatUpdated)%in%currCaseIds]
    currContMat = mergeMatUpdated[,colnames(mergeMatUpdated)%in%currContIds]
    
    currDataMat = cbind(currCaseMat, currContMat)
    currDataLab = c(rep("case",dim(currCaseMat)[2]),
                    rep("cont",dim(currContMat)[2]))
    
    currCaseMeans = rowMeans(currCaseMat)
    currContMeans = rowMeans(currContMat)
    currFoldChange = currCaseMeans / currContMeans
    
    currStats = testRelations(currDataMat, currDataLab, "wilcoxon")
    currStats$dataset = dsId
    currStats$msp = rownames(currStats)
    currStats$foldChange = currFoldChange
    currStats$Case = currCaseMeans
    currStats$Cont = currContMeans
    
    
    currStats = currStats[,3:10]
    currStats$p[is.nan(currStats$p)]=1
    currStats$q[is.nan(currStats$q)]=1
    
    statOut[[dsId]] = currStats
    print(head(currStats))
    
  }
  statFinal = do.call(rbind, statOut)
  
  gssOssTabNewFile = "J://Deposit//Project//2018_oral_catalog_paper//catalogs (gut, oral), 20190523//MSP_set_class_tropism_20190523_short.txt"
  gssOssTabNew = read.table(gssOssTabNewFile, header=T, sep="\t", stringsAsFactors = F)
  rownames(gssOssTabNew) = gssOssTabNew$MSP
  
  statFinal$species = gssOssTabNew$species[match(statFinal$msp, gssOssTabNew$MSP)]
  
  write.table(statFinal, 
              #"C://Data//comparative.analysis.healthy.sweden//atlas.case.cont.comparison.pouyan.20190626.txt",
              "C://Data//comparative.analysis.healthy.sweden//atlas.case.cont.comparison.pouyan.20190827.txt",
              sep="\t")
  
  
  
}

atlasCompModeNew = F
if (atlasCompModeNew) {
  
  normalMat = mergeMatUpdated[,colnames(mergeMatUpdated) %in% newNormalIds]
  
  targetMat = mergeMatUpdated
  phylaMat = getTaxaSumMat(targetMat,taxo,"phylum",T)
  classMat = getTaxaSumMat(targetMat,taxo,"class",T)
  orderMat = getTaxaSumMat(targetMat,taxo,"order",T)
  familyMat = getTaxaSumMat(targetMat,taxo,"family",T)
  genusMat = getTaxaSumMat(targetMat,taxo,"genus",T)
  speciesMat = getTaxaSumMat(targetMat,taxo,"species",T)
  
  
  checkEnrichedByGeo <- function(newNormalAllList, mgsMat) {
    
    numDatasets = length(newNormalAllList)
    nameDatasets = names(newNormalAllList)
    
    
    
    for (targetInd in seq_along(newNormalAllList)) {
      targetSamples = newNormalAllList[[targetInd]]
      targetName = nameDatasets[targetInd]
      
      targetMat = mgsMat[, colnames(mgsMat) %in% targetSamples]
      targetMeans = rowMeans(targetMat)
      
      foldChangeList = list()
      effectSizeList = list()
      
      for (refInd in seq_along(newNormalAllList)) {
        refSamples = newNormalAllList[[refInd]]
        refName = nameDatasets[refInd]
        
        refMat = mgsMat[, colnames(mgsMat) %in% refSamples]
        refMeans = rowMeans(refMat)
        currFoldChange = refMeans/targetMeans
        currEffectSize = getEffectSizeMat(targetMat, refMat)
        
        foldChangeList[[refName]] = currFoldChange
        effectSizeList[[refName]] = currEffectSize
        
        
      }
      
      ##effectSizeMat = do.call(rbind, effectSizeList)
      ##effectSizeMatDigit = effectSizeMat>
      
    }
    
    
  } 
  
  oldMode = F
  if (oldMode) {
    getSampleListByDisease <- function(basicMetaMap, mgsInfo) {
      
      caseDatasets = unique(mgsInfo$ID[mgsInfo$Classification=="disease"])
      caseDatasets2 = caseDatasets[caseDatasets!="id38"]
      caseIDs = basicMetaMap$sample.ID[basicMetaMap$type=="Case" & basicMetaMap$dataset.ID %in% caseDatasets]
      caseIDs2 = basicMetaMap$sample.ID[basicMetaMap$type=="Case" & basicMetaMap$dataset.ID %in% caseDatasets2]
      diseaseOfCaseIDs = sapply(caseIDs, function(currID){
        currDatasetId = basicMetaMap$dataset.ID[basicMetaMap$sample.ID==currID]
        currDisease = unique(mgsInfo$Case[mgsInfo$ID==currDatasetId])
        return(currDisease)
      })
      caseIDsByDisease= split(names(diseaseOfCaseIDs), diseaseOfCaseIDs)
      lenDisease = length(caseIDsByDisease)
      basicMetaMap$disease = ""
      for (ind in 1:lenDisease){
        basicMetaMap$disease[basicMetaMap$sample.ID%in%caseIDsByDisease[[ind]]] = names(caseIDsByDisease)[ind]
      }
      
      
      selectedMetaMap = basicMetaMap[basicMetaMap$sample.ID %in% caseIDs,]
      samplesByDisease = split(selectedMetaMap$sample.ID, selectedMetaMap$disease)
      
      exDisease = c("atherosclerosis")# those samples having less than 5 samples
      samplesByDisease = samplesByDisease[!names(samplesByDisease)%in% exDisease]
      
      return(samplesByDisease)
      
    }
    getSampleListByCountry <- function(basicMetaMap, mgsInfo) {
      mContDatasets = unique(mgsInfo$ID[mgsInfo$Classification=="disease"&!is.na(mgsInfo$Control)])
      hContDatasets = unique(mgsInfo$ID[mgsInfo$Classification=="healthy"&!is.na(mgsInfo$Control)])
      
      healthyDatasets = unique(mgsInfo$ID[mgsInfo$Classification=="healthy"])
      indiDatasets = c("id32","id37","id33","id42")
      notIndiDatasets = healthyDatasets[!healthyDatasets %in% indiDatasets]
      
      healthyIDs = basicMetaMap$sample.ID[basicMetaMap$type=="Case" & basicMetaMap$dataset.ID %in% healthyDatasets]
      indiIDs = basicMetaMap$sample.ID[basicMetaMap$type=="Case" & basicMetaMap$dataset.ID %in% indiDatasets]
      NotIndiHealtyIDs = healthyIDs[!healthyIDs %in% indiIDs]
      
      mContIDs = basicMetaMap$sample.ID[basicMetaMap$type=="Control" & basicMetaMap$dataset.ID %in% mContDatasets]
      hContIDs = basicMetaMap$sample.ID[basicMetaMap$type=="Control" & basicMetaMap$dataset.ID %in% hContDatasets]
      mhContIDs = unique(c(mContIDs, hContIDs))
      normalIDs = unique(c(NotIndiHealtyIDs, mhContIDs))
      
      
      allHealthyIds = unique(c(mhContIDs, normalIDs, indiIDs))
      selectedMetaMap = basicMetaMap[basicMetaMap$sample.ID %in% allHealthyIds,]
      
      samplesByCountry = split(selectedMetaMap$sample.ID, selectedMetaMap$Geography)
      exCountries = c("Finland","Italy")# those samples having less than 5 samples
      samplesByCountry = samplesByCountry[!names(samplesByCountry)%in% exCountries]
      return(samplesByCountry)
    }
    getSampleListByIndiCountry <- function(basicMetaMap, mgsInfo) {
      indiDatasets = c("id32","id37","id33","id42")
      indiIDs = basicMetaMap$sample.ID[basicMetaMap$type=="Case" & basicMetaMap$dataset.ID %in% indiDatasets]
      selectedMetaMap = basicMetaMap[basicMetaMap$sample.ID %in% indiIDs,]
      
      samplesByCountry = split(selectedMetaMap$sample.ID, selectedMetaMap$Geography)
      return(samplesByCountry)
    }
    getSampleListByNotIndiCountry <- function(basicMetaMap, mgsInfo) {
      mContDatasets = unique(mgsInfo$ID[mgsInfo$Classification=="disease"&!is.na(mgsInfo$Control)])
      hContDatasets = unique(mgsInfo$ID[mgsInfo$Classification=="healthy"&!is.na(mgsInfo$Control)])
      
      healthyDatasets = unique(mgsInfo$ID[mgsInfo$Classification=="healthy"])
      notIndiDatasets = healthyDatasets[!healthyDatasets %in% indiDatasets]
      
      healthyIDs = basicMetaMap$sample.ID[basicMetaMap$type=="Case" & basicMetaMap$dataset.ID %in% healthyDatasets]
      NotIndiHealtyIDs = healthyIDs[!healthyIDs %in% indiIDs]
      
      mContIDs = basicMetaMap$sample.ID[basicMetaMap$type=="Control" & basicMetaMap$dataset.ID %in% mContDatasets]
      hContIDs = basicMetaMap$sample.ID[basicMetaMap$type=="Control" & basicMetaMap$dataset.ID %in% hContDatasets]
      mhContIDs = unique(c(mContIDs, hContIDs))
      normalIDs = unique(c(NotIndiHealtyIDs, mhContIDs))
      
      
      allHealthyIds = unique(c(mhContIDs, normalIDs))
      selectedMetaMap = basicMetaMap[basicMetaMap$sample.ID %in% allHealthyIds,]
      
      samplesByCountry = split(selectedMetaMap$sample.ID, selectedMetaMap$Geography)
      exCountries = c("Finland","Italy")# those samples having less than 5 samples
      samplesByCountry = samplesByCountry[!names(samplesByCountry)%in% exCountries]
      return(samplesByCountry)
    }
    
    samplesByDisease = getSampleListByDisease(basicMetaMap, mgsInfo)
    samplesByAllCountry = getSampleListByCountry(basicMetaMap, mgsInfo)
    sampleListByIndiCountry = getSampleListByIndiCountry(basicMetaMap, mgsInfo)
    sampleListByNotIndiCountry = getSampleListByNotIndiCountry(basicMetaMap, mgsInfo)
    
    showFourClassPerBarPlotByCountry <- function(samplesByCountry, target, mergeMat, cutOffs=c(1e-8,  1e-6)) {
      
      perList <- lapply(samplesByCountry, function(currSamples){
        currVec = mergeMat[target,currSamples]
        absent = sum(currVec < cutOffs[1])/length(currVec)
        medium = sum(currVec < cutOffs[2] & currVec >= cutOffs[1] )/length(currVec)
        high = sum(currVec >= cutOffs[2] )/length(currVec)
        #present = sum(currVec >= cutOff)/length(currVec)
        return(c(high, medium, absent))
      })
      perMat <- do.call(cbind, perList)
      rownames(perMat) = c("high","medium","absent")
      cols = c("red","gold","gray")
      cols =  adjustcolor( cols, alpha.f = 0.8)
      par(mar=c(5,10,4,2))
      barplot(perMat, horiz = T, las=1, col=cols)
      legend("topleft",legend=c("high","medium","lowly"),
             fill=cols)
      
      par(mar=c(5.1,4.1,4.1,2.1))
      return(perMat)
    }
    showPerBarPlotByCountry <- function(samplesByCountry, target, mergeMat, cutOff=1e-7) {
      perList <- lapply(samplesByCountry, function(currSamples){
        currVec = mergeMat[target,currSamples]
        present = sum(currVec >= cutOff)/length(currVec)
        absent = sum(currVec < cutOff)/length(currVec)
        return(c(present, absent))
      })
      perMat <- do.call(cbind, perList)
      
      par(mar=c(5,10,4,2))
      barplot(perMat, horiz = T, las=1)
      par(mar=c(5.1,4.1,4.1,2.1))
      return(perMat)
    }
    showBoxPlotsByCountry <- function(samplesByCountry, target, mergeMat) {
      abdList <- lapply(samplesByCountry, function(currSamples){
        return(mergeMat[target,currSamples])
      })
      par(mar=c(5,10,4,2))
      boxplot(abdList,horizontal = T,las=1, ylim=c(0,1e-5))
      par(mar=c(5.1,4.1,4.1,2.1))
      return(abdList)
    }
    showPerBarPlotByDisease <- function(samplesByDisease, target, normalMat, mergeMat, cutOff=1e-7) {
      perList <- lapply(samplesByDisease, function(currSamples){
        currVec = mergeMat[target,currSamples]
        present = sum(currVec >= cutOff)/length(currVec)
        absent = sum(currVec < cutOff)/length(currVec)
        return(c(present, absent))
      })
      normMode = T
      if (normMode) {
        currVec = normalMat[target,]
        present = sum(currVec >= cutOff)/length(currVec)
        absent = sum(currVec < cutOff)/length(currVec)
        perList$normal = c(present, absent)
      }
      perMat <- do.call(cbind, perList)
      
      par(mar=c(5,10,4,2))
      barplot(perMat, horiz = T, las=1)
      par(mar=c(5.1,4.1,4.1,2.1))
      return(perMat)
    }
    showBoxPlotsByDisease <- function(samplesByDisease, target, normalMat, mergeMat, ylims = c(0,1e-5)) {
      abdList <- lapply(samplesByDisease, function(currSamples){
        return(mergeMat[target,currSamples])
      })
      normMode = T
      if (normMode) {
        currVec = normalMat[target,]
        abdList$normal = currVec
      }
      par(mar=c(5,10,4,2))
      boxplot(abdList,horizontal = T,las=1, ylim=ylims)
      par(mar=c(5.1,4.1,4.1,2.1))
      return(abdList)
    }
    
    targetSpecies = "msp_0025"
    # perList=showPerBarPlotByDisease(samplesByDisease, targetSpecies, normalMat, mergeMat )
    getSpeciesName(targetSpecies, taxo)
    ylims= c(0,1e-5)
    outList=showBoxPlotsByDisease(samplesByDisease, targetSpecies, normalMat, mergeMat,ylims )
    
    do.call(c,lapply(outList, mean))
    do.call(c,lapply(outList, median))
    
    
    targetSpecies = "msp_0025"
    perList=showFourClassPerBarPlotByCountry(samplesByAllCountry, targetSpecies, mergeMat, c(1e-8, 1e-6) )
    
    getSpeciesName(targetSpecies, taxo)
    
    outList=showBoxPlotsByCountry(samplesByAllCountry, targetSpecies, familyMat )
    do.call(c,lapply(outList, mean))
    do.call(c,lapply(outList, median))
    
    speciesDiseaseList = statisticsByDisease(samplesByDisease, sampleListByNotIndiCountry, normalMat, mergeMat, esCut=0.2, numCut=17, allCut=8)
    speciesDiseaseGeneralList = statisticsByDiseaseGeneral(samplesByDisease, sampleListByNotIndiCountry, normalMat, mergeMat, esCut=0.3, numCut=17)
    
    speciesSpecificList = statisticsByCountrySpecific(samplesByAllCountry, mergeMat, esCut=0.2, numCut=11)
    
    ####
    speciesUrbanSpecificList = statisticsByUrbanSpecific(sampleListByNotIndiCountry, sampleListByIndiCountry, mergeMat, esCut=0.2)
    speciesIndiSpecificList = statisticsByIndiSpecific(sampleListByNotIndiCountry, sampleListByIndiCountry, mergeMat, esCut=0.2)
    speciesDisSpecificList = statisticsByDiseaseSpecific(sampleListByNotIndiCountry, samplesByDisease,mergeMat, esCut=0.2)
    
    speciesUrbanSpecificTab = melt(speciesUrbanSpecificList)
    speciesIndiSpecificTab = melt(speciesIndiSpecificList)
    speciesDisSpecificTab = melt(speciesDisSpecificList)
    
    speciesUrbanSpecificTab$type="healthy"
    speciesIndiSpecificTab$type="indi"
    speciesDisSpecificTab$type="disease"
    
    speciesAllSpecificTab = rbind(speciesUrbanSpecificTab,
                                  speciesIndiSpecificTab,
                                  speciesDisSpecificTab)
    speciesAllSpecificTab$name = getSpeciesName(speciesAllSpecificTab$value,taxo)
    write.table(speciesAllSpecificTab,"speciesAllSpecificTab.txt",sep='\t')
    
    
    save(speciesUrbanSpecificList,
         speciesIndiSpecificList,
         speciesDisSpecificList,
         file="specific.species.by.category.RData")
    
    load("specific.species.by.category.RData")
    
    
    speciesUrbanSpecificList_es_02 = unique(do.call(c,speciesUrbanSpecificList))
    #177 species
    speciesIndiSpecificList_es_02 = unique(do.call(c,speciesIndiSpecificList))
    #187 species
    speciesDisSpecificList_es_02 = unique(do.call(c,speciesDisSpecificList))
    #72 species
    
    table(speciesUrbanSpecificList_es_02 %in% speciesIndiSpecificList_es_02)
    table(speciesUrbanSpecificList_es_02 %in% speciesDisSpecificList_es_02)
    table(speciesIndiSpecificList_es_02 %in% speciesDisSpecificList_es_02)
    
    
    urbanSpecies = unique(do.call(c,speciesUrbanSpecificList))
    indiSpecies = unique(do.call(c,speciesIndiSpecificList))
    
    getSpeciesName(urbanSpecies, taxo)
    getSpeciesName(indiSpecies, taxo)
    
    
    speciesUrbanList = statisticsByUrbanCountry(sampleListByNotIndiCountry, sampleListByIndiCountry, mergeMat, esCut=0.2)
    speciesIndiList = statisticsByIndiCountry(sampleListByNotIndiCountry, sampleListByIndiCountry, mergeMat, esCut=0.2)
    
    familyUrbanList = statisticsByUrbanCountry(sampleListByNotIndiCountry, sampleListByIndiCountry, familyMat, esCut=0.2)
    familyIndiList = statisticsByIndiCountry(sampleListByNotIndiCountry, sampleListByIndiCountry, familyMat, esCut=0.2)
    
    genusUrbanList = statisticsByUrbanCountry(sampleListByNotIndiCountry, sampleListByIndiCountry, genusMat, esCut=0.2)
    genusIndiList = statisticsByIndiCountry(sampleListByNotIndiCountry, sampleListByIndiCountry, genusMat, esCut=0.2)
    
    
    speciesDiseaseDownList = statisticsDownByDisease(samplesByDisease[names(samplesByDisease)!="preterm"], sampleListByNotIndiCountry, normalMat, mergeMat, esCut=0.2)
    speciesUrbanDownList = statisticsDownByUrbanCountry(sampleListByNotIndiCountry, sampleListByIndiCountry, mergeMat, esCut=0.2)
    speciesIndiDownList = statisticsDownByIndiCountry(sampleListByNotIndiCountry, sampleListByIndiCountry, mergeMat, esCut=0.2)
    
    lapply(speciesUrbanList, function(x)getSpeciesName(x,taxo))
    
    sankeyMode = T
    targetList = speciesIndiSpecificList
    if (sankeyMode) {
      library(networkD3)
      library(tidyverse)
      require(RColorBrewer)
      
      taretMelt = melt(targetList)
      taretMelt$name = sapply(taretMelt$value, function(x) getSpeciesName(x,taxo))
      
      taretMelt$newName = paste(taretMelt$name, taretMelt$value, sep=",")
      taretMelt$value=1
      
      plotMode=T
      if (plotMode) {
        fsSize=12
        qual_col_pals = brewer.pal.info[brewer.pal.info$category == 'qual',]
        col_vector = unlist(mapply(brewer.pal, qual_col_pals$maxcolors, rownames(qual_col_pals)))
        
        targetOut = taretMelt
        elements = c(unique(targetOut$newName),unique(targetOut$L1))
        elementCols = col_vector[1:length(elements)]#
        
        interactions = targetOut[,c("newName","L1","value")]
        numInteractions = dim(interactions)[1]
        
        sourceInds = match(interactions$newName, elements)
        targetInds = match(interactions$L1, elements)
        interactionValues = interactions$value
        
        links = interactions
        colnames(links) = c("source","target","value")
        
        
        # From these flows we need to create a node data frame: it lists every entities involved in the flow
        nodes=data.frame(name=c(as.character(links$source), as.character(links$target)) %>% unique())
        links$IDsource=match(links$source, nodes$name)-1 
        links$IDtarget=match(links$target, nodes$name)-1
        
        
        color_scale <- data.frame(
          range = elementCols,
          domain = elements, 
          nodes = as.character(nodes$name),
          stringsAsFactors = FALSE
        )
        
        # Make the Network. I call my colour scale with the colourScale argument
        sankeyNetwork(Links = links, Nodes = nodes, Source = "IDsource", Target = "IDtarget", Value = "value", NodeID = "name", fontFamily = "arial", fontSize = fsSize, colourScale = JS(
          sprintf(
            'd3.scaleOrdinal()  
            .domain(%s)
            .range(%s)
            ',
            jsonlite::toJSON(color_scale$domain),
            jsonlite::toJSON(color_scale$range)
          )
        ))
      }
      
    }
    
    
    getSampleListByDisease <- function(basicMetaMap) {
      
    }
    
    getSampleListByAge <- function(basicMetaMap) {
      
    }
    
    getSampleListByBmi <- function(basicMetaMap) {
      
    }
    
    
    
    ### country-wise comparison
    getStats <- function()
      
      ### 
  }
  
  
  }

checkLucasKoProfile = F
if (checkLucasKoProfile) {
  gutKoBestMatRData = "j://Deposit/Project/2018_microbiome_atlas/functional.annotation/gutKoBestMat.1990.RData"
  load(gutKoBestMatRData) #gutKoBestMat
  gutKoBestSums = colSums(gutKoBestMat)
  gutKoBestSums = sort(gutKoBestSums, decreasing = T)
  gutKoCountTab = data.frame(KO=names(gutKoBestSums),
                             MSP_COUNT = gutKoBestSums,
                             KO_gene = koDescMap$gene[match(names(gutKoBestSums), koDescMap$KO)],
                             DESCRIPTION=koDescMap$desc[match(names(gutKoBestSums), koDescMap$KO)])
  
  gutKoCountTabFile = "j://Deposit/Project/2018_microbiome_atlas/functional.annotation/gutKoCountTab.for.lucas.txt"
  write.table(gutKoCountTab, file=gutKoCountTabFile, row.names = F, quote = F, sep="\t")
}

diseaseNetworkMode = F
if (diseaseNetworkMode) {
  
  mergeMatUpdatedJaccRData = "J:\\Deposit\\Project\\2018_microbiome_atlas\\microbiome.atlas.Rproj//mergeMatUpdatedJaccard.RData"
  genusMatUpdatedJaccRData = "J:\\Deposit\\Project\\2018_microbiome_atlas\\microbiome.atlas.Rproj//genusMatUpdatedJaccard.RData"
  
  dim(genusMatUpdated)
  
  generateJacc = F
  if (generateJacc) {
    mergeMatUpdatedDigit = (mergeMatUpdated > 0)*1
    mergeMatUpdatedJacc = getJaccMat(mergeMatUpdatedDigit)
    save(mergeMatUpdatedJacc, 
         file=mergeMatUpdatedJaccRData)
    
    genusMatUpdatedDigit = (genusMatUpdated > 0)*1
    genusMatUpdatedJacc = getJaccMat(genusMatUpdatedDigit)
    save(genusMatUpdatedJacc, 
         file=genusMatUpdatedJaccRData)
    
  }
  
  loadJacc = T
  if (loadJacc) {
    load(mergeMatUpdatedJaccRData) #mergeMatUpdatedJacc
    jaccNet = makeJaccNet(mergeMatUpdatedJacc, cutJacc = 0.5)
    
    ### geography network ###
    geoList = newNormalAllList
    geoClusterScores = getClusterScores(jaccNet, geoList)
    
    jaccClusterInfo = annotateModulesByCC(jaccNet, geoList)
    writeModuleNetworkAnnot(jaccClusterInfo, 
                            "nodeTableJaccGeo.txt",
                            "edgeTableJaccGeo.txt")
    
    ### disease network ###
    diseaseList = newDiseaseList
    diseaseClusterScores = getClusterScores(jaccNet, diseaseList)
    
    jaccClusterInfo = annotateModulesByCC(jaccNet, diseaseList)
    writeModuleNetworkAnnot(jaccClusterInfo, 
                            "nodeTableJaccDisease.txt",
                            "edgeTableJaccDisease.txt")
    
    
  }
  
  
}

transmissionMap = F
if (transmissionMap) {
  
  generateMode = F
  if (generateMode) {
    
    freqs = rowSums(mergeMatUpdated>0)
    median(freqs) 
    
    freqSpecies = names(freqs[freqs > median(freqs)])
    rareSpecies = names(freqs[freqs < median(freqs)])
    
    digitMat = (mergeMatUpdated>0)*1
    digitFreqMat = (mergeMatUpdated[freqSpecies,]>0)*1
    digitRareMat = (mergeMatUpdated[rareSpecies,]>0)*1
    digitFreqJaccMat = 1-jaccard(digitFreqMat)
    digitRareJaccMat = 1-jaccard(digitRareMat)
    digitRareJaccMat[is.nan(digitRareJaccMat)]=0
    
    #digitJaccMat = getJaccMat(digitMat)
    
    #require(prabclus)
    digitJaccMat = 1-jaccard(digitMat)
    
    digitJaccMatRData = "C:\\Data\\comparative.analysis.healthy.sweden\\digitJaccMat.RData"
    
    save(digitJaccMat, file=digitJaccMatRData)
    #save(digitJaccMat, digitFreqJaccMat, digitRareJaccMat, file="C:\\Data\\comparative.analysis.healthy.sweden\\digitJaccMat.all.RData")
    #load(digitJaccMatRData)
    
    
    #genusup = getTaxaSumMat(mergeMat, taxo, "genus")
    #digitGenusMat = (genusMat>0)*1
    digitGenusMat = (genusMatUpdated>0)*1
    digitGenusJaccMat = 1-jaccard(digitGenusMat)
    save(digitGenusJaccMat, file="C:\\Data\\comparative.analysis.healthy.sweden\\digitGenusJaccMat.RData")
    
  }
  
  loadMode = T
  if (loadMode) {
    mgsJaccRData = "C:/Data/comparative.analysis.healthy.sweden/digitJaccMat.RData"
    genusJaccRData = "C:/Data/comparative.analysis.healthy.sweden/digitGenusJaccMat.RData"
    load(mgsJaccRData) #digitJaccMat
    load(genusJaccRData)   #digitGenusJaccMat 
    
    plot(density(digitJaccMat[lower.tri(digitJaccMat)]), ylim=c(0,5), main="")
    lines(density(digitGenusJaccMat[lower.tri(digitGenusJaccMat)]), col="red")
    
    samplesBySubjects = split(basicMetaMapUpdated$sample.ID,
                              basicMetaMapUpdated$metadata.ID)
    lenSamplesBySubjects = do.call(c,lapply(samplesBySubjects,length))
    samplesBySubjects = samplesBySubjects[lenSamplesBySubjects>1]
    
    
    allMgsJaccs = digitJaccMat[lower.tri(digitJaccMat)]
    allGenusJaccs = digitGenusJaccMat[lower.tri(digitGenusJaccMat)]
    
    subjectMgsJaccs = do.call(c,lapply(samplesBySubjects, function(currSamples){
      currMat = digitJaccMat[rownames(digitJaccMat)%in%currSamples, colnames(digitJaccMat)%in%currSamples]
      return(currMat[lower.tri(currMat)])
    }))
    plot(density(digitJaccMat[lower.tri(digitJaccMat)]), ylim=c(0,10), main="")
    lines(density(subjectMgsJaccs), col="red")
    
    
    subjectGenusJaccs = do.call(c,lapply(samplesBySubjects, function(currSamples){
      currMat = digitGenusJaccMat[rownames(digitGenusJaccMat)%in%currSamples, colnames(digitGenusJaccMat)%in%currSamples]
      return(currMat[lower.tri(currMat)])
    }))
    plot(density(digitGenusJaccMat[lower.tri(digitGenusJaccMat)]), ylim=c(0,10), main="")
    lines(density(subjectGenusJaccs), col="red")
    
  }
  
  mgsLevel = T
  if (mgsLevel) {
    load("digitJaccMat.RData")
    names(datasetList)
    
    #digitJaccMatCut = (digitJaccMat > 0.5)*1
    digitJaccMatCut = (digitRareJaccMat > 0.5)*1
    
    countsMat = lapply(datasetList, function(cohortI) {
      countsRow = lapply(datasetList,  function(cohortJ) {
        counts = sum(digitJaccMatCut[cohortI, cohortJ])
        return(counts)
        
      })
      
      return(countsRow)
    })
    countsMelt = melt(countsMat)
    countsTransmissionMat = acast(countsMelt, L1~ L2, value.var = "value")
    heatmap.2((countsTransmissionMat>0)*1, trace="none", key=F,
              col = colorRampPalette(c("white","gold"))(256),
              sepcolor = "gray",
              colsep=0:32,rowsep=0:32,
              sepwidth = c(0.01,0.01)
    )
    
    comms = c("Firmicutes","Bacteroidetes","unclassified","Proteobacteria",
              "Actinobacteria","Fusobacteria","Verrucomicrobia","Tenericutes",
              "Euryarchaeota","Synergistetes","unclassified Eukaryota",
              "Elusimicrobia","Candidatus Melainabacteria","unclassified Bacteria")
    
    freqVec = table(getGenusName(freqSpecies,taxo))
    rareVec = table(getGenusName(rareSpecies,taxo))
    comms = names(freqVec)[names(freqVec) %in% names(rareVec)]
    
    freqVec = table(getPhylaName(freqSpecies,taxo))
    rareVec = table(getPhylaName(rareSpecies,taxo))
    
    plot(as.numeric(freqVec[comms]), as.numeric(rareVec[comms]),log="xy")
    
    freqVec[names(rareVec)]
    
    wellnessIDs = basicMetaMap$sample.ID[basicMetaMap$dataset.ID=="id28"]
    wellnessBySubjs = split(wellnessIDs, getSubectIds(wellnessIDs)  )
    
    wellnessJaccBySubs = lapply(wellnessBySubjs, function(x){
      currMat  = digitJaccMat[x,x]
      return(currMat[lower.tri(currMat)])
    })
    
    twinIDs = basicMetaMap$sample.ID[basicMetaMap$dataset.ID=="id39"]
    madaIds = basicMetaMap$sample.ID[basicMetaMap$dataset.ID=="id42"]
    t1dIds = basicMetaMap$sample.ID[basicMetaMap$dataset.ID=="id35"]
    hmpIds = basicMetaMap$sample.ID[basicMetaMap$dataset.ID=="id36"]
    rccIds = basicMetaMap$sample.ID[basicMetaMap$dataset.ID=="id41"]
    rheIds = basicMetaMap$sample.ID[basicMetaMap$dataset.ID=="id31"]
    
    digitTwinMat = digitMat[,twinIDs]
    digitHmpMat = digitMat[,hmpIds]
    digitT1dMat = digitMat[,t1dIds]
    digitRccMat = digitMat[,rccIds]
    digitRheMat = digitMat[,rheIds]
    digitMadaMat = digitMat[,madaIds]
    digitWellMat = digitMat[,wellnessIDs]
    
    
    plot(rowSums(digitTwinMat), rowSums(digitHmpMat))
    plot(rowSums(digitRccMat), rowSums(digitHmpMat))
    plot(rowSums(digitRheMat), rowSums(digitHmpMat))
    plot(rowSums(digitT1dMat), rowSums(digitHmpMat))
    plot(rowSums(digitWellMat), rowSums(digitHmpMat))
    
    
    plot(rowSums(digitTwinMat), rowSums(digitMat))
    plot(rowSums(digitRccMat), rowSums(digitMat))
    plot(rowSums(digitRheMat), rowSums(digitMat))
    plot(rowSums(digitT1dMat), rowSums(digitMat))
    plot(rowSums(digitMadaMat), rowSums(digitMat))
    plot(rowSums(digitWellMat), rowSums(digitMat))
    plot(rowSums(digitHmpMat), rowSums(digitMat))
    
    plot(rowSums(digitMadaMat), rowSums(digitHmpMat))
    
    
    
    
    getSpeciesName(names(sort(rowSums(digitTwinMat),decreasing = T)[1:10]),taxo)
    getSpeciesName(names(sort(rowSums(digitHmpMat),decreasing = T)[1:10]),taxo)
    getSpeciesName(names(sort(rowSums(digitWellMat),decreasing = T)[1:10]),taxo)
    getSpeciesName(names(sort(rowSums(digitWellMat),decreasing = F)[1:10]),taxo)
    
    twinMat = digitJaccMat[twinIDs,twinIDs]
    wellMat = digitJaccMat[wellnessIDs,wellnessIDs]
    madaMat =  digitJaccMat[madaIds,madaIds]
    t1dMat = digitJaccMat[t1dIds,t1dIds]
    hmpMat = digitJaccMat[hmpIds,hmpIds]
    rccMat = digitJaccMat[rccIds,rccIds]
    rheMat = digitJaccMat[rheIds,rheIds]
    
    
    jaccRegList = list(twin=twinMat[lower.tri(twinMat)],
                       wellness=wellMat[lower.tri(wellMat)],
                       madagascar=madaMat[lower.tri(madaMat)],
                       hmp = hmpMat[lower.tri(hmpMat)],
                       t1d=t1dMat[lower.tri(t1dMat)],
                       rcc=rccMat[lower.tri(rccMat)],
                       rhe=rheMat[lower.tri(rheMat)])
    boxplot(jaccRegList,horizontal = T,las=1)
    
    
    boxplot(twinMat[lower.tri(twinMat)])
    boxplot(wellMat[lower.tri(wellMat)])
  }
  
  genusLevel = T
  if (genusLevel) {
    names(datasetList)
    
    digitGenusJaccMatCut = (digitGenusJaccMat > 0.9)*1
    
    countsMat = lapply(datasetList, function(cohortI) {
      countsRow = lapply(datasetList,  function(cohortJ) {
        counts = sum(digitGenusJaccMatCut[cohortI, cohortJ])
        return(counts)
        
      })
      
      return(countsRow)
    })
    countsMelt = melt(countsMat)
    countsTransmissionMat = acast(countsMelt, L1~ L2, value.var = "value")
    heatmap.2((countsTransmissionMat>0)*1, trace="none", key=F,
              col = colorRampPalette(c("white","gold"))(256),
              sepcolor = "gray",
              colsep=0:32,rowsep=0:32,
              sepwidth = c(0.01,0.01)
    )
    
  }
  
}

tsneAfterPcaReductionMode = F
if (tsneAfterPcaReductionMode) {
  
  getTsneAfterPca <- function(targetMat, plotMode = F, numComps = 5, perplexity=50) {
    pcaOut = prcomp(t(targetMat))
    pcaMat = pcaOut$x
    eigs <- pcaOut$sdev^2
    pcaSummary = rbind(
      SD = sqrt(eigs),
      Proportion = eigs/sum(eigs),
      Cumulative = cumsum(eigs)/sum(eigs))
    print(pcaSummary[,1:numComps])
    print("")
    print("... pca done")
    
    set.seed(1)
    tsneOut = Rtsne(pcaMat[,1:numComps], dims = 2, perplexity = perplexity)
    tsneMat = tsneOut$Y
    print("... tSNE done")
    
    if (plotMode) {
      tsneCols = rep("#80808088",dim(tsneMat)[1])
      names(tsneCols) = colnames(targetMat)
      
      
    }
    return(tsneMat)
  }
  
  normMat = mergeMatUpdated[,normUniqAll]
  normTsneMat = getTsneAfterPca(normMat)
  
  tsneCols = rep("#80808088",dim(tsneMat)[1])
  names(tsneCols) = colnames(targetMat)
  tsneCols[names(tsneCols) %in% newNonWesternIds] = "red"
  
  plot(tsneMat, col=tsneCols)
  
}

updateEnteroType = F
if (updateEnteroType) {
  
  genusMode = T
  if (genusMode) {
    
    genusMat = getTaxaSumMat(mergeMatUpdated, taxoNew, "genus")
    genusMatT = round(t(genusMat)*10^9)
    
    generateMode = T
    if (generateMode) {
      set.seed(1)
      genusFit <- mclapply(1:3, dmn, count=genusMatT, verbose=TRUE) 
      save(genusFit, 
           file="J:\\Deposit\\Project\\2018_microbiome_atlas\\enterotypeData\\dirichlet.multinomial.genusFit3.2019.0726.RData")
      
      processMode = T
      if (processMode) {
        
        lplc <- sapply(genusFit, laplace)
        plot(lplc, type="b", xlab="Number of Dirichlet Components",ylab="Model Fit")
        best <- genusFit[[which.min(lplc)]]
        
        p0 <- fitted(genusFit[[1]], scale=TRUE) 
        p3 <- fitted(best, scale=TRUE)
        colnames(p3) <- paste("m", 1:3, sep="")
        (meandiff <- colSums(abs(p3 - as.vector(p0))))
        diff <- rowSums(abs(p3 - as.vector(p0)))
        o <- order(diff, decreasing=TRUE)
        cdiff <- cumsum(diff[o]) / sum(diff)
        df <- head(cbind(Mean=p0[o], p3[o,], diff=diff[o], cdiff), 10)
        
        heatmapdmn(genusMatT, genusFit[[1]], best, 10, lblwidth = 4000)
        
        clusterAssigned = apply(genusFit[[3]]@group, 1, function(x) which.max(x))
        
      }
      
      
    }
    
    loadMode = F
    if (loadMode) {
      dmnNewRData ="J:\\Deposit\\Project\\2018_microbiome_atlas\\enterotypeData\\dirichlet.multinomial.genusFit3.2019.0726.RData"
      load(dmnNewRData)
      
      processMode = T
      if (processMode) {
        lplc <- sapply(genusFit, laplace)
        plot(lplc, type="b", xlab="Number of Dirichlet Components",ylab="Model Fit")
        best <- genusFit[[which.min(lplc)]]
        
        p0 <- fitted(genusFit[[1]], scale=TRUE) 
        p3 <- fitted(best, scale=TRUE)
        colnames(p3) <- paste("m", 1:3, sep="")
        (meandiff <- colSums(abs(p3 - as.vector(p0))))
        diff <- rowSums(abs(p3 - as.vector(p0)))
        o <- order(diff, decreasing=TRUE)
        cdiff <- cumsum(diff[o]) / sum(diff)
        df <- head(cbind(Mean=p0[o], p3[o,], diff=diff[o], cdiff), 10)
        #m1 --> firmicutes
        #m2 --> bacteroides
        #m3 --> prevotella
        
        
        heatmapdmn(genusMatT, genusFit[[1]], best, 10, lblwidth = 4000)
        
        clusterAssigned = apply(genusFit[[3]]@group, 1, function(x) which.max(x))
        
        clusterAssignedList = split(names(clusterAssigned), clusterAssigned)
        
        names(clusterAssignedList) = c("ET-Firmicutes","ET-Bacteroides","ET-Prevotella")
        enteroTypeTab = melt(clusterAssignedList)
        colnames(enteroTypeTab) = c("sample.ID","enteroType")
        basicMetaMapUpdated$enteroType = enteroTypeTab$enteroType[match(basicMetaMapUpdated$sample.ID, enteroTypeTab$sample.ID)]
        save(basicMetaMapUpdated, file=basicMetaCleanRDataUpdated2)
        
        View(basicMetaMapUpdated)
        
      }
    }
    
    
  }
  
  mgsMode = T
  if (mgsMode) {
    mgsMatT = round(t(mergeMatUpdated)*10^9)
    
    set.seed(1)
    mgsFit <- mclapply(1:3, dmn, count=mgsMatT, verbose=TRUE) 
    save(mgsFit, 
         file="J:\\Deposit\\Project\\2018_microbiome_atlas\\enterotypeData\\dirichlet.multinomial.mgsFit3.2019.0726.RData")
    processMode = T
    if (processMode) {
      lplc <- sapply(mgsFit, laplace)
      plot(lplc, type="b", xlab="Number of Dirichlet Components",ylab="Model Fit")
      best <- mgsFit[[which.min(lplc)]]
      
      p0 <- fitted(mgsFit[[1]], scale=TRUE) 
      p3 <- fitted(best, scale=TRUE)
      colnames(p3) <- paste("m", 1:3, sep="")
      (meandiff <- colSums(abs(p3 - as.vector(p0))))
      diff <- rowSums(abs(p3 - as.vector(p0)))
      o <- order(diff, decreasing=TRUE)
      cdiff <- cumsum(diff[o]) / sum(diff)
      df <- head(cbind(Mean=p0[o], p3[o,], diff=diff[o], cdiff), 10)
      #m1 --> firmicutes
      #m2 --> bacteroides
      #m3 --> prevotella
      
      
      heatmapdmn(mgsMatT, mgsFit[[1]], best, 10, lblwidth = 4000)
      
      clusterAssigned = apply(mgsFit[[3]]@group, 1, function(x) which.max(x))
      
      clusterAssignedList = split(names(clusterAssigned), clusterAssigned)
      
      names(clusterAssignedList) = c("ET-Bacteroides","ET-Firmicutes","ET-Prevotella")
      enteroTypeTab = melt(clusterAssignedList)
      colnames(enteroTypeTab) = c("sample.ID","enteroType")
      basicMetaMapUpdated$enteroTypeM = enteroTypeTab$enteroType[match(basicMetaMapUpdated$sample.ID, enteroTypeTab$sample.ID)]
      save(basicMetaMapUpdated, file=basicMetaCleanRDataUpdated2)
      
      View(basicMetaMapUpdated)
      
    }
  }
  
}

